# Comparing `tmp/Pgenlib-0.82.1.tar.gz` & `tmp/Pgenlib-0.83.0.tar.gz`

## filetype from file(1)

```diff
@@ -1 +1 @@
-gzip compressed data, was "Pgenlib-0.82.1.tar", last modified: Sat Mar 11 17:46:16 2023, max compression
+gzip compressed data, was "Pgenlib-0.83.0.tar", last modified: Sun Apr 23 17:46:29 2023, max compression
```

## Comparing `Pgenlib-0.82.1.tar` & `Pgenlib-0.83.0.tar`

### file list

```diff
@@ -1,44 +1,48 @@
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.832733 Pgenlib-0.82.1/
--rw-r--r--   0 chrchang   (501) staff       (20)     7652 2022-06-12 20:06:17.000000 Pgenlib-0.82.1/LICENSE
--rw-r--r--   0 chrchang   (501) staff       (20)     1031 2023-03-11 17:42:00.000000 Pgenlib-0.82.1/MANIFEST.in
--rw-r--r--   0 chrchang   (501) staff       (20)     1405 2023-03-11 17:46:16.832599 Pgenlib-0.82.1/PKG-INFO
--rw-r--r--   0 chrchang   (501) staff       (20)      838 2022-08-26 22:35:15.000000 Pgenlib-0.82.1/README.md
--rw-r--r--   0 chrchang   (501) staff       (20)      143 2022-11-26 19:14:19.000000 Pgenlib-0.82.1/pyproject.toml
--rw-r--r--   0 chrchang   (501) staff       (20)       38 2023-03-11 17:46:16.832775 Pgenlib-0.82.1/setup.cfg
--rw-r--r--   0 chrchang   (501) staff       (20)     1845 2023-03-11 17:43:28.000000 Pgenlib-0.82.1/setup.py
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.811851 Pgenlib-0.82.1/src/
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.813827 Pgenlib-0.82.1/src/Pgenlib.egg-info/
--rw-r--r--   0 chrchang   (501) staff       (20)     1405 2023-03-11 17:46:16.000000 Pgenlib-0.82.1/src/Pgenlib.egg-info/PKG-INFO
--rw-r--r--   0 chrchang   (501) staff       (20)     1337 2023-03-11 17:46:16.000000 Pgenlib-0.82.1/src/Pgenlib.egg-info/SOURCES.txt
--rw-r--r--   0 chrchang   (501) staff       (20)        1 2023-03-11 17:46:16.000000 Pgenlib-0.82.1/src/Pgenlib.egg-info/dependency_links.txt
--rw-r--r--   0 chrchang   (501) staff       (20)        8 2023-03-11 17:46:16.000000 Pgenlib-0.82.1/src/Pgenlib.egg-info/top_level.txt
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.819155 Pgenlib-0.82.1/src/pgenlib/
--rw-r--r--   0 chrchang   (501) staff       (20)  1851952 2023-03-11 17:46:16.000000 Pgenlib-0.82.1/src/pgenlib/pgenlib.cpp
--rw-r--r--   0 chrchang   (501) staff       (20)   107901 2023-03-11 17:25:59.000000 Pgenlib-0.82.1/src/pgenlib/pgenlib.pyx
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.820187 Pgenlib-0.82.1/src/plink2/
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.830410 Pgenlib-0.82.1/src/plink2/include/
--rw-r--r--   0 chrchang   (501) staff       (20)    77513 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/hedley.h
--rw-r--r--   0 chrchang   (501) staff       (20)   125638 2022-12-31 18:46:58.000000 Pgenlib-0.82.1/src/plink2/include/pgenlib_misc.cc
--rw-r--r--   0 chrchang   (501) staff       (20)    45386 2023-01-28 22:50:39.000000 Pgenlib-0.82.1/src/plink2/include/pgenlib_misc.h
--rw-r--r--   0 chrchang   (501) staff       (20)   467355 2023-01-28 19:38:22.000000 Pgenlib-0.82.1/src/plink2/include/pgenlib_read.cc
--rw-r--r--   0 chrchang   (501) staff       (20)    36167 2022-12-31 18:46:00.000000 Pgenlib-0.82.1/src/plink2/include/pgenlib_read.h
--rw-r--r--   0 chrchang   (501) staff       (20)   113300 2022-12-31 18:47:06.000000 Pgenlib-0.82.1/src/plink2/include/pgenlib_write.cc
--rw-r--r--   0 chrchang   (501) staff       (20)    19410 2022-12-31 18:46:06.000000 Pgenlib-0.82.1/src/plink2/include/pgenlib_write.h
--rw-r--r--   0 chrchang   (501) staff       (20)    19155 2022-12-31 18:47:11.000000 Pgenlib-0.82.1/src/plink2/include/plink2_base.cc
--rw-r--r--   0 chrchang   (501) staff       (20)   121745 2023-03-02 03:08:31.000000 Pgenlib-0.82.1/src/plink2/include/plink2_base.h
--rw-r--r--   0 chrchang   (501) staff       (20)   120996 2022-12-31 18:47:24.000000 Pgenlib-0.82.1/src/plink2/include/plink2_bits.cc
--rw-r--r--   0 chrchang   (501) staff       (20)    26110 2022-12-31 18:46:24.000000 Pgenlib-0.82.1/src/plink2/include/plink2_bits.h
--rw-r--r--   0 chrchang   (501) staff       (20)    18009 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-align.h
--rw-r--r--   0 chrchang   (501) staff       (20)    17620 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-arch.h
--rw-r--r--   0 chrchang   (501) staff       (20)    41599 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-common.h
--rw-r--r--   0 chrchang   (501) staff       (20)    18585 2022-09-24 13:42:00.000000 Pgenlib-0.82.1/src/plink2/include/simde-constify.h
--rw-r--r--   0 chrchang   (501) staff       (20)     5054 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-detect-clang.h
--rw-r--r--   0 chrchang   (501) staff       (20)    20930 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-diagnostic.h
--rw-r--r--   0 chrchang   (501) staff       (20)    25041 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-features.h
--rw-r--r--   0 chrchang   (501) staff       (20)    58922 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/simde-math.h
-drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-03-11 17:46:16.831926 Pgenlib-0.82.1/src/plink2/include/x86/
--rw-r--r--   0 chrchang   (501) staff       (20)    76719 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/x86/mmx.h
--rw-r--r--   0 chrchang   (501) staff       (20)   156055 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/x86/sse.h
--rw-r--r--   0 chrchang   (501) staff       (20)   255141 2023-02-13 21:29:14.000000 Pgenlib-0.82.1/src/plink2/include/x86/sse2.h
--rw-r--r--   0 chrchang   (501) staff       (20)    23807 2022-01-01 06:48:54.000000 Pgenlib-0.82.1/src/plink2/pgenlib_ffi_support.cc
--rw-r--r--   0 chrchang   (501) staff       (20)     5716 2022-12-31 18:40:33.000000 Pgenlib-0.82.1/src/plink2/pgenlib_ffi_support.h
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.985729 Pgenlib-0.83.0/
+-rw-r--r--   0 chrchang   (501) staff       (20)     7652 2022-06-12 20:06:17.000000 Pgenlib-0.83.0/LICENSE
+-rw-r--r--   0 chrchang   (501) staff       (20)     1106 2023-04-23 17:45:53.000000 Pgenlib-0.83.0/MANIFEST.in
+-rw-r--r--   0 chrchang   (501) staff       (20)     1464 2023-04-23 17:46:29.985594 Pgenlib-0.83.0/PKG-INFO
+-rw-r--r--   0 chrchang   (501) staff       (20)      897 2023-04-23 17:21:35.000000 Pgenlib-0.83.0/README.md
+-rw-r--r--   0 chrchang   (501) staff       (20)      143 2022-11-26 19:14:19.000000 Pgenlib-0.83.0/pyproject.toml
+-rw-r--r--   0 chrchang   (501) staff       (20)       38 2023-04-23 17:46:29.985770 Pgenlib-0.83.0/setup.cfg
+-rw-r--r--   0 chrchang   (501) staff       (20)     1845 2023-04-18 03:20:48.000000 Pgenlib-0.83.0/setup.py
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.967182 Pgenlib-0.83.0/src/
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.970941 Pgenlib-0.83.0/src/Pgenlib.egg-info/
+-rw-r--r--   0 chrchang   (501) staff       (20)     1464 2023-04-23 17:46:29.000000 Pgenlib-0.83.0/src/Pgenlib.egg-info/PKG-INFO
+-rw-r--r--   0 chrchang   (501) staff       (20)     1418 2023-04-23 17:46:29.000000 Pgenlib-0.83.0/src/Pgenlib.egg-info/SOURCES.txt
+-rw-r--r--   0 chrchang   (501) staff       (20)        1 2023-04-23 17:46:29.000000 Pgenlib-0.83.0/src/Pgenlib.egg-info/dependency_links.txt
+-rw-r--r--   0 chrchang   (501) staff       (20)        8 2023-04-23 17:46:29.000000 Pgenlib-0.83.0/src/Pgenlib.egg-info/top_level.txt
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.975208 Pgenlib-0.83.0/src/pgenlib/
+-rw-r--r--   0 chrchang   (501) staff       (20)  2047820 2023-04-23 17:46:29.000000 Pgenlib-0.83.0/src/pgenlib/pgenlib.cpp
+-rw-r--r--   0 chrchang   (501) staff       (20)   125619 2023-04-23 17:00:44.000000 Pgenlib-0.83.0/src/pgenlib/pgenlib.pyx
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.975822 Pgenlib-0.83.0/src/plink2/
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.983713 Pgenlib-0.83.0/src/plink2/include/
+-rw-r--r--   0 chrchang   (501) staff       (20)     9846 2022-09-24 13:42:00.000000 Pgenlib-0.83.0/src/plink2/include/check.h
+-rw-r--r--   0 chrchang   (501) staff       (20)     3119 2022-09-24 13:42:00.000000 Pgenlib-0.83.0/src/plink2/include/debug-trap.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    77513 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/hedley.h
+-rw-r--r--   0 chrchang   (501) staff       (20)   126324 2023-04-18 02:32:01.000000 Pgenlib-0.83.0/src/plink2/include/pgenlib_misc.cc
+-rw-r--r--   0 chrchang   (501) staff       (20)    45478 2023-04-23 16:10:18.000000 Pgenlib-0.83.0/src/plink2/include/pgenlib_misc.h
+-rw-r--r--   0 chrchang   (501) staff       (20)   467494 2023-04-23 15:58:03.000000 Pgenlib-0.83.0/src/plink2/include/pgenlib_read.cc
+-rw-r--r--   0 chrchang   (501) staff       (20)    36167 2022-12-31 18:46:00.000000 Pgenlib-0.83.0/src/plink2/include/pgenlib_read.h
+-rw-r--r--   0 chrchang   (501) staff       (20)   113428 2023-04-23 16:10:07.000000 Pgenlib-0.83.0/src/plink2/include/pgenlib_write.cc
+-rw-r--r--   0 chrchang   (501) staff       (20)    19411 2023-04-17 00:13:15.000000 Pgenlib-0.83.0/src/plink2/include/pgenlib_write.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    19155 2022-12-31 18:47:11.000000 Pgenlib-0.83.0/src/plink2/include/plink2_base.cc
+-rw-r--r--   0 chrchang   (501) staff       (20)   121745 2023-03-02 03:08:31.000000 Pgenlib-0.83.0/src/plink2/include/plink2_base.h
+-rw-r--r--   0 chrchang   (501) staff       (20)   120996 2022-12-31 18:47:24.000000 Pgenlib-0.83.0/src/plink2/include/plink2_bits.cc
+-rw-r--r--   0 chrchang   (501) staff       (20)    26110 2022-12-31 18:46:24.000000 Pgenlib-0.83.0/src/plink2/include/plink2_bits.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    18009 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-align.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    17620 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-arch.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    41599 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-common.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    18585 2022-09-24 13:42:00.000000 Pgenlib-0.83.0/src/plink2/include/simde-constify.h
+-rw-r--r--   0 chrchang   (501) staff       (20)     5054 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-detect-clang.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    20930 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-diagnostic.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    25041 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-features.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    58922 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/simde-math.h
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.984718 Pgenlib-0.83.0/src/plink2/include/x86/
+-rw-r--r--   0 chrchang   (501) staff       (20)    76719 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/x86/mmx.h
+-rw-r--r--   0 chrchang   (501) staff       (20)   156055 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/x86/sse.h
+-rw-r--r--   0 chrchang   (501) staff       (20)   255141 2023-02-13 21:29:14.000000 Pgenlib-0.83.0/src/plink2/include/x86/sse2.h
+-rw-r--r--   0 chrchang   (501) staff       (20)    32721 2023-04-18 04:03:11.000000 Pgenlib-0.83.0/src/plink2/pgenlib_ffi_support.cc
+-rw-r--r--   0 chrchang   (501) staff       (20)     6814 2023-04-18 00:21:46.000000 Pgenlib-0.83.0/src/plink2/pgenlib_ffi_support.h
+drwxr-xr-x   0 chrchang   (501) staff       (20)        0 2023-04-23 17:46:29.985341 Pgenlib-0.83.0/tests/
+-rw-r--r--   0 chrchang   (501) staff       (20)    19169 2023-04-23 17:35:49.000000 Pgenlib-0.83.0/tests/test_pgenlib.py
```

### Comparing `Pgenlib-0.82.1/LICENSE` & `Pgenlib-0.83.0/LICENSE`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/MANIFEST.in` & `Pgenlib-0.83.0/MANIFEST.in`

 * *Files 10% similar despite different names*

```diff
@@ -7,14 +7,16 @@
 include src/plink2/include/pgenlib_read.h
 include src/plink2/include/pgenlib_write.cc
 include src/plink2/include/pgenlib_write.h
 include src/plink2/include/plink2_base.cc
 include src/plink2/include/plink2_base.h
 include src/plink2/include/plink2_bits.cc
 include src/plink2/include/plink2_bits.h
+include src/plink2/include/check.h
+include src/plink2/include/debug-trap.h
 include src/plink2/include/hedley.h
 include src/plink2/include/simde-common.h
 include src/plink2/include/simde-detect-clang.h
 include src/plink2/include/simde-arch.h
 include src/plink2/include/simde-features.h
 include src/plink2/include/simde-diagnostic.h
 include src/plink2/include/simde-math.h
```

### Comparing `Pgenlib-0.82.1/PKG-INFO` & `Pgenlib-0.83.0/PKG-INFO`

 * *Files 14% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: Pgenlib
-Version: 0.82.1
+Version: 0.83.0
 Summary: Python wrapper for pgenlib's basic reader and writer.
 Home-page: https://github.com/chrchang/plink-ng
 Author: Christopher Chang
 Author-email: chrchang@alumni.caltech.edu
 Project-URL: Bug Tracker, https://github.com/chrchang/plink-ng/issues
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)
@@ -13,24 +13,23 @@
 Description-Content-Type: text/markdown
 License-File: LICENSE
 
 This provides a basic Python API for pgenlib  (See [python_api.txt](python_api.txt) for details.)
 
 
 ### Build instructions
-To build this library you will first need to clone the repository:
+To build from GitHub, clone the repository:
 
 ```
 # clone repo
 git clone https://github.com/chrchang/plink-ng
 # go to python folder
-cd plink-ng/2.0/Python 
+cd plink-ng/2.0/Python
 ```
 
-
 Then install Cython and NumPy:
 ```
 pip3 install "cython>=0.29.21" "numpy>=1.19.0"
 ```
 
 and then build and install the package
 ```
@@ -41,13 +40,15 @@
 
 ##### Example usage:
 ```
 #write a 2 sample file
 import numpy as np
 import pgenlib as pg
 
-with pg.PgenWriter("test.pgen".encode("utf-8"), 2, 3, False) as writer:
+with pg.PgenWriter("test.pgen".encode("utf-8"), 2, variant_ct=3, nonref_flags=False) as writer:
 	writer.append_alleles(np.array([0,1,1,1],dtype=np.int32))
 	writer.append_alleles(np.array([0,1,0,0],dtype=np.int32))
 	writer.append_alleles(np.array([0,0,0,0],dtype=np.int32))
 
 ```
+
+See tests/test_pgenlib.py for more sophisticated examples.
```

### Comparing `Pgenlib-0.82.1/README.md` & `Pgenlib-0.83.0/README.md`

 * *Files 24% similar despite different names*

```diff
@@ -1,21 +1,20 @@
 This provides a basic Python API for pgenlib  (See [python_api.txt](python_api.txt) for details.)
 
 
 ### Build instructions
-To build this library you will first need to clone the repository:
+To build from GitHub, clone the repository:
 
 ```
 # clone repo
 git clone https://github.com/chrchang/plink-ng
 # go to python folder
-cd plink-ng/2.0/Python 
+cd plink-ng/2.0/Python
 ```
 
-
 Then install Cython and NumPy:
 ```
 pip3 install "cython>=0.29.21" "numpy>=1.19.0"
 ```
 
 and then build and install the package
 ```
@@ -26,13 +25,15 @@
 
 ##### Example usage:
 ```
 #write a 2 sample file
 import numpy as np
 import pgenlib as pg
 
-with pg.PgenWriter("test.pgen".encode("utf-8"), 2, 3, False) as writer:
+with pg.PgenWriter("test.pgen".encode("utf-8"), 2, variant_ct=3, nonref_flags=False) as writer:
 	writer.append_alleles(np.array([0,1,1,1],dtype=np.int32))
 	writer.append_alleles(np.array([0,1,0,0],dtype=np.int32))
 	writer.append_alleles(np.array([0,0,0,0],dtype=np.int32))
 
 ```
+
+See tests/test_pgenlib.py for more sophisticated examples.
```

### Comparing `Pgenlib-0.82.1/setup.py` & `Pgenlib-0.83.0/setup.py`

 * *Files 5% similar despite different names*

```diff
@@ -21,15 +21,15 @@
     ]
 
 with open("README.md", "r", encoding="utf-8") as fh:
     long_description = fh.read()
 
 setuptools.setup(
     name="Pgenlib",
-    version="0.82.1",
+    version="0.83.0",
     author="Christopher Chang",
     author_email="chrchang@alumni.caltech.edu",
     description="Python wrapper for pgenlib's basic reader and writer.",
     long_description=long_description,
     long_description_content_type="text/markdown",
     url="https://github.com/chrchang/plink-ng",
     project_urls={
```

### Comparing `Pgenlib-0.82.1/src/Pgenlib.egg-info/PKG-INFO` & `Pgenlib-0.83.0/src/Pgenlib.egg-info/PKG-INFO`

 * *Files 14% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: Pgenlib
-Version: 0.82.1
+Version: 0.83.0
 Summary: Python wrapper for pgenlib's basic reader and writer.
 Home-page: https://github.com/chrchang/plink-ng
 Author: Christopher Chang
 Author-email: chrchang@alumni.caltech.edu
 Project-URL: Bug Tracker, https://github.com/chrchang/plink-ng/issues
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)
@@ -13,24 +13,23 @@
 Description-Content-Type: text/markdown
 License-File: LICENSE
 
 This provides a basic Python API for pgenlib  (See [python_api.txt](python_api.txt) for details.)
 
 
 ### Build instructions
-To build this library you will first need to clone the repository:
+To build from GitHub, clone the repository:
 
 ```
 # clone repo
 git clone https://github.com/chrchang/plink-ng
 # go to python folder
-cd plink-ng/2.0/Python 
+cd plink-ng/2.0/Python
 ```
 
-
 Then install Cython and NumPy:
 ```
 pip3 install "cython>=0.29.21" "numpy>=1.19.0"
 ```
 
 and then build and install the package
 ```
@@ -41,13 +40,15 @@
 
 ##### Example usage:
 ```
 #write a 2 sample file
 import numpy as np
 import pgenlib as pg
 
-with pg.PgenWriter("test.pgen".encode("utf-8"), 2, 3, False) as writer:
+with pg.PgenWriter("test.pgen".encode("utf-8"), 2, variant_ct=3, nonref_flags=False) as writer:
 	writer.append_alleles(np.array([0,1,1,1],dtype=np.int32))
 	writer.append_alleles(np.array([0,1,0,0],dtype=np.int32))
 	writer.append_alleles(np.array([0,0,0,0],dtype=np.int32))
 
 ```
+
+See tests/test_pgenlib.py for more sophisticated examples.
```

### Comparing `Pgenlib-0.82.1/src/Pgenlib.egg-info/SOURCES.txt` & `Pgenlib-0.83.0/src/Pgenlib.egg-info/SOURCES.txt`

 * *Files 2% similar despite different names*

```diff
@@ -18,14 +18,16 @@
 src/Pgenlib.egg-info/SOURCES.txt
 src/Pgenlib.egg-info/dependency_links.txt
 src/Pgenlib.egg-info/top_level.txt
 src/pgenlib/pgenlib.cpp
 src/pgenlib/pgenlib.pyx
 src/plink2/pgenlib_ffi_support.cc
 src/plink2/pgenlib_ffi_support.h
+src/plink2/include/check.h
+src/plink2/include/debug-trap.h
 src/plink2/include/hedley.h
 src/plink2/include/pgenlib_misc.cc
 src/plink2/include/pgenlib_misc.h
 src/plink2/include/pgenlib_read.cc
 src/plink2/include/pgenlib_read.h
 src/plink2/include/pgenlib_write.cc
 src/plink2/include/pgenlib_write.h
@@ -39,8 +41,9 @@
 src/plink2/include/simde-constify.h
 src/plink2/include/simde-detect-clang.h
 src/plink2/include/simde-diagnostic.h
 src/plink2/include/simde-features.h
 src/plink2/include/simde-math.h
 src/plink2/include/x86/mmx.h
 src/plink2/include/x86/sse.h
-src/plink2/include/x86/sse2.h
+src/plink2/include/x86/sse2.h
+tests/test_pgenlib.py
```

### Comparing `Pgenlib-0.82.1/src/pgenlib/pgenlib.cpp` & `Pgenlib-0.83.0/src/pgenlib/pgenlib.cpp`

 * *Files 3% similar despite different names*

```diff
@@ -1,33 +1,34 @@
-/* Generated by Cython 0.29.33 */
+/* Generated by Cython 0.29.34 */
 
 /* BEGIN: Cython Metadata
 {
     "distutils": {
         "depends": [
-            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/core/include/numpy/arrayobject.h",
-            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/core/include/numpy/arrayscalars.h",
-            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/core/include/numpy/ndarrayobject.h",
-            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/core/include/numpy/ndarraytypes.h",
-            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/core/include/numpy/ufuncobject.h",
+            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/core/include/numpy/arrayobject.h",
+            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/core/include/numpy/arrayscalars.h",
+            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/core/include/numpy/ndarrayobject.h",
+            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/core/include/numpy/ndarraytypes.h",
+            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/core/include/numpy/ufuncobject.h",
+            "src/plink2/include/pgenlib_misc.h",
             "src/plink2/include/pgenlib_read.h",
             "src/plink2/include/pgenlib_write.h",
             "src/plink2/pgenlib_ffi_support.h"
         ],
         "extra_compile_args": [
             "-std=c++98",
             "-Wno-unused-function",
             "-Wno-macro-redefined"
         ],
         "extra_link_args": [
             "-std=c++98"
         ],
         "include_dirs": [
             "src/pgenlib",
-            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/core/include"
+            "/private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/core/include"
         ],
         "language": "c++",
         "name": "pgenlib",
         "sources": [
             "src/pgenlib/pgenlib.pyx",
             "src/plink2/pgenlib_ffi_support.cc",
             "src/plink2/include/pgenlib_misc.cc",
@@ -46,16 +47,16 @@
 #endif /* PY_SSIZE_T_CLEAN */
 #include "Python.h"
 #ifndef Py_PYTHON_H
     #error Python headers needed to compile C extensions, please install development version of Python.
 #elif PY_VERSION_HEX < 0x02060000 || (0x03000000 <= PY_VERSION_HEX && PY_VERSION_HEX < 0x03030000)
     #error Cython requires Python 2.6+ or Python 3.3+.
 #else
-#define CYTHON_ABI "0_29_33"
-#define CYTHON_HEX_VERSION 0x001D21F0
+#define CYTHON_ABI "0_29_34"
+#define CYTHON_HEX_VERSION 0x001D22F0
 #define CYTHON_FUTURE_DIVISION 1
 #include <stddef.h>
 #ifndef offsetof
   #define offsetof(type, member) ( (size_t) & ((type*)0) -> member )
 #endif
 #if !defined(WIN32) && !defined(MS_WINDOWS)
   #ifndef __stdcall
@@ -240,15 +241,15 @@
   #elif !defined(CYTHON_USE_ASYNC_SLOTS)
     #define CYTHON_USE_ASYNC_SLOTS 1
   #endif
   #if PY_VERSION_HEX < 0x02070000
     #undef CYTHON_USE_PYLONG_INTERNALS
     #define CYTHON_USE_PYLONG_INTERNALS 0
   #elif !defined(CYTHON_USE_PYLONG_INTERNALS)
-    #define CYTHON_USE_PYLONG_INTERNALS 1
+    #define CYTHON_USE_PYLONG_INTERNALS (PY_VERSION_HEX < 0x030C00A5)
   #endif
   #ifndef CYTHON_USE_PYLIST_INTERNALS
     #define CYTHON_USE_PYLIST_INTERNALS 1
   #endif
   #ifndef CYTHON_USE_UNICODE_INTERNALS
     #define CYTHON_USE_UNICODE_INTERNALS 1
   #endif
@@ -279,15 +280,15 @@
   #ifndef CYTHON_PEP489_MULTI_PHASE_INIT
     #define CYTHON_PEP489_MULTI_PHASE_INIT (PY_VERSION_HEX >= 0x03050000)
   #endif
   #ifndef CYTHON_USE_TP_FINALIZE
     #define CYTHON_USE_TP_FINALIZE (PY_VERSION_HEX >= 0x030400a1)
   #endif
   #ifndef CYTHON_USE_DICT_VERSIONS
-    #define CYTHON_USE_DICT_VERSIONS (PY_VERSION_HEX >= 0x030600B1)
+    #define CYTHON_USE_DICT_VERSIONS ((PY_VERSION_HEX >= 0x030600B1) && (PY_VERSION_HEX < 0x030C00A5))
   #endif
   #if PY_VERSION_HEX >= 0x030B00A4
     #undef CYTHON_USE_EXC_INFO_STACK
     #define CYTHON_USE_EXC_INFO_STACK 0
   #elif !defined(CYTHON_USE_EXC_INFO_STACK)
     #define CYTHON_USE_EXC_INFO_STACK (PY_VERSION_HEX >= 0x030700A3)
   #endif
@@ -803,14 +804,15 @@
 #include "numpy/ndarrayobject.h"
 #include "numpy/ndarraytypes.h"
 #include "numpy/arrayscalars.h"
 #include "numpy/ufuncobject.h"
 
     /* NumPy API declarations from "numpy/__init__.pxd" */
     
+#include "../plink2/include/pgenlib_misc.h"
 #include "../plink2/pgenlib_ffi_support.h"
 #include "../plink2/include/pgenlib_read.h"
 #include "../plink2/include/pgenlib_write.h"
 #ifdef _OPENMP
 #include <omp.h>
 #endif /* _OPENMP */
 
@@ -1079,195 +1081,195 @@
   char enc_type;
   char new_packmode;
   char enc_packmode;
   char is_valid_array;
 } __Pyx_BufFmt_Context;
 
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":690
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":690
  * # in Cython to enable them only on the right systems.
  * 
  * ctypedef npy_int8       int8_t             # <<<<<<<<<<<<<<
  * ctypedef npy_int16      int16_t
  * ctypedef npy_int32      int32_t
  */
 typedef npy_int8 __pyx_t_5numpy_int8_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":691
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":691
  * 
  * ctypedef npy_int8       int8_t
  * ctypedef npy_int16      int16_t             # <<<<<<<<<<<<<<
  * ctypedef npy_int32      int32_t
  * ctypedef npy_int64      int64_t
  */
 typedef npy_int16 __pyx_t_5numpy_int16_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":692
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":692
  * ctypedef npy_int8       int8_t
  * ctypedef npy_int16      int16_t
  * ctypedef npy_int32      int32_t             # <<<<<<<<<<<<<<
  * ctypedef npy_int64      int64_t
  * #ctypedef npy_int96      int96_t
  */
 typedef npy_int32 __pyx_t_5numpy_int32_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":693
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":693
  * ctypedef npy_int16      int16_t
  * ctypedef npy_int32      int32_t
  * ctypedef npy_int64      int64_t             # <<<<<<<<<<<<<<
  * #ctypedef npy_int96      int96_t
  * #ctypedef npy_int128     int128_t
  */
 typedef npy_int64 __pyx_t_5numpy_int64_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":697
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":697
  * #ctypedef npy_int128     int128_t
  * 
  * ctypedef npy_uint8      uint8_t             # <<<<<<<<<<<<<<
  * ctypedef npy_uint16     uint16_t
  * ctypedef npy_uint32     uint32_t
  */
 typedef npy_uint8 __pyx_t_5numpy_uint8_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":698
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":698
  * 
  * ctypedef npy_uint8      uint8_t
  * ctypedef npy_uint16     uint16_t             # <<<<<<<<<<<<<<
  * ctypedef npy_uint32     uint32_t
  * ctypedef npy_uint64     uint64_t
  */
 typedef npy_uint16 __pyx_t_5numpy_uint16_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":699
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":699
  * ctypedef npy_uint8      uint8_t
  * ctypedef npy_uint16     uint16_t
  * ctypedef npy_uint32     uint32_t             # <<<<<<<<<<<<<<
  * ctypedef npy_uint64     uint64_t
  * #ctypedef npy_uint96     uint96_t
  */
 typedef npy_uint32 __pyx_t_5numpy_uint32_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":700
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":700
  * ctypedef npy_uint16     uint16_t
  * ctypedef npy_uint32     uint32_t
  * ctypedef npy_uint64     uint64_t             # <<<<<<<<<<<<<<
  * #ctypedef npy_uint96     uint96_t
  * #ctypedef npy_uint128    uint128_t
  */
 typedef npy_uint64 __pyx_t_5numpy_uint64_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":704
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":704
  * #ctypedef npy_uint128    uint128_t
  * 
  * ctypedef npy_float32    float32_t             # <<<<<<<<<<<<<<
  * ctypedef npy_float64    float64_t
  * #ctypedef npy_float80    float80_t
  */
 typedef npy_float32 __pyx_t_5numpy_float32_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":705
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":705
  * 
  * ctypedef npy_float32    float32_t
  * ctypedef npy_float64    float64_t             # <<<<<<<<<<<<<<
  * #ctypedef npy_float80    float80_t
  * #ctypedef npy_float128   float128_t
  */
 typedef npy_float64 __pyx_t_5numpy_float64_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":714
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":714
  * # The int types are mapped a bit surprising --
  * # numpy.int corresponds to 'l' and numpy.long to 'q'
  * ctypedef npy_long       int_t             # <<<<<<<<<<<<<<
  * ctypedef npy_longlong   long_t
  * ctypedef npy_longlong   longlong_t
  */
 typedef npy_long __pyx_t_5numpy_int_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":715
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":715
  * # numpy.int corresponds to 'l' and numpy.long to 'q'
  * ctypedef npy_long       int_t
  * ctypedef npy_longlong   long_t             # <<<<<<<<<<<<<<
  * ctypedef npy_longlong   longlong_t
  * 
  */
 typedef npy_longlong __pyx_t_5numpy_long_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":716
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":716
  * ctypedef npy_long       int_t
  * ctypedef npy_longlong   long_t
  * ctypedef npy_longlong   longlong_t             # <<<<<<<<<<<<<<
  * 
  * ctypedef npy_ulong      uint_t
  */
 typedef npy_longlong __pyx_t_5numpy_longlong_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":718
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":718
  * ctypedef npy_longlong   longlong_t
  * 
  * ctypedef npy_ulong      uint_t             # <<<<<<<<<<<<<<
  * ctypedef npy_ulonglong  ulong_t
  * ctypedef npy_ulonglong  ulonglong_t
  */
 typedef npy_ulong __pyx_t_5numpy_uint_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":719
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":719
  * 
  * ctypedef npy_ulong      uint_t
  * ctypedef npy_ulonglong  ulong_t             # <<<<<<<<<<<<<<
  * ctypedef npy_ulonglong  ulonglong_t
  * 
  */
 typedef npy_ulonglong __pyx_t_5numpy_ulong_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":720
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":720
  * ctypedef npy_ulong      uint_t
  * ctypedef npy_ulonglong  ulong_t
  * ctypedef npy_ulonglong  ulonglong_t             # <<<<<<<<<<<<<<
  * 
  * ctypedef npy_intp       intp_t
  */
 typedef npy_ulonglong __pyx_t_5numpy_ulonglong_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":722
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":722
  * ctypedef npy_ulonglong  ulonglong_t
  * 
  * ctypedef npy_intp       intp_t             # <<<<<<<<<<<<<<
  * ctypedef npy_uintp      uintp_t
  * 
  */
 typedef npy_intp __pyx_t_5numpy_intp_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":723
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":723
  * 
  * ctypedef npy_intp       intp_t
  * ctypedef npy_uintp      uintp_t             # <<<<<<<<<<<<<<
  * 
  * ctypedef npy_double     float_t
  */
 typedef npy_uintp __pyx_t_5numpy_uintp_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":725
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":725
  * ctypedef npy_uintp      uintp_t
  * 
  * ctypedef npy_double     float_t             # <<<<<<<<<<<<<<
  * ctypedef npy_double     double_t
  * ctypedef npy_longdouble longdouble_t
  */
 typedef npy_double __pyx_t_5numpy_float_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":726
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":726
  * 
  * ctypedef npy_double     float_t
  * ctypedef npy_double     double_t             # <<<<<<<<<<<<<<
  * ctypedef npy_longdouble longdouble_t
  * 
  */
 typedef npy_double __pyx_t_5numpy_double_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":727
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":727
  * ctypedef npy_double     float_t
  * ctypedef npy_double     double_t
  * ctypedef npy_longdouble longdouble_t             # <<<<<<<<<<<<<<
  * 
  * ctypedef npy_cfloat      cfloat_t
  */
 typedef npy_longdouble __pyx_t_5numpy_longdouble_t;
@@ -1296,42 +1298,42 @@
 static CYTHON_INLINE __pyx_t_double_complex __pyx_t_double_complex_from_parts(double, double);
 
 
 /*--- Type declarations ---*/
 struct __pyx_obj_7pgenlib_PgenReader;
 struct __pyx_obj_7pgenlib_PgenWriter;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":729
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":729
  * ctypedef npy_longdouble longdouble_t
  * 
  * ctypedef npy_cfloat      cfloat_t             # <<<<<<<<<<<<<<
  * ctypedef npy_cdouble     cdouble_t
  * ctypedef npy_clongdouble clongdouble_t
  */
 typedef npy_cfloat __pyx_t_5numpy_cfloat_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":730
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":730
  * 
  * ctypedef npy_cfloat      cfloat_t
  * ctypedef npy_cdouble     cdouble_t             # <<<<<<<<<<<<<<
  * ctypedef npy_clongdouble clongdouble_t
  * 
  */
 typedef npy_cdouble __pyx_t_5numpy_cdouble_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":731
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":731
  * ctypedef npy_cfloat      cfloat_t
  * ctypedef npy_cdouble     cdouble_t
  * ctypedef npy_clongdouble clongdouble_t             # <<<<<<<<<<<<<<
  * 
  * ctypedef npy_cdouble     complex_t
  */
 typedef npy_clongdouble __pyx_t_5numpy_clongdouble_t;
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":733
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":733
  * ctypedef npy_clongdouble clongdouble_t
  * 
  * ctypedef npy_cdouble     complex_t             # <<<<<<<<<<<<<<
  * 
  * cdef inline object PyArray_MultiIterNew1(a):
  */
 typedef npy_cdouble __pyx_t_5numpy_complex_t;
@@ -1354,381 +1356,411 @@
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_range;
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list_internal32;
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list_internal64;
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list;
 struct __pyx_opt_args_7pgenlib_10PgenReader_count;
 struct __pyx_opt_args_7pgenlib_10PgenReader_change_sample_subset;
 struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles;
+struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased;
 struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch;
+struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch;
 
-/* "pgenlib.pyx":375
+/* "pgenlib.pyx":438
  * 
  * 
  *     cpdef read(self, uint32_t variant_idx, np.ndarray geno_int_out, uint32_t allele_idx = 1):             # <<<<<<<<<<<<<<
- *         # for full genotype info for multiallelic variants, use read_phased()
+ *         # for full genotype info for multiallelic variants, use read_alleles()
  *         # instead
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read {
   int __pyx_n;
   uint32_t allele_idx;
 };
 
-/* "pgenlib.pyx":410
+/* "pgenlib.pyx":473
  * 
  * 
  *     cpdef read_dosages(self, uint32_t variant_idx, np.ndarray floatarr_out, uint32_t allele_idx = 1):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages {
   int __pyx_n;
   uint32_t allele_idx;
 };
 
-/* "pgenlib.pyx":473
+/* "pgenlib.pyx":532
  * 
  * 
  *     cdef read_range_internal8(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         # todo: benchmark effect of adding @cython.boundscheck(False) and
  *         # @cython.wraparound(False) annotations to these multiple-variant-load
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_range_internal8 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":545
+/* "pgenlib.pyx":604
  *         return
  * 
  *     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_range_internal32 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":611
+/* "pgenlib.pyx":670
  *         return
  * 
  *     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_range_internal64 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":677
+/* "pgenlib.pyx":736
  *         return
  * 
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_range {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":691
+/* "pgenlib.pyx":750
  * 
  * 
  *     cdef read_list_internal8(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_list_internal8 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":763
+/* "pgenlib.pyx":822
  *         return
  * 
  *     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_list_internal32 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":835
+/* "pgenlib.pyx":894
  *         return
  * 
  *     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_list_internal64 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":907
+/* "pgenlib.pyx":966
  *         return
  * 
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if geno_int_out.dtype == np.int8:
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_list {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":919
+/* "pgenlib.pyx":978
  * 
  * 
  *     cpdef read_alleles_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_range {
   int __pyx_n;
   int hap_maj;
 };
 
-/* "pgenlib.pyx":1019
+/* "pgenlib.pyx":1081
  * 
  * 
  *     cpdef read_alleles_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_list {
   int __pyx_n;
   int hap_maj;
 };
 
-/* "pgenlib.pyx":1120
+/* "pgenlib.pyx":1185
  * 
  * 
  *     cpdef read_alleles_and_phasepresent_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_and_phasepresent_range {
   int __pyx_n;
   int hap_maj;
 };
 
-/* "pgenlib.pyx":1161
+/* "pgenlib.pyx":1221
  * 
  * 
  *     cpdef read_alleles_and_phasepresent_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,cast=True,mode="c",ndim=2] phasepresent_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_and_phasepresent_list {
   int __pyx_n;
   int hap_maj;
 };
 
-/* "pgenlib.pyx":1205
+/* "pgenlib.pyx":1263
  * 
  * 
  *     cdef read_dosages_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_range_internal32 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":1229
+/* "pgenlib.pyx":1287
  * 
  * 
  *     cdef read_dosages_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_range_internal64 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":1253
+/* "pgenlib.pyx":1311
  * 
  * 
  *     cpdef read_dosages_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_range {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":1265
+/* "pgenlib.pyx":1323
  * 
  * 
  *     cdef read_dosages_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list_internal32 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":1298
+/* "pgenlib.pyx":1356
  * 
  * 
  *     cdef read_dosages_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list_internal64 {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":1331
+/* "pgenlib.pyx":1389
  * 
  * 
  *     cpdef read_dosages_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list {
   int __pyx_n;
   uint32_t allele_idx;
   int sample_maj;
 };
 
-/* "pgenlib.pyx":1340
+/* "pgenlib.pyx":1398
  *         return
  * 
  *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):             # <<<<<<<<<<<<<<
- *         # todo: multiallelic variants
- *         if allele_idx is None:
+ *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_count {
   int __pyx_n;
   PyObject *allele_idx;
 };
 
-/* "pgenlib.pyx":1356
+/* "pgenlib.pyx":1464
  * 
  * 
  *     cpdef change_sample_subset(self, object sample_subset = None):             # <<<<<<<<<<<<<<
  *         if sample_subset is not None:
  *             self.set_sample_subset_internal(sample_subset)
  */
 struct __pyx_opt_args_7pgenlib_10PgenReader_change_sample_subset {
   int __pyx_n;
   PyObject *sample_subset;
 };
 
-/* "pgenlib.pyx":1494
+/* "pgenlib.pyx":1650
  * 
  * 
- *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False):             # <<<<<<<<<<<<<<
+ *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False, object allele_ct = None):             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
 struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles {
   int __pyx_n;
   int all_phased;
+  PyObject *allele_ct;
+};
+
+/* "pgenlib.pyx":1694
+ * 
+ * 
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):             # <<<<<<<<<<<<<<
+ *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
+ *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
+ */
+struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased {
+  int __pyx_n;
+  PyObject *allele_ct;
 };
 
-/* "pgenlib.pyx":1574
+/* "pgenlib.pyx":1780
  * 
  * 
- *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False):             # <<<<<<<<<<<<<<
+ *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False, object allele_cts = None):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
 struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch {
   int __pyx_n;
   int all_phased;
+  PyObject *allele_cts;
 };
 
-/* "pgenlib.pyx":188
+/* "pgenlib.pyx":1846
+ * 
+ * 
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):             # <<<<<<<<<<<<<<
+ *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
+ *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
+ */
+struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch {
+  int __pyx_n;
+  PyObject *allele_cts;
+};
+
+/* "pgenlib.pyx":227
  * 
  * 
  * cdef class PgenReader:             # <<<<<<<<<<<<<<
- *     # todo: nonref_flags, multiallelic variant support
  *     cdef PgenFileInfo* _info_ptr
+ *     cdef PgenReaderStruct* _state_ptr
  */
 struct __pyx_obj_7pgenlib_PgenReader {
   PyObject_HEAD
   struct __pyx_vtabstruct_7pgenlib_PgenReader *__pyx_vtab;
   plink2::PgenFileInfo *_info_ptr;
   struct plink2::PgenReaderStruct *_state_ptr;
   uintptr_t *_subset_include_vec;
   uintptr_t *_subset_include_interleaved_vec;
   uint32_t *_subset_cumulative_popcounts;
   struct plink2::PgrSampleSubsetIndexStruct _subset_index;
   uint32_t _subset_size;
-  uintptr_t *_genovec;
-  uintptr_t *_phasepresent;
-  uintptr_t *_phaseinfo;
-  uintptr_t *_dosage_present;
-  uint16_t *_dosage_main;
+  struct plink2::PgenVariantStruct _pgv;
   plink2::VecW *_transpose_batch_buf;
   uintptr_t *_multivar_vmaj_geno_buf;
   uintptr_t *_multivar_vmaj_phasepresent_buf;
   uintptr_t *_multivar_vmaj_phaseinfo_buf;
   uintptr_t *_multivar_smaj_geno_batch_buf;
   uintptr_t *_multivar_smaj_phaseinfo_batch_buf;
   uintptr_t *_multivar_smaj_phasepresent_batch_buf;
 };
 
 
-/* "pgenlib.pyx":1412
- *     BytesToBitsUnsafe(boolbytes, sample_ct, bitarr)
+/* "pgenlib.pyx":1524
+ * 
  * 
  * cdef class PgenWriter:             # <<<<<<<<<<<<<<
  *     cdef STPgenWriter* _state_ptr
  *     cdef uintptr_t* _nonref_flags
  */
 struct __pyx_obj_7pgenlib_PgenWriter {
   PyObject_HEAD
   struct __pyx_vtabstruct_7pgenlib_PgenWriter *__pyx_vtab;
   plink2::STPgenWriter *_state_ptr;
   uintptr_t *_nonref_flags;
   plink2::PgenGlobalFlags _phase_dosage_gflags;
+  uint32_t _allele_ct_limit;
   uintptr_t *_genovec;
+  uintptr_t *_patch_01_set;
+  plink2::AlleleCode *_patch_01_vals;
+  uintptr_t *_patch_10_set;
+  plink2::AlleleCode *_patch_10_vals;
   uintptr_t *_phasepresent;
   uintptr_t *_phaseinfo;
   uintptr_t *_dosage_present;
   uint16_t *_dosage_main;
 };
 
 
 
-/* "pgenlib.pyx":188
+/* "pgenlib.pyx":227
  * 
  * 
  * cdef class PgenReader:             # <<<<<<<<<<<<<<
- *     # todo: nonref_flags, multiallelic variant support
  *     cdef PgenFileInfo* _info_ptr
+ *     cdef PgenReaderStruct* _state_ptr
  */
 
 struct __pyx_vtabstruct_7pgenlib_PgenReader {
+  PyObject *(*set_allele_idx_offsets_internal)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *);
   PyObject *(*set_sample_subset_internal)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *);
   PyObject *(*__pyx___enter__)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch);
   PyObject *(*get_raw_sample_ct)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch);
   PyObject *(*get_variant_ct)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch);
   PyObject *(*hardcall_phase_present)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch);
   PyObject *(*read)(struct __pyx_obj_7pgenlib_PgenReader *, uint32_t, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read *__pyx_optional_args);
   PyObject *(*read_dosages)(struct __pyx_obj_7pgenlib_PgenReader *, uint32_t, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages *__pyx_optional_args);
@@ -1756,33 +1788,33 @@
   PyObject *(*change_sample_subset)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_change_sample_subset *__pyx_optional_args);
   PyObject *(*close)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch);
   PyObject *(*__pyx___exit__)(struct __pyx_obj_7pgenlib_PgenReader *, PyObject *, PyObject *, PyObject *, int __pyx_skip_dispatch);
 };
 static struct __pyx_vtabstruct_7pgenlib_PgenReader *__pyx_vtabptr_7pgenlib_PgenReader;
 
 
-/* "pgenlib.pyx":1412
- *     BytesToBitsUnsafe(boolbytes, sample_ct, bitarr)
+/* "pgenlib.pyx":1524
+ * 
  * 
  * cdef class PgenWriter:             # <<<<<<<<<<<<<<
  *     cdef STPgenWriter* _state_ptr
  *     cdef uintptr_t* _nonref_flags
  */
 
 struct __pyx_vtabstruct_7pgenlib_PgenWriter {
   PyObject *(*__pyx___enter__)(struct __pyx_obj_7pgenlib_PgenWriter *, int __pyx_skip_dispatch);
   PyObject *(*append_biallelic)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch);
   PyObject *(*append_alleles)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles *__pyx_optional_args);
-  PyObject *(*append_partially_phased)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch);
+  PyObject *(*append_partially_phased)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased *__pyx_optional_args);
   PyObject *(*append_dosages_internal32)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *);
   PyObject *(*append_dosages_internal64)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *);
   PyObject *(*append_dosages)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch);
   PyObject *(*append_biallelic_batch)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch);
   PyObject *(*append_alleles_batch)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch *__pyx_optional_args);
-  PyObject *(*append_partially_phased_batch)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch);
+  PyObject *(*append_partially_phased_batch)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch *__pyx_optional_args);
   PyObject *(*append_dosages_batch_internal32)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *);
   PyObject *(*append_dosages_batch_internal64)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *);
   PyObject *(*append_dosages_batch)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch);
   PyObject *(*close)(struct __pyx_obj_7pgenlib_PgenWriter *, int __pyx_skip_dispatch);
   PyObject *(*__pyx___exit__)(struct __pyx_obj_7pgenlib_PgenWriter *, PyObject *, PyObject *, PyObject *, int __pyx_skip_dispatch);
 };
 static struct __pyx_vtabstruct_7pgenlib_PgenWriter *__pyx_vtabptr_7pgenlib_PgenWriter;
@@ -1878,21 +1910,18 @@
 static int  __Pyx__GetBufferAndValidate(Py_buffer* buf, PyObject* obj,
     __Pyx_TypeInfo* dtype, int flags, int nd, int cast, __Pyx_BufFmt_StackElem* stack);
 static void __Pyx_ZeroBuffer(Py_buffer* buf);
 static CYTHON_INLINE void __Pyx_SafeReleaseBuffer(Py_buffer* info);
 static Py_ssize_t __Pyx_minusones[] = { -1, -1, -1, -1, -1, -1, -1, -1 };
 static Py_ssize_t __Pyx_zeros[] = { 0, 0, 0, 0, 0, 0, 0, 0 };
 
-/* PyObjectCall.proto */
-#if CYTHON_COMPILING_IN_CPYTHON
-static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw);
-#else
-#define __Pyx_PyObject_Call(func, arg, kw) PyObject_Call(func, arg, kw)
-#endif
+/* BufferIndexError.proto */
+static void __Pyx_RaiseBufferIndexError(int axis);
 
+#define __Pyx_BufPtrCContig1d(type, buf, i0, s0) ((type)buf + i0)
 /* PyThreadStateGet.proto */
 #if CYTHON_FAST_THREAD_STATE
 #define __Pyx_PyThreadState_declare  PyThreadState *__pyx_tstate;
 #define __Pyx_PyThreadState_assign  __pyx_tstate = __Pyx_PyThreadState_Current;
 #define __Pyx_PyErr_Occurred()  __pyx_tstate->curexc_type
 #else
 #define __Pyx_PyThreadState_declare
@@ -1921,21 +1950,24 @@
 #define __Pyx_ErrFetchWithState(type, value, tb)  PyErr_Fetch(type, value, tb)
 #define __Pyx_ErrRestoreInState(tstate, type, value, tb)  PyErr_Restore(type, value, tb)
 #define __Pyx_ErrFetchInState(tstate, type, value, tb)  PyErr_Fetch(type, value, tb)
 #define __Pyx_ErrRestore(type, value, tb)  PyErr_Restore(type, value, tb)
 #define __Pyx_ErrFetch(type, value, tb)  PyErr_Fetch(type, value, tb)
 #endif
 
+/* PyObjectCall.proto */
+#if CYTHON_COMPILING_IN_CPYTHON
+static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw);
+#else
+#define __Pyx_PyObject_Call(func, arg, kw) PyObject_Call(func, arg, kw)
+#endif
+
 /* RaiseException.proto */
 static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb, PyObject *cause);
 
-/* BufferIndexError.proto */
-static void __Pyx_RaiseBufferIndexError(int axis);
-
-#define __Pyx_BufPtrCContig1d(type, buf, i0, s0) ((type)buf + i0)
 /* PyCFunctionFastCall.proto */
 #if CYTHON_FAST_PYCCALL
 static CYTHON_INLINE PyObject *__Pyx_PyCFunction_FastCall(PyObject *func, PyObject **args, Py_ssize_t nargs);
 #else
 #define __Pyx_PyCFunction_FastCall(func, args, nargs)  (assert(0), NULL)
 #endif
 
@@ -1992,20 +2024,20 @@
 
 /* ArgTypeTest.proto */
 #define __Pyx_ArgTypeTest(obj, type, none_allowed, name, exact)\
     ((likely((Py_TYPE(obj) == type) | (none_allowed && (obj == Py_None)))) ? 1 :\
         __Pyx__ArgTypeTest(obj, type, name, exact))
 static int __Pyx__ArgTypeTest(PyObject *obj, PyTypeObject *type, const char *name, int exact);
 
-/* DivInt[long].proto */
-static CYTHON_INLINE long __Pyx_div_long(long, long);
-
 /* ExtTypeTest.proto */
 static CYTHON_INLINE int __Pyx_TypeTest(PyObject *obj, PyTypeObject *type);
 
+/* DivInt[long].proto */
+static CYTHON_INLINE long __Pyx_div_long(long, long);
+
 /* PyDictVersioning.proto */
 #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_TYPE_SLOTS
 #define __PYX_DICT_VERSION_INIT  ((PY_UINT64_T) -1)
 #define __PYX_GET_DICT_VERSION(dict)  (((PyDictObject*)(dict))->ma_version_tag)
 #define __PYX_UPDATE_DICT_CACHE(dict, value, cache_var, version_var)\
     (version_var) = __PYX_GET_DICT_VERSION(dict);\
     (cache_var) = (value);
@@ -2057,20 +2089,32 @@
 #endif
 
 #define __Pyx_BufPtrCContig2d(type, buf, i0, s0, i1, s1) ((type)((char*)buf + i0 * s0) + i1)
 /* ModInt[long].proto */
 static CYTHON_INLINE long __Pyx_mod_long(long, long);
 
 #define __Pyx_BufPtrStrided1d(type, buf, i0, s0) (type)((char*)buf + i0 * s0)
+/* PyIntBinop.proto */
+#if !CYTHON_COMPILING_IN_PYPY
+static PyObject* __Pyx_PyInt_AddObjC(PyObject *op1, PyObject *op2, long intval, int inplace, int zerodivision_check);
+#else
+#define __Pyx_PyInt_AddObjC(op1, op2, intval, inplace, zerodivision_check)\
+    (inplace ? PyNumber_InPlaceAdd(op1, op2) : PyNumber_Add(op1, op2))
+#endif
+
 /* PyIntCompare.proto */
 static CYTHON_INLINE PyObject* __Pyx_PyInt_NeObjC(PyObject *op1, PyObject *op2, long intval, long inplace);
 
 /* PyObjectCall2Args.proto */
 static CYTHON_UNUSED PyObject* __Pyx_PyObject_Call2Args(PyObject* function, PyObject* arg1, PyObject* arg2);
 
+/* UnaryNegOverflows.proto */
+#define UNARY_NEG_WOULD_OVERFLOW(x)\
+        (((x) < 0) & ((unsigned long)(x) == 0-(unsigned long)(x)))
+
 /* GetItemInt.proto */
 #define __Pyx_GetItemInt(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck)\
     (__Pyx_fits_Py_ssize_t(i, type, is_signed) ?\
     __Pyx_GetItemInt_Fast(o, (Py_ssize_t)i, is_list, wraparound, boundscheck) :\
     (is_list ? (PyErr_SetString(PyExc_IndexError, "list index out of range"), (PyObject*)NULL) :\
                __Pyx_GetItemInt_Generic(o, to_py_func(i))))
 #define __Pyx_GetItemInt_List(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck)\
@@ -2085,20 +2129,18 @@
     (PyErr_SetString(PyExc_IndexError, "tuple index out of range"), (PyObject*)NULL))
 static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Tuple_Fast(PyObject *o, Py_ssize_t i,
                                                               int wraparound, int boundscheck);
 static PyObject *__Pyx_GetItemInt_Generic(PyObject *o, PyObject* j);
 static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
                                                      int is_list, int wraparound, int boundscheck);
 
-/* ObjectGetItem.proto */
-#if CYTHON_USE_TYPE_SLOTS
-static CYTHON_INLINE PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* key);
-#else
-#define __Pyx_PyObject_GetItem(obj, key)  PyObject_GetItem(obj, key)
-#endif
+/* WriteUnraisableException.proto */
+static void __Pyx_WriteUnraisable(const char *name, int clineno,
+                                  int lineno, const char *filename,
+                                  int full_traceback, int nogil);
 
 /* GetTopmostException.proto */
 #if CYTHON_USE_EXC_INFO_STACK
 static _PyErr_StackItem * __Pyx_PyErr_GetTopmostException(PyThreadState *tstate);
 #endif
 
 /* SaveResetException.proto */
@@ -2150,20 +2192,28 @@
 
 /* SetupReduce.proto */
 static int __Pyx_setup_reduce(PyObject* type_obj);
 
 /* TypeImport.proto */
 #ifndef __PYX_HAVE_RT_ImportType_proto
 #define __PYX_HAVE_RT_ImportType_proto
+#if __STDC_VERSION__ >= 201112L
+#include <stdalign.h>
+#endif
+#if __STDC_VERSION__ >= 201112L || __cplusplus >= 201103L
+#define __PYX_GET_STRUCT_ALIGNMENT(s) alignof(s)
+#else
+#define __PYX_GET_STRUCT_ALIGNMENT(s) sizeof(void*)
+#endif
 enum __Pyx_ImportType_CheckSize {
    __Pyx_ImportType_CheckSize_Error = 0,
    __Pyx_ImportType_CheckSize_Warn = 1,
    __Pyx_ImportType_CheckSize_Ignore = 2
 };
-static PyTypeObject *__Pyx_ImportType(PyObject* module, const char *module_name, const char *class_name, size_t size, enum __Pyx_ImportType_CheckSize check_size);
+static PyTypeObject *__Pyx_ImportType(PyObject* module, const char *module_name, const char *class_name, size_t size, size_t alignment, enum __Pyx_ImportType_CheckSize check_size);
 #endif
 
 /* Import.proto */
 static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level);
 
 /* CLineInTraceback.proto */
 #ifdef CYTHON_CLINE_IN_TRACEBACK
@@ -2331,20 +2381,20 @@
 
 /* CIntToPy.proto */
 static CYTHON_INLINE PyObject* __Pyx_PyInt_From_plink2_3a__3a_PglErr(plink2::PglErr value);
 
 /* CIntToPy.proto */
 static CYTHON_INLINE PyObject* __Pyx_PyInt_From_Py_intptr_t(Py_intptr_t value);
 
-/* CIntToPy.proto */
-static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value);
-
 /* CIntFromPy.proto */
 static CYTHON_INLINE size_t __Pyx_PyInt_As_size_t(PyObject *);
 
+/* CIntToPy.proto */
+static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value);
+
 /* CIntFromPy.proto */
 static CYTHON_INLINE long __Pyx_PyInt_As_long(PyObject *);
 
 /* CIntFromPy.proto */
 static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *);
 
 /* FastTypeChecks.proto */
@@ -2362,14 +2412,15 @@
 
 /* CheckBinaryVersion.proto */
 static int __Pyx_check_binary_version(void);
 
 /* InitStrings.proto */
 static int __Pyx_InitStrings(__Pyx_StringTabEntry *t);
 
+static PyObject *__pyx_f_7pgenlib_10PgenReader_set_allele_idx_offsets_internal(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyArrayObject *__pyx_v_allele_idx_offsets); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_set_sample_subset_internal(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyArrayObject *__pyx_v_sample_subset); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader___enter__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_get_raw_sample_ct(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_get_variant_ct(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_hardcall_phase_present(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_geno_int_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read *__pyx_optional_args); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_dosages(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_floatarr_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages *__pyx_optional_args); /* proto*/
@@ -2396,35 +2447,35 @@
 static PyObject *__pyx_f_7pgenlib_10PgenReader_count(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_genocount_uint32_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_count *__pyx_optional_args); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_change_sample_subset(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_change_sample_subset *__pyx_optional_args); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_close(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader___exit__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, CYTHON_UNUSED PyObject *__pyx_v_exc_type, CYTHON_UNUSED PyObject *__pyx_v_exc_val, CYTHON_UNUSED PyObject *__pyx_v_exc_tb, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter___enter__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_biallelic(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_geno_int8, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_alleles(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles *__pyx_optional_args); /* proto*/
-static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent, int __pyx_skip_dispatch); /* proto*/
+static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased *__pyx_optional_args); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_dosages_internal32(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_floatarr); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_dosages_internal64(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_doublearr); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_dosages(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_floatarr, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_biallelic_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_geno_int8_batch, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_alleles_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch *__pyx_optional_args); /* proto*/
-static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch, int __pyx_skip_dispatch); /* proto*/
+static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch *__pyx_optional_args); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_dosages_batch_internal32(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_floatarr_batch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_dosages_batch_internal64(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_doublearr_batch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_dosages_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_floatarr_batch, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_close(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter___exit__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, CYTHON_UNUSED PyObject *__pyx_v_exc_type, CYTHON_UNUSED PyObject *__pyx_v_exc_val, CYTHON_UNUSED PyObject *__pyx_v_exc_tb, int __pyx_skip_dispatch); /* proto*/
 
 /* Module declarations from 'libc.stdint' */
 
+/* Module declarations from 'libc.string' */
+
 /* Module declarations from 'cpython.mem' */
 
 /* Module declarations from 'cpython.buffer' */
 
-/* Module declarations from 'libc.string' */
-
 /* Module declarations from 'libc.stdio' */
 
 /* Module declarations from '__builtin__' */
 
 /* Module declarations from 'cpython.type' */
 static PyTypeObject *__pyx_ptype_7cpython_4type_type = 0;
 
@@ -2453,28 +2504,29 @@
 static PyTypeObject *__pyx_ptype_5numpy_character = 0;
 static PyTypeObject *__pyx_ptype_5numpy_ufunc = 0;
 
 /* Module declarations from 'pgenlib' */
 static PyTypeObject *__pyx_ptype_7pgenlib_PgenReader = 0;
 static PyTypeObject *__pyx_ptype_7pgenlib_PgenWriter = 0;
 static PyObject *__pyx_f_7pgenlib_bytes_to_bits_internal(PyArrayObject *, uint32_t, uintptr_t *); /*proto*/
+static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_uintp_t = { "uintp_t", NULL, sizeof(__pyx_t_5numpy_uintp_t), { 0 }, 0, IS_UNSIGNED(__pyx_t_5numpy_uintp_t) ? 'U' : 'I', IS_UNSIGNED(__pyx_t_5numpy_uintp_t), 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t = { "uint32_t", NULL, sizeof(__pyx_t_5numpy_uint32_t), { 0 }, 0, IS_UNSIGNED(__pyx_t_5numpy_uint32_t) ? 'U' : 'I', IS_UNSIGNED(__pyx_t_5numpy_uint32_t), 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t = { "int32_t", NULL, sizeof(__pyx_t_5numpy_int32_t), { 0 }, 0, IS_UNSIGNED(__pyx_t_5numpy_int32_t) ? 'U' : 'I', IS_UNSIGNED(__pyx_t_5numpy_int32_t), 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t = { "uint8_t", NULL, sizeof(__pyx_t_5numpy_uint8_t), { 0 }, 0, IS_UNSIGNED(__pyx_t_5numpy_uint8_t) ? 'U' : 'I', IS_UNSIGNED(__pyx_t_5numpy_uint8_t), 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t = { "int8_t", NULL, sizeof(__pyx_t_5numpy_int8_t), { 0 }, 0, IS_UNSIGNED(__pyx_t_5numpy_int8_t) ? 'U' : 'I', IS_UNSIGNED(__pyx_t_5numpy_int8_t), 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_int64_t = { "int64_t", NULL, sizeof(__pyx_t_5numpy_int64_t), { 0 }, 0, IS_UNSIGNED(__pyx_t_5numpy_int64_t) ? 'U' : 'I', IS_UNSIGNED(__pyx_t_5numpy_int64_t), 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t = { "float32_t", NULL, sizeof(__pyx_t_5numpy_float32_t), { 0 }, 0, 'R', 0, 0 };
 static __Pyx_TypeInfo __Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t = { "float64_t", NULL, sizeof(__pyx_t_5numpy_float64_t), { 0 }, 0, 'R', 0, 0 };
 #define __Pyx_MODULE_NAME "pgenlib"
 extern int __pyx_module_is_main_pgenlib;
 int __pyx_module_is_main_pgenlib = 0;
 
 /* Implementation of 'pgenlib' */
-static PyObject *__pyx_builtin_RuntimeError;
 static PyObject *__pyx_builtin_MemoryError;
+static PyObject *__pyx_builtin_RuntimeError;
 static PyObject *__pyx_builtin_range;
 static PyObject *__pyx_builtin_TypeError;
 static PyObject *__pyx_builtin_ImportError;
 static const char __pyx_k__5[] = ").";
 static const char __pyx_k__9[] = ")";
 static const char __pyx_k_np[] = "np";
 static const char __pyx_k__10[] = ", ";
@@ -2504,21 +2556,23 @@
 static const char __pyx_k_hap_maj[] = "hap_maj";
 static const char __pyx_k_in_file[] = " in file).";
 static const char __pyx_k_exc_type[] = "exc_type";
 static const char __pyx_k_filename[] = "filename";
 static const char __pyx_k_getstate[] = "__getstate__";
 static const char __pyx_k_setstate[] = "__setstate__";
 static const char __pyx_k_TypeError[] = "TypeError";
+static const char __pyx_k_allele_ct[] = "allele_ct";
 static const char __pyx_k_in_file_2[] = " in file)";
 static const char __pyx_k_read_list[] = "read_list";
 static const char __pyx_k_reduce_ex[] = "__reduce_ex__";
 static const char __pyx_k_sample_ct[] = "sample_ct";
 static const char __pyx_k_PgenReader[] = "PgenReader";
 static const char __pyx_k_PgenWriter[] = "PgenWriter";
 static const char __pyx_k_all_phased[] = "all_phased";
+static const char __pyx_k_allele_cts[] = "allele_cts";
 static const char __pyx_k_allele_idx[] = "allele_idx";
 static const char __pyx_k_pyx_vtable[] = "__pyx_vtable__";
 static const char __pyx_k_read_error[] = "read() error ";
 static const char __pyx_k_read_range[] = "read_range";
 static const char __pyx_k_sample_maj[] = "sample_maj";
 static const char __pyx_k_variant_ct[] = "variant_ct";
 static const char __pyx_k_ImportError[] = "ImportError";
@@ -2540,21 +2594,23 @@
 static const char __pyx_k_reduce_cython[] = "__reduce_cython__";
 static const char __pyx_k_sample_subset[] = "sample_subset";
 static const char __pyx_k_variant_idx_2[] = "variant_idx ";
 static const char __pyx_k_append_alleles[] = "append_alleles";
 static const char __pyx_k_append_dosages[] = "append_dosages";
 static const char __pyx_k_dosage_present[] = "dosage_present";
 static const char __pyx_k_get_variant_ct[] = "get_variant_ct";
+static const char __pyx_k_allele_ct_limit[] = "allele_ct_limit";
 static const char __pyx_k_read_list_error[] = "read_list() error ";
 static const char __pyx_k_setstate_cython[] = "__setstate_cython__";
 static const char __pyx_k_variant_idx_end[] = "variant_idx_end";
 static const char __pyx_k_allele_int32_out[] = "allele_int32_out";
 static const char __pyx_k_append_biallelic[] = "append_biallelic";
 static const char __pyx_k_phasepresent_out[] = "phasepresent_out";
 static const char __pyx_k_read_range_error[] = "read_range() error ";
+static const char __pyx_k_variant_ct_limit[] = "variant_ct_limit";
 static const char __pyx_k_get_raw_sample_ct[] = "get_raw_sample_ct";
 static const char __pyx_k_read_alleles_list[] = "read_alleles_list";
 static const char __pyx_k_read_dosages_list[] = "read_dosages_list";
 static const char __pyx_k_variant_idx_start[] = "variant_idx_start";
 static const char __pyx_k_allele_idx_offsets[] = "allele_idx_offsets";
 static const char __pyx_k_allele_int32_batch[] = "allele_int32_batch";
 static const char __pyx_k_cline_in_traceback[] = "cline_in_traceback";
@@ -2594,42 +2650,48 @@
 static const char __pyx_k_Sample_major_read_list_geno_int[] = "Sample-major read_list() geno_int_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_alleles_list[] = "Variant-major read_alleles_list() allele_int32_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_dosages_list[] = "Variant-major read_dosages_list() floatarr_out buffer has too few rows (";
 static const char __pyx_k_and_column_count_should_be_twic[] = ", , and column count should be twice that).";
 static const char __pyx_k_and_row_count_should_be_twice_t[] = ", and row count should be twice that)";
 static const char __pyx_k_append_dosages_cannot_be_called[] = "append_dosages cannot be called when PgenWriter was constructed with dosage_present False";
 static const char __pyx_k_numpy_core_multiarray_failed_to[] = "numpy.core.multiarray failed to import";
+static const char __pyx_k_read_alleles_and_phasepresent_r[] = " read_alleles_and_phasepresent_range() error ";
 static const char __pyx_k_read_alleles_list_variant_index[] = "read_alleles_list() variant index too large (";
 static const char __pyx_k_read_dosages_list_variant_index[] = "read_dosages_list() variant index too large (";
 static const char __pyx_k_unequal_to_initially_declared_v[] = ") unequal to initially declared value (";
 static const char __pyx_k_variant_idx_end_variant_idx_sta[] = "; (variant_idx_end - variant_idx_start) is ";
 static const char __pyx_k_Empty_sample_subset_is_not_curre[] = "Empty sample_subset is not currently permitted.";
 static const char __pyx_k_Haplotype_major_read_alleles_lis[] = "Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (";
 static const char __pyx_k_Haplotype_major_read_alleles_ran[] = "Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (";
 static const char __pyx_k_Invalid_append_dosages_batch_dos[] = "Invalid append_dosages_batch() dosage array element type (float32 or float64 expected).";
 static const char __pyx_k_Invalid_append_dosages_dosage_ar[] = "Invalid append_dosages() dosage array element type (float32 or float64 expected).";
-static const char __pyx_k_Invalid_arguments_for_PgenWriter[] = "Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).";
+static const char __pyx_k_Invalid_arguments_for_PgenWriter[] = "Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).";
 static const char __pyx_k_Invalid_read_dosages_floatarr_ou[] = "Invalid read_dosages() floatarr_out array element type (float32 or float64 expected).";
 static const char __pyx_k_Invalid_read_dosages_list_floata[] = "Invalid read_dosages_list() floatarr_out array element type (float32 or float64 expected).";
 static const char __pyx_k_Invalid_read_dosages_range_float[] = "Invalid read_dosages_range() floatarr_out array element type (float32 or float64 expected).";
 static const char __pyx_k_Invalid_read_list_geno_int_out_a[] = "Invalid read_list() geno_int_out array element type (int8, int32, or int64 expected).";
-static const char __pyx_k_Multiallelic_phase_dosage_datase[] = "Multiallelic + phase/dosage datasets not supported yet";
-static const char __pyx_k_Multiallelic_variants_aren_t_sup[] = "Multiallelic variants aren't supported by PgenWriter yet.";
+static const char __pyx_k_PgenReader_multiallelic_variants[] = "PgenReader: multiallelic variants present, but allele_idx_offsets not provided";
+static const char __pyx_k_PgenWriter___dealloc___SpgwFinis[] = "PgenWriter.__dealloc__(): SpgwFinish() error ";
+static const char __pyx_k_PgenWriter_close_SpgwFinish_erro[] = "PgenWriter.close(): SpgwFinish() error ";
 static const char __pyx_k_PgenWriter_close_called_when_num[] = "PgenWriter.close() called when number of written variants (";
 static const char __pyx_k_Sample_major_read_range_geno_int[] = "Sample-major read_range() geno_int_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_alleles_and_p[] = "Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_alleles_range[] = "Variant-major read_alleles_range() allele_int32_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_list_geno_int[] = "Variant-major read_list() geno_int_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_range_geno_in[] = "Variant-major read_range() geno_int_out buffer has too few rows (";
-static const char __pyx_k_append_alleles_batch_called_with[] = "append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False";
+static const char __pyx_k_append_alleles_batch_called_with[] = "append_alleles_batch called with invalid allele codes";
 static const char __pyx_k_append_alleles_called_with_all_p[] = "append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False";
+static const char __pyx_k_append_alleles_called_with_allel[] = "append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting";
+static const char __pyx_k_append_alleles_called_with_inval[] = "append_alleles called with invalid allele codes";
 static const char __pyx_k_append_dosages_batch_cannot_be_c[] = "append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False";
 static const char __pyx_k_append_partially_phased_batch_ca[] = "append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False";
 static const char __pyx_k_append_partially_phased_batch_er[] = "append_partially_phased_batch() error ";
+static const char __pyx_k_append_partially_phased_called_w[] = "append_partially_phased called with invalid allele codes";
 static const char __pyx_k_append_partially_phased_cannot_b[] = "append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False";
+static const char __pyx_k_count_genocount_uint32_out_buffe[] = "count() genocount_uint32_out buffer is too small for multiallelic variant";
 static const char __pyx_k_no_default___reduce___due_to_non[] = "no default __reduce__ due to non-trivial __cinit__";
 static const char __pyx_k_numpy_core_umath_failed_to_impor[] = "numpy.core.umath failed to import";
 static const char __pyx_k_read_alleles_allele_int32_out_is[] = "read_alleles() allele_int32_out is too small (";
 static const char __pyx_k_read_alleles_and_phasepresent_al[] = "read_alleles_and_phasepresent() allele_int32_out is too small (";
 static const char __pyx_k_read_alleles_and_phasepresent_er[] = "read_alleles_and_phasepresent() error ";
 static const char __pyx_k_read_alleles_and_phasepresent_li[] = "read_alleles_and_phasepresent_list";
 static const char __pyx_k_read_alleles_and_phasepresent_ph[] = "read_alleles_and_phasepresent() phasepresent_out is too small (";
@@ -2650,51 +2712,73 @@
 static const char __pyx_k_read_requires_geno_int_out_to_be[] = "read() requires geno_int_out to be one-dimensional.";
 static const char __pyx_k_sample_subset_is_not_in_strictly[] = "sample_subset is not in strictly increasing order.";
 static const char __pyx_k_Sample_major_read_list_geno_int_2[] = "Sample-major read_list() geno_int_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_alleles_list_2[] = "Variant-major read_alleles_list() allele_int32_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_dosages_list_2[] = "Variant-major read_dosages_list() floatarr_out buffer has too few columns (";
 static const char __pyx_k_and_column_count_should_be_twic_2[] = ", and column count should be twice that)";
 static const char __pyx_k_Haplotype_major_read_alleles_lis_2[] = "Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (";
+static const char __pyx_k_Haplotype_major_read_alleles_lis_3[] = "Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.";
 static const char __pyx_k_Haplotype_major_read_alleles_ran_2[] = "Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (";
+static const char __pyx_k_Haplotype_major_read_alleles_ran_3[] = "Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.";
+static const char __pyx_k_Invalid_arguments_for_PgenWriter_2[] = "Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).";
+static const char __pyx_k_Invalid_arguments_for_PgenWriter_3[] = "Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).";
+static const char __pyx_k_Invalid_arguments_for_PgenWriter_4[] = "Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).";
 static const char __pyx_k_Sample_major_read_range_geno_int_2[] = "Sample-major read_range() geno_int_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_2[] = "Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_3[] = "Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_4[] = "Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_5[] = "Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_6[] = "Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_7[] = "Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_alleles_and_p_8[] = "Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_alleles_range_2[] = "Variant-major read_alleles_range() allele_int32_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_list_geno_int_2[] = "Variant-major read_list() geno_int_out buffer has too few columns (";
 static const char __pyx_k_Variant_major_read_range_geno_in_2[] = "Variant-major read_range() geno_int_out buffer has too few columns (";
+static const char __pyx_k_append_alleles_batch_called_with_2[] = "append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting";
+static const char __pyx_k_append_alleles_batch_called_with_3[] = "append_alleles_batch called with allele codes >= allele_cts[] value";
+static const char __pyx_k_append_alleles_batch_called_with_4[] = "append_alleles_batch called with allele_cts[] value > allele_ct_limit";
+static const char __pyx_k_append_alleles_batch_called_with_5[] = "append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False";
+static const char __pyx_k_append_alleles_called_with_allel_2[] = "append_alleles called with allele codes >= allele_ct argument";
+static const char __pyx_k_append_alleles_called_with_allel_3[] = "append_alleles called with allele_ct > allele_ct_limit";
+static const char __pyx_k_append_partially_phased_batch_ca_2[] = "append_partially_phased_batch called with invalid allele codes";
+static const char __pyx_k_append_partially_phased_batch_ca_3[] = "append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting";
+static const char __pyx_k_append_partially_phased_called_w_2[] = "append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting";
+static const char __pyx_k_append_partially_phased_called_w_3[] = "append_partially_phased called with allele codes >= allele_ct argument";
+static const char __pyx_k_append_partially_phased_called_w_4[] = "append_partially_phased called with allele_ct > allele_ct_limit";
 static const char __pyx_k_read_alleles_and_phasepresent_li_2[] = "read_alleles_and_phasepresent_list() variant index too large (";
 static const char __pyx_k_read_alleles_and_phasepresent_li_3[] = "read_alleles_and_phasepresent_list() error ";
 static const char __pyx_k_read_alleles_and_phasepresent_li_4[] = "read_alleles_and_phasepresent_list() does not support hap_maj == 1 yet.";
 static const char __pyx_k_read_alleles_and_phasepresent_ra_2[] = "read_alleles_and_phasepresent_range() does not support hap_maj == 1 yet.";
 static PyObject *__pyx_kp_u_0_based_sample_idx_too_large;
 static PyObject *__pyx_kp_u_Empty_sample_subset_is_not_curre;
 static PyObject *__pyx_kp_u_Haplotype_major_read_alleles_lis;
 static PyObject *__pyx_kp_u_Haplotype_major_read_alleles_lis_2;
+static PyObject *__pyx_kp_u_Haplotype_major_read_alleles_lis_3;
 static PyObject *__pyx_kp_u_Haplotype_major_read_alleles_ran;
 static PyObject *__pyx_kp_u_Haplotype_major_read_alleles_ran_2;
+static PyObject *__pyx_kp_u_Haplotype_major_read_alleles_ran_3;
 static PyObject *__pyx_n_s_ImportError;
 static PyObject *__pyx_kp_u_Invalid_append_dosages_batch_dos;
 static PyObject *__pyx_kp_u_Invalid_append_dosages_dosage_ar;
 static PyObject *__pyx_kp_u_Invalid_arguments_for_PgenWriter;
+static PyObject *__pyx_kp_u_Invalid_arguments_for_PgenWriter_2;
+static PyObject *__pyx_kp_u_Invalid_arguments_for_PgenWriter_3;
+static PyObject *__pyx_kp_u_Invalid_arguments_for_PgenWriter_4;
 static PyObject *__pyx_kp_u_Invalid_read_dosages_floatarr_ou;
 static PyObject *__pyx_kp_u_Invalid_read_dosages_list_floata;
 static PyObject *__pyx_kp_u_Invalid_read_dosages_range_float;
 static PyObject *__pyx_kp_u_Invalid_read_geno_int_out_array;
 static PyObject *__pyx_kp_u_Invalid_read_list_geno_int_out_a;
 static PyObject *__pyx_kp_u_Invalid_read_range_geno_int_out;
 static PyObject *__pyx_n_s_MemoryError;
-static PyObject *__pyx_kp_u_Multiallelic_phase_dosage_datase;
-static PyObject *__pyx_kp_u_Multiallelic_variants_aren_t_sup;
 static PyObject *__pyx_n_s_PgenReader;
+static PyObject *__pyx_kp_u_PgenReader_multiallelic_variants;
 static PyObject *__pyx_n_s_PgenWriter;
+static PyObject *__pyx_kp_u_PgenWriter___dealloc___SpgwFinis;
+static PyObject *__pyx_kp_u_PgenWriter_close_SpgwFinish_erro;
 static PyObject *__pyx_kp_u_PgenWriter_close_called_when_num;
 static PyObject *__pyx_kp_u_PgrInit_error;
 static PyObject *__pyx_n_s_RuntimeError;
 static PyObject *__pyx_kp_u_Sample_major_read_list_geno_int;
 static PyObject *__pyx_kp_u_Sample_major_read_list_geno_int_2;
 static PyObject *__pyx_kp_u_Sample_major_read_range_geno_int;
 static PyObject *__pyx_kp_u_Sample_major_read_range_geno_int_2;
@@ -2718,50 +2802,68 @@
 static PyObject *__pyx_kp_u_Variant_major_read_list_geno_int_2;
 static PyObject *__pyx_kp_u_Variant_major_read_range_geno_in;
 static PyObject *__pyx_kp_u_Variant_major_read_range_geno_in_2;
 static PyObject *__pyx_kp_u__10;
 static PyObject *__pyx_kp_u__5;
 static PyObject *__pyx_kp_u__9;
 static PyObject *__pyx_n_s_all_phased;
+static PyObject *__pyx_n_s_allele_ct;
+static PyObject *__pyx_n_s_allele_ct_limit;
+static PyObject *__pyx_n_s_allele_cts;
 static PyObject *__pyx_n_s_allele_idx;
 static PyObject *__pyx_n_s_allele_idx_offsets;
 static PyObject *__pyx_n_s_allele_int32;
 static PyObject *__pyx_n_s_allele_int32_batch;
 static PyObject *__pyx_n_s_allele_int32_out;
 static PyObject *__pyx_kp_u_and_column_count_should_be_twic;
 static PyObject *__pyx_kp_u_and_column_count_should_be_twic_2;
 static PyObject *__pyx_kp_u_and_row_count_should_be_twice_t;
 static PyObject *__pyx_n_s_append_alleles;
 static PyObject *__pyx_n_s_append_alleles_batch;
 static PyObject *__pyx_kp_u_append_alleles_batch_called_with;
+static PyObject *__pyx_kp_u_append_alleles_batch_called_with_2;
+static PyObject *__pyx_kp_u_append_alleles_batch_called_with_3;
+static PyObject *__pyx_kp_u_append_alleles_batch_called_with_4;
+static PyObject *__pyx_kp_u_append_alleles_batch_called_with_5;
 static PyObject *__pyx_kp_u_append_alleles_batch_error;
 static PyObject *__pyx_kp_u_append_alleles_called_with_all_p;
+static PyObject *__pyx_kp_u_append_alleles_called_with_allel;
+static PyObject *__pyx_kp_u_append_alleles_called_with_allel_2;
+static PyObject *__pyx_kp_u_append_alleles_called_with_allel_3;
+static PyObject *__pyx_kp_u_append_alleles_called_with_inval;
 static PyObject *__pyx_kp_u_append_alleles_error;
 static PyObject *__pyx_n_s_append_biallelic;
 static PyObject *__pyx_n_s_append_biallelic_batch;
 static PyObject *__pyx_kp_u_append_biallelic_batch_error;
 static PyObject *__pyx_kp_u_append_biallelic_error;
 static PyObject *__pyx_n_s_append_dosages;
 static PyObject *__pyx_n_s_append_dosages_batch;
 static PyObject *__pyx_kp_u_append_dosages_batch_cannot_be_c;
 static PyObject *__pyx_kp_u_append_dosages_batch_error;
 static PyObject *__pyx_kp_u_append_dosages_cannot_be_called;
 static PyObject *__pyx_kp_u_append_dosages_error;
 static PyObject *__pyx_n_s_append_partially_phased;
 static PyObject *__pyx_n_s_append_partially_phased_batch;
 static PyObject *__pyx_kp_u_append_partially_phased_batch_ca;
+static PyObject *__pyx_kp_u_append_partially_phased_batch_ca_2;
+static PyObject *__pyx_kp_u_append_partially_phased_batch_ca_3;
 static PyObject *__pyx_kp_u_append_partially_phased_batch_er;
+static PyObject *__pyx_kp_u_append_partially_phased_called_w;
+static PyObject *__pyx_kp_u_append_partially_phased_called_w_2;
+static PyObject *__pyx_kp_u_append_partially_phased_called_w_3;
+static PyObject *__pyx_kp_u_append_partially_phased_called_w_4;
 static PyObject *__pyx_kp_u_append_partially_phased_cannot_b;
 static PyObject *__pyx_kp_u_append_partially_phased_error;
 static PyObject *__pyx_n_s_change_sample_subset;
 static PyObject *__pyx_n_s_cline_in_traceback;
 static PyObject *__pyx_n_s_close;
 static PyObject *__pyx_kp_u_close_error;
 static PyObject *__pyx_n_s_count;
 static PyObject *__pyx_kp_u_count_error;
+static PyObject *__pyx_kp_u_count_genocount_uint32_out_buffe;
 static PyObject *__pyx_kp_u_current_sample_subset_has_size;
 static PyObject *__pyx_n_s_dosage_phase_present;
 static PyObject *__pyx_n_s_dosage_present;
 static PyObject *__pyx_n_s_dtype;
 static PyObject *__pyx_n_s_enter;
 static PyObject *__pyx_n_s_exc_tb;
 static PyObject *__pyx_n_s_exc_type;
@@ -2806,14 +2908,15 @@
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_al;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_er;
 static PyObject *__pyx_n_s_read_alleles_and_phasepresent_li;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_li_2;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_li_3;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_li_4;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_ph;
+static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_r;
 static PyObject *__pyx_n_s_read_alleles_and_phasepresent_ra;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_ra_2;
 static PyObject *__pyx_kp_u_read_alleles_and_phasepresent_va;
 static PyObject *__pyx_kp_u_read_alleles_error;
 static PyObject *__pyx_n_s_read_alleles_list;
 static PyObject *__pyx_kp_u_read_alleles_list_error;
 static PyObject *__pyx_kp_u_read_alleles_list_variant_index;
@@ -2855,22 +2958,23 @@
 static PyObject *__pyx_n_s_setstate;
 static PyObject *__pyx_n_s_setstate_cython;
 static PyObject *__pyx_n_s_size;
 static PyObject *__pyx_n_s_sys;
 static PyObject *__pyx_n_s_test;
 static PyObject *__pyx_kp_u_unequal_to_initially_declared_v;
 static PyObject *__pyx_n_s_variant_ct;
+static PyObject *__pyx_n_s_variant_ct_limit;
 static PyObject *__pyx_n_s_variant_idx;
 static PyObject *__pyx_kp_u_variant_idx_2;
 static PyObject *__pyx_n_s_variant_idx_end;
 static PyObject *__pyx_kp_u_variant_idx_end_variant_idx_sta;
 static PyObject *__pyx_n_s_variant_idx_start;
 static PyObject *__pyx_n_s_variant_idxs;
 static PyObject *__pyx_kp_u_variant_idxs_length_is;
-static int __pyx_pf_7pgenlib_10PgenReader___cinit__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyObject *__pyx_v_filename, PyObject *__pyx_v_raw_sample_ct, PyObject *__pyx_v_variant_ct, PyObject *__pyx_v_sample_subset); /* proto */
+static int __pyx_pf_7pgenlib_10PgenReader___cinit__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyObject *__pyx_v_filename, PyObject *__pyx_v_raw_sample_ct, PyObject *__pyx_v_variant_ct, PyObject *__pyx_v_sample_subset, PyObject *__pyx_v_allele_idx_offsets); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_2__enter__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_4get_raw_sample_ct(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_6get_variant_ct(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_8hardcall_phase_present(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_10read(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_geno_int_out, uint32_t __pyx_v_allele_idx); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_12read_dosages(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_floatarr_out, uint32_t __pyx_v_allele_idx); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_14read_alleles(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_allele_int32_out); /* proto */
@@ -2886,34 +2990,35 @@
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_34count(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_genocount_uint32_out, PyObject *__pyx_v_allele_idx); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_36change_sample_subset(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyObject *__pyx_v_sample_subset); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_38close(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_40__exit__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyObject *__pyx_v_exc_type, PyObject *__pyx_v_exc_val, PyObject *__pyx_v_exc_tb); /* proto */
 static void __pyx_pf_7pgenlib_10PgenReader_42__dealloc__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_44__reduce_cython__(CYTHON_UNUSED struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenReader_46__setstate_cython__(CYTHON_UNUSED struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, CYTHON_UNUSED PyObject *__pyx_v___pyx_state); /* proto */
-static int __pyx_pf_7pgenlib_10PgenWriter___cinit__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyObject *__pyx_v_filename, uint32_t __pyx_v_sample_ct, uint32_t __pyx_v_variant_ct, PyObject *__pyx_v_nonref_flags, PyObject *__pyx_v_allele_idx_offsets, int __pyx_v_hardcall_phase_present, int __pyx_v_dosage_present, int __pyx_v_dosage_phase_present); /* proto */
+static int __pyx_pf_7pgenlib_10PgenWriter___cinit__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyObject *__pyx_v_filename, uint32_t __pyx_v_sample_ct, PyObject *__pyx_v_variant_ct, PyObject *__pyx_v_nonref_flags, uint32_t __pyx_v_allele_ct_limit, int __pyx_v_hardcall_phase_present, int __pyx_v_dosage_present, int __pyx_v_dosage_phase_present, PyObject *__pyx_v_variant_ct_limit); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_2__enter__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_4append_biallelic(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_geno_int8); /* proto */
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_6append_alleles(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, int __pyx_v_all_phased); /* proto */
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_8append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent); /* proto */
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_6append_alleles(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, int __pyx_v_all_phased, PyObject *__pyx_v_allele_ct); /* proto */
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_8append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent, PyObject *__pyx_v_allele_ct); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_10append_dosages(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_floatarr); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_12append_biallelic_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_geno_int8_batch); /* proto */
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_14append_alleles_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, int __pyx_v_all_phased); /* proto */
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_16append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch); /* proto */
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_14append_alleles_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, int __pyx_v_all_phased, PyObject *__pyx_v_allele_cts); /* proto */
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_16append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch, PyObject *__pyx_v_allele_cts); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_18append_dosages_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_floatarr_batch); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_20close(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_22__exit__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyObject *__pyx_v_exc_type, PyObject *__pyx_v_exc_val, PyObject *__pyx_v_exc_tb); /* proto */
 static void __pyx_pf_7pgenlib_10PgenWriter_24__dealloc__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_26__reduce_cython__(CYTHON_UNUSED struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self); /* proto */
 static PyObject *__pyx_pf_7pgenlib_10PgenWriter_28__setstate_cython__(CYTHON_UNUSED struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, CYTHON_UNUSED PyObject *__pyx_v___pyx_state); /* proto */
 static PyObject *__pyx_tp_new_7pgenlib_PgenReader(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
 static PyObject *__pyx_tp_new_7pgenlib_PgenWriter(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
 static PyObject *__pyx_int_0;
 static PyObject *__pyx_int_1;
 static PyObject *__pyx_int_2;
+static PyObject *__pyx_int_3;
 static PyObject *__pyx_tuple_;
 static PyObject *__pyx_tuple__2;
 static PyObject *__pyx_tuple__3;
 static PyObject *__pyx_tuple__4;
 static PyObject *__pyx_tuple__6;
 static PyObject *__pyx_tuple__7;
 static PyObject *__pyx_tuple__8;
@@ -2937,19 +3042,175 @@
 static PyObject *__pyx_tuple__28;
 static PyObject *__pyx_tuple__29;
 static PyObject *__pyx_tuple__30;
 static PyObject *__pyx_tuple__31;
 static PyObject *__pyx_tuple__32;
 static PyObject *__pyx_tuple__33;
 static PyObject *__pyx_tuple__34;
+static PyObject *__pyx_tuple__35;
+static PyObject *__pyx_tuple__36;
+static PyObject *__pyx_tuple__37;
+static PyObject *__pyx_tuple__38;
+static PyObject *__pyx_tuple__39;
+static PyObject *__pyx_tuple__40;
+static PyObject *__pyx_tuple__41;
+static PyObject *__pyx_tuple__42;
+static PyObject *__pyx_tuple__43;
+static PyObject *__pyx_tuple__44;
+static PyObject *__pyx_tuple__45;
+static PyObject *__pyx_tuple__46;
+static PyObject *__pyx_tuple__47;
+static PyObject *__pyx_tuple__48;
+static PyObject *__pyx_tuple__49;
+static PyObject *__pyx_tuple__50;
+static PyObject *__pyx_tuple__51;
+static PyObject *__pyx_tuple__52;
+static PyObject *__pyx_tuple__53;
 /* Late includes */
 
-/* "pgenlib.pyx":215
+/* "pgenlib.pyx":249
+ *     cdef uintptr_t* _multivar_smaj_phasepresent_batch_buf
+ * 
+ *     cdef set_allele_idx_offsets_internal(self, np.ndarray[np.uintp_t,mode="c",ndim=1] allele_idx_offsets):             # <<<<<<<<<<<<<<
+ *         # Make a copy instead of trying to share this with the caller.
+ *         cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct
+ */
+
+static PyObject *__pyx_f_7pgenlib_10PgenReader_set_allele_idx_offsets_internal(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyArrayObject *__pyx_v_allele_idx_offsets) {
+  uint32_t __pyx_v_nvariant;
+  uintptr_t __pyx_v_nbytes;
+  __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_idx_offsets;
+  __Pyx_Buffer __pyx_pybuffer_allele_idx_offsets;
+  PyObject *__pyx_r = NULL;
+  __Pyx_RefNannyDeclarations
+  uint32_t __pyx_t_1;
+  int __pyx_t_2;
+  Py_ssize_t __pyx_t_3;
+  int __pyx_t_4;
+  int __pyx_lineno = 0;
+  const char *__pyx_filename = NULL;
+  int __pyx_clineno = 0;
+  __Pyx_RefNannySetupContext("set_allele_idx_offsets_internal", 0);
+  __pyx_pybuffer_allele_idx_offsets.pybuffer.buf = NULL;
+  __pyx_pybuffer_allele_idx_offsets.refcount = 0;
+  __pyx_pybuffernd_allele_idx_offsets.data = NULL;
+  __pyx_pybuffernd_allele_idx_offsets.rcbuffer = &__pyx_pybuffer_allele_idx_offsets;
+  {
+    __Pyx_BufFmt_StackElem __pyx_stack[1];
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_idx_offsets.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_idx_offsets, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uintp_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 249, __pyx_L1_error)
+  }
+  __pyx_pybuffernd_allele_idx_offsets.diminfo[0].strides = __pyx_pybuffernd_allele_idx_offsets.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_idx_offsets.diminfo[0].shape = __pyx_pybuffernd_allele_idx_offsets.rcbuffer->pybuffer.shape[0];
+
+  /* "pgenlib.pyx":251
+ *     cdef set_allele_idx_offsets_internal(self, np.ndarray[np.uintp_t,mode="c",ndim=1] allele_idx_offsets):
+ *         # Make a copy instead of trying to share this with the caller.
+ *         cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t nbytes = (1 + nvariant) * sizeof(uintptr_t)
+ *         if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):
+ */
+  __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
+  __pyx_v_nvariant = __pyx_t_1;
+
+  /* "pgenlib.pyx":252
+ *         # Make a copy instead of trying to share this with the caller.
+ *         cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct
+ *         cdef uintptr_t nbytes = (1 + nvariant) * sizeof(uintptr_t)             # <<<<<<<<<<<<<<
+ *         if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):
+ *             raise MemoryError()
+ */
+  __pyx_v_nbytes = ((1 + __pyx_v_nvariant) * (sizeof(uintptr_t)));
+
+  /* "pgenlib.pyx":253
+ *         cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct
+ *         cdef uintptr_t nbytes = (1 + nvariant) * sizeof(uintptr_t)
+ *         if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):             # <<<<<<<<<<<<<<
+ *             raise MemoryError()
+ *         memcpy(self._info_ptr[0].allele_idx_offsets, &(allele_idx_offsets[0]), nbytes)
+ */
+  __pyx_t_2 = (plink2::cachealigned_malloc(__pyx_v_nbytes, (&(__pyx_v_self->_info_ptr[0]).allele_idx_offsets)) != 0);
+  if (unlikely(__pyx_t_2)) {
+
+    /* "pgenlib.pyx":254
+ *         cdef uintptr_t nbytes = (1 + nvariant) * sizeof(uintptr_t)
+ *         if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):
+ *             raise MemoryError()             # <<<<<<<<<<<<<<
+ *         memcpy(self._info_ptr[0].allele_idx_offsets, &(allele_idx_offsets[0]), nbytes)
+ *         self._info_ptr[0].max_allele_ct = PglComputeMaxAlleleCt(self._info_ptr[0].allele_idx_offsets, nvariant)
+ */
+    PyErr_NoMemory(); __PYX_ERR(0, 254, __pyx_L1_error)
+
+    /* "pgenlib.pyx":253
+ *         cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct
+ *         cdef uintptr_t nbytes = (1 + nvariant) * sizeof(uintptr_t)
+ *         if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):             # <<<<<<<<<<<<<<
+ *             raise MemoryError()
+ *         memcpy(self._info_ptr[0].allele_idx_offsets, &(allele_idx_offsets[0]), nbytes)
+ */
+  }
+
+  /* "pgenlib.pyx":255
+ *         if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):
+ *             raise MemoryError()
+ *         memcpy(self._info_ptr[0].allele_idx_offsets, &(allele_idx_offsets[0]), nbytes)             # <<<<<<<<<<<<<<
+ *         self._info_ptr[0].max_allele_ct = PglComputeMaxAlleleCt(self._info_ptr[0].allele_idx_offsets, nvariant)
+ * 
+ */
+  __pyx_t_3 = 0;
+  __pyx_t_4 = -1;
+  if (__pyx_t_3 < 0) {
+    __pyx_t_3 += __pyx_pybuffernd_allele_idx_offsets.diminfo[0].shape;
+    if (unlikely(__pyx_t_3 < 0)) __pyx_t_4 = 0;
+  } else if (unlikely(__pyx_t_3 >= __pyx_pybuffernd_allele_idx_offsets.diminfo[0].shape)) __pyx_t_4 = 0;
+  if (unlikely(__pyx_t_4 != -1)) {
+    __Pyx_RaiseBufferIndexError(__pyx_t_4);
+    __PYX_ERR(0, 255, __pyx_L1_error)
+  }
+  (void)(memcpy((__pyx_v_self->_info_ptr[0]).allele_idx_offsets, (&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uintp_t *, __pyx_pybuffernd_allele_idx_offsets.rcbuffer->pybuffer.buf, __pyx_t_3, __pyx_pybuffernd_allele_idx_offsets.diminfo[0].strides))), __pyx_v_nbytes));
+
+  /* "pgenlib.pyx":256
+ *             raise MemoryError()
+ *         memcpy(self._info_ptr[0].allele_idx_offsets, &(allele_idx_offsets[0]), nbytes)
+ *         self._info_ptr[0].max_allele_ct = PglComputeMaxAlleleCt(self._info_ptr[0].allele_idx_offsets, nvariant)             # <<<<<<<<<<<<<<
+ * 
+ *     cdef set_sample_subset_internal(self, np.ndarray[np.uint32_t,mode="c",ndim=1] sample_subset):
+ */
+  (__pyx_v_self->_info_ptr[0]).max_allele_ct = plink2::PglComputeMaxAlleleCt((__pyx_v_self->_info_ptr[0]).allele_idx_offsets, __pyx_v_nvariant);
+
+  /* "pgenlib.pyx":249
  *     cdef uintptr_t* _multivar_smaj_phasepresent_batch_buf
  * 
+ *     cdef set_allele_idx_offsets_internal(self, np.ndarray[np.uintp_t,mode="c",ndim=1] allele_idx_offsets):             # <<<<<<<<<<<<<<
+ *         # Make a copy instead of trying to share this with the caller.
+ *         cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct
+ */
+
+  /* function exit code */
+  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
+  goto __pyx_L0;
+  __pyx_L1_error:;
+  { PyObject *__pyx_type, *__pyx_value, *__pyx_tb;
+    __Pyx_PyThreadState_declare
+    __Pyx_PyThreadState_assign
+    __Pyx_ErrFetch(&__pyx_type, &__pyx_value, &__pyx_tb);
+    __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_allele_idx_offsets.rcbuffer->pybuffer);
+  __Pyx_ErrRestore(__pyx_type, __pyx_value, __pyx_tb);}
+  __Pyx_AddTraceback("pgenlib.PgenReader.set_allele_idx_offsets_internal", __pyx_clineno, __pyx_lineno, __pyx_filename);
+  __pyx_r = 0;
+  goto __pyx_L2;
+  __pyx_L0:;
+  __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_allele_idx_offsets.rcbuffer->pybuffer);
+  __pyx_L2:;
+  __Pyx_XGIVEREF(__pyx_r);
+  __Pyx_RefNannyFinishContext();
+  return __pyx_r;
+}
+
+/* "pgenlib.pyx":258
+ *         self._info_ptr[0].max_allele_ct = PglComputeMaxAlleleCt(self._info_ptr[0].allele_idx_offsets, nvariant)
+ * 
  *     cdef set_sample_subset_internal(self, np.ndarray[np.uint32_t,mode="c",ndim=1] sample_subset):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_sample_ct = self._info_ptr[0].raw_sample_ct
  *         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)
  */
 
 static PyObject *__pyx_f_7pgenlib_10PgenReader_set_sample_subset_internal(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyArrayObject *__pyx_v_sample_subset) {
   uint32_t __pyx_v_raw_sample_ct;
@@ -2983,403 +3244,403 @@
   __Pyx_RefNannySetupContext("set_sample_subset_internal", 0);
   __pyx_pybuffer_sample_subset.pybuffer.buf = NULL;
   __pyx_pybuffer_sample_subset.refcount = 0;
   __pyx_pybuffernd_sample_subset.data = NULL;
   __pyx_pybuffernd_sample_subset.rcbuffer = &__pyx_pybuffer_sample_subset;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_sample_subset.rcbuffer->pybuffer, (PyObject*)__pyx_v_sample_subset, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 215, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_sample_subset.rcbuffer->pybuffer, (PyObject*)__pyx_v_sample_subset, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 258, __pyx_L1_error)
   }
   __pyx_pybuffernd_sample_subset.diminfo[0].strides = __pyx_pybuffernd_sample_subset.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_sample_subset.diminfo[0].shape = __pyx_pybuffernd_sample_subset.rcbuffer->pybuffer.shape[0];
 
-  /* "pgenlib.pyx":216
+  /* "pgenlib.pyx":259
  * 
  *     cdef set_sample_subset_internal(self, np.ndarray[np.uint32_t,mode="c",ndim=1] sample_subset):
  *         cdef uint32_t raw_sample_ct = self._info_ptr[0].raw_sample_ct             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)
  *         cdef uint32_t raw_sample_ctaw = raw_sample_ctv * kWordsPerVec
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_sample_ct;
   __pyx_v_raw_sample_ct = __pyx_t_1;
 
-  /* "pgenlib.pyx":217
+  /* "pgenlib.pyx":260
  *     cdef set_sample_subset_internal(self, np.ndarray[np.uint32_t,mode="c",ndim=1] sample_subset):
  *         cdef uint32_t raw_sample_ct = self._info_ptr[0].raw_sample_ct
  *         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_sample_ctaw = raw_sample_ctv * kWordsPerVec
  *         cdef uintptr_t* sample_include = self._subset_include_vec
  */
   __pyx_v_raw_sample_ctv = plink2::DivUp(__pyx_v_raw_sample_ct, plink2::kBitsPerVec);
 
-  /* "pgenlib.pyx":218
+  /* "pgenlib.pyx":261
  *         cdef uint32_t raw_sample_ct = self._info_ptr[0].raw_sample_ct
  *         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)
  *         cdef uint32_t raw_sample_ctaw = raw_sample_ctv * kWordsPerVec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* sample_include = self._subset_include_vec
  *         ZeroWArr(raw_sample_ctaw, sample_include)
  */
   __pyx_v_raw_sample_ctaw = (__pyx_v_raw_sample_ctv * plink2::kWordsPerVec);
 
-  /* "pgenlib.pyx":219
+  /* "pgenlib.pyx":262
  *         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)
  *         cdef uint32_t raw_sample_ctaw = raw_sample_ctv * kWordsPerVec
  *         cdef uintptr_t* sample_include = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         ZeroWArr(raw_sample_ctaw, sample_include)
  *         cdef uint32_t subset_size = sample_subset.size
  */
   __pyx_t_2 = __pyx_v_self->_subset_include_vec;
   __pyx_v_sample_include = __pyx_t_2;
 
-  /* "pgenlib.pyx":220
+  /* "pgenlib.pyx":263
  *         cdef uint32_t raw_sample_ctaw = raw_sample_ctv * kWordsPerVec
  *         cdef uintptr_t* sample_include = self._subset_include_vec
  *         ZeroWArr(raw_sample_ctaw, sample_include)             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = sample_subset.size
  *         if subset_size == 0:
  */
   plink2::ZeroWArr(__pyx_v_raw_sample_ctaw, __pyx_v_sample_include);
 
-  /* "pgenlib.pyx":221
+  /* "pgenlib.pyx":264
  *         cdef uintptr_t* sample_include = self._subset_include_vec
  *         ZeroWArr(raw_sample_ctaw, sample_include)
  *         cdef uint32_t subset_size = sample_subset.size             # <<<<<<<<<<<<<<
  *         if subset_size == 0:
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")
  */
-  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_sample_subset), __pyx_n_s_size); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 221, __pyx_L1_error)
+  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_sample_subset), __pyx_n_s_size); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 264, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
-  __pyx_t_1 = __Pyx_PyInt_As_uint32_t(__pyx_t_3); if (unlikely((__pyx_t_1 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 221, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyInt_As_uint32_t(__pyx_t_3); if (unlikely((__pyx_t_1 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 264, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
   __pyx_v_subset_size = __pyx_t_1;
 
-  /* "pgenlib.pyx":222
+  /* "pgenlib.pyx":265
  *         ZeroWArr(raw_sample_ctaw, sample_include)
  *         cdef uint32_t subset_size = sample_subset.size
  *         if subset_size == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")
  *         cdef sample_uidx = sample_subset[0]
  */
   __pyx_t_4 = ((__pyx_v_subset_size == 0) != 0);
   if (unlikely(__pyx_t_4)) {
 
-    /* "pgenlib.pyx":223
+    /* "pgenlib.pyx":266
  *         cdef uint32_t subset_size = sample_subset.size
  *         if subset_size == 0:
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")             # <<<<<<<<<<<<<<
  *         cdef sample_uidx = sample_subset[0]
  *         cdef uint32_t idx = 0
  */
-    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple_, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 223, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple_, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 266, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
     __Pyx_Raise(__pyx_t_3, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __PYX_ERR(0, 223, __pyx_L1_error)
+    __PYX_ERR(0, 266, __pyx_L1_error)
 
-    /* "pgenlib.pyx":222
+    /* "pgenlib.pyx":265
  *         ZeroWArr(raw_sample_ctaw, sample_include)
  *         cdef uint32_t subset_size = sample_subset.size
  *         if subset_size == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")
  *         cdef sample_uidx = sample_subset[0]
  */
   }
 
-  /* "pgenlib.pyx":224
+  /* "pgenlib.pyx":267
  *         if subset_size == 0:
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")
  *         cdef sample_uidx = sample_subset[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t idx = 0
  *         cdef next_uidx
  */
   __pyx_t_5 = 0;
   __pyx_t_6 = -1;
   if (__pyx_t_5 < 0) {
     __pyx_t_5 += __pyx_pybuffernd_sample_subset.diminfo[0].shape;
     if (unlikely(__pyx_t_5 < 0)) __pyx_t_6 = 0;
   } else if (unlikely(__pyx_t_5 >= __pyx_pybuffernd_sample_subset.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 224, __pyx_L1_error)
+    __PYX_ERR(0, 267, __pyx_L1_error)
   }
-  __pyx_t_3 = __Pyx_PyInt_From_npy_uint32((*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_sample_subset.rcbuffer->pybuffer.buf, __pyx_t_5, __pyx_pybuffernd_sample_subset.diminfo[0].strides))); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 224, __pyx_L1_error)
+  __pyx_t_3 = __Pyx_PyInt_From_npy_uint32((*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_sample_subset.rcbuffer->pybuffer.buf, __pyx_t_5, __pyx_pybuffernd_sample_subset.diminfo[0].strides))); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 267, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
   __pyx_v_sample_uidx = __pyx_t_3;
   __pyx_t_3 = 0;
 
-  /* "pgenlib.pyx":225
+  /* "pgenlib.pyx":268
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")
  *         cdef sample_uidx = sample_subset[0]
  *         cdef uint32_t idx = 0             # <<<<<<<<<<<<<<
  *         cdef next_uidx
  *         while True:
  */
   __pyx_v_idx = 0;
 
-  /* "pgenlib.pyx":227
+  /* "pgenlib.pyx":270
  *         cdef uint32_t idx = 0
  *         cdef next_uidx
  *         while True:             # <<<<<<<<<<<<<<
  *             if sample_uidx >= raw_sample_ct:
  *                 raise RuntimeError("0-based sample idx too large (" + str(sample_uidx) + "; only " + str(raw_sample_ct) + " in file).")
  */
   while (1) {
 
-    /* "pgenlib.pyx":228
+    /* "pgenlib.pyx":271
  *         cdef next_uidx
  *         while True:
  *             if sample_uidx >= raw_sample_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("0-based sample idx too large (" + str(sample_uidx) + "; only " + str(raw_sample_ct) + " in file).")
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)
  */
-    __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_sample_ct); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 228, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_sample_ct); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 271, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
-    __pyx_t_7 = PyObject_RichCompare(__pyx_v_sample_uidx, __pyx_t_3, Py_GE); __Pyx_XGOTREF(__pyx_t_7); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 228, __pyx_L1_error)
+    __pyx_t_7 = PyObject_RichCompare(__pyx_v_sample_uidx, __pyx_t_3, Py_GE); __Pyx_XGOTREF(__pyx_t_7); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 271, __pyx_L1_error)
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __pyx_t_4 = __Pyx_PyObject_IsTrue(__pyx_t_7); if (unlikely(__pyx_t_4 < 0)) __PYX_ERR(0, 228, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_IsTrue(__pyx_t_7); if (unlikely(__pyx_t_4 < 0)) __PYX_ERR(0, 271, __pyx_L1_error)
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     if (unlikely(__pyx_t_4)) {
 
-      /* "pgenlib.pyx":229
+      /* "pgenlib.pyx":272
  *         while True:
  *             if sample_uidx >= raw_sample_ct:
  *                 raise RuntimeError("0-based sample idx too large (" + str(sample_uidx) + "; only " + str(raw_sample_ct) + " in file).")             # <<<<<<<<<<<<<<
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)
  *             idx += 1
  */
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_v_sample_uidx); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_v_sample_uidx); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_kp_u_0_based_sample_idx_too_large, __pyx_t_7); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_kp_u_0_based_sample_idx_too_large, __pyx_t_7); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_3);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_3, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_3, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-      __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_sample_ct); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_sample_ct); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_3);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_3); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_3); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-      __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_3);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_3, __pyx_kp_u_in_file); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_3, __pyx_kp_u_in_file); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 229, __pyx_L1_error)
+      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 272, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_3);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_3, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-      __PYX_ERR(0, 229, __pyx_L1_error)
+      __PYX_ERR(0, 272, __pyx_L1_error)
 
-      /* "pgenlib.pyx":228
+      /* "pgenlib.pyx":271
  *         cdef next_uidx
  *         while True:
  *             if sample_uidx >= raw_sample_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("0-based sample idx too large (" + str(sample_uidx) + "; only " + str(raw_sample_ct) + " in file).")
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)
  */
     }
 
-    /* "pgenlib.pyx":230
+    /* "pgenlib.pyx":273
  *             if sample_uidx >= raw_sample_ct:
  *                 raise RuntimeError("0-based sample idx too large (" + str(sample_uidx) + "; only " + str(raw_sample_ct) + " in file).")
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)             # <<<<<<<<<<<<<<
  *             idx += 1
  *             if idx == subset_size:
  */
-    __pyx_t_3 = __Pyx_PyInt_From_int(plink2::kBitsPerWord); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyInt_From_int(plink2::kBitsPerWord); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
-    __pyx_t_8 = PyNumber_FloorDivide(__pyx_v_sample_uidx, __pyx_t_3); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_8 = PyNumber_FloorDivide(__pyx_v_sample_uidx, __pyx_t_3); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __pyx_t_9 = __Pyx_PyIndex_AsSsize_t(__pyx_t_8); if (unlikely((__pyx_t_9 == (Py_ssize_t)-1) && PyErr_Occurred())) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_9 = __Pyx_PyIndex_AsSsize_t(__pyx_t_8); if (unlikely((__pyx_t_9 == (Py_ssize_t)-1) && PyErr_Occurred())) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_FromSize_t((__pyx_v_sample_include[__pyx_t_9])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_FromSize_t((__pyx_v_sample_include[__pyx_t_9])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_3 = __Pyx_PyInt_From_int(plink2::k1LU); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyInt_From_int(plink2::k1LU); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
-    __pyx_t_7 = __Pyx_PyInt_From_int(plink2::kBitsPerWord); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_int(plink2::kBitsPerWord); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_10 = PyNumber_Remainder(__pyx_v_sample_uidx, __pyx_t_7); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_10 = PyNumber_Remainder(__pyx_v_sample_uidx, __pyx_t_7); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_10);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = PyNumber_Lshift(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_7 = PyNumber_Lshift(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
     __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-    __pyx_t_10 = PyNumber_InPlaceOr(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_10 = PyNumber_InPlaceOr(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_10);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_11 = __Pyx_PyInt_As_size_t(__pyx_t_10); if (unlikely((__pyx_t_11 == ((uintptr_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 230, __pyx_L1_error)
+    __pyx_t_11 = __Pyx_PyInt_As_size_t(__pyx_t_10); if (unlikely((__pyx_t_11 == ((uintptr_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 273, __pyx_L1_error)
     __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
     (__pyx_v_sample_include[__pyx_t_9]) = __pyx_t_11;
 
-    /* "pgenlib.pyx":231
+    /* "pgenlib.pyx":274
  *                 raise RuntimeError("0-based sample idx too large (" + str(sample_uidx) + "; only " + str(raw_sample_ct) + " in file).")
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)
  *             idx += 1             # <<<<<<<<<<<<<<
  *             if idx == subset_size:
  *                 break
  */
     __pyx_v_idx = (__pyx_v_idx + 1);
 
-    /* "pgenlib.pyx":232
+    /* "pgenlib.pyx":275
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)
  *             idx += 1
  *             if idx == subset_size:             # <<<<<<<<<<<<<<
  *                 break
  *             next_uidx = sample_subset[idx]
  */
     __pyx_t_4 = ((__pyx_v_idx == __pyx_v_subset_size) != 0);
     if (__pyx_t_4) {
 
-      /* "pgenlib.pyx":233
+      /* "pgenlib.pyx":276
  *             idx += 1
  *             if idx == subset_size:
  *                 break             # <<<<<<<<<<<<<<
  *             next_uidx = sample_subset[idx]
  * 
  */
       goto __pyx_L5_break;
 
-      /* "pgenlib.pyx":232
+      /* "pgenlib.pyx":275
  *             sample_include[sample_uidx // kBitsPerWord] |= k1LU << (sample_uidx % kBitsPerWord)
  *             idx += 1
  *             if idx == subset_size:             # <<<<<<<<<<<<<<
  *                 break
  *             next_uidx = sample_subset[idx]
  */
     }
 
-    /* "pgenlib.pyx":234
+    /* "pgenlib.pyx":277
  *             if idx == subset_size:
  *                 break
  *             next_uidx = sample_subset[idx]             # <<<<<<<<<<<<<<
  * 
  *             # prohibit this since it implies that the caller expects genotypes
  */
     __pyx_t_12 = __pyx_v_idx;
     __pyx_t_6 = -1;
     if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_sample_subset.diminfo[0].shape)) __pyx_t_6 = 0;
     if (unlikely(__pyx_t_6 != -1)) {
       __Pyx_RaiseBufferIndexError(__pyx_t_6);
-      __PYX_ERR(0, 234, __pyx_L1_error)
+      __PYX_ERR(0, 277, __pyx_L1_error)
     }
-    __pyx_t_10 = __Pyx_PyInt_From_npy_uint32((*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_sample_subset.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_sample_subset.diminfo[0].strides))); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 234, __pyx_L1_error)
+    __pyx_t_10 = __Pyx_PyInt_From_npy_uint32((*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_sample_subset.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_sample_subset.diminfo[0].strides))); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 277, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_10);
     __Pyx_XDECREF_SET(__pyx_v_next_uidx, __pyx_t_10);
     __pyx_t_10 = 0;
 
-    /* "pgenlib.pyx":238
+    /* "pgenlib.pyx":281
  *             # prohibit this since it implies that the caller expects genotypes
  *             # to be returned in a different order
  *             if next_uidx <= sample_uidx:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("sample_subset is not in strictly increasing order.")
  * 
  */
-    __pyx_t_10 = PyObject_RichCompare(__pyx_v_next_uidx, __pyx_v_sample_uidx, Py_LE); __Pyx_XGOTREF(__pyx_t_10); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 238, __pyx_L1_error)
-    __pyx_t_4 = __Pyx_PyObject_IsTrue(__pyx_t_10); if (unlikely(__pyx_t_4 < 0)) __PYX_ERR(0, 238, __pyx_L1_error)
+    __pyx_t_10 = PyObject_RichCompare(__pyx_v_next_uidx, __pyx_v_sample_uidx, Py_LE); __Pyx_XGOTREF(__pyx_t_10); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 281, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_IsTrue(__pyx_t_10); if (unlikely(__pyx_t_4 < 0)) __PYX_ERR(0, 281, __pyx_L1_error)
     __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
     if (unlikely(__pyx_t_4)) {
 
-      /* "pgenlib.pyx":239
+      /* "pgenlib.pyx":282
  *             # to be returned in a different order
  *             if next_uidx <= sample_uidx:
  *                 raise RuntimeError("sample_subset is not in strictly increasing order.")             # <<<<<<<<<<<<<<
  * 
  *             sample_uidx = next_uidx
  */
-      __pyx_t_10 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__2, NULL); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 239, __pyx_L1_error)
+      __pyx_t_10 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__2, NULL); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 282, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_10);
       __Pyx_Raise(__pyx_t_10, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-      __PYX_ERR(0, 239, __pyx_L1_error)
+      __PYX_ERR(0, 282, __pyx_L1_error)
 
-      /* "pgenlib.pyx":238
+      /* "pgenlib.pyx":281
  *             # prohibit this since it implies that the caller expects genotypes
  *             # to be returned in a different order
  *             if next_uidx <= sample_uidx:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("sample_subset is not in strictly increasing order.")
  * 
  */
     }
 
-    /* "pgenlib.pyx":241
+    /* "pgenlib.pyx":284
  *                 raise RuntimeError("sample_subset is not in strictly increasing order.")
  * 
  *             sample_uidx = next_uidx             # <<<<<<<<<<<<<<
  * 
  *         FillInterleavedMaskVec(sample_include, raw_sample_ctv, self._subset_include_interleaved_vec)
  */
     __Pyx_INCREF(__pyx_v_next_uidx);
     __Pyx_DECREF_SET(__pyx_v_sample_uidx, __pyx_v_next_uidx);
   }
   __pyx_L5_break:;
 
-  /* "pgenlib.pyx":243
+  /* "pgenlib.pyx":286
  *             sample_uidx = next_uidx
  * 
  *         FillInterleavedMaskVec(sample_include, raw_sample_ctv, self._subset_include_interleaved_vec)             # <<<<<<<<<<<<<<
  * 
  *         cdef uint32_t raw_sample_ctl = DivUp(raw_sample_ct, kBitsPerWord)
  */
   plink2::FillInterleavedMaskVec(__pyx_v_sample_include, __pyx_v_raw_sample_ctv, __pyx_v_self->_subset_include_interleaved_vec);
 
-  /* "pgenlib.pyx":245
+  /* "pgenlib.pyx":288
  *         FillInterleavedMaskVec(sample_include, raw_sample_ctv, self._subset_include_interleaved_vec)
  * 
  *         cdef uint32_t raw_sample_ctl = DivUp(raw_sample_ct, kBitsPerWord)             # <<<<<<<<<<<<<<
  *         # er, this isn't a safe usage pattern.
  *         FillCumulativePopcounts(sample_include, raw_sample_ctl, self._subset_cumulative_popcounts)
  */
   __pyx_v_raw_sample_ctl = plink2::DivUp(__pyx_v_raw_sample_ct, plink2::kBitsPerWord);
 
-  /* "pgenlib.pyx":247
+  /* "pgenlib.pyx":290
  *         cdef uint32_t raw_sample_ctl = DivUp(raw_sample_ct, kBitsPerWord)
  *         # er, this isn't a safe usage pattern.
  *         FillCumulativePopcounts(sample_include, raw_sample_ctl, self._subset_cumulative_popcounts)             # <<<<<<<<<<<<<<
  *         PgrSetSampleSubsetIndex(self._subset_cumulative_popcounts, self._state_ptr, &(self._subset_index))
  * 
  */
   plink2::FillCumulativePopcounts(__pyx_v_sample_include, __pyx_v_raw_sample_ctl, __pyx_v_self->_subset_cumulative_popcounts);
 
-  /* "pgenlib.pyx":248
+  /* "pgenlib.pyx":291
  *         # er, this isn't a safe usage pattern.
  *         FillCumulativePopcounts(sample_include, raw_sample_ctl, self._subset_cumulative_popcounts)
  *         PgrSetSampleSubsetIndex(self._subset_cumulative_popcounts, self._state_ptr, &(self._subset_index))             # <<<<<<<<<<<<<<
  * 
  *         self._subset_size = subset_size
  */
   plink2::PgrSetSampleSubsetIndex(__pyx_v_self->_subset_cumulative_popcounts, __pyx_v_self->_state_ptr, (&__pyx_v_self->_subset_index));
 
-  /* "pgenlib.pyx":250
+  /* "pgenlib.pyx":293
  *         PgrSetSampleSubsetIndex(self._subset_cumulative_popcounts, self._state_ptr, &(self._subset_index))
  * 
  *         self._subset_size = subset_size             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   __pyx_v_self->_subset_size = __pyx_v_subset_size;
 
-  /* "pgenlib.pyx":251
+  /* "pgenlib.pyx":294
  * 
  *         self._subset_size = subset_size
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":215
- *     cdef uintptr_t* _multivar_smaj_phasepresent_batch_buf
+  /* "pgenlib.pyx":258
+ *         self._info_ptr[0].max_allele_ct = PglComputeMaxAlleleCt(self._info_ptr[0].allele_idx_offsets, nvariant)
  * 
  *     cdef set_sample_subset_internal(self, np.ndarray[np.uint32_t,mode="c",ndim=1] sample_subset):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_sample_ct = self._info_ptr[0].raw_sample_ct
  *         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)
  */
 
   /* function exit code */
@@ -3403,53 +3664,65 @@
   __Pyx_XDECREF(__pyx_v_sample_uidx);
   __Pyx_XDECREF(__pyx_v_next_uidx);
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":254
+/* "pgenlib.pyx":297
  * 
  * 
  *     def __cinit__(self, bytes filename, object raw_sample_ct = None,             # <<<<<<<<<<<<<<
- *                   object variant_ct = None, object sample_subset = None):
- *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
+ *                   object variant_ct = None, object sample_subset = None,
+ *                   object allele_idx_offsets = None):
  */
 
 /* Python wrapper */
 static int __pyx_pw_7pgenlib_10PgenReader_1__cinit__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static int __pyx_pw_7pgenlib_10PgenReader_1__cinit__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
   PyObject *__pyx_v_filename = 0;
   PyObject *__pyx_v_raw_sample_ct = 0;
   PyObject *__pyx_v_variant_ct = 0;
   PyObject *__pyx_v_sample_subset = 0;
+  PyObject *__pyx_v_allele_idx_offsets = 0;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   int __pyx_r;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("__cinit__ (wrapper)", 0);
   {
-    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_filename,&__pyx_n_s_raw_sample_ct,&__pyx_n_s_variant_ct,&__pyx_n_s_sample_subset,0};
-    PyObject* values[4] = {0,0,0,0};
+    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_filename,&__pyx_n_s_raw_sample_ct,&__pyx_n_s_variant_ct,&__pyx_n_s_sample_subset,&__pyx_n_s_allele_idx_offsets,0};
+    PyObject* values[5] = {0,0,0,0,0};
     values[1] = ((PyObject *)Py_None);
 
-    /* "pgenlib.pyx":255
+    /* "pgenlib.pyx":298
  * 
  *     def __cinit__(self, bytes filename, object raw_sample_ct = None,
- *                   object variant_ct = None, object sample_subset = None):             # <<<<<<<<<<<<<<
+ *                   object variant_ct = None, object sample_subset = None,             # <<<<<<<<<<<<<<
+ *                   object allele_idx_offsets = None):
  *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
- *         if not self._info_ptr:
  */
     values[2] = ((PyObject *)Py_None);
     values[3] = ((PyObject *)Py_None);
+
+    /* "pgenlib.pyx":299
+ *     def __cinit__(self, bytes filename, object raw_sample_ct = None,
+ *                   object variant_ct = None, object sample_subset = None,
+ *                   object allele_idx_offsets = None):             # <<<<<<<<<<<<<<
+ *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
+ *         if not self._info_ptr:
+ */
+    values[4] = ((PyObject *)Py_None);
     if (unlikely(__pyx_kwds)) {
       Py_ssize_t kw_args;
       const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
       switch (pos_args) {
+        case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
+        CYTHON_FALLTHROUGH;
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
@@ -3476,20 +3749,28 @@
         }
         CYTHON_FALLTHROUGH;
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_subset);
           if (value) { values[3] = value; kw_args--; }
         }
+        CYTHON_FALLTHROUGH;
+        case  4:
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx_offsets);
+          if (value) { values[4] = value; kw_args--; }
+        }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__cinit__") < 0)) __PYX_ERR(0, 254, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__cinit__") < 0)) __PYX_ERR(0, 297, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
+        case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
+        CYTHON_FALLTHROUGH;
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
@@ -3497,59 +3778,63 @@
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_filename = ((PyObject*)values[0]);
     __pyx_v_raw_sample_ct = values[1];
     __pyx_v_variant_ct = values[2];
     __pyx_v_sample_subset = values[3];
+    __pyx_v_allele_idx_offsets = values[4];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 1, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 254, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 1, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 297, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.__cinit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return -1;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_filename), (&PyBytes_Type), 1, "filename", 1))) __PYX_ERR(0, 254, __pyx_L1_error)
-  __pyx_r = __pyx_pf_7pgenlib_10PgenReader___cinit__(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_filename, __pyx_v_raw_sample_ct, __pyx_v_variant_ct, __pyx_v_sample_subset);
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_filename), (&PyBytes_Type), 1, "filename", 1))) __PYX_ERR(0, 297, __pyx_L1_error)
+  __pyx_r = __pyx_pf_7pgenlib_10PgenReader___cinit__(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_filename, __pyx_v_raw_sample_ct, __pyx_v_variant_ct, __pyx_v_sample_subset, __pyx_v_allele_idx_offsets);
 
-  /* "pgenlib.pyx":254
+  /* "pgenlib.pyx":297
  * 
  * 
  *     def __cinit__(self, bytes filename, object raw_sample_ct = None,             # <<<<<<<<<<<<<<
- *                   object variant_ct = None, object sample_subset = None):
- *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
+ *                   object variant_ct = None, object sample_subset = None,
+ *                   object allele_idx_offsets = None):
  */
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = -1;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-static int __pyx_pf_7pgenlib_10PgenReader___cinit__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyObject *__pyx_v_filename, PyObject *__pyx_v_raw_sample_ct, PyObject *__pyx_v_variant_ct, PyObject *__pyx_v_sample_subset) {
+static int __pyx_pf_7pgenlib_10PgenReader___cinit__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyObject *__pyx_v_filename, PyObject *__pyx_v_raw_sample_ct, PyObject *__pyx_v_variant_ct, PyObject *__pyx_v_sample_subset, PyObject *__pyx_v_allele_idx_offsets) {
   uint32_t __pyx_v_cur_sample_ct;
   uint32_t __pyx_v_cur_variant_ct;
   char const *__pyx_v_fname;
   plink2::PgenHeaderCtrl __pyx_v_header_ctrl;
   uintptr_t __pyx_v_pgfi_alloc_cacheline_ct;
   char __pyx_v_errstr_buf[plink2::kPglErrstrBufBlen];
+  uint32_t __pyx_v_file_variant_ct;
   uintptr_t __pyx_v_nonref_flags_byte_ct;
   uint32_t __pyx_v_file_sample_ct;
   unsigned char *__pyx_v_pgfi_alloc;
   uint32_t __pyx_v_max_vrec_width;
   uintptr_t __pyx_v_pgr_alloc_cacheline_ct;
   uintptr_t __pyx_v_pgr_alloc_main_byte_ct;
   uintptr_t __pyx_v_sample_subset_byte_ct;
   uintptr_t __pyx_v_cumulative_popcounts_byte_ct;
   uintptr_t __pyx_v_genovec_byte_ct;
+  uintptr_t __pyx_v_patch_01_vals_byte_ct;
+  uintptr_t __pyx_v_patch_10_vals_byte_ct;
   uintptr_t __pyx_v_dosage_main_byte_ct;
   unsigned char *__pyx_v_pgr_alloc;
   plink2::PglErr __pyx_v_reterr;
   unsigned char *__pyx_v_pgr_alloc_iter;
   int __pyx_r;
   __Pyx_RefNannyDeclarations
   int __pyx_t_1;
@@ -3560,1013 +3845,1199 @@
   PyObject *__pyx_t_6 = NULL;
   plink2::PglErr __pyx_t_7;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__cinit__", 0);
 
-  /* "pgenlib.pyx":256
- *     def __cinit__(self, bytes filename, object raw_sample_ct = None,
- *                   object variant_ct = None, object sample_subset = None):
+  /* "pgenlib.pyx":300
+ *                   object variant_ct = None, object sample_subset = None,
+ *                   object allele_idx_offsets = None):
  *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))             # <<<<<<<<<<<<<<
  *         if not self._info_ptr:
  *             raise MemoryError()
  */
   __pyx_v_self->_info_ptr = ((plink2::PgenFileInfo *)PyMem_Malloc((sizeof(plink2::PgenFileInfo))));
 
-  /* "pgenlib.pyx":257
- *                   object variant_ct = None, object sample_subset = None):
+  /* "pgenlib.pyx":301
+ *                   object allele_idx_offsets = None):
  *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
  *         if not self._info_ptr:             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         PreinitPgfi(self._info_ptr)
  */
   __pyx_t_1 = ((!(__pyx_v_self->_info_ptr != 0)) != 0);
   if (unlikely(__pyx_t_1)) {
 
-    /* "pgenlib.pyx":258
+    /* "pgenlib.pyx":302
  *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
  *         if not self._info_ptr:
  *             raise MemoryError()             # <<<<<<<<<<<<<<
  *         PreinitPgfi(self._info_ptr)
  *         # this depends on pgenlib_internal implementation.  could save
  */
-    PyErr_NoMemory(); __PYX_ERR(0, 258, __pyx_L1_error)
+    PyErr_NoMemory(); __PYX_ERR(0, 302, __pyx_L1_error)
 
-    /* "pgenlib.pyx":257
- *                   object variant_ct = None, object sample_subset = None):
+    /* "pgenlib.pyx":301
+ *                   object allele_idx_offsets = None):
  *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
  *         if not self._info_ptr:             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         PreinitPgfi(self._info_ptr)
  */
   }
 
-  /* "pgenlib.pyx":259
+  /* "pgenlib.pyx":303
  *         if not self._info_ptr:
  *             raise MemoryError()
  *         PreinitPgfi(self._info_ptr)             # <<<<<<<<<<<<<<
  *         # this depends on pgenlib_internal implementation.  could save
  *         # pgfi_alloc and pgr_alloc instead.
  */
   plink2::PreinitPgfi(__pyx_v_self->_info_ptr);
 
-  /* "pgenlib.pyx":262
+  /* "pgenlib.pyx":306
  *         # this depends on pgenlib_internal implementation.  could save
  *         # pgfi_alloc and pgr_alloc instead.
  *         self._info_ptr[0].vrtypes = NULL             # <<<<<<<<<<<<<<
  *         cdef uint32_t cur_sample_ct = 0xffffffffU
  *         if raw_sample_ct is not None:
  */
   (__pyx_v_self->_info_ptr[0]).vrtypes = NULL;
 
-  /* "pgenlib.pyx":263
+  /* "pgenlib.pyx":307
  *         # pgfi_alloc and pgr_alloc instead.
  *         self._info_ptr[0].vrtypes = NULL
  *         cdef uint32_t cur_sample_ct = 0xffffffffU             # <<<<<<<<<<<<<<
  *         if raw_sample_ct is not None:
  *             cur_sample_ct = raw_sample_ct
  */
   __pyx_v_cur_sample_ct = 0xffffffffU;
 
-  /* "pgenlib.pyx":264
+  /* "pgenlib.pyx":308
  *         self._info_ptr[0].vrtypes = NULL
  *         cdef uint32_t cur_sample_ct = 0xffffffffU
  *         if raw_sample_ct is not None:             # <<<<<<<<<<<<<<
  *             cur_sample_ct = raw_sample_ct
  *         cdef uint32_t cur_variant_ct = 0xffffffffU
  */
   __pyx_t_1 = (__pyx_v_raw_sample_ct != Py_None);
   __pyx_t_2 = (__pyx_t_1 != 0);
   if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":265
+    /* "pgenlib.pyx":309
  *         cdef uint32_t cur_sample_ct = 0xffffffffU
  *         if raw_sample_ct is not None:
  *             cur_sample_ct = raw_sample_ct             # <<<<<<<<<<<<<<
  *         cdef uint32_t cur_variant_ct = 0xffffffffU
  *         if variant_ct is not None:
  */
-    __pyx_t_3 = __Pyx_PyInt_As_uint32_t(__pyx_v_raw_sample_ct); if (unlikely((__pyx_t_3 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 265, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyInt_As_uint32_t(__pyx_v_raw_sample_ct); if (unlikely((__pyx_t_3 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 309, __pyx_L1_error)
     __pyx_v_cur_sample_ct = __pyx_t_3;
 
-    /* "pgenlib.pyx":264
+    /* "pgenlib.pyx":308
  *         self._info_ptr[0].vrtypes = NULL
  *         cdef uint32_t cur_sample_ct = 0xffffffffU
  *         if raw_sample_ct is not None:             # <<<<<<<<<<<<<<
  *             cur_sample_ct = raw_sample_ct
  *         cdef uint32_t cur_variant_ct = 0xffffffffU
  */
   }
 
-  /* "pgenlib.pyx":266
+  /* "pgenlib.pyx":310
  *         if raw_sample_ct is not None:
  *             cur_sample_ct = raw_sample_ct
  *         cdef uint32_t cur_variant_ct = 0xffffffffU             # <<<<<<<<<<<<<<
  *         if variant_ct is not None:
  *             cur_variant_ct = variant_ct
  */
   __pyx_v_cur_variant_ct = 0xffffffffU;
 
-  /* "pgenlib.pyx":267
+  /* "pgenlib.pyx":311
  *             cur_sample_ct = raw_sample_ct
  *         cdef uint32_t cur_variant_ct = 0xffffffffU
  *         if variant_ct is not None:             # <<<<<<<<<<<<<<
  *             cur_variant_ct = variant_ct
  *         cdef const char* fname = <const char*>filename
  */
   __pyx_t_2 = (__pyx_v_variant_ct != Py_None);
   __pyx_t_1 = (__pyx_t_2 != 0);
   if (__pyx_t_1) {
 
-    /* "pgenlib.pyx":268
+    /* "pgenlib.pyx":312
  *         cdef uint32_t cur_variant_ct = 0xffffffffU
  *         if variant_ct is not None:
  *             cur_variant_ct = variant_ct             # <<<<<<<<<<<<<<
  *         cdef const char* fname = <const char*>filename
  *         cdef PgenHeaderCtrl header_ctrl
  */
-    __pyx_t_3 = __Pyx_PyInt_As_uint32_t(__pyx_v_variant_ct); if (unlikely((__pyx_t_3 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 268, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyInt_As_uint32_t(__pyx_v_variant_ct); if (unlikely((__pyx_t_3 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 312, __pyx_L1_error)
     __pyx_v_cur_variant_ct = __pyx_t_3;
 
-    /* "pgenlib.pyx":267
+    /* "pgenlib.pyx":311
  *             cur_sample_ct = raw_sample_ct
  *         cdef uint32_t cur_variant_ct = 0xffffffffU
  *         if variant_ct is not None:             # <<<<<<<<<<<<<<
  *             cur_variant_ct = variant_ct
  *         cdef const char* fname = <const char*>filename
  */
   }
 
-  /* "pgenlib.pyx":269
+  /* "pgenlib.pyx":313
  *         if variant_ct is not None:
  *             cur_variant_ct = variant_ct
  *         cdef const char* fname = <const char*>filename             # <<<<<<<<<<<<<<
  *         cdef PgenHeaderCtrl header_ctrl
  *         cdef uintptr_t pgfi_alloc_cacheline_ct
  */
   if (unlikely(__pyx_v_filename == Py_None)) {
     PyErr_SetString(PyExc_TypeError, "expected bytes, NoneType found");
-    __PYX_ERR(0, 269, __pyx_L1_error)
+    __PYX_ERR(0, 313, __pyx_L1_error)
   }
-  __pyx_t_4 = __Pyx_PyBytes_AsString(__pyx_v_filename); if (unlikely((!__pyx_t_4) && PyErr_Occurred())) __PYX_ERR(0, 269, __pyx_L1_error)
+  __pyx_t_4 = __Pyx_PyBytes_AsString(__pyx_v_filename); if (unlikely((!__pyx_t_4) && PyErr_Occurred())) __PYX_ERR(0, 313, __pyx_L1_error)
   __pyx_v_fname = ((char const *)__pyx_t_4);
 
-  /* "pgenlib.pyx":273
+  /* "pgenlib.pyx":317
  *         cdef uintptr_t pgfi_alloc_cacheline_ct
  *         cdef char errstr_buf[kPglErrstrBufBlen]
  *         if PgfiInitPhase1(fname, NULL, cur_variant_ct, cur_sample_ct, &header_ctrl, self._info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError(errstr_buf[7:])
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
  */
   __pyx_t_1 = ((plink2::PgfiInitPhase1(__pyx_v_fname, NULL, __pyx_v_cur_variant_ct, __pyx_v_cur_sample_ct, (&__pyx_v_header_ctrl), __pyx_v_self->_info_ptr, (&__pyx_v_pgfi_alloc_cacheline_ct), __pyx_v_errstr_buf) != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_1)) {
 
-    /* "pgenlib.pyx":274
+    /* "pgenlib.pyx":318
  *         cdef char errstr_buf[kPglErrstrBufBlen]
  *         if PgfiInitPhase1(fname, NULL, cur_variant_ct, cur_sample_ct, &header_ctrl, self._info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != kPglRetSuccess:
  *             raise RuntimeError(errstr_buf[7:])             # <<<<<<<<<<<<<<
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts
- *         cdef uintptr_t nonref_flags_byte_ct = DivUp(self._info_ptr[0].raw_variant_ct, kBitsPerWord) * kBytesPerWord
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
+ *         if allele_idx_offsets is not None:
  */
-    __pyx_t_5 = __Pyx_PyBytes_FromString(((const char*)__pyx_v_errstr_buf) + 7); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 274, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyBytes_FromString(((const char*)__pyx_v_errstr_buf) + 7); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 318, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 274, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 318, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
     __Pyx_Raise(__pyx_t_6, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __PYX_ERR(0, 274, __pyx_L1_error)
+    __PYX_ERR(0, 318, __pyx_L1_error)
 
-    /* "pgenlib.pyx":273
+    /* "pgenlib.pyx":317
  *         cdef uintptr_t pgfi_alloc_cacheline_ct
  *         cdef char errstr_buf[kPglErrstrBufBlen]
  *         if PgfiInitPhase1(fname, NULL, cur_variant_ct, cur_sample_ct, &header_ctrl, self._info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError(errstr_buf[7:])
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
  */
   }
 
-  /* "pgenlib.pyx":275
+  /* "pgenlib.pyx":319
  *         if PgfiInitPhase1(fname, NULL, cur_variant_ct, cur_sample_ct, &header_ctrl, self._info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != kPglRetSuccess:
  *             raise RuntimeError(errstr_buf[7:])
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts             # <<<<<<<<<<<<<<
- *         cdef uintptr_t nonref_flags_byte_ct = DivUp(self._info_ptr[0].raw_variant_ct, kBitsPerWord) * kBytesPerWord
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
+ *         if allele_idx_offsets is not None:
+ *             self.set_allele_idx_offsets_internal(allele_idx_offsets)
+ */
+  __pyx_t_3 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
+  __pyx_v_file_variant_ct = __pyx_t_3;
+
+  /* "pgenlib.pyx":320
+ *             raise RuntimeError(errstr_buf[7:])
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
+ *         if allele_idx_offsets is not None:             # <<<<<<<<<<<<<<
+ *             self.set_allele_idx_offsets_internal(allele_idx_offsets)
+ *         else:
+ */
+  __pyx_t_1 = (__pyx_v_allele_idx_offsets != Py_None);
+  __pyx_t_2 = (__pyx_t_1 != 0);
+  if (__pyx_t_2) {
+
+    /* "pgenlib.pyx":321
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
+ *         if allele_idx_offsets is not None:
+ *             self.set_allele_idx_offsets_internal(allele_idx_offsets)             # <<<<<<<<<<<<<<
+ *         else:
+ *             if (header_ctrl & 0x30) != 0:
+ */
+    if (!(likely(((__pyx_v_allele_idx_offsets) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_allele_idx_offsets, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 321, __pyx_L1_error)
+    __pyx_t_6 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->set_allele_idx_offsets_internal(__pyx_v_self, ((PyArrayObject *)__pyx_v_allele_idx_offsets)); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 321, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_6);
+    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
+
+    /* "pgenlib.pyx":320
+ *             raise RuntimeError(errstr_buf[7:])
+ *         cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
+ *         if allele_idx_offsets is not None:             # <<<<<<<<<<<<<<
+ *             self.set_allele_idx_offsets_internal(allele_idx_offsets)
+ *         else:
+ */
+    goto __pyx_L7;
+  }
+
+  /* "pgenlib.pyx":323
+ *             self.set_allele_idx_offsets_internal(allele_idx_offsets)
+ *         else:
+ *             if (header_ctrl & 0x30) != 0:             # <<<<<<<<<<<<<<
+ *                 # Not tested yet, since C pgenlib doesn't yet have a mode which
+ *                 # embeds a copy of allele_idx_offsets in the .pgen.
+ */
+  /*else*/ {
+    __pyx_t_2 = (((__pyx_v_header_ctrl & 0x30) != 0) != 0);
+    if (__pyx_t_2) {
+
+      /* "pgenlib.pyx":326
+ *                 # Not tested yet, since C pgenlib doesn't yet have a mode which
+ *                 # embeds a copy of allele_idx_offsets in the .pgen.
+ *                 if cachealigned_malloc((file_variant_ct + 1) * sizeof(uintptr_t), &self._info_ptr[0].allele_idx_offsets):             # <<<<<<<<<<<<<<
+ *                     raise MemoryError()
+ *         cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord
+ */
+      __pyx_t_2 = (plink2::cachealigned_malloc(((__pyx_v_file_variant_ct + 1) * (sizeof(uintptr_t))), (&(__pyx_v_self->_info_ptr[0]).allele_idx_offsets)) != 0);
+      if (unlikely(__pyx_t_2)) {
+
+        /* "pgenlib.pyx":327
+ *                 # embeds a copy of allele_idx_offsets in the .pgen.
+ *                 if cachealigned_malloc((file_variant_ct + 1) * sizeof(uintptr_t), &self._info_ptr[0].allele_idx_offsets):
+ *                     raise MemoryError()             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord
  *         if (header_ctrl & 0xc0) == 0xc0:
  */
-  #ifndef CYTHON_WITHOUT_ASSERTIONS
-  if (unlikely(!Py_OptimizeFlag)) {
-    if (unlikely(!(((__pyx_v_header_ctrl & 0x30) == 0) != 0))) {
-      PyErr_SetNone(PyExc_AssertionError);
-      __PYX_ERR(0, 275, __pyx_L1_error)
+        PyErr_NoMemory(); __PYX_ERR(0, 327, __pyx_L1_error)
+
+        /* "pgenlib.pyx":326
+ *                 # Not tested yet, since C pgenlib doesn't yet have a mode which
+ *                 # embeds a copy of allele_idx_offsets in the .pgen.
+ *                 if cachealigned_malloc((file_variant_ct + 1) * sizeof(uintptr_t), &self._info_ptr[0].allele_idx_offsets):             # <<<<<<<<<<<<<<
+ *                     raise MemoryError()
+ *         cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord
+ */
+      }
+
+      /* "pgenlib.pyx":323
+ *             self.set_allele_idx_offsets_internal(allele_idx_offsets)
+ *         else:
+ *             if (header_ctrl & 0x30) != 0:             # <<<<<<<<<<<<<<
+ *                 # Not tested yet, since C pgenlib doesn't yet have a mode which
+ *                 # embeds a copy of allele_idx_offsets in the .pgen.
+ */
     }
   }
-  #endif
+  __pyx_L7:;
 
-  /* "pgenlib.pyx":276
- *             raise RuntimeError(errstr_buf[7:])
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts
- *         cdef uintptr_t nonref_flags_byte_ct = DivUp(self._info_ptr[0].raw_variant_ct, kBitsPerWord) * kBytesPerWord             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":328
+ *                 if cachealigned_malloc((file_variant_ct + 1) * sizeof(uintptr_t), &self._info_ptr[0].allele_idx_offsets):
+ *                     raise MemoryError()
+ *         cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord             # <<<<<<<<<<<<<<
  *         if (header_ctrl & 0xc0) == 0xc0:
  *             # explicit nonref_flags
  */
-  __pyx_v_nonref_flags_byte_ct = (plink2::DivUp((__pyx_v_self->_info_ptr[0]).raw_variant_ct, plink2::kBitsPerWord) * plink2::kBytesPerWord);
+  __pyx_v_nonref_flags_byte_ct = (plink2::DivUp(__pyx_v_file_variant_ct, plink2::kBitsPerWord) * plink2::kBytesPerWord);
 
-  /* "pgenlib.pyx":277
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts
- *         cdef uintptr_t nonref_flags_byte_ct = DivUp(self._info_ptr[0].raw_variant_ct, kBitsPerWord) * kBytesPerWord
+  /* "pgenlib.pyx":329
+ *                     raise MemoryError()
+ *         cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord
  *         if (header_ctrl & 0xc0) == 0xc0:             # <<<<<<<<<<<<<<
  *             # explicit nonref_flags
  *             # might want to modify PgfiInitPhase2 to let us not load this.
  */
-  __pyx_t_1 = (((__pyx_v_header_ctrl & 0xc0) == 0xc0) != 0);
-  if (__pyx_t_1) {
+  __pyx_t_2 = (((__pyx_v_header_ctrl & 0xc0) == 0xc0) != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":280
+    /* "pgenlib.pyx":332
  *             # explicit nonref_flags
  *             # might want to modify PgfiInitPhase2 to let us not load this.
  *             if cachealigned_malloc(nonref_flags_byte_ct, &self._info_ptr[0].nonref_flags):             # <<<<<<<<<<<<<<
  *                 raise MemoryError()
  *         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct
  */
-    __pyx_t_1 = (plink2::cachealigned_malloc(__pyx_v_nonref_flags_byte_ct, (&(__pyx_v_self->_info_ptr[0]).nonref_flags)) != 0);
-    if (unlikely(__pyx_t_1)) {
+    __pyx_t_2 = (plink2::cachealigned_malloc(__pyx_v_nonref_flags_byte_ct, (&(__pyx_v_self->_info_ptr[0]).nonref_flags)) != 0);
+    if (unlikely(__pyx_t_2)) {
 
-      /* "pgenlib.pyx":281
+      /* "pgenlib.pyx":333
  *             # might want to modify PgfiInitPhase2 to let us not load this.
  *             if cachealigned_malloc(nonref_flags_byte_ct, &self._info_ptr[0].nonref_flags):
  *                 raise MemoryError()             # <<<<<<<<<<<<<<
  *         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct
  *         assert file_sample_ct != 0
  */
-      PyErr_NoMemory(); __PYX_ERR(0, 281, __pyx_L1_error)
+      PyErr_NoMemory(); __PYX_ERR(0, 333, __pyx_L1_error)
 
-      /* "pgenlib.pyx":280
+      /* "pgenlib.pyx":332
  *             # explicit nonref_flags
  *             # might want to modify PgfiInitPhase2 to let us not load this.
  *             if cachealigned_malloc(nonref_flags_byte_ct, &self._info_ptr[0].nonref_flags):             # <<<<<<<<<<<<<<
  *                 raise MemoryError()
  *         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct
  */
     }
 
-    /* "pgenlib.pyx":277
- *         assert (header_ctrl & 0x30) == 0 # no alt allele counts
- *         cdef uintptr_t nonref_flags_byte_ct = DivUp(self._info_ptr[0].raw_variant_ct, kBitsPerWord) * kBytesPerWord
+    /* "pgenlib.pyx":329
+ *                     raise MemoryError()
+ *         cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord
  *         if (header_ctrl & 0xc0) == 0xc0:             # <<<<<<<<<<<<<<
  *             # explicit nonref_flags
  *             # might want to modify PgfiInitPhase2 to let us not load this.
  */
   }
 
-  /* "pgenlib.pyx":282
+  /* "pgenlib.pyx":334
  *             if cachealigned_malloc(nonref_flags_byte_ct, &self._info_ptr[0].nonref_flags):
  *                 raise MemoryError()
  *         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct             # <<<<<<<<<<<<<<
  *         assert file_sample_ct != 0
  *         cdef unsigned char* pgfi_alloc = NULL
  */
   __pyx_t_3 = (__pyx_v_self->_info_ptr[0]).raw_sample_ct;
   __pyx_v_file_sample_ct = __pyx_t_3;
 
-  /* "pgenlib.pyx":283
+  /* "pgenlib.pyx":335
  *                 raise MemoryError()
  *         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct
  *         assert file_sample_ct != 0             # <<<<<<<<<<<<<<
  *         cdef unsigned char* pgfi_alloc = NULL
  *         if pgfi_alloc_cacheline_ct != 0:
  */
   #ifndef CYTHON_WITHOUT_ASSERTIONS
   if (unlikely(!Py_OptimizeFlag)) {
     if (unlikely(!((__pyx_v_file_sample_ct != 0) != 0))) {
       PyErr_SetNone(PyExc_AssertionError);
-      __PYX_ERR(0, 283, __pyx_L1_error)
+      __PYX_ERR(0, 335, __pyx_L1_error)
     }
   }
   #endif
 
-  /* "pgenlib.pyx":284
+  /* "pgenlib.pyx":336
  *         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct
  *         assert file_sample_ct != 0
  *         cdef unsigned char* pgfi_alloc = NULL             # <<<<<<<<<<<<<<
  *         if pgfi_alloc_cacheline_ct != 0:
  *             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):
  */
   __pyx_v_pgfi_alloc = NULL;
 
-  /* "pgenlib.pyx":285
+  /* "pgenlib.pyx":337
  *         assert file_sample_ct != 0
  *         cdef unsigned char* pgfi_alloc = NULL
  *         if pgfi_alloc_cacheline_ct != 0:             # <<<<<<<<<<<<<<
  *             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):
  *                 raise MemoryError()
  */
-  __pyx_t_1 = ((__pyx_v_pgfi_alloc_cacheline_ct != 0) != 0);
-  if (__pyx_t_1) {
+  __pyx_t_2 = ((__pyx_v_pgfi_alloc_cacheline_ct != 0) != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":286
+    /* "pgenlib.pyx":338
  *         cdef unsigned char* pgfi_alloc = NULL
  *         if pgfi_alloc_cacheline_ct != 0:
  *             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):             # <<<<<<<<<<<<<<
  *                 raise MemoryError()
  *         cdef uint32_t max_vrec_width
  */
-    __pyx_t_1 = (plink2::cachealigned_malloc((__pyx_v_pgfi_alloc_cacheline_ct * plink2::kCacheline), (&__pyx_v_pgfi_alloc)) != 0);
-    if (unlikely(__pyx_t_1)) {
+    __pyx_t_2 = (plink2::cachealigned_malloc((__pyx_v_pgfi_alloc_cacheline_ct * plink2::kCacheline), (&__pyx_v_pgfi_alloc)) != 0);
+    if (unlikely(__pyx_t_2)) {
 
-      /* "pgenlib.pyx":287
+      /* "pgenlib.pyx":339
  *         if pgfi_alloc_cacheline_ct != 0:
  *             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):
  *                 raise MemoryError()             # <<<<<<<<<<<<<<
  *         cdef uint32_t max_vrec_width
  *         cdef uintptr_t pgr_alloc_cacheline_ct
  */
-      PyErr_NoMemory(); __PYX_ERR(0, 287, __pyx_L1_error)
+      PyErr_NoMemory(); __PYX_ERR(0, 339, __pyx_L1_error)
 
-      /* "pgenlib.pyx":286
+      /* "pgenlib.pyx":338
  *         cdef unsigned char* pgfi_alloc = NULL
  *         if pgfi_alloc_cacheline_ct != 0:
  *             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):             # <<<<<<<<<<<<<<
  *                 raise MemoryError()
  *         cdef uint32_t max_vrec_width
  */
     }
 
-    /* "pgenlib.pyx":285
+    /* "pgenlib.pyx":337
  *         assert file_sample_ct != 0
  *         cdef unsigned char* pgfi_alloc = NULL
  *         if pgfi_alloc_cacheline_ct != 0:             # <<<<<<<<<<<<<<
  *             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):
  *                 raise MemoryError()
  */
   }
 
-  /* "pgenlib.pyx":290
+  /* "pgenlib.pyx":342
  *         cdef uint32_t max_vrec_width
  *         cdef uintptr_t pgr_alloc_cacheline_ct
- *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, self._info_ptr[0].raw_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):             # <<<<<<<<<<<<<<
+ *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, file_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):             # <<<<<<<<<<<<<<
  *             if pgfi_alloc and not self._info_ptr[0].vrtypes:
  *                 aligned_free(pgfi_alloc)
  */
-  __pyx_t_7 = plink2::PgfiInitPhase2(__pyx_v_header_ctrl, 1, 0, 0, 0, (__pyx_v_self->_info_ptr[0]).raw_variant_ct, (&__pyx_v_max_vrec_width), __pyx_v_self->_info_ptr, __pyx_v_pgfi_alloc, (&__pyx_v_pgr_alloc_cacheline_ct), __pyx_v_errstr_buf);
+  __pyx_t_7 = plink2::PgfiInitPhase2(__pyx_v_header_ctrl, 1, 0, 0, 0, __pyx_v_file_variant_ct, (&__pyx_v_max_vrec_width), __pyx_v_self->_info_ptr, __pyx_v_pgfi_alloc, (&__pyx_v_pgr_alloc_cacheline_ct), __pyx_v_errstr_buf);
   if (__pyx_t_7) {
 
-    /* "pgenlib.pyx":291
+    /* "pgenlib.pyx":343
  *         cdef uintptr_t pgr_alloc_cacheline_ct
- *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, self._info_ptr[0].raw_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
+ *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, file_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
  *             if pgfi_alloc and not self._info_ptr[0].vrtypes:             # <<<<<<<<<<<<<<
  *                 aligned_free(pgfi_alloc)
  *             raise RuntimeError(errstr_buf[7:])
  */
-    __pyx_t_2 = (__pyx_v_pgfi_alloc != 0);
-    if (__pyx_t_2) {
+    __pyx_t_1 = (__pyx_v_pgfi_alloc != 0);
+    if (__pyx_t_1) {
     } else {
-      __pyx_t_1 = __pyx_t_2;
-      goto __pyx_L13_bool_binop_done;
+      __pyx_t_2 = __pyx_t_1;
+      goto __pyx_L16_bool_binop_done;
     }
-    __pyx_t_2 = ((!((__pyx_v_self->_info_ptr[0]).vrtypes != 0)) != 0);
-    __pyx_t_1 = __pyx_t_2;
-    __pyx_L13_bool_binop_done:;
-    if (__pyx_t_1) {
+    __pyx_t_1 = ((!((__pyx_v_self->_info_ptr[0]).vrtypes != 0)) != 0);
+    __pyx_t_2 = __pyx_t_1;
+    __pyx_L16_bool_binop_done:;
+    if (__pyx_t_2) {
 
-      /* "pgenlib.pyx":292
- *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, self._info_ptr[0].raw_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
+      /* "pgenlib.pyx":344
+ *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, file_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
  *             if pgfi_alloc and not self._info_ptr[0].vrtypes:
  *                 aligned_free(pgfi_alloc)             # <<<<<<<<<<<<<<
  *             raise RuntimeError(errstr_buf[7:])
  *         if self._info_ptr[0].gflags & kfPgenGlobalMultiallelicHardcallFound:
  */
       plink2::aligned_free(__pyx_v_pgfi_alloc);
 
-      /* "pgenlib.pyx":291
+      /* "pgenlib.pyx":343
  *         cdef uintptr_t pgr_alloc_cacheline_ct
- *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, self._info_ptr[0].raw_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
+ *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, file_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
  *             if pgfi_alloc and not self._info_ptr[0].vrtypes:             # <<<<<<<<<<<<<<
  *                 aligned_free(pgfi_alloc)
  *             raise RuntimeError(errstr_buf[7:])
  */
     }
 
-    /* "pgenlib.pyx":293
+    /* "pgenlib.pyx":345
  *             if pgfi_alloc and not self._info_ptr[0].vrtypes:
  *                 aligned_free(pgfi_alloc)
  *             raise RuntimeError(errstr_buf[7:])             # <<<<<<<<<<<<<<
  *         if self._info_ptr[0].gflags & kfPgenGlobalMultiallelicHardcallFound:
- *             # todo: support this by wrapping pvar loader the same way as the R
+ *             # todo: support easy initialization of allele_idx_offsets by
  */
-    __pyx_t_6 = __Pyx_PyBytes_FromString(((const char*)__pyx_v_errstr_buf) + 7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 293, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyBytes_FromString(((const char*)__pyx_v_errstr_buf) + 7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 345, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
-    __pyx_t_5 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 293, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 345, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_5, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __PYX_ERR(0, 293, __pyx_L1_error)
+    __PYX_ERR(0, 345, __pyx_L1_error)
 
-    /* "pgenlib.pyx":290
+    /* "pgenlib.pyx":342
  *         cdef uint32_t max_vrec_width
  *         cdef uintptr_t pgr_alloc_cacheline_ct
- *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, self._info_ptr[0].raw_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):             # <<<<<<<<<<<<<<
+ *         if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, file_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):             # <<<<<<<<<<<<<<
  *             if pgfi_alloc and not self._info_ptr[0].vrtypes:
  *                 aligned_free(pgfi_alloc)
  */
   }
 
-  /* "pgenlib.pyx":294
+  /* "pgenlib.pyx":346
  *                 aligned_free(pgfi_alloc)
  *             raise RuntimeError(errstr_buf[7:])
  *         if self._info_ptr[0].gflags & kfPgenGlobalMultiallelicHardcallFound:             # <<<<<<<<<<<<<<
- *             # todo: support this by wrapping pvar loader the same way as the R
- *             # interface
+ *             # todo: support easy initialization of allele_idx_offsets by
+ *             # wrapping pvar loader the same way as the R interface
  */
-  __pyx_t_1 = (((__pyx_v_self->_info_ptr[0]).gflags & plink2::kfPgenGlobalMultiallelicHardcallFound) != 0);
-  if (unlikely(__pyx_t_1)) {
+  __pyx_t_2 = (((__pyx_v_self->_info_ptr[0]).gflags & plink2::kfPgenGlobalMultiallelicHardcallFound) != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":297
- *             # todo: support this by wrapping pvar loader the same way as the R
- *             # interface
- *             raise RuntimeError("Multiallelic + phase/dosage datasets not supported yet")             # <<<<<<<<<<<<<<
+    /* "pgenlib.pyx":349
+ *             # todo: support easy initialization of allele_idx_offsets by
+ *             # wrapping pvar loader the same way as the R interface
+ *             if self._info_ptr[0].allele_idx_offsets == NULL:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("PgenReader: multiallelic variants present, but allele_idx_offsets not provided")
+ * 
+ */
+    __pyx_t_2 = (((__pyx_v_self->_info_ptr[0]).allele_idx_offsets == NULL) != 0);
+    if (unlikely(__pyx_t_2)) {
+
+      /* "pgenlib.pyx":350
+ *             # wrapping pvar loader the same way as the R interface
+ *             if self._info_ptr[0].allele_idx_offsets == NULL:
+ *                 raise RuntimeError("PgenReader: multiallelic variants present, but allele_idx_offsets not provided")             # <<<<<<<<<<<<<<
  * 
  *         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))
  */
-    __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__3, NULL); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 297, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_5);
-    __Pyx_Raise(__pyx_t_5, 0, 0, 0);
-    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __PYX_ERR(0, 297, __pyx_L1_error)
+      __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__3, NULL); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 350, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_5);
+      __Pyx_Raise(__pyx_t_5, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
+      __PYX_ERR(0, 350, __pyx_L1_error)
+
+      /* "pgenlib.pyx":349
+ *             # todo: support easy initialization of allele_idx_offsets by
+ *             # wrapping pvar loader the same way as the R interface
+ *             if self._info_ptr[0].allele_idx_offsets == NULL:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("PgenReader: multiallelic variants present, but allele_idx_offsets not provided")
+ * 
+ */
+    }
 
-    /* "pgenlib.pyx":294
+    /* "pgenlib.pyx":346
  *                 aligned_free(pgfi_alloc)
  *             raise RuntimeError(errstr_buf[7:])
  *         if self._info_ptr[0].gflags & kfPgenGlobalMultiallelicHardcallFound:             # <<<<<<<<<<<<<<
- *             # todo: support this by wrapping pvar loader the same way as the R
- *             # interface
+ *             # todo: support easy initialization of allele_idx_offsets by
+ *             # wrapping pvar loader the same way as the R interface
  */
   }
 
-  /* "pgenlib.pyx":299
- *             raise RuntimeError("Multiallelic + phase/dosage datasets not supported yet")
+  /* "pgenlib.pyx":352
+ *                 raise RuntimeError("PgenReader: multiallelic variants present, but allele_idx_offsets not provided")
  * 
  *         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))             # <<<<<<<<<<<<<<
  *         if not self._state_ptr:
  *             raise MemoryError()
  */
   __pyx_v_self->_state_ptr = ((struct plink2::PgenReaderStruct *)PyMem_Malloc((sizeof(struct plink2::PgenReaderStruct))));
 
-  /* "pgenlib.pyx":300
+  /* "pgenlib.pyx":353
  * 
  *         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))
  *         if not self._state_ptr:             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         PreinitPgr(self._state_ptr)
  */
-  __pyx_t_1 = ((!(__pyx_v_self->_state_ptr != 0)) != 0);
-  if (unlikely(__pyx_t_1)) {
+  __pyx_t_2 = ((!(__pyx_v_self->_state_ptr != 0)) != 0);
+  if (unlikely(__pyx_t_2)) {
 
-    /* "pgenlib.pyx":301
+    /* "pgenlib.pyx":354
  *         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))
  *         if not self._state_ptr:
  *             raise MemoryError()             # <<<<<<<<<<<<<<
  *         PreinitPgr(self._state_ptr)
  *         PgrSetFreadBuf(NULL, self._state_ptr)
  */
-    PyErr_NoMemory(); __PYX_ERR(0, 301, __pyx_L1_error)
+    PyErr_NoMemory(); __PYX_ERR(0, 354, __pyx_L1_error)
 
-    /* "pgenlib.pyx":300
+    /* "pgenlib.pyx":353
  * 
  *         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))
  *         if not self._state_ptr:             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         PreinitPgr(self._state_ptr)
  */
   }
 
-  /* "pgenlib.pyx":302
+  /* "pgenlib.pyx":355
  *         if not self._state_ptr:
  *             raise MemoryError()
  *         PreinitPgr(self._state_ptr)             # <<<<<<<<<<<<<<
  *         PgrSetFreadBuf(NULL, self._state_ptr)
  *         cdef uintptr_t pgr_alloc_main_byte_ct = pgr_alloc_cacheline_ct * kCacheline
  */
   plink2::PreinitPgr(__pyx_v_self->_state_ptr);
 
-  /* "pgenlib.pyx":303
+  /* "pgenlib.pyx":356
  *             raise MemoryError()
  *         PreinitPgr(self._state_ptr)
  *         PgrSetFreadBuf(NULL, self._state_ptr)             # <<<<<<<<<<<<<<
  *         cdef uintptr_t pgr_alloc_main_byte_ct = pgr_alloc_cacheline_ct * kCacheline
  *         cdef uintptr_t sample_subset_byte_ct = DivUp(file_sample_ct, kBitsPerVec) * kBytesPerVec
  */
   plink2::PgrSetFreadBuf(NULL, __pyx_v_self->_state_ptr);
 
-  /* "pgenlib.pyx":304
+  /* "pgenlib.pyx":357
  *         PreinitPgr(self._state_ptr)
  *         PgrSetFreadBuf(NULL, self._state_ptr)
  *         cdef uintptr_t pgr_alloc_main_byte_ct = pgr_alloc_cacheline_ct * kCacheline             # <<<<<<<<<<<<<<
  *         cdef uintptr_t sample_subset_byte_ct = DivUp(file_sample_ct, kBitsPerVec) * kBytesPerVec
  *         cdef uintptr_t cumulative_popcounts_byte_ct = DivUp(file_sample_ct, kBitsPerWord * kInt32PerVec) * kBytesPerVec
  */
   __pyx_v_pgr_alloc_main_byte_ct = (__pyx_v_pgr_alloc_cacheline_ct * plink2::kCacheline);
 
-  /* "pgenlib.pyx":305
+  /* "pgenlib.pyx":358
  *         PgrSetFreadBuf(NULL, self._state_ptr)
  *         cdef uintptr_t pgr_alloc_main_byte_ct = pgr_alloc_cacheline_ct * kCacheline
  *         cdef uintptr_t sample_subset_byte_ct = DivUp(file_sample_ct, kBitsPerVec) * kBytesPerVec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t cumulative_popcounts_byte_ct = DivUp(file_sample_ct, kBitsPerWord * kInt32PerVec) * kBytesPerVec
  *         cdef uintptr_t genovec_byte_ct = DivUp(file_sample_ct, kNypsPerVec) * kBytesPerVec
  */
   __pyx_v_sample_subset_byte_ct = (plink2::DivUp(__pyx_v_file_sample_ct, plink2::kBitsPerVec) * plink2::kBytesPerVec);
 
-  /* "pgenlib.pyx":306
+  /* "pgenlib.pyx":359
  *         cdef uintptr_t pgr_alloc_main_byte_ct = pgr_alloc_cacheline_ct * kCacheline
  *         cdef uintptr_t sample_subset_byte_ct = DivUp(file_sample_ct, kBitsPerVec) * kBytesPerVec
  *         cdef uintptr_t cumulative_popcounts_byte_ct = DivUp(file_sample_ct, kBitsPerWord * kInt32PerVec) * kBytesPerVec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t genovec_byte_ct = DivUp(file_sample_ct, kNypsPerVec) * kBytesPerVec
- *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
+ *         cdef uintptr_t patch_01_vals_byte_ct = RoundUpPow2(file_sample_ct * sizeof(AlleleCode), kBytesPerVec)
  */
   __pyx_v_cumulative_popcounts_byte_ct = (plink2::DivUp(__pyx_v_file_sample_ct, (plink2::kBitsPerWord * plink2::kInt32PerVec)) * plink2::kBytesPerVec);
 
-  /* "pgenlib.pyx":307
+  /* "pgenlib.pyx":360
  *         cdef uintptr_t sample_subset_byte_ct = DivUp(file_sample_ct, kBitsPerVec) * kBytesPerVec
  *         cdef uintptr_t cumulative_popcounts_byte_ct = DivUp(file_sample_ct, kBitsPerWord * kInt32PerVec) * kBytesPerVec
  *         cdef uintptr_t genovec_byte_ct = DivUp(file_sample_ct, kNypsPerVec) * kBytesPerVec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
- *         cdef unsigned char* pgr_alloc
+ *         cdef uintptr_t patch_01_vals_byte_ct = RoundUpPow2(file_sample_ct * sizeof(AlleleCode), kBytesPerVec)
+ *         cdef uintptr_t patch_10_vals_byte_ct = RoundUpPow2(file_sample_ct * 2 * sizeof(AlleleCode), kBytesPerVec)
  */
   __pyx_v_genovec_byte_ct = (plink2::DivUp(__pyx_v_file_sample_ct, plink2::kNypsPerVec) * plink2::kBytesPerVec);
 
-  /* "pgenlib.pyx":308
+  /* "pgenlib.pyx":361
  *         cdef uintptr_t cumulative_popcounts_byte_ct = DivUp(file_sample_ct, kBitsPerWord * kInt32PerVec) * kBytesPerVec
  *         cdef uintptr_t genovec_byte_ct = DivUp(file_sample_ct, kNypsPerVec) * kBytesPerVec
+ *         cdef uintptr_t patch_01_vals_byte_ct = RoundUpPow2(file_sample_ct * sizeof(AlleleCode), kBytesPerVec)             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t patch_10_vals_byte_ct = RoundUpPow2(file_sample_ct * 2 * sizeof(AlleleCode), kBytesPerVec)
+ *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
+ */
+  __pyx_v_patch_01_vals_byte_ct = plink2::RoundUpPow2((__pyx_v_file_sample_ct * (sizeof(plink2::AlleleCode))), plink2::kBytesPerVec);
+
+  /* "pgenlib.pyx":362
+ *         cdef uintptr_t genovec_byte_ct = DivUp(file_sample_ct, kNypsPerVec) * kBytesPerVec
+ *         cdef uintptr_t patch_01_vals_byte_ct = RoundUpPow2(file_sample_ct * sizeof(AlleleCode), kBytesPerVec)
+ *         cdef uintptr_t patch_10_vals_byte_ct = RoundUpPow2(file_sample_ct * 2 * sizeof(AlleleCode), kBytesPerVec)             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
+ *         cdef unsigned char* pgr_alloc
+ */
+  __pyx_v_patch_10_vals_byte_ct = plink2::RoundUpPow2(((__pyx_v_file_sample_ct * 2) * (sizeof(plink2::AlleleCode))), plink2::kBytesPerVec);
+
+  /* "pgenlib.pyx":363
+ *         cdef uintptr_t patch_01_vals_byte_ct = RoundUpPow2(file_sample_ct * sizeof(AlleleCode), kBytesPerVec)
+ *         cdef uintptr_t patch_10_vals_byte_ct = RoundUpPow2(file_sample_ct * 2 * sizeof(AlleleCode), kBytesPerVec)
  *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec             # <<<<<<<<<<<<<<
  *         cdef unsigned char* pgr_alloc
- *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 5) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
+ *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 7) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + patch_01_vals_byte_ct + patch_10_vals_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
  */
   __pyx_v_dosage_main_byte_ct = (plink2::DivUp(__pyx_v_file_sample_ct, (2 * plink2::kInt32PerVec)) * plink2::kBytesPerVec);
 
-  /* "pgenlib.pyx":310
+  /* "pgenlib.pyx":365
  *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
  *         cdef unsigned char* pgr_alloc
- *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 5) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):             # <<<<<<<<<<<<<<
+ *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 7) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + patch_01_vals_byte_ct + patch_10_vals_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  */
-  __pyx_t_1 = (plink2::cachealigned_malloc(((((((__pyx_v_pgr_alloc_main_byte_ct + (((2 * plink2::kPglNypTransposeBatch) + 5) * __pyx_v_sample_subset_byte_ct)) + __pyx_v_cumulative_popcounts_byte_ct) + ((1 + plink2::kPglNypTransposeBatch) * __pyx_v_genovec_byte_ct)) + __pyx_v_dosage_main_byte_ct) + plink2::kPglBitTransposeBufbytes) + (4 * __Pyx_div_long((plink2::kPglNypTransposeBatch * plink2::kPglNypTransposeBatch), 8))), (&__pyx_v_pgr_alloc)) != 0);
-  if (unlikely(__pyx_t_1)) {
+  __pyx_t_2 = (plink2::cachealigned_malloc(((((((((__pyx_v_pgr_alloc_main_byte_ct + (((2 * plink2::kPglNypTransposeBatch) + 7) * __pyx_v_sample_subset_byte_ct)) + __pyx_v_cumulative_popcounts_byte_ct) + ((1 + plink2::kPglNypTransposeBatch) * __pyx_v_genovec_byte_ct)) + __pyx_v_patch_01_vals_byte_ct) + __pyx_v_patch_10_vals_byte_ct) + __pyx_v_dosage_main_byte_ct) + plink2::kPglBitTransposeBufbytes) + (4 * __Pyx_div_long((plink2::kPglNypTransposeBatch * plink2::kPglNypTransposeBatch), 8))), (&__pyx_v_pgr_alloc)) != 0);
+  if (unlikely(__pyx_t_2)) {
 
-    /* "pgenlib.pyx":311
+    /* "pgenlib.pyx":366
  *         cdef unsigned char* pgr_alloc
- *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 5) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
+ *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 7) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + patch_01_vals_byte_ct + patch_10_vals_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
  *             raise MemoryError()             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  *         if reterr != kPglRetSuccess:
  */
-    PyErr_NoMemory(); __PYX_ERR(0, 311, __pyx_L1_error)
+    PyErr_NoMemory(); __PYX_ERR(0, 366, __pyx_L1_error)
 
-    /* "pgenlib.pyx":310
+    /* "pgenlib.pyx":365
  *         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
  *         cdef unsigned char* pgr_alloc
- *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 5) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):             # <<<<<<<<<<<<<<
+ *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 7) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + patch_01_vals_byte_ct + patch_10_vals_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  */
   }
 
-  /* "pgenlib.pyx":312
- *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 5) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
+  /* "pgenlib.pyx":367
+ *         if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 7) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + patch_01_vals_byte_ct + patch_10_vals_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
  *             raise MemoryError()
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             if not PgrGetFreadBuf(self._state_ptr):
  */
   __pyx_v_reterr = plink2::PgrInit(__pyx_v_fname, __pyx_v_max_vrec_width, __pyx_v_self->_info_ptr, __pyx_v_self->_state_ptr, __pyx_v_pgr_alloc);
 
-  /* "pgenlib.pyx":313
+  /* "pgenlib.pyx":368
  *             raise MemoryError()
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             if not PgrGetFreadBuf(self._state_ptr):
  *                 aligned_free(pgr_alloc)
  */
-  __pyx_t_1 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-  if (__pyx_t_1) {
+  __pyx_t_2 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":314
+    /* "pgenlib.pyx":369
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  *         if reterr != kPglRetSuccess:
  *             if not PgrGetFreadBuf(self._state_ptr):             # <<<<<<<<<<<<<<
  *                 aligned_free(pgr_alloc)
  *             raise RuntimeError("PgrInit() error " + str(reterr))
  */
-    __pyx_t_1 = ((!(plink2::PgrGetFreadBuf(__pyx_v_self->_state_ptr) != 0)) != 0);
-    if (__pyx_t_1) {
+    __pyx_t_2 = ((!(plink2::PgrGetFreadBuf(__pyx_v_self->_state_ptr) != 0)) != 0);
+    if (__pyx_t_2) {
 
-      /* "pgenlib.pyx":315
+      /* "pgenlib.pyx":370
  *         if reterr != kPglRetSuccess:
  *             if not PgrGetFreadBuf(self._state_ptr):
  *                 aligned_free(pgr_alloc)             # <<<<<<<<<<<<<<
  *             raise RuntimeError("PgrInit() error " + str(reterr))
  *         cdef unsigned char* pgr_alloc_iter = &(pgr_alloc[pgr_alloc_main_byte_ct])
  */
       plink2::aligned_free(__pyx_v_pgr_alloc);
 
-      /* "pgenlib.pyx":314
+      /* "pgenlib.pyx":369
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  *         if reterr != kPglRetSuccess:
  *             if not PgrGetFreadBuf(self._state_ptr):             # <<<<<<<<<<<<<<
  *                 aligned_free(pgr_alloc)
  *             raise RuntimeError("PgrInit() error " + str(reterr))
  */
     }
 
-    /* "pgenlib.pyx":316
+    /* "pgenlib.pyx":371
  *             if not PgrGetFreadBuf(self._state_ptr):
  *                 aligned_free(pgr_alloc)
  *             raise RuntimeError("PgrInit() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         cdef unsigned char* pgr_alloc_iter = &(pgr_alloc[pgr_alloc_main_byte_ct])
  *         self._subset_include_vec = <uintptr_t*>pgr_alloc_iter
  */
-    __pyx_t_5 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 316, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 371, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 316, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 371, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_kp_u_PgrInit_error, __pyx_t_6); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 316, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_kp_u_PgrInit_error, __pyx_t_6); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 371, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 316, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 371, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
     __Pyx_Raise(__pyx_t_6, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __PYX_ERR(0, 316, __pyx_L1_error)
+    __PYX_ERR(0, 371, __pyx_L1_error)
 
-    /* "pgenlib.pyx":313
+    /* "pgenlib.pyx":368
  *             raise MemoryError()
  *         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             if not PgrGetFreadBuf(self._state_ptr):
  *                 aligned_free(pgr_alloc)
  */
   }
 
-  /* "pgenlib.pyx":317
+  /* "pgenlib.pyx":372
  *                 aligned_free(pgr_alloc)
  *             raise RuntimeError("PgrInit() error " + str(reterr))
  *         cdef unsigned char* pgr_alloc_iter = &(pgr_alloc[pgr_alloc_main_byte_ct])             # <<<<<<<<<<<<<<
  *         self._subset_include_vec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc[__pyx_v_pgr_alloc_main_byte_ct]));
 
-  /* "pgenlib.pyx":318
+  /* "pgenlib.pyx":373
  *             raise RuntimeError("PgrInit() error " + str(reterr))
  *         cdef unsigned char* pgr_alloc_iter = &(pgr_alloc[pgr_alloc_main_byte_ct])
  *         self._subset_include_vec = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  *         self._subset_include_interleaved_vec = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_subset_include_vec = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":319
+  /* "pgenlib.pyx":374
  *         cdef unsigned char* pgr_alloc_iter = &(pgr_alloc[pgr_alloc_main_byte_ct])
  *         self._subset_include_vec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
  *         self._subset_include_interleaved_vec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
 
-  /* "pgenlib.pyx":320
+  /* "pgenlib.pyx":375
  *         self._subset_include_vec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  *         self._subset_include_interleaved_vec = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  * 
  */
   __pyx_v_self->_subset_include_interleaved_vec = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":321
+  /* "pgenlib.pyx":376
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  *         self._subset_include_interleaved_vec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
  * 
  *         # assumes kWordsPerVec <= 2
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
 
-  /* "pgenlib.pyx":324
+  /* "pgenlib.pyx":379
  * 
  *         # assumes kWordsPerVec <= 2
  *         self._subset_include_interleaved_vec[-1] = 0             # <<<<<<<<<<<<<<
  * 
  *         self._subset_cumulative_popcounts = <uint32_t*>pgr_alloc_iter
  */
   (__pyx_v_self->_subset_include_interleaved_vec[-1L]) = 0;
 
-  /* "pgenlib.pyx":326
+  /* "pgenlib.pyx":381
  *         self._subset_include_interleaved_vec[-1] = 0
  * 
  *         self._subset_cumulative_popcounts = <uint32_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[cumulative_popcounts_byte_ct])
- *         self._genovec = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.genovec = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_subset_cumulative_popcounts = ((uint32_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":327
+  /* "pgenlib.pyx":382
  * 
  *         self._subset_cumulative_popcounts = <uint32_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[cumulative_popcounts_byte_ct])             # <<<<<<<<<<<<<<
- *         self._genovec = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.genovec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_cumulative_popcounts_byte_ct]));
 
-  /* "pgenlib.pyx":328
+  /* "pgenlib.pyx":383
  *         self._subset_cumulative_popcounts = <uint32_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[cumulative_popcounts_byte_ct])
- *         self._genovec = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         self._pgv.genovec = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct])
- *         self._phasepresent = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.patch_01_set = <uintptr_t*>pgr_alloc_iter
  */
-  __pyx_v_self->_genovec = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+  __pyx_v_self->_pgv.genovec = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":329
+  /* "pgenlib.pyx":384
  *         pgr_alloc_iter = &(pgr_alloc_iter[cumulative_popcounts_byte_ct])
- *         self._genovec = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.genovec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct])             # <<<<<<<<<<<<<<
- *         self._phasepresent = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.patch_01_set = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_genovec_byte_ct]));
 
-  /* "pgenlib.pyx":330
- *         self._genovec = <uintptr_t*>pgr_alloc_iter
+  /* "pgenlib.pyx":385
+ *         self._pgv.genovec = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct])
- *         self._phasepresent = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         self._pgv.patch_01_set = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._phaseinfo = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.patch_01_vals = <AlleleCode*>pgr_alloc_iter
  */
-  __pyx_v_self->_phasepresent = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+  __pyx_v_self->_pgv.patch_01_set = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":331
+  /* "pgenlib.pyx":386
  *         pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct])
- *         self._phasepresent = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.patch_01_set = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
- *         self._phaseinfo = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.patch_01_vals = <AlleleCode*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_01_vals_byte_ct])
+ */
+  __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
+
+  /* "pgenlib.pyx":387
+ *         self._pgv.patch_01_set = <uintptr_t*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.patch_01_vals = <AlleleCode*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_01_vals_byte_ct])
+ *         self._pgv.patch_10_set = <uintptr_t*>pgr_alloc_iter
+ */
+  __pyx_v_self->_pgv.patch_01_vals = ((plink2::AlleleCode *)__pyx_v_pgr_alloc_iter);
+
+  /* "pgenlib.pyx":388
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.patch_01_vals = <AlleleCode*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_01_vals_byte_ct])             # <<<<<<<<<<<<<<
+ *         self._pgv.patch_10_set = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  */
+  __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_patch_01_vals_byte_ct]));
+
+  /* "pgenlib.pyx":389
+ *         self._pgv.patch_01_vals = <AlleleCode*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_01_vals_byte_ct])
+ *         self._pgv.patch_10_set = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.patch_10_vals = <AlleleCode*>pgr_alloc_iter
+ */
+  __pyx_v_self->_pgv.patch_10_set = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+
+  /* "pgenlib.pyx":390
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_01_vals_byte_ct])
+ *         self._pgv.patch_10_set = <uintptr_t*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
+ *         self._pgv.patch_10_vals = <AlleleCode*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_10_vals_byte_ct])
+ */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
 
-  /* "pgenlib.pyx":332
- *         self._phasepresent = <uintptr_t*>pgr_alloc_iter
+  /* "pgenlib.pyx":391
+ *         self._pgv.patch_10_set = <uintptr_t*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.patch_10_vals = <AlleleCode*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_10_vals_byte_ct])
+ *         self._pgv.phasepresent = <uintptr_t*>pgr_alloc_iter
+ */
+  __pyx_v_self->_pgv.patch_10_vals = ((plink2::AlleleCode *)__pyx_v_pgr_alloc_iter);
+
+  /* "pgenlib.pyx":392
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._phaseinfo = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         self._pgv.patch_10_vals = <AlleleCode*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_10_vals_byte_ct])             # <<<<<<<<<<<<<<
+ *         self._pgv.phasepresent = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._dosage_present = <uintptr_t*>pgr_alloc_iter
  */
-  __pyx_v_self->_phaseinfo = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+  __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_patch_10_vals_byte_ct]));
 
-  /* "pgenlib.pyx":333
+  /* "pgenlib.pyx":393
+ *         self._pgv.patch_10_vals = <AlleleCode*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_10_vals_byte_ct])
+ *         self._pgv.phasepresent = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._phaseinfo = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.phaseinfo = <uintptr_t*>pgr_alloc_iter
+ */
+  __pyx_v_self->_pgv.phasepresent = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+
+  /* "pgenlib.pyx":394
+ *         pgr_alloc_iter = &(pgr_alloc_iter[patch_10_vals_byte_ct])
+ *         self._pgv.phasepresent = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
- *         self._dosage_present = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.phaseinfo = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
 
-  /* "pgenlib.pyx":334
- *         self._phaseinfo = <uintptr_t*>pgr_alloc_iter
+  /* "pgenlib.pyx":395
+ *         self._pgv.phasepresent = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._dosage_present = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         self._pgv.phaseinfo = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._dosage_main = <uint16_t*>pgr_alloc_iter
+ *         self._pgv.dosage_present = <uintptr_t*>pgr_alloc_iter
  */
-  __pyx_v_self->_dosage_present = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+  __pyx_v_self->_pgv.phaseinfo = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":335
+  /* "pgenlib.pyx":396
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.phaseinfo = <uintptr_t*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
+ *         self._pgv.dosage_present = <uintptr_t*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ */
+  __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
+
+  /* "pgenlib.pyx":397
+ *         self._pgv.phaseinfo = <uintptr_t*>pgr_alloc_iter
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.dosage_present = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._dosage_present = <uintptr_t*>pgr_alloc_iter
+ *         self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter
+ */
+  __pyx_v_self->_pgv.dosage_present = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
+
+  /* "pgenlib.pyx":398
+ *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+ *         self._pgv.dosage_present = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])             # <<<<<<<<<<<<<<
- *         self._dosage_main = <uint16_t*>pgr_alloc_iter
+ *         self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_sample_subset_byte_ct]));
 
-  /* "pgenlib.pyx":336
- *         self._dosage_present = <uintptr_t*>pgr_alloc_iter
+  /* "pgenlib.pyx":399
+ *         self._pgv.dosage_present = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._dosage_main = <uint16_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
+ *         self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])
  *         if sample_subset is not None:
  */
-  __pyx_v_self->_dosage_main = ((uint16_t *)__pyx_v_pgr_alloc_iter);
+  __pyx_v_self->_pgv.dosage_main = ((uint16_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":337
+  /* "pgenlib.pyx":400
  *         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
- *         self._dosage_main = <uint16_t*>pgr_alloc_iter
+ *         self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])             # <<<<<<<<<<<<<<
  *         if sample_subset is not None:
  *             self.set_sample_subset_internal(sample_subset)
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__pyx_v_dosage_main_byte_ct]));
 
-  /* "pgenlib.pyx":338
- *         self._dosage_main = <uint16_t*>pgr_alloc_iter
+  /* "pgenlib.pyx":401
+ *         self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])
  *         if sample_subset is not None:             # <<<<<<<<<<<<<<
  *             self.set_sample_subset_internal(sample_subset)
  *         else:
  */
-  __pyx_t_1 = (__pyx_v_sample_subset != Py_None);
-  __pyx_t_2 = (__pyx_t_1 != 0);
-  if (__pyx_t_2) {
+  __pyx_t_2 = (__pyx_v_sample_subset != Py_None);
+  __pyx_t_1 = (__pyx_t_2 != 0);
+  if (__pyx_t_1) {
 
-    /* "pgenlib.pyx":339
+    /* "pgenlib.pyx":402
  *         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])
  *         if sample_subset is not None:
  *             self.set_sample_subset_internal(sample_subset)             # <<<<<<<<<<<<<<
  *         else:
  *             self._subset_size = file_sample_ct
  */
-    if (!(likely(((__pyx_v_sample_subset) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_sample_subset, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 339, __pyx_L1_error)
-    __pyx_t_6 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->set_sample_subset_internal(__pyx_v_self, ((PyArrayObject *)__pyx_v_sample_subset)); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 339, __pyx_L1_error)
+    if (!(likely(((__pyx_v_sample_subset) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_sample_subset, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 402, __pyx_L1_error)
+    __pyx_t_6 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->set_sample_subset_internal(__pyx_v_self, ((PyArrayObject *)__pyx_v_sample_subset)); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 402, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
 
-    /* "pgenlib.pyx":338
- *         self._dosage_main = <uint16_t*>pgr_alloc_iter
+    /* "pgenlib.pyx":401
+ *         self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])
  *         if sample_subset is not None:             # <<<<<<<<<<<<<<
  *             self.set_sample_subset_internal(sample_subset)
  *         else:
  */
-    goto __pyx_L20;
+    goto __pyx_L24;
   }
 
-  /* "pgenlib.pyx":341
+  /* "pgenlib.pyx":404
  *             self.set_sample_subset_internal(sample_subset)
  *         else:
  *             self._subset_size = file_sample_ct             # <<<<<<<<<<<<<<
  *         self._transpose_batch_buf = <VecW*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglBitTransposeBufbytes])
  */
   /*else*/ {
     __pyx_v_self->_subset_size = __pyx_v_file_sample_ct;
   }
-  __pyx_L20:;
+  __pyx_L24:;
 
-  /* "pgenlib.pyx":342
+  /* "pgenlib.pyx":405
  *         else:
  *             self._subset_size = file_sample_ct
  *         self._transpose_batch_buf = <VecW*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglBitTransposeBufbytes])
  *         self._multivar_vmaj_geno_buf = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_transpose_batch_buf = ((plink2::VecW *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":343
+  /* "pgenlib.pyx":406
  *             self._subset_size = file_sample_ct
  *         self._transpose_batch_buf = <VecW*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglBitTransposeBufbytes])             # <<<<<<<<<<<<<<
  *         self._multivar_vmaj_geno_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * genovec_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[plink2::kPglBitTransposeBufbytes]));
 
-  /* "pgenlib.pyx":344
+  /* "pgenlib.pyx":407
  *         self._transpose_batch_buf = <VecW*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglBitTransposeBufbytes])
  *         self._multivar_vmaj_geno_buf = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * genovec_byte_ct])
  *         self._multivar_vmaj_phasepresent_buf = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_multivar_vmaj_geno_buf = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":345
+  /* "pgenlib.pyx":408
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglBitTransposeBufbytes])
  *         self._multivar_vmaj_geno_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * genovec_byte_ct])             # <<<<<<<<<<<<<<
  *         self._multivar_vmaj_phasepresent_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[(plink2::kPglNypTransposeBatch * __pyx_v_genovec_byte_ct)]));
 
-  /* "pgenlib.pyx":346
+  /* "pgenlib.pyx":409
  *         self._multivar_vmaj_geno_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * genovec_byte_ct])
  *         self._multivar_vmaj_phasepresent_buf = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  *         self._multivar_vmaj_phaseinfo_buf = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_multivar_vmaj_phasepresent_buf = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":347
+  /* "pgenlib.pyx":410
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * genovec_byte_ct])
  *         self._multivar_vmaj_phasepresent_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])             # <<<<<<<<<<<<<<
  *         self._multivar_vmaj_phaseinfo_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[(plink2::kPglNypTransposeBatch * __pyx_v_sample_subset_byte_ct)]));
 
-  /* "pgenlib.pyx":348
+  /* "pgenlib.pyx":411
  *         self._multivar_vmaj_phasepresent_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  *         self._multivar_vmaj_phaseinfo_buf = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  *         self._multivar_smaj_geno_batch_buf = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_multivar_vmaj_phaseinfo_buf = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":349
+  /* "pgenlib.pyx":412
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  *         self._multivar_vmaj_phaseinfo_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])             # <<<<<<<<<<<<<<
  *         self._multivar_smaj_geno_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 4])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[(plink2::kPglNypTransposeBatch * __pyx_v_sample_subset_byte_ct)]));
 
-  /* "pgenlib.pyx":350
+  /* "pgenlib.pyx":413
  *         self._multivar_vmaj_phaseinfo_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  *         self._multivar_smaj_geno_batch_buf = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 4])
  *         self._multivar_smaj_phaseinfo_batch_buf = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_multivar_smaj_geno_batch_buf = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":351
+  /* "pgenlib.pyx":414
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * sample_subset_byte_ct])
  *         self._multivar_smaj_geno_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 4])             # <<<<<<<<<<<<<<
  *         self._multivar_smaj_phaseinfo_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__Pyx_div_long((plink2::kPglNypTransposeBatch * plink2::kPglNypTransposeBatch), 4)]));
 
-  /* "pgenlib.pyx":352
+  /* "pgenlib.pyx":415
  *         self._multivar_smaj_geno_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 4])
  *         self._multivar_smaj_phaseinfo_batch_buf = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])
  *         self._multivar_smaj_phasepresent_batch_buf = <uintptr_t*>pgr_alloc_iter
  */
   __pyx_v_self->_multivar_smaj_phaseinfo_batch_buf = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":353
+  /* "pgenlib.pyx":416
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 4])
  *         self._multivar_smaj_phaseinfo_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])             # <<<<<<<<<<<<<<
  *         self._multivar_smaj_phasepresent_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         # pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])
  */
   __pyx_v_pgr_alloc_iter = (&(__pyx_v_pgr_alloc_iter[__Pyx_div_long((plink2::kPglNypTransposeBatch * plink2::kPglNypTransposeBatch), 8)]));
 
-  /* "pgenlib.pyx":354
+  /* "pgenlib.pyx":417
  *         self._multivar_smaj_phaseinfo_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])
  *         self._multivar_smaj_phasepresent_batch_buf = <uintptr_t*>pgr_alloc_iter             # <<<<<<<<<<<<<<
  *         # pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])
  *         return
  */
   __pyx_v_self->_multivar_smaj_phasepresent_batch_buf = ((uintptr_t *)__pyx_v_pgr_alloc_iter);
 
-  /* "pgenlib.pyx":356
+  /* "pgenlib.pyx":419
  *         self._multivar_smaj_phasepresent_batch_buf = <uintptr_t*>pgr_alloc_iter
  *         # pgr_alloc_iter = &(pgr_alloc_iter[kPglNypTransposeBatch * kPglNypTransposeBatch // 8])
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __pyx_r = 0;
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":254
+  /* "pgenlib.pyx":297
  * 
  * 
  *     def __cinit__(self, bytes filename, object raw_sample_ct = None,             # <<<<<<<<<<<<<<
- *                   object variant_ct = None, object sample_subset = None):
- *         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
+ *                   object variant_ct = None, object sample_subset = None,
+ *                   object allele_idx_offsets = None):
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_5);
   __Pyx_XDECREF(__pyx_t_6);
   __Pyx_AddTraceback("pgenlib.PgenReader.__cinit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = -1;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":359
+/* "pgenlib.pyx":422
  * 
  * 
  *     cpdef __enter__(self):             # <<<<<<<<<<<<<<
  *         return self
  * 
  */
 
@@ -4587,15 +5058,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_enter); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 359, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_enter); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 422, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_3__enter__)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -4604,15 +5075,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 359, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 422, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -4625,27 +5096,27 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":360
+  /* "pgenlib.pyx":423
  * 
  *     cpdef __enter__(self):
  *         return self             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __Pyx_INCREF(((PyObject *)__pyx_v_self));
   __pyx_r = ((PyObject *)__pyx_v_self);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":359
+  /* "pgenlib.pyx":422
  * 
  * 
  *     cpdef __enter__(self):             # <<<<<<<<<<<<<<
  *         return self
  * 
  */
 
@@ -4681,15 +5152,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__enter__", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader___enter__(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 359, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader___enter__(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 422, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -4698,15 +5169,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":363
+/* "pgenlib.pyx":426
  * 
  * 
  *     cpdef get_raw_sample_ct(self):             # <<<<<<<<<<<<<<
  *         return self._info_ptr[0].raw_sample_ct
  * 
  */
 
@@ -4727,15 +5198,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_get_raw_sample_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 363, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_get_raw_sample_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 426, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_5get_raw_sample_ct)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -4744,15 +5215,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 363, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 426, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -4765,29 +5236,29 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":364
+  /* "pgenlib.pyx":427
  * 
  *     cpdef get_raw_sample_ct(self):
  *         return self._info_ptr[0].raw_sample_ct             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_sample_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 364, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_sample_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 427, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":363
+  /* "pgenlib.pyx":426
  * 
  * 
  *     cpdef get_raw_sample_ct(self):             # <<<<<<<<<<<<<<
  *         return self._info_ptr[0].raw_sample_ct
  * 
  */
 
@@ -4823,15 +5294,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("get_raw_sample_ct", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_get_raw_sample_ct(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 363, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_get_raw_sample_ct(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 426, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -4840,15 +5311,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":367
+/* "pgenlib.pyx":430
  * 
  * 
  *     cpdef get_variant_ct(self):             # <<<<<<<<<<<<<<
  *         return self._info_ptr[0].raw_variant_ct
  * 
  */
 
@@ -4869,15 +5340,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_get_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 367, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_get_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 430, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_7get_variant_ct)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -4886,15 +5357,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 367, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 430, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -4907,29 +5378,29 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":368
+  /* "pgenlib.pyx":431
  * 
  *     cpdef get_variant_ct(self):
  *         return self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 368, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 431, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":367
+  /* "pgenlib.pyx":430
  * 
  * 
  *     cpdef get_variant_ct(self):             # <<<<<<<<<<<<<<
  *         return self._info_ptr[0].raw_variant_ct
  * 
  */
 
@@ -4965,15 +5436,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("get_variant_ct", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_get_variant_ct(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 367, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_get_variant_ct(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 430, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -4982,15 +5453,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":371
+/* "pgenlib.pyx":434
  * 
  * 
  *     cpdef hardcall_phase_present(self):             # <<<<<<<<<<<<<<
  *         return ((self._info_ptr[0].gflags & kfPgenGlobalHardcallPhasePresent) != 0)
  * 
  */
 
@@ -5011,15 +5482,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hardcall_phase_present); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 371, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hardcall_phase_present); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 434, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_9hardcall_phase_present)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -5028,15 +5499,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 371, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 434, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -5049,29 +5520,29 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":372
+  /* "pgenlib.pyx":435
  * 
  *     cpdef hardcall_phase_present(self):
  *         return ((self._info_ptr[0].gflags & kfPgenGlobalHardcallPhasePresent) != 0)             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __Pyx_PyBool_FromLong((((__pyx_v_self->_info_ptr[0]).gflags & plink2::kfPgenGlobalHardcallPhasePresent) != 0)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 372, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyBool_FromLong((((__pyx_v_self->_info_ptr[0]).gflags & plink2::kfPgenGlobalHardcallPhasePresent) != 0)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 435, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":371
+  /* "pgenlib.pyx":434
  * 
  * 
  *     cpdef hardcall_phase_present(self):             # <<<<<<<<<<<<<<
  *         return ((self._info_ptr[0].gflags & kfPgenGlobalHardcallPhasePresent) != 0)
  * 
  */
 
@@ -5107,15 +5578,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("hardcall_phase_present", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_hardcall_phase_present(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 371, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_hardcall_phase_present(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 434, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -5124,19 +5595,19 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":375
+/* "pgenlib.pyx":438
  * 
  * 
  *     cpdef read(self, uint32_t variant_idx, np.ndarray geno_int_out, uint32_t allele_idx = 1):             # <<<<<<<<<<<<<<
- *         # for full genotype info for multiallelic variants, use read_phased()
+ *         # for full genotype info for multiallelic variants, use read_alleles()
  *         # instead
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_11read(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_geno_int_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read *__pyx_optional_args) {
   uint32_t __pyx_v_allele_idx = ((uint32_t)1);
   uint32_t __pyx_v_subset_size;
@@ -5172,21 +5643,21 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 375, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_11read)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 375, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 438, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 375, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 438, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_5 = __pyx_t_1; __pyx_t_6 = NULL;
         __pyx_t_7 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_5))) {
           __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_5);
           if (likely(__pyx_t_6)) {
@@ -5196,47 +5667,47 @@
             __Pyx_DECREF_SET(__pyx_t_5, function);
             __pyx_t_7 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[4] = {__pyx_t_6, __pyx_t_3, ((PyObject *)__pyx_v_geno_int_out), __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 375, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 438, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[4] = {__pyx_t_6, __pyx_t_3, ((PyObject *)__pyx_v_geno_int_out), __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 375, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 438, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         {
-          __pyx_t_8 = PyTuple_New(3+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 375, __pyx_L1_error)
+          __pyx_t_8 = PyTuple_New(3+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 438, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_8);
           if (__pyx_t_6) {
             __Pyx_GIVEREF(__pyx_t_6); PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_6); __pyx_t_6 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_8, 0+__pyx_t_7, __pyx_t_3);
           __Pyx_INCREF(((PyObject *)__pyx_v_geno_int_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_geno_int_out));
           PyTuple_SET_ITEM(__pyx_t_8, 1+__pyx_t_7, ((PyObject *)__pyx_v_geno_int_out));
           __Pyx_GIVEREF(__pyx_t_4);
           PyTuple_SET_ITEM(__pyx_t_8, 2+__pyx_t_7, __pyx_t_4);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 375, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 438, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         }
         __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -5251,407 +5722,407 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":381
+  /* "pgenlib.pyx":444
  *         # when this is too much bounds-checking, caller should be using
  *         # read_range() or read_list()
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.ndim != 1:
  */
   __pyx_t_9 = ((__pyx_v_variant_idx >= (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":382
+    /* "pgenlib.pyx":445
  *         # read_range() or read_list()
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         if geno_int_out.ndim != 1:
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_variant_idx_too_large, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_variant_idx_too_large, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 382, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 445, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 382, __pyx_L1_error)
+    __PYX_ERR(0, 445, __pyx_L1_error)
 
-    /* "pgenlib.pyx":381
+    /* "pgenlib.pyx":444
  *         # when this is too much bounds-checking, caller should be using
  *         # read_range() or read_list()
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.ndim != 1:
  */
   }
 
-  /* "pgenlib.pyx":383
+  /* "pgenlib.pyx":446
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.ndim != 1:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  */
   __pyx_t_9 = ((__pyx_v_geno_int_out->nd != 1) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":384
+    /* "pgenlib.pyx":447
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.ndim != 1:
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         if geno_int_out.shape[0] < subset_size:
  */
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__4, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 384, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__4, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 447, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 384, __pyx_L1_error)
+    __PYX_ERR(0, 447, __pyx_L1_error)
 
-    /* "pgenlib.pyx":383
+    /* "pgenlib.pyx":446
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.ndim != 1:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  */
   }
 
-  /* "pgenlib.pyx":385
+  /* "pgenlib.pyx":448
  *         if geno_int_out.ndim != 1:
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         if geno_int_out.shape[0] < subset_size:
  *             raise RuntimeError("read() geno_int_out is too small (" + str(geno_int_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  */
   __pyx_t_10 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_10;
 
-  /* "pgenlib.pyx":386
+  /* "pgenlib.pyx":449
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  *         if geno_int_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() geno_int_out is too small (" + str(geno_int_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   __pyx_t_9 = (((__pyx_v_geno_int_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":387
+    /* "pgenlib.pyx":450
  *         cdef uint32_t subset_size = self._subset_size
  *         if geno_int_out.shape[0] < subset_size:
  *             raise RuntimeError("read() geno_int_out is too small (" + str(geno_int_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")             # <<<<<<<<<<<<<<
  * 
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
-    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_geno_int_out_is_too_small, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_geno_int_out_is_too_small, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_5, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_5, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 387, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 450, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 387, __pyx_L1_error)
+    __PYX_ERR(0, 450, __pyx_L1_error)
 
-    /* "pgenlib.pyx":386
+    /* "pgenlib.pyx":449
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  *         if geno_int_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() geno_int_out is too small (" + str(geno_int_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   }
 
-  /* "pgenlib.pyx":389
+  /* "pgenlib.pyx":452
  *             raise RuntimeError("read() geno_int_out is too small (" + str(geno_int_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = PgrGet1(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, genovec)
  *         if reterr != kPglRetSuccess:
  */
-  __pyx_t_11 = __pyx_v_self->_genovec;
+  __pyx_t_11 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_11;
 
-  /* "pgenlib.pyx":390
+  /* "pgenlib.pyx":453
  * 
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef PglErr reterr = PgrGet1(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, genovec)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read() error " + str(reterr))
  */
   __pyx_v_reterr = plink2::PgrGet1(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_self->_state_ptr, __pyx_v_genovec);
 
-  /* "pgenlib.pyx":391
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":454
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef PglErr reterr = PgrGet1(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, genovec)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() error " + str(reterr))
  *         cdef int8_t* data8_ptr
  */
   __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":392
+    /* "pgenlib.pyx":455
  *         cdef PglErr reterr = PgrGet1(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, genovec)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         cdef int8_t* data8_ptr
  *         cdef int32_t* data32_ptr
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 392, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 392, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 455, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 392, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 392, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 455, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 392, __pyx_L1_error)
+    __PYX_ERR(0, 455, __pyx_L1_error)
 
-    /* "pgenlib.pyx":391
- *         cdef uintptr_t* genovec = self._genovec
+    /* "pgenlib.pyx":454
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef PglErr reterr = PgrGet1(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, genovec)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read() error " + str(reterr))
  *         cdef int8_t* data8_ptr
  */
   }
 
-  /* "pgenlib.pyx":396
+  /* "pgenlib.pyx":459
  *         cdef int32_t* data32_ptr
  *         cdef int64_t* data64_ptr
  *         if geno_int_out.dtype == np.int8:             # <<<<<<<<<<<<<<
  *             data8_ptr = <int8_t*>geno_int_out.data
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 396, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 459, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 396, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 459, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_int8); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 396, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_int8); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 459, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_1 = PyObject_RichCompare(__pyx_t_2, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 396, __pyx_L1_error)
+  __pyx_t_1 = PyObject_RichCompare(__pyx_t_2, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 459, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 396, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 459, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   if (__pyx_t_9) {
 
-    /* "pgenlib.pyx":397
+    /* "pgenlib.pyx":460
  *         cdef int64_t* data64_ptr
  *         if geno_int_out.dtype == np.int8:
  *             data8_ptr = <int8_t*>geno_int_out.data             # <<<<<<<<<<<<<<
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)
  *         elif geno_int_out.dtype == np.int32:
  */
     __pyx_v_data8_ptr = ((int8_t *)__pyx_v_geno_int_out->data);
 
-    /* "pgenlib.pyx":398
+    /* "pgenlib.pyx":461
  *         if geno_int_out.dtype == np.int8:
  *             data8_ptr = <int8_t*>geno_int_out.data
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)             # <<<<<<<<<<<<<<
  *         elif geno_int_out.dtype == np.int32:
  *             data32_ptr = <int32_t*>geno_int_out.data
  */
     plink2::GenoarrToBytesMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data8_ptr);
 
-    /* "pgenlib.pyx":396
+    /* "pgenlib.pyx":459
  *         cdef int32_t* data32_ptr
  *         cdef int64_t* data64_ptr
  *         if geno_int_out.dtype == np.int8:             # <<<<<<<<<<<<<<
  *             data8_ptr = <int8_t*>geno_int_out.data
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)
  */
     goto __pyx_L7;
   }
 
-  /* "pgenlib.pyx":399
+  /* "pgenlib.pyx":462
  *             data8_ptr = <int8_t*>geno_int_out.data
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)
  *         elif geno_int_out.dtype == np.int32:             # <<<<<<<<<<<<<<
  *             data32_ptr = <int32_t*>geno_int_out.data
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 399, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 462, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 399, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 462, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_int32); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 399, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_int32); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 462, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_5 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 399, __pyx_L1_error)
+  __pyx_t_5 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 462, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 399, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 462, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
   if (__pyx_t_9) {
 
-    /* "pgenlib.pyx":400
+    /* "pgenlib.pyx":463
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)
  *         elif geno_int_out.dtype == np.int32:
  *             data32_ptr = <int32_t*>geno_int_out.data             # <<<<<<<<<<<<<<
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)
  *         elif geno_int_out.dtype == np.int64:
  */
     __pyx_v_data32_ptr = ((int32_t *)__pyx_v_geno_int_out->data);
 
-    /* "pgenlib.pyx":401
+    /* "pgenlib.pyx":464
  *         elif geno_int_out.dtype == np.int32:
  *             data32_ptr = <int32_t*>geno_int_out.data
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)             # <<<<<<<<<<<<<<
  *         elif geno_int_out.dtype == np.int64:
  *             data64_ptr = <int64_t*>geno_int_out.data
  */
     plink2::GenoarrToInt32sMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data32_ptr);
 
-    /* "pgenlib.pyx":399
+    /* "pgenlib.pyx":462
  *             data8_ptr = <int8_t*>geno_int_out.data
  *             GenoarrToBytesMinus9(genovec, subset_size, data8_ptr)
  *         elif geno_int_out.dtype == np.int32:             # <<<<<<<<<<<<<<
  *             data32_ptr = <int32_t*>geno_int_out.data
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)
  */
     goto __pyx_L7;
   }
 
-  /* "pgenlib.pyx":402
+  /* "pgenlib.pyx":465
  *             data32_ptr = <int32_t*>geno_int_out.data
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)
  *         elif geno_int_out.dtype == np.int64:             # <<<<<<<<<<<<<<
  *             data64_ptr = <int64_t*>geno_int_out.data
  *             GenoarrToInt64sMinus9(genovec, subset_size, data64_ptr)
  */
-  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 402, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 465, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 402, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 465, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_int64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 402, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_int64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 465, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_5, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 402, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_5, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 465, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 402, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 465, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (likely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":403
+    /* "pgenlib.pyx":466
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)
  *         elif geno_int_out.dtype == np.int64:
  *             data64_ptr = <int64_t*>geno_int_out.data             # <<<<<<<<<<<<<<
  *             GenoarrToInt64sMinus9(genovec, subset_size, data64_ptr)
  *         else:
  */
     __pyx_v_data64_ptr = ((int64_t *)__pyx_v_geno_int_out->data);
 
-    /* "pgenlib.pyx":404
+    /* "pgenlib.pyx":467
  *         elif geno_int_out.dtype == np.int64:
  *             data64_ptr = <int64_t*>geno_int_out.data
  *             GenoarrToInt64sMinus9(genovec, subset_size, data64_ptr)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid read() geno_int_out array element type (int8, int32, or int64 expected).")
  */
     plink2::GenoarrToInt64sMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data64_ptr);
 
-    /* "pgenlib.pyx":402
+    /* "pgenlib.pyx":465
  *             data32_ptr = <int32_t*>geno_int_out.data
  *             GenoarrToInt32sMinus9(genovec, subset_size, data32_ptr)
  *         elif geno_int_out.dtype == np.int64:             # <<<<<<<<<<<<<<
  *             data64_ptr = <int64_t*>geno_int_out.data
  *             GenoarrToInt64sMinus9(genovec, subset_size, data64_ptr)
  */
     goto __pyx_L7;
   }
 
-  /* "pgenlib.pyx":406
+  /* "pgenlib.pyx":469
  *             GenoarrToInt64sMinus9(genovec, subset_size, data64_ptr)
  *         else:
  *             raise RuntimeError("Invalid read() geno_int_out array element type (int8, int32, or int64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 406, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 469, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 406, __pyx_L1_error)
+    __PYX_ERR(0, 469, __pyx_L1_error)
   }
   __pyx_L7:;
 
-  /* "pgenlib.pyx":407
+  /* "pgenlib.pyx":470
  *         else:
  *             raise RuntimeError("Invalid read() geno_int_out array element type (int8, int32, or int64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":375
+  /* "pgenlib.pyx":438
  * 
  * 
  *     cpdef read(self, uint32_t variant_idx, np.ndarray geno_int_out, uint32_t allele_idx = 1):             # <<<<<<<<<<<<<<
- *         # for full genotype info for multiallelic variants, use read_phased()
+ *         # for full genotype info for multiallelic variants, use read_alleles()
  *         # instead
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
   __Pyx_XDECREF(__pyx_t_2);
@@ -5701,53 +6172,53 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_geno_int_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read", 0, 2, 3, 1); __PYX_ERR(0, 375, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read", 0, 2, 3, 1); __PYX_ERR(0, 438, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[2] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read") < 0)) __PYX_ERR(0, 375, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read") < 0)) __PYX_ERR(0, 438, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 375, __pyx_L3_error)
+    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 438, __pyx_L3_error)
     __pyx_v_geno_int_out = ((PyArrayObject *)values[1]);
     if (values[2]) {
-      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 375, __pyx_L3_error)
+      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 438, __pyx_L3_error)
     } else {
       __pyx_v_allele_idx = ((uint32_t)1);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 375, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 438, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int_out), __pyx_ptype_5numpy_ndarray, 1, "geno_int_out", 0))) __PYX_ERR(0, 375, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int_out), __pyx_ptype_5numpy_ndarray, 1, "geno_int_out", 0))) __PYX_ERR(0, 438, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_10read(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx, __pyx_v_geno_int_out, __pyx_v_allele_idx);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -5763,15 +6234,15 @@
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("read", 0);
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_geno_int_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 375, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_geno_int_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -5780,15 +6251,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":410
+/* "pgenlib.pyx":473
  * 
  * 
  *     cpdef read_dosages(self, uint32_t variant_idx, np.ndarray floatarr_out, uint32_t allele_idx = 1):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -5826,21 +6297,21 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_dosages); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 410, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_dosages); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 473, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_13read_dosages)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 410, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 473, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 410, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 473, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_5 = __pyx_t_1; __pyx_t_6 = NULL;
         __pyx_t_7 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_5))) {
           __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_5);
           if (likely(__pyx_t_6)) {
@@ -5850,47 +6321,47 @@
             __Pyx_DECREF_SET(__pyx_t_5, function);
             __pyx_t_7 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[4] = {__pyx_t_6, __pyx_t_3, ((PyObject *)__pyx_v_floatarr_out), __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 410, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 473, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[4] = {__pyx_t_6, __pyx_t_3, ((PyObject *)__pyx_v_floatarr_out), __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 410, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 3+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 473, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         {
-          __pyx_t_8 = PyTuple_New(3+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 410, __pyx_L1_error)
+          __pyx_t_8 = PyTuple_New(3+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 473, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_8);
           if (__pyx_t_6) {
             __Pyx_GIVEREF(__pyx_t_6); PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_6); __pyx_t_6 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_8, 0+__pyx_t_7, __pyx_t_3);
           __Pyx_INCREF(((PyObject *)__pyx_v_floatarr_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_floatarr_out));
           PyTuple_SET_ITEM(__pyx_t_8, 1+__pyx_t_7, ((PyObject *)__pyx_v_floatarr_out));
           __Pyx_GIVEREF(__pyx_t_4);
           PyTuple_SET_ITEM(__pyx_t_8, 2+__pyx_t_7, __pyx_t_4);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 410, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 473, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         }
         __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -5905,344 +6376,344 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":411
+  /* "pgenlib.pyx":474
  * 
  *     cpdef read_dosages(self, uint32_t variant_idx, np.ndarray floatarr_out, uint32_t allele_idx = 1):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.ndim != 1:
  */
   __pyx_t_9 = ((__pyx_v_variant_idx >= (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":412
+    /* "pgenlib.pyx":475
  *     cpdef read_dosages(self, uint32_t variant_idx, np.ndarray floatarr_out, uint32_t allele_idx = 1):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         if floatarr_out.ndim != 1:
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_variant_idx_too_lar, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_variant_idx_too_lar, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 412, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 475, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 412, __pyx_L1_error)
+    __PYX_ERR(0, 475, __pyx_L1_error)
 
-    /* "pgenlib.pyx":411
+    /* "pgenlib.pyx":474
  * 
  *     cpdef read_dosages(self, uint32_t variant_idx, np.ndarray floatarr_out, uint32_t allele_idx = 1):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.ndim != 1:
  */
   }
 
-  /* "pgenlib.pyx":413
+  /* "pgenlib.pyx":476
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.ndim != 1:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  */
   __pyx_t_9 = ((__pyx_v_floatarr_out->nd != 1) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":414
+    /* "pgenlib.pyx":477
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.ndim != 1:
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         if floatarr_out.shape[0] < subset_size:
  */
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__7, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 414, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__7, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 477, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 414, __pyx_L1_error)
+    __PYX_ERR(0, 477, __pyx_L1_error)
 
-    /* "pgenlib.pyx":413
+    /* "pgenlib.pyx":476
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.ndim != 1:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  */
   }
 
-  /* "pgenlib.pyx":415
+  /* "pgenlib.pyx":478
  *         if floatarr_out.ndim != 1:
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         if floatarr_out.shape[0] < subset_size:
  *             raise RuntimeError("read_dosages() floatarr_out is too small (" + str(floatarr_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  */
   __pyx_t_10 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_10;
 
-  /* "pgenlib.pyx":416
+  /* "pgenlib.pyx":479
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  *         if floatarr_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() floatarr_out is too small (" + str(floatarr_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   __pyx_t_9 = (((__pyx_v_floatarr_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":417
+    /* "pgenlib.pyx":480
  *         cdef uint32_t subset_size = self._subset_size
  *         if floatarr_out.shape[0] < subset_size:
  *             raise RuntimeError("read_dosages() floatarr_out is too small (" + str(floatarr_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")             # <<<<<<<<<<<<<<
  * 
  *         cdef uint32_t dosage_ct
  */
-    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_floatarr_out_is_too, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_floatarr_out_is_too, __pyx_t_5); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_5, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_5, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 417, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 480, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 417, __pyx_L1_error)
+    __PYX_ERR(0, 480, __pyx_L1_error)
 
-    /* "pgenlib.pyx":416
+    /* "pgenlib.pyx":479
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
  *         cdef uint32_t subset_size = self._subset_size
  *         if floatarr_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() floatarr_out is too small (" + str(floatarr_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   }
 
-  /* "pgenlib.pyx":420
+  /* "pgenlib.pyx":483
  * 
  *         cdef uint32_t dosage_ct
- *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._genovec, self._dosage_present, self._dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
+ *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_dosages() error " + str(reterr))
  */
-  __pyx_v_reterr = plink2::PgrGet1D(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_self->_state_ptr, __pyx_v_self->_genovec, __pyx_v_self->_dosage_present, __pyx_v_self->_dosage_main, (&__pyx_v_dosage_ct));
+  __pyx_v_reterr = plink2::PgrGet1D(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_self->_state_ptr, __pyx_v_self->_pgv.genovec, __pyx_v_self->_pgv.dosage_present, __pyx_v_self->_pgv.dosage_main, (&__pyx_v_dosage_ct));
 
-  /* "pgenlib.pyx":421
+  /* "pgenlib.pyx":484
  *         cdef uint32_t dosage_ct
- *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._genovec, self._dosage_present, self._dosage_main, &dosage_ct)
+ *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, &dosage_ct)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() error " + str(reterr))
  *         cdef float* data32_ptr
  */
   __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":422
- *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._genovec, self._dosage_present, self._dosage_main, &dosage_ct)
+    /* "pgenlib.pyx":485
+ *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, &dosage_ct)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_dosages() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         cdef float* data32_ptr
  *         cdef double* data64_ptr
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 422, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 485, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 422, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 485, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 422, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 485, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 422, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 485, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 422, __pyx_L1_error)
+    __PYX_ERR(0, 485, __pyx_L1_error)
 
-    /* "pgenlib.pyx":421
+    /* "pgenlib.pyx":484
  *         cdef uint32_t dosage_ct
- *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._genovec, self._dosage_present, self._dosage_main, &dosage_ct)
+ *         cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, &dosage_ct)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages() error " + str(reterr))
  *         cdef float* data32_ptr
  */
   }
 
-  /* "pgenlib.pyx":425
+  /* "pgenlib.pyx":488
  *         cdef float* data32_ptr
  *         cdef double* data64_ptr
  *         if floatarr_out.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             data32_ptr = <float*>floatarr_out.data
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 425, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 488, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 425, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 488, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_float32); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 425, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_float32); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 488, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_1 = PyObject_RichCompare(__pyx_t_2, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 425, __pyx_L1_error)
+  __pyx_t_1 = PyObject_RichCompare(__pyx_t_2, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 488, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 425, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 488, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   if (__pyx_t_9) {
 
-    /* "pgenlib.pyx":426
+    /* "pgenlib.pyx":489
  *         cdef double* data64_ptr
  *         if floatarr_out.dtype == np.float32:
  *             data32_ptr = <float*>floatarr_out.data             # <<<<<<<<<<<<<<
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
  *         elif floatarr_out.dtype == np.float64:
  */
     __pyx_v_data32_ptr = ((float *)__pyx_v_floatarr_out->data);
 
-    /* "pgenlib.pyx":427
+    /* "pgenlib.pyx":490
  *         if floatarr_out.dtype == np.float32:
  *             data32_ptr = <float*>floatarr_out.data
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)             # <<<<<<<<<<<<<<
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)             # <<<<<<<<<<<<<<
  *         elif floatarr_out.dtype == np.float64:
  *             data64_ptr = <double*>floatarr_out.data
  */
-    plink2::Dosage16ToFloatsMinus9(__pyx_v_self->_genovec, __pyx_v_self->_dosage_present, __pyx_v_self->_dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data32_ptr);
+    plink2::Dosage16ToFloatsMinus9(__pyx_v_self->_pgv.genovec, __pyx_v_self->_pgv.dosage_present, __pyx_v_self->_pgv.dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data32_ptr);
 
-    /* "pgenlib.pyx":425
+    /* "pgenlib.pyx":488
  *         cdef float* data32_ptr
  *         cdef double* data64_ptr
  *         if floatarr_out.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             data32_ptr = <float*>floatarr_out.data
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
  */
     goto __pyx_L7;
   }
 
-  /* "pgenlib.pyx":428
+  /* "pgenlib.pyx":491
  *             data32_ptr = <float*>floatarr_out.data
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
  *         elif floatarr_out.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             data64_ptr = <double*>floatarr_out.data
- *             Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)
+ *             Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 428, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 491, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 428, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 491, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_float64); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 428, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_float64); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 491, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_5 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 428, __pyx_L1_error)
+  __pyx_t_5 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 491, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 428, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 491, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
   if (likely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":429
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+    /* "pgenlib.pyx":492
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
  *         elif floatarr_out.dtype == np.float64:
  *             data64_ptr = <double*>floatarr_out.data             # <<<<<<<<<<<<<<
- *             Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)
+ *             Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)
  *         else:
  */
     __pyx_v_data64_ptr = ((double *)__pyx_v_floatarr_out->data);
 
-    /* "pgenlib.pyx":430
+    /* "pgenlib.pyx":493
  *         elif floatarr_out.dtype == np.float64:
  *             data64_ptr = <double*>floatarr_out.data
- *             Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)             # <<<<<<<<<<<<<<
+ *             Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid read_dosages() floatarr_out array element type (float32 or float64 expected).")
  */
-    plink2::Dosage16ToDoublesMinus9(__pyx_v_self->_genovec, __pyx_v_self->_dosage_present, __pyx_v_self->_dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data64_ptr);
+    plink2::Dosage16ToDoublesMinus9(__pyx_v_self->_pgv.genovec, __pyx_v_self->_pgv.dosage_present, __pyx_v_self->_pgv.dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data64_ptr);
 
-    /* "pgenlib.pyx":428
+    /* "pgenlib.pyx":491
  *             data32_ptr = <float*>floatarr_out.data
- *             Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+ *             Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
  *         elif floatarr_out.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             data64_ptr = <double*>floatarr_out.data
- *             Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)
+ *             Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)
  */
     goto __pyx_L7;
   }
 
-  /* "pgenlib.pyx":432
- *             Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)
+  /* "pgenlib.pyx":495
+ *             Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)
  *         else:
  *             raise RuntimeError("Invalid read_dosages() floatarr_out array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__8, NULL); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 432, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__8, NULL); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 495, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_Raise(__pyx_t_5, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __PYX_ERR(0, 432, __pyx_L1_error)
+    __PYX_ERR(0, 495, __pyx_L1_error)
   }
   __pyx_L7:;
 
-  /* "pgenlib.pyx":433
+  /* "pgenlib.pyx":496
  *         else:
  *             raise RuntimeError("Invalid read_dosages() floatarr_out array element type (float32 or float64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":410
+  /* "pgenlib.pyx":473
  * 
  * 
  *     cpdef read_dosages(self, uint32_t variant_idx, np.ndarray floatarr_out, uint32_t allele_idx = 1):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -6296,53 +6767,53 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_floatarr_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_dosages", 0, 2, 3, 1); __PYX_ERR(0, 410, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_dosages", 0, 2, 3, 1); __PYX_ERR(0, 473, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[2] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_dosages") < 0)) __PYX_ERR(0, 410, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_dosages") < 0)) __PYX_ERR(0, 473, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 410, __pyx_L3_error)
+    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 473, __pyx_L3_error)
     __pyx_v_floatarr_out = ((PyArrayObject *)values[1]);
     if (values[2]) {
-      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 410, __pyx_L3_error)
+      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 473, __pyx_L3_error)
     } else {
       __pyx_v_allele_idx = ((uint32_t)1);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_dosages", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 410, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_dosages", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 473, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_dosages", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_out), __pyx_ptype_5numpy_ndarray, 1, "floatarr_out", 0))) __PYX_ERR(0, 410, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_out), __pyx_ptype_5numpy_ndarray, 1, "floatarr_out", 0))) __PYX_ERR(0, 473, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_12read_dosages(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx, __pyx_v_floatarr_out, __pyx_v_allele_idx);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -6358,15 +6829,15 @@
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("read_dosages", 0);
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_dosages(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_floatarr_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 410, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_dosages(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_floatarr_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 473, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -6375,26 +6846,25 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":436
+/* "pgenlib.pyx":499
  * 
  * 
  *     cpdef read_alleles(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_15read_alleles(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_alleles(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_allele_int32_out, int __pyx_skip_dispatch) {
   uint32_t __pyx_v_subset_size;
-  uint32_t __pyx_v_phasepresent_ct;
   plink2::PglErr __pyx_v_reterr;
   int32_t *__pyx_v_main_data_ptr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_out;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_out;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
@@ -6413,31 +6883,31 @@
   __Pyx_RefNannySetupContext("read_alleles", 0);
   __pyx_pybuffer_allele_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_out.refcount = 0;
   __pyx_pybuffernd_allele_int32_out.data = NULL;
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 436, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 499, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 436, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 499, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_15read_alleles)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 436, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 499, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -6447,42 +6917,42 @@
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[3] = {__pyx_t_5, __pyx_t_3, ((PyObject *)__pyx_v_allele_int32_out)};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 436, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 499, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[3] = {__pyx_t_5, __pyx_t_3, ((PyObject *)__pyx_v_allele_int32_out)};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 436, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 499, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(2+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 436, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(2+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 499, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, __pyx_t_3);
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32_out));
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, ((PyObject *)__pyx_v_allele_int32_out));
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 436, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 499, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -6497,226 +6967,226 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":437
+  /* "pgenlib.pyx":500
  * 
  *     cpdef read_alleles(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  */
   __pyx_t_8 = ((__pyx_v_variant_idx >= (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":438
+    /* "pgenlib.pyx":501
  *     cpdef read_alleles(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < subset_size:
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_variant_idx_too_lar, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_variant_idx_too_lar, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 438, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 501, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 438, __pyx_L1_error)
+    __PYX_ERR(0, 501, __pyx_L1_error)
 
-    /* "pgenlib.pyx":437
+    /* "pgenlib.pyx":500
  * 
  *     cpdef read_alleles(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  */
   }
 
-  /* "pgenlib.pyx":439
+  /* "pgenlib.pyx":502
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         if allele_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("read_alleles() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  */
   __pyx_t_9 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_9;
 
-  /* "pgenlib.pyx":440
+  /* "pgenlib.pyx":503
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   __pyx_t_8 = (((__pyx_v_allele_int32_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":441
+    /* "pgenlib.pyx":504
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("read_alleles() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")             # <<<<<<<<<<<<<<
  * 
- *         cdef uint32_t phasepresent_ct
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  */
-    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_allele_int32_out_is, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_allele_int32_out_is, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 441, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 504, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 441, __pyx_L1_error)
+    __PYX_ERR(0, 504, __pyx_L1_error)
 
-    /* "pgenlib.pyx":440
+    /* "pgenlib.pyx":503
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   }
 
-  /* "pgenlib.pyx":445
- *         cdef uint32_t phasepresent_ct
- *         # upgrade to multiallelic version of this function in the future
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":506
+ *             raise RuntimeError("read_alleles() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
+ * 
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_alleles() error " + str(reterr))
  */
-  __pyx_v_reterr = plink2::PgrGetP(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_self->_state_ptr, __pyx_v_self->_genovec, __pyx_v_self->_phasepresent, __pyx_v_self->_phaseinfo, (&__pyx_v_phasepresent_ct));
+  __pyx_v_reterr = plink2::PgrGetMP(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_self->_state_ptr, (&__pyx_v_self->_pgv));
 
-  /* "pgenlib.pyx":446
- *         # upgrade to multiallelic version of this function in the future
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+  /* "pgenlib.pyx":507
+ * 
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  */
   __pyx_t_8 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":447
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+    /* "pgenlib.pyx":508
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_alleles() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 447, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 508, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 447, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 508, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 447, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 508, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 447, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 508, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 447, __pyx_L1_error)
+    __PYX_ERR(0, 508, __pyx_L1_error)
 
-    /* "pgenlib.pyx":446
- *         # upgrade to multiallelic version of this function in the future
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+    /* "pgenlib.pyx":507
+ * 
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  */
   }
 
-  /* "pgenlib.pyx":448
+  /* "pgenlib.pyx":509
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_alleles() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))             # <<<<<<<<<<<<<<
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *         return
  */
   __pyx_t_10 = 0;
   __pyx_t_6 = -1;
   if (__pyx_t_10 < 0) {
     __pyx_t_10 += __pyx_pybuffernd_allele_int32_out.diminfo[0].shape;
     if (unlikely(__pyx_t_10 < 0)) __pyx_t_6 = 0;
   } else if (unlikely(__pyx_t_10 >= __pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 448, __pyx_L1_error)
+    __PYX_ERR(0, 509, __pyx_L1_error)
   }
   __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_10, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":449
+  /* "pgenlib.pyx":510
  *             raise RuntimeError("read_alleles() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)             # <<<<<<<<<<<<<<
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  plink2::GenoarrPhasedToAlleleCodesMinus9(__pyx_v_self->_genovec, __pyx_v_self->_phasepresent, __pyx_v_self->_phaseinfo, __pyx_v_subset_size, __pyx_v_phasepresent_ct, NULL, __pyx_v_main_data_ptr);
+  plink2::GenoarrMPToAlleleCodesMinus9((&__pyx_v_self->_pgv), __pyx_v_subset_size, NULL, __pyx_v_main_data_ptr);
 
-  /* "pgenlib.pyx":450
+  /* "pgenlib.pyx":511
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":436
+  /* "pgenlib.pyx":499
  * 
  * 
  *     cpdef read_alleles(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -6775,38 +7245,38 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles", 1, 2, 2, 1); __PYX_ERR(0, 436, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles", 1, 2, 2, 1); __PYX_ERR(0, 499, __pyx_L3_error)
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles") < 0)) __PYX_ERR(0, 436, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles") < 0)) __PYX_ERR(0, 499, __pyx_L3_error)
       }
     } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
       goto __pyx_L5_argtuple_error;
     } else {
       values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
       values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
     }
-    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 436, __pyx_L3_error)
+    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 499, __pyx_L3_error)
     __pyx_v_allele_int32_out = ((PyArrayObject *)values[1]);
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_alleles", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 436, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_alleles", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 499, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_alleles", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 436, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 499, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_14read_alleles(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx, __pyx_v_allele_int32_out);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -6826,19 +7296,19 @@
   __Pyx_RefNannySetupContext("read_alleles", 0);
   __pyx_pybuffer_allele_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_out.refcount = 0;
   __pyx_pybuffernd_allele_int32_out.data = NULL;
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 436, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 499, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_read_alleles(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_allele_int32_out, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 436, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_read_alleles(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_allele_int32_out, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 499, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -6856,26 +7326,25 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":453
+/* "pgenlib.pyx":514
  * 
  * 
  *     cpdef read_alleles_and_phasepresent(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True] phasepresent_out):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_17read_alleles_and_phasepresent(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_alleles_and_phasepresent(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_allele_int32_out, PyArrayObject *__pyx_v_phasepresent_out, int __pyx_skip_dispatch) {
   uint32_t __pyx_v_subset_size;
-  uint32_t __pyx_v_phasepresent_ct;
   plink2::PglErr __pyx_v_reterr;
   int32_t *__pyx_v_main_data_ptr;
   unsigned char *__pyx_v_phasepresent_data_ptr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_out;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_out;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent_out;
   __Pyx_Buffer __pyx_pybuffer_phasepresent_out;
@@ -6901,36 +7370,36 @@
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   __pyx_pybuffer_phasepresent_out.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_out.refcount = 0;
   __pyx_pybuffernd_phasepresent_out.data = NULL;
   __pyx_pybuffernd_phasepresent_out.rcbuffer = &__pyx_pybuffer_phasepresent_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 453, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 514, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 453, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 514, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_out.diminfo[0].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_out.diminfo[0].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_and_phasepresent); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 453, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_and_phasepresent); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 514, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_17read_alleles_and_phasepresent)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 453, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 514, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -6940,45 +7409,45 @@
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[4] = {__pyx_t_5, __pyx_t_3, ((PyObject *)__pyx_v_allele_int32_out), ((PyObject *)__pyx_v_phasepresent_out)};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 453, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 514, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[4] = {__pyx_t_5, __pyx_t_3, ((PyObject *)__pyx_v_allele_int32_out), ((PyObject *)__pyx_v_phasepresent_out)};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 453, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 514, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 453, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 514, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, __pyx_t_3);
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32_out));
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, ((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_INCREF(((PyObject *)__pyx_v_phasepresent_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_phasepresent_out));
           PyTuple_SET_ITEM(__pyx_t_7, 2+__pyx_t_6, ((PyObject *)__pyx_v_phasepresent_out));
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 453, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 514, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -6993,301 +7462,301 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":454
+  /* "pgenlib.pyx":515
  * 
  *     cpdef read_alleles_and_phasepresent(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True] phasepresent_out):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  */
   __pyx_t_8 = ((__pyx_v_variant_idx >= (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":455
+    /* "pgenlib.pyx":516
  *     cpdef read_alleles_and_phasepresent(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True] phasepresent_out):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_va, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_va, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 455, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 516, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 455, __pyx_L1_error)
+    __PYX_ERR(0, 516, __pyx_L1_error)
 
-    /* "pgenlib.pyx":454
+    /* "pgenlib.pyx":515
  * 
  *     cpdef read_alleles_and_phasepresent(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True] phasepresent_out):
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  */
   }
 
-  /* "pgenlib.pyx":456
+  /* "pgenlib.pyx":517
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
  */
   __pyx_t_9 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_9;
 
-  /* "pgenlib.pyx":457
+  /* "pgenlib.pyx":518
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
  *         if phasepresent_out.shape[0] < subset_size:
  */
   __pyx_t_8 = (((__pyx_v_allele_int32_out->dimensions[0]) < (2 * __pyx_v_subset_size)) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":458
+    /* "pgenlib.pyx":519
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")             # <<<<<<<<<<<<<<
  *         if phasepresent_out.shape[0] < subset_size:
  *             raise RuntimeError("read_alleles_and_phasepresent() phasepresent_out is too small (" + str(phasepresent_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  */
-    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_al, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_al, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 458, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 519, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 458, __pyx_L1_error)
+    __PYX_ERR(0, 519, __pyx_L1_error)
 
-    /* "pgenlib.pyx":457
+    /* "pgenlib.pyx":518
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef uint32_t subset_size = self._subset_size
  *         if allele_int32_out.shape[0] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
  *         if phasepresent_out.shape[0] < subset_size:
  */
   }
 
-  /* "pgenlib.pyx":459
+  /* "pgenlib.pyx":520
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
  *         if phasepresent_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() phasepresent_out is too small (" + str(phasepresent_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   __pyx_t_8 = (((__pyx_v_phasepresent_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":460
+    /* "pgenlib.pyx":521
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
  *         if phasepresent_out.shape[0] < subset_size:
  *             raise RuntimeError("read_alleles_and_phasepresent() phasepresent_out is too small (" + str(phasepresent_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")             # <<<<<<<<<<<<<<
  * 
- *         cdef uint32_t phasepresent_ct
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  */
-    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_ph, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_ph, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 460, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 521, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 460, __pyx_L1_error)
+    __PYX_ERR(0, 521, __pyx_L1_error)
 
-    /* "pgenlib.pyx":459
+    /* "pgenlib.pyx":520
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
  *         if phasepresent_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() phasepresent_out is too small (" + str(phasepresent_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
  * 
  */
   }
 
-  /* "pgenlib.pyx":464
- *         cdef uint32_t phasepresent_ct
- *         # upgrade to multiallelic version of this function in the future
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":523
+ *             raise RuntimeError("read_alleles_and_phasepresent() phasepresent_out is too small (" + str(phasepresent_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
+ * 
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))
  */
-  __pyx_v_reterr = plink2::PgrGetP(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_self->_state_ptr, __pyx_v_self->_genovec, __pyx_v_self->_phasepresent, __pyx_v_self->_phaseinfo, (&__pyx_v_phasepresent_ct));
+  __pyx_v_reterr = plink2::PgrGetMP(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_self->_state_ptr, (&__pyx_v_self->_pgv));
 
-  /* "pgenlib.pyx":465
- *         # upgrade to multiallelic version of this function in the future
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+  /* "pgenlib.pyx":524
+ * 
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  */
   __pyx_t_8 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_8)) {
 
-    /* "pgenlib.pyx":466
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+    /* "pgenlib.pyx":525
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  *         cdef unsigned char* phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[0]))
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 466, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 525, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 466, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 525, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_er, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 466, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_er, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 525, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 466, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 525, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_4, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __PYX_ERR(0, 466, __pyx_L1_error)
+    __PYX_ERR(0, 525, __pyx_L1_error)
 
-    /* "pgenlib.pyx":465
- *         # upgrade to multiallelic version of this function in the future
- *         cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+    /* "pgenlib.pyx":524
+ * 
+ *         cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  */
   }
 
-  /* "pgenlib.pyx":467
+  /* "pgenlib.pyx":526
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))             # <<<<<<<<<<<<<<
  *         cdef unsigned char* phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[0]))
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  */
   __pyx_t_10 = 0;
   __pyx_t_6 = -1;
   if (__pyx_t_10 < 0) {
     __pyx_t_10 += __pyx_pybuffernd_allele_int32_out.diminfo[0].shape;
     if (unlikely(__pyx_t_10 < 0)) __pyx_t_6 = 0;
   } else if (unlikely(__pyx_t_10 >= __pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 467, __pyx_L1_error)
+    __PYX_ERR(0, 526, __pyx_L1_error)
   }
   __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_10, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":468
+  /* "pgenlib.pyx":527
  *             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  *         cdef unsigned char* phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[0]))             # <<<<<<<<<<<<<<
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *         return
  */
   __pyx_t_10 = 0;
   __pyx_t_6 = -1;
   if (__pyx_t_10 < 0) {
     __pyx_t_10 += __pyx_pybuffernd_phasepresent_out.diminfo[0].shape;
     if (unlikely(__pyx_t_10 < 0)) __pyx_t_6 = 0;
   } else if (unlikely(__pyx_t_10 >= __pyx_pybuffernd_phasepresent_out.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 468, __pyx_L1_error)
+    __PYX_ERR(0, 527, __pyx_L1_error)
   }
   __pyx_v_phasepresent_data_ptr = ((unsigned char *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint8_t *, __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.buf, __pyx_t_10, __pyx_pybuffernd_phasepresent_out.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":469
+  /* "pgenlib.pyx":528
  *         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
  *         cdef unsigned char* phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[0]))
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)             # <<<<<<<<<<<<<<
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  plink2::GenoarrPhasedToAlleleCodesMinus9(__pyx_v_self->_genovec, __pyx_v_self->_phasepresent, __pyx_v_self->_phaseinfo, __pyx_v_subset_size, __pyx_v_phasepresent_ct, __pyx_v_phasepresent_data_ptr, __pyx_v_main_data_ptr);
+  plink2::GenoarrMPToAlleleCodesMinus9((&__pyx_v_self->_pgv), __pyx_v_subset_size, __pyx_v_phasepresent_data_ptr, __pyx_v_main_data_ptr);
 
-  /* "pgenlib.pyx":470
+  /* "pgenlib.pyx":529
  *         cdef unsigned char* phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[0]))
- *         GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *         GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":453
+  /* "pgenlib.pyx":514
  * 
  * 
  *     cpdef read_alleles_and_phasepresent(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True] phasepresent_out):             # <<<<<<<<<<<<<<
  *         if variant_idx >= self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -7351,47 +7820,47 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent", 1, 3, 3, 1); __PYX_ERR(0, 453, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent", 1, 3, 3, 1); __PYX_ERR(0, 514, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_phasepresent_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent", 1, 3, 3, 2); __PYX_ERR(0, 453, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent", 1, 3, 3, 2); __PYX_ERR(0, 514, __pyx_L3_error)
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_and_phasepresent") < 0)) __PYX_ERR(0, 453, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_and_phasepresent") < 0)) __PYX_ERR(0, 514, __pyx_L3_error)
       }
     } else if (PyTuple_GET_SIZE(__pyx_args) != 3) {
       goto __pyx_L5_argtuple_error;
     } else {
       values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
       values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
       values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
     }
-    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 453, __pyx_L3_error)
+    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 514, __pyx_L3_error)
     __pyx_v_allele_int32_out = ((PyArrayObject *)values[1]);
     __pyx_v_phasepresent_out = ((PyArrayObject *)values[2]);
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 453, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 514, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_alleles_and_phasepresent", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 453, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_out), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_out", 0))) __PYX_ERR(0, 453, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 514, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_out), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_out", 0))) __PYX_ERR(0, 514, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_16read_alleles_and_phasepresent(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -7417,24 +7886,24 @@
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   __pyx_pybuffer_phasepresent_out.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_out.refcount = 0;
   __pyx_pybuffernd_phasepresent_out.data = NULL;
   __pyx_pybuffernd_phasepresent_out.rcbuffer = &__pyx_pybuffer_phasepresent_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 453, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 514, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 453, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 514, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_out.diminfo[0].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_out.diminfo[0].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_read_alleles_and_phasepresent(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 453, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_read_alleles_and_phasepresent(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 514, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -7454,15 +7923,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":473
+/* "pgenlib.pyx":532
  * 
  * 
  *     cdef read_range_internal8(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         # todo: benchmark effect of adding @cython.boundscheck(False) and
  *         # @cython.wraparound(False) annotations to these multiple-variant-load
  */
 
@@ -7532,262 +8001,262 @@
   }
   __pyx_pybuffer_geno_int8_out.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int8_out.refcount = 0;
   __pyx_pybuffernd_geno_int8_out.data = NULL;
   __pyx_pybuffernd_geno_int8_out.rcbuffer = &__pyx_pybuffer_geno_int8_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 473, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 532, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int8_out.diminfo[0].strides = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int8_out.diminfo[0].shape = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int8_out.diminfo[1].strides = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int8_out.diminfo[1].shape = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":480
+  /* "pgenlib.pyx":539
  *         # large jobs (this probably belongs under pgenlib_ffi_support, rather
  *         # than here)
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_1 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_1;
 
-  /* "pgenlib.pyx":481
+  /* "pgenlib.pyx":540
  *         # than here)
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_2 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_2;
 
-  /* "pgenlib.pyx":482
+  /* "pgenlib.pyx":541
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
   __pyx_t_3 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_3;
 
-  /* "pgenlib.pyx":483
+  /* "pgenlib.pyx":542
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_1 = __pyx_v_self->_genovec;
+  __pyx_t_1 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":484
+  /* "pgenlib.pyx":543
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int8_t* data_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":485
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":544
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int8_t* data_ptr
  *         cdef uint32_t variant_idx
  */
   __pyx_t_4 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_4;
 
-  /* "pgenlib.pyx":489
+  /* "pgenlib.pyx":548
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_5 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":490
+    /* "pgenlib.pyx":549
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int8_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:
  */
     __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":491
+      /* "pgenlib.pyx":550
  *         if sample_maj == 0:
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if geno_int8_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 491, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 550, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 491, __pyx_L1_error)
+      __PYX_ERR(0, 550, __pyx_L1_error)
 
-      /* "pgenlib.pyx":490
+      /* "pgenlib.pyx":549
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int8_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":492
+    /* "pgenlib.pyx":551
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":493
+      /* "pgenlib.pyx":552
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 493, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 552, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 493, __pyx_L1_error)
+      __PYX_ERR(0, 552, __pyx_L1_error)
 
-      /* "pgenlib.pyx":492
+      /* "pgenlib.pyx":551
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     }
 
-    /* "pgenlib.pyx":494
+    /* "pgenlib.pyx":553
  *             if geno_int8_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_4 = __pyx_v_variant_idx_end;
     __pyx_t_9 = __pyx_t_4;
     for (__pyx_t_10 = __pyx_v_variant_idx_start; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
       __pyx_v_variant_idx = __pyx_t_10;
 
-      /* "pgenlib.pyx":495
+      /* "pgenlib.pyx":554
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec);
 
-      /* "pgenlib.pyx":496
+      /* "pgenlib.pyx":555
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[(variant_idx - variant_idx_start), 0])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":497
+        /* "pgenlib.pyx":556
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data_ptr = &(geno_int8_out[(variant_idx - variant_idx_start), 0])
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  */
-        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 497, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 556, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 497, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 556, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 497, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 556, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 497, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 556, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 497, __pyx_L1_error)
+        __PYX_ERR(0, 556, __pyx_L1_error)
 
-        /* "pgenlib.pyx":496
+        /* "pgenlib.pyx":555
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[(variant_idx - variant_idx_start), 0])
  */
       }
 
-      /* "pgenlib.pyx":498
+      /* "pgenlib.pyx":557
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[(variant_idx - variant_idx_start), 0])             # <<<<<<<<<<<<<<
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return
  */
       __pyx_t_11 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
@@ -7796,579 +8265,579 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int8_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_12 < 0) {
         __pyx_t_12 += __pyx_pybuffernd_geno_int8_out.diminfo[1].shape;
         if (unlikely(__pyx_t_12 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_12 >= __pyx_pybuffernd_geno_int8_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 498, __pyx_L1_error)
+        __PYX_ERR(0, 557, __pyx_L1_error)
       }
       __pyx_v_data_ptr = (&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int8_t *, __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int8_out.diminfo[0].strides, __pyx_t_12, __pyx_pybuffernd_geno_int8_out.diminfo[1].strides)));
 
-      /* "pgenlib.pyx":499
+      /* "pgenlib.pyx":558
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[(variant_idx - variant_idx_start), 0])
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if variant_idx_start >= variant_idx_end:
  */
       plink2::GenoarrToBytesMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
     }
 
-    /* "pgenlib.pyx":500
+    /* "pgenlib.pyx":559
  *                 data_ptr = &(geno_int8_out[(variant_idx - variant_idx_start), 0])
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":489
+    /* "pgenlib.pyx":548
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":501
+  /* "pgenlib.pyx":560
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int8_out.shape[0] < subset_size:
  */
   __pyx_t_5 = ((__pyx_v_variant_idx_start >= __pyx_v_variant_idx_end) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":502
+    /* "pgenlib.pyx":561
  *             return
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_start_var, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_start_var, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__10); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__10); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 502, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 561, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 502, __pyx_L1_error)
+    __PYX_ERR(0, 561, __pyx_L1_error)
 
-    /* "pgenlib.pyx":501
+    /* "pgenlib.pyx":560
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int8_out.shape[0] < subset_size:
  */
   }
 
-  /* "pgenlib.pyx":503
+  /* "pgenlib.pyx":562
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int8_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":504
+    /* "pgenlib.pyx":563
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 504, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 504, __pyx_L1_error)
+    __PYX_ERR(0, 563, __pyx_L1_error)
 
-    /* "pgenlib.pyx":503
+    /* "pgenlib.pyx":562
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int8_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":505
+  /* "pgenlib.pyx":564
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":506
+    /* "pgenlib.pyx":565
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 506, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 565, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 506, __pyx_L1_error)
+    __PYX_ERR(0, 565, __pyx_L1_error)
 
-    /* "pgenlib.pyx":505
+    /* "pgenlib.pyx":564
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":507
+  /* "pgenlib.pyx":566
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":508
+  /* "pgenlib.pyx":567
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":509
+  /* "pgenlib.pyx":568
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_variant_idx_offset = __pyx_v_variant_idx_start;
 
-  /* "pgenlib.pyx":510
+  /* "pgenlib.pyx":569
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":511
+  /* "pgenlib.pyx":570
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":512
+  /* "pgenlib.pyx":571
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_14 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_14;
 
-  /* "pgenlib.pyx":513
+  /* "pgenlib.pyx":572
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* vmaj_iter
  */
   __pyx_t_1 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_1;
 
-  /* "pgenlib.pyx":514
+  /* "pgenlib.pyx":573
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_iter
  *         cdef uintptr_t* smaj_iter
  */
   __pyx_t_1 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_1;
 
-  /* "pgenlib.pyx":521
+  /* "pgenlib.pyx":580
  *         cdef uint32_t sample_batch_idx
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_4 = __pyx_v_variant_batch_ct;
   __pyx_t_9 = __pyx_t_4;
   for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_10;
 
-    /* "pgenlib.pyx":522
+    /* "pgenlib.pyx":581
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     __pyx_t_5 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":523
+      /* "pgenlib.pyx":582
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  */
       __pyx_t_15 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 523, __pyx_L1_error)
+        __PYX_ERR(0, 582, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":522
+      /* "pgenlib.pyx":581
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     }
 
-    /* "pgenlib.pyx":524
+    /* "pgenlib.pyx":583
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":525
+    /* "pgenlib.pyx":584
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_16 = __pyx_v_variant_batch_size;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_uii = __pyx_t_18;
 
-      /* "pgenlib.pyx":526
+      /* "pgenlib.pyx":585
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, (__pyx_v_uii + __pyx_v_variant_idx_offset), __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_vmaj_iter);
 
-      /* "pgenlib.pyx":527
+      /* "pgenlib.pyx":586
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":528
+        /* "pgenlib.pyx":587
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  */
-        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 528, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 587, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 528, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 587, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 528, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 587, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 528, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 587, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 528, __pyx_L1_error)
+        __PYX_ERR(0, 587, __pyx_L1_error)
 
-        /* "pgenlib.pyx":527
+        /* "pgenlib.pyx":586
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       }
 
-      /* "pgenlib.pyx":529
+      /* "pgenlib.pyx":588
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[__pyx_v_sample_ctaw2]));
     }
 
-    /* "pgenlib.pyx":530
+    /* "pgenlib.pyx":589
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":531
+    /* "pgenlib.pyx":590
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":532
+    /* "pgenlib.pyx":591
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_16 = __pyx_v_sample_batch_ct;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_18;
 
-      /* "pgenlib.pyx":533
+      /* "pgenlib.pyx":592
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_5 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":534
+        /* "pgenlib.pyx":593
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  */
         __pyx_t_15 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 534, __pyx_L1_error)
+          __PYX_ERR(0, 593, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":533
+        /* "pgenlib.pyx":592
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":535
+      /* "pgenlib.pyx":594
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  */
       __pyx_v_smaj_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":536
+      /* "pgenlib.pyx":595
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":537
+      /* "pgenlib.pyx":596
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  */
       __pyx_t_19 = __pyx_v_sample_batch_size;
       __pyx_t_20 = __pyx_t_19;
       for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
         __pyx_v_uii = __pyx_t_21;
 
-        /* "pgenlib.pyx":538
+        /* "pgenlib.pyx":597
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])             # <<<<<<<<<<<<<<
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  */
         __pyx_t_11 = (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch));
         __pyx_t_22 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch);
         __pyx_t_13 = -1;
         if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int8_out.diminfo[0].shape)) __pyx_t_13 = 0;
         if (unlikely(__pyx_t_22 >= (size_t)__pyx_pybuffernd_geno_int8_out.diminfo[1].shape)) __pyx_t_13 = 1;
         if (unlikely(__pyx_t_13 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_13);
-          __PYX_ERR(0, 538, __pyx_L1_error)
+          __PYX_ERR(0, 597, __pyx_L1_error)
         }
         __pyx_v_data_ptr = (&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int8_t *, __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int8_out.diminfo[0].strides, __pyx_t_22, __pyx_pybuffernd_geno_int8_out.diminfo[1].strides)));
 
-        /* "pgenlib.pyx":539
+        /* "pgenlib.pyx":598
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  */
         plink2::GenoarrToBytesMinus9(__pyx_v_smaj_iter, __pyx_v_variant_batch_size, __pyx_v_data_ptr);
 
-        /* "pgenlib.pyx":540
+        /* "pgenlib.pyx":599
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch
  */
         __pyx_v_smaj_iter = (&(__pyx_v_smaj_iter[plink2::kPglNypTransposeWords]));
       }
 
-      /* "pgenlib.pyx":541
+      /* "pgenlib.pyx":600
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[plink2::kPglNypTransposeWords]));
     }
 
-    /* "pgenlib.pyx":542
+    /* "pgenlib.pyx":601
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_idx_offset = (__pyx_v_variant_idx_offset + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":543
+  /* "pgenlib.pyx":602
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":473
+  /* "pgenlib.pyx":532
  * 
  * 
  *     cdef read_range_internal8(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         # todo: benchmark effect of adding @cython.boundscheck(False) and
  *         # @cython.wraparound(False) annotations to these multiple-variant-load
  */
 
@@ -8390,15 +8859,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":545
+/* "pgenlib.pyx":604
  *         return
  * 
  *     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -8468,262 +8937,262 @@
   }
   __pyx_pybuffer_geno_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int32_out.refcount = 0;
   __pyx_pybuffernd_geno_int32_out.data = NULL;
   __pyx_pybuffernd_geno_int32_out.rcbuffer = &__pyx_pybuffer_geno_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 545, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 604, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int32_out.diminfo[0].strides = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int32_out.diminfo[0].shape = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int32_out.diminfo[1].strides = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int32_out.diminfo[1].shape = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":546
+  /* "pgenlib.pyx":605
  * 
  *     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_1 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_1;
 
-  /* "pgenlib.pyx":547
+  /* "pgenlib.pyx":606
  *     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_2 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_2;
 
-  /* "pgenlib.pyx":548
+  /* "pgenlib.pyx":607
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
   __pyx_t_3 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_3;
 
-  /* "pgenlib.pyx":549
+  /* "pgenlib.pyx":608
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_1 = __pyx_v_self->_genovec;
+  __pyx_t_1 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":550
+  /* "pgenlib.pyx":609
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int32_t* data_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":551
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":610
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int32_t* data_ptr
  *         cdef uint32_t variant_idx
  */
   __pyx_t_4 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_4;
 
-  /* "pgenlib.pyx":555
+  /* "pgenlib.pyx":614
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_5 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":556
+    /* "pgenlib.pyx":615
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:
  */
     __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":557
+      /* "pgenlib.pyx":616
  *         if sample_maj == 0:
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if geno_int32_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 557, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 616, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 557, __pyx_L1_error)
+      __PYX_ERR(0, 616, __pyx_L1_error)
 
-      /* "pgenlib.pyx":556
+      /* "pgenlib.pyx":615
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":558
+    /* "pgenlib.pyx":617
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":559
+      /* "pgenlib.pyx":618
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 559, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 618, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 559, __pyx_L1_error)
+      __PYX_ERR(0, 618, __pyx_L1_error)
 
-      /* "pgenlib.pyx":558
+      /* "pgenlib.pyx":617
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     }
 
-    /* "pgenlib.pyx":560
+    /* "pgenlib.pyx":619
  *             if geno_int32_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_4 = __pyx_v_variant_idx_end;
     __pyx_t_9 = __pyx_t_4;
     for (__pyx_t_10 = __pyx_v_variant_idx_start; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
       __pyx_v_variant_idx = __pyx_t_10;
 
-      /* "pgenlib.pyx":561
+      /* "pgenlib.pyx":620
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec);
 
-      /* "pgenlib.pyx":562
+      /* "pgenlib.pyx":621
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[(variant_idx - variant_idx_start), 0]))
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":563
+        /* "pgenlib.pyx":622
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data_ptr = <int32_t*>(&(geno_int32_out[(variant_idx - variant_idx_start), 0]))
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  */
-        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 563, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 622, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 622, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 563, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 622, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 563, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 622, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 563, __pyx_L1_error)
+        __PYX_ERR(0, 622, __pyx_L1_error)
 
-        /* "pgenlib.pyx":562
+        /* "pgenlib.pyx":621
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[(variant_idx - variant_idx_start), 0]))
  */
       }
 
-      /* "pgenlib.pyx":564
+      /* "pgenlib.pyx":623
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[(variant_idx - variant_idx_start), 0]))             # <<<<<<<<<<<<<<
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return
  */
       __pyx_t_11 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
@@ -8732,579 +9201,579 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int32_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_12 < 0) {
         __pyx_t_12 += __pyx_pybuffernd_geno_int32_out.diminfo[1].shape;
         if (unlikely(__pyx_t_12 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_12 >= __pyx_pybuffernd_geno_int32_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 564, __pyx_L1_error)
+        __PYX_ERR(0, 623, __pyx_L1_error)
       }
       __pyx_v_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int32_out.diminfo[0].strides, __pyx_t_12, __pyx_pybuffernd_geno_int32_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":565
+      /* "pgenlib.pyx":624
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[(variant_idx - variant_idx_start), 0]))
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if variant_idx_start >= variant_idx_end:
  */
       plink2::GenoarrToInt32sMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
     }
 
-    /* "pgenlib.pyx":566
+    /* "pgenlib.pyx":625
  *                 data_ptr = <int32_t*>(&(geno_int32_out[(variant_idx - variant_idx_start), 0]))
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":555
+    /* "pgenlib.pyx":614
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":567
+  /* "pgenlib.pyx":626
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int32_out.shape[0] < subset_size:
  */
   __pyx_t_5 = ((__pyx_v_variant_idx_start >= __pyx_v_variant_idx_end) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":568
+    /* "pgenlib.pyx":627
  *             return
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_start_var, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_start_var, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__10); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__10); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 568, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 627, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 568, __pyx_L1_error)
+    __PYX_ERR(0, 627, __pyx_L1_error)
 
-    /* "pgenlib.pyx":567
+    /* "pgenlib.pyx":626
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int32_out.shape[0] < subset_size:
  */
   }
 
-  /* "pgenlib.pyx":569
+  /* "pgenlib.pyx":628
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int32_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":570
+    /* "pgenlib.pyx":629
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 570, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 570, __pyx_L1_error)
+    __PYX_ERR(0, 629, __pyx_L1_error)
 
-    /* "pgenlib.pyx":569
+    /* "pgenlib.pyx":628
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int32_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":571
+  /* "pgenlib.pyx":630
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":572
+    /* "pgenlib.pyx":631
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 572, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 631, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 572, __pyx_L1_error)
+    __PYX_ERR(0, 631, __pyx_L1_error)
 
-    /* "pgenlib.pyx":571
+    /* "pgenlib.pyx":630
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":573
+  /* "pgenlib.pyx":632
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":574
+  /* "pgenlib.pyx":633
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":575
+  /* "pgenlib.pyx":634
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_variant_idx_offset = __pyx_v_variant_idx_start;
 
-  /* "pgenlib.pyx":576
+  /* "pgenlib.pyx":635
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":577
+  /* "pgenlib.pyx":636
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":578
+  /* "pgenlib.pyx":637
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_14 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_14;
 
-  /* "pgenlib.pyx":579
+  /* "pgenlib.pyx":638
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* vmaj_iter
  */
   __pyx_t_1 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_1;
 
-  /* "pgenlib.pyx":580
+  /* "pgenlib.pyx":639
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_iter
  *         cdef uintptr_t* smaj_iter
  */
   __pyx_t_1 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_1;
 
-  /* "pgenlib.pyx":587
+  /* "pgenlib.pyx":646
  *         cdef uint32_t sample_batch_idx
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_4 = __pyx_v_variant_batch_ct;
   __pyx_t_9 = __pyx_t_4;
   for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_10;
 
-    /* "pgenlib.pyx":588
+    /* "pgenlib.pyx":647
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     __pyx_t_5 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":589
+      /* "pgenlib.pyx":648
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  */
       __pyx_t_15 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 589, __pyx_L1_error)
+        __PYX_ERR(0, 648, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":588
+      /* "pgenlib.pyx":647
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     }
 
-    /* "pgenlib.pyx":590
+    /* "pgenlib.pyx":649
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":591
+    /* "pgenlib.pyx":650
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_16 = __pyx_v_variant_batch_size;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_uii = __pyx_t_18;
 
-      /* "pgenlib.pyx":592
+      /* "pgenlib.pyx":651
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, (__pyx_v_uii + __pyx_v_variant_idx_offset), __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_vmaj_iter);
 
-      /* "pgenlib.pyx":593
+      /* "pgenlib.pyx":652
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":594
+        /* "pgenlib.pyx":653
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  */
-        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 594, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 653, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 594, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 653, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 594, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 653, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 594, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 653, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 594, __pyx_L1_error)
+        __PYX_ERR(0, 653, __pyx_L1_error)
 
-        /* "pgenlib.pyx":593
+        /* "pgenlib.pyx":652
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       }
 
-      /* "pgenlib.pyx":595
+      /* "pgenlib.pyx":654
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[__pyx_v_sample_ctaw2]));
     }
 
-    /* "pgenlib.pyx":596
+    /* "pgenlib.pyx":655
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":597
+    /* "pgenlib.pyx":656
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":598
+    /* "pgenlib.pyx":657
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_16 = __pyx_v_sample_batch_ct;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_18;
 
-      /* "pgenlib.pyx":599
+      /* "pgenlib.pyx":658
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_5 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":600
+        /* "pgenlib.pyx":659
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  */
         __pyx_t_15 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 600, __pyx_L1_error)
+          __PYX_ERR(0, 659, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":599
+        /* "pgenlib.pyx":658
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":601
+      /* "pgenlib.pyx":660
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  */
       __pyx_v_smaj_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":602
+      /* "pgenlib.pyx":661
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":603
+      /* "pgenlib.pyx":662
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  */
       __pyx_t_19 = __pyx_v_sample_batch_size;
       __pyx_t_20 = __pyx_t_19;
       for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
         __pyx_v_uii = __pyx_t_21;
 
-        /* "pgenlib.pyx":604
+        /* "pgenlib.pyx":663
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))             # <<<<<<<<<<<<<<
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  */
         __pyx_t_11 = (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch));
         __pyx_t_22 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch);
         __pyx_t_13 = -1;
         if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int32_out.diminfo[0].shape)) __pyx_t_13 = 0;
         if (unlikely(__pyx_t_22 >= (size_t)__pyx_pybuffernd_geno_int32_out.diminfo[1].shape)) __pyx_t_13 = 1;
         if (unlikely(__pyx_t_13 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_13);
-          __PYX_ERR(0, 604, __pyx_L1_error)
+          __PYX_ERR(0, 663, __pyx_L1_error)
         }
         __pyx_v_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int32_out.diminfo[0].strides, __pyx_t_22, __pyx_pybuffernd_geno_int32_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":605
+        /* "pgenlib.pyx":664
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  */
         plink2::GenoarrToInt32sMinus9(__pyx_v_smaj_iter, __pyx_v_variant_batch_size, __pyx_v_data_ptr);
 
-        /* "pgenlib.pyx":606
+        /* "pgenlib.pyx":665
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch
  */
         __pyx_v_smaj_iter = (&(__pyx_v_smaj_iter[plink2::kPglNypTransposeWords]));
       }
 
-      /* "pgenlib.pyx":607
+      /* "pgenlib.pyx":666
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[plink2::kPglNypTransposeWords]));
     }
 
-    /* "pgenlib.pyx":608
+    /* "pgenlib.pyx":667
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_idx_offset = (__pyx_v_variant_idx_offset + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":609
+  /* "pgenlib.pyx":668
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":545
+  /* "pgenlib.pyx":604
  *         return
  * 
  *     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -9326,15 +9795,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":611
+/* "pgenlib.pyx":670
  *         return
  * 
  *     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -9404,262 +9873,262 @@
   }
   __pyx_pybuffer_geno_int64_out.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int64_out.refcount = 0;
   __pyx_pybuffernd_geno_int64_out.data = NULL;
   __pyx_pybuffernd_geno_int64_out.rcbuffer = &__pyx_pybuffer_geno_int64_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int64_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 611, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int64_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 670, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int64_out.diminfo[0].strides = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int64_out.diminfo[0].shape = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int64_out.diminfo[1].strides = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int64_out.diminfo[1].shape = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":612
+  /* "pgenlib.pyx":671
  * 
  *     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_1 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_1;
 
-  /* "pgenlib.pyx":613
+  /* "pgenlib.pyx":672
  *     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_2 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_2;
 
-  /* "pgenlib.pyx":614
+  /* "pgenlib.pyx":673
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
   __pyx_t_3 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_3;
 
-  /* "pgenlib.pyx":615
+  /* "pgenlib.pyx":674
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_1 = __pyx_v_self->_genovec;
+  __pyx_t_1 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":616
+  /* "pgenlib.pyx":675
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int64_t* data_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":617
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":676
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int64_t* data_ptr
  *         cdef uint32_t variant_idx
  */
   __pyx_t_4 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_4;
 
-  /* "pgenlib.pyx":621
+  /* "pgenlib.pyx":680
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_5 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":622
+    /* "pgenlib.pyx":681
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int64_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:
  */
     __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":623
+      /* "pgenlib.pyx":682
  *         if sample_maj == 0:
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if geno_int64_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 623, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 682, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 623, __pyx_L1_error)
+      __PYX_ERR(0, 682, __pyx_L1_error)
 
-      /* "pgenlib.pyx":622
+      /* "pgenlib.pyx":681
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int64_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":624
+    /* "pgenlib.pyx":683
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":625
+      /* "pgenlib.pyx":684
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 625, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 684, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 625, __pyx_L1_error)
+      __PYX_ERR(0, 684, __pyx_L1_error)
 
-      /* "pgenlib.pyx":624
+      /* "pgenlib.pyx":683
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     }
 
-    /* "pgenlib.pyx":626
+    /* "pgenlib.pyx":685
  *             if geno_int64_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_4 = __pyx_v_variant_idx_end;
     __pyx_t_9 = __pyx_t_4;
     for (__pyx_t_10 = __pyx_v_variant_idx_start; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
       __pyx_v_variant_idx = __pyx_t_10;
 
-      /* "pgenlib.pyx":627
+      /* "pgenlib.pyx":686
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec);
 
-      /* "pgenlib.pyx":628
+      /* "pgenlib.pyx":687
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[(variant_idx - variant_idx_start), 0])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":629
+        /* "pgenlib.pyx":688
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data_ptr = &(geno_int64_out[(variant_idx - variant_idx_start), 0])
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  */
-        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 629, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 688, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 688, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 629, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 688, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 629, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 688, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 629, __pyx_L1_error)
+        __PYX_ERR(0, 688, __pyx_L1_error)
 
-        /* "pgenlib.pyx":628
+        /* "pgenlib.pyx":687
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[(variant_idx - variant_idx_start), 0])
  */
       }
 
-      /* "pgenlib.pyx":630
+      /* "pgenlib.pyx":689
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[(variant_idx - variant_idx_start), 0])             # <<<<<<<<<<<<<<
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return
  */
       __pyx_t_11 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
@@ -9668,579 +10137,579 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int64_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_12 < 0) {
         __pyx_t_12 += __pyx_pybuffernd_geno_int64_out.diminfo[1].shape;
         if (unlikely(__pyx_t_12 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_12 >= __pyx_pybuffernd_geno_int64_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 630, __pyx_L1_error)
+        __PYX_ERR(0, 689, __pyx_L1_error)
       }
       __pyx_v_data_ptr = ((int64_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int64_t *, __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int64_out.diminfo[0].strides, __pyx_t_12, __pyx_pybuffernd_geno_int64_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":631
+      /* "pgenlib.pyx":690
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[(variant_idx - variant_idx_start), 0])
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if variant_idx_start >= variant_idx_end:
  */
       plink2::GenoarrToInt64sMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
     }
 
-    /* "pgenlib.pyx":632
+    /* "pgenlib.pyx":691
  *                 data_ptr = &(geno_int64_out[(variant_idx - variant_idx_start), 0])
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":621
+    /* "pgenlib.pyx":680
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":633
+  /* "pgenlib.pyx":692
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int64_out.shape[0] < subset_size:
  */
   __pyx_t_5 = ((__pyx_v_variant_idx_start >= __pyx_v_variant_idx_end) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":634
+    /* "pgenlib.pyx":693
  *             return
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_start_var, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_start_var, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__10); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__10); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 634, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 693, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 634, __pyx_L1_error)
+    __PYX_ERR(0, 693, __pyx_L1_error)
 
-    /* "pgenlib.pyx":633
+    /* "pgenlib.pyx":692
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int64_out.shape[0] < subset_size:
  */
   }
 
-  /* "pgenlib.pyx":635
+  /* "pgenlib.pyx":694
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int64_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":636
+    /* "pgenlib.pyx":695
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 636, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 695, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 636, __pyx_L1_error)
+    __PYX_ERR(0, 695, __pyx_L1_error)
 
-    /* "pgenlib.pyx":635
+    /* "pgenlib.pyx":694
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if geno_int64_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":637
+  /* "pgenlib.pyx":696
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":638
+    /* "pgenlib.pyx":697
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 638, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 697, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 638, __pyx_L1_error)
+    __PYX_ERR(0, 697, __pyx_L1_error)
 
-    /* "pgenlib.pyx":637
+    /* "pgenlib.pyx":696
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":639
+  /* "pgenlib.pyx":698
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":640
+  /* "pgenlib.pyx":699
  *             raise RuntimeError("Sample-major read_range() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":641
+  /* "pgenlib.pyx":700
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_variant_idx_offset = __pyx_v_variant_idx_start;
 
-  /* "pgenlib.pyx":642
+  /* "pgenlib.pyx":701
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":643
+  /* "pgenlib.pyx":702
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":644
+  /* "pgenlib.pyx":703
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_14 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_14;
 
-  /* "pgenlib.pyx":645
+  /* "pgenlib.pyx":704
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* vmaj_iter
  */
   __pyx_t_1 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_1;
 
-  /* "pgenlib.pyx":646
+  /* "pgenlib.pyx":705
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_iter
  *         cdef uintptr_t* smaj_iter
  */
   __pyx_t_1 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_1;
 
-  /* "pgenlib.pyx":653
+  /* "pgenlib.pyx":712
  *         cdef uint32_t sample_batch_idx
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_4 = __pyx_v_variant_batch_ct;
   __pyx_t_9 = __pyx_t_4;
   for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_10;
 
-    /* "pgenlib.pyx":654
+    /* "pgenlib.pyx":713
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     __pyx_t_5 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":655
+      /* "pgenlib.pyx":714
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  */
       __pyx_t_15 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 655, __pyx_L1_error)
+        __PYX_ERR(0, 714, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":654
+      /* "pgenlib.pyx":713
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     }
 
-    /* "pgenlib.pyx":656
+    /* "pgenlib.pyx":715
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":657
+    /* "pgenlib.pyx":716
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_16 = __pyx_v_variant_batch_size;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_uii = __pyx_t_18;
 
-      /* "pgenlib.pyx":658
+      /* "pgenlib.pyx":717
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, (__pyx_v_uii + __pyx_v_variant_idx_offset), __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_vmaj_iter);
 
-      /* "pgenlib.pyx":659
+      /* "pgenlib.pyx":718
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":660
+        /* "pgenlib.pyx":719
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  */
-        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 660, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 719, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 660, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 660, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 719, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 660, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 660, __pyx_L1_error)
+        __PYX_ERR(0, 719, __pyx_L1_error)
 
-        /* "pgenlib.pyx":659
+        /* "pgenlib.pyx":718
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       }
 
-      /* "pgenlib.pyx":661
+      /* "pgenlib.pyx":720
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[__pyx_v_sample_ctaw2]));
     }
 
-    /* "pgenlib.pyx":662
+    /* "pgenlib.pyx":721
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":663
+    /* "pgenlib.pyx":722
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":664
+    /* "pgenlib.pyx":723
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_16 = __pyx_v_sample_batch_ct;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_18;
 
-      /* "pgenlib.pyx":665
+      /* "pgenlib.pyx":724
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_5 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":666
+        /* "pgenlib.pyx":725
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  */
         __pyx_t_15 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 666, __pyx_L1_error)
+          __PYX_ERR(0, 725, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":665
+        /* "pgenlib.pyx":724
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":667
+      /* "pgenlib.pyx":726
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  */
       __pyx_v_smaj_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":668
+      /* "pgenlib.pyx":727
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":669
+      /* "pgenlib.pyx":728
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  */
       __pyx_t_19 = __pyx_v_sample_batch_size;
       __pyx_t_20 = __pyx_t_19;
       for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
         __pyx_v_uii = __pyx_t_21;
 
-        /* "pgenlib.pyx":670
+        /* "pgenlib.pyx":729
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])             # <<<<<<<<<<<<<<
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  */
         __pyx_t_11 = (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch));
         __pyx_t_22 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch);
         __pyx_t_13 = -1;
         if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int64_out.diminfo[0].shape)) __pyx_t_13 = 0;
         if (unlikely(__pyx_t_22 >= (size_t)__pyx_pybuffernd_geno_int64_out.diminfo[1].shape)) __pyx_t_13 = 1;
         if (unlikely(__pyx_t_13 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_13);
-          __PYX_ERR(0, 670, __pyx_L1_error)
+          __PYX_ERR(0, 729, __pyx_L1_error)
         }
         __pyx_v_data_ptr = ((int64_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int64_t *, __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int64_out.diminfo[0].strides, __pyx_t_22, __pyx_pybuffernd_geno_int64_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":671
+        /* "pgenlib.pyx":730
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  */
         plink2::GenoarrToInt64sMinus9(__pyx_v_smaj_iter, __pyx_v_variant_batch_size, __pyx_v_data_ptr);
 
-        /* "pgenlib.pyx":672
+        /* "pgenlib.pyx":731
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch
  */
         __pyx_v_smaj_iter = (&(__pyx_v_smaj_iter[plink2::kPglNypTransposeWords]));
       }
 
-      /* "pgenlib.pyx":673
+      /* "pgenlib.pyx":732
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[plink2::kPglNypTransposeWords]));
     }
 
-    /* "pgenlib.pyx":674
+    /* "pgenlib.pyx":733
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_idx_offset = (__pyx_v_variant_idx_offset + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":675
+  /* "pgenlib.pyx":734
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":611
+  /* "pgenlib.pyx":670
  *         return
  * 
  *     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -10262,15 +10731,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":677
+/* "pgenlib.pyx":736
  *         return
  * 
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -10311,25 +10780,25 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_range); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 677, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_range); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 736, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_19read_range)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 677, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 736, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 677, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 736, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_5 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 677, __pyx_L1_error)
+        __pyx_t_5 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 736, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_5);
-        __pyx_t_6 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 677, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 736, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_7 = __pyx_t_1; __pyx_t_8 = NULL;
         __pyx_t_9 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_7))) {
           __pyx_t_8 = PyMethod_GET_SELF(__pyx_t_7);
           if (likely(__pyx_t_8)) {
@@ -10339,37 +10808,37 @@
             __Pyx_DECREF_SET(__pyx_t_7, function);
             __pyx_t_9 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_7)) {
           PyObject *__pyx_temp[6] = {__pyx_t_8, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_geno_int_out), __pyx_t_5, __pyx_t_6};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 677, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 736, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_8); __pyx_t_8 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_7)) {
           PyObject *__pyx_temp[6] = {__pyx_t_8, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_geno_int_out), __pyx_t_5, __pyx_t_6};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 677, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 736, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_8); __pyx_t_8 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         } else
         #endif
         {
-          __pyx_t_10 = PyTuple_New(5+__pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 677, __pyx_L1_error)
+          __pyx_t_10 = PyTuple_New(5+__pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 736, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_10);
           if (__pyx_t_8) {
             __Pyx_GIVEREF(__pyx_t_8); PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_8); __pyx_t_8 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_10, 0+__pyx_t_9, __pyx_t_3);
           __Pyx_GIVEREF(__pyx_t_4);
@@ -10381,15 +10850,15 @@
           PyTuple_SET_ITEM(__pyx_t_10, 3+__pyx_t_9, __pyx_t_5);
           __Pyx_GIVEREF(__pyx_t_6);
           PyTuple_SET_ITEM(__pyx_t_10, 4+__pyx_t_9, __pyx_t_6);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
           __pyx_t_5 = 0;
           __pyx_t_6 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_7, __pyx_t_10, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 677, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_7, __pyx_t_10, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 736, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
         }
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -10404,233 +10873,233 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":678
+  /* "pgenlib.pyx":737
  * 
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.dtype == np.int8:
  */
   __pyx_t_11 = ((__pyx_v_variant_idx_end > (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_11)) {
 
-    /* "pgenlib.pyx":679
+    /* "pgenlib.pyx":738
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         if geno_int_out.dtype == np.int8:
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_end_too_l, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_variant_idx_end_too_l, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 679, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 738, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 679, __pyx_L1_error)
+    __PYX_ERR(0, 738, __pyx_L1_error)
 
-    /* "pgenlib.pyx":678
+    /* "pgenlib.pyx":737
  * 
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.dtype == np.int8:
  */
   }
 
-  /* "pgenlib.pyx":680
+  /* "pgenlib.pyx":739
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.dtype == np.int8:             # <<<<<<<<<<<<<<
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 680, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 739, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_7, __pyx_n_s_np); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 680, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_7, __pyx_n_s_np); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 739, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_7);
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_7, __pyx_n_s_int8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 680, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_7, __pyx_n_s_int8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 739, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-  __pyx_t_7 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_7); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 680, __pyx_L1_error)
+  __pyx_t_7 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_7); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 739, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_7); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 680, __pyx_L1_error)
+  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_7); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 739, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
   if (__pyx_t_11) {
 
-    /* "pgenlib.pyx":681
+    /* "pgenlib.pyx":740
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.dtype == np.int8:
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  */
     __pyx_t_12.__pyx_n = 2;
     __pyx_t_12.allele_idx = __pyx_v_allele_idx;
     __pyx_t_12.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_7 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_range_internal8(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_12); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 681, __pyx_L1_error)
+    __pyx_t_7 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_range_internal8(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_12); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 740, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
 
-    /* "pgenlib.pyx":680
+    /* "pgenlib.pyx":739
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.dtype == np.int8:             # <<<<<<<<<<<<<<
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":682
+  /* "pgenlib.pyx":741
  *         if geno_int_out.dtype == np.int8:
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:             # <<<<<<<<<<<<<<
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:
  */
-  __pyx_t_7 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 682, __pyx_L1_error)
+  __pyx_t_7 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 741, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_7);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 682, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 741, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_int32); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 682, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_int32); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 741, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_7, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 682, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_7, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 741, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 682, __pyx_L1_error)
+  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 741, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (__pyx_t_11) {
 
-    /* "pgenlib.pyx":683
+    /* "pgenlib.pyx":742
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         elif geno_int_out.dtype == np.int64:
  *             self.read_range_internal64(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  */
     __pyx_t_13.__pyx_n = 2;
     __pyx_t_13.allele_idx = __pyx_v_allele_idx;
     __pyx_t_13.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_range_internal32(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_13); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 683, __pyx_L1_error)
+    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_range_internal32(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_13); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 742, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
 
-    /* "pgenlib.pyx":682
+    /* "pgenlib.pyx":741
  *         if geno_int_out.dtype == np.int8:
  *             self.read_range_internal8(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:             # <<<<<<<<<<<<<<
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":684
+  /* "pgenlib.pyx":743
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:             # <<<<<<<<<<<<<<
  *             self.read_range_internal64(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         else:
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 684, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 743, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 684, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 743, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __pyx_t_7 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_int64); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 684, __pyx_L1_error)
+  __pyx_t_7 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_int64); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 743, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_7);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_1 = PyObject_RichCompare(__pyx_t_2, __pyx_t_7, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 684, __pyx_L1_error)
+  __pyx_t_1 = PyObject_RichCompare(__pyx_t_2, __pyx_t_7, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 743, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 684, __pyx_L1_error)
+  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 743, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   if (likely(__pyx_t_11)) {
 
-    /* "pgenlib.pyx":685
+    /* "pgenlib.pyx":744
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:
  *             self.read_range_internal64(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid read_range() geno_int_out array element type (int8, int32, or int64 expected).")
  */
     __pyx_t_14.__pyx_n = 2;
     __pyx_t_14.allele_idx = __pyx_v_allele_idx;
     __pyx_t_14.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_range_internal64(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_14); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 685, __pyx_L1_error)
+    __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_range_internal64(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_14); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 744, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-    /* "pgenlib.pyx":684
+    /* "pgenlib.pyx":743
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_range_internal32(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:             # <<<<<<<<<<<<<<
  *             self.read_range_internal64(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         else:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":687
+  /* "pgenlib.pyx":746
  *             self.read_range_internal64(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_range() geno_int_out array element type (int8, int32, or int64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__11, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 687, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__11, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 746, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 687, __pyx_L1_error)
+    __PYX_ERR(0, 746, __pyx_L1_error)
   }
   __pyx_L4:;
 
-  /* "pgenlib.pyx":688
+  /* "pgenlib.pyx":747
  *         else:
  *             raise RuntimeError("Invalid read_range() geno_int_out array element type (int8, int32, or int64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":677
+  /* "pgenlib.pyx":736
  *         return
  * 
  *     cpdef read_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -10692,21 +11161,21 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_start)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_end)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_range", 0, 3, 5, 1); __PYX_ERR(0, 677, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_range", 0, 3, 5, 1); __PYX_ERR(0, 736, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_geno_int_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_range", 0, 3, 5, 2); __PYX_ERR(0, 677, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_range", 0, 3, 5, 2); __PYX_ERR(0, 736, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[3] = value; kw_args--; }
         }
@@ -10714,52 +11183,52 @@
         case  4:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_maj);
           if (value) { values[4] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_range") < 0)) __PYX_ERR(0, 677, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_range") < 0)) __PYX_ERR(0, 736, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
         CYTHON_FALLTHROUGH;
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 677, __pyx_L3_error)
-    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 677, __pyx_L3_error)
+    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 736, __pyx_L3_error)
+    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 736, __pyx_L3_error)
     __pyx_v_geno_int_out = ((PyArrayObject *)values[2]);
     if (values[3]) {
-      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[3]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 677, __pyx_L3_error)
+      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[3]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 736, __pyx_L3_error)
     } else {
       __pyx_v_allele_idx = ((uint32_t)1);
     }
     if (values[4]) {
-      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[4]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 677, __pyx_L3_error)
+      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[4]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 736, __pyx_L3_error)
     } else {
       __pyx_v_sample_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_range", 0, 3, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 677, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_range", 0, 3, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 736, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_range", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int_out), __pyx_ptype_5numpy_ndarray, 1, "geno_int_out", 0))) __PYX_ERR(0, 677, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int_out), __pyx_ptype_5numpy_ndarray, 1, "geno_int_out", 0))) __PYX_ERR(0, 736, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_18read_range(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_geno_int_out, __pyx_v_allele_idx, __pyx_v_sample_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -10776,15 +11245,15 @@
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("read_range", 0);
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 2;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
   __pyx_t_2.sample_maj = __pyx_v_sample_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_geno_int_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 677, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_geno_int_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 736, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -10793,15 +11262,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":691
+/* "pgenlib.pyx":750
  * 
  * 
  *     cdef read_list_internal8(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -10878,349 +11347,349 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_geno_int8_out.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int8_out.refcount = 0;
   __pyx_pybuffernd_geno_int8_out.data = NULL;
   __pyx_pybuffernd_geno_int8_out.rcbuffer = &__pyx_pybuffer_geno_int8_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 691, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 750, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 691, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 750, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int8_out.diminfo[0].strides = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int8_out.diminfo[0].shape = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int8_out.diminfo[1].strides = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int8_out.diminfo[1].shape = __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":692
+  /* "pgenlib.pyx":751
  * 
  *     cdef read_list_internal8(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_1;
 
-  /* "pgenlib.pyx":693
+  /* "pgenlib.pyx":752
  *     cdef read_list_internal8(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_2 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_2;
 
-  /* "pgenlib.pyx":694
+  /* "pgenlib.pyx":753
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_3 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_3;
 
-  /* "pgenlib.pyx":695
+  /* "pgenlib.pyx":754
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
   __pyx_t_4 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_4;
 
-  /* "pgenlib.pyx":696
+  /* "pgenlib.pyx":755
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_2 = __pyx_v_self->_genovec;
+  __pyx_t_2 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_2;
 
-  /* "pgenlib.pyx":697
+  /* "pgenlib.pyx":756
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int8_t* data_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":698
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":757
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int8_t* data_ptr
  *         cdef uint32_t variant_list_idx
  */
   __pyx_t_1 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_1;
 
-  /* "pgenlib.pyx":703
+  /* "pgenlib.pyx":762
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_5 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":704
+    /* "pgenlib.pyx":763
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int8_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:
  */
     __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":705
+      /* "pgenlib.pyx":764
  *         if sample_maj == 0:
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if geno_int8_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 705, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 764, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 705, __pyx_L1_error)
+      __PYX_ERR(0, 764, __pyx_L1_error)
 
-      /* "pgenlib.pyx":704
+      /* "pgenlib.pyx":763
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int8_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":706
+    /* "pgenlib.pyx":765
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":707
+      /* "pgenlib.pyx":766
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 707, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 707, __pyx_L1_error)
+      __PYX_ERR(0, 766, __pyx_L1_error)
 
-      /* "pgenlib.pyx":706
+      /* "pgenlib.pyx":765
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int8_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":708
+    /* "pgenlib.pyx":767
  *             if geno_int8_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_1 = __pyx_v_variant_idx_ct;
     __pyx_t_9 = __pyx_t_1;
     for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
       __pyx_v_variant_list_idx = __pyx_t_10;
 
-      /* "pgenlib.pyx":709
+      /* "pgenlib.pyx":768
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_11 = __pyx_v_variant_list_idx;
       __pyx_t_12 = -1;
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_12 = 0;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 709, __pyx_L1_error)
+        __PYX_ERR(0, 768, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":710
+      /* "pgenlib.pyx":769
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
       __pyx_t_5 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":711
+        /* "pgenlib.pyx":770
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 711, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 770, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_Raise(__pyx_t_6, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __PYX_ERR(0, 711, __pyx_L1_error)
+        __PYX_ERR(0, 770, __pyx_L1_error)
 
-        /* "pgenlib.pyx":710
+        /* "pgenlib.pyx":769
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
       }
 
-      /* "pgenlib.pyx":712
+      /* "pgenlib.pyx":771
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec);
 
-      /* "pgenlib.pyx":713
+      /* "pgenlib.pyx":772
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[variant_list_idx, 0])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":714
+        /* "pgenlib.pyx":773
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data_ptr = &(geno_int8_out[variant_list_idx, 0])
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  */
-        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 714, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 773, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 714, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 773, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 714, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 773, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 714, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 773, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 714, __pyx_L1_error)
+        __PYX_ERR(0, 773, __pyx_L1_error)
 
-        /* "pgenlib.pyx":713
+        /* "pgenlib.pyx":772
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[variant_list_idx, 0])
  */
       }
 
-      /* "pgenlib.pyx":715
+      /* "pgenlib.pyx":774
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[variant_list_idx, 0])             # <<<<<<<<<<<<<<
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return
  */
       __pyx_t_11 = __pyx_v_variant_list_idx;
@@ -11229,595 +11698,595 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int8_out.diminfo[0].shape)) __pyx_t_12 = 0;
       if (__pyx_t_13 < 0) {
         __pyx_t_13 += __pyx_pybuffernd_geno_int8_out.diminfo[1].shape;
         if (unlikely(__pyx_t_13 < 0)) __pyx_t_12 = 1;
       } else if (unlikely(__pyx_t_13 >= __pyx_pybuffernd_geno_int8_out.diminfo[1].shape)) __pyx_t_12 = 1;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 715, __pyx_L1_error)
+        __PYX_ERR(0, 774, __pyx_L1_error)
       }
       __pyx_v_data_ptr = (&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int8_t *, __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int8_out.diminfo[0].strides, __pyx_t_13, __pyx_pybuffernd_geno_int8_out.diminfo[1].strides)));
 
-      /* "pgenlib.pyx":716
+      /* "pgenlib.pyx":775
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int8_out[variant_list_idx, 0])
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if geno_int8_out.shape[0] < subset_size:
  */
       plink2::GenoarrToBytesMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
     }
 
-    /* "pgenlib.pyx":717
+    /* "pgenlib.pyx":776
  *                 data_ptr = &(geno_int8_out[variant_list_idx, 0])
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":703
+    /* "pgenlib.pyx":762
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int8_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":718
+  /* "pgenlib.pyx":777
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if geno_int8_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":719
+    /* "pgenlib.pyx":778
  *             return
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[0])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 719, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 778, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_Raise(__pyx_t_8, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 719, __pyx_L1_error)
+    __PYX_ERR(0, 778, __pyx_L1_error)
 
-    /* "pgenlib.pyx":718
+    /* "pgenlib.pyx":777
  *                 GenoarrToBytesMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if geno_int8_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":720
+  /* "pgenlib.pyx":779
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_5 = (((__pyx_v_geno_int8_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":721
+    /* "pgenlib.pyx":780
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int8_out->dimensions[1])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 721, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 780, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_8, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 721, __pyx_L1_error)
+    __PYX_ERR(0, 780, __pyx_L1_error)
 
-    /* "pgenlib.pyx":720
+    /* "pgenlib.pyx":779
  *         if geno_int8_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int8_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int8_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":722
+  /* "pgenlib.pyx":781
  *         if geno_int8_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":723
+  /* "pgenlib.pyx":782
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int8_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":724
+  /* "pgenlib.pyx":783
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":725
+  /* "pgenlib.pyx":784
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":726
+  /* "pgenlib.pyx":785
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_14 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_14;
 
-  /* "pgenlib.pyx":727
+  /* "pgenlib.pyx":786
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* vmaj_iter
  */
   __pyx_t_2 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_2;
 
-  /* "pgenlib.pyx":728
+  /* "pgenlib.pyx":787
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_iter
  *         cdef uintptr_t* smaj_iter
  */
   __pyx_t_2 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_2;
 
-  /* "pgenlib.pyx":735
+  /* "pgenlib.pyx":794
  *         cdef uint32_t sample_batch_idx
  *         cdef uint32_t uii
  *         variant_list_idx = 0             # <<<<<<<<<<<<<<
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  */
   __pyx_v_variant_list_idx = 0;
 
-  /* "pgenlib.pyx":736
+  /* "pgenlib.pyx":795
  *         cdef uint32_t uii
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_1 = __pyx_v_variant_batch_ct;
   __pyx_t_9 = __pyx_t_1;
   for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_10;
 
-    /* "pgenlib.pyx":737
+    /* "pgenlib.pyx":796
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     __pyx_t_5 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":738
+      /* "pgenlib.pyx":797
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  */
       __pyx_t_15 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 738, __pyx_L1_error)
+        __PYX_ERR(0, 797, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":737
+      /* "pgenlib.pyx":796
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     }
 
-    /* "pgenlib.pyx":739
+    /* "pgenlib.pyx":798
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":740
+    /* "pgenlib.pyx":799
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_16 = __pyx_v_variant_batch_size;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_uii = __pyx_t_18;
 
-      /* "pgenlib.pyx":741
+      /* "pgenlib.pyx":800
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_11 = (__pyx_v_uii + __pyx_v_variant_list_idx);
       __pyx_t_12 = -1;
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_12 = 0;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 741, __pyx_L1_error)
+        __PYX_ERR(0, 800, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":742
+      /* "pgenlib.pyx":801
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  */
       __pyx_t_5 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":743
+        /* "pgenlib.pyx":802
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_only); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_only); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 743, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 802, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 743, __pyx_L1_error)
+        __PYX_ERR(0, 802, __pyx_L1_error)
 
-        /* "pgenlib.pyx":742
+        /* "pgenlib.pyx":801
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  */
       }
 
-      /* "pgenlib.pyx":744
+      /* "pgenlib.pyx":803
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_vmaj_iter);
 
-      /* "pgenlib.pyx":745
+      /* "pgenlib.pyx":804
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":746
+        /* "pgenlib.pyx":805
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  */
-        __pyx_t_8 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 746, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 805, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 746, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 805, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_error, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 746, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_error, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 805, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 746, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 805, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 746, __pyx_L1_error)
+        __PYX_ERR(0, 805, __pyx_L1_error)
 
-        /* "pgenlib.pyx":745
+        /* "pgenlib.pyx":804
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       }
 
-      /* "pgenlib.pyx":747
+      /* "pgenlib.pyx":806
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[__pyx_v_sample_ctaw2]));
     }
 
-    /* "pgenlib.pyx":748
+    /* "pgenlib.pyx":807
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":749
+    /* "pgenlib.pyx":808
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":750
+    /* "pgenlib.pyx":809
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_16 = __pyx_v_sample_batch_ct;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_18;
 
-      /* "pgenlib.pyx":751
+      /* "pgenlib.pyx":810
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_5 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":752
+        /* "pgenlib.pyx":811
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  */
         __pyx_t_15 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 752, __pyx_L1_error)
+          __PYX_ERR(0, 811, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":751
+        /* "pgenlib.pyx":810
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":753
+      /* "pgenlib.pyx":812
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  */
       __pyx_v_smaj_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":754
+      /* "pgenlib.pyx":813
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":755
+      /* "pgenlib.pyx":814
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  */
       __pyx_t_19 = __pyx_v_sample_batch_size;
       __pyx_t_20 = __pyx_t_19;
       for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
         __pyx_v_uii = __pyx_t_21;
 
-        /* "pgenlib.pyx":756
+        /* "pgenlib.pyx":815
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])             # <<<<<<<<<<<<<<
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  */
         __pyx_t_11 = (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch));
         __pyx_t_22 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch);
         __pyx_t_12 = -1;
         if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int8_out.diminfo[0].shape)) __pyx_t_12 = 0;
         if (unlikely(__pyx_t_22 >= (size_t)__pyx_pybuffernd_geno_int8_out.diminfo[1].shape)) __pyx_t_12 = 1;
         if (unlikely(__pyx_t_12 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_12);
-          __PYX_ERR(0, 756, __pyx_L1_error)
+          __PYX_ERR(0, 815, __pyx_L1_error)
         }
         __pyx_v_data_ptr = (&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int8_t *, __pyx_pybuffernd_geno_int8_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int8_out.diminfo[0].strides, __pyx_t_22, __pyx_pybuffernd_geno_int8_out.diminfo[1].strides)));
 
-        /* "pgenlib.pyx":757
+        /* "pgenlib.pyx":816
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  */
         plink2::GenoarrToBytesMinus9(__pyx_v_smaj_iter, __pyx_v_variant_batch_size, __pyx_v_data_ptr);
 
-        /* "pgenlib.pyx":758
+        /* "pgenlib.pyx":817
  *                     data_ptr = &(geno_int8_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch
  */
         __pyx_v_smaj_iter = (&(__pyx_v_smaj_iter[plink2::kPglNypTransposeWords]));
       }
 
-      /* "pgenlib.pyx":759
+      /* "pgenlib.pyx":818
  *                     GenoarrToBytesMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *             variant_list_idx += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[plink2::kPglNypTransposeWords]));
     }
 
-    /* "pgenlib.pyx":760
+    /* "pgenlib.pyx":819
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_list_idx = (__pyx_v_variant_list_idx + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":761
+  /* "pgenlib.pyx":820
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":691
+  /* "pgenlib.pyx":750
  * 
  * 
  *     cdef read_list_internal8(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -11841,15 +12310,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":763
+/* "pgenlib.pyx":822
  *         return
  * 
  *     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -11926,349 +12395,349 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_geno_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int32_out.refcount = 0;
   __pyx_pybuffernd_geno_int32_out.data = NULL;
   __pyx_pybuffernd_geno_int32_out.rcbuffer = &__pyx_pybuffer_geno_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 763, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 822, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 763, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 822, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int32_out.diminfo[0].strides = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int32_out.diminfo[0].shape = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int32_out.diminfo[1].strides = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int32_out.diminfo[1].shape = __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":764
+  /* "pgenlib.pyx":823
  * 
  *     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_1;
 
-  /* "pgenlib.pyx":765
+  /* "pgenlib.pyx":824
  *     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_2 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_2;
 
-  /* "pgenlib.pyx":766
+  /* "pgenlib.pyx":825
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_3 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_3;
 
-  /* "pgenlib.pyx":767
+  /* "pgenlib.pyx":826
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
   __pyx_t_4 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_4;
 
-  /* "pgenlib.pyx":768
+  /* "pgenlib.pyx":827
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_2 = __pyx_v_self->_genovec;
+  __pyx_t_2 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_2;
 
-  /* "pgenlib.pyx":769
+  /* "pgenlib.pyx":828
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int32_t* data_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":770
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":829
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int32_t* data_ptr
  *         cdef uint32_t variant_list_idx
  */
   __pyx_t_1 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_1;
 
-  /* "pgenlib.pyx":775
+  /* "pgenlib.pyx":834
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_5 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":776
+    /* "pgenlib.pyx":835
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:
  */
     __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":777
+      /* "pgenlib.pyx":836
  *         if sample_maj == 0:
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if geno_int32_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 777, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 836, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 777, __pyx_L1_error)
+      __PYX_ERR(0, 836, __pyx_L1_error)
 
-      /* "pgenlib.pyx":776
+      /* "pgenlib.pyx":835
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":778
+    /* "pgenlib.pyx":837
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":779
+      /* "pgenlib.pyx":838
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 779, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 838, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 779, __pyx_L1_error)
+      __PYX_ERR(0, 838, __pyx_L1_error)
 
-      /* "pgenlib.pyx":778
+      /* "pgenlib.pyx":837
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int32_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":780
+    /* "pgenlib.pyx":839
  *             if geno_int32_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_1 = __pyx_v_variant_idx_ct;
     __pyx_t_9 = __pyx_t_1;
     for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
       __pyx_v_variant_list_idx = __pyx_t_10;
 
-      /* "pgenlib.pyx":781
+      /* "pgenlib.pyx":840
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_11 = __pyx_v_variant_list_idx;
       __pyx_t_12 = -1;
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_12 = 0;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 781, __pyx_L1_error)
+        __PYX_ERR(0, 840, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":782
+      /* "pgenlib.pyx":841
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
       __pyx_t_5 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":783
+        /* "pgenlib.pyx":842
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 783, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_Raise(__pyx_t_6, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __PYX_ERR(0, 783, __pyx_L1_error)
+        __PYX_ERR(0, 842, __pyx_L1_error)
 
-        /* "pgenlib.pyx":782
+        /* "pgenlib.pyx":841
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
       }
 
-      /* "pgenlib.pyx":784
+      /* "pgenlib.pyx":843
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec);
 
-      /* "pgenlib.pyx":785
+      /* "pgenlib.pyx":844
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[variant_list_idx, 0]))
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":786
+        /* "pgenlib.pyx":845
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data_ptr = <int32_t*>(&(geno_int32_out[variant_list_idx, 0]))
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  */
-        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 786, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 845, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 786, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 845, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 786, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 845, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 786, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 845, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 786, __pyx_L1_error)
+        __PYX_ERR(0, 845, __pyx_L1_error)
 
-        /* "pgenlib.pyx":785
+        /* "pgenlib.pyx":844
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[variant_list_idx, 0]))
  */
       }
 
-      /* "pgenlib.pyx":787
+      /* "pgenlib.pyx":846
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[variant_list_idx, 0]))             # <<<<<<<<<<<<<<
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return
  */
       __pyx_t_11 = __pyx_v_variant_list_idx;
@@ -12277,595 +12746,595 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int32_out.diminfo[0].shape)) __pyx_t_12 = 0;
       if (__pyx_t_13 < 0) {
         __pyx_t_13 += __pyx_pybuffernd_geno_int32_out.diminfo[1].shape;
         if (unlikely(__pyx_t_13 < 0)) __pyx_t_12 = 1;
       } else if (unlikely(__pyx_t_13 >= __pyx_pybuffernd_geno_int32_out.diminfo[1].shape)) __pyx_t_12 = 1;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 787, __pyx_L1_error)
+        __PYX_ERR(0, 846, __pyx_L1_error)
       }
       __pyx_v_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int32_out.diminfo[0].strides, __pyx_t_13, __pyx_pybuffernd_geno_int32_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":788
+      /* "pgenlib.pyx":847
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = <int32_t*>(&(geno_int32_out[variant_list_idx, 0]))
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if geno_int32_out.shape[0] < subset_size:
  */
       plink2::GenoarrToInt32sMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
     }
 
-    /* "pgenlib.pyx":789
+    /* "pgenlib.pyx":848
  *                 data_ptr = <int32_t*>(&(geno_int32_out[variant_list_idx, 0]))
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":775
+    /* "pgenlib.pyx":834
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":790
+  /* "pgenlib.pyx":849
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if geno_int32_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":791
+    /* "pgenlib.pyx":850
  *             return
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[0])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 791, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 850, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_Raise(__pyx_t_8, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 791, __pyx_L1_error)
+    __PYX_ERR(0, 850, __pyx_L1_error)
 
-    /* "pgenlib.pyx":790
+    /* "pgenlib.pyx":849
  *                 GenoarrToInt32sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if geno_int32_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":792
+  /* "pgenlib.pyx":851
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_5 = (((__pyx_v_geno_int32_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":793
+    /* "pgenlib.pyx":852
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int32_out->dimensions[1])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 793, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 852, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_8, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 793, __pyx_L1_error)
+    __PYX_ERR(0, 852, __pyx_L1_error)
 
-    /* "pgenlib.pyx":792
+    /* "pgenlib.pyx":851
  *         if geno_int32_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":794
+  /* "pgenlib.pyx":853
  *         if geno_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":795
+  /* "pgenlib.pyx":854
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":796
+  /* "pgenlib.pyx":855
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":797
+  /* "pgenlib.pyx":856
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":798
+  /* "pgenlib.pyx":857
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_14 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_14;
 
-  /* "pgenlib.pyx":799
+  /* "pgenlib.pyx":858
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* vmaj_iter
  */
   __pyx_t_2 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_2;
 
-  /* "pgenlib.pyx":800
+  /* "pgenlib.pyx":859
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_iter
  *         cdef uintptr_t* smaj_iter
  */
   __pyx_t_2 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_2;
 
-  /* "pgenlib.pyx":807
+  /* "pgenlib.pyx":866
  *         cdef uint32_t sample_batch_idx
  *         cdef uint32_t uii
  *         variant_list_idx = 0             # <<<<<<<<<<<<<<
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  */
   __pyx_v_variant_list_idx = 0;
 
-  /* "pgenlib.pyx":808
+  /* "pgenlib.pyx":867
  *         cdef uint32_t uii
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_1 = __pyx_v_variant_batch_ct;
   __pyx_t_9 = __pyx_t_1;
   for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_10;
 
-    /* "pgenlib.pyx":809
+    /* "pgenlib.pyx":868
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     __pyx_t_5 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":810
+      /* "pgenlib.pyx":869
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  */
       __pyx_t_15 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 810, __pyx_L1_error)
+        __PYX_ERR(0, 869, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":809
+      /* "pgenlib.pyx":868
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     }
 
-    /* "pgenlib.pyx":811
+    /* "pgenlib.pyx":870
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":812
+    /* "pgenlib.pyx":871
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_16 = __pyx_v_variant_batch_size;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_uii = __pyx_t_18;
 
-      /* "pgenlib.pyx":813
+      /* "pgenlib.pyx":872
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_11 = (__pyx_v_uii + __pyx_v_variant_list_idx);
       __pyx_t_12 = -1;
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_12 = 0;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 813, __pyx_L1_error)
+        __PYX_ERR(0, 872, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":814
+      /* "pgenlib.pyx":873
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  */
       __pyx_t_5 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":815
+        /* "pgenlib.pyx":874
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_only); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_only); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 815, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 874, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 815, __pyx_L1_error)
+        __PYX_ERR(0, 874, __pyx_L1_error)
 
-        /* "pgenlib.pyx":814
+        /* "pgenlib.pyx":873
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  */
       }
 
-      /* "pgenlib.pyx":816
+      /* "pgenlib.pyx":875
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_vmaj_iter);
 
-      /* "pgenlib.pyx":817
+      /* "pgenlib.pyx":876
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":818
+        /* "pgenlib.pyx":877
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  */
-        __pyx_t_8 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 818, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 877, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 818, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 877, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_error, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 818, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_error, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 877, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 818, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 877, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 818, __pyx_L1_error)
+        __PYX_ERR(0, 877, __pyx_L1_error)
 
-        /* "pgenlib.pyx":817
+        /* "pgenlib.pyx":876
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       }
 
-      /* "pgenlib.pyx":819
+      /* "pgenlib.pyx":878
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[__pyx_v_sample_ctaw2]));
     }
 
-    /* "pgenlib.pyx":820
+    /* "pgenlib.pyx":879
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":821
+    /* "pgenlib.pyx":880
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":822
+    /* "pgenlib.pyx":881
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_16 = __pyx_v_sample_batch_ct;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_18;
 
-      /* "pgenlib.pyx":823
+      /* "pgenlib.pyx":882
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_5 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":824
+        /* "pgenlib.pyx":883
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  */
         __pyx_t_15 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 824, __pyx_L1_error)
+          __PYX_ERR(0, 883, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":823
+        /* "pgenlib.pyx":882
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":825
+      /* "pgenlib.pyx":884
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  */
       __pyx_v_smaj_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":826
+      /* "pgenlib.pyx":885
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":827
+      /* "pgenlib.pyx":886
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  */
       __pyx_t_19 = __pyx_v_sample_batch_size;
       __pyx_t_20 = __pyx_t_19;
       for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
         __pyx_v_uii = __pyx_t_21;
 
-        /* "pgenlib.pyx":828
+        /* "pgenlib.pyx":887
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))             # <<<<<<<<<<<<<<
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  */
         __pyx_t_11 = (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch));
         __pyx_t_22 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch);
         __pyx_t_12 = -1;
         if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int32_out.diminfo[0].shape)) __pyx_t_12 = 0;
         if (unlikely(__pyx_t_22 >= (size_t)__pyx_pybuffernd_geno_int32_out.diminfo[1].shape)) __pyx_t_12 = 1;
         if (unlikely(__pyx_t_12 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_12);
-          __PYX_ERR(0, 828, __pyx_L1_error)
+          __PYX_ERR(0, 887, __pyx_L1_error)
         }
         __pyx_v_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_geno_int32_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int32_out.diminfo[0].strides, __pyx_t_22, __pyx_pybuffernd_geno_int32_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":829
+        /* "pgenlib.pyx":888
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  */
         plink2::GenoarrToInt32sMinus9(__pyx_v_smaj_iter, __pyx_v_variant_batch_size, __pyx_v_data_ptr);
 
-        /* "pgenlib.pyx":830
+        /* "pgenlib.pyx":889
  *                     data_ptr = <int32_t*>(&(geno_int32_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch
  */
         __pyx_v_smaj_iter = (&(__pyx_v_smaj_iter[plink2::kPglNypTransposeWords]));
       }
 
-      /* "pgenlib.pyx":831
+      /* "pgenlib.pyx":890
  *                     GenoarrToInt32sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *             variant_list_idx += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[plink2::kPglNypTransposeWords]));
     }
 
-    /* "pgenlib.pyx":832
+    /* "pgenlib.pyx":891
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_list_idx = (__pyx_v_variant_list_idx + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":833
+  /* "pgenlib.pyx":892
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":763
+  /* "pgenlib.pyx":822
  *         return
  * 
  *     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -12889,15 +13358,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":835
+/* "pgenlib.pyx":894
  *         return
  * 
  *     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -12974,349 +13443,349 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_geno_int64_out.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int64_out.refcount = 0;
   __pyx_pybuffernd_geno_int64_out.data = NULL;
   __pyx_pybuffernd_geno_int64_out.rcbuffer = &__pyx_pybuffer_geno_int64_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 835, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 894, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int64_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 835, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int64_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 894, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int64_out.diminfo[0].strides = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int64_out.diminfo[0].shape = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int64_out.diminfo[1].strides = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int64_out.diminfo[1].shape = __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":836
+  /* "pgenlib.pyx":895
  * 
  *     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_1;
 
-  /* "pgenlib.pyx":837
+  /* "pgenlib.pyx":896
  *     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_2 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_2;
 
-  /* "pgenlib.pyx":838
+  /* "pgenlib.pyx":897
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_3 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_3;
 
-  /* "pgenlib.pyx":839
+  /* "pgenlib.pyx":898
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
   __pyx_t_4 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_4;
 
-  /* "pgenlib.pyx":840
+  /* "pgenlib.pyx":899
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_2 = __pyx_v_self->_genovec;
+  __pyx_t_2 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_2;
 
-  /* "pgenlib.pyx":841
+  /* "pgenlib.pyx":900
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int64_t* data_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":842
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":901
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int64_t* data_ptr
  *         cdef uint32_t variant_list_idx
  */
   __pyx_t_1 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_1;
 
-  /* "pgenlib.pyx":847
+  /* "pgenlib.pyx":906
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_5 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":848
+    /* "pgenlib.pyx":907
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int64_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:
  */
     __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":849
+      /* "pgenlib.pyx":908
  *         if sample_maj == 0:
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if geno_int64_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 849, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 908, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 849, __pyx_L1_error)
+      __PYX_ERR(0, 908, __pyx_L1_error)
 
-      /* "pgenlib.pyx":848
+      /* "pgenlib.pyx":907
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if geno_int64_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":850
+    /* "pgenlib.pyx":909
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":851
+      /* "pgenlib.pyx":910
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 851, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 910, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
       __Pyx_Raise(__pyx_t_6, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __PYX_ERR(0, 851, __pyx_L1_error)
+      __PYX_ERR(0, 910, __pyx_L1_error)
 
-      /* "pgenlib.pyx":850
+      /* "pgenlib.pyx":909
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if geno_int64_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":852
+    /* "pgenlib.pyx":911
  *             if geno_int64_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_1 = __pyx_v_variant_idx_ct;
     __pyx_t_9 = __pyx_t_1;
     for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
       __pyx_v_variant_list_idx = __pyx_t_10;
 
-      /* "pgenlib.pyx":853
+      /* "pgenlib.pyx":912
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_11 = __pyx_v_variant_list_idx;
       __pyx_t_12 = -1;
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_12 = 0;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 853, __pyx_L1_error)
+        __PYX_ERR(0, 912, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":854
+      /* "pgenlib.pyx":913
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
       __pyx_t_5 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":855
+        /* "pgenlib.pyx":914
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_only); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 855, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 914, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_Raise(__pyx_t_6, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __PYX_ERR(0, 855, __pyx_L1_error)
+        __PYX_ERR(0, 914, __pyx_L1_error)
 
-        /* "pgenlib.pyx":854
+        /* "pgenlib.pyx":913
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  */
       }
 
-      /* "pgenlib.pyx":856
+      /* "pgenlib.pyx":915
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec);
 
-      /* "pgenlib.pyx":857
+      /* "pgenlib.pyx":916
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[variant_list_idx, 0])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":858
+        /* "pgenlib.pyx":917
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data_ptr = &(geno_int64_out[variant_list_idx, 0])
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  */
-        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 858, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 917, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 858, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 917, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 858, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 917, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 858, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 917, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 858, __pyx_L1_error)
+        __PYX_ERR(0, 917, __pyx_L1_error)
 
-        /* "pgenlib.pyx":857
+        /* "pgenlib.pyx":916
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[variant_list_idx, 0])
  */
       }
 
-      /* "pgenlib.pyx":859
+      /* "pgenlib.pyx":918
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[variant_list_idx, 0])             # <<<<<<<<<<<<<<
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return
  */
       __pyx_t_11 = __pyx_v_variant_list_idx;
@@ -13325,595 +13794,595 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int64_out.diminfo[0].shape)) __pyx_t_12 = 0;
       if (__pyx_t_13 < 0) {
         __pyx_t_13 += __pyx_pybuffernd_geno_int64_out.diminfo[1].shape;
         if (unlikely(__pyx_t_13 < 0)) __pyx_t_12 = 1;
       } else if (unlikely(__pyx_t_13 >= __pyx_pybuffernd_geno_int64_out.diminfo[1].shape)) __pyx_t_12 = 1;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 859, __pyx_L1_error)
+        __PYX_ERR(0, 918, __pyx_L1_error)
       }
       __pyx_v_data_ptr = ((int64_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int64_t *, __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int64_out.diminfo[0].strides, __pyx_t_13, __pyx_pybuffernd_geno_int64_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":860
+      /* "pgenlib.pyx":919
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data_ptr = &(geno_int64_out[variant_list_idx, 0])
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if geno_int64_out.shape[0] < subset_size:
  */
       plink2::GenoarrToInt64sMinus9(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
     }
 
-    /* "pgenlib.pyx":861
+    /* "pgenlib.pyx":920
  *                 data_ptr = &(geno_int64_out[variant_list_idx, 0])
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":847
+    /* "pgenlib.pyx":906
  *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if geno_int64_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":862
+  /* "pgenlib.pyx":921
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if geno_int64_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[0]) < __pyx_v_subset_size) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":863
+    /* "pgenlib.pyx":922
  *             return
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[0])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 863, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 922, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_Raise(__pyx_t_8, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 863, __pyx_L1_error)
+    __PYX_ERR(0, 922, __pyx_L1_error)
 
-    /* "pgenlib.pyx":862
+    /* "pgenlib.pyx":921
  *                 GenoarrToInt64sMinus9(genovec, subset_size, data_ptr)
  *             return
  *         if geno_int64_out.shape[0] < subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":864
+  /* "pgenlib.pyx":923
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_5 = (((__pyx_v_geno_int64_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":865
+    /* "pgenlib.pyx":924
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_geno_int64_out->dimensions[1])); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 865, __pyx_L1_error)
+    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 924, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_8);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_8, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 865, __pyx_L1_error)
+    __PYX_ERR(0, 924, __pyx_L1_error)
 
-    /* "pgenlib.pyx":864
+    /* "pgenlib.pyx":923
  *         if geno_int64_out.shape[0] < subset_size:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few rows (" + str(geno_int64_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ")")
  *         if geno_int64_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":866
+  /* "pgenlib.pyx":925
  *         if geno_int64_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":867
+  /* "pgenlib.pyx":926
  *             raise RuntimeError("Sample-major read_list() geno_int_out buffer has too few columns (" + str(geno_int64_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":868
+  /* "pgenlib.pyx":927
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":869
+  /* "pgenlib.pyx":928
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":870
+  /* "pgenlib.pyx":929
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_14 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_14;
 
-  /* "pgenlib.pyx":871
+  /* "pgenlib.pyx":930
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* vmaj_iter
  */
   __pyx_t_2 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_2;
 
-  /* "pgenlib.pyx":872
+  /* "pgenlib.pyx":931
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_iter
  *         cdef uintptr_t* smaj_iter
  */
   __pyx_t_2 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_2;
 
-  /* "pgenlib.pyx":879
+  /* "pgenlib.pyx":938
  *         cdef uint32_t sample_batch_idx
  *         cdef uint32_t uii
  *         variant_list_idx = 0             # <<<<<<<<<<<<<<
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  */
   __pyx_v_variant_list_idx = 0;
 
-  /* "pgenlib.pyx":880
+  /* "pgenlib.pyx":939
  *         cdef uint32_t uii
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_1 = __pyx_v_variant_batch_ct;
   __pyx_t_9 = __pyx_t_1;
   for (__pyx_t_10 = 0; __pyx_t_10 < __pyx_t_9; __pyx_t_10+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_10;
 
-    /* "pgenlib.pyx":881
+    /* "pgenlib.pyx":940
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     __pyx_t_5 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":882
+      /* "pgenlib.pyx":941
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  */
       __pyx_t_15 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 882, __pyx_L1_error)
+        __PYX_ERR(0, 941, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":881
+      /* "pgenlib.pyx":940
  *         variant_list_idx = 0
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
     }
 
-    /* "pgenlib.pyx":883
+    /* "pgenlib.pyx":942
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":884
+    /* "pgenlib.pyx":943
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_16 = __pyx_v_variant_batch_size;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_uii = __pyx_t_18;
 
-      /* "pgenlib.pyx":885
+      /* "pgenlib.pyx":944
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_11 = (__pyx_v_uii + __pyx_v_variant_list_idx);
       __pyx_t_12 = -1;
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_12 = 0;
       if (unlikely(__pyx_t_12 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_12);
-        __PYX_ERR(0, 885, __pyx_L1_error)
+        __PYX_ERR(0, 944, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":886
+      /* "pgenlib.pyx":945
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  */
       __pyx_t_5 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":887
+        /* "pgenlib.pyx":946
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_variant_index_too_larg, __pyx_t_6); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_only); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_only); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 887, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 946, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_8, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 887, __pyx_L1_error)
+        __PYX_ERR(0, 946, __pyx_L1_error)
 
-        /* "pgenlib.pyx":886
+        /* "pgenlib.pyx":945
  *             for uii in range(variant_batch_size):
  *                 variant_idx = variant_idxs[uii + variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  */
       }
 
-      /* "pgenlib.pyx":888
+      /* "pgenlib.pyx":947
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_vmaj_iter);
 
-      /* "pgenlib.pyx":889
+      /* "pgenlib.pyx":948
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_5)) {
 
-        /* "pgenlib.pyx":890
+        /* "pgenlib.pyx":949
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  */
-        __pyx_t_8 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 890, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 949, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 890, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 949, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_error, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 890, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_list_error, __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 949, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 890, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 949, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 890, __pyx_L1_error)
+        __PYX_ERR(0, 949, __pyx_L1_error)
 
-        /* "pgenlib.pyx":889
+        /* "pgenlib.pyx":948
  *                     raise RuntimeError("read_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, vmaj_iter)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  */
       }
 
-      /* "pgenlib.pyx":891
+      /* "pgenlib.pyx":950
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[__pyx_v_sample_ctaw2]));
     }
 
-    /* "pgenlib.pyx":892
+    /* "pgenlib.pyx":951
  *                     raise RuntimeError("read_list() error " + str(reterr))
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":893
+    /* "pgenlib.pyx":952
  *                 vmaj_iter = &(vmaj_iter[sample_ctaw2])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":894
+    /* "pgenlib.pyx":953
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_16 = __pyx_v_sample_batch_ct;
     __pyx_t_17 = __pyx_t_16;
     for (__pyx_t_18 = 0; __pyx_t_18 < __pyx_t_17; __pyx_t_18+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_18;
 
-      /* "pgenlib.pyx":895
+      /* "pgenlib.pyx":954
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_5 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":896
+        /* "pgenlib.pyx":955
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  */
         __pyx_t_15 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 896, __pyx_L1_error)
+          __PYX_ERR(0, 955, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_15, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":895
+        /* "pgenlib.pyx":954
  *             vmaj_iter = multivar_vmaj_geno_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":897
+      /* "pgenlib.pyx":956
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  */
       __pyx_v_smaj_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":898
+      /* "pgenlib.pyx":957
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":899
+      /* "pgenlib.pyx":958
  *                 smaj_iter = multivar_smaj_geno_batch_buf
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  */
       __pyx_t_19 = __pyx_v_sample_batch_size;
       __pyx_t_20 = __pyx_t_19;
       for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
         __pyx_v_uii = __pyx_t_21;
 
-        /* "pgenlib.pyx":900
+        /* "pgenlib.pyx":959
  *                 TransposeNypblock(vmaj_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])             # <<<<<<<<<<<<<<
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  */
         __pyx_t_11 = (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch));
         __pyx_t_22 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch);
         __pyx_t_12 = -1;
         if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_geno_int64_out.diminfo[0].shape)) __pyx_t_12 = 0;
         if (unlikely(__pyx_t_22 >= (size_t)__pyx_pybuffernd_geno_int64_out.diminfo[1].shape)) __pyx_t_12 = 1;
         if (unlikely(__pyx_t_12 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_12);
-          __PYX_ERR(0, 900, __pyx_L1_error)
+          __PYX_ERR(0, 959, __pyx_L1_error)
         }
         __pyx_v_data_ptr = ((int64_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int64_t *, __pyx_pybuffernd_geno_int64_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_geno_int64_out.diminfo[0].strides, __pyx_t_22, __pyx_pybuffernd_geno_int64_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":901
+        /* "pgenlib.pyx":960
  *                 for uii in range(sample_batch_size):
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  */
         plink2::GenoarrToInt64sMinus9(__pyx_v_smaj_iter, __pyx_v_variant_batch_size, __pyx_v_data_ptr);
 
-        /* "pgenlib.pyx":902
+        /* "pgenlib.pyx":961
  *                     data_ptr = &(geno_int64_out[uii + sample_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch])
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch
  */
         __pyx_v_smaj_iter = (&(__pyx_v_smaj_iter[plink2::kPglNypTransposeWords]));
       }
 
-      /* "pgenlib.pyx":903
+      /* "pgenlib.pyx":962
  *                     GenoarrToInt64sMinus9(smaj_iter, variant_batch_size, data_ptr)
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *             variant_list_idx += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_iter = (&(__pyx_v_vmaj_iter[plink2::kPglNypTransposeWords]));
     }
 
-    /* "pgenlib.pyx":904
+    /* "pgenlib.pyx":963
  *                     smaj_iter = &(smaj_iter[kPglNypTransposeWords])
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_list_idx = (__pyx_v_variant_list_idx + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":905
+  /* "pgenlib.pyx":964
  *                 vmaj_iter = &(vmaj_iter[kPglNypTransposeWords])
  *             variant_list_idx += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":835
+  /* "pgenlib.pyx":894
  *         return
  * 
  *     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -13937,15 +14406,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":907
+/* "pgenlib.pyx":966
  *         return
  * 
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if geno_int_out.dtype == np.int8:
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  */
 
@@ -13983,33 +14452,33 @@
   }
   __pyx_pybuffer_variant_idxs.pybuffer.buf = NULL;
   __pyx_pybuffer_variant_idxs.refcount = 0;
   __pyx_pybuffernd_variant_idxs.data = NULL;
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 907, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 966, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_list); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 907, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_list); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 966, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_21read_list)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 907, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 966, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 907, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 966, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_5 = __pyx_t_1; __pyx_t_6 = NULL;
         __pyx_t_7 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_5))) {
           __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_5);
           if (likely(__pyx_t_6)) {
@@ -14019,33 +14488,33 @@
             __Pyx_DECREF_SET(__pyx_t_5, function);
             __pyx_t_7 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[5] = {__pyx_t_6, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_geno_int_out), __pyx_t_3, __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 907, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 966, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[5] = {__pyx_t_6, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_geno_int_out), __pyx_t_3, __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 907, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 966, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         {
-          __pyx_t_8 = PyTuple_New(4+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 907, __pyx_L1_error)
+          __pyx_t_8 = PyTuple_New(4+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 966, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_8);
           if (__pyx_t_6) {
             __Pyx_GIVEREF(__pyx_t_6); PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_6); __pyx_t_6 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_variant_idxs));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_variant_idxs));
           PyTuple_SET_ITEM(__pyx_t_8, 0+__pyx_t_7, ((PyObject *)__pyx_v_variant_idxs));
@@ -14054,15 +14523,15 @@
           PyTuple_SET_ITEM(__pyx_t_8, 1+__pyx_t_7, ((PyObject *)__pyx_v_geno_int_out));
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_8, 2+__pyx_t_7, __pyx_t_3);
           __Pyx_GIVEREF(__pyx_t_4);
           PyTuple_SET_ITEM(__pyx_t_8, 3+__pyx_t_7, __pyx_t_4);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 907, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 966, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         }
         __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -14077,177 +14546,177 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":908
+  /* "pgenlib.pyx":967
  * 
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if geno_int_out.dtype == np.int8:             # <<<<<<<<<<<<<<
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 908, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 967, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 908, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 967, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_int8); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 908, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_int8); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 967, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 908, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 967, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 908, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 967, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (__pyx_t_9) {
 
-    /* "pgenlib.pyx":909
+    /* "pgenlib.pyx":968
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if geno_int_out.dtype == np.int8:
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)
  */
     __pyx_t_10.__pyx_n = 2;
     __pyx_t_10.allele_idx = __pyx_v_allele_idx;
     __pyx_t_10.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_list_internal8(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_10); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 909, __pyx_L1_error)
+    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_list_internal8(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_10); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 968, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
 
-    /* "pgenlib.pyx":908
+    /* "pgenlib.pyx":967
  * 
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if geno_int_out.dtype == np.int8:             # <<<<<<<<<<<<<<
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":910
+  /* "pgenlib.pyx":969
  *         if geno_int_out.dtype == np.int8:
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:             # <<<<<<<<<<<<<<
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 910, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 969, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 910, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 969, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_int32); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 910, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_int32); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 969, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_5 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 910, __pyx_L1_error)
+  __pyx_t_5 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 969, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 910, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 969, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
   if (__pyx_t_9) {
 
-    /* "pgenlib.pyx":911
+    /* "pgenlib.pyx":970
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         elif geno_int_out.dtype == np.int64:
  *             self.read_list_internal64(variant_idxs, geno_int_out, allele_idx, sample_maj)
  */
     __pyx_t_11.__pyx_n = 2;
     __pyx_t_11.allele_idx = __pyx_v_allele_idx;
     __pyx_t_11.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_5 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_list_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_11); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 911, __pyx_L1_error)
+    __pyx_t_5 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_list_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_11); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 970, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
 
-    /* "pgenlib.pyx":910
+    /* "pgenlib.pyx":969
  *         if geno_int_out.dtype == np.int8:
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int32:             # <<<<<<<<<<<<<<
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":912
+  /* "pgenlib.pyx":971
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:             # <<<<<<<<<<<<<<
  *             self.read_list_internal64(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         else:
  */
-  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 912, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_geno_int_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 971, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
-  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 912, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_1, __pyx_n_s_np); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 971, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_int64); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 912, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_int64); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 971, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_1 = PyObject_RichCompare(__pyx_t_5, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 912, __pyx_L1_error)
+  __pyx_t_1 = PyObject_RichCompare(__pyx_t_5, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 971, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 912, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 971, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   if (likely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":913
+    /* "pgenlib.pyx":972
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:
  *             self.read_list_internal64(variant_idxs, geno_int_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid read_list() geno_int_out array element type (int8, int32, or int64 expected).")
  */
     __pyx_t_12.__pyx_n = 2;
     __pyx_t_12.allele_idx = __pyx_v_allele_idx;
     __pyx_t_12.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_list_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_12); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 913, __pyx_L1_error)
+    __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_list_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_geno_int_out), &__pyx_t_12); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 972, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-    /* "pgenlib.pyx":912
+    /* "pgenlib.pyx":971
  *         elif geno_int_out.dtype == np.int32:
  *             self.read_list_internal32(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         elif geno_int_out.dtype == np.int64:             # <<<<<<<<<<<<<<
  *             self.read_list_internal64(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         else:
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":915
+  /* "pgenlib.pyx":974
  *             self.read_list_internal64(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_list() geno_int_out array element type (int8, int32, or int64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__12, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 915, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__12, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 974, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 915, __pyx_L1_error)
+    __PYX_ERR(0, 974, __pyx_L1_error)
   }
   __pyx_L3:;
 
-  /* "pgenlib.pyx":916
+  /* "pgenlib.pyx":975
  *         else:
  *             raise RuntimeError("Invalid read_list() geno_int_out array element type (int8, int32, or int64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":907
+  /* "pgenlib.pyx":966
  *         return
  * 
  *     cpdef read_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray geno_int_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if geno_int_out.dtype == np.int8:
  *             self.read_list_internal8(variant_idxs, geno_int_out, allele_idx, sample_maj)
  */
 
@@ -14313,15 +14782,15 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idxs)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_geno_int_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_list", 0, 2, 4, 1); __PYX_ERR(0, 907, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_list", 0, 2, 4, 1); __PYX_ERR(0, 966, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[2] = value; kw_args--; }
         }
@@ -14329,15 +14798,15 @@
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_maj);
           if (value) { values[3] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_list") < 0)) __PYX_ERR(0, 907, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_list") < 0)) __PYX_ERR(0, 966, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
@@ -14346,34 +14815,34 @@
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_variant_idxs = ((PyArrayObject *)values[0]);
     __pyx_v_geno_int_out = ((PyArrayObject *)values[1]);
     if (values[2]) {
-      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 907, __pyx_L3_error)
+      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 966, __pyx_L3_error)
     } else {
       __pyx_v_allele_idx = ((uint32_t)1);
     }
     if (values[3]) {
-      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 907, __pyx_L3_error)
+      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 966, __pyx_L3_error)
     } else {
       __pyx_v_sample_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_list", 0, 2, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 907, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_list", 0, 2, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 966, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_list", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 907, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int_out), __pyx_ptype_5numpy_ndarray, 1, "geno_int_out", 0))) __PYX_ERR(0, 907, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 966, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int_out), __pyx_ptype_5numpy_ndarray, 1, "geno_int_out", 0))) __PYX_ERR(0, 966, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_20read_list(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idxs, __pyx_v_geno_int_out, __pyx_v_allele_idx, __pyx_v_sample_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -14394,22 +14863,22 @@
   __Pyx_RefNannySetupContext("read_list", 0);
   __pyx_pybuffer_variant_idxs.pybuffer.buf = NULL;
   __pyx_pybuffer_variant_idxs.refcount = 0;
   __pyx_pybuffernd_variant_idxs.data = NULL;
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 907, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 966, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 2;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
   __pyx_t_2.sample_maj = __pyx_v_sample_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_geno_int_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 907, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_geno_int_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 966, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -14427,37 +14896,37 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":919
+/* "pgenlib.pyx":978
  * 
  * 
  *     cpdef read_alleles_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_23read_alleles_range(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_alleles_range(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx_start, uint32_t __pyx_v_variant_idx_end, PyArrayObject *__pyx_v_allele_int32_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_range *__pyx_optional_args) {
   int __pyx_v_hap_maj = ((int)0);
   uintptr_t const *__pyx_v_subset_include_vec;
   struct plink2::PgrSampleSubsetIndexStruct __pyx_v_subset_index;
   struct plink2::PgenReaderStruct *__pyx_v_pgrp;
-  uintptr_t *__pyx_v_genovec;
+  CYTHON_UNUSED uintptr_t *__pyx_v_genovec;
   uintptr_t *__pyx_v_phasepresent;
-  uintptr_t *__pyx_v_phaseinfo;
+  CYTHON_UNUSED uintptr_t *__pyx_v_phaseinfo;
   uint32_t __pyx_v_variant_idx_ct;
   uint32_t __pyx_v_subset_size;
   int32_t *__pyx_v_main_data_ptr;
   uint32_t __pyx_v_variant_idx;
-  uint32_t __pyx_v_phasepresent_ct;
   plink2::PglErr __pyx_v_reterr;
+  uintptr_t *__pyx_v_allele_idx_offsets;
   uint32_t __pyx_v_variant_batch_ct;
   uint32_t __pyx_v_variant_batch_size;
   CYTHON_UNUSED uint32_t __pyx_v_variant_batch_sizel;
   uint32_t __pyx_v_variant_idx_offset;
   uint32_t __pyx_v_sample_ctaw2;
   uint32_t __pyx_v_sample_ctaw;
   uint32_t __pyx_v_sample_batch_ct;
@@ -14470,14 +14939,15 @@
   uintptr_t *__pyx_v_vmaj_phaseinfo_iter;
   uintptr_t *__pyx_v_smaj_geno_iter;
   uintptr_t *__pyx_v_smaj_phaseinfo_iter;
   int32_t *__pyx_v_main_data1_ptr;
   uint32_t __pyx_v_variant_batch_idx;
   uint32_t __pyx_v_sample_batch_size;
   uint32_t __pyx_v_sample_batch_idx;
+  uint32_t __pyx_v_phasepresent_ct;
   uint32_t __pyx_v_uii;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_out;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_out;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   PyObject *__pyx_t_2 = NULL;
@@ -14516,35 +14986,35 @@
   }
   __pyx_pybuffer_allele_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_out.refcount = 0;
   __pyx_pybuffernd_allele_int32_out.data = NULL;
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 919, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 978, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_range); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 919, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_range); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 978, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_23read_alleles_range)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 919, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 978, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 919, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 978, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_5 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 919, __pyx_L1_error)
+        __pyx_t_5 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 978, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_5);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_6 = __pyx_t_1; __pyx_t_7 = NULL;
         __pyx_t_8 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_6))) {
           __pyx_t_7 = PyMethod_GET_SELF(__pyx_t_6);
           if (likely(__pyx_t_7)) {
@@ -14554,35 +15024,35 @@
             __Pyx_DECREF_SET(__pyx_t_6, function);
             __pyx_t_8 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_6)) {
           PyObject *__pyx_temp[5] = {__pyx_t_7, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_allele_int32_out), __pyx_t_5};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 4+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 919, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 4+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 978, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_7); __pyx_t_7 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_6)) {
           PyObject *__pyx_temp[5] = {__pyx_t_7, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_allele_int32_out), __pyx_t_5};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 4+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 919, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 4+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 978, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_7); __pyx_t_7 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         } else
         #endif
         {
-          __pyx_t_9 = PyTuple_New(4+__pyx_t_8); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 919, __pyx_L1_error)
+          __pyx_t_9 = PyTuple_New(4+__pyx_t_8); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 978, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_9);
           if (__pyx_t_7) {
             __Pyx_GIVEREF(__pyx_t_7); PyTuple_SET_ITEM(__pyx_t_9, 0, __pyx_t_7); __pyx_t_7 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_9, 0+__pyx_t_8, __pyx_t_3);
           __Pyx_GIVEREF(__pyx_t_4);
@@ -14591,15 +15061,15 @@
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32_out));
           PyTuple_SET_ITEM(__pyx_t_9, 2+__pyx_t_8, ((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_GIVEREF(__pyx_t_5);
           PyTuple_SET_ITEM(__pyx_t_9, 3+__pyx_t_8, __pyx_t_5);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
           __pyx_t_5 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_9, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 919, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_9, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 978, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
         }
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -14614,991 +15084,1052 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":924
+  /* "pgenlib.pyx":983
  *         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
  *         #   rows, variant_idx_ct columns
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
   __pyx_t_10 = ((__pyx_v_variant_idx_end > (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_10)) {
 
-    /* "pgenlib.pyx":925
+    /* "pgenlib.pyx":984
  *         #   rows, variant_idx_ct columns
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_range_variant_idx_e, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_range_variant_idx_e, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 925, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 984, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 925, __pyx_L1_error)
+    __PYX_ERR(0, 984, __pyx_L1_error)
 
-    /* "pgenlib.pyx":924
+    /* "pgenlib.pyx":983
  *         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
  *         #   rows, variant_idx_ct columns
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
   }
 
-  /* "pgenlib.pyx":926
+  /* "pgenlib.pyx":985
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_11 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_11;
 
-  /* "pgenlib.pyx":927
+  /* "pgenlib.pyx":986
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_12 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_12;
 
-  /* "pgenlib.pyx":928
+  /* "pgenlib.pyx":987
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
  */
   __pyx_t_13 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_13;
 
-  /* "pgenlib.pyx":929
+  /* "pgenlib.pyx":988
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  */
-  __pyx_t_11 = __pyx_v_self->_genovec;
+  __pyx_t_11 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_11;
 
-  /* "pgenlib.pyx":930
+  /* "pgenlib.pyx":989
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
-  __pyx_t_11 = __pyx_v_self->_phasepresent;
+  __pyx_t_11 = __pyx_v_self->_pgv.phasepresent;
   __pyx_v_phasepresent = __pyx_t_11;
 
-  /* "pgenlib.pyx":931
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":990
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_11 = __pyx_v_self->_phaseinfo;
+  __pyx_t_11 = __pyx_v_self->_pgv.phaseinfo;
   __pyx_v_phaseinfo = __pyx_t_11;
 
-  /* "pgenlib.pyx":932
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":991
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int32_t* main_data_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":933
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":992
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int32_t* main_data_ptr
  *         cdef uint32_t variant_idx
  */
   __pyx_t_14 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_14;
 
-  /* "pgenlib.pyx":938
- *         cdef uint32_t phasepresent_ct
+  /* "pgenlib.pyx":996
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_10 = ((__pyx_v_hap_maj == 0) != 0);
   if (__pyx_t_10) {
 
-    /* "pgenlib.pyx":939
+    /* "pgenlib.pyx":997
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     __pyx_t_10 = (((__pyx_v_allele_int32_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_10)) {
 
-      /* "pgenlib.pyx":940
+      /* "pgenlib.pyx":998
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_range, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_range, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 940, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 998, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 940, __pyx_L1_error)
+      __PYX_ERR(0, 998, __pyx_L1_error)
 
-      /* "pgenlib.pyx":939
+      /* "pgenlib.pyx":997
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     }
 
-    /* "pgenlib.pyx":941
+    /* "pgenlib.pyx":999
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     __pyx_t_10 = (((__pyx_v_allele_int32_out->dimensions[1]) < (2 * __pyx_v_subset_size)) != 0);
     if (unlikely(__pyx_t_10)) {
 
-      /* "pgenlib.pyx":942
+      /* "pgenlib.pyx":1000
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
- *                 # upgrade to multiallelic version of this function later
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_range_2, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_range_2, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 942, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1000, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 942, __pyx_L1_error)
+      __PYX_ERR(0, 1000, __pyx_L1_error)
 
-      /* "pgenlib.pyx":941
+      /* "pgenlib.pyx":999
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     }
 
-    /* "pgenlib.pyx":943
+    /* "pgenlib.pyx":1001
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_14 = __pyx_v_variant_idx_end;
     __pyx_t_15 = __pyx_t_14;
     for (__pyx_t_16 = __pyx_v_variant_idx_start; __pyx_t_16 < __pyx_t_15; __pyx_t_16+=1) {
       __pyx_v_variant_idx = __pyx_t_16;
 
-      /* "pgenlib.pyx":945
+      /* "pgenlib.pyx":1002
+ *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
  */
-      __pyx_v_reterr = plink2::PgrGetP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, (&__pyx_v_phasepresent_ct));
+      __pyx_v_reterr = plink2::PgrGetMP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, (&__pyx_v_self->_pgv));
 
-      /* "pgenlib.pyx":946
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+      /* "pgenlib.pyx":1003
+ *             for variant_idx in range(variant_idx_start, variant_idx_end):
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  */
       __pyx_t_10 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_10)) {
 
-        /* "pgenlib.pyx":947
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1004
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  */
-        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_variant_idx_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_variant_idx_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_read_alleles_range_error); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_read_alleles_range_error); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 947, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1004, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         __Pyx_Raise(__pyx_t_2, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __PYX_ERR(0, 947, __pyx_L1_error)
+        __PYX_ERR(0, 1004, __pyx_L1_error)
 
-        /* "pgenlib.pyx":946
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1003
+ *             for variant_idx in range(variant_idx_start, variant_idx_end):
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  */
       }
 
-      /* "pgenlib.pyx":948
+      /* "pgenlib.pyx":1005
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))             # <<<<<<<<<<<<<<
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return
  */
       __pyx_t_17 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
       __pyx_t_18 = 0;
       __pyx_t_8 = -1;
       if (unlikely(__pyx_t_17 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_8 = 0;
       if (__pyx_t_18 < 0) {
         __pyx_t_18 += __pyx_pybuffernd_allele_int32_out.diminfo[1].shape;
         if (unlikely(__pyx_t_18 < 0)) __pyx_t_8 = 1;
       } else if (unlikely(__pyx_t_18 >= __pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_8 = 1;
       if (unlikely(__pyx_t_8 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_8);
-        __PYX_ERR(0, 948, __pyx_L1_error)
+        __PYX_ERR(0, 1005, __pyx_L1_error)
       }
       __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_17, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_18, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":949
+      /* "pgenlib.pyx":1006
  *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)             # <<<<<<<<<<<<<<
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if variant_idx_start >= variant_idx_end:
  */
-      plink2::GenoarrPhasedToAlleleCodesMinus9(__pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, __pyx_v_subset_size, __pyx_v_phasepresent_ct, NULL, __pyx_v_main_data_ptr);
+      plink2::GenoarrMPToAlleleCodesMinus9((&__pyx_v_self->_pgv), __pyx_v_subset_size, NULL, __pyx_v_main_data_ptr);
     }
 
-    /* "pgenlib.pyx":950
+    /* "pgenlib.pyx":1007
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":938
- *         cdef uint32_t phasepresent_ct
+    /* "pgenlib.pyx":996
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":951
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+  /* "pgenlib.pyx":1008
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  */
   __pyx_t_10 = ((__pyx_v_variant_idx_start >= __pyx_v_variant_idx_end) != 0);
   if (unlikely(__pyx_t_10)) {
 
-    /* "pgenlib.pyx":952
+    /* "pgenlib.pyx":1009
  *             return
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")             # <<<<<<<<<<<<<<
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  */
-    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_range_variant_idx_s, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_range_variant_idx_s, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u__10); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u__10); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 952, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1009, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 952, __pyx_L1_error)
+    __PYX_ERR(0, 1009, __pyx_L1_error)
 
-    /* "pgenlib.pyx":951
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+    /* "pgenlib.pyx":1008
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return
  *         if variant_idx_start >= variant_idx_end:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  */
   }
 
-  /* "pgenlib.pyx":953
+  /* "pgenlib.pyx":1010
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if allele_int32_out.shape[0] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_10 = (((__pyx_v_allele_int32_out->dimensions[0]) < (2 * __pyx_v_subset_size)) != 0);
   if (unlikely(__pyx_t_10)) {
 
-    /* "pgenlib.pyx":954
+    /* "pgenlib.pyx":1011
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")             # <<<<<<<<<<<<<<
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_2 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_ran, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_ran, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_and_row_count_should_be_twice_t); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_and_row_count_should_be_twice_t); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 954, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1011, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 954, __pyx_L1_error)
+    __PYX_ERR(0, 1011, __pyx_L1_error)
 
-    /* "pgenlib.pyx":953
+    /* "pgenlib.pyx":1010
  *         if variant_idx_start >= variant_idx_end:
  *             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
  *         if allele_int32_out.shape[0] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":955
+  /* "pgenlib.pyx":1012
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
- *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
  */
   __pyx_t_10 = (((__pyx_v_allele_int32_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_10)) {
 
-    /* "pgenlib.pyx":956
+    /* "pgenlib.pyx":1013
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
- *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
- *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *         if allele_idx_offsets != NULL:
  */
-    __pyx_t_2 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_ran_2, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_ran_2, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 956, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1013, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 956, __pyx_L1_error)
+    __PYX_ERR(0, 1013, __pyx_L1_error)
 
-    /* "pgenlib.pyx":955
+    /* "pgenlib.pyx":1012
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
- *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
  */
   }
 
-  /* "pgenlib.pyx":957
+  /* "pgenlib.pyx":1014
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets             # <<<<<<<<<<<<<<
+ *         if allele_idx_offsets != NULL:
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+ */
+  __pyx_t_11 = (__pyx_v_self->_info_ptr[0]).allele_idx_offsets;
+  __pyx_v_allele_idx_offsets = __pyx_t_11;
+
+  /* "pgenlib.pyx":1015
+ *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *         if allele_idx_offsets != NULL:             # <<<<<<<<<<<<<<
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
+ */
+  __pyx_t_10 = ((__pyx_v_allele_idx_offsets != NULL) != 0);
+  if (__pyx_t_10) {
+
+    /* "pgenlib.pyx":1016
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *         if allele_idx_offsets != NULL:
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
+ *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
+ */
+    __pyx_t_10 = ((((__pyx_v_allele_idx_offsets[__pyx_v_variant_idx_end]) - (__pyx_v_allele_idx_offsets[__pyx_v_variant_idx_start])) != (((__pyx_v_variant_idx_end - __pyx_v_variant_idx_start) * plink2::k1LU) * 2)) != 0);
+    if (unlikely(__pyx_t_10)) {
+
+      /* "pgenlib.pyx":1017
+ *         if allele_idx_offsets != NULL:
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")             # <<<<<<<<<<<<<<
+ *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
+ *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
+ */
+      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__13, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1017, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_2);
+      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+      __PYX_ERR(0, 1017, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1016
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *         if allele_idx_offsets != NULL:
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
+ *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
+ */
+    }
+
+    /* "pgenlib.pyx":1015
+ *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *         if allele_idx_offsets != NULL:             # <<<<<<<<<<<<<<
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
+ */
+  }
+
+  /* "pgenlib.pyx":1018
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":958
- *             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
+  /* "pgenlib.pyx":1019
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":959
+  /* "pgenlib.pyx":1020
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_sizel = plink2::DivUp(__pyx_v_variant_batch_size, plink2::kBitsPerWord);
 
-  /* "pgenlib.pyx":960
+  /* "pgenlib.pyx":1021
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *         cdef uint32_t variant_idx_offset = variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  */
   __pyx_v_variant_idx_offset = __pyx_v_variant_idx_start;
 
-  /* "pgenlib.pyx":961
+  /* "pgenlib.pyx":1022
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":962
+  /* "pgenlib.pyx":1023
  *         cdef uint32_t variant_idx_offset = variant_idx_start
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWord));
 
-  /* "pgenlib.pyx":963
+  /* "pgenlib.pyx":1024
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":964
+  /* "pgenlib.pyx":1025
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  */
   __pyx_t_19 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_19;
 
-  /* "pgenlib.pyx":965
+  /* "pgenlib.pyx":1026
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_11 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_11;
 
-  /* "pgenlib.pyx":966
+  /* "pgenlib.pyx":1027
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf
  */
   __pyx_t_11 = __pyx_v_self->_multivar_vmaj_phaseinfo_buf;
   __pyx_v_multivar_vmaj_phaseinfo_buf = __pyx_t_11;
 
-  /* "pgenlib.pyx":967
+  /* "pgenlib.pyx":1028
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf
  *         cdef uintptr_t* vmaj_geno_iter
  */
   __pyx_t_11 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_11;
 
-  /* "pgenlib.pyx":968
+  /* "pgenlib.pyx":1029
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* vmaj_geno_iter
  *         cdef uintptr_t* vmaj_phaseinfo_iter
  */
   __pyx_t_11 = __pyx_v_self->_multivar_smaj_phaseinfo_batch_buf;
   __pyx_v_multivar_smaj_phaseinfo_batch_buf = __pyx_t_11;
 
-  /* "pgenlib.pyx":978
- *         cdef uint32_t sample_batch_idx
+  /* "pgenlib.pyx":1040
+ *         cdef uint32_t phasepresent_ct
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_14 = __pyx_v_variant_batch_ct;
   __pyx_t_15 = __pyx_t_14;
   for (__pyx_t_16 = 0; __pyx_t_16 < __pyx_t_15; __pyx_t_16+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_16;
 
-    /* "pgenlib.pyx":979
+    /* "pgenlib.pyx":1041
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  */
     __pyx_t_10 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_10) {
 
-      /* "pgenlib.pyx":980
+      /* "pgenlib.pyx":1042
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  */
       __pyx_t_20 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 980, __pyx_L1_error)
+        __PYX_ERR(0, 1042, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_20, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":981
+      /* "pgenlib.pyx":1043
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)             # <<<<<<<<<<<<<<
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  */
       __pyx_v_variant_batch_sizel = plink2::DivUp(__pyx_v_variant_batch_size, plink2::kBitsPerWord);
 
-      /* "pgenlib.pyx":979
+      /* "pgenlib.pyx":1041
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  */
     }
 
-    /* "pgenlib.pyx":982
+    /* "pgenlib.pyx":1044
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *             vmaj_geno_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for uii in range(variant_batch_size):
  */
     __pyx_v_vmaj_geno_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":983
+    /* "pgenlib.pyx":1045
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf             # <<<<<<<<<<<<<<
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  */
     __pyx_v_vmaj_phaseinfo_iter = __pyx_v_multivar_vmaj_phaseinfo_buf;
 
-    /* "pgenlib.pyx":984
+    /* "pgenlib.pyx":1046
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for uii in range(variant_batch_size):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_21 = __pyx_v_variant_batch_size;
     __pyx_t_22 = __pyx_t_21;
     for (__pyx_t_23 = 0; __pyx_t_23 < __pyx_t_22; __pyx_t_23+=1) {
       __pyx_v_uii = __pyx_t_23;
 
-      /* "pgenlib.pyx":985
+      /* "pgenlib.pyx":1047
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGetP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, (__pyx_v_uii + __pyx_v_variant_idx_offset), __pyx_v_pgrp, __pyx_v_vmaj_geno_iter, __pyx_v_phasepresent, __pyx_v_vmaj_phaseinfo_iter, (&__pyx_v_phasepresent_ct));
 
-      /* "pgenlib.pyx":986
+      /* "pgenlib.pyx":1048
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))
  *                 if phasepresent_ct == 0:
  */
       __pyx_t_10 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_10)) {
 
-        /* "pgenlib.pyx":987
+        /* "pgenlib.pyx":1049
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 if phasepresent_ct == 0:
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
  */
-        __pyx_t_2 = __Pyx_PyInt_From_uint32_t((__pyx_v_uii + __pyx_v_variant_idx_offset)); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyInt_From_uint32_t((__pyx_v_uii + __pyx_v_variant_idx_offset)); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_variant_idx_2, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_variant_idx_2, __pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_read_alleles_range_error); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_kp_u_read_alleles_range_error); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 987, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1049, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
         __Pyx_Raise(__pyx_t_1, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __PYX_ERR(0, 987, __pyx_L1_error)
+        __PYX_ERR(0, 1049, __pyx_L1_error)
 
-        /* "pgenlib.pyx":986
+        /* "pgenlib.pyx":1048
  *             for uii in range(variant_batch_size):
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, uii + variant_idx_offset, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))
  *                 if phasepresent_ct == 0:
  */
       }
 
-      /* "pgenlib.pyx":988
+      /* "pgenlib.pyx":1050
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))
  *                 if phasepresent_ct == 0:             # <<<<<<<<<<<<<<
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
  *                 # else:
  */
       __pyx_t_10 = ((__pyx_v_phasepresent_ct == 0) != 0);
       if (__pyx_t_10) {
 
-        /* "pgenlib.pyx":989
+        /* "pgenlib.pyx":1051
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))
  *                 if phasepresent_ct == 0:
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)             # <<<<<<<<<<<<<<
  *                 # else:
  *                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
  */
         plink2::ZeroWArr(__pyx_v_sample_ctaw, __pyx_v_vmaj_phaseinfo_iter);
 
-        /* "pgenlib.pyx":988
+        /* "pgenlib.pyx":1050
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("variant_idx " + str(uii + variant_idx_offset) + " read_alleles_range() error " + str(reterr))
  *                 if phasepresent_ct == 0:             # <<<<<<<<<<<<<<
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
  *                 # else:
  */
       }
 
-      /* "pgenlib.pyx":992
+      /* "pgenlib.pyx":1054
  *                 # else:
  *                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
  *                 vmaj_geno_iter = &(vmaj_geno_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])
  *             sample_batch_size = kPglNypTransposeBatch
  */
       __pyx_v_vmaj_geno_iter = (&(__pyx_v_vmaj_geno_iter[__pyx_v_sample_ctaw2]));
 
-      /* "pgenlib.pyx":993
+      /* "pgenlib.pyx":1055
  *                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
  *                 vmaj_geno_iter = &(vmaj_geno_iter[sample_ctaw2])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_phaseinfo_iter = (&(__pyx_v_vmaj_phaseinfo_iter[__pyx_v_sample_ctaw]));
     }
 
-    /* "pgenlib.pyx":994
+    /* "pgenlib.pyx":1056
  *                 vmaj_geno_iter = &(vmaj_geno_iter[sample_ctaw2])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":995
+    /* "pgenlib.pyx":1057
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_geno_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_vmaj_geno_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":996
+    /* "pgenlib.pyx":1058
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_phaseinfo_iter = __pyx_v_multivar_vmaj_phaseinfo_buf;
 
-    /* "pgenlib.pyx":997
+    /* "pgenlib.pyx":1059
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_21 = __pyx_v_sample_batch_ct;
     __pyx_t_22 = __pyx_t_21;
     for (__pyx_t_23 = 0; __pyx_t_23 < __pyx_t_22; __pyx_t_23+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_23;
 
-      /* "pgenlib.pyx":998
+      /* "pgenlib.pyx":1060
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_10 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_10) {
 
-        /* "pgenlib.pyx":999
+        /* "pgenlib.pyx":1061
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf
  */
         __pyx_t_20 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 999, __pyx_L1_error)
+          __PYX_ERR(0, 1061, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_20, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":998
+        /* "pgenlib.pyx":1060
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":1000
+      /* "pgenlib.pyx":1062
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf
  *                 TransposeNypblock(vmaj_geno_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_geno_iter, transpose_batch_buf)
  */
       __pyx_v_smaj_geno_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":1001
+      /* "pgenlib.pyx":1063
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_geno_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_geno_iter, transpose_batch_buf)
  *                 # todo: skip bitblock transpose when all phasepresent_ct values
  */
       __pyx_v_smaj_phaseinfo_iter = __pyx_v_multivar_smaj_phaseinfo_batch_buf;
 
-      /* "pgenlib.pyx":1002
+      /* "pgenlib.pyx":1064
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf
  *                 TransposeNypblock(vmaj_geno_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_geno_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 # todo: skip bitblock transpose when all phasepresent_ct values
  *                 #       are zero, etc.
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_geno_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_geno_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":1005
+      /* "pgenlib.pyx":1067
  *                 # todo: skip bitblock transpose when all phasepresent_ct values
  *                 #       are zero, etc.
  *                 TransposeBitblock(vmaj_phaseinfo_iter, sample_ctaw, <uint32_t>(kPglNypTransposeWords // 2), variant_batch_size, sample_batch_size, smaj_phaseinfo_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     # bugfix (11 Mar 2023): first index was incorrect
  */
       plink2::TransposeBitblock(__pyx_v_vmaj_phaseinfo_iter, __pyx_v_sample_ctaw, ((uint32_t)__Pyx_div_long(plink2::kPglNypTransposeWords, 2)), __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_phaseinfo_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":1006
+      /* "pgenlib.pyx":1068
  *                 #       are zero, etc.
  *                 TransposeBitblock(vmaj_phaseinfo_iter, sample_ctaw, <uint32_t>(kPglNypTransposeWords // 2), variant_batch_size, sample_batch_size, smaj_phaseinfo_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     # bugfix (11 Mar 2023): first index was incorrect
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  */
       __pyx_t_24 = __pyx_v_sample_batch_size;
       __pyx_t_25 = __pyx_t_24;
       for (__pyx_t_26 = 0; __pyx_t_26 < __pyx_t_25; __pyx_t_26+=1) {
         __pyx_v_uii = __pyx_t_26;
 
-        /* "pgenlib.pyx":1008
+        /* "pgenlib.pyx":1070
  *                 for uii in range(sample_batch_size):
  *                     # bugfix (11 Mar 2023): first index was incorrect
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))             # <<<<<<<<<<<<<<
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  */
         __pyx_t_18 = (2 * (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch)));
@@ -15607,19 +16138,19 @@
         if (__pyx_t_18 < 0) {
           __pyx_t_18 += __pyx_pybuffernd_allele_int32_out.diminfo[0].shape;
           if (unlikely(__pyx_t_18 < 0)) __pyx_t_8 = 0;
         } else if (unlikely(__pyx_t_18 >= __pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_8 = 0;
         if (unlikely(__pyx_t_17 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_8 = 1;
         if (unlikely(__pyx_t_8 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_8);
-          __PYX_ERR(0, 1008, __pyx_L1_error)
+          __PYX_ERR(0, 1070, __pyx_L1_error)
         }
         __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_18, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_17, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":1009
+        /* "pgenlib.pyx":1071
  *                     # bugfix (11 Mar 2023): first index was incorrect
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))             # <<<<<<<<<<<<<<
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  */
         __pyx_t_18 = ((2 * (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch))) + 1);
@@ -15628,87 +16159,87 @@
         if (__pyx_t_18 < 0) {
           __pyx_t_18 += __pyx_pybuffernd_allele_int32_out.diminfo[0].shape;
           if (unlikely(__pyx_t_18 < 0)) __pyx_t_8 = 0;
         } else if (unlikely(__pyx_t_18 >= __pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_8 = 0;
         if (unlikely(__pyx_t_17 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_8 = 1;
         if (unlikely(__pyx_t_8 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_8);
-          __PYX_ERR(0, 1009, __pyx_L1_error)
+          __PYX_ERR(0, 1071, __pyx_L1_error)
         }
         __pyx_v_main_data1_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_18, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_17, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":1010
+        /* "pgenlib.pyx":1072
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  */
         plink2::GenoarrPhasedToHapCodes(__pyx_v_smaj_geno_iter, __pyx_v_smaj_phaseinfo_iter, __pyx_v_variant_batch_size, __pyx_v_main_data_ptr, __pyx_v_main_data1_ptr);
 
-        /* "pgenlib.pyx":1011
+        /* "pgenlib.pyx":1073
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  */
         __pyx_v_smaj_geno_iter = (&(__pyx_v_smaj_geno_iter[plink2::kPglNypTransposeWords]));
 
-        /* "pgenlib.pyx":1012
+        /* "pgenlib.pyx":1074
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])             # <<<<<<<<<<<<<<
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  */
         __pyx_v_smaj_phaseinfo_iter = (&(__pyx_v_smaj_phaseinfo_iter[__Pyx_div_long(plink2::kPglNypTransposeWords, 2)]));
       }
 
-      /* "pgenlib.pyx":1013
+      /* "pgenlib.pyx":1075
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *             variant_idx_offset += kPglNypTransposeBatch
  */
       __pyx_v_vmaj_geno_iter = (&(__pyx_v_vmaj_geno_iter[plink2::kPglNypTransposeWords]));
 
-      /* "pgenlib.pyx":1014
+      /* "pgenlib.pyx":1076
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])             # <<<<<<<<<<<<<<
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return
  */
       __pyx_v_vmaj_phaseinfo_iter = (&(__pyx_v_vmaj_phaseinfo_iter[__Pyx_div_long(plink2::kPglNypTransposeWords, 2)]));
     }
 
-    /* "pgenlib.pyx":1015
+    /* "pgenlib.pyx":1077
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *             variant_idx_offset += kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_variant_idx_offset = (__pyx_v_variant_idx_offset + plink2::kPglNypTransposeBatch);
   }
 
-  /* "pgenlib.pyx":1016
+  /* "pgenlib.pyx":1078
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *             variant_idx_offset += kPglNypTransposeBatch
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":919
+  /* "pgenlib.pyx":978
  * 
  * 
  *     cpdef read_alleles_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
@@ -15775,61 +16306,61 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_start)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_end)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_range", 0, 3, 4, 1); __PYX_ERR(0, 919, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_range", 0, 3, 4, 1); __PYX_ERR(0, 978, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_range", 0, 3, 4, 2); __PYX_ERR(0, 919, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_range", 0, 3, 4, 2); __PYX_ERR(0, 978, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_hap_maj);
           if (value) { values[3] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_range") < 0)) __PYX_ERR(0, 919, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_range") < 0)) __PYX_ERR(0, 978, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 919, __pyx_L3_error)
-    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 919, __pyx_L3_error)
+    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 978, __pyx_L3_error)
+    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 978, __pyx_L3_error)
     __pyx_v_allele_int32_out = ((PyArrayObject *)values[2]);
     if (values[3]) {
-      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 919, __pyx_L3_error)
+      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 978, __pyx_L3_error)
     } else {
       __pyx_v_hap_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_alleles_range", 0, 3, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 919, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_alleles_range", 0, 3, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 978, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_alleles_range", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 919, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 978, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_22read_alleles_range(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_allele_int32_out, __pyx_v_hap_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -15850,21 +16381,21 @@
   __Pyx_RefNannySetupContext("read_alleles_range", 0);
   __pyx_pybuffer_allele_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_out.refcount = 0;
   __pyx_pybuffernd_allele_int32_out.data = NULL;
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 919, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 978, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.hap_maj = __pyx_v_hap_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_allele_int32_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 919, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_allele_int32_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 978, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -15882,58 +16413,59 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1019
+/* "pgenlib.pyx":1081
  * 
  * 
  *     cpdef read_alleles_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_25read_alleles_list(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_alleles_list(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyArrayObject *__pyx_v_variant_idxs, PyArrayObject *__pyx_v_allele_int32_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_list *__pyx_optional_args) {
   int __pyx_v_hap_maj = ((int)0);
   uint32_t __pyx_v_raw_variant_ct;
   uintptr_t const *__pyx_v_subset_include_vec;
   struct plink2::PgrSampleSubsetIndexStruct __pyx_v_subset_index;
   struct plink2::PgenReaderStruct *__pyx_v_pgrp;
-  uintptr_t *__pyx_v_genovec;
+  CYTHON_UNUSED uintptr_t *__pyx_v_genovec;
   uintptr_t *__pyx_v_phasepresent;
-  uintptr_t *__pyx_v_phaseinfo;
+  CYTHON_UNUSED uintptr_t *__pyx_v_phaseinfo;
   uint32_t __pyx_v_variant_idx_ct;
   uint32_t __pyx_v_subset_size;
   int32_t *__pyx_v_main_data_ptr;
   uint32_t __pyx_v_variant_list_idx;
   uint32_t __pyx_v_variant_idx;
-  uint32_t __pyx_v_phasepresent_ct;
   plink2::PglErr __pyx_v_reterr;
   uint32_t __pyx_v_variant_batch_ct;
   uint32_t __pyx_v_variant_batch_size;
   CYTHON_UNUSED uint32_t __pyx_v_variant_batch_sizel;
   uint32_t __pyx_v_sample_ctaw2;
   uint32_t __pyx_v_sample_ctaw;
   uint32_t __pyx_v_sample_batch_ct;
   plink2::VecW *__pyx_v_transpose_batch_buf;
   uintptr_t *__pyx_v_multivar_vmaj_geno_buf;
   uintptr_t *__pyx_v_multivar_vmaj_phaseinfo_buf;
   uintptr_t *__pyx_v_multivar_smaj_geno_batch_buf;
   uintptr_t *__pyx_v_multivar_smaj_phaseinfo_batch_buf;
+  uintptr_t *__pyx_v_allele_idx_offsets;
   uintptr_t *__pyx_v_vmaj_geno_iter;
   uintptr_t *__pyx_v_vmaj_phaseinfo_iter;
   uintptr_t *__pyx_v_smaj_geno_iter;
   uintptr_t *__pyx_v_smaj_phaseinfo_iter;
   int32_t *__pyx_v_main_data1_ptr;
   uint32_t __pyx_v_variant_batch_idx;
   uint32_t __pyx_v_sample_batch_size;
   uint32_t __pyx_v_sample_batch_idx;
+  uint32_t __pyx_v_phasepresent_ct;
   uint32_t __pyx_v_uii;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_out;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_out;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_variant_idxs;
   __Pyx_Buffer __pyx_pybuffer_variant_idxs;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
@@ -15976,36 +16508,36 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_allele_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_out.refcount = 0;
   __pyx_pybuffernd_allele_int32_out.data = NULL;
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1019, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1081, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1019, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1081, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_list); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1019, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_list); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1081, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_25read_alleles_list)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1019, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1081, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -16015,45 +16547,45 @@
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[4] = {__pyx_t_5, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_allele_int32_out), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1019, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1081, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[4] = {__pyx_t_5, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_allele_int32_out), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1019, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1081, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1019, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1081, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_variant_idxs));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_variant_idxs));
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, ((PyObject *)__pyx_v_variant_idxs));
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32_out));
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, ((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 2+__pyx_t_6, __pyx_t_3);
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1019, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1081, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -16068,1000 +16600,1061 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1024
+  /* "pgenlib.pyx":1086
  *         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
  *         #   rows, variant_idx_ct columns
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_8 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_8;
 
-  /* "pgenlib.pyx":1025
+  /* "pgenlib.pyx":1087
  *         #   rows, variant_idx_ct columns
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_9 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_9;
 
-  /* "pgenlib.pyx":1026
+  /* "pgenlib.pyx":1088
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_10 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_10;
 
-  /* "pgenlib.pyx":1027
+  /* "pgenlib.pyx":1089
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
  */
   __pyx_t_11 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_11;
 
-  /* "pgenlib.pyx":1028
+  /* "pgenlib.pyx":1090
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  */
-  __pyx_t_9 = __pyx_v_self->_genovec;
+  __pyx_t_9 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_9;
 
-  /* "pgenlib.pyx":1029
+  /* "pgenlib.pyx":1091
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
-  __pyx_t_9 = __pyx_v_self->_phasepresent;
+  __pyx_t_9 = __pyx_v_self->_pgv.phasepresent;
   __pyx_v_phasepresent = __pyx_t_9;
 
-  /* "pgenlib.pyx":1030
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1092
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_9 = __pyx_v_self->_phaseinfo;
+  __pyx_t_9 = __pyx_v_self->_pgv.phaseinfo;
   __pyx_v_phaseinfo = __pyx_t_9;
 
-  /* "pgenlib.pyx":1031
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":1093
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int32_t* main_data_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":1032
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":1094
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int32_t* main_data_ptr
  *         cdef uint32_t variant_list_idx
  */
   __pyx_t_8 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_8;
 
-  /* "pgenlib.pyx":1038
- *         cdef uint32_t phasepresent_ct
+  /* "pgenlib.pyx":1099
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_12 = ((__pyx_v_hap_maj == 0) != 0);
   if (__pyx_t_12) {
 
-    /* "pgenlib.pyx":1039
+    /* "pgenlib.pyx":1100
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     __pyx_t_12 = (((__pyx_v_allele_int32_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1040
+      /* "pgenlib.pyx":1101
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_list, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_list, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1040, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1101, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1040, __pyx_L1_error)
+      __PYX_ERR(0, 1101, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1039
+      /* "pgenlib.pyx":1100
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1041
+    /* "pgenlib.pyx":1102
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_12 = (((__pyx_v_allele_int32_out->dimensions[1]) < (2 * __pyx_v_subset_size)) != 0);
     if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1042
+      /* "pgenlib.pyx":1103
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_list_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_list_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1042, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1103, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1042, __pyx_L1_error)
+      __PYX_ERR(0, 1103, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1041
+      /* "pgenlib.pyx":1102
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":1043
+    /* "pgenlib.pyx":1104
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_8 = __pyx_v_variant_idx_ct;
     __pyx_t_13 = __pyx_t_8;
     for (__pyx_t_14 = 0; __pyx_t_14 < __pyx_t_13; __pyx_t_14+=1) {
       __pyx_v_variant_list_idx = __pyx_t_14;
 
-      /* "pgenlib.pyx":1044
+      /* "pgenlib.pyx":1105
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_15 = __pyx_v_variant_list_idx;
       __pyx_t_6 = -1;
       if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_6 = 0;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1044, __pyx_L1_error)
+        __PYX_ERR(0, 1105, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_15, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":1045
+      /* "pgenlib.pyx":1106
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 # upgrade to multiallelic version of this function later
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  */
       __pyx_t_12 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_12)) {
 
-        /* "pgenlib.pyx":1046
+        /* "pgenlib.pyx":1107
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_variant_index, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_variant_index, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1046, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1107, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __Pyx_Raise(__pyx_t_1, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __PYX_ERR(0, 1046, __pyx_L1_error)
+        __PYX_ERR(0, 1107, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1045
+        /* "pgenlib.pyx":1106
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 # upgrade to multiallelic version of this function later
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  */
       }
 
-      /* "pgenlib.pyx":1048
+      /* "pgenlib.pyx":1108
+ *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  */
-      __pyx_v_reterr = plink2::PgrGetP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, (&__pyx_v_phasepresent_ct));
+      __pyx_v_reterr = plink2::PgrGetMP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, (&__pyx_v_self->_pgv));
 
-      /* "pgenlib.pyx":1049
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+      /* "pgenlib.pyx":1109
+ *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  */
       __pyx_t_12 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_12)) {
 
-        /* "pgenlib.pyx":1050
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1110
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  */
-        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1050, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1110, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1050, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1110, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_error, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1050, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_error, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1110, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1050, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1110, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         __Pyx_Raise(__pyx_t_4, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __PYX_ERR(0, 1050, __pyx_L1_error)
+        __PYX_ERR(0, 1110, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1049
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1109
+ *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  */
       }
 
-      /* "pgenlib.pyx":1051
+      /* "pgenlib.pyx":1111
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))             # <<<<<<<<<<<<<<
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return
  */
       __pyx_t_15 = __pyx_v_variant_list_idx;
       __pyx_t_16 = 0;
       __pyx_t_6 = -1;
       if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_6 = 0;
       if (__pyx_t_16 < 0) {
         __pyx_t_16 += __pyx_pybuffernd_allele_int32_out.diminfo[1].shape;
         if (unlikely(__pyx_t_16 < 0)) __pyx_t_6 = 1;
       } else if (unlikely(__pyx_t_16 >= __pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_6 = 1;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1051, __pyx_L1_error)
+        __PYX_ERR(0, 1111, __pyx_L1_error)
       }
       __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_15, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_16, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1052
+      /* "pgenlib.pyx":1112
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)             # <<<<<<<<<<<<<<
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  */
-      plink2::GenoarrPhasedToAlleleCodesMinus9(__pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, __pyx_v_subset_size, __pyx_v_phasepresent_ct, NULL, __pyx_v_main_data_ptr);
+      plink2::GenoarrMPToAlleleCodesMinus9((&__pyx_v_self->_pgv), __pyx_v_subset_size, NULL, __pyx_v_main_data_ptr);
     }
 
-    /* "pgenlib.pyx":1053
+    /* "pgenlib.pyx":1113
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1038
- *         cdef uint32_t phasepresent_ct
+    /* "pgenlib.pyx":1099
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":1054
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+  /* "pgenlib.pyx":1114
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return
  *         if allele_int32_out.shape[0] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  */
   __pyx_t_12 = (((__pyx_v_allele_int32_out->dimensions[0]) < (2 * __pyx_v_subset_size)) != 0);
   if (unlikely(__pyx_t_12)) {
 
-    /* "pgenlib.pyx":1055
+    /* "pgenlib.pyx":1115
  *             return
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")             # <<<<<<<<<<<<<<
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
-    __pyx_t_4 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_lis, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_lis, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_and_row_count_should_be_twice_t); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_and_row_count_should_be_twice_t); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1055, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1115, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_Raise(__pyx_t_4, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __PYX_ERR(0, 1055, __pyx_L1_error)
+    __PYX_ERR(0, 1115, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1054
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+    /* "pgenlib.pyx":1114
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
  *             return
  *         if allele_int32_out.shape[0] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  */
   }
 
-  /* "pgenlib.pyx":1056
+  /* "pgenlib.pyx":1116
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   __pyx_t_12 = (((__pyx_v_allele_int32_out->dimensions[1]) < __pyx_v_variant_idx_ct) != 0);
   if (unlikely(__pyx_t_12)) {
 
-    /* "pgenlib.pyx":1057
+    /* "pgenlib.pyx":1117
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  */
-    __pyx_t_4 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_lis_2, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Haplotype_major_read_alleles_lis_2, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u__9); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u__9); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1057, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1117, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_4, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    __PYX_ERR(0, 1057, __pyx_L1_error)
+    __PYX_ERR(0, 1117, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1056
+    /* "pgenlib.pyx":1116
  *         if allele_int32_out.shape[0] < 2 * subset_size:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
  *         if allele_int32_out.shape[1] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  */
   }
 
-  /* "pgenlib.pyx":1058
+  /* "pgenlib.pyx":1118
  *         if allele_int32_out.shape[1] < variant_idx_ct:
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  */
   __pyx_v_variant_batch_ct = plink2::DivUp(__pyx_v_variant_idx_ct, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":1059
+  /* "pgenlib.pyx":1119
  *             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  */
   __pyx_v_variant_batch_size = plink2::kPglNypTransposeBatch;
 
-  /* "pgenlib.pyx":1060
+  /* "pgenlib.pyx":1120
  *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  */
   __pyx_v_variant_batch_sizel = plink2::DivUp(__pyx_v_variant_batch_size, plink2::kBitsPerWord);
 
-  /* "pgenlib.pyx":1061
+  /* "pgenlib.pyx":1121
  *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  */
   __pyx_v_sample_ctaw2 = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWordD2));
 
-  /* "pgenlib.pyx":1062
+  /* "pgenlib.pyx":1122
  *         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)             # <<<<<<<<<<<<<<
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  */
   __pyx_v_sample_ctaw = (plink2::kWordsPerVec * plink2::DivUp(__pyx_v_subset_size, plink2::kBitsPerWord));
 
-  /* "pgenlib.pyx":1063
+  /* "pgenlib.pyx":1123
  *         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  */
   __pyx_v_sample_batch_ct = plink2::DivUp(__pyx_v_subset_size, plink2::kPglNypTransposeBatch);
 
-  /* "pgenlib.pyx":1064
+  /* "pgenlib.pyx":1124
  *         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  */
   __pyx_t_17 = __pyx_v_self->_transpose_batch_buf;
   __pyx_v_transpose_batch_buf = __pyx_t_17;
 
-  /* "pgenlib.pyx":1065
+  /* "pgenlib.pyx":1125
  *         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  */
   __pyx_t_9 = __pyx_v_self->_multivar_vmaj_geno_buf;
   __pyx_v_multivar_vmaj_geno_buf = __pyx_t_9;
 
-  /* "pgenlib.pyx":1066
+  /* "pgenlib.pyx":1126
  *         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf
  */
   __pyx_t_9 = __pyx_v_self->_multivar_vmaj_phaseinfo_buf;
   __pyx_v_multivar_vmaj_phaseinfo_buf = __pyx_t_9;
 
-  /* "pgenlib.pyx":1067
+  /* "pgenlib.pyx":1127
  *         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf
- *         cdef uintptr_t* vmaj_geno_iter
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
  */
   __pyx_t_9 = __pyx_v_self->_multivar_smaj_geno_batch_buf;
   __pyx_v_multivar_smaj_geno_batch_buf = __pyx_t_9;
 
-  /* "pgenlib.pyx":1068
+  /* "pgenlib.pyx":1128
  *         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
  *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
  *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
  *         cdef uintptr_t* vmaj_geno_iter
- *         cdef uintptr_t* vmaj_phaseinfo_iter
  */
   __pyx_t_9 = __pyx_v_self->_multivar_smaj_phaseinfo_batch_buf;
   __pyx_v_multivar_smaj_phaseinfo_batch_buf = __pyx_t_9;
 
-  /* "pgenlib.pyx":1078
- *         cdef uint32_t sample_batch_idx
+  /* "pgenlib.pyx":1129
+ *         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
+ *         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf
+ *         cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* vmaj_geno_iter
+ *         cdef uintptr_t* vmaj_phaseinfo_iter
+ */
+  __pyx_t_9 = (__pyx_v_self->_info_ptr[0]).allele_idx_offsets;
+  __pyx_v_allele_idx_offsets = __pyx_t_9;
+
+  /* "pgenlib.pyx":1140
+ *         cdef uint32_t phasepresent_ct
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):             # <<<<<<<<<<<<<<
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  */
   __pyx_t_8 = __pyx_v_variant_batch_ct;
   __pyx_t_13 = __pyx_t_8;
   for (__pyx_t_14 = 0; __pyx_t_14 < __pyx_t_13; __pyx_t_14+=1) {
     __pyx_v_variant_batch_idx = __pyx_t_14;
 
-    /* "pgenlib.pyx":1079
+    /* "pgenlib.pyx":1141
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  */
     __pyx_t_12 = ((__pyx_v_variant_batch_idx == (__pyx_v_variant_batch_ct - 1)) != 0);
     if (__pyx_t_12) {
 
-      /* "pgenlib.pyx":1080
+      /* "pgenlib.pyx":1142
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  */
       __pyx_t_18 = (__pyx_v_variant_idx_ct - 1);
       if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
         PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-        __PYX_ERR(0, 1080, __pyx_L1_error)
+        __PYX_ERR(0, 1142, __pyx_L1_error)
       }
       __pyx_v_variant_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_18, plink2::kPglNypTransposeBatch)));
 
-      /* "pgenlib.pyx":1081
+      /* "pgenlib.pyx":1143
  *             if variant_batch_idx == (variant_batch_ct - 1):
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)             # <<<<<<<<<<<<<<
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  */
       __pyx_v_variant_batch_sizel = plink2::DivUp(__pyx_v_variant_batch_size, plink2::kBitsPerWord);
 
-      /* "pgenlib.pyx":1079
+      /* "pgenlib.pyx":1141
  *         cdef uint32_t uii
  *         for variant_batch_idx in range(variant_batch_ct):
  *             if variant_batch_idx == (variant_batch_ct - 1):             # <<<<<<<<<<<<<<
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  */
     }
 
-    /* "pgenlib.pyx":1082
+    /* "pgenlib.pyx":1144
  *                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *             vmaj_geno_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):
  */
     __pyx_v_vmaj_geno_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":1083
+    /* "pgenlib.pyx":1145
  *                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
     __pyx_v_vmaj_phaseinfo_iter = __pyx_v_multivar_vmaj_phaseinfo_buf;
 
-    /* "pgenlib.pyx":1084
+    /* "pgenlib.pyx":1146
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_19 = ((__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch) + __pyx_v_variant_batch_size);
     __pyx_t_20 = __pyx_t_19;
     for (__pyx_t_21 = (__pyx_v_variant_batch_idx * plink2::kPglNypTransposeBatch); __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
       __pyx_v_variant_list_idx = __pyx_t_21;
 
-      /* "pgenlib.pyx":1085
+      /* "pgenlib.pyx":1147
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_15 = __pyx_v_variant_list_idx;
       __pyx_t_6 = -1;
       if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_6 = 0;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1085, __pyx_L1_error)
+        __PYX_ERR(0, 1147, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_15, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":1086
+      /* "pgenlib.pyx":1148
  *             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
+ *                 if allele_idx_offsets != NULL:
  */
       __pyx_t_12 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_12)) {
 
-        /* "pgenlib.pyx":1087
+        /* "pgenlib.pyx":1149
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
- *                 if reterr != kPglRetSuccess:
+ *                 if allele_idx_offsets != NULL:
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
  */
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_variant_index, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_variant_index, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_only); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_only); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1087, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1149, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
         __Pyx_Raise(__pyx_t_4, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __PYX_ERR(0, 1087, __pyx_L1_error)
+        __PYX_ERR(0, 1149, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1086
+        /* "pgenlib.pyx":1148
  *             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
+ *                 if allele_idx_offsets != NULL:
  */
       }
 
-      /* "pgenlib.pyx":1088
+      /* "pgenlib.pyx":1150
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 if allele_idx_offsets != NULL:             # <<<<<<<<<<<<<<
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
+ */
+      __pyx_t_12 = ((__pyx_v_allele_idx_offsets != NULL) != 0);
+      if (__pyx_t_12) {
+
+        /* "pgenlib.pyx":1151
+ *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 if allele_idx_offsets != NULL:
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
+ *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
+ */
+        __pyx_t_12 = ((((__pyx_v_allele_idx_offsets[(__pyx_v_variant_idx + 1)]) - (__pyx_v_allele_idx_offsets[__pyx_v_variant_idx])) != 2) != 0);
+        if (unlikely(__pyx_t_12)) {
+
+          /* "pgenlib.pyx":1152
+ *                 if allele_idx_offsets != NULL:
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
+ *                 if reterr != kPglRetSuccess:
+ */
+          __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__14, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1152, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_4);
+          __Pyx_Raise(__pyx_t_4, 0, 0, 0);
+          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+          __PYX_ERR(0, 1152, __pyx_L1_error)
+
+          /* "pgenlib.pyx":1151
+ *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 if allele_idx_offsets != NULL:
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
+ *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
+ */
+        }
+
+        /* "pgenlib.pyx":1150
+ *                 if variant_idx >= raw_variant_ct:
+ *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 if allele_idx_offsets != NULL:             # <<<<<<<<<<<<<<
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
+ */
+      }
+
+      /* "pgenlib.pyx":1153
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGetP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, __pyx_v_vmaj_geno_iter, __pyx_v_phasepresent, __pyx_v_vmaj_phaseinfo_iter, (&__pyx_v_phasepresent_ct));
 
-      /* "pgenlib.pyx":1089
- *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+      /* "pgenlib.pyx":1154
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 if phasepresent_ct == 0:
  */
       __pyx_t_12 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_12)) {
 
-        /* "pgenlib.pyx":1090
+        /* "pgenlib.pyx":1155
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 if phasepresent_ct == 0:
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
  */
-        __pyx_t_4 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1090, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1155, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1090, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1155, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_error, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1090, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_list_error, __pyx_t_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1155, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1090, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1155, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __Pyx_Raise(__pyx_t_2, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __PYX_ERR(0, 1090, __pyx_L1_error)
+        __PYX_ERR(0, 1155, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1089
- *                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+        /* "pgenlib.pyx":1154
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
  *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 if phasepresent_ct == 0:
  */
       }
 
-      /* "pgenlib.pyx":1091
+      /* "pgenlib.pyx":1156
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 if phasepresent_ct == 0:             # <<<<<<<<<<<<<<
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
  *                 # else:
  */
       __pyx_t_12 = ((__pyx_v_phasepresent_ct == 0) != 0);
       if (__pyx_t_12) {
 
-        /* "pgenlib.pyx":1092
+        /* "pgenlib.pyx":1157
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 if phasepresent_ct == 0:
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)             # <<<<<<<<<<<<<<
  *                 # else:
  *                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
  */
         plink2::ZeroWArr(__pyx_v_sample_ctaw, __pyx_v_vmaj_phaseinfo_iter);
 
-        /* "pgenlib.pyx":1091
+        /* "pgenlib.pyx":1156
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_list() error " + str(reterr))
  *                 if phasepresent_ct == 0:             # <<<<<<<<<<<<<<
  *                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
  *                 # else:
  */
       }
 
-      /* "pgenlib.pyx":1095
+      /* "pgenlib.pyx":1160
  *                 # else:
  *                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
  *                 vmaj_geno_iter = &(vmaj_geno_iter[sample_ctaw2])             # <<<<<<<<<<<<<<
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])
  *             sample_batch_size = kPglNypTransposeBatch
  */
       __pyx_v_vmaj_geno_iter = (&(__pyx_v_vmaj_geno_iter[__pyx_v_sample_ctaw2]));
 
-      /* "pgenlib.pyx":1096
+      /* "pgenlib.pyx":1161
  *                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
  *                 vmaj_geno_iter = &(vmaj_geno_iter[sample_ctaw2])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])             # <<<<<<<<<<<<<<
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  */
       __pyx_v_vmaj_phaseinfo_iter = (&(__pyx_v_vmaj_phaseinfo_iter[__pyx_v_sample_ctaw]));
     }
 
-    /* "pgenlib.pyx":1097
+    /* "pgenlib.pyx":1162
  *                 vmaj_geno_iter = &(vmaj_geno_iter[sample_ctaw2])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])
  *             sample_batch_size = kPglNypTransposeBatch             # <<<<<<<<<<<<<<
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  */
     __pyx_v_sample_batch_size = plink2::kPglNypTransposeBatch;
 
-    /* "pgenlib.pyx":1098
+    /* "pgenlib.pyx":1163
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[sample_ctaw])
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_geno_iter = multivar_vmaj_geno_buf             # <<<<<<<<<<<<<<
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  */
     __pyx_v_vmaj_geno_iter = __pyx_v_multivar_vmaj_geno_buf;
 
-    /* "pgenlib.pyx":1099
+    /* "pgenlib.pyx":1164
  *             sample_batch_size = kPglNypTransposeBatch
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf             # <<<<<<<<<<<<<<
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  */
     __pyx_v_vmaj_phaseinfo_iter = __pyx_v_multivar_vmaj_phaseinfo_buf;
 
-    /* "pgenlib.pyx":1100
+    /* "pgenlib.pyx":1165
  *             vmaj_geno_iter = multivar_vmaj_geno_buf
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):             # <<<<<<<<<<<<<<
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  */
     __pyx_t_19 = __pyx_v_sample_batch_ct;
     __pyx_t_20 = __pyx_t_19;
     for (__pyx_t_21 = 0; __pyx_t_21 < __pyx_t_20; __pyx_t_21+=1) {
       __pyx_v_sample_batch_idx = __pyx_t_21;
 
-      /* "pgenlib.pyx":1101
+      /* "pgenlib.pyx":1166
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  */
       __pyx_t_12 = ((__pyx_v_sample_batch_idx == (__pyx_v_sample_batch_ct - 1)) != 0);
       if (__pyx_t_12) {
 
-        /* "pgenlib.pyx":1102
+        /* "pgenlib.pyx":1167
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)             # <<<<<<<<<<<<<<
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf
  */
         __pyx_t_18 = (__pyx_v_subset_size - 1);
         if (unlikely(plink2::kPglNypTransposeBatch == 0)) {
           PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
-          __PYX_ERR(0, 1102, __pyx_L1_error)
+          __PYX_ERR(0, 1167, __pyx_L1_error)
         }
         __pyx_v_sample_batch_size = (1 + ((uint32_t)__Pyx_mod_long(__pyx_t_18, plink2::kPglNypTransposeBatch)));
 
-        /* "pgenlib.pyx":1101
+        /* "pgenlib.pyx":1166
  *             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
  *             for sample_batch_idx in range(sample_batch_ct):
  *                 if sample_batch_idx == sample_batch_ct - 1:             # <<<<<<<<<<<<<<
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  */
       }
 
-      /* "pgenlib.pyx":1103
+      /* "pgenlib.pyx":1168
  *                 if sample_batch_idx == sample_batch_ct - 1:
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf             # <<<<<<<<<<<<<<
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf
  *                 TransposeNypblock(vmaj_geno_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_geno_iter, transpose_batch_buf)
  */
       __pyx_v_smaj_geno_iter = __pyx_v_multivar_smaj_geno_batch_buf;
 
-      /* "pgenlib.pyx":1104
+      /* "pgenlib.pyx":1169
  *                     sample_batch_size = 1 + <uint32_t>((subset_size - 1) % kPglNypTransposeBatch)
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf             # <<<<<<<<<<<<<<
  *                 TransposeNypblock(vmaj_geno_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_geno_iter, transpose_batch_buf)
  *                 # todo: skip bitblock transpose when all phasepresent_ct values
  */
       __pyx_v_smaj_phaseinfo_iter = __pyx_v_multivar_smaj_phaseinfo_batch_buf;
 
-      /* "pgenlib.pyx":1105
+      /* "pgenlib.pyx":1170
  *                 smaj_geno_iter = multivar_smaj_geno_batch_buf
  *                 smaj_phaseinfo_iter = multivar_smaj_phaseinfo_batch_buf
  *                 TransposeNypblock(vmaj_geno_iter, sample_ctaw2, kPglNypTransposeWords, variant_batch_size, sample_batch_size, smaj_geno_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 # todo: skip bitblock transpose when all phasepresent_ct values
  *                 #       are zero, etc.
  */
       plink2::TransposeNypblock(__pyx_v_vmaj_geno_iter, __pyx_v_sample_ctaw2, plink2::kPglNypTransposeWords, __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_geno_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":1108
+      /* "pgenlib.pyx":1173
  *                 # todo: skip bitblock transpose when all phasepresent_ct values
  *                 #       are zero, etc.
  *                 TransposeBitblock(vmaj_phaseinfo_iter, sample_ctaw, <uint32_t>(kPglNypTransposeWords // 2), variant_batch_size, sample_batch_size, smaj_phaseinfo_iter, transpose_batch_buf)             # <<<<<<<<<<<<<<
  *                 for uii in range(sample_batch_size):
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  */
       plink2::TransposeBitblock(__pyx_v_vmaj_phaseinfo_iter, __pyx_v_sample_ctaw, ((uint32_t)__Pyx_div_long(plink2::kPglNypTransposeWords, 2)), __pyx_v_variant_batch_size, __pyx_v_sample_batch_size, __pyx_v_smaj_phaseinfo_iter, __pyx_v_transpose_batch_buf);
 
-      /* "pgenlib.pyx":1109
+      /* "pgenlib.pyx":1174
  *                 #       are zero, etc.
  *                 TransposeBitblock(vmaj_phaseinfo_iter, sample_ctaw, <uint32_t>(kPglNypTransposeWords // 2), variant_batch_size, sample_batch_size, smaj_phaseinfo_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):             # <<<<<<<<<<<<<<
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  */
       __pyx_t_22 = __pyx_v_sample_batch_size;
       __pyx_t_23 = __pyx_t_22;
       for (__pyx_t_24 = 0; __pyx_t_24 < __pyx_t_23; __pyx_t_24+=1) {
         __pyx_v_uii = __pyx_t_24;
 
-        /* "pgenlib.pyx":1110
+        /* "pgenlib.pyx":1175
  *                 TransposeBitblock(vmaj_phaseinfo_iter, sample_ctaw, <uint32_t>(kPglNypTransposeWords // 2), variant_batch_size, sample_batch_size, smaj_phaseinfo_iter, transpose_batch_buf)
  *                 for uii in range(sample_batch_size):
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))             # <<<<<<<<<<<<<<
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  */
         __pyx_t_16 = (2 * (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch)));
@@ -17070,19 +17663,19 @@
         if (__pyx_t_16 < 0) {
           __pyx_t_16 += __pyx_pybuffernd_allele_int32_out.diminfo[0].shape;
           if (unlikely(__pyx_t_16 < 0)) __pyx_t_6 = 0;
         } else if (unlikely(__pyx_t_16 >= __pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_6 = 0;
         if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_6 = 1;
         if (unlikely(__pyx_t_6 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_6);
-          __PYX_ERR(0, 1110, __pyx_L1_error)
+          __PYX_ERR(0, 1175, __pyx_L1_error)
         }
         __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_16, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_15, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":1111
+        /* "pgenlib.pyx":1176
  *                 for uii in range(sample_batch_size):
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))             # <<<<<<<<<<<<<<
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  */
         __pyx_t_16 = ((2 * (__pyx_v_uii + (__pyx_v_sample_batch_idx * plink2::kPglNypTransposeBatch))) + 1);
@@ -17091,78 +17684,78 @@
         if (__pyx_t_16 < 0) {
           __pyx_t_16 += __pyx_pybuffernd_allele_int32_out.diminfo[0].shape;
           if (unlikely(__pyx_t_16 < 0)) __pyx_t_6 = 0;
         } else if (unlikely(__pyx_t_16 >= __pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_6 = 0;
         if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_6 = 1;
         if (unlikely(__pyx_t_6 != -1)) {
           __Pyx_RaiseBufferIndexError(__pyx_t_6);
-          __PYX_ERR(0, 1111, __pyx_L1_error)
+          __PYX_ERR(0, 1176, __pyx_L1_error)
         }
         __pyx_v_main_data1_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_16, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_15, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-        /* "pgenlib.pyx":1112
+        /* "pgenlib.pyx":1177
  *                     main_data_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch), variant_batch_idx * kPglNypTransposeBatch]))
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)             # <<<<<<<<<<<<<<
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  */
         plink2::GenoarrPhasedToHapCodes(__pyx_v_smaj_geno_iter, __pyx_v_smaj_phaseinfo_iter, __pyx_v_variant_batch_size, __pyx_v_main_data_ptr, __pyx_v_main_data1_ptr);
 
-        /* "pgenlib.pyx":1113
+        /* "pgenlib.pyx":1178
  *                     main_data1_ptr = <int32_t*>(&(allele_int32_out[2 * (uii + sample_batch_idx * kPglNypTransposeBatch) + 1, variant_batch_idx * kPglNypTransposeBatch]))
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  */
         __pyx_v_smaj_geno_iter = (&(__pyx_v_smaj_geno_iter[plink2::kPglNypTransposeWords]));
 
-        /* "pgenlib.pyx":1114
+        /* "pgenlib.pyx":1179
  *                     GenoarrPhasedToHapCodes(smaj_geno_iter, smaj_phaseinfo_iter, variant_batch_size, main_data_ptr, main_data1_ptr)
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])             # <<<<<<<<<<<<<<
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  */
         __pyx_v_smaj_phaseinfo_iter = (&(__pyx_v_smaj_phaseinfo_iter[__Pyx_div_long(plink2::kPglNypTransposeWords, 2)]));
       }
 
-      /* "pgenlib.pyx":1115
+      /* "pgenlib.pyx":1180
  *                     smaj_geno_iter = &(smaj_geno_iter[kPglNypTransposeWords])
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])             # <<<<<<<<<<<<<<
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *         return
  */
       __pyx_v_vmaj_geno_iter = (&(__pyx_v_vmaj_geno_iter[plink2::kPglNypTransposeWords]));
 
-      /* "pgenlib.pyx":1116
+      /* "pgenlib.pyx":1181
  *                     smaj_phaseinfo_iter = &(smaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])             # <<<<<<<<<<<<<<
  *         return
  * 
  */
       __pyx_v_vmaj_phaseinfo_iter = (&(__pyx_v_vmaj_phaseinfo_iter[__Pyx_div_long(plink2::kPglNypTransposeWords, 2)]));
     }
   }
 
-  /* "pgenlib.pyx":1117
+  /* "pgenlib.pyx":1182
  *                 vmaj_geno_iter = &(vmaj_geno_iter[kPglNypTransposeWords])
  *                 vmaj_phaseinfo_iter = &(vmaj_phaseinfo_iter[kPglNypTransposeWords // 2])
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1019
+  /* "pgenlib.pyx":1081
  * 
  * 
  *     cpdef read_alleles_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
@@ -17226,54 +17819,54 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idxs)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_list", 0, 2, 3, 1); __PYX_ERR(0, 1019, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_list", 0, 2, 3, 1); __PYX_ERR(0, 1081, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_hap_maj);
           if (value) { values[2] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_list") < 0)) __PYX_ERR(0, 1019, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_list") < 0)) __PYX_ERR(0, 1081, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_variant_idxs = ((PyArrayObject *)values[0]);
     __pyx_v_allele_int32_out = ((PyArrayObject *)values[1]);
     if (values[2]) {
-      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[2]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1019, __pyx_L3_error)
+      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[2]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1081, __pyx_L3_error)
     } else {
       __pyx_v_hap_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_alleles_list", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1019, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_alleles_list", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1081, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_alleles_list", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 1019, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 1019, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 1081, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 1081, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_24read_alleles_list(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idxs, __pyx_v_allele_int32_out, __pyx_v_hap_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -17300,26 +17893,26 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_allele_int32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_out.refcount = 0;
   __pyx_pybuffernd_allele_int32_out.data = NULL;
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1019, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1081, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1019, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1081, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.hap_maj = __pyx_v_hap_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_allele_int32_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1019, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_allele_int32_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1081, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -17339,37 +17932,33 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1120
+/* "pgenlib.pyx":1185
  * 
  * 
  *     cpdef read_alleles_and_phasepresent_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_27read_alleles_and_phasepresent_range(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_alleles_and_phasepresent_range(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx_start, uint32_t __pyx_v_variant_idx_end, PyArrayObject *__pyx_v_allele_int32_out, PyArrayObject *__pyx_v_phasepresent_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_and_phasepresent_range *__pyx_optional_args) {
   int __pyx_v_hap_maj = ((int)0);
   uintptr_t const *__pyx_v_subset_include_vec;
   struct plink2::PgrSampleSubsetIndexStruct __pyx_v_subset_index;
   struct plink2::PgenReaderStruct *__pyx_v_pgrp;
-  uintptr_t *__pyx_v_genovec;
-  uintptr_t *__pyx_v_phasepresent;
-  uintptr_t *__pyx_v_phaseinfo;
   uint32_t __pyx_v_variant_idx_ct;
   uint32_t __pyx_v_subset_size;
   int32_t *__pyx_v_main_data_ptr;
   unsigned char *__pyx_v_phasepresent_data_ptr;
   uint32_t __pyx_v_variant_idx;
-  uint32_t __pyx_v_phasepresent_ct;
   plink2::PglErr __pyx_v_reterr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_out;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_out;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent_out;
   __Pyx_Buffer __pyx_pybuffer_phasepresent_out;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
@@ -17406,40 +17995,40 @@
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   __pyx_pybuffer_phasepresent_out.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_out.refcount = 0;
   __pyx_pybuffernd_phasepresent_out.data = NULL;
   __pyx_pybuffernd_phasepresent_out.rcbuffer = &__pyx_pybuffer_phasepresent_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1120, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1185, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1120, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1185, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_out.diminfo[0].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_out.diminfo[0].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_phasepresent_out.diminfo[1].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_phasepresent_out.diminfo[1].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_and_phasepresent_ra); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1120, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_and_phasepresent_ra); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_27read_alleles_and_phasepresent_range)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1120, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1185, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1120, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1185, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_5 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1120, __pyx_L1_error)
+        __pyx_t_5 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1185, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_5);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_6 = __pyx_t_1; __pyx_t_7 = NULL;
         __pyx_t_8 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_6))) {
           __pyx_t_7 = PyMethod_GET_SELF(__pyx_t_6);
           if (likely(__pyx_t_7)) {
@@ -17449,35 +18038,35 @@
             __Pyx_DECREF_SET(__pyx_t_6, function);
             __pyx_t_8 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_6)) {
           PyObject *__pyx_temp[6] = {__pyx_t_7, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_allele_int32_out), ((PyObject *)__pyx_v_phasepresent_out), __pyx_t_5};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 5+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1120, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 5+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1185, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_7); __pyx_t_7 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_6)) {
           PyObject *__pyx_temp[6] = {__pyx_t_7, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_allele_int32_out), ((PyObject *)__pyx_v_phasepresent_out), __pyx_t_5};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 5+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1120, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_6, __pyx_temp+1-__pyx_t_8, 5+__pyx_t_8); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1185, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_7); __pyx_t_7 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         } else
         #endif
         {
-          __pyx_t_9 = PyTuple_New(5+__pyx_t_8); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1120, __pyx_L1_error)
+          __pyx_t_9 = PyTuple_New(5+__pyx_t_8); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1185, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_9);
           if (__pyx_t_7) {
             __Pyx_GIVEREF(__pyx_t_7); PyTuple_SET_ITEM(__pyx_t_9, 0, __pyx_t_7); __pyx_t_7 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_9, 0+__pyx_t_8, __pyx_t_3);
           __Pyx_GIVEREF(__pyx_t_4);
@@ -17489,15 +18078,15 @@
           __Pyx_GIVEREF(((PyObject *)__pyx_v_phasepresent_out));
           PyTuple_SET_ITEM(__pyx_t_9, 3+__pyx_t_8, ((PyObject *)__pyx_v_phasepresent_out));
           __Pyx_GIVEREF(__pyx_t_5);
           PyTuple_SET_ITEM(__pyx_t_9, 4+__pyx_t_8, __pyx_t_5);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
           __pyx_t_5 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_9, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1120, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_9, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1185, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
         }
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -17512,543 +18101,513 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1125
+  /* "pgenlib.pyx":1190
  *         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
  *         #   rows, variant_idx_ct columns
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
   __pyx_t_10 = ((__pyx_v_variant_idx_end > (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_10)) {
 
-    /* "pgenlib.pyx":1126
+    /* "pgenlib.pyx":1191
  *         #   rows, variant_idx_ct columns
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_range_variant_idx_e, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_range_variant_idx_e, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1126, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1191, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 1126, __pyx_L1_error)
+    __PYX_ERR(0, 1191, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1125
+    /* "pgenlib.pyx":1190
  *         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
  *         #   rows, variant_idx_ct columns
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
   }
 
-  /* "pgenlib.pyx":1127
+  /* "pgenlib.pyx":1192
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_11 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_11;
 
-  /* "pgenlib.pyx":1128
+  /* "pgenlib.pyx":1193
  *             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
   __pyx_t_12 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_12;
 
-  /* "pgenlib.pyx":1129
+  /* "pgenlib.pyx":1194
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
+ *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
+ *         cdef uint32_t subset_size = self._subset_size
  */
   __pyx_t_13 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_13;
 
-  /* "pgenlib.pyx":1130
+  /* "pgenlib.pyx":1195
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
- */
-  __pyx_t_11 = __pyx_v_self->_genovec;
-  __pyx_v_genovec = __pyx_t_11;
-
-  /* "pgenlib.pyx":1131
- *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
- *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
- */
-  __pyx_t_11 = __pyx_v_self->_phasepresent;
-  __pyx_v_phasepresent = __pyx_t_11;
-
-  /* "pgenlib.pyx":1132
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
- *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
- *         cdef uint32_t subset_size = self._subset_size
- */
-  __pyx_t_11 = __pyx_v_self->_phaseinfo;
-  __pyx_v_phaseinfo = __pyx_t_11;
-
-  /* "pgenlib.pyx":1133
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int32_t* main_data_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":1134
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":1196
+ *         cdef PgenReaderStruct* pgrp = self._state_ptr
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int32_t* main_data_ptr
  *         cdef unsigned char* phasepresent_data_ptr
  */
   __pyx_t_14 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_14;
 
-  /* "pgenlib.pyx":1140
- *         cdef uint32_t phasepresent_ct
+  /* "pgenlib.pyx":1201
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_10 = ((__pyx_v_hap_maj == 0) != 0);
   if (__pyx_t_10) {
 
-    /* "pgenlib.pyx":1141
+    /* "pgenlib.pyx":1202
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  */
     __pyx_t_10 = (((__pyx_v_allele_int32_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_10)) {
 
-      /* "pgenlib.pyx":1142
+      /* "pgenlib.pyx":1203
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1142, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1203, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1142, __pyx_L1_error)
+      __PYX_ERR(0, 1203, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1141
+      /* "pgenlib.pyx":1202
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  */
     }
 
-    /* "pgenlib.pyx":1143
+    /* "pgenlib.pyx":1204
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     __pyx_t_10 = (((__pyx_v_phasepresent_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_10)) {
 
-      /* "pgenlib.pyx":1144
+      /* "pgenlib.pyx":1205
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_2, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_2, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idx_end_variant_idx_sta); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1144, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1205, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1144, __pyx_L1_error)
+      __PYX_ERR(0, 1205, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1143
+      /* "pgenlib.pyx":1204
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1145
+    /* "pgenlib.pyx":1206
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:
  */
     __pyx_t_10 = (((__pyx_v_allele_int32_out->dimensions[1]) < (2 * __pyx_v_subset_size)) != 0);
     if (unlikely(__pyx_t_10)) {
 
-      /* "pgenlib.pyx":1146
+      /* "pgenlib.pyx":1207
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")             # <<<<<<<<<<<<<<
  *             if phasepresent_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_3, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_3, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1146, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1207, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1146, __pyx_L1_error)
+      __PYX_ERR(0, 1207, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1145
+      /* "pgenlib.pyx":1206
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1147
+    /* "pgenlib.pyx":1208
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     __pyx_t_10 = (((__pyx_v_phasepresent_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_10)) {
 
-      /* "pgenlib.pyx":1148
+      /* "pgenlib.pyx":1209
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
- *                 # upgrade to multiallelic version of this function later
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1148, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1209, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1148, __pyx_L1_error)
+      __PYX_ERR(0, 1209, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1147
+      /* "pgenlib.pyx":1208
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  */
     }
 
-    /* "pgenlib.pyx":1149
+    /* "pgenlib.pyx":1210
  *             if phasepresent_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_14 = __pyx_v_variant_idx_end;
     __pyx_t_15 = __pyx_t_14;
     for (__pyx_t_16 = __pyx_v_variant_idx_start; __pyx_t_16 < __pyx_t_15; __pyx_t_16+=1) {
       __pyx_v_variant_idx = __pyx_t_16;
 
-      /* "pgenlib.pyx":1151
+      /* "pgenlib.pyx":1211
+ *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
- *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
+ *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))
  */
-      __pyx_v_reterr = plink2::PgrGetP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, (&__pyx_v_phasepresent_ct));
+      __pyx_v_reterr = plink2::PgrGetMP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, (&__pyx_v_self->_pgv));
 
-      /* "pgenlib.pyx":1152
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+      /* "pgenlib.pyx":1212
+ *             for variant_idx in range(variant_idx_start, variant_idx_end):
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
- *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
+ *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  */
       __pyx_t_10 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_10)) {
 
-        /* "pgenlib.pyx":1153
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1213
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:
- *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[(variant_idx - variant_idx_start), 0]))
  */
-        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_variant_idx_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_variant_idx_2, __pyx_t_6); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_read_alleles_range_error); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_read_alleles_and_phasepresent_r); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_6, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1153, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1213, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         __Pyx_Raise(__pyx_t_2, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __PYX_ERR(0, 1153, __pyx_L1_error)
+        __PYX_ERR(0, 1213, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1152
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1212
+ *             for variant_idx in range(variant_idx_start, variant_idx_end):
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
- *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
+ *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  */
       }
 
-      /* "pgenlib.pyx":1154
+      /* "pgenlib.pyx":1214
  *                 if reterr != kPglRetSuccess:
- *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
+ *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))             # <<<<<<<<<<<<<<
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[(variant_idx - variant_idx_start), 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  */
       __pyx_t_17 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
       __pyx_t_18 = 0;
       __pyx_t_8 = -1;
       if (unlikely(__pyx_t_17 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_8 = 0;
       if (__pyx_t_18 < 0) {
         __pyx_t_18 += __pyx_pybuffernd_allele_int32_out.diminfo[1].shape;
         if (unlikely(__pyx_t_18 < 0)) __pyx_t_8 = 1;
       } else if (unlikely(__pyx_t_18 >= __pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_8 = 1;
       if (unlikely(__pyx_t_8 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_8);
-        __PYX_ERR(0, 1154, __pyx_L1_error)
+        __PYX_ERR(0, 1214, __pyx_L1_error)
       }
       __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_17, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_18, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1155
- *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
+      /* "pgenlib.pyx":1215
+ *                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[(variant_idx - variant_idx_start), 0]))             # <<<<<<<<<<<<<<
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return
  */
       __pyx_t_17 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
       __pyx_t_18 = 0;
       __pyx_t_8 = -1;
       if (unlikely(__pyx_t_17 >= (size_t)__pyx_pybuffernd_phasepresent_out.diminfo[0].shape)) __pyx_t_8 = 0;
       if (__pyx_t_18 < 0) {
         __pyx_t_18 += __pyx_pybuffernd_phasepresent_out.diminfo[1].shape;
         if (unlikely(__pyx_t_18 < 0)) __pyx_t_8 = 1;
       } else if (unlikely(__pyx_t_18 >= __pyx_pybuffernd_phasepresent_out.diminfo[1].shape)) __pyx_t_8 = 1;
       if (unlikely(__pyx_t_8 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_8);
-        __PYX_ERR(0, 1155, __pyx_L1_error)
+        __PYX_ERR(0, 1215, __pyx_L1_error)
       }
       __pyx_v_phasepresent_data_ptr = ((unsigned char *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_uint8_t *, __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.buf, __pyx_t_17, __pyx_pybuffernd_phasepresent_out.diminfo[0].strides, __pyx_t_18, __pyx_pybuffernd_phasepresent_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1156
+      /* "pgenlib.pyx":1216
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[(variant_idx - variant_idx_start), 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)             # <<<<<<<<<<<<<<
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         raise RuntimeError("read_alleles_and_phasepresent_range() does not support hap_maj == 1 yet.")
  */
-      plink2::GenoarrPhasedToAlleleCodesMinus9(__pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, __pyx_v_subset_size, __pyx_v_phasepresent_ct, __pyx_v_phasepresent_data_ptr, __pyx_v_main_data_ptr);
+      plink2::GenoarrMPToAlleleCodesMinus9((&__pyx_v_self->_pgv), __pyx_v_subset_size, __pyx_v_phasepresent_data_ptr, __pyx_v_main_data_ptr);
     }
 
-    /* "pgenlib.pyx":1157
+    /* "pgenlib.pyx":1217
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[(variant_idx - variant_idx_start), 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         raise RuntimeError("read_alleles_and_phasepresent_range() does not support hap_maj == 1 yet.")
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1140
- *         cdef uint32_t phasepresent_ct
+    /* "pgenlib.pyx":1201
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":1158
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+  /* "pgenlib.pyx":1218
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return
  *         raise RuntimeError("read_alleles_and_phasepresent_range() does not support hap_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__13, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1158, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__15, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1218, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
   __Pyx_Raise(__pyx_t_2, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __PYX_ERR(0, 1158, __pyx_L1_error)
+  __PYX_ERR(0, 1218, __pyx_L1_error)
 
-  /* "pgenlib.pyx":1120
+  /* "pgenlib.pyx":1185
  * 
  * 
  *     cpdef read_alleles_and_phasepresent_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
@@ -18120,70 +18679,70 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_start)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_end)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, 1); __PYX_ERR(0, 1120, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, 1); __PYX_ERR(0, 1185, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, 2); __PYX_ERR(0, 1120, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, 2); __PYX_ERR(0, 1185, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  3:
         if (likely((values[3] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_phasepresent_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, 3); __PYX_ERR(0, 1120, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, 3); __PYX_ERR(0, 1185, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  4:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_hap_maj);
           if (value) { values[4] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_and_phasepresent_range") < 0)) __PYX_ERR(0, 1120, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_and_phasepresent_range") < 0)) __PYX_ERR(0, 1185, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
         CYTHON_FALLTHROUGH;
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1120, __pyx_L3_error)
-    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1120, __pyx_L3_error)
+    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1185, __pyx_L3_error)
+    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1185, __pyx_L3_error)
     __pyx_v_allele_int32_out = ((PyArrayObject *)values[2]);
     __pyx_v_phasepresent_out = ((PyArrayObject *)values[3]);
     if (values[4]) {
-      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[4]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1120, __pyx_L3_error)
+      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[4]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1185, __pyx_L3_error)
     } else {
       __pyx_v_hap_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1120, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_range", 0, 4, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1185, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_alleles_and_phasepresent_range", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 1120, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_out), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_out", 0))) __PYX_ERR(0, 1120, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 1185, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_out), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_out", 0))) __PYX_ERR(0, 1185, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_26read_alleles_and_phasepresent_range(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, __pyx_v_hap_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -18210,26 +18769,26 @@
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   __pyx_pybuffer_phasepresent_out.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_out.refcount = 0;
   __pyx_pybuffernd_phasepresent_out.data = NULL;
   __pyx_pybuffernd_phasepresent_out.rcbuffer = &__pyx_pybuffer_phasepresent_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1120, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1185, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1120, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1185, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_out.diminfo[0].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_out.diminfo[0].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_phasepresent_out.diminfo[1].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_phasepresent_out.diminfo[1].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.hap_maj = __pyx_v_hap_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_and_phasepresent_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1120, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_and_phasepresent_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -18249,39 +18808,38 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1161
+/* "pgenlib.pyx":1221
  * 
  * 
  *     cpdef read_alleles_and_phasepresent_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,cast=True,mode="c",ndim=2] phasepresent_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_29read_alleles_and_phasepresent_list(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_read_alleles_and_phasepresent_list(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, PyArrayObject *__pyx_v_variant_idxs, PyArrayObject *__pyx_v_allele_int32_out, PyArrayObject *__pyx_v_phasepresent_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_alleles_and_phasepresent_list *__pyx_optional_args) {
   int __pyx_v_hap_maj = ((int)0);
   uint32_t __pyx_v_raw_variant_ct;
   uintptr_t const *__pyx_v_subset_include_vec;
   struct plink2::PgrSampleSubsetIndexStruct __pyx_v_subset_index;
   struct plink2::PgenReaderStruct *__pyx_v_pgrp;
-  uintptr_t *__pyx_v_genovec;
-  uintptr_t *__pyx_v_phasepresent;
-  uintptr_t *__pyx_v_phaseinfo;
+  CYTHON_UNUSED uintptr_t *__pyx_v_genovec;
+  CYTHON_UNUSED uintptr_t *__pyx_v_phasepresent;
+  CYTHON_UNUSED uintptr_t *__pyx_v_phaseinfo;
   uint32_t __pyx_v_variant_idx_ct;
   uint32_t __pyx_v_subset_size;
   int32_t *__pyx_v_main_data_ptr;
   unsigned char *__pyx_v_phasepresent_data_ptr;
   uint32_t __pyx_v_variant_list_idx;
   uint32_t __pyx_v_variant_idx;
-  uint32_t __pyx_v_phasepresent_ct;
   plink2::PglErr __pyx_v_reterr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_out;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_out;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent_out;
   __Pyx_Buffer __pyx_pybuffer_phasepresent_out;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_variant_idxs;
   __Pyx_Buffer __pyx_pybuffer_variant_idxs;
@@ -18322,41 +18880,41 @@
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   __pyx_pybuffer_phasepresent_out.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_out.refcount = 0;
   __pyx_pybuffernd_phasepresent_out.data = NULL;
   __pyx_pybuffernd_phasepresent_out.rcbuffer = &__pyx_pybuffer_phasepresent_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_out.diminfo[0].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_out.diminfo[0].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_phasepresent_out.diminfo[1].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_phasepresent_out.diminfo[1].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_and_phasepresent_li); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_alleles_and_phasepresent_li); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1221, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_29read_alleles_and_phasepresent_list)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1161, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_hap_maj); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1221, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -18366,31 +18924,31 @@
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[5] = {__pyx_t_5, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_allele_int32_out), ((PyObject *)__pyx_v_phasepresent_out), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 4+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1161, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 4+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1221, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[5] = {__pyx_t_5, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_allele_int32_out), ((PyObject *)__pyx_v_phasepresent_out), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 4+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1161, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 4+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1221, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(4+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1161, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(4+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1221, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_variant_idxs));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_variant_idxs));
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, ((PyObject *)__pyx_v_variant_idxs));
@@ -18399,15 +18957,15 @@
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, ((PyObject *)__pyx_v_allele_int32_out));
           __Pyx_INCREF(((PyObject *)__pyx_v_phasepresent_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_phasepresent_out));
           PyTuple_SET_ITEM(__pyx_t_7, 2+__pyx_t_6, ((PyObject *)__pyx_v_phasepresent_out));
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 3+__pyx_t_6, __pyx_t_3);
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1161, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1221, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -18422,557 +18980,557 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1166
+  /* "pgenlib.pyx":1226
  *         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
  *         #   rows, variant_idx_ct columns
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_8 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_8;
 
-  /* "pgenlib.pyx":1167
+  /* "pgenlib.pyx":1227
  *         #   rows, variant_idx_ct columns
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_9 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_9;
 
-  /* "pgenlib.pyx":1168
+  /* "pgenlib.pyx":1228
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_10 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_10;
 
-  /* "pgenlib.pyx":1169
+  /* "pgenlib.pyx":1229
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
  */
   __pyx_t_11 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_11;
 
-  /* "pgenlib.pyx":1170
+  /* "pgenlib.pyx":1230
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  */
-  __pyx_t_9 = __pyx_v_self->_genovec;
+  __pyx_t_9 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_9;
 
-  /* "pgenlib.pyx":1171
+  /* "pgenlib.pyx":1231
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
-  __pyx_t_9 = __pyx_v_self->_phasepresent;
+  __pyx_t_9 = __pyx_v_self->_pgv.phasepresent;
   __pyx_v_phasepresent = __pyx_t_9;
 
-  /* "pgenlib.pyx":1172
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1232
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_9 = __pyx_v_self->_phaseinfo;
+  __pyx_t_9 = __pyx_v_self->_pgv.phaseinfo;
   __pyx_v_phaseinfo = __pyx_t_9;
 
-  /* "pgenlib.pyx":1173
- *         cdef uintptr_t* phasepresent = self._phasepresent
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":1233
+ *         cdef uintptr_t* phasepresent = self._pgv.phasepresent
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef int32_t* main_data_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":1174
- *         cdef uintptr_t* phaseinfo = self._phaseinfo
+  /* "pgenlib.pyx":1234
+ *         cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef int32_t* main_data_ptr
  *         cdef unsigned char* phasepresent_data_ptr
  */
   __pyx_t_8 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_8;
 
-  /* "pgenlib.pyx":1181
- *         cdef uint32_t phasepresent_ct
+  /* "pgenlib.pyx":1240
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_12 = ((__pyx_v_hap_maj == 0) != 0);
   if (__pyx_t_12) {
 
-    /* "pgenlib.pyx":1182
+    /* "pgenlib.pyx":1241
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  */
     __pyx_t_12 = (((__pyx_v_allele_int32_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1183
+      /* "pgenlib.pyx":1242
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_5, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_5, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1183, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1242, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1183, __pyx_L1_error)
+      __PYX_ERR(0, 1242, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1182
+      /* "pgenlib.pyx":1241
  *         cdef PglErr reterr
  *         if hap_maj == 0:
  *             if allele_int32_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  */
     }
 
-    /* "pgenlib.pyx":1184
+    /* "pgenlib.pyx":1243
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     __pyx_t_12 = (((__pyx_v_phasepresent_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1185
+      /* "pgenlib.pyx":1244
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[0])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_6, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_6, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1185, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1244, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1185, __pyx_L1_error)
+      __PYX_ERR(0, 1244, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1184
+      /* "pgenlib.pyx":1243
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if phasepresent_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1186
+    /* "pgenlib.pyx":1245
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:
  */
     __pyx_t_12 = (((__pyx_v_allele_int32_out->dimensions[1]) < (2 * __pyx_v_subset_size)) != 0);
     if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1187
+      /* "pgenlib.pyx":1246
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")             # <<<<<<<<<<<<<<
  *             if phasepresent_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_allele_int32_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_7, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_7, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_and_column_count_should_be_twic_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1187, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1246, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1187, __pyx_L1_error)
+      __PYX_ERR(0, 1246, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1186
+      /* "pgenlib.pyx":1245
  *             if phasepresent_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if allele_int32_out.shape[1] < 2 * subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1188
+    /* "pgenlib.pyx":1247
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_12 = (((__pyx_v_phasepresent_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1189
+      /* "pgenlib.pyx":1248
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_phasepresent_out->dimensions[1])); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_8, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_alleles_and_p_8, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_4, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1189, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1248, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1189, __pyx_L1_error)
+      __PYX_ERR(0, 1248, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1188
+      /* "pgenlib.pyx":1247
  *             if allele_int32_out.shape[1] < 2 * subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
  *             if phasepresent_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":1190
+    /* "pgenlib.pyx":1249
  *             if phasepresent_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_8 = __pyx_v_variant_idx_ct;
     __pyx_t_13 = __pyx_t_8;
     for (__pyx_t_14 = 0; __pyx_t_14 < __pyx_t_13; __pyx_t_14+=1) {
       __pyx_v_variant_list_idx = __pyx_t_14;
 
-      /* "pgenlib.pyx":1191
+      /* "pgenlib.pyx":1250
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_15 = __pyx_v_variant_list_idx;
       __pyx_t_6 = -1;
       if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_6 = 0;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1191, __pyx_L1_error)
+        __PYX_ERR(0, 1250, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_15, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":1192
+      /* "pgenlib.pyx":1251
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 # upgrade to multiallelic version of this function later
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  */
       __pyx_t_12 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_12)) {
 
-        /* "pgenlib.pyx":1193
+        /* "pgenlib.pyx":1252
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_li_2, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_li_2, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1193, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1252, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __Pyx_Raise(__pyx_t_1, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __PYX_ERR(0, 1193, __pyx_L1_error)
+        __PYX_ERR(0, 1252, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1192
+        /* "pgenlib.pyx":1251
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 # upgrade to multiallelic version of this function later
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  */
       }
 
-      /* "pgenlib.pyx":1195
+      /* "pgenlib.pyx":1253
+ *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))
  */
-      __pyx_v_reterr = plink2::PgrGetP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, (&__pyx_v_phasepresent_ct));
+      __pyx_v_reterr = plink2::PgrGetMP(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, (&__pyx_v_self->_pgv));
 
-      /* "pgenlib.pyx":1196
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+      /* "pgenlib.pyx":1254
+ *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  */
       __pyx_t_12 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_12)) {
 
-        /* "pgenlib.pyx":1197
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1255
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[variant_list_idx, 0]))
  */
-        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1197, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1197, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1255, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_li_3, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1197, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_alleles_and_phasepresent_li_3, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1197, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1255, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         __Pyx_Raise(__pyx_t_4, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-        __PYX_ERR(0, 1197, __pyx_L1_error)
+        __PYX_ERR(0, 1255, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1196
- *                 # upgrade to multiallelic version of this function later
- *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+        /* "pgenlib.pyx":1254
+ *                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+ *                 reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  */
       }
 
-      /* "pgenlib.pyx":1198
+      /* "pgenlib.pyx":1256
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))             # <<<<<<<<<<<<<<
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[variant_list_idx, 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  */
       __pyx_t_15 = __pyx_v_variant_list_idx;
       __pyx_t_16 = 0;
       __pyx_t_6 = -1;
       if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_allele_int32_out.diminfo[0].shape)) __pyx_t_6 = 0;
       if (__pyx_t_16 < 0) {
         __pyx_t_16 += __pyx_pybuffernd_allele_int32_out.diminfo[1].shape;
         if (unlikely(__pyx_t_16 < 0)) __pyx_t_6 = 1;
       } else if (unlikely(__pyx_t_16 >= __pyx_pybuffernd_allele_int32_out.diminfo[1].shape)) __pyx_t_6 = 1;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1198, __pyx_L1_error)
+        __PYX_ERR(0, 1256, __pyx_L1_error)
       }
       __pyx_v_main_data_ptr = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.buf, __pyx_t_15, __pyx_pybuffernd_allele_int32_out.diminfo[0].strides, __pyx_t_16, __pyx_pybuffernd_allele_int32_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1199
+      /* "pgenlib.pyx":1257
  *                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[variant_list_idx, 0]))             # <<<<<<<<<<<<<<
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return
  */
       __pyx_t_15 = __pyx_v_variant_list_idx;
       __pyx_t_16 = 0;
       __pyx_t_6 = -1;
       if (unlikely(__pyx_t_15 >= (size_t)__pyx_pybuffernd_phasepresent_out.diminfo[0].shape)) __pyx_t_6 = 0;
       if (__pyx_t_16 < 0) {
         __pyx_t_16 += __pyx_pybuffernd_phasepresent_out.diminfo[1].shape;
         if (unlikely(__pyx_t_16 < 0)) __pyx_t_6 = 1;
       } else if (unlikely(__pyx_t_16 >= __pyx_pybuffernd_phasepresent_out.diminfo[1].shape)) __pyx_t_6 = 1;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1199, __pyx_L1_error)
+        __PYX_ERR(0, 1257, __pyx_L1_error)
       }
       __pyx_v_phasepresent_data_ptr = ((unsigned char *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_uint8_t *, __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.buf, __pyx_t_15, __pyx_pybuffernd_phasepresent_out.diminfo[0].strides, __pyx_t_16, __pyx_pybuffernd_phasepresent_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1200
+      /* "pgenlib.pyx":1258
  *                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[variant_list_idx, 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)             # <<<<<<<<<<<<<<
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         raise RuntimeError("read_alleles_and_phasepresent_list() does not support hap_maj == 1 yet.")
  */
-      plink2::GenoarrPhasedToAlleleCodesMinus9(__pyx_v_genovec, __pyx_v_phasepresent, __pyx_v_phaseinfo, __pyx_v_subset_size, __pyx_v_phasepresent_ct, __pyx_v_phasepresent_data_ptr, __pyx_v_main_data_ptr);
+      plink2::GenoarrMPToAlleleCodesMinus9((&__pyx_v_self->_pgv), __pyx_v_subset_size, __pyx_v_phasepresent_data_ptr, __pyx_v_main_data_ptr);
     }
 
-    /* "pgenlib.pyx":1201
+    /* "pgenlib.pyx":1259
  *                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[variant_list_idx, 0]))
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         raise RuntimeError("read_alleles_and_phasepresent_list() does not support hap_maj == 1 yet.")
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1181
- *         cdef uint32_t phasepresent_ct
+    /* "pgenlib.pyx":1240
+ *         cdef uint32_t variant_idx
  *         cdef PglErr reterr
  *         if hap_maj == 0:             # <<<<<<<<<<<<<<
  *             if allele_int32_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":1202
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+  /* "pgenlib.pyx":1260
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return
  *         raise RuntimeError("read_alleles_and_phasepresent_list() does not support hap_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__14, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1202, __pyx_L1_error)
+  __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__16, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1260, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_4);
   __Pyx_Raise(__pyx_t_4, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-  __PYX_ERR(0, 1202, __pyx_L1_error)
+  __PYX_ERR(0, 1260, __pyx_L1_error)
 
-  /* "pgenlib.pyx":1161
+  /* "pgenlib.pyx":1221
  * 
  * 
  *     cpdef read_alleles_and_phasepresent_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,cast=True,mode="c",ndim=2] phasepresent_out, bint hap_maj = 0):             # <<<<<<<<<<<<<<
  *         # if hap_maj == False, allele_int32_out must have at least
  *         #   variant_idx_ct rows, 2 * sample_ct columns
  */
 
@@ -19041,31 +19599,31 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idxs)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_list", 0, 3, 4, 1); __PYX_ERR(0, 1161, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_list", 0, 3, 4, 1); __PYX_ERR(0, 1221, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_phasepresent_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_list", 0, 3, 4, 2); __PYX_ERR(0, 1161, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_list", 0, 3, 4, 2); __PYX_ERR(0, 1221, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_hap_maj);
           if (value) { values[3] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_and_phasepresent_list") < 0)) __PYX_ERR(0, 1161, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_alleles_and_phasepresent_list") < 0)) __PYX_ERR(0, 1221, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
@@ -19074,30 +19632,30 @@
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_variant_idxs = ((PyArrayObject *)values[0]);
     __pyx_v_allele_int32_out = ((PyArrayObject *)values[1]);
     __pyx_v_phasepresent_out = ((PyArrayObject *)values[2]);
     if (values[3]) {
-      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1161, __pyx_L3_error)
+      __pyx_v_hap_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_hap_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1221, __pyx_L3_error)
     } else {
       __pyx_v_hap_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_list", 0, 3, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1161, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_alleles_and_phasepresent_list", 0, 3, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1221, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_alleles_and_phasepresent_list", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 1161, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 1161, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_out), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_out", 0))) __PYX_ERR(0, 1161, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 1221, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_out), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_out", 0))) __PYX_ERR(0, 1221, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_out), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_out", 0))) __PYX_ERR(0, 1221, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_28read_alleles_and_phasepresent_list(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idxs, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, __pyx_v_hap_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -19130,31 +19688,31 @@
   __pyx_pybuffernd_allele_int32_out.rcbuffer = &__pyx_pybuffer_allele_int32_out;
   __pyx_pybuffer_phasepresent_out.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_out.refcount = 0;
   __pyx_pybuffernd_phasepresent_out.data = NULL;
   __pyx_pybuffernd_phasepresent_out.rcbuffer = &__pyx_pybuffer_phasepresent_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_out.diminfo[0].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_out.diminfo[0].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_out.diminfo[1].strides = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_out.diminfo[1].shape = __pyx_pybuffernd_allele_int32_out.rcbuffer->pybuffer.shape[1];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_out.diminfo[0].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_out.diminfo[0].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_phasepresent_out.diminfo[1].strides = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_phasepresent_out.diminfo[1].shape = __pyx_pybuffernd_phasepresent_out.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.hap_maj = __pyx_v_hap_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_and_phasepresent_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1161, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_alleles_and_phasepresent_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_allele_int32_out, __pyx_v_phasepresent_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1221, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -19176,15 +19734,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1205
+/* "pgenlib.pyx":1263
  * 
  * 
  *     cdef read_dosages_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -19234,170 +19792,170 @@
   }
   __pyx_pybuffer_floatarr_out.pybuffer.buf = NULL;
   __pyx_pybuffer_floatarr_out.refcount = 0;
   __pyx_pybuffernd_floatarr_out.data = NULL;
   __pyx_pybuffernd_floatarr_out.rcbuffer = &__pyx_pybuffer_floatarr_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1205, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1263, __pyx_L1_error)
   }
   __pyx_pybuffernd_floatarr_out.diminfo[0].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_floatarr_out.diminfo[0].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_floatarr_out.diminfo[1].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_floatarr_out.diminfo[1].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":1206
+  /* "pgenlib.pyx":1264
  * 
  *     cdef read_dosages_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_1 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1207
+  /* "pgenlib.pyx":1265
  *     cdef read_dosages_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_2 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_2;
 
-  /* "pgenlib.pyx":1208
+  /* "pgenlib.pyx":1266
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
  */
   __pyx_t_3 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_3;
 
-  /* "pgenlib.pyx":1209
+  /* "pgenlib.pyx":1267
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  */
-  __pyx_t_1 = __pyx_v_self->_genovec;
+  __pyx_t_1 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1210
+  /* "pgenlib.pyx":1268
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present             # <<<<<<<<<<<<<<
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
-  __pyx_t_1 = __pyx_v_self->_dosage_present;
+  __pyx_t_1 = __pyx_v_self->_pgv.dosage_present;
   __pyx_v_dosage_present = __pyx_t_1;
 
-  /* "pgenlib.pyx":1211
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1269
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_4 = __pyx_v_self->_dosage_main;
+  __pyx_t_4 = __pyx_v_self->_pgv.dosage_main;
   __pyx_v_dosage_main = __pyx_t_4;
 
-  /* "pgenlib.pyx":1212
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1270
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef float* data32_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":1213
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1271
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef float* data32_ptr
  *         cdef uint32_t variant_idx
  */
   __pyx_t_5 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_5;
 
-  /* "pgenlib.pyx":1218
+  /* "pgenlib.pyx":1276
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
   __pyx_t_6 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_6) {
 
-    /* "pgenlib.pyx":1219
+    /* "pgenlib.pyx":1277
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_5 = __pyx_v_variant_idx_end;
     __pyx_t_7 = __pyx_t_5;
     for (__pyx_t_8 = __pyx_v_variant_idx_start; __pyx_t_8 < __pyx_t_7; __pyx_t_8+=1) {
       __pyx_v_variant_idx = __pyx_t_8;
 
-      /* "pgenlib.pyx":1220
+      /* "pgenlib.pyx":1278
  *         if sample_maj == 0:
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1D(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-      /* "pgenlib.pyx":1221
+      /* "pgenlib.pyx":1279
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  */
       __pyx_t_6 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_6)) {
 
-        /* "pgenlib.pyx":1222
+        /* "pgenlib.pyx":1280
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data32_ptr = <float*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  */
-        __pyx_t_9 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1222, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1280, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
-        __pyx_t_10 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1222, __pyx_L1_error)
+        __pyx_t_10 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1280, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_10);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_range_error, __pyx_t_10); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1222, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_range_error, __pyx_t_10); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1280, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-        __pyx_t_10 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1222, __pyx_L1_error)
+        __pyx_t_10 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1280, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_10);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
         __Pyx_Raise(__pyx_t_10, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-        __PYX_ERR(0, 1222, __pyx_L1_error)
+        __PYX_ERR(0, 1280, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1221
+        /* "pgenlib.pyx":1279
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  */
       }
 
-      /* "pgenlib.pyx":1223
+      /* "pgenlib.pyx":1281
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))             # <<<<<<<<<<<<<<
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return
  */
       __pyx_t_11 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
@@ -19406,62 +19964,62 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_floatarr_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_12 < 0) {
         __pyx_t_12 += __pyx_pybuffernd_floatarr_out.diminfo[1].shape;
         if (unlikely(__pyx_t_12 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_12 >= __pyx_pybuffernd_floatarr_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 1223, __pyx_L1_error)
+        __PYX_ERR(0, 1281, __pyx_L1_error)
       }
       __pyx_v_data32_ptr = ((float *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_float32_t *, __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_floatarr_out.diminfo[0].strides, __pyx_t_12, __pyx_pybuffernd_floatarr_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1224
+      /* "pgenlib.pyx":1282
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")
  */
       plink2::Dosage16ToFloatsMinus9(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data32_ptr);
     }
 
-    /* "pgenlib.pyx":1225
+    /* "pgenlib.pyx":1283
  *                 data32_ptr = <float*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1218
+    /* "pgenlib.pyx":1276
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
   }
 
-  /* "pgenlib.pyx":1226
+  /* "pgenlib.pyx":1284
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_t_10 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__15, NULL); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1226, __pyx_L1_error)
+  __pyx_t_10 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__17, NULL); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1284, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_10);
   __Pyx_Raise(__pyx_t_10, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-  __PYX_ERR(0, 1226, __pyx_L1_error)
+  __PYX_ERR(0, 1284, __pyx_L1_error)
 
-  /* "pgenlib.pyx":1205
+  /* "pgenlib.pyx":1263
  * 
  * 
  *     cdef read_dosages_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -19482,15 +20040,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1229
+/* "pgenlib.pyx":1287
  * 
  * 
  *     cdef read_dosages_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -19540,170 +20098,170 @@
   }
   __pyx_pybuffer_floatarr_out.pybuffer.buf = NULL;
   __pyx_pybuffer_floatarr_out.refcount = 0;
   __pyx_pybuffernd_floatarr_out.data = NULL;
   __pyx_pybuffernd_floatarr_out.rcbuffer = &__pyx_pybuffer_floatarr_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1229, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1287, __pyx_L1_error)
   }
   __pyx_pybuffernd_floatarr_out.diminfo[0].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_floatarr_out.diminfo[0].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_floatarr_out.diminfo[1].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_floatarr_out.diminfo[1].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":1230
+  /* "pgenlib.pyx":1288
  * 
  *     cdef read_dosages_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_1 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1231
+  /* "pgenlib.pyx":1289
  *     cdef read_dosages_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_2 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_2;
 
-  /* "pgenlib.pyx":1232
+  /* "pgenlib.pyx":1290
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
  */
   __pyx_t_3 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_3;
 
-  /* "pgenlib.pyx":1233
+  /* "pgenlib.pyx":1291
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  */
-  __pyx_t_1 = __pyx_v_self->_genovec;
+  __pyx_t_1 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1234
+  /* "pgenlib.pyx":1292
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present             # <<<<<<<<<<<<<<
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  */
-  __pyx_t_1 = __pyx_v_self->_dosage_present;
+  __pyx_t_1 = __pyx_v_self->_pgv.dosage_present;
   __pyx_v_dosage_present = __pyx_t_1;
 
-  /* "pgenlib.pyx":1235
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1293
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_4 = __pyx_v_self->_dosage_main;
+  __pyx_t_4 = __pyx_v_self->_pgv.dosage_main;
   __pyx_v_dosage_main = __pyx_t_4;
 
-  /* "pgenlib.pyx":1236
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1294
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef double* data64_ptr
  */
   __pyx_v_variant_idx_ct = (__pyx_v_variant_idx_end - __pyx_v_variant_idx_start);
 
-  /* "pgenlib.pyx":1237
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1295
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef double* data64_ptr
  *         cdef uint32_t variant_idx
  */
   __pyx_t_5 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_5;
 
-  /* "pgenlib.pyx":1242
+  /* "pgenlib.pyx":1300
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
   __pyx_t_6 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_6) {
 
-    /* "pgenlib.pyx":1243
+    /* "pgenlib.pyx":1301
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             for variant_idx in range(variant_idx_start, variant_idx_end):             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_5 = __pyx_v_variant_idx_end;
     __pyx_t_7 = __pyx_t_5;
     for (__pyx_t_8 = __pyx_v_variant_idx_start; __pyx_t_8 < __pyx_t_7; __pyx_t_8+=1) {
       __pyx_v_variant_idx = __pyx_t_8;
 
-      /* "pgenlib.pyx":1244
+      /* "pgenlib.pyx":1302
  *         if sample_maj == 0:
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1D(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-      /* "pgenlib.pyx":1245
+      /* "pgenlib.pyx":1303
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  */
       __pyx_t_6 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_6)) {
 
-        /* "pgenlib.pyx":1246
+        /* "pgenlib.pyx":1304
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data64_ptr = <double*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  */
-        __pyx_t_9 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1246, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1304, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
-        __pyx_t_10 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1246, __pyx_L1_error)
+        __pyx_t_10 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1304, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_10);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_range_error, __pyx_t_10); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1246, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_range_error, __pyx_t_10); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1304, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-        __pyx_t_10 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1246, __pyx_L1_error)
+        __pyx_t_10 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1304, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_10);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
         __Pyx_Raise(__pyx_t_10, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-        __PYX_ERR(0, 1246, __pyx_L1_error)
+        __PYX_ERR(0, 1304, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1245
+        /* "pgenlib.pyx":1303
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  */
       }
 
-      /* "pgenlib.pyx":1247
+      /* "pgenlib.pyx":1305
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))             # <<<<<<<<<<<<<<
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  *             return
  */
       __pyx_t_11 = (__pyx_v_variant_idx - __pyx_v_variant_idx_start);
@@ -19712,62 +20270,62 @@
       if (unlikely(__pyx_t_11 >= (size_t)__pyx_pybuffernd_floatarr_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_12 < 0) {
         __pyx_t_12 += __pyx_pybuffernd_floatarr_out.diminfo[1].shape;
         if (unlikely(__pyx_t_12 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_12 >= __pyx_pybuffernd_floatarr_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 1247, __pyx_L1_error)
+        __PYX_ERR(0, 1305, __pyx_L1_error)
       }
       __pyx_v_data64_ptr = ((double *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_float64_t *, __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.buf, __pyx_t_11, __pyx_pybuffernd_floatarr_out.diminfo[0].strides, __pyx_t_12, __pyx_pybuffernd_floatarr_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1248
+      /* "pgenlib.pyx":1306
  *                     raise RuntimeError("read_dosages_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")
  */
       plink2::Dosage16ToDoublesMinus9(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data64_ptr);
     }
 
-    /* "pgenlib.pyx":1249
+    /* "pgenlib.pyx":1307
  *                 data64_ptr = <double*>(&(floatarr_out[(variant_idx - variant_idx_start), 0]))
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1242
+    /* "pgenlib.pyx":1300
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             for variant_idx in range(variant_idx_start, variant_idx_end):
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
   }
 
-  /* "pgenlib.pyx":1250
+  /* "pgenlib.pyx":1308
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  *             return
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_t_10 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__15, NULL); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1250, __pyx_L1_error)
+  __pyx_t_10 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__17, NULL); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1308, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_10);
   __Pyx_Raise(__pyx_t_10, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-  __PYX_ERR(0, 1250, __pyx_L1_error)
+  __PYX_ERR(0, 1308, __pyx_L1_error)
 
-  /* "pgenlib.pyx":1229
+  /* "pgenlib.pyx":1287
  * 
  * 
  *     cdef read_dosages_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
@@ -19788,15 +20346,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1253
+/* "pgenlib.pyx":1311
  * 
  * 
  *     cpdef read_dosages_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -19836,25 +20394,25 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_dosages_range); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1253, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_dosages_range); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1311, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_31read_dosages_range)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1253, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_start); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1311, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1253, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1311, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
-        __pyx_t_5 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1253, __pyx_L1_error)
+        __pyx_t_5 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1311, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_5);
-        __pyx_t_6 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1253, __pyx_L1_error)
+        __pyx_t_6 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1311, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_6);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_7 = __pyx_t_1; __pyx_t_8 = NULL;
         __pyx_t_9 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_7))) {
           __pyx_t_8 = PyMethod_GET_SELF(__pyx_t_7);
           if (likely(__pyx_t_8)) {
@@ -19864,37 +20422,37 @@
             __Pyx_DECREF_SET(__pyx_t_7, function);
             __pyx_t_9 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_7)) {
           PyObject *__pyx_temp[6] = {__pyx_t_8, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_floatarr_out), __pyx_t_5, __pyx_t_6};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1253, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1311, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_8); __pyx_t_8 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_7)) {
           PyObject *__pyx_temp[6] = {__pyx_t_8, __pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_floatarr_out), __pyx_t_5, __pyx_t_6};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1253, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_7, __pyx_temp+1-__pyx_t_9, 5+__pyx_t_9); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1311, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_8); __pyx_t_8 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         } else
         #endif
         {
-          __pyx_t_10 = PyTuple_New(5+__pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1253, __pyx_L1_error)
+          __pyx_t_10 = PyTuple_New(5+__pyx_t_9); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1311, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_10);
           if (__pyx_t_8) {
             __Pyx_GIVEREF(__pyx_t_8); PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_8); __pyx_t_8 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_10, 0+__pyx_t_9, __pyx_t_3);
           __Pyx_GIVEREF(__pyx_t_4);
@@ -19906,15 +20464,15 @@
           PyTuple_SET_ITEM(__pyx_t_10, 3+__pyx_t_9, __pyx_t_5);
           __Pyx_GIVEREF(__pyx_t_6);
           PyTuple_SET_ITEM(__pyx_t_10, 4+__pyx_t_9, __pyx_t_6);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
           __pyx_t_5 = 0;
           __pyx_t_6 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_7, __pyx_t_10, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1253, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_7, __pyx_t_10, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1311, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
         }
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -19929,188 +20487,188 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1254
+  /* "pgenlib.pyx":1312
  * 
  *     cpdef read_dosages_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.dtype == np.float32:
  */
   __pyx_t_11 = ((__pyx_v_variant_idx_end > (__pyx_v_self->_info_ptr[0]).raw_variant_ct) != 0);
   if (unlikely(__pyx_t_11)) {
 
-    /* "pgenlib.pyx":1255
+    /* "pgenlib.pyx":1313
  *     cpdef read_dosages_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  */
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_end); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_range_variant_idx_e, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_range_variant_idx_e, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_only); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t((__pyx_v_self->_info_ptr[0]).raw_variant_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1255, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1313, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 1255, __pyx_L1_error)
+    __PYX_ERR(0, 1313, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1254
+    /* "pgenlib.pyx":1312
  * 
  *     cpdef read_dosages_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.dtype == np.float32:
  */
   }
 
-  /* "pgenlib.pyx":1256
+  /* "pgenlib.pyx":1314
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1256, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1314, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_7, __pyx_n_s_np); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1256, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_7, __pyx_n_s_np); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1314, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_7);
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_7, __pyx_n_s_float32); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1256, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_7, __pyx_n_s_float32); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1314, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-  __pyx_t_7 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_7); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1256, __pyx_L1_error)
+  __pyx_t_7 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_7); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1314, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_7); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 1256, __pyx_L1_error)
+  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_7); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 1314, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
   if (__pyx_t_11) {
 
-    /* "pgenlib.pyx":1257
+    /* "pgenlib.pyx":1315
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         elif floatarr_out.dtype == np.float64:
  *             self.read_dosages_range_internal64(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  */
     __pyx_t_12.__pyx_n = 2;
     __pyx_t_12.allele_idx = __pyx_v_allele_idx;
     __pyx_t_12.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_7 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_range_internal32(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_12); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1257, __pyx_L1_error)
+    __pyx_t_7 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_range_internal32(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_12); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1315, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
 
-    /* "pgenlib.pyx":1256
+    /* "pgenlib.pyx":1314
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":1258
+  /* "pgenlib.pyx":1316
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.read_dosages_range_internal64(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         else:
  */
-  __pyx_t_7 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1258, __pyx_L1_error)
+  __pyx_t_7 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1316, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_7);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1258, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1316, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1258, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1316, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_7, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1258, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_7, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1316, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 1258, __pyx_L1_error)
+  __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_11 < 0)) __PYX_ERR(0, 1316, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (likely(__pyx_t_11)) {
 
-    /* "pgenlib.pyx":1259
+    /* "pgenlib.pyx":1317
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:
  *             self.read_dosages_range_internal64(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid read_dosages_range() floatarr_out array element type (float32 or float64 expected).")
  */
     __pyx_t_13.__pyx_n = 2;
     __pyx_t_13.allele_idx = __pyx_v_allele_idx;
     __pyx_t_13.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_range_internal64(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_13); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1259, __pyx_L1_error)
+    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_range_internal64(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_13); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1317, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
 
-    /* "pgenlib.pyx":1258
+    /* "pgenlib.pyx":1316
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_range_internal32(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.read_dosages_range_internal64(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         else:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":1261
+  /* "pgenlib.pyx":1319
  *             self.read_dosages_range_internal64(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_dosages_range() floatarr_out array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__16, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1261, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__18, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1319, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 1261, __pyx_L1_error)
+    __PYX_ERR(0, 1319, __pyx_L1_error)
   }
   __pyx_L4:;
 
-  /* "pgenlib.pyx":1262
+  /* "pgenlib.pyx":1320
  *         else:
  *             raise RuntimeError("Invalid read_dosages_range() floatarr_out array element type (float32 or float64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1253
+  /* "pgenlib.pyx":1311
  * 
  * 
  *     cpdef read_dosages_range(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
  *             raise RuntimeError("read_dosages_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  */
 
@@ -20172,21 +20730,21 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_start)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx_end)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_dosages_range", 0, 3, 5, 1); __PYX_ERR(0, 1253, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_dosages_range", 0, 3, 5, 1); __PYX_ERR(0, 1311, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_floatarr_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_dosages_range", 0, 3, 5, 2); __PYX_ERR(0, 1253, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_dosages_range", 0, 3, 5, 2); __PYX_ERR(0, 1311, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[3] = value; kw_args--; }
         }
@@ -20194,52 +20752,52 @@
         case  4:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_maj);
           if (value) { values[4] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_dosages_range") < 0)) __PYX_ERR(0, 1253, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_dosages_range") < 0)) __PYX_ERR(0, 1311, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
         CYTHON_FALLTHROUGH;
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1253, __pyx_L3_error)
-    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1253, __pyx_L3_error)
+    __pyx_v_variant_idx_start = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx_start == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1311, __pyx_L3_error)
+    __pyx_v_variant_idx_end = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_variant_idx_end == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1311, __pyx_L3_error)
     __pyx_v_floatarr_out = ((PyArrayObject *)values[2]);
     if (values[3]) {
-      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[3]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1253, __pyx_L3_error)
+      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[3]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1311, __pyx_L3_error)
     } else {
       __pyx_v_allele_idx = ((uint32_t)1);
     }
     if (values[4]) {
-      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[4]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1253, __pyx_L3_error)
+      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[4]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1311, __pyx_L3_error)
     } else {
       __pyx_v_sample_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_dosages_range", 0, 3, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1253, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_dosages_range", 0, 3, 5, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1311, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_dosages_range", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_out), __pyx_ptype_5numpy_ndarray, 1, "floatarr_out", 0))) __PYX_ERR(0, 1253, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_out), __pyx_ptype_5numpy_ndarray, 1, "floatarr_out", 0))) __PYX_ERR(0, 1311, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_30read_dosages_range(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_floatarr_out, __pyx_v_allele_idx, __pyx_v_sample_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -20256,15 +20814,15 @@
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("read_dosages_range", 0);
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 2;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
   __pyx_t_2.sample_maj = __pyx_v_sample_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_dosages_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_floatarr_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1253, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_dosages_range(__pyx_v_self, __pyx_v_variant_idx_start, __pyx_v_variant_idx_end, __pyx_v_floatarr_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1311, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -20273,15 +20831,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1265
+/* "pgenlib.pyx":1323
  * 
  * 
  *     cdef read_dosages_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -20340,369 +20898,369 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_floatarr_out.pybuffer.buf = NULL;
   __pyx_pybuffer_floatarr_out.refcount = 0;
   __pyx_pybuffernd_floatarr_out.data = NULL;
   __pyx_pybuffernd_floatarr_out.rcbuffer = &__pyx_pybuffer_floatarr_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1265, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1323, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1265, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1323, __pyx_L1_error)
   }
   __pyx_pybuffernd_floatarr_out.diminfo[0].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_floatarr_out.diminfo[0].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_floatarr_out.diminfo[1].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_floatarr_out.diminfo[1].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":1266
+  /* "pgenlib.pyx":1324
  * 
  *     cdef read_dosages_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_1;
 
-  /* "pgenlib.pyx":1267
+  /* "pgenlib.pyx":1325
  *     cdef read_dosages_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_2 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_2;
 
-  /* "pgenlib.pyx":1268
+  /* "pgenlib.pyx":1326
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_3 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_3;
 
-  /* "pgenlib.pyx":1269
+  /* "pgenlib.pyx":1327
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
  */
   __pyx_t_4 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_4;
 
-  /* "pgenlib.pyx":1270
+  /* "pgenlib.pyx":1328
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  */
-  __pyx_t_2 = __pyx_v_self->_genovec;
+  __pyx_t_2 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_2;
 
-  /* "pgenlib.pyx":1271
+  /* "pgenlib.pyx":1329
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present             # <<<<<<<<<<<<<<
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
-  __pyx_t_2 = __pyx_v_self->_dosage_present;
+  __pyx_t_2 = __pyx_v_self->_pgv.dosage_present;
   __pyx_v_dosage_present = __pyx_t_2;
 
-  /* "pgenlib.pyx":1272
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1330
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_5 = __pyx_v_self->_dosage_main;
+  __pyx_t_5 = __pyx_v_self->_pgv.dosage_main;
   __pyx_v_dosage_main = __pyx_t_5;
 
-  /* "pgenlib.pyx":1273
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1331
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef float* data32_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":1274
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1332
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef float* data32_ptr
  *         cdef uint32_t variant_list_idx
  */
   __pyx_t_1 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_1;
 
-  /* "pgenlib.pyx":1280
+  /* "pgenlib.pyx":1338
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_6 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_6) {
 
-    /* "pgenlib.pyx":1281
+    /* "pgenlib.pyx":1339
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if floatarr_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:
  */
     __pyx_t_6 = (((__pyx_v_floatarr_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_6)) {
 
-      /* "pgenlib.pyx":1282
+      /* "pgenlib.pyx":1340
  *         if sample_maj == 0:
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if floatarr_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1282, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1340, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
       __Pyx_Raise(__pyx_t_7, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __PYX_ERR(0, 1282, __pyx_L1_error)
+      __PYX_ERR(0, 1340, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1281
+      /* "pgenlib.pyx":1339
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if floatarr_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1283
+    /* "pgenlib.pyx":1341
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_6 = (((__pyx_v_floatarr_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_6)) {
 
-      /* "pgenlib.pyx":1284
+      /* "pgenlib.pyx":1342
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list_2, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list_2, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_9, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_9, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1284, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1342, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_7, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __PYX_ERR(0, 1284, __pyx_L1_error)
+      __PYX_ERR(0, 1342, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1283
+      /* "pgenlib.pyx":1341
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":1285
+    /* "pgenlib.pyx":1343
  *             if floatarr_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_1 = __pyx_v_variant_idx_ct;
     __pyx_t_10 = __pyx_t_1;
     for (__pyx_t_11 = 0; __pyx_t_11 < __pyx_t_10; __pyx_t_11+=1) {
       __pyx_v_variant_list_idx = __pyx_t_11;
 
-      /* "pgenlib.pyx":1286
+      /* "pgenlib.pyx":1344
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_12 = __pyx_v_variant_list_idx;
       __pyx_t_13 = -1;
       if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_13 = 0;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 1286, __pyx_L1_error)
+        __PYX_ERR(0, 1344, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":1287
+      /* "pgenlib.pyx":1345
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
       __pyx_t_6 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_6)) {
 
-        /* "pgenlib.pyx":1288
+        /* "pgenlib.pyx":1346
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_list_variant_index, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_list_variant_index, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_only); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_only); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1288, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1346, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 1288, __pyx_L1_error)
+        __PYX_ERR(0, 1346, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1287
+        /* "pgenlib.pyx":1345
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
       }
 
-      /* "pgenlib.pyx":1289
+      /* "pgenlib.pyx":1347
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1D(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-      /* "pgenlib.pyx":1290
+      /* "pgenlib.pyx":1348
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[variant_list_idx, 0]))
  */
       __pyx_t_6 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_6)) {
 
-        /* "pgenlib.pyx":1291
+        /* "pgenlib.pyx":1349
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data32_ptr = <float*>(&(floatarr_out[variant_list_idx, 0]))
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  */
-        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1291, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1349, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1291, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1349, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1291, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1349, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __pyx_t_9 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1291, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1349, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_9, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __PYX_ERR(0, 1291, __pyx_L1_error)
+        __PYX_ERR(0, 1349, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1290
+        /* "pgenlib.pyx":1348
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[variant_list_idx, 0]))
  */
       }
 
-      /* "pgenlib.pyx":1292
+      /* "pgenlib.pyx":1350
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[variant_list_idx, 0]))             # <<<<<<<<<<<<<<
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return
  */
       __pyx_t_12 = __pyx_v_variant_list_idx;
@@ -20711,62 +21269,62 @@
       if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_floatarr_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_14 < 0) {
         __pyx_t_14 += __pyx_pybuffernd_floatarr_out.diminfo[1].shape;
         if (unlikely(__pyx_t_14 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_14 >= __pyx_pybuffernd_floatarr_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 1292, __pyx_L1_error)
+        __PYX_ERR(0, 1350, __pyx_L1_error)
       }
       __pyx_v_data32_ptr = ((float *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_float32_t *, __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_floatarr_out.diminfo[0].strides, __pyx_t_14, __pyx_pybuffernd_floatarr_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1293
+      /* "pgenlib.pyx":1351
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data32_ptr = <float*>(&(floatarr_out[variant_list_idx, 0]))
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")
  */
       plink2::Dosage16ToFloatsMinus9(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data32_ptr);
     }
 
-    /* "pgenlib.pyx":1294
+    /* "pgenlib.pyx":1352
  *                 data32_ptr = <float*>(&(floatarr_out[variant_list_idx, 0]))
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1280
+    /* "pgenlib.pyx":1338
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":1295
+  /* "pgenlib.pyx":1353
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_t_9 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__17, NULL); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1295, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__19, NULL); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1353, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_9);
   __Pyx_Raise(__pyx_t_9, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-  __PYX_ERR(0, 1295, __pyx_L1_error)
+  __PYX_ERR(0, 1353, __pyx_L1_error)
 
-  /* "pgenlib.pyx":1265
+  /* "pgenlib.pyx":1323
  * 
  * 
  *     cdef read_dosages_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -20790,15 +21348,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1298
+/* "pgenlib.pyx":1356
  * 
  * 
  *     cdef read_dosages_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -20857,369 +21415,369 @@
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   __pyx_pybuffer_floatarr_out.pybuffer.buf = NULL;
   __pyx_pybuffer_floatarr_out.refcount = 0;
   __pyx_pybuffernd_floatarr_out.data = NULL;
   __pyx_pybuffernd_floatarr_out.rcbuffer = &__pyx_pybuffer_floatarr_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1298, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1356, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1298, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1356, __pyx_L1_error)
   }
   __pyx_pybuffernd_floatarr_out.diminfo[0].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_floatarr_out.diminfo[0].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_floatarr_out.diminfo[1].strides = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_floatarr_out.diminfo[1].shape = __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":1299
+  /* "pgenlib.pyx":1357
  * 
  *     cdef read_dosages_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct             # <<<<<<<<<<<<<<
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr[0]).raw_variant_ct;
   __pyx_v_raw_variant_ct = __pyx_t_1;
 
-  /* "pgenlib.pyx":1300
+  /* "pgenlib.pyx":1358
  *     cdef read_dosages_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
  */
   __pyx_t_2 = __pyx_v_self->_subset_include_vec;
   __pyx_v_subset_include_vec = __pyx_t_2;
 
-  /* "pgenlib.pyx":1301
+  /* "pgenlib.pyx":1359
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* genovec = self._pgv.genovec
  */
   __pyx_t_3 = __pyx_v_self->_subset_index;
   __pyx_v_subset_index = __pyx_t_3;
 
-  /* "pgenlib.pyx":1302
+  /* "pgenlib.pyx":1360
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
  */
   __pyx_t_4 = __pyx_v_self->_state_ptr;
   __pyx_v_pgrp = __pyx_t_4;
 
-  /* "pgenlib.pyx":1303
+  /* "pgenlib.pyx":1361
  *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  */
-  __pyx_t_2 = __pyx_v_self->_genovec;
+  __pyx_t_2 = __pyx_v_self->_pgv.genovec;
   __pyx_v_genovec = __pyx_t_2;
 
-  /* "pgenlib.pyx":1304
+  /* "pgenlib.pyx":1362
  *         cdef PgenReaderStruct* pgrp = self._state_ptr
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
- *         cdef uint16_t* dosage_main = self._dosage_main
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present             # <<<<<<<<<<<<<<
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  */
-  __pyx_t_2 = __pyx_v_self->_dosage_present;
+  __pyx_t_2 = __pyx_v_self->_pgv.dosage_present;
   __pyx_v_dosage_present = __pyx_t_2;
 
-  /* "pgenlib.pyx":1305
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1363
+ *         cdef uintptr_t* genovec = self._pgv.genovec
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_5 = __pyx_v_self->_dosage_main;
+  __pyx_t_5 = __pyx_v_self->_pgv.dosage_main;
   __pyx_v_dosage_main = __pyx_t_5;
 
-  /* "pgenlib.pyx":1306
- *         cdef uintptr_t* dosage_present = self._dosage_present
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1364
+ *         cdef uintptr_t* dosage_present = self._pgv.dosage_present
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         cdef double* data64_ptr
  */
   __pyx_v_variant_idx_ct = ((uint32_t)(__pyx_v_variant_idxs->dimensions[0]));
 
-  /* "pgenlib.pyx":1307
- *         cdef uint16_t* dosage_main = self._dosage_main
+  /* "pgenlib.pyx":1365
+ *         cdef uint16_t* dosage_main = self._pgv.dosage_main
  *         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
  *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
  *         cdef double* data64_ptr
  *         cdef uint32_t variant_list_idx
  */
   __pyx_t_1 = __pyx_v_self->_subset_size;
   __pyx_v_subset_size = __pyx_t_1;
 
-  /* "pgenlib.pyx":1313
+  /* "pgenlib.pyx":1371
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   __pyx_t_6 = ((__pyx_v_sample_maj == 0) != 0);
   if (__pyx_t_6) {
 
-    /* "pgenlib.pyx":1314
+    /* "pgenlib.pyx":1372
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if floatarr_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:
  */
     __pyx_t_6 = (((__pyx_v_floatarr_out->dimensions[0]) < __pyx_v_variant_idx_ct) != 0);
     if (unlikely(__pyx_t_6)) {
 
-      /* "pgenlib.pyx":1315
+      /* "pgenlib.pyx":1373
  *         if sample_maj == 0:
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")             # <<<<<<<<<<<<<<
  *             if floatarr_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  */
-      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[0])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_variant_idxs_length_is); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1315, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1373, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
       __Pyx_Raise(__pyx_t_7, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __PYX_ERR(0, 1315, __pyx_L1_error)
+      __PYX_ERR(0, 1373, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1314
+      /* "pgenlib.pyx":1372
  *         cdef PglErr reterr
  *         if sample_maj == 0:
  *             if floatarr_out.shape[0] < variant_idx_ct:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:
  */
     }
 
-    /* "pgenlib.pyx":1316
+    /* "pgenlib.pyx":1374
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     __pyx_t_6 = (((__pyx_v_floatarr_out->dimensions[1]) < __pyx_v_subset_size) != 0);
     if (unlikely(__pyx_t_6)) {
 
-      /* "pgenlib.pyx":1317
+      /* "pgenlib.pyx":1375
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")             # <<<<<<<<<<<<<<
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  */
-      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_Py_intptr_t((__pyx_v_floatarr_out->dimensions[1])); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list_2, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_Variant_major_read_dosages_list_2, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_current_sample_subset_has_size); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_9);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_subset_size); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_9, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_9, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u__9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1317, __pyx_L1_error)
+      __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1375, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_7);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __Pyx_Raise(__pyx_t_7, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __PYX_ERR(0, 1317, __pyx_L1_error)
+      __PYX_ERR(0, 1375, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1316
+      /* "pgenlib.pyx":1374
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  *             if floatarr_out.shape[1] < subset_size:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  */
     }
 
-    /* "pgenlib.pyx":1318
+    /* "pgenlib.pyx":1376
  *             if floatarr_out.shape[1] < subset_size:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):             # <<<<<<<<<<<<<<
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  */
     __pyx_t_1 = __pyx_v_variant_idx_ct;
     __pyx_t_10 = __pyx_t_1;
     for (__pyx_t_11 = 0; __pyx_t_11 < __pyx_t_10; __pyx_t_11+=1) {
       __pyx_v_variant_list_idx = __pyx_t_11;
 
-      /* "pgenlib.pyx":1319
+      /* "pgenlib.pyx":1377
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few columns (" + str(floatarr_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]             # <<<<<<<<<<<<<<
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  */
       __pyx_t_12 = __pyx_v_variant_list_idx;
       __pyx_t_13 = -1;
       if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_variant_idxs.diminfo[0].shape)) __pyx_t_13 = 0;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 1319, __pyx_L1_error)
+        __PYX_ERR(0, 1377, __pyx_L1_error)
       }
       __pyx_v_variant_idx = (*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_variant_idxs.diminfo[0].strides));
 
-      /* "pgenlib.pyx":1320
+      /* "pgenlib.pyx":1378
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
       __pyx_t_6 = ((__pyx_v_variant_idx >= __pyx_v_raw_variant_ct) != 0);
       if (unlikely(__pyx_t_6)) {
 
-        /* "pgenlib.pyx":1321
+        /* "pgenlib.pyx":1379
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")             # <<<<<<<<<<<<<<
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  */
-        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_list_variant_index, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_dosages_list_variant_index, __pyx_t_8); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_only); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_8 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_only); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_8);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_uint32_t(__pyx_v_raw_variant_ct); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_t_8, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyUnicode_Concat(__pyx_t_7, __pyx_kp_u_in_file_2); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1321, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1379, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
         __Pyx_Raise(__pyx_t_7, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __PYX_ERR(0, 1321, __pyx_L1_error)
+        __PYX_ERR(0, 1379, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1320
+        /* "pgenlib.pyx":1378
  *             for variant_list_idx in range(variant_idx_ct):
  *                 variant_idx = variant_idxs[variant_list_idx]
  *                 if variant_idx >= raw_variant_ct:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  */
       }
 
-      /* "pgenlib.pyx":1322
+      /* "pgenlib.pyx":1380
  *                 if variant_idx >= raw_variant_ct:
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  */
       __pyx_v_reterr = plink2::PgrGet1D(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_allele_idx, __pyx_v_pgrp, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-      /* "pgenlib.pyx":1323
+      /* "pgenlib.pyx":1381
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[variant_list_idx, 0]))
  */
       __pyx_t_6 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
       if (unlikely(__pyx_t_6)) {
 
-        /* "pgenlib.pyx":1324
+        /* "pgenlib.pyx":1382
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))             # <<<<<<<<<<<<<<
  *                 data64_ptr = <double*>(&(floatarr_out[variant_list_idx, 0]))
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  */
-        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1324, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1382, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
-        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1324, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1382, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1324, __pyx_L1_error)
+        __pyx_t_7 = __Pyx_PyUnicode_Concat(__pyx_kp_u_read_range_error, __pyx_t_9); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1382, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_7);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __pyx_t_9 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1324, __pyx_L1_error)
+        __pyx_t_9 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_7); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1382, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_9);
         __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         __Pyx_Raise(__pyx_t_9, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-        __PYX_ERR(0, 1324, __pyx_L1_error)
+        __PYX_ERR(0, 1382, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1323
+        /* "pgenlib.pyx":1381
  *                     raise RuntimeError("read_dosages_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
  *                 reterr = PgrGet1D(subset_include_vec, subset_index, subset_size, variant_idx, allele_idx, pgrp, genovec, dosage_present, dosage_main, &dosage_ct)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[variant_list_idx, 0]))
  */
       }
 
-      /* "pgenlib.pyx":1325
+      /* "pgenlib.pyx":1383
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[variant_list_idx, 0]))             # <<<<<<<<<<<<<<
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  *             return
  */
       __pyx_t_12 = __pyx_v_variant_list_idx;
@@ -21228,62 +21786,62 @@
       if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_floatarr_out.diminfo[0].shape)) __pyx_t_13 = 0;
       if (__pyx_t_14 < 0) {
         __pyx_t_14 += __pyx_pybuffernd_floatarr_out.diminfo[1].shape;
         if (unlikely(__pyx_t_14 < 0)) __pyx_t_13 = 1;
       } else if (unlikely(__pyx_t_14 >= __pyx_pybuffernd_floatarr_out.diminfo[1].shape)) __pyx_t_13 = 1;
       if (unlikely(__pyx_t_13 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_13);
-        __PYX_ERR(0, 1325, __pyx_L1_error)
+        __PYX_ERR(0, 1383, __pyx_L1_error)
       }
       __pyx_v_data64_ptr = ((double *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_float64_t *, __pyx_pybuffernd_floatarr_out.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_floatarr_out.diminfo[0].strides, __pyx_t_14, __pyx_pybuffernd_floatarr_out.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1326
+      /* "pgenlib.pyx":1384
  *                     raise RuntimeError("read_range() error " + str(reterr))
  *                 data64_ptr = <double*>(&(floatarr_out[variant_list_idx, 0]))
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)             # <<<<<<<<<<<<<<
  *             return
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")
  */
       plink2::Dosage16ToDoublesMinus9(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_subset_size, __pyx_v_dosage_ct, __pyx_v_data64_ptr);
     }
 
-    /* "pgenlib.pyx":1327
+    /* "pgenlib.pyx":1385
  *                 data64_ptr = <double*>(&(floatarr_out[variant_list_idx, 0]))
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  *             return             # <<<<<<<<<<<<<<
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1313
+    /* "pgenlib.pyx":1371
  *         cdef uint32_t dosage_ct
  *         cdef PglErr reterr
  *         if sample_maj == 0:             # <<<<<<<<<<<<<<
  *             if floatarr_out.shape[0] < variant_idx_ct:
  *                 raise RuntimeError("Variant-major read_dosages_list() floatarr_out buffer has too few rows (" + str(floatarr_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
  */
   }
 
-  /* "pgenlib.pyx":1328
+  /* "pgenlib.pyx":1386
  *                 Dosage16ToDoublesMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data64_ptr)
  *             return
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_t_9 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__17, NULL); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1328, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__19, NULL); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 1386, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_9);
   __Pyx_Raise(__pyx_t_9, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
-  __PYX_ERR(0, 1328, __pyx_L1_error)
+  __PYX_ERR(0, 1386, __pyx_L1_error)
 
-  /* "pgenlib.pyx":1298
+  /* "pgenlib.pyx":1356
  * 
  * 
  *     cdef read_dosages_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
  *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
  */
 
@@ -21307,15 +21865,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1331
+/* "pgenlib.pyx":1389
  * 
  * 
  *     cpdef read_dosages_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  */
 
@@ -21352,33 +21910,33 @@
   }
   __pyx_pybuffer_variant_idxs.pybuffer.buf = NULL;
   __pyx_pybuffer_variant_idxs.refcount = 0;
   __pyx_pybuffernd_variant_idxs.data = NULL;
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1331, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1389, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_dosages_list); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1331, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_read_dosages_list); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1389, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_33read_dosages_list)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1331, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1389, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
-        __pyx_t_4 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1331, __pyx_L1_error)
+        __pyx_t_4 = __Pyx_PyBool_FromLong(__pyx_v_sample_maj); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1389, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_4);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_5 = __pyx_t_1; __pyx_t_6 = NULL;
         __pyx_t_7 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_5))) {
           __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_5);
           if (likely(__pyx_t_6)) {
@@ -21388,33 +21946,33 @@
             __Pyx_DECREF_SET(__pyx_t_5, function);
             __pyx_t_7 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[5] = {__pyx_t_6, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_floatarr_out), __pyx_t_3, __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1331, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1389, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_5)) {
           PyObject *__pyx_temp[5] = {__pyx_t_6, ((PyObject *)__pyx_v_variant_idxs), ((PyObject *)__pyx_v_floatarr_out), __pyx_t_3, __pyx_t_4};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1331, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_5, __pyx_temp+1-__pyx_t_7, 4+__pyx_t_7); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1389, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_6); __pyx_t_6 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
           __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         } else
         #endif
         {
-          __pyx_t_8 = PyTuple_New(4+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1331, __pyx_L1_error)
+          __pyx_t_8 = PyTuple_New(4+__pyx_t_7); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1389, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_8);
           if (__pyx_t_6) {
             __Pyx_GIVEREF(__pyx_t_6); PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_6); __pyx_t_6 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_variant_idxs));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_variant_idxs));
           PyTuple_SET_ITEM(__pyx_t_8, 0+__pyx_t_7, ((PyObject *)__pyx_v_variant_idxs));
@@ -21423,15 +21981,15 @@
           PyTuple_SET_ITEM(__pyx_t_8, 1+__pyx_t_7, ((PyObject *)__pyx_v_floatarr_out));
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_8, 2+__pyx_t_7, __pyx_t_3);
           __Pyx_GIVEREF(__pyx_t_4);
           PyTuple_SET_ITEM(__pyx_t_8, 3+__pyx_t_7, __pyx_t_4);
           __pyx_t_3 = 0;
           __pyx_t_4 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1331, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1389, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
         }
         __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -21446,132 +22004,132 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1332
+  /* "pgenlib.pyx":1390
  * 
  *     cpdef read_dosages_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if floatarr_out.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1332, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1390, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1332, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1390, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float32); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1332, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float32); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1390, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1332, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_5, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1390, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 1332, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 1390, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (__pyx_t_9) {
 
-    /* "pgenlib.pyx":1333
+    /* "pgenlib.pyx":1391
  *     cpdef read_dosages_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         elif floatarr_out.dtype == np.float64:
  *             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)
  */
     __pyx_t_10.__pyx_n = 2;
     __pyx_t_10.allele_idx = __pyx_v_allele_idx;
     __pyx_t_10.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_list_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_10); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1333, __pyx_L1_error)
+    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_list_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_10); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1391, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
 
-    /* "pgenlib.pyx":1332
+    /* "pgenlib.pyx":1390
  * 
  *     cpdef read_dosages_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
  *         if floatarr_out.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":1334
+  /* "pgenlib.pyx":1392
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         else:
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1334, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_out), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1392, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1334, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_5, __pyx_n_s_np); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1392, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_5);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1334, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_5, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1392, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-  __pyx_t_5 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1334, __pyx_L1_error)
+  __pyx_t_5 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1392, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 1334, __pyx_L1_error)
+  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 1392, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
   if (likely(__pyx_t_9)) {
 
-    /* "pgenlib.pyx":1335
+    /* "pgenlib.pyx":1393
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:
  *             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid read_dosages_list() floatarr_out array element type (float32 or float64 expected).")
  */
     __pyx_t_11.__pyx_n = 2;
     __pyx_t_11.allele_idx = __pyx_v_allele_idx;
     __pyx_t_11.sample_maj = __pyx_v_sample_maj;
-    __pyx_t_5 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_list_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_11); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1335, __pyx_L1_error)
+    __pyx_t_5 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->read_dosages_list_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_variant_idxs), ((PyArrayObject *)__pyx_v_floatarr_out), &__pyx_t_11); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1393, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
 
-    /* "pgenlib.pyx":1334
+    /* "pgenlib.pyx":1392
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         elif floatarr_out.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         else:
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":1337
+  /* "pgenlib.pyx":1395
  *             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_dosages_list() floatarr_out array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__18, NULL); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1337, __pyx_L1_error)
+    __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__20, NULL); if (unlikely(!__pyx_t_5)) __PYX_ERR(0, 1395, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_5);
     __Pyx_Raise(__pyx_t_5, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
-    __PYX_ERR(0, 1337, __pyx_L1_error)
+    __PYX_ERR(0, 1395, __pyx_L1_error)
   }
   __pyx_L3:;
 
-  /* "pgenlib.pyx":1338
+  /* "pgenlib.pyx":1396
  *         else:
  *             raise RuntimeError("Invalid read_dosages_list() floatarr_out array element type (float32 or float64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1331
+  /* "pgenlib.pyx":1389
  * 
  * 
  *     cpdef read_dosages_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):             # <<<<<<<<<<<<<<
  *         if floatarr_out.dtype == np.float32:
  *             self.read_dosages_list_internal32(variant_idxs, floatarr_out, allele_idx, sample_maj)
  */
 
@@ -21637,15 +22195,15 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idxs)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_floatarr_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("read_dosages_list", 0, 2, 4, 1); __PYX_ERR(0, 1331, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("read_dosages_list", 0, 2, 4, 1); __PYX_ERR(0, 1389, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[2] = value; kw_args--; }
         }
@@ -21653,15 +22211,15 @@
         case  3:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_maj);
           if (value) { values[3] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_dosages_list") < 0)) __PYX_ERR(0, 1331, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "read_dosages_list") < 0)) __PYX_ERR(0, 1389, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
         CYTHON_FALLTHROUGH;
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
@@ -21670,34 +22228,34 @@
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_variant_idxs = ((PyArrayObject *)values[0]);
     __pyx_v_floatarr_out = ((PyArrayObject *)values[1]);
     if (values[2]) {
-      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1331, __pyx_L3_error)
+      __pyx_v_allele_idx = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_allele_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1389, __pyx_L3_error)
     } else {
       __pyx_v_allele_idx = ((uint32_t)1);
     }
     if (values[3]) {
-      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1331, __pyx_L3_error)
+      __pyx_v_sample_maj = __Pyx_PyObject_IsTrue(values[3]); if (unlikely((__pyx_v_sample_maj == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1389, __pyx_L3_error)
     } else {
       __pyx_v_sample_maj = ((int)0);
     }
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("read_dosages_list", 0, 2, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1331, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("read_dosages_list", 0, 2, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1389, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.read_dosages_list", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 1331, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_out), __pyx_ptype_5numpy_ndarray, 1, "floatarr_out", 0))) __PYX_ERR(0, 1331, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_variant_idxs), __pyx_ptype_5numpy_ndarray, 1, "variant_idxs", 0))) __PYX_ERR(0, 1389, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_out), __pyx_ptype_5numpy_ndarray, 1, "floatarr_out", 0))) __PYX_ERR(0, 1389, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_32read_dosages_list(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idxs, __pyx_v_floatarr_out, __pyx_v_allele_idx, __pyx_v_sample_maj);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -21718,22 +22276,22 @@
   __Pyx_RefNannySetupContext("read_dosages_list", 0);
   __pyx_pybuffer_variant_idxs.pybuffer.buf = NULL;
   __pyx_pybuffer_variant_idxs.refcount = 0;
   __pyx_pybuffernd_variant_idxs.data = NULL;
   __pyx_pybuffernd_variant_idxs.rcbuffer = &__pyx_pybuffer_variant_idxs;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1331, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer, (PyObject*)__pyx_v_variant_idxs, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1389, __pyx_L1_error)
   }
   __pyx_pybuffernd_variant_idxs.diminfo[0].strides = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_variant_idxs.diminfo[0].shape = __pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 2;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
   __pyx_t_2.sample_maj = __pyx_v_sample_maj;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_dosages_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_floatarr_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1331, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->read_dosages_list(__pyx_v_self, __pyx_v_variant_idxs, __pyx_v_floatarr_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1389, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -21751,42 +22309,67 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_variant_idxs.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1340
+/* "pgenlib.pyx":1398
  *         return
  * 
  *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):             # <<<<<<<<<<<<<<
- *         # todo: multiallelic variants
- *         if allele_idx is None:
+ *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenReader_35count(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenReader_count(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self, uint32_t __pyx_v_variant_idx, PyArrayObject *__pyx_v_genocount_uint32_out, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_count *__pyx_optional_args) {
   PyObject *__pyx_v_allele_idx = ((PyObject *)__pyx_int_1);
+  uintptr_t const *__pyx_v_subset_include_vec;
+  struct plink2::PgrSampleSubsetIndexStruct __pyx_v_subset_index;
+  uint32_t __pyx_v_subset_size;
+  struct plink2::PgenReaderStruct *__pyx_v_pgrp;
   uint32_t *__pyx_v_data_ptr;
+  uintptr_t *__pyx_v_allele_idx_offsets;
+  uintptr_t *__pyx_v_genovec;
+  plink2::AlleleCode *__pyx_v_patch_01_vals;
+  plink2::AlleleCode *__pyx_v_patch_10_vals;
+  uintptr_t __pyx_v_type_ct;
+  uintptr_t __pyx_v_allele_ct;
+  uint32_t __pyx_v_patch_01_ct;
+  uint32_t __pyx_v_patch_10_ct;
+  uintptr_t __pyx_v_ac0;
+  uintptr_t __pyx_v_ac1;
   plink2::PglErr __pyx_v_reterr;
+  PyObject *__pyx_v_i = NULL;
   uint32_t __pyx_v_tmp;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_genocount_uint32_out;
   __Pyx_Buffer __pyx_pybuffer_genocount_uint32_out;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   PyObject *__pyx_t_2 = NULL;
   PyObject *__pyx_t_3 = NULL;
   PyObject *__pyx_t_4 = NULL;
   PyObject *__pyx_t_5 = NULL;
   int __pyx_t_6;
   PyObject *__pyx_t_7 = NULL;
-  int __pyx_t_8;
-  int __pyx_t_9;
-  Py_ssize_t __pyx_t_10;
+  uintptr_t *__pyx_t_8;
+  struct plink2::PgrSampleSubsetIndexStruct __pyx_t_9;
+  uint32_t __pyx_t_10;
+  struct plink2::PgenReaderStruct *__pyx_t_11;
+  Py_ssize_t __pyx_t_12;
+  int __pyx_t_13;
+  int __pyx_t_14;
+  Py_ssize_t __pyx_t_15;
+  PyObject *(*__pyx_t_16)(PyObject *);
+  Py_ssize_t __pyx_t_17;
+  long __pyx_t_18;
+  plink2::AlleleCode *__pyx_t_19;
+  uintptr_t __pyx_t_20;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("count", 0);
   if (__pyx_optional_args) {
     if (__pyx_optional_args->__pyx_n > 0) {
       __pyx_v_allele_idx = __pyx_optional_args->allele_idx;
@@ -21795,31 +22378,31 @@
   __Pyx_INCREF(__pyx_v_allele_idx);
   __pyx_pybuffer_genocount_uint32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_genocount_uint32_out.refcount = 0;
   __pyx_pybuffernd_genocount_uint32_out.data = NULL;
   __pyx_pybuffernd_genocount_uint32_out.rcbuffer = &__pyx_pybuffer_genocount_uint32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_genocount_uint32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1340, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_genocount_uint32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1398, __pyx_L1_error)
   }
   __pyx_pybuffernd_genocount_uint32_out.diminfo[0].strides = __pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_genocount_uint32_out.diminfo[0].shape = __pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_count); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1340, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_count); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1398, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_35count)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1340, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyInt_From_uint32_t(__pyx_v_variant_idx); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1398, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -21829,45 +22412,45 @@
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[4] = {__pyx_t_5, __pyx_t_3, ((PyObject *)__pyx_v_genocount_uint32_out), __pyx_v_allele_idx};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1340, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1398, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
           PyObject *__pyx_temp[4] = {__pyx_t_5, __pyx_t_3, ((PyObject *)__pyx_v_genocount_uint32_out), __pyx_v_allele_idx};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1340, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1398, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1340, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1398, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, __pyx_t_3);
           __Pyx_INCREF(((PyObject *)__pyx_v_genocount_uint32_out));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_genocount_uint32_out));
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, ((PyObject *)__pyx_v_genocount_uint32_out));
           __Pyx_INCREF(__pyx_v_allele_idx);
           __Pyx_GIVEREF(__pyx_v_allele_idx);
           PyTuple_SET_ITEM(__pyx_t_7, 2+__pyx_t_6, __pyx_v_allele_idx);
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1340, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1398, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -21882,190 +22465,805 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1342
+  /* "pgenlib.pyx":1399
+ * 
  *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):
- *         # todo: multiallelic variants
- *         if allele_idx is None:             # <<<<<<<<<<<<<<
- *             allele_idx = 1
- *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
+ *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec             # <<<<<<<<<<<<<<
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
+ *         cdef uint32_t subset_size = self._subset_size
  */
-  __pyx_t_8 = (__pyx_v_allele_idx == Py_None);
-  __pyx_t_9 = (__pyx_t_8 != 0);
-  if (__pyx_t_9) {
+  __pyx_t_8 = __pyx_v_self->_subset_include_vec;
+  __pyx_v_subset_include_vec = __pyx_t_8;
 
-    /* "pgenlib.pyx":1343
- *         # todo: multiallelic variants
- *         if allele_idx is None:
- *             allele_idx = 1             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1400
+ *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):
+ *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index             # <<<<<<<<<<<<<<
+ *         cdef uint32_t subset_size = self._subset_size
+ *         cdef PgenReaderStruct* pgrp = self._state_ptr
+ */
+  __pyx_t_9 = __pyx_v_self->_subset_index;
+  __pyx_v_subset_index = __pyx_t_9;
+
+  /* "pgenlib.pyx":1401
+ *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
+ *         cdef uint32_t subset_size = self._subset_size             # <<<<<<<<<<<<<<
+ *         cdef PgenReaderStruct* pgrp = self._state_ptr
  *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
- *         cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)
  */
-    __Pyx_INCREF(__pyx_int_1);
-    __Pyx_DECREF_SET(__pyx_v_allele_idx, __pyx_int_1);
+  __pyx_t_10 = __pyx_v_self->_subset_size;
+  __pyx_v_subset_size = __pyx_t_10;
 
-    /* "pgenlib.pyx":1342
- *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):
- *         # todo: multiallelic variants
- *         if allele_idx is None:             # <<<<<<<<<<<<<<
- *             allele_idx = 1
+  /* "pgenlib.pyx":1402
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
+ *         cdef uint32_t subset_size = self._subset_size
+ *         cdef PgenReaderStruct* pgrp = self._state_ptr             # <<<<<<<<<<<<<<
  *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
+ *         cdef uintptr_t* allele_idx_offsets
  */
-  }
+  __pyx_t_11 = __pyx_v_self->_state_ptr;
+  __pyx_v_pgrp = __pyx_t_11;
 
-  /* "pgenlib.pyx":1344
- *         if allele_idx is None:
- *             allele_idx = 1
+  /* "pgenlib.pyx":1403
+ *         cdef uint32_t subset_size = self._subset_size
+ *         cdef PgenReaderStruct* pgrp = self._state_ptr
  *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))             # <<<<<<<<<<<<<<
- *         cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)
- *         if reterr != kPglRetSuccess:
+ *         cdef uintptr_t* allele_idx_offsets
+ *         cdef uintptr_t* genovec
  */
-  __pyx_t_10 = 0;
+  __pyx_t_12 = 0;
   __pyx_t_6 = -1;
-  if (__pyx_t_10 < 0) {
-    __pyx_t_10 += __pyx_pybuffernd_genocount_uint32_out.diminfo[0].shape;
-    if (unlikely(__pyx_t_10 < 0)) __pyx_t_6 = 0;
-  } else if (unlikely(__pyx_t_10 >= __pyx_pybuffernd_genocount_uint32_out.diminfo[0].shape)) __pyx_t_6 = 0;
+  if (__pyx_t_12 < 0) {
+    __pyx_t_12 += __pyx_pybuffernd_genocount_uint32_out.diminfo[0].shape;
+    if (unlikely(__pyx_t_12 < 0)) __pyx_t_6 = 0;
+  } else if (unlikely(__pyx_t_12 >= __pyx_pybuffernd_genocount_uint32_out.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 1344, __pyx_L1_error)
+    __PYX_ERR(0, 1403, __pyx_L1_error)
+  }
+  __pyx_v_data_ptr = ((uint32_t *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_genocount_uint32_out.diminfo[0].strides))));
+
+  /* "pgenlib.pyx":1415
+ *         cdef uintptr_t ac1
+ *         cdef PglErr reterr
+ *         if allele_idx is None:             # <<<<<<<<<<<<<<
+ *             allele_ct = 2
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ */
+  __pyx_t_13 = (__pyx_v_allele_idx == Py_None);
+  __pyx_t_14 = (__pyx_t_13 != 0);
+  if (__pyx_t_14) {
+
+    /* "pgenlib.pyx":1416
+ *         cdef PglErr reterr
+ *         if allele_idx is None:
+ *             allele_ct = 2             # <<<<<<<<<<<<<<
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *             if allele_idx_offsets != NULL:
+ */
+    __pyx_v_allele_ct = 2;
+
+    /* "pgenlib.pyx":1417
+ *         if allele_idx is None:
+ *             allele_ct = 2
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets             # <<<<<<<<<<<<<<
+ *             if allele_idx_offsets != NULL:
+ *                 allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]
+ */
+    __pyx_t_8 = (__pyx_v_self->_info_ptr[0]).allele_idx_offsets;
+    __pyx_v_allele_idx_offsets = __pyx_t_8;
+
+    /* "pgenlib.pyx":1418
+ *             allele_ct = 2
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *             if allele_idx_offsets != NULL:             # <<<<<<<<<<<<<<
+ *                 allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]
+ *             if allele_ct > 2:
+ */
+    __pyx_t_14 = ((__pyx_v_allele_idx_offsets != NULL) != 0);
+    if (__pyx_t_14) {
+
+      /* "pgenlib.pyx":1419
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *             if allele_idx_offsets != NULL:
+ *                 allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]             # <<<<<<<<<<<<<<
+ *             if allele_ct > 2:
+ *                 # Multiallelic special case.
+ */
+      __pyx_v_allele_ct = ((__pyx_v_allele_idx_offsets[(__pyx_v_variant_idx + 1)]) - (__pyx_v_allele_idx_offsets[__pyx_v_variant_idx]));
+
+      /* "pgenlib.pyx":1418
+ *             allele_ct = 2
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ *             if allele_idx_offsets != NULL:             # <<<<<<<<<<<<<<
+ *                 allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]
+ *             if allele_ct > 2:
+ */
+    }
+
+    /* "pgenlib.pyx":1420
+ *             if allele_idx_offsets != NULL:
+ *                 allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]
+ *             if allele_ct > 2:             # <<<<<<<<<<<<<<
+ *                 # Multiallelic special case.
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1
+ */
+    __pyx_t_14 = ((__pyx_v_allele_ct > 2) != 0);
+    if (__pyx_t_14) {
+
+      /* "pgenlib.pyx":1422
+ *             if allele_ct > 2:
+ *                 # Multiallelic special case.
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1             # <<<<<<<<<<<<<<
+ *                 if genocount_uint32_out.shape[0] <= type_ct:
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+ */
+      __pyx_v_type_ct = ((__pyx_v_allele_ct * (__pyx_v_allele_ct + 1)) >> 1);
+
+      /* "pgenlib.pyx":1423
+ *                 # Multiallelic special case.
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1
+ *                 if genocount_uint32_out.shape[0] <= type_ct:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ */
+      __pyx_t_14 = (((__pyx_v_genocount_uint32_out->dimensions[0]) <= __pyx_v_type_ct) != 0);
+      if (unlikely(__pyx_t_14)) {
+
+        /* "pgenlib.pyx":1424
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1
+ *                 if genocount_uint32_out.shape[0] <= type_ct:
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
+ */
+        __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__21, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1424, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __PYX_ERR(0, 1424, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1423
+ *                 # Multiallelic special case.
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1
+ *                 if genocount_uint32_out.shape[0] <= type_ct:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ */
+      }
+
+      /* "pgenlib.pyx":1425
+ *                 if genocount_uint32_out.shape[0] <= type_ct:
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)             # <<<<<<<<<<<<<<
+ *                 if reterr != kPglRetSuccess:
+ *                     raise RuntimeError("count() error " + str(reterr))
+ */
+      __pyx_v_reterr = plink2::PgrGetM(__pyx_v_subset_include_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, (&__pyx_v_self->_pgv));
+
+      /* "pgenlib.pyx":1426
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("count() error " + str(reterr))
+ *                 genovec = self._pgv.genovec
+ */
+      __pyx_t_14 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+      if (unlikely(__pyx_t_14)) {
+
+        /* "pgenlib.pyx":1427
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
+ *                     raise RuntimeError("count() error " + str(reterr))             # <<<<<<<<<<<<<<
+ *                 genovec = self._pgv.genovec
+ *                 ZeroTrailingNyps(subset_size, genovec)
+ */
+        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1427, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1427, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_count_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1427, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1427, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __PYX_ERR(0, 1427, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1426
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("count() error " + str(reterr))
+ *                 genovec = self._pgv.genovec
+ */
+      }
+
+      /* "pgenlib.pyx":1428
+ *                 if reterr != kPglRetSuccess:
+ *                     raise RuntimeError("count() error " + str(reterr))
+ *                 genovec = self._pgv.genovec             # <<<<<<<<<<<<<<
+ *                 ZeroTrailingNyps(subset_size, genovec)
+ *                 GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)
+ */
+      __pyx_t_8 = __pyx_v_self->_pgv.genovec;
+      __pyx_v_genovec = __pyx_t_8;
+
+      /* "pgenlib.pyx":1429
+ *                     raise RuntimeError("count() error " + str(reterr))
+ *                 genovec = self._pgv.genovec
+ *                 ZeroTrailingNyps(subset_size, genovec)             # <<<<<<<<<<<<<<
+ *                 GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)
+ *                 data_ptr[type_ct] = data_ptr[3]
+ */
+      plink2::ZeroTrailingNyps(__pyx_v_subset_size, __pyx_v_genovec);
+
+      /* "pgenlib.pyx":1430
+ *                 genovec = self._pgv.genovec
+ *                 ZeroTrailingNyps(subset_size, genovec)
+ *                 GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)             # <<<<<<<<<<<<<<
+ *                 data_ptr[type_ct] = data_ptr[3]
+ *                 for i in range(3, type_ct):
+ */
+      plink2::GenoarrCountFreqsUnsafe(__pyx_v_genovec, __pyx_v_subset_size, __pyx_v_data_ptr);
+
+      /* "pgenlib.pyx":1431
+ *                 ZeroTrailingNyps(subset_size, genovec)
+ *                 GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)
+ *                 data_ptr[type_ct] = data_ptr[3]             # <<<<<<<<<<<<<<
+ *                 for i in range(3, type_ct):
+ *                     data_ptr[i] = 0
+ */
+      (__pyx_v_data_ptr[__pyx_v_type_ct]) = (__pyx_v_data_ptr[3]);
+
+      /* "pgenlib.pyx":1432
+ *                 GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)
+ *                 data_ptr[type_ct] = data_ptr[3]
+ *                 for i in range(3, type_ct):             # <<<<<<<<<<<<<<
+ *                     data_ptr[i] = 0
+ *                 # Now correct for patch_01 and patch_10.
+ */
+      __pyx_t_2 = __Pyx_PyInt_FromSize_t(__pyx_v_type_ct); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1432, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_2);
+      __pyx_t_1 = PyTuple_New(2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1432, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __Pyx_INCREF(__pyx_int_3);
+      __Pyx_GIVEREF(__pyx_int_3);
+      PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_int_3);
+      __Pyx_GIVEREF(__pyx_t_2);
+      PyTuple_SET_ITEM(__pyx_t_1, 1, __pyx_t_2);
+      __pyx_t_2 = 0;
+      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1432, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_2);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
+        __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_15 = 0;
+        __pyx_t_16 = NULL;
+      } else {
+        __pyx_t_15 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1432, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __pyx_t_16 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_16)) __PYX_ERR(0, 1432, __pyx_L1_error)
+      }
+      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+      for (;;) {
+        if (likely(!__pyx_t_16)) {
+          if (likely(PyList_CheckExact(__pyx_t_1))) {
+            if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_1)) break;
+            #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+            __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_2); __pyx_t_15++; if (unlikely(0 < 0)) __PYX_ERR(0, 1432, __pyx_L1_error)
+            #else
+            __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1432, __pyx_L1_error)
+            __Pyx_GOTREF(__pyx_t_2);
+            #endif
+          } else {
+            if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
+            #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+            __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_2); __pyx_t_15++; if (unlikely(0 < 0)) __PYX_ERR(0, 1432, __pyx_L1_error)
+            #else
+            __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1432, __pyx_L1_error)
+            __Pyx_GOTREF(__pyx_t_2);
+            #endif
+          }
+        } else {
+          __pyx_t_2 = __pyx_t_16(__pyx_t_1);
+          if (unlikely(!__pyx_t_2)) {
+            PyObject* exc_type = PyErr_Occurred();
+            if (exc_type) {
+              if (likely(__Pyx_PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
+              else __PYX_ERR(0, 1432, __pyx_L1_error)
+            }
+            break;
+          }
+          __Pyx_GOTREF(__pyx_t_2);
+        }
+        __Pyx_XDECREF_SET(__pyx_v_i, __pyx_t_2);
+        __pyx_t_2 = 0;
+
+        /* "pgenlib.pyx":1433
+ *                 data_ptr[type_ct] = data_ptr[3]
+ *                 for i in range(3, type_ct):
+ *                     data_ptr[i] = 0             # <<<<<<<<<<<<<<
+ *                 # Now correct for patch_01 and patch_10.
+ *                 patch_01_ct = self._pgv.patch_01_ct
+ */
+        __pyx_t_17 = __Pyx_PyIndex_AsSsize_t(__pyx_v_i); if (unlikely((__pyx_t_17 == (Py_ssize_t)-1) && PyErr_Occurred())) __PYX_ERR(0, 1433, __pyx_L1_error)
+        (__pyx_v_data_ptr[__pyx_t_17]) = 0;
+
+        /* "pgenlib.pyx":1432
+ *                 GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)
+ *                 data_ptr[type_ct] = data_ptr[3]
+ *                 for i in range(3, type_ct):             # <<<<<<<<<<<<<<
+ *                     data_ptr[i] = 0
+ *                 # Now correct for patch_01 and patch_10.
+ */
+      }
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+
+      /* "pgenlib.pyx":1435
+ *                     data_ptr[i] = 0
+ *                 # Now correct for patch_01 and patch_10.
+ *                 patch_01_ct = self._pgv.patch_01_ct             # <<<<<<<<<<<<<<
+ *                 if patch_01_ct > 0:
+ *                     data_ptr[1] -= patch_01_ct
+ */
+      __pyx_t_10 = __pyx_v_self->_pgv.patch_01_ct;
+      __pyx_v_patch_01_ct = __pyx_t_10;
+
+      /* "pgenlib.pyx":1436
+ *                 # Now correct for patch_01 and patch_10.
+ *                 patch_01_ct = self._pgv.patch_01_ct
+ *                 if patch_01_ct > 0:             # <<<<<<<<<<<<<<
+ *                     data_ptr[1] -= patch_01_ct
+ *                     patch_01_vals = self._pgv.patch_01_vals
+ */
+      __pyx_t_14 = ((__pyx_v_patch_01_ct > 0) != 0);
+      if (__pyx_t_14) {
+
+        /* "pgenlib.pyx":1437
+ *                 patch_01_ct = self._pgv.patch_01_ct
+ *                 if patch_01_ct > 0:
+ *                     data_ptr[1] -= patch_01_ct             # <<<<<<<<<<<<<<
+ *                     patch_01_vals = self._pgv.patch_01_vals
+ *                     # could special-case allele_ct == 3, etc.
+ */
+        __pyx_t_18 = 1;
+        (__pyx_v_data_ptr[__pyx_t_18]) = ((__pyx_v_data_ptr[__pyx_t_18]) - __pyx_v_patch_01_ct);
+
+        /* "pgenlib.pyx":1438
+ *                 if patch_01_ct > 0:
+ *                     data_ptr[1] -= patch_01_ct
+ *                     patch_01_vals = self._pgv.patch_01_vals             # <<<<<<<<<<<<<<
+ *                     # could special-case allele_ct == 3, etc.
+ *                     for i in range(patch_01_ct):
+ */
+        __pyx_t_19 = __pyx_v_self->_pgv.patch_01_vals;
+        __pyx_v_patch_01_vals = __pyx_t_19;
+
+        /* "pgenlib.pyx":1440
+ *                     patch_01_vals = self._pgv.patch_01_vals
+ *                     # could special-case allele_ct == 3, etc.
+ *                     for i in range(patch_01_ct):             # <<<<<<<<<<<<<<
+ *                         ac1 = patch_01_vals[i]
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+ */
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_patch_01_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1440, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_range, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1440, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
+          __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_15 = 0;
+          __pyx_t_16 = NULL;
+        } else {
+          __pyx_t_15 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1440, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_1);
+          __pyx_t_16 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_16)) __PYX_ERR(0, 1440, __pyx_L1_error)
+        }
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        for (;;) {
+          if (likely(!__pyx_t_16)) {
+            if (likely(PyList_CheckExact(__pyx_t_1))) {
+              if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_1)) break;
+              #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+              __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_2); __pyx_t_15++; if (unlikely(0 < 0)) __PYX_ERR(0, 1440, __pyx_L1_error)
+              #else
+              __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1440, __pyx_L1_error)
+              __Pyx_GOTREF(__pyx_t_2);
+              #endif
+            } else {
+              if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
+              #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+              __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_2); __pyx_t_15++; if (unlikely(0 < 0)) __PYX_ERR(0, 1440, __pyx_L1_error)
+              #else
+              __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1440, __pyx_L1_error)
+              __Pyx_GOTREF(__pyx_t_2);
+              #endif
+            }
+          } else {
+            __pyx_t_2 = __pyx_t_16(__pyx_t_1);
+            if (unlikely(!__pyx_t_2)) {
+              PyObject* exc_type = PyErr_Occurred();
+              if (exc_type) {
+                if (likely(__Pyx_PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
+                else __PYX_ERR(0, 1440, __pyx_L1_error)
+              }
+              break;
+            }
+            __Pyx_GOTREF(__pyx_t_2);
+          }
+          __Pyx_XDECREF_SET(__pyx_v_i, __pyx_t_2);
+          __pyx_t_2 = 0;
+
+          /* "pgenlib.pyx":1441
+ *                     # could special-case allele_ct == 3, etc.
+ *                     for i in range(patch_01_ct):
+ *                         ac1 = patch_01_vals[i]             # <<<<<<<<<<<<<<
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+ *                 patch_10_ct = self._pgv.patch_10_ct
+ */
+          __pyx_t_17 = __Pyx_PyIndex_AsSsize_t(__pyx_v_i); if (unlikely((__pyx_t_17 == (Py_ssize_t)-1) && PyErr_Occurred())) __PYX_ERR(0, 1441, __pyx_L1_error)
+          __pyx_v_ac1 = (__pyx_v_patch_01_vals[__pyx_t_17]);
+
+          /* "pgenlib.pyx":1442
+ *                     for i in range(patch_01_ct):
+ *                         ac1 = patch_01_vals[i]
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1             # <<<<<<<<<<<<<<
+ *                 patch_10_ct = self._pgv.patch_10_ct
+ *                 if patch_10_ct > 0:
+ */
+          __pyx_t_20 = ((__pyx_v_ac1 * (__pyx_v_ac1 + 1)) >> 1);
+          (__pyx_v_data_ptr[__pyx_t_20]) = ((__pyx_v_data_ptr[__pyx_t_20]) + 1);
+
+          /* "pgenlib.pyx":1440
+ *                     patch_01_vals = self._pgv.patch_01_vals
+ *                     # could special-case allele_ct == 3, etc.
+ *                     for i in range(patch_01_ct):             # <<<<<<<<<<<<<<
+ *                         ac1 = patch_01_vals[i]
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+ */
+        }
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+
+        /* "pgenlib.pyx":1436
+ *                 # Now correct for patch_01 and patch_10.
+ *                 patch_01_ct = self._pgv.patch_01_ct
+ *                 if patch_01_ct > 0:             # <<<<<<<<<<<<<<
+ *                     data_ptr[1] -= patch_01_ct
+ *                     patch_01_vals = self._pgv.patch_01_vals
+ */
+      }
+
+      /* "pgenlib.pyx":1443
+ *                         ac1 = patch_01_vals[i]
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+ *                 patch_10_ct = self._pgv.patch_10_ct             # <<<<<<<<<<<<<<
+ *                 if patch_10_ct > 0:
+ *                     data_ptr[2] -= patch_10_ct
+ */
+      __pyx_t_10 = __pyx_v_self->_pgv.patch_10_ct;
+      __pyx_v_patch_10_ct = __pyx_t_10;
+
+      /* "pgenlib.pyx":1444
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+ *                 patch_10_ct = self._pgv.patch_10_ct
+ *                 if patch_10_ct > 0:             # <<<<<<<<<<<<<<
+ *                     data_ptr[2] -= patch_10_ct
+ *                     patch_10_vals = self._pgv.patch_10_vals
+ */
+      __pyx_t_14 = ((__pyx_v_patch_10_ct > 0) != 0);
+      if (__pyx_t_14) {
+
+        /* "pgenlib.pyx":1445
+ *                 patch_10_ct = self._pgv.patch_10_ct
+ *                 if patch_10_ct > 0:
+ *                     data_ptr[2] -= patch_10_ct             # <<<<<<<<<<<<<<
+ *                     patch_10_vals = self._pgv.patch_10_vals
+ *                     for i in range(patch_10_ct):
+ */
+        __pyx_t_18 = 2;
+        (__pyx_v_data_ptr[__pyx_t_18]) = ((__pyx_v_data_ptr[__pyx_t_18]) - __pyx_v_patch_10_ct);
+
+        /* "pgenlib.pyx":1446
+ *                 if patch_10_ct > 0:
+ *                     data_ptr[2] -= patch_10_ct
+ *                     patch_10_vals = self._pgv.patch_10_vals             # <<<<<<<<<<<<<<
+ *                     for i in range(patch_10_ct):
+ *                         ac0 = patch_10_vals[2*i]
+ */
+        __pyx_t_19 = __pyx_v_self->_pgv.patch_10_vals;
+        __pyx_v_patch_10_vals = __pyx_t_19;
+
+        /* "pgenlib.pyx":1447
+ *                     data_ptr[2] -= patch_10_ct
+ *                     patch_10_vals = self._pgv.patch_10_vals
+ *                     for i in range(patch_10_ct):             # <<<<<<<<<<<<<<
+ *                         ac0 = patch_10_vals[2*i]
+ *                         ac1 = patch_10_vals[2*i+1]
+ */
+        __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_patch_10_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1447, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_range, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1447, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
+          __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_15 = 0;
+          __pyx_t_16 = NULL;
+        } else {
+          __pyx_t_15 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1447, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_1);
+          __pyx_t_16 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_16)) __PYX_ERR(0, 1447, __pyx_L1_error)
+        }
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        for (;;) {
+          if (likely(!__pyx_t_16)) {
+            if (likely(PyList_CheckExact(__pyx_t_1))) {
+              if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_1)) break;
+              #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+              __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_2); __pyx_t_15++; if (unlikely(0 < 0)) __PYX_ERR(0, 1447, __pyx_L1_error)
+              #else
+              __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1447, __pyx_L1_error)
+              __Pyx_GOTREF(__pyx_t_2);
+              #endif
+            } else {
+              if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
+              #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+              __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_2); __pyx_t_15++; if (unlikely(0 < 0)) __PYX_ERR(0, 1447, __pyx_L1_error)
+              #else
+              __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1447, __pyx_L1_error)
+              __Pyx_GOTREF(__pyx_t_2);
+              #endif
+            }
+          } else {
+            __pyx_t_2 = __pyx_t_16(__pyx_t_1);
+            if (unlikely(!__pyx_t_2)) {
+              PyObject* exc_type = PyErr_Occurred();
+              if (exc_type) {
+                if (likely(__Pyx_PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
+                else __PYX_ERR(0, 1447, __pyx_L1_error)
+              }
+              break;
+            }
+            __Pyx_GOTREF(__pyx_t_2);
+          }
+          __Pyx_XDECREF_SET(__pyx_v_i, __pyx_t_2);
+          __pyx_t_2 = 0;
+
+          /* "pgenlib.pyx":1448
+ *                     patch_10_vals = self._pgv.patch_10_vals
+ *                     for i in range(patch_10_ct):
+ *                         ac0 = patch_10_vals[2*i]             # <<<<<<<<<<<<<<
+ *                         ac1 = patch_10_vals[2*i+1]
+ *                         data_ptr[ac0 + ((ac1 * (ac1 + 1)) >> 1)] += 1
+ */
+          __pyx_t_2 = PyNumber_Multiply(__pyx_int_2, __pyx_v_i); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1448, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_2);
+          __pyx_t_17 = __Pyx_PyIndex_AsSsize_t(__pyx_t_2); if (unlikely((__pyx_t_17 == (Py_ssize_t)-1) && PyErr_Occurred())) __PYX_ERR(0, 1448, __pyx_L1_error)
+          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+          __pyx_v_ac0 = (__pyx_v_patch_10_vals[__pyx_t_17]);
+
+          /* "pgenlib.pyx":1449
+ *                     for i in range(patch_10_ct):
+ *                         ac0 = patch_10_vals[2*i]
+ *                         ac1 = patch_10_vals[2*i+1]             # <<<<<<<<<<<<<<
+ *                         data_ptr[ac0 + ((ac1 * (ac1 + 1)) >> 1)] += 1
+ *                 return
+ */
+          __pyx_t_2 = PyNumber_Multiply(__pyx_int_2, __pyx_v_i); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1449, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_2);
+          __pyx_t_4 = __Pyx_PyInt_AddObjC(__pyx_t_2, __pyx_int_1, 1, 0, 0); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1449, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_4);
+          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+          __pyx_t_17 = __Pyx_PyIndex_AsSsize_t(__pyx_t_4); if (unlikely((__pyx_t_17 == (Py_ssize_t)-1) && PyErr_Occurred())) __PYX_ERR(0, 1449, __pyx_L1_error)
+          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+          __pyx_v_ac1 = (__pyx_v_patch_10_vals[__pyx_t_17]);
+
+          /* "pgenlib.pyx":1450
+ *                         ac0 = patch_10_vals[2*i]
+ *                         ac1 = patch_10_vals[2*i+1]
+ *                         data_ptr[ac0 + ((ac1 * (ac1 + 1)) >> 1)] += 1             # <<<<<<<<<<<<<<
+ *                 return
+ *             allele_idx = 1
+ */
+          __pyx_t_20 = (__pyx_v_ac0 + ((__pyx_v_ac1 * (__pyx_v_ac1 + 1)) >> 1));
+          (__pyx_v_data_ptr[__pyx_t_20]) = ((__pyx_v_data_ptr[__pyx_t_20]) + 1);
+
+          /* "pgenlib.pyx":1447
+ *                     data_ptr[2] -= patch_10_ct
+ *                     patch_10_vals = self._pgv.patch_10_vals
+ *                     for i in range(patch_10_ct):             # <<<<<<<<<<<<<<
+ *                         ac0 = patch_10_vals[2*i]
+ *                         ac1 = patch_10_vals[2*i+1]
+ */
+        }
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+
+        /* "pgenlib.pyx":1444
+ *                         data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+ *                 patch_10_ct = self._pgv.patch_10_ct
+ *                 if patch_10_ct > 0:             # <<<<<<<<<<<<<<
+ *                     data_ptr[2] -= patch_10_ct
+ *                     patch_10_vals = self._pgv.patch_10_vals
+ */
+      }
+
+      /* "pgenlib.pyx":1451
+ *                         ac1 = patch_10_vals[2*i+1]
+ *                         data_ptr[ac0 + ((ac1 * (ac1 + 1)) >> 1)] += 1
+ *                 return             # <<<<<<<<<<<<<<
+ *             allele_idx = 1
+ *         reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)
+ */
+      __Pyx_XDECREF(__pyx_r);
+      __pyx_r = Py_None; __Pyx_INCREF(Py_None);
+      goto __pyx_L0;
+
+      /* "pgenlib.pyx":1420
+ *             if allele_idx_offsets != NULL:
+ *                 allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]
+ *             if allele_ct > 2:             # <<<<<<<<<<<<<<
+ *                 # Multiallelic special case.
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1
+ */
+    }
+
+    /* "pgenlib.pyx":1452
+ *                         data_ptr[ac0 + ((ac1 * (ac1 + 1)) >> 1)] += 1
+ *                 return
+ *             allele_idx = 1             # <<<<<<<<<<<<<<
+ *         reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)
+ *         if reterr != kPglRetSuccess:
+ */
+    __Pyx_INCREF(__pyx_int_1);
+    __Pyx_DECREF_SET(__pyx_v_allele_idx, __pyx_int_1);
+
+    /* "pgenlib.pyx":1415
+ *         cdef uintptr_t ac1
+ *         cdef PglErr reterr
+ *         if allele_idx is None:             # <<<<<<<<<<<<<<
+ *             allele_ct = 2
+ *             allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+ */
   }
-  __pyx_v_data_ptr = ((uint32_t *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_uint32_t *, __pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer.buf, __pyx_t_10, __pyx_pybuffernd_genocount_uint32_out.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":1345
+  /* "pgenlib.pyx":1453
+ *                 return
  *             allele_idx = 1
- *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
- *         cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)             # <<<<<<<<<<<<<<
+ *         reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("count() error " + str(reterr))
  */
-  __pyx_v_reterr = plink2::PgrGetCounts(__pyx_v_self->_subset_include_vec, __pyx_v_self->_subset_include_interleaved_vec, __pyx_v_self->_subset_index, __pyx_v_self->_subset_size, __pyx_v_variant_idx, __pyx_v_self->_state_ptr, __pyx_v_data_ptr);
+  __pyx_v_reterr = plink2::PgrGetCounts(__pyx_v_subset_include_vec, __pyx_v_self->_subset_include_interleaved_vec, __pyx_v_subset_index, __pyx_v_subset_size, __pyx_v_variant_idx, __pyx_v_pgrp, __pyx_v_data_ptr);
 
-  /* "pgenlib.pyx":1346
- *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
- *         cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)
+  /* "pgenlib.pyx":1454
+ *             allele_idx = 1
+ *         reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("count() error " + str(reterr))
  *         if allele_idx != 0:
  */
-  __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-  if (unlikely(__pyx_t_9)) {
+  __pyx_t_14 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+  if (unlikely(__pyx_t_14)) {
 
-    /* "pgenlib.pyx":1347
- *         cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)
+    /* "pgenlib.pyx":1455
+ *         reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("count() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         if allele_idx != 0:
  *             return
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1347, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1455, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1347, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_2);
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1455, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_count_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1347, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_count_error, __pyx_t_4); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1455, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1347, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_2);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1455, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __Pyx_Raise(__pyx_t_2, 0, 0, 0);
-    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 1347, __pyx_L1_error)
+    __Pyx_Raise(__pyx_t_4, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __PYX_ERR(0, 1455, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1346
- *         cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
- *         cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)
+    /* "pgenlib.pyx":1454
+ *             allele_idx = 1
+ *         reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("count() error " + str(reterr))
  *         if allele_idx != 0:
  */
   }
 
-  /* "pgenlib.pyx":1348
+  /* "pgenlib.pyx":1456
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("count() error " + str(reterr))
  *         if allele_idx != 0:             # <<<<<<<<<<<<<<
  *             return
  *         cdef uint32_t tmp = data_ptr[0]
  */
-  __pyx_t_2 = __Pyx_PyInt_NeObjC(__pyx_v_allele_idx, __pyx_int_0, 0, 0); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1348, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) __PYX_ERR(0, 1348, __pyx_L1_error)
-  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  if (__pyx_t_9) {
+  __pyx_t_4 = __Pyx_PyInt_NeObjC(__pyx_v_allele_idx, __pyx_int_0, 0, 0); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1456, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_t_4);
+  __pyx_t_14 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_14 < 0)) __PYX_ERR(0, 1456, __pyx_L1_error)
+  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+  if (__pyx_t_14) {
 
-    /* "pgenlib.pyx":1349
+    /* "pgenlib.pyx":1457
  *             raise RuntimeError("count() error " + str(reterr))
  *         if allele_idx != 0:
  *             return             # <<<<<<<<<<<<<<
  *         cdef uint32_t tmp = data_ptr[0]
  *         data_ptr[0] = data_ptr[2]
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "pgenlib.pyx":1348
+    /* "pgenlib.pyx":1456
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("count() error " + str(reterr))
  *         if allele_idx != 0:             # <<<<<<<<<<<<<<
  *             return
  *         cdef uint32_t tmp = data_ptr[0]
  */
   }
 
-  /* "pgenlib.pyx":1350
+  /* "pgenlib.pyx":1458
  *         if allele_idx != 0:
  *             return
  *         cdef uint32_t tmp = data_ptr[0]             # <<<<<<<<<<<<<<
  *         data_ptr[0] = data_ptr[2]
  *         data_ptr[2] = tmp
  */
   __pyx_v_tmp = (__pyx_v_data_ptr[0]);
 
-  /* "pgenlib.pyx":1351
+  /* "pgenlib.pyx":1459
  *             return
  *         cdef uint32_t tmp = data_ptr[0]
  *         data_ptr[0] = data_ptr[2]             # <<<<<<<<<<<<<<
  *         data_ptr[2] = tmp
  *         return
  */
   (__pyx_v_data_ptr[0]) = (__pyx_v_data_ptr[2]);
 
-  /* "pgenlib.pyx":1352
+  /* "pgenlib.pyx":1460
  *         cdef uint32_t tmp = data_ptr[0]
  *         data_ptr[0] = data_ptr[2]
  *         data_ptr[2] = tmp             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   (__pyx_v_data_ptr[2]) = __pyx_v_tmp;
 
-  /* "pgenlib.pyx":1353
+  /* "pgenlib.pyx":1461
  *         data_ptr[0] = data_ptr[2]
  *         data_ptr[2] = tmp
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1340
+  /* "pgenlib.pyx":1398
  *         return
  * 
  *     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):             # <<<<<<<<<<<<<<
- *         # todo: multiallelic variants
- *         if allele_idx is None:
+ *         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
+ *         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
   __Pyx_XDECREF(__pyx_t_2);
   __Pyx_XDECREF(__pyx_t_3);
@@ -22080,14 +23278,15 @@
   __Pyx_ErrRestore(__pyx_type, __pyx_value, __pyx_tb);}
   __Pyx_AddTraceback("pgenlib.PgenReader.count", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = 0;
   goto __pyx_L2;
   __pyx_L0:;
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer);
   __pyx_L2:;
+  __Pyx_XDECREF(__pyx_v_i);
   __Pyx_XDECREF(__pyx_v_allele_idx);
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
 /* Python wrapper */
@@ -22124,49 +23323,49 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_idx)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_genocount_uint32_out)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("count", 0, 2, 3, 1); __PYX_ERR(0, 1340, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("count", 0, 2, 3, 1); __PYX_ERR(0, 1398, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx);
           if (value) { values[2] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "count") < 0)) __PYX_ERR(0, 1340, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "count") < 0)) __PYX_ERR(0, 1398, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
         CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
-    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1340, __pyx_L3_error)
+    __pyx_v_variant_idx = __Pyx_PyInt_As_uint32_t(values[0]); if (unlikely((__pyx_v_variant_idx == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1398, __pyx_L3_error)
     __pyx_v_genocount_uint32_out = ((PyArrayObject *)values[1]);
     __pyx_v_allele_idx = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("count", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1340, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("count", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1398, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.count", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_genocount_uint32_out), __pyx_ptype_5numpy_ndarray, 1, "genocount_uint32_out", 0))) __PYX_ERR(0, 1340, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_genocount_uint32_out), __pyx_ptype_5numpy_ndarray, 1, "genocount_uint32_out", 0))) __PYX_ERR(0, 1398, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_34count(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_variant_idx, __pyx_v_genocount_uint32_out, __pyx_v_allele_idx);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -22187,21 +23386,21 @@
   __Pyx_RefNannySetupContext("count", 0);
   __pyx_pybuffer_genocount_uint32_out.pybuffer.buf = NULL;
   __pyx_pybuffer_genocount_uint32_out.refcount = 0;
   __pyx_pybuffernd_genocount_uint32_out.data = NULL;
   __pyx_pybuffernd_genocount_uint32_out.rcbuffer = &__pyx_pybuffer_genocount_uint32_out;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_genocount_uint32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1340, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer, (PyObject*)__pyx_v_genocount_uint32_out, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1398, __pyx_L1_error)
   }
   __pyx_pybuffernd_genocount_uint32_out.diminfo[0].strides = __pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_genocount_uint32_out.diminfo[0].shape = __pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.allele_idx = __pyx_v_allele_idx;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->count(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_genocount_uint32_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1340, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->count(__pyx_v_self, __pyx_v_variant_idx, __pyx_v_genocount_uint32_out, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1398, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -22219,15 +23418,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_genocount_uint32_out.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1356
+/* "pgenlib.pyx":1464
  * 
  * 
  *     cpdef change_sample_subset(self, object sample_subset = None):             # <<<<<<<<<<<<<<
  *         if sample_subset is not None:
  *             self.set_sample_subset_internal(sample_subset)
  */
 
@@ -22257,15 +23456,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_change_sample_subset); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1356, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_change_sample_subset); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1464, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_37change_sample_subset)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -22274,15 +23473,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_Call2Args(__pyx_t_3, __pyx_t_4, __pyx_v_sample_subset) : __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_sample_subset);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1356, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1464, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -22295,81 +23494,81 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1357
+  /* "pgenlib.pyx":1465
  * 
  *     cpdef change_sample_subset(self, object sample_subset = None):
  *         if sample_subset is not None:             # <<<<<<<<<<<<<<
  *             self.set_sample_subset_internal(sample_subset)
  *         else:
  */
   __pyx_t_5 = (__pyx_v_sample_subset != Py_None);
   __pyx_t_6 = (__pyx_t_5 != 0);
   if (__pyx_t_6) {
 
-    /* "pgenlib.pyx":1358
+    /* "pgenlib.pyx":1466
  *     cpdef change_sample_subset(self, object sample_subset = None):
  *         if sample_subset is not None:
  *             self.set_sample_subset_internal(sample_subset)             # <<<<<<<<<<<<<<
  *         else:
  *             PgrClearSampleSubsetIndex(self._state_ptr, &(self._subset_index))
  */
-    if (!(likely(((__pyx_v_sample_subset) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_sample_subset, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 1358, __pyx_L1_error)
-    __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->set_sample_subset_internal(__pyx_v_self, ((PyArrayObject *)__pyx_v_sample_subset)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1358, __pyx_L1_error)
+    if (!(likely(((__pyx_v_sample_subset) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_sample_subset, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 1466, __pyx_L1_error)
+    __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->set_sample_subset_internal(__pyx_v_self, ((PyArrayObject *)__pyx_v_sample_subset)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1466, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-    /* "pgenlib.pyx":1357
+    /* "pgenlib.pyx":1465
  * 
  *     cpdef change_sample_subset(self, object sample_subset = None):
  *         if sample_subset is not None:             # <<<<<<<<<<<<<<
  *             self.set_sample_subset_internal(sample_subset)
  *         else:
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":1360
+  /* "pgenlib.pyx":1468
  *             self.set_sample_subset_internal(sample_subset)
  *         else:
  *             PgrClearSampleSubsetIndex(self._state_ptr, &(self._subset_index))             # <<<<<<<<<<<<<<
  *             self._subset_size = self._info_ptr[0].raw_sample_ct
  *         return
  */
   /*else*/ {
     plink2::PgrClearSampleSubsetIndex(__pyx_v_self->_state_ptr, (&__pyx_v_self->_subset_index));
 
-    /* "pgenlib.pyx":1361
+    /* "pgenlib.pyx":1469
  *         else:
  *             PgrClearSampleSubsetIndex(self._state_ptr, &(self._subset_index))
  *             self._subset_size = self._info_ptr[0].raw_sample_ct             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_t_7 = (__pyx_v_self->_info_ptr[0]).raw_sample_ct;
     __pyx_v_self->_subset_size = __pyx_t_7;
   }
   __pyx_L3:;
 
-  /* "pgenlib.pyx":1362
+  /* "pgenlib.pyx":1470
  *             PgrClearSampleSubsetIndex(self._state_ptr, &(self._subset_index))
  *             self._subset_size = self._info_ptr[0].raw_sample_ct
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1356
+  /* "pgenlib.pyx":1464
  * 
  * 
  *     cpdef change_sample_subset(self, object sample_subset = None):             # <<<<<<<<<<<<<<
  *         if sample_subset is not None:
  *             self.set_sample_subset_internal(sample_subset)
  */
 
@@ -22415,29 +23614,29 @@
         case  0:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_subset);
           if (value) { values[0] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "change_sample_subset") < 0)) __PYX_ERR(0, 1356, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "change_sample_subset") < 0)) __PYX_ERR(0, 1464, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         CYTHON_FALLTHROUGH;
         case  0: break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_sample_subset = values[0];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("change_sample_subset", 0, 0, 1, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1356, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("change_sample_subset", 0, 0, 1, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1464, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.change_sample_subset", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_36change_sample_subset(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_sample_subset);
 
@@ -22454,15 +23653,15 @@
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("change_sample_subset", 0);
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_2.__pyx_n = 1;
   __pyx_t_2.sample_subset = __pyx_v_sample_subset;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->change_sample_subset(__pyx_v_self, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1356, __pyx_L1_error)
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenReader->change_sample_subset(__pyx_v_self, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1464, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -22471,15 +23670,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1365
+/* "pgenlib.pyx":1473
  * 
  * 
  *     cpdef close(self):             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:
  */
 
@@ -22502,15 +23701,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_close); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1365, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_close); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1473, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_39close)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -22519,15 +23718,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1365, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1473, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -22540,252 +23739,280 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1366
+  /* "pgenlib.pyx":1474
  * 
  *     cpdef close(self):
  *         cdef PglErr reterr = kPglRetSuccess             # <<<<<<<<<<<<<<
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)
  */
   __pyx_v_reterr = plink2::kPglRetSuccess;
 
-  /* "pgenlib.pyx":1367
+  /* "pgenlib.pyx":1475
  *     cpdef close(self):
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:             # <<<<<<<<<<<<<<
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:
  */
   __pyx_t_5 = (__pyx_v_self->_info_ptr != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":1368
+    /* "pgenlib.pyx":1476
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)             # <<<<<<<<<<<<<<
  *             if self._info_ptr[0].vrtypes:
  *                 aligned_free(self._info_ptr[0].vrtypes)
  */
     (void)(plink2::CleanupPgfi(__pyx_v_self->_info_ptr, (&__pyx_v_reterr)));
 
-    /* "pgenlib.pyx":1369
+    /* "pgenlib.pyx":1477
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].vrtypes)
- *             if self._info_ptr[0].nonref_flags:
+ *             if self._info_ptr[0].allele_idx_offsets:
  */
     __pyx_t_5 = ((__pyx_v_self->_info_ptr[0]).vrtypes != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":1370
+      /* "pgenlib.pyx":1478
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:
  *                 aligned_free(self._info_ptr[0].vrtypes)             # <<<<<<<<<<<<<<
- *             if self._info_ptr[0].nonref_flags:
- *                 aligned_free(self._info_ptr[0].nonref_flags)
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  */
       plink2::aligned_free((__pyx_v_self->_info_ptr[0]).vrtypes);
 
-      /* "pgenlib.pyx":1369
+      /* "pgenlib.pyx":1477
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].vrtypes)
- *             if self._info_ptr[0].nonref_flags:
+ *             if self._info_ptr[0].allele_idx_offsets:
  */
     }
 
-    /* "pgenlib.pyx":1371
+    /* "pgenlib.pyx":1479
  *             if self._info_ptr[0].vrtypes:
  *                 aligned_free(self._info_ptr[0].vrtypes)
+ *             if self._info_ptr[0].allele_idx_offsets:             # <<<<<<<<<<<<<<
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
+ *             if self._info_ptr[0].nonref_flags:
+ */
+    __pyx_t_5 = ((__pyx_v_self->_info_ptr[0]).allele_idx_offsets != 0);
+    if (__pyx_t_5) {
+
+      /* "pgenlib.pyx":1480
+ *                 aligned_free(self._info_ptr[0].vrtypes)
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)             # <<<<<<<<<<<<<<
+ *             if self._info_ptr[0].nonref_flags:
+ *                 aligned_free(self._info_ptr[0].nonref_flags)
+ */
+      plink2::aligned_free((__pyx_v_self->_info_ptr[0]).allele_idx_offsets);
+
+      /* "pgenlib.pyx":1479
+ *             if self._info_ptr[0].vrtypes:
+ *                 aligned_free(self._info_ptr[0].vrtypes)
+ *             if self._info_ptr[0].allele_idx_offsets:             # <<<<<<<<<<<<<<
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
+ *             if self._info_ptr[0].nonref_flags:
+ */
+    }
+
+    /* "pgenlib.pyx":1481
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  *             if self._info_ptr[0].nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:
  */
     __pyx_t_5 = ((__pyx_v_self->_info_ptr[0]).nonref_flags != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":1372
- *                 aligned_free(self._info_ptr[0].vrtypes)
+      /* "pgenlib.pyx":1482
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  *             if self._info_ptr[0].nonref_flags:
  *                 aligned_free(self._info_ptr[0].nonref_flags)             # <<<<<<<<<<<<<<
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)
  */
       plink2::aligned_free((__pyx_v_self->_info_ptr[0]).nonref_flags);
 
-      /* "pgenlib.pyx":1371
- *             if self._info_ptr[0].vrtypes:
- *                 aligned_free(self._info_ptr[0].vrtypes)
+      /* "pgenlib.pyx":1481
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  *             if self._info_ptr[0].nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:
  */
     }
 
-    /* "pgenlib.pyx":1373
+    /* "pgenlib.pyx":1483
  *             if self._info_ptr[0].nonref_flags:
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:             # <<<<<<<<<<<<<<
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):
  */
     __pyx_t_5 = (__pyx_v_self->_state_ptr != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":1374
+      /* "pgenlib.pyx":1484
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)             # <<<<<<<<<<<<<<
  *                 if PgrGetFreadBuf(self._state_ptr):
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  */
       (void)(plink2::CleanupPgr(__pyx_v_self->_state_ptr, (&__pyx_v_reterr)));
 
-      /* "pgenlib.pyx":1375
+      /* "pgenlib.pyx":1485
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):             # <<<<<<<<<<<<<<
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)
  */
       __pyx_t_5 = (plink2::PgrGetFreadBuf(__pyx_v_self->_state_ptr) != 0);
       if (__pyx_t_5) {
 
-        /* "pgenlib.pyx":1376
+        /* "pgenlib.pyx":1486
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))             # <<<<<<<<<<<<<<
  *                 PyMem_Free(self._state_ptr)
  *                 self._state_ptr = NULL
  */
         plink2::aligned_free(plink2::PgrGetFreadBuf(__pyx_v_self->_state_ptr));
 
-        /* "pgenlib.pyx":1375
+        /* "pgenlib.pyx":1485
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):             # <<<<<<<<<<<<<<
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)
  */
       }
 
-      /* "pgenlib.pyx":1377
+      /* "pgenlib.pyx":1487
  *                 if PgrGetFreadBuf(self._state_ptr):
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)             # <<<<<<<<<<<<<<
  *                 self._state_ptr = NULL
  *             PyMem_Free(self._info_ptr)
  */
       PyMem_Free(__pyx_v_self->_state_ptr);
 
-      /* "pgenlib.pyx":1378
+      /* "pgenlib.pyx":1488
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)
  *                 self._state_ptr = NULL             # <<<<<<<<<<<<<<
  *             PyMem_Free(self._info_ptr)
  *             self._info_ptr = NULL
  */
       __pyx_v_self->_state_ptr = NULL;
 
-      /* "pgenlib.pyx":1373
+      /* "pgenlib.pyx":1483
  *             if self._info_ptr[0].nonref_flags:
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:             # <<<<<<<<<<<<<<
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):
  */
     }
 
-    /* "pgenlib.pyx":1379
+    /* "pgenlib.pyx":1489
  *                 PyMem_Free(self._state_ptr)
  *                 self._state_ptr = NULL
  *             PyMem_Free(self._info_ptr)             # <<<<<<<<<<<<<<
  *             self._info_ptr = NULL
  *             if reterr != kPglRetSuccess:
  */
     PyMem_Free(__pyx_v_self->_info_ptr);
 
-    /* "pgenlib.pyx":1380
+    /* "pgenlib.pyx":1490
  *                 self._state_ptr = NULL
  *             PyMem_Free(self._info_ptr)
  *             self._info_ptr = NULL             # <<<<<<<<<<<<<<
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("close() error " + str(reterr))
  */
     __pyx_v_self->_info_ptr = NULL;
 
-    /* "pgenlib.pyx":1381
+    /* "pgenlib.pyx":1491
  *             PyMem_Free(self._info_ptr)
  *             self._info_ptr = NULL
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("close() error " + str(reterr))
  *         return
  */
     __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":1382
+      /* "pgenlib.pyx":1492
  *             self._info_ptr = NULL
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("close() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1382, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1492, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1382, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1492, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_close_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1382, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_close_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1492, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1382, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1492, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       __Pyx_Raise(__pyx_t_2, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __PYX_ERR(0, 1382, __pyx_L1_error)
+      __PYX_ERR(0, 1492, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1381
+      /* "pgenlib.pyx":1491
  *             PyMem_Free(self._info_ptr)
  *             self._info_ptr = NULL
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("close() error " + str(reterr))
  *         return
  */
     }
 
-    /* "pgenlib.pyx":1367
+    /* "pgenlib.pyx":1475
  *     cpdef close(self):
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:             # <<<<<<<<<<<<<<
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:
  */
   }
 
-  /* "pgenlib.pyx":1383
+  /* "pgenlib.pyx":1493
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("close() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1365
+  /* "pgenlib.pyx":1473
  * 
  * 
  *     cpdef close(self):             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:
  */
 
@@ -22821,15 +24048,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("close", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_close(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1365, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader_close(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1473, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -22838,15 +24065,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1386
+/* "pgenlib.pyx":1496
  * 
  * 
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):             # <<<<<<<<<<<<<<
  *         self.close()
  *         return
  */
 
@@ -22869,15 +24096,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_exit); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1386, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_exit); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1496, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenReader_41__exit__)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         __pyx_t_5 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
@@ -22889,43 +24116,43 @@
             __Pyx_DECREF_SET(__pyx_t_3, function);
             __pyx_t_5 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_3)) {
           PyObject *__pyx_temp[4] = {__pyx_t_4, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1386, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1496, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_3)) {
           PyObject *__pyx_temp[4] = {__pyx_t_4, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1386, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1496, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         {
-          __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1386, __pyx_L1_error)
+          __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1496, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_6);
           if (__pyx_t_4) {
             __Pyx_GIVEREF(__pyx_t_4); PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __pyx_t_4 = NULL;
           }
           __Pyx_INCREF(__pyx_v_exc_type);
           __Pyx_GIVEREF(__pyx_v_exc_type);
           PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, __pyx_v_exc_type);
           __Pyx_INCREF(__pyx_v_exc_val);
           __Pyx_GIVEREF(__pyx_v_exc_val);
           PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_exc_val);
           __Pyx_INCREF(__pyx_v_exc_tb);
           __Pyx_GIVEREF(__pyx_v_exc_tb);
           PyTuple_SET_ITEM(__pyx_t_6, 2+__pyx_t_5, __pyx_v_exc_tb);
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1386, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1496, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         }
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -22940,37 +24167,37 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1387
+  /* "pgenlib.pyx":1497
  * 
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):
  *         self.close()             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->close(__pyx_v_self, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1387, __pyx_L1_error)
+  __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenReader *)__pyx_v_self->__pyx_vtab)->close(__pyx_v_self, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1497, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-  /* "pgenlib.pyx":1388
+  /* "pgenlib.pyx":1498
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):
  *         self.close()
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1386
+  /* "pgenlib.pyx":1496
  * 
  * 
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):             # <<<<<<<<<<<<<<
  *         self.close()
  *         return
  */
 
@@ -23022,40 +24249,40 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_exc_type)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_exc_val)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 1); __PYX_ERR(0, 1386, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 1); __PYX_ERR(0, 1496, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_exc_tb)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 2); __PYX_ERR(0, 1386, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 2); __PYX_ERR(0, 1496, __pyx_L3_error)
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__exit__") < 0)) __PYX_ERR(0, 1386, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__exit__") < 0)) __PYX_ERR(0, 1496, __pyx_L3_error)
       }
     } else if (PyTuple_GET_SIZE(__pyx_args) != 3) {
       goto __pyx_L5_argtuple_error;
     } else {
       values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
       values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
       values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
     }
     __pyx_v_exc_type = values[0];
     __pyx_v_exc_val = values[1];
     __pyx_v_exc_tb = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1386, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1496, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenReader.__exit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
   __pyx_r = __pyx_pf_7pgenlib_10PgenReader_40__exit__(((struct __pyx_obj_7pgenlib_PgenReader *)__pyx_v_self), __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb);
 
@@ -23069,15 +24296,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__exit__", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader___exit__(__pyx_v_self, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1386, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenReader___exit__(__pyx_v_self, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1496, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -23086,15 +24313,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1391
+/* "pgenlib.pyx":1501
  * 
  * 
  *     def __dealloc__(self):             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:
  */
 
@@ -23111,191 +24338,219 @@
 
 static void __pyx_pf_7pgenlib_10PgenReader_42__dealloc__(struct __pyx_obj_7pgenlib_PgenReader *__pyx_v_self) {
   plink2::PglErr __pyx_v_reterr;
   __Pyx_RefNannyDeclarations
   int __pyx_t_1;
   __Pyx_RefNannySetupContext("__dealloc__", 0);
 
-  /* "pgenlib.pyx":1392
+  /* "pgenlib.pyx":1502
  * 
  *     def __dealloc__(self):
  *         cdef PglErr reterr = kPglRetSuccess             # <<<<<<<<<<<<<<
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)
  */
   __pyx_v_reterr = plink2::kPglRetSuccess;
 
-  /* "pgenlib.pyx":1393
+  /* "pgenlib.pyx":1503
  *     def __dealloc__(self):
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:             # <<<<<<<<<<<<<<
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:
  */
   __pyx_t_1 = (__pyx_v_self->_info_ptr != 0);
   if (__pyx_t_1) {
 
-    /* "pgenlib.pyx":1394
+    /* "pgenlib.pyx":1504
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)             # <<<<<<<<<<<<<<
  *             if self._info_ptr[0].vrtypes:
  *                 aligned_free(self._info_ptr[0].vrtypes)
  */
     (void)(plink2::CleanupPgfi(__pyx_v_self->_info_ptr, (&__pyx_v_reterr)));
 
-    /* "pgenlib.pyx":1395
+    /* "pgenlib.pyx":1505
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].vrtypes)
- *             if self._info_ptr[0].nonref_flags:
+ *             if self._info_ptr[0].allele_idx_offsets:
  */
     __pyx_t_1 = ((__pyx_v_self->_info_ptr[0]).vrtypes != 0);
     if (__pyx_t_1) {
 
-      /* "pgenlib.pyx":1396
+      /* "pgenlib.pyx":1506
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:
  *                 aligned_free(self._info_ptr[0].vrtypes)             # <<<<<<<<<<<<<<
- *             if self._info_ptr[0].nonref_flags:
- *                 aligned_free(self._info_ptr[0].nonref_flags)
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  */
       plink2::aligned_free((__pyx_v_self->_info_ptr[0]).vrtypes);
 
-      /* "pgenlib.pyx":1395
+      /* "pgenlib.pyx":1505
  *         if self._info_ptr:
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].vrtypes)
- *             if self._info_ptr[0].nonref_flags:
+ *             if self._info_ptr[0].allele_idx_offsets:
  */
     }
 
-    /* "pgenlib.pyx":1397
+    /* "pgenlib.pyx":1507
+ *             if self._info_ptr[0].vrtypes:
+ *                 aligned_free(self._info_ptr[0].vrtypes)
+ *             if self._info_ptr[0].allele_idx_offsets:             # <<<<<<<<<<<<<<
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
+ *             if self._info_ptr[0].nonref_flags:
+ */
+    __pyx_t_1 = ((__pyx_v_self->_info_ptr[0]).allele_idx_offsets != 0);
+    if (__pyx_t_1) {
+
+      /* "pgenlib.pyx":1508
+ *                 aligned_free(self._info_ptr[0].vrtypes)
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)             # <<<<<<<<<<<<<<
+ *             if self._info_ptr[0].nonref_flags:
+ *                 aligned_free(self._info_ptr[0].nonref_flags)
+ */
+      plink2::aligned_free((__pyx_v_self->_info_ptr[0]).allele_idx_offsets);
+
+      /* "pgenlib.pyx":1507
  *             if self._info_ptr[0].vrtypes:
  *                 aligned_free(self._info_ptr[0].vrtypes)
+ *             if self._info_ptr[0].allele_idx_offsets:             # <<<<<<<<<<<<<<
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
+ *             if self._info_ptr[0].nonref_flags:
+ */
+    }
+
+    /* "pgenlib.pyx":1509
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  *             if self._info_ptr[0].nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:
  */
     __pyx_t_1 = ((__pyx_v_self->_info_ptr[0]).nonref_flags != 0);
     if (__pyx_t_1) {
 
-      /* "pgenlib.pyx":1398
- *                 aligned_free(self._info_ptr[0].vrtypes)
+      /* "pgenlib.pyx":1510
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  *             if self._info_ptr[0].nonref_flags:
  *                 aligned_free(self._info_ptr[0].nonref_flags)             # <<<<<<<<<<<<<<
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)
  */
       plink2::aligned_free((__pyx_v_self->_info_ptr[0]).nonref_flags);
 
-      /* "pgenlib.pyx":1397
- *             if self._info_ptr[0].vrtypes:
- *                 aligned_free(self._info_ptr[0].vrtypes)
+      /* "pgenlib.pyx":1509
+ *             if self._info_ptr[0].allele_idx_offsets:
+ *                 aligned_free(self._info_ptr[0].allele_idx_offsets)
  *             if self._info_ptr[0].nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:
  */
     }
 
-    /* "pgenlib.pyx":1399
+    /* "pgenlib.pyx":1511
  *             if self._info_ptr[0].nonref_flags:
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:             # <<<<<<<<<<<<<<
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):
  */
     __pyx_t_1 = (__pyx_v_self->_state_ptr != 0);
     if (__pyx_t_1) {
 
-      /* "pgenlib.pyx":1400
+      /* "pgenlib.pyx":1512
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)             # <<<<<<<<<<<<<<
  *                 if PgrGetFreadBuf(self._state_ptr):
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  */
       (void)(plink2::CleanupPgr(__pyx_v_self->_state_ptr, (&__pyx_v_reterr)));
 
-      /* "pgenlib.pyx":1401
+      /* "pgenlib.pyx":1513
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):             # <<<<<<<<<<<<<<
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)
  */
       __pyx_t_1 = (plink2::PgrGetFreadBuf(__pyx_v_self->_state_ptr) != 0);
       if (__pyx_t_1) {
 
-        /* "pgenlib.pyx":1402
+        /* "pgenlib.pyx":1514
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))             # <<<<<<<<<<<<<<
  *                 PyMem_Free(self._state_ptr)
  *             PyMem_Free(self._info_ptr)
  */
         plink2::aligned_free(plink2::PgrGetFreadBuf(__pyx_v_self->_state_ptr));
 
-        /* "pgenlib.pyx":1401
+        /* "pgenlib.pyx":1513
  *             if self._state_ptr:
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):             # <<<<<<<<<<<<<<
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)
  */
       }
 
-      /* "pgenlib.pyx":1403
+      /* "pgenlib.pyx":1515
  *                 if PgrGetFreadBuf(self._state_ptr):
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)             # <<<<<<<<<<<<<<
  *             PyMem_Free(self._info_ptr)
  *         return
  */
       PyMem_Free(__pyx_v_self->_state_ptr);
 
-      /* "pgenlib.pyx":1399
+      /* "pgenlib.pyx":1511
  *             if self._info_ptr[0].nonref_flags:
  *                 aligned_free(self._info_ptr[0].nonref_flags)
  *             if self._state_ptr:             # <<<<<<<<<<<<<<
  *                 CleanupPgr(self._state_ptr, &reterr)
  *                 if PgrGetFreadBuf(self._state_ptr):
  */
     }
 
-    /* "pgenlib.pyx":1404
+    /* "pgenlib.pyx":1516
  *                     aligned_free(PgrGetFreadBuf(self._state_ptr))
  *                 PyMem_Free(self._state_ptr)
  *             PyMem_Free(self._info_ptr)             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     PyMem_Free(__pyx_v_self->_info_ptr);
 
-    /* "pgenlib.pyx":1393
+    /* "pgenlib.pyx":1503
  *     def __dealloc__(self):
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:             # <<<<<<<<<<<<<<
  *             CleanupPgfi(self._info_ptr, &reterr)
  *             if self._info_ptr[0].vrtypes:
  */
   }
 
-  /* "pgenlib.pyx":1405
+  /* "pgenlib.pyx":1517
  *                 PyMem_Free(self._state_ptr)
  *             PyMem_Free(self._info_ptr)
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1391
+  /* "pgenlib.pyx":1501
  * 
  * 
  *     def __dealloc__(self):             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._info_ptr:
  */
 
@@ -23334,15 +24589,15 @@
 
   /* "(tree fragment)":2
  * def __reduce_cython__(self):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  */
-  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__19, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 2, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__22, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 2, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_Raise(__pyx_t_1, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __PYX_ERR(1, 2, __pyx_L1_error)
 
   /* "(tree fragment)":1
  * def __reduce_cython__(self):             # <<<<<<<<<<<<<<
@@ -23390,15 +24645,15 @@
   __Pyx_RefNannySetupContext("__setstate_cython__", 0);
 
   /* "(tree fragment)":4
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  */
-  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__20, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 4, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__23, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 4, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_Raise(__pyx_t_1, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __PYX_ERR(1, 4, __pyx_L1_error)
 
   /* "(tree fragment)":3
  * def __reduce_cython__(self):
@@ -23413,15 +24668,15 @@
   __Pyx_AddTraceback("pgenlib.PgenReader.__setstate_cython__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = NULL;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1409
+/* "pgenlib.pyx":1520
  * 
  * 
  * cdef bytes_to_bits_internal(np.ndarray[np.uint8_t,mode="c",cast=True] boolbytes, uint32_t sample_ct, uintptr_t* bitarr):             # <<<<<<<<<<<<<<
  *     BytesToBitsUnsafe(boolbytes, sample_ct, bitarr)
  * 
  */
 
@@ -23437,29 +24692,29 @@
   __Pyx_RefNannySetupContext("bytes_to_bits_internal", 0);
   __pyx_pybuffer_boolbytes.pybuffer.buf = NULL;
   __pyx_pybuffer_boolbytes.refcount = 0;
   __pyx_pybuffernd_boolbytes.data = NULL;
   __pyx_pybuffernd_boolbytes.rcbuffer = &__pyx_pybuffer_boolbytes;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_boolbytes.rcbuffer->pybuffer, (PyObject*)__pyx_v_boolbytes, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1409, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_boolbytes.rcbuffer->pybuffer, (PyObject*)__pyx_v_boolbytes, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1520, __pyx_L1_error)
   }
   __pyx_pybuffernd_boolbytes.diminfo[0].strides = __pyx_pybuffernd_boolbytes.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_boolbytes.diminfo[0].shape = __pyx_pybuffernd_boolbytes.rcbuffer->pybuffer.shape[0];
 
-  /* "pgenlib.pyx":1410
+  /* "pgenlib.pyx":1521
  * 
  * cdef bytes_to_bits_internal(np.ndarray[np.uint8_t,mode="c",cast=True] boolbytes, uint32_t sample_ct, uintptr_t* bitarr):
  *     BytesToBitsUnsafe(boolbytes, sample_ct, bitarr)             # <<<<<<<<<<<<<<
  * 
- * cdef class PgenWriter:
+ * 
  */
-  __pyx_t_1 = __Pyx_PyObject_AsUString(((PyObject *)__pyx_v_boolbytes)); if (unlikely((!__pyx_t_1) && PyErr_Occurred())) __PYX_ERR(0, 1410, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_AsUString(((PyObject *)__pyx_v_boolbytes)); if (unlikely((!__pyx_t_1) && PyErr_Occurred())) __PYX_ERR(0, 1521, __pyx_L1_error)
   plink2::BytesToBitsUnsafe(__pyx_t_1, __pyx_v_sample_ct, __pyx_v_bitarr);
 
-  /* "pgenlib.pyx":1409
+  /* "pgenlib.pyx":1520
  * 
  * 
  * cdef bytes_to_bits_internal(np.ndarray[np.uint8_t,mode="c",cast=True] boolbytes, uint32_t sample_ct, uintptr_t* bitarr):             # <<<<<<<<<<<<<<
  *     BytesToBitsUnsafe(boolbytes, sample_ct, bitarr)
  * 
  */
 
@@ -23480,55 +24735,68 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_boolbytes.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1424
- * 
+/* "pgenlib.pyx":1540
+ *     cdef uint16_t* _dosage_main
  * 
  *     def __cinit__(self, bytes filename, uint32_t sample_ct,             # <<<<<<<<<<<<<<
- *                   uint32_t variant_ct, object nonref_flags,
- *                   object allele_idx_offsets = None,
+ *                   object variant_ct = None, object nonref_flags = True,
+ *                   uint32_t allele_ct_limit = 2,
  */
 
 /* Python wrapper */
 static int __pyx_pw_7pgenlib_10PgenWriter_1__cinit__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static int __pyx_pw_7pgenlib_10PgenWriter_1__cinit__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
   PyObject *__pyx_v_filename = 0;
   uint32_t __pyx_v_sample_ct;
-  uint32_t __pyx_v_variant_ct;
+  PyObject *__pyx_v_variant_ct = 0;
   PyObject *__pyx_v_nonref_flags = 0;
-  PyObject *__pyx_v_allele_idx_offsets = 0;
+  uint32_t __pyx_v_allele_ct_limit;
   int __pyx_v_hardcall_phase_present;
   int __pyx_v_dosage_present;
   int __pyx_v_dosage_phase_present;
+  PyObject *__pyx_v_variant_ct_limit = 0;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   int __pyx_r;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("__cinit__ (wrapper)", 0);
   {
-    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_filename,&__pyx_n_s_sample_ct,&__pyx_n_s_variant_ct,&__pyx_n_s_nonref_flags,&__pyx_n_s_allele_idx_offsets,&__pyx_n_s_hardcall_phase_present,&__pyx_n_s_dosage_present,&__pyx_n_s_dosage_phase_present,0};
-    PyObject* values[8] = {0,0,0,0,0,0,0,0};
+    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_filename,&__pyx_n_s_sample_ct,&__pyx_n_s_variant_ct,&__pyx_n_s_nonref_flags,&__pyx_n_s_allele_ct_limit,&__pyx_n_s_hardcall_phase_present,&__pyx_n_s_dosage_present,&__pyx_n_s_dosage_phase_present,&__pyx_n_s_variant_ct_limit,0};
+    PyObject* values[9] = {0,0,0,0,0,0,0,0,0};
 
-    /* "pgenlib.pyx":1426
+    /* "pgenlib.pyx":1541
+ * 
  *     def __cinit__(self, bytes filename, uint32_t sample_ct,
- *                   uint32_t variant_ct, object nonref_flags,
- *                   object allele_idx_offsets = None,             # <<<<<<<<<<<<<<
+ *                   object variant_ct = None, object nonref_flags = True,             # <<<<<<<<<<<<<<
+ *                   uint32_t allele_ct_limit = 2,
  *                   bint hardcall_phase_present = False,
+ */
+    values[2] = ((PyObject *)Py_None);
+    values[3] = ((PyObject *)Py_True);
+
+    /* "pgenlib.pyx":1546
  *                   bint dosage_present = False,
+ *                   bint dosage_phase_present = False,
+ *                   object variant_ct_limit = None):             # <<<<<<<<<<<<<<
+ *         cdef PgenWriteMode write_mode = kPgenWriteBackwardSeek
+ *         cdef uint32_t cur_variant_ct_limit
  */
-    values[4] = ((PyObject *)Py_None);
+    values[8] = ((PyObject *)Py_None);
     if (unlikely(__pyx_kwds)) {
       Py_ssize_t kw_args;
       const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
       switch (pos_args) {
+        case  9: values[8] = PyTuple_GET_ITEM(__pyx_args, 8);
+        CYTHON_FALLTHROUGH;
         case  8: values[7] = PyTuple_GET_ITEM(__pyx_args, 7);
         CYTHON_FALLTHROUGH;
         case  7: values[6] = PyTuple_GET_ITEM(__pyx_args, 6);
         CYTHON_FALLTHROUGH;
         case  6: values[5] = PyTuple_GET_ITEM(__pyx_args, 5);
         CYTHON_FALLTHROUGH;
         case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
@@ -23549,32 +24817,32 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_filename)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_sample_ct)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 4, 8, 1); __PYX_ERR(0, 1424, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 2, 9, 1); __PYX_ERR(0, 1540, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
-        if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_ct)) != 0)) kw_args--;
-        else {
-          __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 4, 8, 2); __PYX_ERR(0, 1424, __pyx_L3_error)
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_ct);
+          if (value) { values[2] = value; kw_args--; }
         }
         CYTHON_FALLTHROUGH;
         case  3:
-        if (likely((values[3] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_nonref_flags)) != 0)) kw_args--;
-        else {
-          __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 4, 8, 3); __PYX_ERR(0, 1424, __pyx_L3_error)
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_nonref_flags);
+          if (value) { values[3] = value; kw_args--; }
         }
         CYTHON_FALLTHROUGH;
         case  4:
         if (kw_args > 0) {
-          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_idx_offsets);
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_ct_limit);
           if (value) { values[4] = value; kw_args--; }
         }
         CYTHON_FALLTHROUGH;
         case  5:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_hardcall_phase_present);
           if (value) { values[5] = value; kw_args--; }
@@ -23587,788 +24855,1061 @@
         }
         CYTHON_FALLTHROUGH;
         case  7:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_dosage_phase_present);
           if (value) { values[7] = value; kw_args--; }
         }
+        CYTHON_FALLTHROUGH;
+        case  8:
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_variant_ct_limit);
+          if (value) { values[8] = value; kw_args--; }
+        }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__cinit__") < 0)) __PYX_ERR(0, 1424, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__cinit__") < 0)) __PYX_ERR(0, 1540, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
+        case  9: values[8] = PyTuple_GET_ITEM(__pyx_args, 8);
+        CYTHON_FALLTHROUGH;
         case  8: values[7] = PyTuple_GET_ITEM(__pyx_args, 7);
         CYTHON_FALLTHROUGH;
         case  7: values[6] = PyTuple_GET_ITEM(__pyx_args, 6);
         CYTHON_FALLTHROUGH;
         case  6: values[5] = PyTuple_GET_ITEM(__pyx_args, 5);
         CYTHON_FALLTHROUGH;
         case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
         CYTHON_FALLTHROUGH;
         case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
-        values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
-        values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
+        CYTHON_FALLTHROUGH;
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
+        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_filename = ((PyObject*)values[0]);
-    __pyx_v_sample_ct = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_sample_ct == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1424, __pyx_L3_error)
-    __pyx_v_variant_ct = __Pyx_PyInt_As_uint32_t(values[2]); if (unlikely((__pyx_v_variant_ct == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1425, __pyx_L3_error)
+    __pyx_v_sample_ct = __Pyx_PyInt_As_uint32_t(values[1]); if (unlikely((__pyx_v_sample_ct == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1540, __pyx_L3_error)
+    __pyx_v_variant_ct = values[2];
     __pyx_v_nonref_flags = values[3];
-    __pyx_v_allele_idx_offsets = values[4];
+    if (values[4]) {
+      __pyx_v_allele_ct_limit = __Pyx_PyInt_As_uint32_t(values[4]); if (unlikely((__pyx_v_allele_ct_limit == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1542, __pyx_L3_error)
+    } else {
+      __pyx_v_allele_ct_limit = ((uint32_t)2);
+    }
     if (values[5]) {
-      __pyx_v_hardcall_phase_present = __Pyx_PyObject_IsTrue(values[5]); if (unlikely((__pyx_v_hardcall_phase_present == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1427, __pyx_L3_error)
+      __pyx_v_hardcall_phase_present = __Pyx_PyObject_IsTrue(values[5]); if (unlikely((__pyx_v_hardcall_phase_present == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1543, __pyx_L3_error)
     } else {
 
-      /* "pgenlib.pyx":1427
- *                   uint32_t variant_ct, object nonref_flags,
- *                   object allele_idx_offsets = None,
+      /* "pgenlib.pyx":1543
+ *                   object variant_ct = None, object nonref_flags = True,
+ *                   uint32_t allele_ct_limit = 2,
  *                   bint hardcall_phase_present = False,             # <<<<<<<<<<<<<<
  *                   bint dosage_present = False,
- *                   bint dosage_phase_present = False):
+ *                   bint dosage_phase_present = False,
  */
       __pyx_v_hardcall_phase_present = ((int)0);
     }
     if (values[6]) {
-      __pyx_v_dosage_present = __Pyx_PyObject_IsTrue(values[6]); if (unlikely((__pyx_v_dosage_present == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1428, __pyx_L3_error)
+      __pyx_v_dosage_present = __Pyx_PyObject_IsTrue(values[6]); if (unlikely((__pyx_v_dosage_present == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1544, __pyx_L3_error)
     } else {
 
-      /* "pgenlib.pyx":1428
- *                   object allele_idx_offsets = None,
+      /* "pgenlib.pyx":1544
+ *                   uint32_t allele_ct_limit = 2,
  *                   bint hardcall_phase_present = False,
  *                   bint dosage_present = False,             # <<<<<<<<<<<<<<
- *                   bint dosage_phase_present = False):
- *         if dosage_phase_present and not dosage_present:
+ *                   bint dosage_phase_present = False,
+ *                   object variant_ct_limit = None):
  */
       __pyx_v_dosage_present = ((int)0);
     }
     if (values[7]) {
-      __pyx_v_dosage_phase_present = __Pyx_PyObject_IsTrue(values[7]); if (unlikely((__pyx_v_dosage_phase_present == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1429, __pyx_L3_error)
+      __pyx_v_dosage_phase_present = __Pyx_PyObject_IsTrue(values[7]); if (unlikely((__pyx_v_dosage_phase_present == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1545, __pyx_L3_error)
     } else {
 
-      /* "pgenlib.pyx":1429
+      /* "pgenlib.pyx":1545
  *                   bint hardcall_phase_present = False,
  *                   bint dosage_present = False,
- *                   bint dosage_phase_present = False):             # <<<<<<<<<<<<<<
- *         if dosage_phase_present and not dosage_present:
- *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
+ *                   bint dosage_phase_present = False,             # <<<<<<<<<<<<<<
+ *                   object variant_ct_limit = None):
+ *         cdef PgenWriteMode write_mode = kPgenWriteBackwardSeek
  */
       __pyx_v_dosage_phase_present = ((int)0);
     }
+    __pyx_v_variant_ct_limit = values[8];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 4, 8, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1424, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("__cinit__", 0, 2, 9, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1540, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenWriter.__cinit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return -1;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_filename), (&PyBytes_Type), 1, "filename", 1))) __PYX_ERR(0, 1424, __pyx_L1_error)
-  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter___cinit__(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_filename, __pyx_v_sample_ct, __pyx_v_variant_ct, __pyx_v_nonref_flags, __pyx_v_allele_idx_offsets, __pyx_v_hardcall_phase_present, __pyx_v_dosage_present, __pyx_v_dosage_phase_present);
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_filename), (&PyBytes_Type), 1, "filename", 1))) __PYX_ERR(0, 1540, __pyx_L1_error)
+  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter___cinit__(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_filename, __pyx_v_sample_ct, __pyx_v_variant_ct, __pyx_v_nonref_flags, __pyx_v_allele_ct_limit, __pyx_v_hardcall_phase_present, __pyx_v_dosage_present, __pyx_v_dosage_phase_present, __pyx_v_variant_ct_limit);
 
-  /* "pgenlib.pyx":1424
- * 
+  /* "pgenlib.pyx":1540
+ *     cdef uint16_t* _dosage_main
  * 
  *     def __cinit__(self, bytes filename, uint32_t sample_ct,             # <<<<<<<<<<<<<<
- *                   uint32_t variant_ct, object nonref_flags,
- *                   object allele_idx_offsets = None,
+ *                   object variant_ct = None, object nonref_flags = True,
+ *                   uint32_t allele_ct_limit = 2,
  */
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = -1;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-static int __pyx_pf_7pgenlib_10PgenWriter___cinit__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyObject *__pyx_v_filename, uint32_t __pyx_v_sample_ct, uint32_t __pyx_v_variant_ct, PyObject *__pyx_v_nonref_flags, PyObject *__pyx_v_allele_idx_offsets, int __pyx_v_hardcall_phase_present, int __pyx_v_dosage_present, int __pyx_v_dosage_phase_present) {
-  PyObject *__pyx_v_uii = NULL;
+static int __pyx_pf_7pgenlib_10PgenWriter___cinit__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyObject *__pyx_v_filename, uint32_t __pyx_v_sample_ct, PyObject *__pyx_v_variant_ct, PyObject *__pyx_v_nonref_flags, uint32_t __pyx_v_allele_ct_limit, int __pyx_v_hardcall_phase_present, int __pyx_v_dosage_present, int __pyx_v_dosage_phase_present, PyObject *__pyx_v_variant_ct_limit) {
+  plink2::PgenWriteMode __pyx_v_write_mode;
+  uint32_t __pyx_v_cur_variant_ct_limit;
   uint32_t __pyx_v_nonref_flags_storage;
   uint32_t __pyx_v_bitvec_cacheline_ct;
   char const *__pyx_v_fname;
   plink2::PgenGlobalFlags __pyx_v_phase_dosage_gflags;
   uintptr_t __pyx_v_alloc_cacheline_ct;
   uint32_t __pyx_v_max_vrec_len;
   plink2::PglErr __pyx_v_reterr;
   uint32_t __pyx_v_genovec_cacheline_ct;
+  uint32_t __pyx_v_patch_01_vals_cacheline_ct;
+  uint32_t __pyx_v_patch_10_vals_cacheline_ct;
   uint32_t __pyx_v_dosage_main_cacheline_ct;
   unsigned char *__pyx_v_spgw_alloc;
+  unsigned char *__pyx_v_spgw_alloc_iter;
   int __pyx_r;
   __Pyx_RefNannyDeclarations
   int __pyx_t_1;
   int __pyx_t_2;
-  PyObject *__pyx_t_3 = NULL;
+  uint32_t __pyx_t_3;
   PyObject *__pyx_t_4 = NULL;
-  Py_ssize_t __pyx_t_5;
-  PyObject *(*__pyx_t_6)(PyObject *);
-  PyObject *__pyx_t_7 = NULL;
-  PyObject *__pyx_t_8 = NULL;
-  char const *__pyx_t_9;
+  char const *__pyx_t_5;
+  PyObject *__pyx_t_6 = NULL;
+  long __pyx_t_7;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__cinit__", 0);
 
-  /* "pgenlib.pyx":1430
- *                   bint dosage_present = False,
- *                   bint dosage_phase_present = False):
+  /* "pgenlib.pyx":1547
+ *                   bint dosage_phase_present = False,
+ *                   object variant_ct_limit = None):
+ *         cdef PgenWriteMode write_mode = kPgenWriteBackwardSeek             # <<<<<<<<<<<<<<
+ *         cdef uint32_t cur_variant_ct_limit
+ *         if variant_ct is not None:
+ */
+  __pyx_v_write_mode = plink2::kPgenWriteBackwardSeek;
+
+  /* "pgenlib.pyx":1549
+ *         cdef PgenWriteMode write_mode = kPgenWriteBackwardSeek
+ *         cdef uint32_t cur_variant_ct_limit
+ *         if variant_ct is not None:             # <<<<<<<<<<<<<<
+ *             # Could also enforce variant_ct >= variant_ct_limit when both are
+ *             # provided?
+ */
+  __pyx_t_1 = (__pyx_v_variant_ct != Py_None);
+  __pyx_t_2 = (__pyx_t_1 != 0);
+  if (__pyx_t_2) {
+
+    /* "pgenlib.pyx":1552
+ *             # Could also enforce variant_ct >= variant_ct_limit when both are
+ *             # provided?
+ *             cur_variant_ct_limit = variant_ct             # <<<<<<<<<<<<<<
+ *         elif variant_ct_limit is not None:
+ *             write_mode = kPgenWriteAndCopy
+ */
+    __pyx_t_3 = __Pyx_PyInt_As_uint32_t(__pyx_v_variant_ct); if (unlikely((__pyx_t_3 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1552, __pyx_L1_error)
+    __pyx_v_cur_variant_ct_limit = __pyx_t_3;
+
+    /* "pgenlib.pyx":1549
+ *         cdef PgenWriteMode write_mode = kPgenWriteBackwardSeek
+ *         cdef uint32_t cur_variant_ct_limit
+ *         if variant_ct is not None:             # <<<<<<<<<<<<<<
+ *             # Could also enforce variant_ct >= variant_ct_limit when both are
+ *             # provided?
+ */
+    goto __pyx_L3;
+  }
+
+  /* "pgenlib.pyx":1553
+ *             # provided?
+ *             cur_variant_ct_limit = variant_ct
+ *         elif variant_ct_limit is not None:             # <<<<<<<<<<<<<<
+ *             write_mode = kPgenWriteAndCopy
+ *             cur_variant_ct_limit = variant_ct_limit
+ */
+  __pyx_t_2 = (__pyx_v_variant_ct_limit != Py_None);
+  __pyx_t_1 = (__pyx_t_2 != 0);
+  if (likely(__pyx_t_1)) {
+
+    /* "pgenlib.pyx":1554
+ *             cur_variant_ct_limit = variant_ct
+ *         elif variant_ct_limit is not None:
+ *             write_mode = kPgenWriteAndCopy             # <<<<<<<<<<<<<<
+ *             cur_variant_ct_limit = variant_ct_limit
+ *         else:
+ */
+    __pyx_v_write_mode = plink2::kPgenWriteAndCopy;
+
+    /* "pgenlib.pyx":1555
+ *         elif variant_ct_limit is not None:
+ *             write_mode = kPgenWriteAndCopy
+ *             cur_variant_ct_limit = variant_ct_limit             # <<<<<<<<<<<<<<
+ *         else:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")
+ */
+    __pyx_t_3 = __Pyx_PyInt_As_uint32_t(__pyx_v_variant_ct_limit); if (unlikely((__pyx_t_3 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1555, __pyx_L1_error)
+    __pyx_v_cur_variant_ct_limit = __pyx_t_3;
+
+    /* "pgenlib.pyx":1553
+ *             # provided?
+ *             cur_variant_ct_limit = variant_ct
+ *         elif variant_ct_limit is not None:             # <<<<<<<<<<<<<<
+ *             write_mode = kPgenWriteAndCopy
+ *             cur_variant_ct_limit = variant_ct_limit
+ */
+    goto __pyx_L3;
+  }
+
+  /* "pgenlib.pyx":1557
+ *             cur_variant_ct_limit = variant_ct_limit
+ *         else:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")             # <<<<<<<<<<<<<<
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
+ */
+  /*else*/ {
+    __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__24, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1557, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
+    __Pyx_Raise(__pyx_t_4, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __PYX_ERR(0, 1557, __pyx_L1_error)
+  }
+  __pyx_L3:;
+
+  /* "pgenlib.pyx":1558
+ *         else:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
+ *         if dosage_phase_present and not dosage_present:
+ */
+  __pyx_t_2 = ((__pyx_v_cur_variant_ct_limit == 0) != 0);
+  if (!__pyx_t_2) {
+  } else {
+    __pyx_t_1 = __pyx_t_2;
+    goto __pyx_L5_bool_binop_done;
+  }
+  __pyx_t_2 = ((__pyx_v_cur_variant_ct_limit > plink2::kPglMaxVariantCt) != 0);
+  __pyx_t_1 = __pyx_t_2;
+  __pyx_L5_bool_binop_done:;
+  if (unlikely(__pyx_t_1)) {
+
+    /* "pgenlib.pyx":1559
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")             # <<<<<<<<<<<<<<
+ *         if dosage_phase_present and not dosage_present:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
+ */
+    __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__25, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1559, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
+    __Pyx_Raise(__pyx_t_4, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __PYX_ERR(0, 1559, __pyx_L1_error)
+
+    /* "pgenlib.pyx":1558
+ *         else:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
+ *         if dosage_phase_present and not dosage_present:
+ */
+  }
+
+  /* "pgenlib.pyx":1560
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
  *         if dosage_phase_present and not dosage_present:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
- *         if allele_idx_offsets is not None:
+ *         if allele_ct_limit != 2:
  */
   __pyx_t_2 = (__pyx_v_dosage_phase_present != 0);
   if (__pyx_t_2) {
   } else {
     __pyx_t_1 = __pyx_t_2;
-    goto __pyx_L4_bool_binop_done;
+    goto __pyx_L8_bool_binop_done;
   }
   __pyx_t_2 = ((!(__pyx_v_dosage_present != 0)) != 0);
   __pyx_t_1 = __pyx_t_2;
-  __pyx_L4_bool_binop_done:;
+  __pyx_L8_bool_binop_done:;
   if (unlikely(__pyx_t_1)) {
 
-    /* "pgenlib.pyx":1431
- *                   bint dosage_phase_present = False):
+    /* "pgenlib.pyx":1561
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
  *         if dosage_phase_present and not dosage_present:
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")             # <<<<<<<<<<<<<<
- *         if allele_idx_offsets is not None:
- *             for uii in range(variant_ct + 1):
+ *         if allele_ct_limit != 2:
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
  */
-    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__21, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1431, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_3);
-    __Pyx_Raise(__pyx_t_3, 0, 0, 0);
-    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __PYX_ERR(0, 1431, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__26, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1561, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
+    __Pyx_Raise(__pyx_t_4, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __PYX_ERR(0, 1561, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1430
- *                   bint dosage_present = False,
- *                   bint dosage_phase_present = False):
+    /* "pgenlib.pyx":1560
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
  *         if dosage_phase_present and not dosage_present:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
- *         if allele_idx_offsets is not None:
+ *         if allele_ct_limit != 2:
  */
   }
 
-  /* "pgenlib.pyx":1432
+  /* "pgenlib.pyx":1562
  *         if dosage_phase_present and not dosage_present:
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
- *         if allele_idx_offsets is not None:             # <<<<<<<<<<<<<<
- *             for uii in range(variant_ct + 1):
- *                 if allele_idx_offsets[uii] != uii * 2:
+ *         if allele_ct_limit != 2:             # <<<<<<<<<<<<<<
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")
  */
-  __pyx_t_1 = (__pyx_v_allele_idx_offsets != Py_None);
-  __pyx_t_2 = (__pyx_t_1 != 0);
-  if (__pyx_t_2) {
+  __pyx_t_1 = ((__pyx_v_allele_ct_limit != 2) != 0);
+  if (__pyx_t_1) {
 
-    /* "pgenlib.pyx":1433
+    /* "pgenlib.pyx":1563
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
- *         if allele_idx_offsets is not None:
- *             for uii in range(variant_ct + 1):             # <<<<<<<<<<<<<<
- *                 if allele_idx_offsets[uii] != uii * 2:
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")
+ *         if allele_ct_limit != 2:
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")
+ *             write_mode = kPgenWriteAndCopy
  */
-    __pyx_t_3 = __Pyx_PyInt_From_long((__pyx_v_variant_ct + 1)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1433, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_3);
-    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_builtin_range, __pyx_t_3); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1433, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_4);
-    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    if (likely(PyList_CheckExact(__pyx_t_4)) || PyTuple_CheckExact(__pyx_t_4)) {
-      __pyx_t_3 = __pyx_t_4; __Pyx_INCREF(__pyx_t_3); __pyx_t_5 = 0;
-      __pyx_t_6 = NULL;
+    __pyx_t_2 = ((__pyx_v_allele_ct_limit < 2) != 0);
+    if (!__pyx_t_2) {
     } else {
-      __pyx_t_5 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_4); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1433, __pyx_L1_error)
-      __Pyx_GOTREF(__pyx_t_3);
-      __pyx_t_6 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1433, __pyx_L1_error)
+      __pyx_t_1 = __pyx_t_2;
+      goto __pyx_L12_bool_binop_done;
     }
-    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-    for (;;) {
-      if (likely(!__pyx_t_6)) {
-        if (likely(PyList_CheckExact(__pyx_t_3))) {
-          if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_3)) break;
-          #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
-          __pyx_t_4 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_5); __Pyx_INCREF(__pyx_t_4); __pyx_t_5++; if (unlikely(0 < 0)) __PYX_ERR(0, 1433, __pyx_L1_error)
-          #else
-          __pyx_t_4 = PySequence_ITEM(__pyx_t_3, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1433, __pyx_L1_error)
-          __Pyx_GOTREF(__pyx_t_4);
-          #endif
-        } else {
-          if (__pyx_t_5 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
-          #if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
-          __pyx_t_4 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_5); __Pyx_INCREF(__pyx_t_4); __pyx_t_5++; if (unlikely(0 < 0)) __PYX_ERR(0, 1433, __pyx_L1_error)
-          #else
-          __pyx_t_4 = PySequence_ITEM(__pyx_t_3, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1433, __pyx_L1_error)
-          __Pyx_GOTREF(__pyx_t_4);
-          #endif
-        }
-      } else {
-        __pyx_t_4 = __pyx_t_6(__pyx_t_3);
-        if (unlikely(!__pyx_t_4)) {
-          PyObject* exc_type = PyErr_Occurred();
-          if (exc_type) {
-            if (likely(__Pyx_PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
-            else __PYX_ERR(0, 1433, __pyx_L1_error)
-          }
-          break;
-        }
-        __Pyx_GOTREF(__pyx_t_4);
-      }
-      __Pyx_XDECREF_SET(__pyx_v_uii, __pyx_t_4);
-      __pyx_t_4 = 0;
+    __pyx_t_2 = ((__pyx_v_allele_ct_limit > plink2::kPglMaxAlleleCt) != 0);
+    __pyx_t_1 = __pyx_t_2;
+    __pyx_L12_bool_binop_done:;
+    if (unlikely(__pyx_t_1)) {
 
-      /* "pgenlib.pyx":1434
- *         if allele_idx_offsets is not None:
- *             for uii in range(variant_ct + 1):
- *                 if allele_idx_offsets[uii] != uii * 2:             # <<<<<<<<<<<<<<
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")
+      /* "pgenlib.pyx":1564
+ *         if allele_ct_limit != 2:
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")             # <<<<<<<<<<<<<<
+ *             write_mode = kPgenWriteAndCopy
  * 
  */
-      __pyx_t_4 = __Pyx_PyObject_GetItem(__pyx_v_allele_idx_offsets, __pyx_v_uii); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1434, __pyx_L1_error)
+      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__27, NULL); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1564, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_4);
-      __pyx_t_7 = PyNumber_Multiply(__pyx_v_uii, __pyx_int_2); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1434, __pyx_L1_error)
-      __Pyx_GOTREF(__pyx_t_7);
-      __pyx_t_8 = PyObject_RichCompare(__pyx_t_4, __pyx_t_7, Py_NE); __Pyx_XGOTREF(__pyx_t_8); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1434, __pyx_L1_error)
+      __Pyx_Raise(__pyx_t_4, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
-      __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-      __pyx_t_2 = __Pyx_PyObject_IsTrue(__pyx_t_8); if (unlikely(__pyx_t_2 < 0)) __PYX_ERR(0, 1434, __pyx_L1_error)
-      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-      if (unlikely(__pyx_t_2)) {
+      __PYX_ERR(0, 1564, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1435
- *             for uii in range(variant_ct + 1):
- *                 if allele_idx_offsets[uii] != uii * 2:
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")             # <<<<<<<<<<<<<<
- * 
- *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
+      /* "pgenlib.pyx":1563
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
+ *         if allele_ct_limit != 2:
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")
+ *             write_mode = kPgenWriteAndCopy
  */
-        __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__22, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1435, __pyx_L1_error)
-        __Pyx_GOTREF(__pyx_t_8);
-        __Pyx_Raise(__pyx_t_8, 0, 0, 0);
-        __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-        __PYX_ERR(0, 1435, __pyx_L1_error)
+    }
 
-        /* "pgenlib.pyx":1434
- *         if allele_idx_offsets is not None:
- *             for uii in range(variant_ct + 1):
- *                 if allele_idx_offsets[uii] != uii * 2:             # <<<<<<<<<<<<<<
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")
+    /* "pgenlib.pyx":1565
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")
+ *             write_mode = kPgenWriteAndCopy             # <<<<<<<<<<<<<<
  * 
+ *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
  */
-      }
-
-      /* "pgenlib.pyx":1433
- *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
- *         if allele_idx_offsets is not None:
- *             for uii in range(variant_ct + 1):             # <<<<<<<<<<<<<<
- *                 if allele_idx_offsets[uii] != uii * 2:
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")
- */
-    }
-    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
+    __pyx_v_write_mode = plink2::kPgenWriteAndCopy;
 
-    /* "pgenlib.pyx":1432
+    /* "pgenlib.pyx":1562
  *         if dosage_phase_present and not dosage_present:
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
- *         if allele_idx_offsets is not None:             # <<<<<<<<<<<<<<
- *             for uii in range(variant_ct + 1):
- *                 if allele_idx_offsets[uii] != uii * 2:
+ *         if allele_ct_limit != 2:             # <<<<<<<<<<<<<<
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")
  */
   }
 
-  /* "pgenlib.pyx":1437
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")
+  /* "pgenlib.pyx":1567
+ *             write_mode = kPgenWriteAndCopy
  * 
  *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))             # <<<<<<<<<<<<<<
  *         if not self._state_ptr:
  *             raise MemoryError()
  */
   __pyx_v_self->_state_ptr = ((plink2::STPgenWriter *)PyMem_Malloc((sizeof(plink2::STPgenWriter))));
 
-  /* "pgenlib.pyx":1438
+  /* "pgenlib.pyx":1568
  * 
  *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
  *         if not self._state_ptr:             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         self._nonref_flags = NULL
  */
-  __pyx_t_2 = ((!(__pyx_v_self->_state_ptr != 0)) != 0);
-  if (unlikely(__pyx_t_2)) {
+  __pyx_t_1 = ((!(__pyx_v_self->_state_ptr != 0)) != 0);
+  if (unlikely(__pyx_t_1)) {
 
-    /* "pgenlib.pyx":1439
+    /* "pgenlib.pyx":1569
  *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
  *         if not self._state_ptr:
  *             raise MemoryError()             # <<<<<<<<<<<<<<
  *         self._nonref_flags = NULL
  *         cdef uint32_t nonref_flags_storage = 0
  */
-    PyErr_NoMemory(); __PYX_ERR(0, 1439, __pyx_L1_error)
+    PyErr_NoMemory(); __PYX_ERR(0, 1569, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1438
+    /* "pgenlib.pyx":1568
  * 
  *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
  *         if not self._state_ptr:             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         self._nonref_flags = NULL
  */
   }
 
-  /* "pgenlib.pyx":1440
+  /* "pgenlib.pyx":1570
  *         if not self._state_ptr:
  *             raise MemoryError()
  *         self._nonref_flags = NULL             # <<<<<<<<<<<<<<
  *         cdef uint32_t nonref_flags_storage = 0
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
  */
   __pyx_v_self->_nonref_flags = NULL;
 
-  /* "pgenlib.pyx":1441
+  /* "pgenlib.pyx":1571
  *             raise MemoryError()
  *         self._nonref_flags = NULL
  *         cdef uint32_t nonref_flags_storage = 0             # <<<<<<<<<<<<<<
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
  *         if nonref_flags is not None:
  */
   __pyx_v_nonref_flags_storage = 0;
 
-  /* "pgenlib.pyx":1442
+  /* "pgenlib.pyx":1572
  *         self._nonref_flags = NULL
  *         cdef uint32_t nonref_flags_storage = 0
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)             # <<<<<<<<<<<<<<
  *         if nonref_flags is not None:
  *             if type(nonref_flags) == type(True):
  */
   __pyx_v_bitvec_cacheline_ct = plink2::DivUp(__pyx_v_sample_ct, plink2::kBitsPerCacheline);
 
-  /* "pgenlib.pyx":1443
+  /* "pgenlib.pyx":1573
  *         cdef uint32_t nonref_flags_storage = 0
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
  *         if nonref_flags is not None:             # <<<<<<<<<<<<<<
  *             if type(nonref_flags) == type(True):
  *                 if nonref_flags:
  */
-  __pyx_t_2 = (__pyx_v_nonref_flags != Py_None);
-  __pyx_t_1 = (__pyx_t_2 != 0);
-  if (__pyx_t_1) {
+  __pyx_t_1 = (__pyx_v_nonref_flags != Py_None);
+  __pyx_t_2 = (__pyx_t_1 != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":1444
+    /* "pgenlib.pyx":1574
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
  *         if nonref_flags is not None:
  *             if type(nonref_flags) == type(True):             # <<<<<<<<<<<<<<
  *                 if nonref_flags:
  *                     nonref_flags_storage = 2
  */
-    __pyx_t_3 = PyObject_RichCompare(((PyObject *)Py_TYPE(__pyx_v_nonref_flags)), ((PyObject *)Py_TYPE(Py_True)), Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1444, __pyx_L1_error)
-    __pyx_t_1 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_1 < 0)) __PYX_ERR(0, 1444, __pyx_L1_error)
-    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    if (__pyx_t_1) {
+    __pyx_t_4 = PyObject_RichCompare(((PyObject *)Py_TYPE(__pyx_v_nonref_flags)), ((PyObject *)Py_TYPE(Py_True)), Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1574, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_2 < 0)) __PYX_ERR(0, 1574, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    if (__pyx_t_2) {
 
-      /* "pgenlib.pyx":1445
+      /* "pgenlib.pyx":1575
  *         if nonref_flags is not None:
  *             if type(nonref_flags) == type(True):
  *                 if nonref_flags:             # <<<<<<<<<<<<<<
  *                     nonref_flags_storage = 2
  *                 else:
  */
-      __pyx_t_1 = __Pyx_PyObject_IsTrue(__pyx_v_nonref_flags); if (unlikely(__pyx_t_1 < 0)) __PYX_ERR(0, 1445, __pyx_L1_error)
-      if (__pyx_t_1) {
+      __pyx_t_2 = __Pyx_PyObject_IsTrue(__pyx_v_nonref_flags); if (unlikely(__pyx_t_2 < 0)) __PYX_ERR(0, 1575, __pyx_L1_error)
+      if (__pyx_t_2) {
 
-        /* "pgenlib.pyx":1446
+        /* "pgenlib.pyx":1576
  *             if type(nonref_flags) == type(True):
  *                 if nonref_flags:
  *                     nonref_flags_storage = 2             # <<<<<<<<<<<<<<
  *                 else:
  *                     nonref_flags_storage = 1
  */
         __pyx_v_nonref_flags_storage = 2;
 
-        /* "pgenlib.pyx":1445
+        /* "pgenlib.pyx":1575
  *         if nonref_flags is not None:
  *             if type(nonref_flags) == type(True):
  *                 if nonref_flags:             # <<<<<<<<<<<<<<
  *                     nonref_flags_storage = 2
  *                 else:
  */
-        goto __pyx_L13;
+        goto __pyx_L17;
       }
 
-      /* "pgenlib.pyx":1448
+      /* "pgenlib.pyx":1578
  *                     nonref_flags_storage = 2
  *                 else:
  *                     nonref_flags_storage = 1             # <<<<<<<<<<<<<<
  *             else:
  *                 nonref_flags_storage = 3
  */
       /*else*/ {
         __pyx_v_nonref_flags_storage = 1;
       }
-      __pyx_L13:;
+      __pyx_L17:;
 
-      /* "pgenlib.pyx":1444
+      /* "pgenlib.pyx":1574
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
  *         if nonref_flags is not None:
  *             if type(nonref_flags) == type(True):             # <<<<<<<<<<<<<<
  *                 if nonref_flags:
  *                     nonref_flags_storage = 2
  */
-      goto __pyx_L12;
+      goto __pyx_L16;
     }
 
-    /* "pgenlib.pyx":1450
+    /* "pgenlib.pyx":1580
  *                     nonref_flags_storage = 1
  *             else:
  *                 nonref_flags_storage = 3             # <<<<<<<<<<<<<<
  *                 if cachealigned_malloc(bitvec_cacheline_ct * kCacheline, &(self._nonref_flags)):
  *                     raise MemoryError()
  */
     /*else*/ {
       __pyx_v_nonref_flags_storage = 3;
 
-      /* "pgenlib.pyx":1451
+      /* "pgenlib.pyx":1581
  *             else:
  *                 nonref_flags_storage = 3
  *                 if cachealigned_malloc(bitvec_cacheline_ct * kCacheline, &(self._nonref_flags)):             # <<<<<<<<<<<<<<
  *                     raise MemoryError()
  *                 bytes_to_bits_internal(nonref_flags, sample_ct, self._nonref_flags)
  */
-      __pyx_t_1 = (plink2::cachealigned_malloc((__pyx_v_bitvec_cacheline_ct * plink2::kCacheline), (&__pyx_v_self->_nonref_flags)) != 0);
-      if (unlikely(__pyx_t_1)) {
+      __pyx_t_2 = (plink2::cachealigned_malloc((__pyx_v_bitvec_cacheline_ct * plink2::kCacheline), (&__pyx_v_self->_nonref_flags)) != 0);
+      if (unlikely(__pyx_t_2)) {
 
-        /* "pgenlib.pyx":1452
+        /* "pgenlib.pyx":1582
  *                 nonref_flags_storage = 3
  *                 if cachealigned_malloc(bitvec_cacheline_ct * kCacheline, &(self._nonref_flags)):
  *                     raise MemoryError()             # <<<<<<<<<<<<<<
  *                 bytes_to_bits_internal(nonref_flags, sample_ct, self._nonref_flags)
  *         cdef const char* fname = <const char*>filename
  */
-        PyErr_NoMemory(); __PYX_ERR(0, 1452, __pyx_L1_error)
+        PyErr_NoMemory(); __PYX_ERR(0, 1582, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1451
+        /* "pgenlib.pyx":1581
  *             else:
  *                 nonref_flags_storage = 3
  *                 if cachealigned_malloc(bitvec_cacheline_ct * kCacheline, &(self._nonref_flags)):             # <<<<<<<<<<<<<<
  *                     raise MemoryError()
  *                 bytes_to_bits_internal(nonref_flags, sample_ct, self._nonref_flags)
  */
       }
 
-      /* "pgenlib.pyx":1453
+      /* "pgenlib.pyx":1583
  *                 if cachealigned_malloc(bitvec_cacheline_ct * kCacheline, &(self._nonref_flags)):
  *                     raise MemoryError()
  *                 bytes_to_bits_internal(nonref_flags, sample_ct, self._nonref_flags)             # <<<<<<<<<<<<<<
  *         cdef const char* fname = <const char*>filename
  *         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0
  */
-      if (!(likely(((__pyx_v_nonref_flags) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_nonref_flags, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 1453, __pyx_L1_error)
-      __pyx_t_3 = __pyx_f_7pgenlib_bytes_to_bits_internal(((PyArrayObject *)__pyx_v_nonref_flags), __pyx_v_sample_ct, __pyx_v_self->_nonref_flags); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1453, __pyx_L1_error)
-      __Pyx_GOTREF(__pyx_t_3);
-      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
+      if (!(likely(((__pyx_v_nonref_flags) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_nonref_flags, __pyx_ptype_5numpy_ndarray))))) __PYX_ERR(0, 1583, __pyx_L1_error)
+      __pyx_t_4 = __pyx_f_7pgenlib_bytes_to_bits_internal(((PyArrayObject *)__pyx_v_nonref_flags), __pyx_v_sample_ct, __pyx_v_self->_nonref_flags); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1583, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_4);
+      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
     }
-    __pyx_L12:;
+    __pyx_L16:;
 
-    /* "pgenlib.pyx":1443
+    /* "pgenlib.pyx":1573
  *         cdef uint32_t nonref_flags_storage = 0
  *         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
  *         if nonref_flags is not None:             # <<<<<<<<<<<<<<
  *             if type(nonref_flags) == type(True):
  *                 if nonref_flags:
  */
   }
 
-  /* "pgenlib.pyx":1454
+  /* "pgenlib.pyx":1584
  *                     raise MemoryError()
  *                 bytes_to_bits_internal(nonref_flags, sample_ct, self._nonref_flags)
  *         cdef const char* fname = <const char*>filename             # <<<<<<<<<<<<<<
  *         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0
  *         if hardcall_phase_present:
  */
   if (unlikely(__pyx_v_filename == Py_None)) {
     PyErr_SetString(PyExc_TypeError, "expected bytes, NoneType found");
-    __PYX_ERR(0, 1454, __pyx_L1_error)
+    __PYX_ERR(0, 1584, __pyx_L1_error)
   }
-  __pyx_t_9 = __Pyx_PyBytes_AsString(__pyx_v_filename); if (unlikely((!__pyx_t_9) && PyErr_Occurred())) __PYX_ERR(0, 1454, __pyx_L1_error)
-  __pyx_v_fname = ((char const *)__pyx_t_9);
+  __pyx_t_5 = __Pyx_PyBytes_AsString(__pyx_v_filename); if (unlikely((!__pyx_t_5) && PyErr_Occurred())) __PYX_ERR(0, 1584, __pyx_L1_error)
+  __pyx_v_fname = ((char const *)__pyx_t_5);
 
-  /* "pgenlib.pyx":1455
+  /* "pgenlib.pyx":1585
  *                 bytes_to_bits_internal(nonref_flags, sample_ct, self._nonref_flags)
  *         cdef const char* fname = <const char*>filename
  *         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0             # <<<<<<<<<<<<<<
  *         if hardcall_phase_present:
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
  */
   __pyx_v_phase_dosage_gflags = plink2::kfPgenGlobal0;
 
-  /* "pgenlib.pyx":1456
+  /* "pgenlib.pyx":1586
  *         cdef const char* fname = <const char*>filename
  *         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0
  *         if hardcall_phase_present:             # <<<<<<<<<<<<<<
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
  *         if dosage_present:
  */
-  __pyx_t_1 = (__pyx_v_hardcall_phase_present != 0);
-  if (__pyx_t_1) {
+  __pyx_t_2 = (__pyx_v_hardcall_phase_present != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":1457
+    /* "pgenlib.pyx":1587
  *         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0
  *         if hardcall_phase_present:
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent             # <<<<<<<<<<<<<<
  *         if dosage_present:
  *             phase_dosage_gflags |= kfPgenGlobalDosagePresent
  */
     __pyx_v_phase_dosage_gflags = (__pyx_v_phase_dosage_gflags | plink2::kfPgenGlobalHardcallPhasePresent);
 
-    /* "pgenlib.pyx":1456
+    /* "pgenlib.pyx":1586
  *         cdef const char* fname = <const char*>filename
  *         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0
  *         if hardcall_phase_present:             # <<<<<<<<<<<<<<
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
  *         if dosage_present:
  */
   }
 
-  /* "pgenlib.pyx":1458
+  /* "pgenlib.pyx":1588
  *         if hardcall_phase_present:
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
  *         if dosage_present:             # <<<<<<<<<<<<<<
  *             phase_dosage_gflags |= kfPgenGlobalDosagePresent
  *         self._phase_dosage_gflags = phase_dosage_gflags
  */
-  __pyx_t_1 = (__pyx_v_dosage_present != 0);
-  if (__pyx_t_1) {
+  __pyx_t_2 = (__pyx_v_dosage_present != 0);
+  if (__pyx_t_2) {
 
-    /* "pgenlib.pyx":1459
+    /* "pgenlib.pyx":1589
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
  *         if dosage_present:
  *             phase_dosage_gflags |= kfPgenGlobalDosagePresent             # <<<<<<<<<<<<<<
  *         self._phase_dosage_gflags = phase_dosage_gflags
  *         assert not dosage_phase_present
  */
     __pyx_v_phase_dosage_gflags = (__pyx_v_phase_dosage_gflags | plink2::kfPgenGlobalDosagePresent);
 
-    /* "pgenlib.pyx":1458
+    /* "pgenlib.pyx":1588
  *         if hardcall_phase_present:
  *             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
  *         if dosage_present:             # <<<<<<<<<<<<<<
  *             phase_dosage_gflags |= kfPgenGlobalDosagePresent
  *         self._phase_dosage_gflags = phase_dosage_gflags
  */
   }
 
-  /* "pgenlib.pyx":1460
+  /* "pgenlib.pyx":1590
  *         if dosage_present:
  *             phase_dosage_gflags |= kfPgenGlobalDosagePresent
  *         self._phase_dosage_gflags = phase_dosage_gflags             # <<<<<<<<<<<<<<
  *         assert not dosage_phase_present
- *         cdef uintptr_t alloc_cacheline_ct
+ * 
  */
   __pyx_v_self->_phase_dosage_gflags = __pyx_v_phase_dosage_gflags;
 
-  /* "pgenlib.pyx":1461
+  /* "pgenlib.pyx":1591
  *             phase_dosage_gflags |= kfPgenGlobalDosagePresent
  *         self._phase_dosage_gflags = phase_dosage_gflags
  *         assert not dosage_phase_present             # <<<<<<<<<<<<<<
+ * 
  *         cdef uintptr_t alloc_cacheline_ct
- *         cdef uint32_t max_vrec_len
  */
   #ifndef CYTHON_WITHOUT_ASSERTIONS
   if (unlikely(!Py_OptimizeFlag)) {
     if (unlikely(!((!(__pyx_v_dosage_phase_present != 0)) != 0))) {
       PyErr_SetNone(PyExc_AssertionError);
-      __PYX_ERR(0, 1461, __pyx_L1_error)
+      __PYX_ERR(0, 1591, __pyx_L1_error)
     }
   }
   #endif
 
-  /* "pgenlib.pyx":1464
+  /* "pgenlib.pyx":1595
  *         cdef uintptr_t alloc_cacheline_ct
  *         cdef uint32_t max_vrec_len
- *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, variant_ct, sample_ct, 0, 0, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)             # <<<<<<<<<<<<<<
+ *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, cur_variant_ct_limit, sample_ct, allele_ct_limit, write_mode, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))
  */
-  __pyx_v_reterr = plink2::SpgwInitPhase1(__pyx_v_fname, NULL, __pyx_v_self->_nonref_flags, __pyx_v_variant_ct, __pyx_v_sample_ct, 0, 0, __pyx_v_phase_dosage_gflags, __pyx_v_nonref_flags_storage, __pyx_v_self->_state_ptr, (&__pyx_v_alloc_cacheline_ct), (&__pyx_v_max_vrec_len));
+  __pyx_v_reterr = plink2::SpgwInitPhase1(__pyx_v_fname, NULL, __pyx_v_self->_nonref_flags, __pyx_v_cur_variant_ct_limit, __pyx_v_sample_ct, __pyx_v_allele_ct_limit, __pyx_v_write_mode, __pyx_v_phase_dosage_gflags, __pyx_v_nonref_flags_storage, __pyx_v_self->_state_ptr, (&__pyx_v_alloc_cacheline_ct), (&__pyx_v_max_vrec_len));
 
-  /* "pgenlib.pyx":1465
+  /* "pgenlib.pyx":1596
  *         cdef uint32_t max_vrec_len
- *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, variant_ct, sample_ct, 0, 0, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
+ *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, cur_variant_ct_limit, sample_ct, allele_ct_limit, write_mode, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))
  *         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)
  */
-  __pyx_t_1 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-  if (unlikely(__pyx_t_1)) {
+  __pyx_t_2 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+  if (unlikely(__pyx_t_2)) {
 
-    /* "pgenlib.pyx":1466
- *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, variant_ct, sample_ct, 0, 0, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
+    /* "pgenlib.pyx":1597
+ *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, cur_variant_ct_limit, sample_ct, allele_ct_limit, write_mode, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)
- *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
+ *         cdef uint32_t patch_01_vals_cacheline_ct = DivUp(sample_ct * sizeof(AlleleCode), kCacheline)
  */
-    __pyx_t_3 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1466, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_3);
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_3); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1466, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_8);
-    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_kp_u_SpgwInitPhase1_error, __pyx_t_8); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1466, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_3);
-    __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_3); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 1466, __pyx_L1_error)
-    __Pyx_GOTREF(__pyx_t_8);
-    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __Pyx_Raise(__pyx_t_8, 0, 0, 0);
-    __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
-    __PYX_ERR(0, 1466, __pyx_L1_error)
+    __pyx_t_4 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1597, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_4); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1597, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_6);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __pyx_t_4 = __Pyx_PyUnicode_Concat(__pyx_kp_u_SpgwInitPhase1_error, __pyx_t_6); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 1597, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_4);
+    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
+    __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_4); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1597, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_6);
+    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
+    __Pyx_Raise(__pyx_t_6, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
+    __PYX_ERR(0, 1597, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1465
+    /* "pgenlib.pyx":1596
  *         cdef uint32_t max_vrec_len
- *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, variant_ct, sample_ct, 0, 0, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
+ *         cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, cur_variant_ct_limit, sample_ct, allele_ct_limit, write_mode, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))
  *         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)
  */
   }
 
-  /* "pgenlib.pyx":1467
+  /* "pgenlib.pyx":1598
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))
  *         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)             # <<<<<<<<<<<<<<
- *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
- *         cdef unsigned char* spgw_alloc
+ *         cdef uint32_t patch_01_vals_cacheline_ct = DivUp(sample_ct * sizeof(AlleleCode), kCacheline)
+ *         cdef uint32_t patch_10_vals_cacheline_ct = DivUp(sample_ct * 2 * sizeof(AlleleCode), kCacheline)
  */
   __pyx_v_genovec_cacheline_ct = plink2::DivUp(__pyx_v_sample_ct, plink2::kNypsPerCacheline);
 
-  /* "pgenlib.pyx":1468
+  /* "pgenlib.pyx":1599
  *             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))
  *         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)
+ *         cdef uint32_t patch_01_vals_cacheline_ct = DivUp(sample_ct * sizeof(AlleleCode), kCacheline)             # <<<<<<<<<<<<<<
+ *         cdef uint32_t patch_10_vals_cacheline_ct = DivUp(sample_ct * 2 * sizeof(AlleleCode), kCacheline)
+ *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
+ */
+  __pyx_v_patch_01_vals_cacheline_ct = plink2::DivUp((__pyx_v_sample_ct * (sizeof(plink2::AlleleCode))), plink2::kCacheline);
+
+  /* "pgenlib.pyx":1600
+ *         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)
+ *         cdef uint32_t patch_01_vals_cacheline_ct = DivUp(sample_ct * sizeof(AlleleCode), kCacheline)
+ *         cdef uint32_t patch_10_vals_cacheline_ct = DivUp(sample_ct * 2 * sizeof(AlleleCode), kCacheline)             # <<<<<<<<<<<<<<
+ *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
+ *         cdef unsigned char* spgw_alloc
+ */
+  __pyx_v_patch_10_vals_cacheline_ct = plink2::DivUp(((__pyx_v_sample_ct * 2) * (sizeof(plink2::AlleleCode))), plink2::kCacheline);
+
+  /* "pgenlib.pyx":1601
+ *         cdef uint32_t patch_01_vals_cacheline_ct = DivUp(sample_ct * sizeof(AlleleCode), kCacheline)
+ *         cdef uint32_t patch_10_vals_cacheline_ct = DivUp(sample_ct * 2 * sizeof(AlleleCode), kCacheline)
  *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))             # <<<<<<<<<<<<<<
  *         cdef unsigned char* spgw_alloc
- *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
+ *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 5 * bitvec_cacheline_ct + patch_01_vals_cacheline_ct + patch_10_vals_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
  */
   __pyx_v_dosage_main_cacheline_ct = plink2::DivUp(__pyx_v_sample_ct, (2 * plink2::kInt32PerCacheline));
 
-  /* "pgenlib.pyx":1470
+  /* "pgenlib.pyx":1603
  *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
  *         cdef unsigned char* spgw_alloc
- *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):             # <<<<<<<<<<<<<<
+ *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 5 * bitvec_cacheline_ct + patch_01_vals_cacheline_ct + patch_10_vals_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)
  */
-  __pyx_t_1 = (plink2::cachealigned_malloc(((((__pyx_v_alloc_cacheline_ct + __pyx_v_genovec_cacheline_ct) + (3 * __pyx_v_bitvec_cacheline_ct)) + __pyx_v_dosage_main_cacheline_ct) * plink2::kCacheline), (&__pyx_v_spgw_alloc)) != 0);
-  if (unlikely(__pyx_t_1)) {
+  __pyx_t_2 = (plink2::cachealigned_malloc(((((((__pyx_v_alloc_cacheline_ct + __pyx_v_genovec_cacheline_ct) + (5 * __pyx_v_bitvec_cacheline_ct)) + __pyx_v_patch_01_vals_cacheline_ct) + __pyx_v_patch_10_vals_cacheline_ct) + __pyx_v_dosage_main_cacheline_ct) * plink2::kCacheline), (&__pyx_v_spgw_alloc)) != 0);
+  if (unlikely(__pyx_t_2)) {
 
-    /* "pgenlib.pyx":1471
+    /* "pgenlib.pyx":1604
  *         cdef unsigned char* spgw_alloc
- *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
+ *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 5 * bitvec_cacheline_ct + patch_01_vals_cacheline_ct + patch_10_vals_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
  *             raise MemoryError()             # <<<<<<<<<<<<<<
  *         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)
- *         self._genovec = <uintptr_t*>(&(spgw_alloc[alloc_cacheline_ct * kCacheline]))
+ *         cdef unsigned char* spgw_alloc_iter = &(spgw_alloc[alloc_cacheline_ct * kCacheline])
  */
-    PyErr_NoMemory(); __PYX_ERR(0, 1471, __pyx_L1_error)
+    PyErr_NoMemory(); __PYX_ERR(0, 1604, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1470
+    /* "pgenlib.pyx":1603
  *         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
  *         cdef unsigned char* spgw_alloc
- *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):             # <<<<<<<<<<<<<<
+ *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 5 * bitvec_cacheline_ct + patch_01_vals_cacheline_ct + patch_10_vals_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):             # <<<<<<<<<<<<<<
  *             raise MemoryError()
  *         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)
  */
   }
 
-  /* "pgenlib.pyx":1472
- *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
+  /* "pgenlib.pyx":1605
+ *         if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 5 * bitvec_cacheline_ct + patch_01_vals_cacheline_ct + patch_10_vals_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
  *             raise MemoryError()
  *         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)             # <<<<<<<<<<<<<<
- *         self._genovec = <uintptr_t*>(&(spgw_alloc[alloc_cacheline_ct * kCacheline]))
- *         self._phasepresent = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct) * kCacheline]))
+ *         cdef unsigned char* spgw_alloc_iter = &(spgw_alloc[alloc_cacheline_ct * kCacheline])
+ *         self._allele_ct_limit = allele_ct_limit
  */
   plink2::SpgwInitPhase2(__pyx_v_max_vrec_len, __pyx_v_self->_state_ptr, __pyx_v_spgw_alloc);
 
-  /* "pgenlib.pyx":1473
+  /* "pgenlib.pyx":1606
  *             raise MemoryError()
  *         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)
- *         self._genovec = <uintptr_t*>(&(spgw_alloc[alloc_cacheline_ct * kCacheline]))             # <<<<<<<<<<<<<<
- *         self._phasepresent = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct) * kCacheline]))
- *         self._phaseinfo = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + bitvec_cacheline_ct) * kCacheline]))
+ *         cdef unsigned char* spgw_alloc_iter = &(spgw_alloc[alloc_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._allele_ct_limit = allele_ct_limit
+ *         self._genovec = <uintptr_t*>(spgw_alloc_iter)
  */
-  __pyx_v_self->_genovec = ((uintptr_t *)(&(__pyx_v_spgw_alloc[(__pyx_v_alloc_cacheline_ct * plink2::kCacheline)])));
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc[(__pyx_v_alloc_cacheline_ct * plink2::kCacheline)]));
 
-  /* "pgenlib.pyx":1474
+  /* "pgenlib.pyx":1607
  *         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)
- *         self._genovec = <uintptr_t*>(&(spgw_alloc[alloc_cacheline_ct * kCacheline]))
- *         self._phasepresent = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct) * kCacheline]))             # <<<<<<<<<<<<<<
- *         self._phaseinfo = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + bitvec_cacheline_ct) * kCacheline]))
- *         self._dosage_present = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 2 * bitvec_cacheline_ct) * kCacheline]))
+ *         cdef unsigned char* spgw_alloc_iter = &(spgw_alloc[alloc_cacheline_ct * kCacheline])
+ *         self._allele_ct_limit = allele_ct_limit             # <<<<<<<<<<<<<<
+ *         self._genovec = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[genovec_cacheline_ct * kCacheline])
  */
-  __pyx_v_self->_phasepresent = ((uintptr_t *)(&(__pyx_v_spgw_alloc[((__pyx_v_alloc_cacheline_ct + __pyx_v_genovec_cacheline_ct) * plink2::kCacheline)])));
+  __pyx_v_self->_allele_ct_limit = __pyx_v_allele_ct_limit;
 
-  /* "pgenlib.pyx":1475
- *         self._genovec = <uintptr_t*>(&(spgw_alloc[alloc_cacheline_ct * kCacheline]))
- *         self._phasepresent = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct) * kCacheline]))
- *         self._phaseinfo = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + bitvec_cacheline_ct) * kCacheline]))             # <<<<<<<<<<<<<<
- *         self._dosage_present = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 2 * bitvec_cacheline_ct) * kCacheline]))
- *         self._dosage_main = <uint16_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct) * kCacheline]))
+  /* "pgenlib.pyx":1608
+ *         cdef unsigned char* spgw_alloc_iter = &(spgw_alloc[alloc_cacheline_ct * kCacheline])
+ *         self._allele_ct_limit = allele_ct_limit
+ *         self._genovec = <uintptr_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[genovec_cacheline_ct * kCacheline])
+ * 
  */
-  __pyx_v_self->_phaseinfo = ((uintptr_t *)(&(__pyx_v_spgw_alloc[(((__pyx_v_alloc_cacheline_ct + __pyx_v_genovec_cacheline_ct) + __pyx_v_bitvec_cacheline_ct) * plink2::kCacheline)])));
+  __pyx_v_self->_genovec = ((uintptr_t *)__pyx_v_spgw_alloc_iter);
 
-  /* "pgenlib.pyx":1476
- *         self._phasepresent = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct) * kCacheline]))
- *         self._phaseinfo = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + bitvec_cacheline_ct) * kCacheline]))
- *         self._dosage_present = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 2 * bitvec_cacheline_ct) * kCacheline]))             # <<<<<<<<<<<<<<
- *         self._dosage_main = <uint16_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct) * kCacheline]))
- *         return
+  /* "pgenlib.pyx":1609
+ *         self._allele_ct_limit = allele_ct_limit
+ *         self._genovec = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[genovec_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ * 
+ *         # Can't skimp on patch_{01,10}_{set,vals} allocations even when
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_genovec_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1615
+ *         # works.
+ *         # Could skimp on dosage/phase, but that doesn't gain us much.
+ *         self._patch_01_set = <uintptr_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._patch_01_vals = <AlleleCode*>(spgw_alloc_iter)
+ */
+  __pyx_v_self->_patch_01_set = ((uintptr_t *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1616
+ *         # Could skimp on dosage/phase, but that doesn't gain us much.
+ *         self._patch_01_set = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._patch_01_vals = <AlleleCode*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_01_vals_cacheline_ct * kCacheline])
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_bitvec_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1617
+ *         self._patch_01_set = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._patch_01_vals = <AlleleCode*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_01_vals_cacheline_ct * kCacheline])
+ *         self._patch_10_set = <uintptr_t*>(spgw_alloc_iter)
+ */
+  __pyx_v_self->_patch_01_vals = ((plink2::AlleleCode *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1618
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._patch_01_vals = <AlleleCode*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_01_vals_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._patch_10_set = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_patch_01_vals_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1619
+ *         self._patch_01_vals = <AlleleCode*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_01_vals_cacheline_ct * kCacheline])
+ *         self._patch_10_set = <uintptr_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._patch_10_vals = <AlleleCode*>(spgw_alloc_iter)
  */
-  __pyx_v_self->_dosage_present = ((uintptr_t *)(&(__pyx_v_spgw_alloc[(((__pyx_v_alloc_cacheline_ct + __pyx_v_genovec_cacheline_ct) + (2 * __pyx_v_bitvec_cacheline_ct)) * plink2::kCacheline)])));
+  __pyx_v_self->_patch_10_set = ((uintptr_t *)__pyx_v_spgw_alloc_iter);
 
-  /* "pgenlib.pyx":1477
- *         self._phaseinfo = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + bitvec_cacheline_ct) * kCacheline]))
- *         self._dosage_present = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 2 * bitvec_cacheline_ct) * kCacheline]))
- *         self._dosage_main = <uint16_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct) * kCacheline]))             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1620
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_01_vals_cacheline_ct * kCacheline])
+ *         self._patch_10_set = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._patch_10_vals = <AlleleCode*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_10_vals_cacheline_ct * kCacheline])
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_bitvec_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1621
+ *         self._patch_10_set = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._patch_10_vals = <AlleleCode*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_10_vals_cacheline_ct * kCacheline])
+ *         self._phasepresent = <uintptr_t*>(spgw_alloc_iter)
+ */
+  __pyx_v_self->_patch_10_vals = ((plink2::AlleleCode *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1622
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._patch_10_vals = <AlleleCode*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_10_vals_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._phasepresent = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_patch_10_vals_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1623
+ *         self._patch_10_vals = <AlleleCode*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_10_vals_cacheline_ct * kCacheline])
+ *         self._phasepresent = <uintptr_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._phaseinfo = <uintptr_t*>(spgw_alloc_iter)
+ */
+  __pyx_v_self->_phasepresent = ((uintptr_t *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1624
+ *         spgw_alloc_iter = &(spgw_alloc_iter[patch_10_vals_cacheline_ct * kCacheline])
+ *         self._phasepresent = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._phaseinfo = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_bitvec_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1625
+ *         self._phasepresent = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._phaseinfo = <uintptr_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._dosage_present = <uintptr_t*>(spgw_alloc_iter)
+ */
+  __pyx_v_self->_phaseinfo = ((uintptr_t *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1626
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._phaseinfo = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._dosage_present = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_bitvec_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1627
+ *         self._phaseinfo = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._dosage_present = <uintptr_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._dosage_main = <uint16_t*>(spgw_alloc_iter)
+ */
+  __pyx_v_self->_dosage_present = ((uintptr_t *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1628
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._dosage_present = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])             # <<<<<<<<<<<<<<
+ *         self._dosage_main = <uint16_t*>(spgw_alloc_iter)
+ *         # bugfix (16 Apr 2023): SpgwAppendBiallelicGenovec[Hphase] assumes
+ */
+  __pyx_v_spgw_alloc_iter = (&(__pyx_v_spgw_alloc_iter[(__pyx_v_bitvec_cacheline_ct * plink2::kCacheline)]));
+
+  /* "pgenlib.pyx":1629
+ *         self._dosage_present = <uintptr_t*>(spgw_alloc_iter)
+ *         spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+ *         self._dosage_main = <uint16_t*>(spgw_alloc_iter)             # <<<<<<<<<<<<<<
+ *         # bugfix (16 Apr 2023): SpgwAppendBiallelicGenovec[Hphase] assumes
+ *         # trailing bits are clear
+ */
+  __pyx_v_self->_dosage_main = ((uint16_t *)__pyx_v_spgw_alloc_iter);
+
+  /* "pgenlib.pyx":1632
+ *         # bugfix (16 Apr 2023): SpgwAppendBiallelicGenovec[Hphase] assumes
+ *         # trailing bits are clear
+ *         self._genovec[(sample_ct - 1) // kBitsPerWordD2] = 0             # <<<<<<<<<<<<<<
+ *         self._phasepresent[(sample_ct - 1) // kBitsPerWord] = 0
+ *         return
+ */
+  __pyx_t_7 = (__pyx_v_sample_ct - 1);
+  if (unlikely(plink2::kBitsPerWordD2 == 0)) {
+    PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
+    __PYX_ERR(0, 1632, __pyx_L1_error)
+  }
+  else if (sizeof(long) == sizeof(long) && (!(((int)-1) > 0)) && unlikely(plink2::kBitsPerWordD2 == (int)-1)  && unlikely(UNARY_NEG_WOULD_OVERFLOW(__pyx_t_7))) {
+    PyErr_SetString(PyExc_OverflowError, "value too large to perform division");
+    __PYX_ERR(0, 1632, __pyx_L1_error)
+  }
+  (__pyx_v_self->_genovec[__Pyx_div_long(__pyx_t_7, plink2::kBitsPerWordD2)]) = 0;
+
+  /* "pgenlib.pyx":1633
+ *         # trailing bits are clear
+ *         self._genovec[(sample_ct - 1) // kBitsPerWordD2] = 0
+ *         self._phasepresent[(sample_ct - 1) // kBitsPerWord] = 0             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_v_self->_dosage_main = ((uint16_t *)(&(__pyx_v_spgw_alloc[(((__pyx_v_alloc_cacheline_ct + __pyx_v_genovec_cacheline_ct) + (3 * __pyx_v_bitvec_cacheline_ct)) * plink2::kCacheline)])));
+  __pyx_t_7 = (__pyx_v_sample_ct - 1);
+  if (unlikely(plink2::kBitsPerWord == 0)) {
+    PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
+    __PYX_ERR(0, 1633, __pyx_L1_error)
+  }
+  else if (sizeof(long) == sizeof(long) && (!(((int)-1) > 0)) && unlikely(plink2::kBitsPerWord == (int)-1)  && unlikely(UNARY_NEG_WOULD_OVERFLOW(__pyx_t_7))) {
+    PyErr_SetString(PyExc_OverflowError, "value too large to perform division");
+    __PYX_ERR(0, 1633, __pyx_L1_error)
+  }
+  (__pyx_v_self->_phasepresent[__Pyx_div_long(__pyx_t_7, plink2::kBitsPerWord)]) = 0;
 
-  /* "pgenlib.pyx":1478
- *         self._dosage_present = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 2 * bitvec_cacheline_ct) * kCacheline]))
- *         self._dosage_main = <uint16_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct) * kCacheline]))
+  /* "pgenlib.pyx":1634
+ *         self._genovec[(sample_ct - 1) // kBitsPerWordD2] = 0
+ *         self._phasepresent[(sample_ct - 1) // kBitsPerWord] = 0
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __pyx_r = 0;
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1424
- * 
+  /* "pgenlib.pyx":1540
+ *     cdef uint16_t* _dosage_main
  * 
  *     def __cinit__(self, bytes filename, uint32_t sample_ct,             # <<<<<<<<<<<<<<
- *                   uint32_t variant_ct, object nonref_flags,
- *                   object allele_idx_offsets = None,
+ *                   object variant_ct = None, object nonref_flags = True,
+ *                   uint32_t allele_ct_limit = 2,
  */
 
   /* function exit code */
   __pyx_L1_error:;
-  __Pyx_XDECREF(__pyx_t_3);
   __Pyx_XDECREF(__pyx_t_4);
-  __Pyx_XDECREF(__pyx_t_7);
-  __Pyx_XDECREF(__pyx_t_8);
+  __Pyx_XDECREF(__pyx_t_6);
   __Pyx_AddTraceback("pgenlib.PgenWriter.__cinit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = -1;
   __pyx_L0:;
-  __Pyx_XDECREF(__pyx_v_uii);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1481
+/* "pgenlib.pyx":1637
  * 
  * 
  *     cpdef __enter__(self):             # <<<<<<<<<<<<<<
  *         return self
  * 
  */
 
@@ -24389,15 +25930,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_enter); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1481, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_enter); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1637, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_3__enter__)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -24406,15 +25947,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1481, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1637, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -24427,27 +25968,27 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1482
+  /* "pgenlib.pyx":1638
  * 
  *     cpdef __enter__(self):
  *         return self             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __Pyx_INCREF(((PyObject *)__pyx_v_self));
   __pyx_r = ((PyObject *)__pyx_v_self);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1481
+  /* "pgenlib.pyx":1637
  * 
  * 
  *     cpdef __enter__(self):             # <<<<<<<<<<<<<<
  *         return self
  * 
  */
 
@@ -24483,15 +26024,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__enter__", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter___enter__(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1481, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter___enter__(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1637, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -24500,15 +26041,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1485
+/* "pgenlib.pyx":1641
  * 
  * 
  *     cpdef append_biallelic(self, np.ndarray[np.int8_t,mode="c"] geno_int8):             # <<<<<<<<<<<<<<
  *         cdef int8_t* genobytes = &(geno_int8[0])
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  */
 
@@ -24533,27 +26074,27 @@
   __Pyx_RefNannySetupContext("append_biallelic", 0);
   __pyx_pybuffer_geno_int8.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int8.refcount = 0;
   __pyx_pybuffernd_geno_int8.data = NULL;
   __pyx_pybuffernd_geno_int8.rcbuffer = &__pyx_pybuffer_geno_int8;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1485, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1641, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int8.diminfo[0].strides = __pyx_pybuffernd_geno_int8.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int8.diminfo[0].shape = __pyx_pybuffernd_geno_int8.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_biallelic); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1485, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_biallelic); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1641, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_5append_biallelic)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -24562,15 +26103,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_Call2Args(__pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_geno_int8)) : __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_geno_int8));
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1485, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1641, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -24583,104 +26124,104 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1486
+  /* "pgenlib.pyx":1642
  * 
  *     cpdef append_biallelic(self, np.ndarray[np.int8_t,mode="c"] geno_int8):
  *         cdef int8_t* genobytes = &(geno_int8[0])             # <<<<<<<<<<<<<<
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  */
   __pyx_t_5 = 0;
   __pyx_t_6 = -1;
   if (__pyx_t_5 < 0) {
     __pyx_t_5 += __pyx_pybuffernd_geno_int8.diminfo[0].shape;
     if (unlikely(__pyx_t_5 < 0)) __pyx_t_6 = 0;
   } else if (unlikely(__pyx_t_5 >= __pyx_pybuffernd_geno_int8.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 1486, __pyx_L1_error)
+    __PYX_ERR(0, 1642, __pyx_L1_error)
   }
   __pyx_v_genobytes = (&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_int8_t *, __pyx_pybuffernd_geno_int8.rcbuffer->pybuffer.buf, __pyx_t_5, __pyx_pybuffernd_geno_int8.diminfo[0].strides)));
 
-  /* "pgenlib.pyx":1487
+  /* "pgenlib.pyx":1643
  *     cpdef append_biallelic(self, np.ndarray[np.int8_t,mode="c"] geno_int8):
  *         cdef int8_t* genobytes = &(geno_int8[0])
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  */
   plink2::BytesToGenoarrUnsafe(__pyx_v_genobytes, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_self->_genovec);
 
-  /* "pgenlib.pyx":1488
+  /* "pgenlib.pyx":1644
  *         cdef int8_t* genobytes = &(geno_int8[0])
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_biallelic() error " + str(reterr))
  */
   __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovec(__pyx_v_self->_genovec, __pyx_v_self->_state_ptr);
 
-  /* "pgenlib.pyx":1489
+  /* "pgenlib.pyx":1645
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_biallelic() error " + str(reterr))
  *         return
  */
   __pyx_t_7 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_7)) {
 
-    /* "pgenlib.pyx":1490
+    /* "pgenlib.pyx":1646
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_biallelic() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1490, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1646, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1490, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1646, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_biallelic_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1490, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_biallelic_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1646, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1490, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1646, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 1490, __pyx_L1_error)
+    __PYX_ERR(0, 1646, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1489
+    /* "pgenlib.pyx":1645
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_biallelic() error " + str(reterr))
  *         return
  */
   }
 
-  /* "pgenlib.pyx":1491
+  /* "pgenlib.pyx":1647
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_biallelic() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1485
+  /* "pgenlib.pyx":1641
  * 
  * 
  *     cpdef append_biallelic(self, np.ndarray[np.int8_t,mode="c"] geno_int8):             # <<<<<<<<<<<<<<
  *         cdef int8_t* genobytes = &(geno_int8[0])
  *         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  */
 
@@ -24712,15 +26253,15 @@
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_5append_biallelic(PyObject *__pyx_v_self, PyObject *__pyx_v_geno_int8) {
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_biallelic (wrapper)", 0);
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int8), __pyx_ptype_5numpy_ndarray, 1, "geno_int8", 0))) __PYX_ERR(0, 1485, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int8), __pyx_ptype_5numpy_ndarray, 1, "geno_int8", 0))) __PYX_ERR(0, 1641, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_4append_biallelic(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), ((PyArrayObject *)__pyx_v_geno_int8));
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -24740,19 +26281,19 @@
   __Pyx_RefNannySetupContext("append_biallelic", 0);
   __pyx_pybuffer_geno_int8.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int8.refcount = 0;
   __pyx_pybuffernd_geno_int8.data = NULL;
   __pyx_pybuffernd_geno_int8.rcbuffer = &__pyx_pybuffer_geno_int8;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1485, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1641, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int8.diminfo[0].strides = __pyx_pybuffernd_geno_int8.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int8.diminfo[0].shape = __pyx_pybuffernd_geno_int8.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_biallelic(__pyx_v_self, __pyx_v_geno_int8, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1485, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_biallelic(__pyx_v_self, __pyx_v_geno_int8, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1641, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -24770,74 +26311,92 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_geno_int8.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1494
+/* "pgenlib.pyx":1650
  * 
  * 
- *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False):             # <<<<<<<<<<<<<<
+ *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False, object allele_ct = None):             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_7append_alleles(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_alleles(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles *__pyx_optional_args) {
   int __pyx_v_all_phased = ((int)0);
+  PyObject *__pyx_v_allele_ct = ((PyObject *)Py_None);
   int32_t *__pyx_v_allele_codes;
+  uint32_t __pyx_v_sample_ct;
+  uint32_t __pyx_v_allele_ct_limit;
   uintptr_t *__pyx_v_genovec;
+  uintptr_t *__pyx_v_patch_01_set;
+  plink2::AlleleCode *__pyx_v_patch_01_vals;
+  uintptr_t *__pyx_v_patch_10_set;
+  plink2::AlleleCode *__pyx_v_patch_10_vals;
+  uintptr_t *__pyx_v_phaseinfo;
+  uint32_t __pyx_v_patch_01_ct;
+  uint32_t __pyx_v_patch_10_ct;
+  int32_t __pyx_v_observed_allele_ct;
+  uint32_t __pyx_v_write_allele_ct;
   plink2::PglErr __pyx_v_reterr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32;
   __Pyx_Buffer __pyx_pybuffer_allele_int32;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   PyObject *__pyx_t_2 = NULL;
   PyObject *__pyx_t_3 = NULL;
   PyObject *__pyx_t_4 = NULL;
   PyObject *__pyx_t_5 = NULL;
   int __pyx_t_6;
   PyObject *__pyx_t_7 = NULL;
   Py_ssize_t __pyx_t_8;
-  uintptr_t *__pyx_t_9;
-  int __pyx_t_10;
+  uint32_t __pyx_t_9;
+  uintptr_t *__pyx_t_10;
+  plink2::AlleleCode *__pyx_t_11;
+  int __pyx_t_12;
+  int __pyx_t_13;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_alleles", 0);
   if (__pyx_optional_args) {
     if (__pyx_optional_args->__pyx_n > 0) {
       __pyx_v_all_phased = __pyx_optional_args->all_phased;
+      if (__pyx_optional_args->__pyx_n > 1) {
+        __pyx_v_allele_ct = __pyx_optional_args->allele_ct;
+      }
     }
   }
   __pyx_pybuffer_allele_int32.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32.refcount = 0;
   __pyx_pybuffernd_allele_int32.data = NULL;
   __pyx_pybuffernd_allele_int32.rcbuffer = &__pyx_pybuffer_allele_int32;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1494, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1650, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32.diminfo[0].strides = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32.diminfo[0].shape = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_alleles); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1494, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_alleles); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1650, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_7append_alleles)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_all_phased); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1494, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_all_phased); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1650, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -24846,43 +26405,46 @@
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1494, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32), __pyx_t_3, __pyx_v_allele_ct};
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1650, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1494, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32), __pyx_t_3, __pyx_v_allele_ct};
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1650, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(2+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1494, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1650, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32));
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, ((PyObject *)__pyx_v_allele_int32));
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, __pyx_t_3);
+          __Pyx_INCREF(__pyx_v_allele_ct);
+          __Pyx_GIVEREF(__pyx_v_allele_ct);
+          PyTuple_SET_ITEM(__pyx_t_7, 2+__pyx_t_6, __pyx_v_allele_ct);
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1494, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1650, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -24897,192 +26459,537 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1495
+  /* "pgenlib.pyx":1651
  * 
- *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False):
+ *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False, object allele_ct = None):
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef PglErr reterr
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  */
   __pyx_t_8 = 0;
   __pyx_t_6 = -1;
   if (__pyx_t_8 < 0) {
     __pyx_t_8 += __pyx_pybuffernd_allele_int32.diminfo[0].shape;
     if (unlikely(__pyx_t_8 < 0)) __pyx_t_6 = 0;
   } else if (unlikely(__pyx_t_8 >= __pyx_pybuffernd_allele_int32.diminfo[0].shape)) __pyx_t_6 = 0;
   if (unlikely(__pyx_t_6 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_6);
-    __PYX_ERR(0, 1495, __pyx_L1_error)
+    __PYX_ERR(0, 1651, __pyx_L1_error)
   }
   __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.buf, __pyx_t_8, __pyx_pybuffernd_allele_int32.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":1496
- *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False):
+  /* "pgenlib.pyx":1652
+ *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False, object allele_ct = None):
+ *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)             # <<<<<<<<<<<<<<
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
+ *         cdef uintptr_t* genovec = self._genovec
+ */
+  __pyx_v_sample_ct = plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr);
+
+  /* "pgenlib.pyx":1653
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ */
+  __pyx_t_9 = __pyx_v_self->_allele_ct_limit;
+  __pyx_v_allele_ct_limit = __pyx_t_9;
+
+  /* "pgenlib.pyx":1654
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
- *         cdef PglErr reterr
- *         if not all_phased:
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
  */
-  __pyx_t_9 = __pyx_v_self->_genovec;
-  __pyx_v_genovec = __pyx_t_9;
+  __pyx_t_10 = __pyx_v_self->_genovec;
+  __pyx_v_genovec = __pyx_t_10;
 
-  /* "pgenlib.pyx":1498
+  /* "pgenlib.pyx":1655
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  *         cdef uintptr_t* genovec = self._genovec
- *         cdef PglErr reterr
- *         if not all_phased:             # <<<<<<<<<<<<<<
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *             reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ */
+  __pyx_t_10 = __pyx_v_self->_patch_01_set;
+  __pyx_v_patch_01_set = __pyx_t_10;
+
+  /* "pgenlib.pyx":1656
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ */
+  __pyx_t_11 = __pyx_v_self->_patch_01_vals;
+  __pyx_v_patch_01_vals = __pyx_t_11;
+
+  /* "pgenlib.pyx":1657
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ *         cdef uintptr_t* phaseinfo = NULL
  */
-  __pyx_t_10 = ((!(__pyx_v_all_phased != 0)) != 0);
-  if (__pyx_t_10) {
+  __pyx_t_10 = __pyx_v_self->_patch_10_set;
+  __pyx_v_patch_10_set = __pyx_t_10;
 
-    /* "pgenlib.pyx":1499
- *         cdef PglErr reterr
- *         if not all_phased:
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)             # <<<<<<<<<<<<<<
- *             reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
- *         else:
+  /* "pgenlib.pyx":1658
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* phaseinfo = NULL
+ *         if all_phased:
  */
-    plink2::AlleleCodesToGenoarrUnsafe(__pyx_v_allele_codes, NULL, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_genovec, NULL, NULL);
+  __pyx_t_11 = __pyx_v_self->_patch_10_vals;
+  __pyx_v_patch_10_vals = __pyx_t_11;
 
-    /* "pgenlib.pyx":1500
- *         if not all_phased:
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *             reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)             # <<<<<<<<<<<<<<
- *         else:
+  /* "pgenlib.pyx":1659
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ *         cdef uintptr_t* phaseinfo = NULL             # <<<<<<<<<<<<<<
+ *         if all_phased:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  */
-    __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovec(__pyx_v_genovec, __pyx_v_self->_state_ptr);
+  __pyx_v_phaseinfo = NULL;
 
-    /* "pgenlib.pyx":1498
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef PglErr reterr
- *         if not all_phased:             # <<<<<<<<<<<<<<
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *             reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+  /* "pgenlib.pyx":1660
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ *         cdef uintptr_t* phaseinfo = NULL
+ *         if all_phased:             # <<<<<<<<<<<<<<
+ *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
+ *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
  */
-    goto __pyx_L3;
-  }
+  __pyx_t_12 = (__pyx_v_all_phased != 0);
+  if (__pyx_t_12) {
 
-  /* "pgenlib.pyx":1502
- *             reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
- *         else:
+    /* "pgenlib.pyx":1661
+ *         cdef uintptr_t* phaseinfo = NULL
+ *         if all_phased:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
+ *             phaseinfo = self._phaseinfo
  */
-  /*else*/ {
-    __pyx_t_10 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalHardcallPhasePresent) == 0) != 0);
-    if (unlikely(__pyx_t_10)) {
+    __pyx_t_12 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalHardcallPhasePresent) == 0) != 0);
+    if (unlikely(__pyx_t_12)) {
 
-      /* "pgenlib.pyx":1503
- *         else:
+      /* "pgenlib.pyx":1662
+ *         if all_phased:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ *             phaseinfo = self._phaseinfo
+ *         cdef uint32_t patch_01_ct
  */
-      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__23, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1503, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__28, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1662, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1503, __pyx_L1_error)
+      __PYX_ERR(0, 1662, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1502
- *             reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
- *         else:
+      /* "pgenlib.pyx":1661
+ *         cdef uintptr_t* phaseinfo = NULL
+ *         if all_phased:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
+ *             phaseinfo = self._phaseinfo
  */
     }
 
-    /* "pgenlib.pyx":1504
+    /* "pgenlib.pyx":1663
+ *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
+ *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
+ *             phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
+ *         cdef uint32_t patch_01_ct
+ *         cdef uint32_t patch_10_ct
+ */
+    __pyx_t_10 = __pyx_v_self->_phaseinfo;
+    __pyx_v_phaseinfo = __pyx_t_10;
+
+    /* "pgenlib.pyx":1660
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ *         cdef uintptr_t* phaseinfo = NULL
+ *         if all_phased:             # <<<<<<<<<<<<<<
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)             # <<<<<<<<<<<<<<
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
- *         if reterr != kPglRetSuccess:
  */
-    plink2::AlleleCodesToGenoarrUnsafe(__pyx_v_allele_codes, NULL, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_genovec, NULL, __pyx_v_self->_phaseinfo);
+  }
 
-    /* "pgenlib.pyx":1505
- *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1666
+ *         cdef uint32_t patch_01_ct
+ *         cdef uint32_t patch_10_ct
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, phaseinfo)             # <<<<<<<<<<<<<<
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_alleles called with invalid allele codes")
+ */
+  __pyx_v_observed_allele_ct = plink2::ConvertMultiAlleleCodesUnsafe(__pyx_v_allele_codes, NULL, __pyx_v_sample_ct, __pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, (&__pyx_v_patch_01_ct), (&__pyx_v_patch_10_ct), NULL, __pyx_v_phaseinfo);
+
+  /* "pgenlib.pyx":1667
+ *         cdef uint32_t patch_10_ct
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, phaseinfo)
+ *         if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_alleles called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+  __pyx_t_12 = ((__pyx_v_observed_allele_ct == -1L) != 0);
+  if (unlikely(__pyx_t_12)) {
+
+    /* "pgenlib.pyx":1668
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, phaseinfo)
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_alleles called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ */
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__29, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1668, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_1);
+    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    __PYX_ERR(0, 1668, __pyx_L1_error)
+
+    /* "pgenlib.pyx":1667
+ *         cdef uint32_t patch_10_ct
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, phaseinfo)
+ *         if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_alleles called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+  }
+
+  /* "pgenlib.pyx":1669
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_alleles called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)             # <<<<<<<<<<<<<<
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ */
+  __pyx_v_write_allele_ct = ((uint32_t)__pyx_v_observed_allele_ct);
+
+  /* "pgenlib.pyx":1670
+ *             raise RuntimeError("append_alleles called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ */
+  __pyx_t_12 = ((__pyx_v_write_allele_ct > __pyx_v_allele_ct_limit) != 0);
+  if (unlikely(__pyx_t_12)) {
+
+    /* "pgenlib.pyx":1671
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ */
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__30, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1671, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_1);
+    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    __PYX_ERR(0, 1671, __pyx_L1_error)
+
+    /* "pgenlib.pyx":1670
+ *             raise RuntimeError("append_alleles called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ */
+  }
+
+  /* "pgenlib.pyx":1672
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:             # <<<<<<<<<<<<<<
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ */
+  __pyx_t_12 = (__pyx_v_allele_ct != Py_None);
+  __pyx_t_13 = (__pyx_t_12 != 0);
+  if (__pyx_t_13) {
+
+    /* "pgenlib.pyx":1673
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ */
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_write_allele_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1673, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_1);
+    __pyx_t_2 = PyObject_RichCompare(__pyx_v_allele_ct, __pyx_t_1, Py_LT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1673, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_13 < 0)) __PYX_ERR(0, 1673, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+    if (unlikely(__pyx_t_13)) {
+
+      /* "pgenlib.pyx":1674
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")             # <<<<<<<<<<<<<<
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")
+ */
+      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__31, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1674, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_2);
+      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+      __PYX_ERR(0, 1674, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1673
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ */
+    }
+
+    /* "pgenlib.pyx":1675
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")
+ *             write_allele_ct = allele_ct
+ */
+    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_ct_limit); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1675, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_2);
+    __pyx_t_1 = PyObject_RichCompare(__pyx_v_allele_ct, __pyx_t_2, Py_GT); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1675, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+    __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_13 < 0)) __PYX_ERR(0, 1675, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    if (unlikely(__pyx_t_13)) {
+
+      /* "pgenlib.pyx":1676
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ */
+      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__32, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1676, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __PYX_ERR(0, 1676, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1675
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")
+ *             write_allele_ct = allele_ct
+ */
+    }
+
+    /* "pgenlib.pyx":1677
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")
+ *             write_allele_ct = allele_ct             # <<<<<<<<<<<<<<
+ *         cdef PglErr reterr
+ *         if not all_phased:
+ */
+    __pyx_t_9 = __Pyx_PyInt_As_uint32_t(__pyx_v_allele_ct); if (unlikely((__pyx_t_9 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1677, __pyx_L1_error)
+    __pyx_v_write_allele_ct = __pyx_t_9;
+
+    /* "pgenlib.pyx":1672
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:             # <<<<<<<<<<<<<<
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ */
+  }
+
+  /* "pgenlib.pyx":1679
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ *         if not all_phased:             # <<<<<<<<<<<<<<
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ */
+  __pyx_t_13 = ((!(__pyx_v_all_phased != 0)) != 0);
+  if (__pyx_t_13) {
+
+    /* "pgenlib.pyx":1680
+ *         cdef PglErr reterr
+ *         if not all_phased:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *             else:
+ */
+    __pyx_t_12 = ((__pyx_v_patch_01_ct == 0) != 0);
+    if (__pyx_t_12) {
+    } else {
+      __pyx_t_13 = __pyx_t_12;
+      goto __pyx_L12_bool_binop_done;
+    }
+    __pyx_t_12 = ((__pyx_v_patch_10_ct == 0) != 0);
+    __pyx_t_13 = __pyx_t_12;
+    __pyx_L12_bool_binop_done:;
+    if (__pyx_t_13) {
+
+      /* "pgenlib.pyx":1681
+ *         if not all_phased:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)             # <<<<<<<<<<<<<<
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ */
+      __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovec(__pyx_v_genovec, __pyx_v_self->_state_ptr);
+
+      /* "pgenlib.pyx":1680
+ *         cdef PglErr reterr
+ *         if not all_phased:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *             else:
+ */
+      goto __pyx_L11;
+    }
+
+    /* "pgenlib.pyx":1683
+ *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)             # <<<<<<<<<<<<<<
+ *         else:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ */
+    /*else*/ {
+      __pyx_v_reterr = plink2::SpgwAppendMultiallelicSparse(__pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, __pyx_v_write_allele_ct, __pyx_v_patch_01_ct, __pyx_v_patch_10_ct, __pyx_v_self->_state_ptr);
+    }
+    __pyx_L11:;
+
+    /* "pgenlib.pyx":1679
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ *         if not all_phased:             # <<<<<<<<<<<<<<
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ */
+    goto __pyx_L10;
+  }
+
+  /* "pgenlib.pyx":1685
+ *                 reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ *         else:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, phaseinfo, self._state_ptr)
+ *             else:
+ */
+  /*else*/ {
+    __pyx_t_12 = ((__pyx_v_patch_01_ct == 0) != 0);
+    if (__pyx_t_12) {
+    } else {
+      __pyx_t_13 = __pyx_t_12;
+      goto __pyx_L15_bool_binop_done;
+    }
+    __pyx_t_12 = ((__pyx_v_patch_10_ct == 0) != 0);
+    __pyx_t_13 = __pyx_t_12;
+    __pyx_L15_bool_binop_done:;
+    if (__pyx_t_13) {
+
+      /* "pgenlib.pyx":1686
+ *         else:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ */
+      __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, NULL, __pyx_v_phaseinfo, __pyx_v_self->_state_ptr);
+
+      /* "pgenlib.pyx":1685
+ *                 reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ *         else:
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, phaseinfo, self._state_ptr)
+ *             else:
+ */
+      goto __pyx_L14;
+    }
+
+    /* "pgenlib.pyx":1688
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, phaseinfo, self._state_ptr)
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_alleles() error " + str(reterr))
  */
-    __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, NULL, __pyx_v_self->_phaseinfo, __pyx_v_self->_state_ptr);
+    /*else*/ {
+      __pyx_v_reterr = plink2::SpgwAppendMultiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, NULL, __pyx_v_phaseinfo, __pyx_v_write_allele_ct, __pyx_v_patch_01_ct, __pyx_v_patch_10_ct, __pyx_v_self->_state_ptr);
+    }
+    __pyx_L14:;
   }
-  __pyx_L3:;
+  __pyx_L10:;
 
-  /* "pgenlib.pyx":1506
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+  /* "pgenlib.pyx":1689
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_alleles() error " + str(reterr))
  *         return
  */
-  __pyx_t_10 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-  if (unlikely(__pyx_t_10)) {
+  __pyx_t_13 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+  if (unlikely(__pyx_t_13)) {
 
-    /* "pgenlib.pyx":1507
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+    /* "pgenlib.pyx":1690
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_alleles() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1507, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1690, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1507, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1690, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_alleles_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1507, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_alleles_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1690, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1507, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1690, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 1507, __pyx_L1_error)
+    __PYX_ERR(0, 1690, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1506
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+    /* "pgenlib.pyx":1689
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_alleles() error " + str(reterr))
  *         return
  */
   }
 
-  /* "pgenlib.pyx":1508
+  /* "pgenlib.pyx":1691
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_alleles() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1494
+  /* "pgenlib.pyx":1650
  * 
  * 
- *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False):             # <<<<<<<<<<<<<<
+ *     cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False, object allele_ct = None):             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
   __Pyx_XDECREF(__pyx_t_2);
   __Pyx_XDECREF(__pyx_t_3);
@@ -25107,27 +27014,31 @@
 }
 
 /* Python wrapper */
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_7append_alleles(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_7append_alleles(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
   PyArrayObject *__pyx_v_allele_int32 = 0;
   int __pyx_v_all_phased;
+  PyObject *__pyx_v_allele_ct = 0;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_alleles (wrapper)", 0);
   {
-    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32,&__pyx_n_s_all_phased,0};
-    PyObject* values[2] = {0,0};
+    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32,&__pyx_n_s_all_phased,&__pyx_n_s_allele_ct,0};
+    PyObject* values[3] = {0,0,0};
+    values[2] = ((PyObject *)Py_None);
     if (unlikely(__pyx_kwds)) {
       Py_ssize_t kw_args;
       const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
       switch (pos_args) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         CYTHON_FALLTHROUGH;
         case  0: break;
         default: goto __pyx_L5_argtuple_error;
       }
@@ -25138,55 +27049,64 @@
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_all_phased);
           if (value) { values[1] = value; kw_args--; }
         }
+        CYTHON_FALLTHROUGH;
+        case  2:
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_ct);
+          if (value) { values[2] = value; kw_args--; }
+        }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_alleles") < 0)) __PYX_ERR(0, 1494, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_alleles") < 0)) __PYX_ERR(0, 1650, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_allele_int32 = ((PyArrayObject *)values[0]);
     if (values[1]) {
-      __pyx_v_all_phased = __Pyx_PyObject_IsTrue(values[1]); if (unlikely((__pyx_v_all_phased == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1494, __pyx_L3_error)
+      __pyx_v_all_phased = __Pyx_PyObject_IsTrue(values[1]); if (unlikely((__pyx_v_all_phased == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1650, __pyx_L3_error)
     } else {
       __pyx_v_all_phased = ((int)0);
     }
+    __pyx_v_allele_ct = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("append_alleles", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1494, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("append_alleles", 0, 1, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1650, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenWriter.append_alleles", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32), __pyx_ptype_5numpy_ndarray, 1, "allele_int32", 0))) __PYX_ERR(0, 1494, __pyx_L1_error)
-  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_6append_alleles(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32, __pyx_v_all_phased);
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32), __pyx_ptype_5numpy_ndarray, 1, "allele_int32", 0))) __PYX_ERR(0, 1650, __pyx_L1_error)
+  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_6append_alleles(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32, __pyx_v_all_phased, __pyx_v_allele_ct);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_6append_alleles(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, int __pyx_v_all_phased) {
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_6append_alleles(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, int __pyx_v_all_phased, PyObject *__pyx_v_allele_ct) {
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32;
   __Pyx_Buffer __pyx_pybuffer_allele_int32;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles __pyx_t_2;
   int __pyx_lineno = 0;
@@ -25195,21 +27115,22 @@
   __Pyx_RefNannySetupContext("append_alleles", 0);
   __pyx_pybuffer_allele_int32.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32.refcount = 0;
   __pyx_pybuffernd_allele_int32.data = NULL;
   __pyx_pybuffernd_allele_int32.rcbuffer = &__pyx_pybuffer_allele_int32;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1494, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1650, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32.diminfo[0].strides = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32.diminfo[0].shape = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_2.__pyx_n = 1;
+  __pyx_t_2.__pyx_n = 2;
   __pyx_t_2.all_phased = __pyx_v_all_phased;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenWriter->append_alleles(__pyx_v_self, __pyx_v_allele_int32, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1494, __pyx_L1_error)
+  __pyx_t_2.allele_ct = __pyx_v_allele_ct;
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenWriter->append_alleles(__pyx_v_self, __pyx_v_allele_int32, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1650, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -25227,29 +27148,40 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1511
+/* "pgenlib.pyx":1694
  * 
  * 
- *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):             # <<<<<<<<<<<<<<
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_9append_partially_phased(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
-static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent, int __pyx_skip_dispatch) {
+static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased *__pyx_optional_args) {
+  PyObject *__pyx_v_allele_ct = ((PyObject *)Py_None);
   int32_t *__pyx_v_allele_codes;
   unsigned char *__pyx_v_phasepresent_bytes;
+  uint32_t __pyx_v_sample_ct;
+  uint32_t __pyx_v_allele_ct_limit;
   uintptr_t *__pyx_v_genovec;
   uintptr_t *__pyx_v_phasepresent_buf;
   uintptr_t *__pyx_v_phaseinfo;
+  uintptr_t *__pyx_v_patch_01_set;
+  plink2::AlleleCode *__pyx_v_patch_01_vals;
+  uintptr_t *__pyx_v_patch_10_set;
+  plink2::AlleleCode *__pyx_v_patch_10_vals;
+  uint32_t __pyx_v_patch_01_ct;
+  uint32_t __pyx_v_patch_10_ct;
+  int32_t __pyx_v_observed_allele_ct;
+  uint32_t __pyx_v_write_allele_ct;
   plink2::PglErr __pyx_v_reterr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32;
   __Pyx_Buffer __pyx_pybuffer_allele_int32;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent;
   __Pyx_Buffer __pyx_pybuffer_phasepresent;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
@@ -25257,47 +27189,55 @@
   PyObject *__pyx_t_2 = NULL;
   PyObject *__pyx_t_3 = NULL;
   PyObject *__pyx_t_4 = NULL;
   int __pyx_t_5;
   PyObject *__pyx_t_6 = NULL;
   int __pyx_t_7;
   Py_ssize_t __pyx_t_8;
-  uintptr_t *__pyx_t_9;
+  uint32_t __pyx_t_9;
+  uintptr_t *__pyx_t_10;
+  plink2::AlleleCode *__pyx_t_11;
+  int __pyx_t_12;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_partially_phased", 0);
+  if (__pyx_optional_args) {
+    if (__pyx_optional_args->__pyx_n > 0) {
+      __pyx_v_allele_ct = __pyx_optional_args->allele_ct;
+    }
+  }
   __pyx_pybuffer_allele_int32.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32.refcount = 0;
   __pyx_pybuffernd_allele_int32.data = NULL;
   __pyx_pybuffernd_allele_int32.rcbuffer = &__pyx_pybuffer_allele_int32;
   __pyx_pybuffer_phasepresent.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent.refcount = 0;
   __pyx_pybuffernd_phasepresent.data = NULL;
   __pyx_pybuffernd_phasepresent.rcbuffer = &__pyx_pybuffer_phasepresent;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1511, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1694, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32.diminfo[0].strides = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32.diminfo[0].shape = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1511, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1694, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent.diminfo[0].strides = __pyx_pybuffernd_phasepresent.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent.diminfo[0].shape = __pyx_pybuffernd_phasepresent.rcbuffer->pybuffer.shape[0];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_partially_phased); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1511, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_partially_phased); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1694, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_9append_partially_phased)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         __pyx_t_5 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
@@ -25308,41 +27248,44 @@
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
             __pyx_t_5 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_3)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32), ((PyObject *)__pyx_v_phasepresent)};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 2+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1511, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32), ((PyObject *)__pyx_v_phasepresent), __pyx_v_allele_ct};
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1694, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_3)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32), ((PyObject *)__pyx_v_phasepresent)};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 2+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1511, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32), ((PyObject *)__pyx_v_phasepresent), __pyx_v_allele_ct};
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1694, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         {
-          __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1511, __pyx_L1_error)
+          __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1694, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_6);
           if (__pyx_t_4) {
             __Pyx_GIVEREF(__pyx_t_4); PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __pyx_t_4 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32));
           PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_allele_int32));
           __Pyx_INCREF(((PyObject *)__pyx_v_phasepresent));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_phasepresent));
           PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_phasepresent));
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1511, __pyx_L1_error)
+          __Pyx_INCREF(__pyx_v_allele_ct);
+          __Pyx_GIVEREF(__pyx_v_allele_ct);
+          PyTuple_SET_ITEM(__pyx_t_6, 2+__pyx_t_5, __pyx_v_allele_ct);
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1694, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         }
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -25357,188 +27300,464 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1512
+  /* "pgenlib.pyx":1695
  * 
- *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
  */
   __pyx_t_7 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalHardcallPhasePresent) == 0) != 0);
   if (unlikely(__pyx_t_7)) {
 
-    /* "pgenlib.pyx":1513
- *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):
+    /* "pgenlib.pyx":1696
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
  *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
  */
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__24, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1513, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__33, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1696, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 1513, __pyx_L1_error)
+    __PYX_ERR(0, 1696, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1512
+    /* "pgenlib.pyx":1695
  * 
- *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
  */
   }
 
-  /* "pgenlib.pyx":1514
+  /* "pgenlib.pyx":1697
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))             # <<<<<<<<<<<<<<
  *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
   __pyx_t_8 = 0;
   __pyx_t_5 = -1;
   if (__pyx_t_8 < 0) {
     __pyx_t_8 += __pyx_pybuffernd_allele_int32.diminfo[0].shape;
     if (unlikely(__pyx_t_8 < 0)) __pyx_t_5 = 0;
   } else if (unlikely(__pyx_t_8 >= __pyx_pybuffernd_allele_int32.diminfo[0].shape)) __pyx_t_5 = 0;
   if (unlikely(__pyx_t_5 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_5);
-    __PYX_ERR(0, 1514, __pyx_L1_error)
+    __PYX_ERR(0, 1697, __pyx_L1_error)
   }
   __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.buf, __pyx_t_8, __pyx_pybuffernd_allele_int32.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":1515
+  /* "pgenlib.pyx":1698
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
  *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent_buf = self._phasepresent
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  */
   __pyx_t_8 = 0;
   __pyx_t_5 = -1;
   if (__pyx_t_8 < 0) {
     __pyx_t_8 += __pyx_pybuffernd_phasepresent.diminfo[0].shape;
     if (unlikely(__pyx_t_8 < 0)) __pyx_t_5 = 0;
   } else if (unlikely(__pyx_t_8 >= __pyx_pybuffernd_phasepresent.diminfo[0].shape)) __pyx_t_5 = 0;
   if (unlikely(__pyx_t_5 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_5);
-    __PYX_ERR(0, 1515, __pyx_L1_error)
+    __PYX_ERR(0, 1698, __pyx_L1_error)
   }
   __pyx_v_phasepresent_bytes = ((unsigned char *)(&(*__Pyx_BufPtrStrided1d(__pyx_t_5numpy_uint8_t *, __pyx_pybuffernd_phasepresent.rcbuffer->pybuffer.buf, __pyx_t_8, __pyx_pybuffernd_phasepresent.diminfo[0].strides))));
 
-  /* "pgenlib.pyx":1516
+  /* "pgenlib.pyx":1699
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
  *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)             # <<<<<<<<<<<<<<
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
+ *         cdef uintptr_t* genovec = self._genovec
+ */
+  __pyx_v_sample_ct = plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr);
+
+  /* "pgenlib.pyx":1700
+ *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* phasepresent_buf = self._phasepresent
+ */
+  __pyx_t_9 = __pyx_v_self->_allele_ct_limit;
+  __pyx_v_allele_ct_limit = __pyx_t_9;
+
+  /* "pgenlib.pyx":1701
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent
  *         cdef uintptr_t* phaseinfo = self._phaseinfo
  */
-  __pyx_t_9 = __pyx_v_self->_genovec;
-  __pyx_v_genovec = __pyx_t_9;
+  __pyx_t_10 = __pyx_v_self->_genovec;
+  __pyx_v_genovec = __pyx_t_10;
 
-  /* "pgenlib.pyx":1517
- *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
+  /* "pgenlib.pyx":1702
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* phaseinfo = self._phaseinfo
- *         AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
  */
-  __pyx_t_9 = __pyx_v_self->_phasepresent;
-  __pyx_v_phasepresent_buf = __pyx_t_9;
+  __pyx_t_10 = __pyx_v_self->_phasepresent;
+  __pyx_v_phasepresent_buf = __pyx_t_10;
 
-  /* "pgenlib.pyx":1518
+  /* "pgenlib.pyx":1703
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent
  *         cdef uintptr_t* phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
- *         AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *         cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
  */
-  __pyx_t_9 = __pyx_v_self->_phaseinfo;
-  __pyx_v_phaseinfo = __pyx_t_9;
+  __pyx_t_10 = __pyx_v_self->_phaseinfo;
+  __pyx_v_phaseinfo = __pyx_t_10;
 
-  /* "pgenlib.pyx":1519
+  /* "pgenlib.pyx":1704
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent
  *         cdef uintptr_t* phaseinfo = self._phaseinfo
- *         AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)             # <<<<<<<<<<<<<<
- *         cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
- *         if reterr != kPglRetSuccess:
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
  */
-  plink2::AlleleCodesToGenoarrUnsafe(__pyx_v_allele_codes, __pyx_v_phasepresent_bytes, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_genovec, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo);
+  __pyx_t_10 = __pyx_v_self->_patch_01_set;
+  __pyx_v_patch_01_set = __pyx_t_10;
 
-  /* "pgenlib.pyx":1520
+  /* "pgenlib.pyx":1705
  *         cdef uintptr_t* phaseinfo = self._phaseinfo
- *         AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *         cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ */
+  __pyx_t_11 = __pyx_v_self->_patch_01_vals;
+  __pyx_v_patch_01_vals = __pyx_t_11;
+
+  /* "pgenlib.pyx":1706
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ *         cdef uint32_t patch_01_ct
+ */
+  __pyx_t_10 = __pyx_v_self->_patch_10_set;
+  __pyx_v_patch_10_set = __pyx_t_10;
+
+  /* "pgenlib.pyx":1707
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals             # <<<<<<<<<<<<<<
+ *         cdef uint32_t patch_01_ct
+ *         cdef uint32_t patch_10_ct
+ */
+  __pyx_t_11 = __pyx_v_self->_patch_10_vals;
+  __pyx_v_patch_10_vals = __pyx_t_11;
+
+  /* "pgenlib.pyx":1710
+ *         cdef uint32_t patch_01_ct
+ *         cdef uint32_t patch_10_ct
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)             # <<<<<<<<<<<<<<
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")
+ */
+  __pyx_v_observed_allele_ct = plink2::ConvertMultiAlleleCodesUnsafe(__pyx_v_allele_codes, __pyx_v_phasepresent_bytes, __pyx_v_sample_ct, __pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, (&__pyx_v_patch_01_ct), (&__pyx_v_patch_10_ct), __pyx_v_phasepresent_buf, __pyx_v_phaseinfo);
+
+  /* "pgenlib.pyx":1711
+ *         cdef uint32_t patch_10_ct
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *         if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+  __pyx_t_7 = ((__pyx_v_observed_allele_ct == -1L) != 0);
+  if (unlikely(__pyx_t_7)) {
+
+    /* "pgenlib.pyx":1712
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ */
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__34, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1712, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_1);
+    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    __PYX_ERR(0, 1712, __pyx_L1_error)
+
+    /* "pgenlib.pyx":1711
+ *         cdef uint32_t patch_10_ct
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *         if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+  }
+
+  /* "pgenlib.pyx":1713
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)             # <<<<<<<<<<<<<<
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ */
+  __pyx_v_write_allele_ct = ((uint32_t)__pyx_v_observed_allele_ct);
+
+  /* "pgenlib.pyx":1714
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ */
+  __pyx_t_7 = ((__pyx_v_write_allele_ct > __pyx_v_allele_ct_limit) != 0);
+  if (unlikely(__pyx_t_7)) {
+
+    /* "pgenlib.pyx":1715
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ */
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__35, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1715, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_1);
+    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    __PYX_ERR(0, 1715, __pyx_L1_error)
+
+    /* "pgenlib.pyx":1714
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ */
+  }
+
+  /* "pgenlib.pyx":1716
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:             # <<<<<<<<<<<<<<
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ */
+  __pyx_t_7 = (__pyx_v_allele_ct != Py_None);
+  __pyx_t_12 = (__pyx_t_7 != 0);
+  if (__pyx_t_12) {
+
+    /* "pgenlib.pyx":1717
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ */
+    __pyx_t_1 = __Pyx_PyInt_From_uint32_t(__pyx_v_write_allele_ct); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1717, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_1);
+    __pyx_t_2 = PyObject_RichCompare(__pyx_v_allele_ct, __pyx_t_1, Py_LT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1717, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    __pyx_t_12 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_12 < 0)) __PYX_ERR(0, 1717, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+    if (unlikely(__pyx_t_12)) {
+
+      /* "pgenlib.pyx":1718
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")             # <<<<<<<<<<<<<<
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")
+ */
+      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__36, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1718, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_2);
+      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+      __PYX_ERR(0, 1718, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1717
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ */
+    }
+
+    /* "pgenlib.pyx":1719
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")
+ *             write_allele_ct = allele_ct
+ */
+    __pyx_t_2 = __Pyx_PyInt_From_uint32_t(__pyx_v_allele_ct_limit); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1719, __pyx_L1_error)
+    __Pyx_GOTREF(__pyx_t_2);
+    __pyx_t_1 = PyObject_RichCompare(__pyx_v_allele_ct, __pyx_t_2, Py_GT); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1719, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+    __pyx_t_12 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_12 < 0)) __PYX_ERR(0, 1719, __pyx_L1_error)
+    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+    if (unlikely(__pyx_t_12)) {
+
+      /* "pgenlib.pyx":1720
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ */
+      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__37, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1720, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __PYX_ERR(0, 1720, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1719
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")
+ *             write_allele_ct = allele_ct
+ */
+    }
+
+    /* "pgenlib.pyx":1721
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")
+ *             write_allele_ct = allele_ct             # <<<<<<<<<<<<<<
+ *         cdef PglErr reterr
+ *         if (patch_01_ct == 0) and (patch_10_ct == 0):
+ */
+    __pyx_t_9 = __Pyx_PyInt_As_uint32_t(__pyx_v_allele_ct); if (unlikely((__pyx_t_9 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1721, __pyx_L1_error)
+    __pyx_v_write_allele_ct = __pyx_t_9;
+
+    /* "pgenlib.pyx":1716
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *         if allele_ct is not None:             # <<<<<<<<<<<<<<
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ */
+  }
+
+  /* "pgenlib.pyx":1723
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ *         if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *         else:
+ */
+  __pyx_t_7 = ((__pyx_v_patch_01_ct == 0) != 0);
+  if (__pyx_t_7) {
+  } else {
+    __pyx_t_12 = __pyx_t_7;
+    goto __pyx_L10_bool_binop_done;
+  }
+  __pyx_t_7 = ((__pyx_v_patch_10_ct == 0) != 0);
+  __pyx_t_12 = __pyx_t_7;
+  __pyx_L10_bool_binop_done:;
+  if (__pyx_t_12) {
+
+    /* "pgenlib.pyx":1724
+ *         cdef PglErr reterr
+ *         if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *         else:
+ *             reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ */
+    __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo, __pyx_v_self->_state_ptr);
+
+    /* "pgenlib.pyx":1723
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ *         if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *         else:
+ */
+    goto __pyx_L9;
+  }
+
+  /* "pgenlib.pyx":1726
+ *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *         else:
+ *             reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_partially_phased() error " + str(reterr))
  */
-  __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo, __pyx_v_self->_state_ptr);
+  /*else*/ {
+    __pyx_v_reterr = plink2::SpgwAppendMultiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo, __pyx_v_write_allele_ct, __pyx_v_patch_01_ct, __pyx_v_patch_10_ct, __pyx_v_self->_state_ptr);
+  }
+  __pyx_L9:;
 
-  /* "pgenlib.pyx":1521
- *         AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *         cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+  /* "pgenlib.pyx":1727
+ *         else:
+ *             reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_partially_phased() error " + str(reterr))
  *         return
  */
-  __pyx_t_7 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-  if (unlikely(__pyx_t_7)) {
+  __pyx_t_12 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+  if (unlikely(__pyx_t_12)) {
 
-    /* "pgenlib.pyx":1522
- *         cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+    /* "pgenlib.pyx":1728
+ *             reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_partially_phased() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1522, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1728, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1522, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1728, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_partially_phased_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1522, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_partially_phased_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1728, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1522, __pyx_L1_error)
+    __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1728, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_Raise(__pyx_t_2, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-    __PYX_ERR(0, 1522, __pyx_L1_error)
+    __PYX_ERR(0, 1728, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1521
- *         AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *         cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+    /* "pgenlib.pyx":1727
+ *         else:
+ *             reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_partially_phased() error " + str(reterr))
  *         return
  */
   }
 
-  /* "pgenlib.pyx":1523
+  /* "pgenlib.pyx":1729
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_partially_phased() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1511
+  /* "pgenlib.pyx":1694
  * 
  * 
- *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):             # <<<<<<<<<<<<<<
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
@@ -25566,27 +27785,31 @@
 }
 
 /* Python wrapper */
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_9append_partially_phased(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_9append_partially_phased(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
   PyArrayObject *__pyx_v_allele_int32 = 0;
   PyArrayObject *__pyx_v_phasepresent = 0;
+  PyObject *__pyx_v_allele_ct = 0;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_partially_phased (wrapper)", 0);
   {
-    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32,&__pyx_n_s_phasepresent,0};
-    PyObject* values[2] = {0,0};
+    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32,&__pyx_n_s_phasepresent,&__pyx_n_s_allele_ct,0};
+    PyObject* values[3] = {0,0,0};
+    values[2] = ((PyObject *)Py_None);
     if (unlikely(__pyx_kwds)) {
       Py_ssize_t kw_args;
       const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
       switch (pos_args) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         CYTHON_FALLTHROUGH;
         case  0: break;
         default: goto __pyx_L5_argtuple_error;
       }
@@ -25595,82 +27818,96 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_phasepresent)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("append_partially_phased", 1, 2, 2, 1); __PYX_ERR(0, 1511, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("append_partially_phased", 0, 2, 3, 1); __PYX_ERR(0, 1694, __pyx_L3_error)
+        }
+        CYTHON_FALLTHROUGH;
+        case  2:
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_ct);
+          if (value) { values[2] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_partially_phased") < 0)) __PYX_ERR(0, 1511, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_partially_phased") < 0)) __PYX_ERR(0, 1694, __pyx_L3_error)
       }
-    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
-      goto __pyx_L5_argtuple_error;
     } else {
-      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
-      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
+      switch (PyTuple_GET_SIZE(__pyx_args)) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
+        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
+        values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
+        break;
+        default: goto __pyx_L5_argtuple_error;
+      }
     }
     __pyx_v_allele_int32 = ((PyArrayObject *)values[0]);
     __pyx_v_phasepresent = ((PyArrayObject *)values[1]);
+    __pyx_v_allele_ct = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("append_partially_phased", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1511, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("append_partially_phased", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1694, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenWriter.append_partially_phased", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32), __pyx_ptype_5numpy_ndarray, 1, "allele_int32", 0))) __PYX_ERR(0, 1511, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent), __pyx_ptype_5numpy_ndarray, 1, "phasepresent", 0))) __PYX_ERR(0, 1511, __pyx_L1_error)
-  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_8append_partially_phased(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32, __pyx_v_phasepresent);
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32), __pyx_ptype_5numpy_ndarray, 1, "allele_int32", 0))) __PYX_ERR(0, 1694, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent), __pyx_ptype_5numpy_ndarray, 1, "phasepresent", 0))) __PYX_ERR(0, 1694, __pyx_L1_error)
+  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_8append_partially_phased(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32, __pyx_v_phasepresent, __pyx_v_allele_ct);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_8append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent) {
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_8append_partially_phased(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32, PyArrayObject *__pyx_v_phasepresent, PyObject *__pyx_v_allele_ct) {
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32;
   __Pyx_Buffer __pyx_pybuffer_allele_int32;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent;
   __Pyx_Buffer __pyx_pybuffer_phasepresent;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
+  struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased __pyx_t_2;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_partially_phased", 0);
   __pyx_pybuffer_allele_int32.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32.refcount = 0;
   __pyx_pybuffernd_allele_int32.data = NULL;
   __pyx_pybuffernd_allele_int32.rcbuffer = &__pyx_pybuffer_allele_int32;
   __pyx_pybuffer_phasepresent.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent.refcount = 0;
   __pyx_pybuffernd_phasepresent.data = NULL;
   __pyx_pybuffernd_phasepresent.rcbuffer = &__pyx_pybuffer_phasepresent;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1511, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1694, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32.diminfo[0].strides = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32.diminfo[0].shape = __pyx_pybuffernd_allele_int32.rcbuffer->pybuffer.shape[0];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1511, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_STRIDES, 1, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1694, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent.diminfo[0].strides = __pyx_pybuffernd_phasepresent.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent.diminfo[0].shape = __pyx_pybuffernd_phasepresent.rcbuffer->pybuffer.shape[0];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_partially_phased(__pyx_v_self, __pyx_v_allele_int32, __pyx_v_phasepresent, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1511, __pyx_L1_error)
+  __pyx_t_2.__pyx_n = 1;
+  __pyx_t_2.allele_ct = __pyx_v_allele_ct;
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenWriter->append_partially_phased(__pyx_v_self, __pyx_v_allele_int32, __pyx_v_phasepresent, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1694, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -25690,15 +27927,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_phasepresent.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1526
+/* "pgenlib.pyx":1732
  * 
  * 
  *     cdef append_dosages_internal32(self, np.ndarray[np.float32_t,mode="c"] floatarr):             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  */
 
@@ -25725,129 +27962,129 @@
   __Pyx_RefNannySetupContext("append_dosages_internal32", 0);
   __pyx_pybuffer_floatarr.pybuffer.buf = NULL;
   __pyx_pybuffer_floatarr.refcount = 0;
   __pyx_pybuffernd_floatarr.data = NULL;
   __pyx_pybuffernd_floatarr.rcbuffer = &__pyx_pybuffer_floatarr;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1526, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1732, __pyx_L1_error)
   }
   __pyx_pybuffernd_floatarr.diminfo[0].strides = __pyx_pybuffernd_floatarr.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_floatarr.diminfo[0].shape = __pyx_pybuffernd_floatarr.rcbuffer->pybuffer.shape[0];
 
-  /* "pgenlib.pyx":1527
+  /* "pgenlib.pyx":1733
  * 
  *     cdef append_dosages_internal32(self, np.ndarray[np.float32_t,mode="c"] floatarr):
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main
  */
   __pyx_t_1 = __pyx_v_self->_genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1528
+  /* "pgenlib.pyx":1734
  *     cdef append_dosages_internal32(self, np.ndarray[np.float32_t,mode="c"] floatarr):
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
  *         cdef uint16_t* dosage_main = self._dosage_main
  *         cdef uint32_t dosage_ct
  */
   __pyx_t_1 = __pyx_v_self->_dosage_present;
   __pyx_v_dosage_present = __pyx_t_1;
 
-  /* "pgenlib.pyx":1529
+  /* "pgenlib.pyx":1735
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t dosage_ct
  *         FloatsToDosage16(<float*>(&(floatarr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  */
   __pyx_t_2 = __pyx_v_self->_dosage_main;
   __pyx_v_dosage_main = __pyx_t_2;
 
-  /* "pgenlib.pyx":1531
+  /* "pgenlib.pyx":1737
  *         cdef uint16_t* dosage_main = self._dosage_main
  *         cdef uint32_t dosage_ct
  *         FloatsToDosage16(<float*>(&(floatarr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  */
   __pyx_t_3 = 0;
   __pyx_t_4 = -1;
   if (__pyx_t_3 < 0) {
     __pyx_t_3 += __pyx_pybuffernd_floatarr.diminfo[0].shape;
     if (unlikely(__pyx_t_3 < 0)) __pyx_t_4 = 0;
   } else if (unlikely(__pyx_t_3 >= __pyx_pybuffernd_floatarr.diminfo[0].shape)) __pyx_t_4 = 0;
   if (unlikely(__pyx_t_4 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_4);
-    __PYX_ERR(0, 1531, __pyx_L1_error)
+    __PYX_ERR(0, 1737, __pyx_L1_error)
   }
   plink2::FloatsToDosage16(((float *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_float32_t *, __pyx_pybuffernd_floatarr.rcbuffer->pybuffer.buf, __pyx_t_3, __pyx_pybuffernd_floatarr.diminfo[0].strides)))), plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), 0x199A, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-  /* "pgenlib.pyx":1532
+  /* "pgenlib.pyx":1738
  *         cdef uint32_t dosage_ct
  *         FloatsToDosage16(<float*>(&(floatarr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  */
   __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecDosage16(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_dosage_ct, __pyx_v_self->_state_ptr);
 
-  /* "pgenlib.pyx":1533
+  /* "pgenlib.pyx":1739
  *         FloatsToDosage16(<float*>(&(floatarr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  *         return
  */
   __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":1534
+    /* "pgenlib.pyx":1740
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_dosages() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-    __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1534, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1740, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1534, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1740, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1534, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1740, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1534, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1740, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 1534, __pyx_L1_error)
+    __PYX_ERR(0, 1740, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1533
+    /* "pgenlib.pyx":1739
  *         FloatsToDosage16(<float*>(&(floatarr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  *         return
  */
   }
 
-  /* "pgenlib.pyx":1535
+  /* "pgenlib.pyx":1741
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cdef append_dosages_internal64(self, np.ndarray[np.float64_t,mode="c"] doublearr):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1526
+  /* "pgenlib.pyx":1732
  * 
  * 
  *     cdef append_dosages_internal32(self, np.ndarray[np.float32_t,mode="c"] floatarr):             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  */
 
@@ -25868,15 +28105,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_floatarr.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1537
+/* "pgenlib.pyx":1743
  *         return
  * 
  *     cdef append_dosages_internal64(self, np.ndarray[np.float64_t,mode="c"] doublearr):             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  */
 
@@ -25903,129 +28140,129 @@
   __Pyx_RefNannySetupContext("append_dosages_internal64", 0);
   __pyx_pybuffer_doublearr.pybuffer.buf = NULL;
   __pyx_pybuffer_doublearr.refcount = 0;
   __pyx_pybuffernd_doublearr.data = NULL;
   __pyx_pybuffernd_doublearr.rcbuffer = &__pyx_pybuffer_doublearr;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_doublearr.rcbuffer->pybuffer, (PyObject*)__pyx_v_doublearr, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1537, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_doublearr.rcbuffer->pybuffer, (PyObject*)__pyx_v_doublearr, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 1, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1743, __pyx_L1_error)
   }
   __pyx_pybuffernd_doublearr.diminfo[0].strides = __pyx_pybuffernd_doublearr.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_doublearr.diminfo[0].shape = __pyx_pybuffernd_doublearr.rcbuffer->pybuffer.shape[0];
 
-  /* "pgenlib.pyx":1538
+  /* "pgenlib.pyx":1744
  * 
  *     cdef append_dosages_internal64(self, np.ndarray[np.float64_t,mode="c"] doublearr):
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main
  */
   __pyx_t_1 = __pyx_v_self->_genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1539
+  /* "pgenlib.pyx":1745
  *     cdef append_dosages_internal64(self, np.ndarray[np.float64_t,mode="c"] doublearr):
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
  *         cdef uint16_t* dosage_main = self._dosage_main
  *         cdef uint32_t dosage_ct
  */
   __pyx_t_1 = __pyx_v_self->_dosage_present;
   __pyx_v_dosage_present = __pyx_t_1;
 
-  /* "pgenlib.pyx":1540
+  /* "pgenlib.pyx":1746
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t dosage_ct
  *         DoublesToDosage16(<double*>(&(doublearr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  */
   __pyx_t_2 = __pyx_v_self->_dosage_main;
   __pyx_v_dosage_main = __pyx_t_2;
 
-  /* "pgenlib.pyx":1542
+  /* "pgenlib.pyx":1748
  *         cdef uint16_t* dosage_main = self._dosage_main
  *         cdef uint32_t dosage_ct
  *         DoublesToDosage16(<double*>(&(doublearr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  */
   __pyx_t_3 = 0;
   __pyx_t_4 = -1;
   if (__pyx_t_3 < 0) {
     __pyx_t_3 += __pyx_pybuffernd_doublearr.diminfo[0].shape;
     if (unlikely(__pyx_t_3 < 0)) __pyx_t_4 = 0;
   } else if (unlikely(__pyx_t_3 >= __pyx_pybuffernd_doublearr.diminfo[0].shape)) __pyx_t_4 = 0;
   if (unlikely(__pyx_t_4 != -1)) {
     __Pyx_RaiseBufferIndexError(__pyx_t_4);
-    __PYX_ERR(0, 1542, __pyx_L1_error)
+    __PYX_ERR(0, 1748, __pyx_L1_error)
   }
   plink2::DoublesToDosage16(((double *)(&(*__Pyx_BufPtrCContig1d(__pyx_t_5numpy_float64_t *, __pyx_pybuffernd_doublearr.rcbuffer->pybuffer.buf, __pyx_t_3, __pyx_pybuffernd_doublearr.diminfo[0].strides)))), plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), 0x199A, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-  /* "pgenlib.pyx":1543
+  /* "pgenlib.pyx":1749
  *         cdef uint32_t dosage_ct
  *         DoublesToDosage16(<double*>(&(doublearr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  */
   __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecDosage16(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_dosage_ct, __pyx_v_self->_state_ptr);
 
-  /* "pgenlib.pyx":1544
+  /* "pgenlib.pyx":1750
  *         DoublesToDosage16(<double*>(&(doublearr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  *         return
  */
   __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":1545
+    /* "pgenlib.pyx":1751
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_dosages() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-    __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1545, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1751, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1545, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1751, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
-    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1545, __pyx_L1_error)
+    __pyx_t_6 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_error, __pyx_t_7); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1751, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_6);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1545, __pyx_L1_error)
+    __pyx_t_7 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1751, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_7);
     __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
     __Pyx_Raise(__pyx_t_7, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
-    __PYX_ERR(0, 1545, __pyx_L1_error)
+    __PYX_ERR(0, 1751, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1544
+    /* "pgenlib.pyx":1750
  *         DoublesToDosage16(<double*>(&(doublearr[0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *         cdef PglErr reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *         if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  *         return
  */
   }
 
-  /* "pgenlib.pyx":1546
+  /* "pgenlib.pyx":1752
  *         if reterr != kPglRetSuccess:
  *             raise RuntimeError("append_dosages() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cpdef append_dosages(self, np.ndarray floatarr):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1537
+  /* "pgenlib.pyx":1743
  *         return
  * 
  *     cdef append_dosages_internal64(self, np.ndarray[np.float64_t,mode="c"] doublearr):             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  */
 
@@ -26046,15 +28283,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_doublearr.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1548
+/* "pgenlib.pyx":1754
  *         return
  * 
  *     cpdef append_dosages(self, np.ndarray floatarr):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  */
 
@@ -26076,15 +28313,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_dosages); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1548, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_dosages); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1754, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_11append_dosages)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -26093,15 +28330,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_Call2Args(__pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_floatarr)) : __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_floatarr));
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1548, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1754, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -26114,158 +28351,158 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1549
+  /* "pgenlib.pyx":1755
  * 
  *     cpdef append_dosages(self, np.ndarray floatarr):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr.dtype == np.float32:
  */
   __pyx_t_5 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalDosagePresent) == 0) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":1550
+    /* "pgenlib.pyx":1756
  *     cpdef append_dosages(self, np.ndarray floatarr):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")             # <<<<<<<<<<<<<<
  *         if floatarr.dtype == np.float32:
  *             self.append_dosages_internal32(floatarr)
  */
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__25, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1550, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__38, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1756, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 1550, __pyx_L1_error)
+    __PYX_ERR(0, 1756, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1549
+    /* "pgenlib.pyx":1755
  * 
  *     cpdef append_dosages(self, np.ndarray floatarr):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr.dtype == np.float32:
  */
   }
 
-  /* "pgenlib.pyx":1551
+  /* "pgenlib.pyx":1757
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.append_dosages_internal32(floatarr)
  *         elif floatarr.dtype == np.float64:
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1551, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1757, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1551, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1757, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float32); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1551, __pyx_L1_error)
+  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float32); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1757, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1551, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1757, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1551, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1757, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":1552
+    /* "pgenlib.pyx":1758
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr.dtype == np.float32:
  *             self.append_dosages_internal32(floatarr)             # <<<<<<<<<<<<<<
  *         elif floatarr.dtype == np.float64:
  *             self.append_dosages_internal64(floatarr)
  */
-    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr)); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1552, __pyx_L1_error)
+    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr)); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1758, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
 
-    /* "pgenlib.pyx":1551
+    /* "pgenlib.pyx":1757
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.append_dosages_internal32(floatarr)
  *         elif floatarr.dtype == np.float64:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":1553
+  /* "pgenlib.pyx":1759
  *         if floatarr.dtype == np.float32:
  *             self.append_dosages_internal32(floatarr)
  *         elif floatarr.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.append_dosages_internal64(floatarr)
  *         else:
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1553, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1759, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_3, __pyx_n_s_np); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1553, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_3, __pyx_n_s_np); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1759, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1553, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1759, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-  __pyx_t_3 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1553, __pyx_L1_error)
+  __pyx_t_3 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1759, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1553, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1759, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
   if (likely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":1554
+    /* "pgenlib.pyx":1760
  *             self.append_dosages_internal32(floatarr)
  *         elif floatarr.dtype == np.float64:
  *             self.append_dosages_internal64(floatarr)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid append_dosages() dosage array element type (float32 or float64 expected).")
  */
-    __pyx_t_3 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1554, __pyx_L1_error)
+    __pyx_t_3 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1760, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
 
-    /* "pgenlib.pyx":1553
+    /* "pgenlib.pyx":1759
  *         if floatarr.dtype == np.float32:
  *             self.append_dosages_internal32(floatarr)
  *         elif floatarr.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.append_dosages_internal64(floatarr)
  *         else:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":1556
+  /* "pgenlib.pyx":1762
  *             self.append_dosages_internal64(floatarr)
  *         else:
  *             raise RuntimeError("Invalid append_dosages() dosage array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__26, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1556, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__39, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1762, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
     __Pyx_Raise(__pyx_t_3, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __PYX_ERR(0, 1556, __pyx_L1_error)
+    __PYX_ERR(0, 1762, __pyx_L1_error)
   }
   __pyx_L4:;
 
-  /* "pgenlib.pyx":1557
+  /* "pgenlib.pyx":1763
  *         else:
  *             raise RuntimeError("Invalid append_dosages() dosage array element type (float32 or float64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1548
+  /* "pgenlib.pyx":1754
  *         return
  * 
  *     cpdef append_dosages(self, np.ndarray floatarr):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")
  */
 
@@ -26288,15 +28525,15 @@
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_11append_dosages(PyObject *__pyx_v_self, PyObject *__pyx_v_floatarr) {
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_dosages (wrapper)", 0);
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr), __pyx_ptype_5numpy_ndarray, 1, "floatarr", 0))) __PYX_ERR(0, 1548, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr), __pyx_ptype_5numpy_ndarray, 1, "floatarr", 0))) __PYX_ERR(0, 1754, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_10append_dosages(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), ((PyArrayObject *)__pyx_v_floatarr));
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -26309,15 +28546,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_dosages", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_dosages(__pyx_v_self, __pyx_v_floatarr, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1548, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_dosages(__pyx_v_self, __pyx_v_floatarr, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1754, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -26326,15 +28563,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1560
+/* "pgenlib.pyx":1766
  * 
  * 
  *     cpdef append_biallelic_batch(self, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_batch):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>geno_int8_batch.shape[0]
  *         cdef int8_t* genobytes
  */
 
@@ -26365,27 +28602,27 @@
   __Pyx_RefNannySetupContext("append_biallelic_batch", 0);
   __pyx_pybuffer_geno_int8_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int8_batch.refcount = 0;
   __pyx_pybuffernd_geno_int8_batch.data = NULL;
   __pyx_pybuffernd_geno_int8_batch.rcbuffer = &__pyx_pybuffer_geno_int8_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1560, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1766, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int8_batch.diminfo[0].strides = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int8_batch.diminfo[0].shape = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int8_batch.diminfo[1].strides = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int8_batch.diminfo[1].shape = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_biallelic_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1560, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_biallelic_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1766, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_13append_biallelic_batch)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -26394,15 +28631,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_Call2Args(__pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_geno_int8_batch)) : __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_geno_int8_batch));
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1560, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1766, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -26415,36 +28652,36 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1561
+  /* "pgenlib.pyx":1767
  * 
  *     cpdef append_biallelic_batch(self, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_batch):
  *         cdef uint32_t batch_size = <uint32_t>geno_int8_batch.shape[0]             # <<<<<<<<<<<<<<
  *         cdef int8_t* genobytes
  *         cdef uint32_t uii
  */
   __pyx_v_batch_size = ((uint32_t)(__pyx_v_geno_int8_batch->dimensions[0]));
 
-  /* "pgenlib.pyx":1565
+  /* "pgenlib.pyx":1771
  *         cdef uint32_t uii
  *         cdef PglErr reterr
  *         for uii in range(batch_size):             # <<<<<<<<<<<<<<
  *             genobytes = &(geno_int8_batch[uii, 0])
  *             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  */
   __pyx_t_5 = __pyx_v_batch_size;
   __pyx_t_6 = __pyx_t_5;
   for (__pyx_t_7 = 0; __pyx_t_7 < __pyx_t_6; __pyx_t_7+=1) {
     __pyx_v_uii = __pyx_t_7;
 
-    /* "pgenlib.pyx":1566
+    /* "pgenlib.pyx":1772
  *         cdef PglErr reterr
  *         for uii in range(batch_size):
  *             genobytes = &(geno_int8_batch[uii, 0])             # <<<<<<<<<<<<<<
  *             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  */
     __pyx_t_8 = __pyx_v_uii;
@@ -26453,90 +28690,90 @@
     if (unlikely(__pyx_t_8 >= (size_t)__pyx_pybuffernd_geno_int8_batch.diminfo[0].shape)) __pyx_t_10 = 0;
     if (__pyx_t_9 < 0) {
       __pyx_t_9 += __pyx_pybuffernd_geno_int8_batch.diminfo[1].shape;
       if (unlikely(__pyx_t_9 < 0)) __pyx_t_10 = 1;
     } else if (unlikely(__pyx_t_9 >= __pyx_pybuffernd_geno_int8_batch.diminfo[1].shape)) __pyx_t_10 = 1;
     if (unlikely(__pyx_t_10 != -1)) {
       __Pyx_RaiseBufferIndexError(__pyx_t_10);
-      __PYX_ERR(0, 1566, __pyx_L1_error)
+      __PYX_ERR(0, 1772, __pyx_L1_error)
     }
     __pyx_v_genobytes = (&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int8_t *, __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.buf, __pyx_t_8, __pyx_pybuffernd_geno_int8_batch.diminfo[0].strides, __pyx_t_9, __pyx_pybuffernd_geno_int8_batch.diminfo[1].strides)));
 
-    /* "pgenlib.pyx":1567
+    /* "pgenlib.pyx":1773
  *         for uii in range(batch_size):
  *             genobytes = &(geno_int8_batch[uii, 0])
  *             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)             # <<<<<<<<<<<<<<
  *             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  */
     plink2::BytesToGenoarrUnsafe(__pyx_v_genobytes, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_self->_genovec);
 
-    /* "pgenlib.pyx":1568
+    /* "pgenlib.pyx":1774
  *             genobytes = &(geno_int8_batch[uii, 0])
  *             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)             # <<<<<<<<<<<<<<
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_biallelic_batch() error " + str(reterr))
  */
     __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovec(__pyx_v_self->_genovec, __pyx_v_self->_state_ptr);
 
-    /* "pgenlib.pyx":1569
+    /* "pgenlib.pyx":1775
  *             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_biallelic_batch() error " + str(reterr))
  *         return
  */
     __pyx_t_11 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
     if (unlikely(__pyx_t_11)) {
 
-      /* "pgenlib.pyx":1570
+      /* "pgenlib.pyx":1776
  *             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_biallelic_batch() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1570, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1776, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1570, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1776, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_biallelic_batch_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1570, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_biallelic_batch_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1776, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1570, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1776, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       __Pyx_Raise(__pyx_t_2, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __PYX_ERR(0, 1570, __pyx_L1_error)
+      __PYX_ERR(0, 1776, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1569
+      /* "pgenlib.pyx":1775
  *             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
  *             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_biallelic_batch() error " + str(reterr))
  *         return
  */
     }
   }
 
-  /* "pgenlib.pyx":1571
+  /* "pgenlib.pyx":1777
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_biallelic_batch() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1560
+  /* "pgenlib.pyx":1766
  * 
  * 
  *     cpdef append_biallelic_batch(self, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_batch):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>geno_int8_batch.shape[0]
  *         cdef int8_t* genobytes
  */
 
@@ -26568,15 +28805,15 @@
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_13append_biallelic_batch(PyObject *__pyx_v_self, PyObject *__pyx_v_geno_int8_batch) {
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_biallelic_batch (wrapper)", 0);
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int8_batch), __pyx_ptype_5numpy_ndarray, 1, "geno_int8_batch", 0))) __PYX_ERR(0, 1560, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_geno_int8_batch), __pyx_ptype_5numpy_ndarray, 1, "geno_int8_batch", 0))) __PYX_ERR(0, 1766, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_12append_biallelic_batch(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), ((PyArrayObject *)__pyx_v_geno_int8_batch));
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -26596,19 +28833,19 @@
   __Pyx_RefNannySetupContext("append_biallelic_batch", 0);
   __pyx_pybuffer_geno_int8_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_geno_int8_batch.refcount = 0;
   __pyx_pybuffernd_geno_int8_batch.data = NULL;
   __pyx_pybuffernd_geno_int8_batch.rcbuffer = &__pyx_pybuffer_geno_int8_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1560, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_geno_int8_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1766, __pyx_L1_error)
   }
   __pyx_pybuffernd_geno_int8_batch.diminfo[0].strides = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_geno_int8_batch.diminfo[0].shape = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_geno_int8_batch.diminfo[1].strides = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_geno_int8_batch.diminfo[1].shape = __pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_biallelic_batch(__pyx_v_self, __pyx_v_geno_int8_batch, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1560, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_biallelic_batch(__pyx_v_self, __pyx_v_geno_int8_batch, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1766, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -26626,80 +28863,98 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_geno_int8_batch.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1574
+/* "pgenlib.pyx":1780
  * 
  * 
- *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False):             # <<<<<<<<<<<<<<
+ *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False, object allele_cts = None):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_15append_alleles_batch(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_alleles_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch *__pyx_optional_args) {
   int __pyx_v_all_phased = ((int)0);
+  PyObject *__pyx_v_allele_cts = ((PyObject *)Py_None);
   uint32_t __pyx_v_batch_size;
+  uint32_t __pyx_v_sample_ct;
+  uint32_t __pyx_v_allele_ct_limit;
   uintptr_t *__pyx_v_genovec;
   int32_t *__pyx_v_allele_codes;
+  uintptr_t *__pyx_v_patch_01_set;
+  plink2::AlleleCode *__pyx_v_patch_01_vals;
+  uintptr_t *__pyx_v_patch_10_set;
+  plink2::AlleleCode *__pyx_v_patch_10_vals;
   uint32_t __pyx_v_uii;
+  int32_t __pyx_v_observed_allele_ct;
+  uint32_t __pyx_v_write_allele_ct;
+  uint32_t __pyx_v_casted_allele_ct;
+  uint32_t __pyx_v_patch_01_ct;
+  uint32_t __pyx_v_patch_10_ct;
   plink2::PglErr __pyx_v_reterr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_batch;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_batch;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   PyObject *__pyx_t_2 = NULL;
   PyObject *__pyx_t_3 = NULL;
   PyObject *__pyx_t_4 = NULL;
   PyObject *__pyx_t_5 = NULL;
   int __pyx_t_6;
   PyObject *__pyx_t_7 = NULL;
-  uintptr_t *__pyx_t_8;
-  int __pyx_t_9;
-  uint32_t __pyx_t_10;
-  uint32_t __pyx_t_11;
+  uint32_t __pyx_t_8;
+  uintptr_t *__pyx_t_9;
+  plink2::AlleleCode *__pyx_t_10;
+  int __pyx_t_11;
   uint32_t __pyx_t_12;
-  size_t __pyx_t_13;
-  Py_ssize_t __pyx_t_14;
+  uint32_t __pyx_t_13;
+  size_t __pyx_t_14;
+  Py_ssize_t __pyx_t_15;
+  int __pyx_t_16;
+  uint32_t __pyx_t_17;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_alleles_batch", 0);
   if (__pyx_optional_args) {
     if (__pyx_optional_args->__pyx_n > 0) {
       __pyx_v_all_phased = __pyx_optional_args->all_phased;
+      if (__pyx_optional_args->__pyx_n > 1) {
+        __pyx_v_allele_cts = __pyx_optional_args->allele_cts;
+      }
     }
   }
   __pyx_pybuffer_allele_int32_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_batch.refcount = 0;
   __pyx_pybuffernd_allele_int32_batch.data = NULL;
   __pyx_pybuffernd_allele_int32_batch.rcbuffer = &__pyx_pybuffer_allele_int32_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1574, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1780, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[0].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_alleles_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1574, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_alleles_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1780, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_15append_alleles_batch)) {
         __Pyx_XDECREF(__pyx_r);
-        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_all_phased); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1574, __pyx_L1_error)
+        __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_all_phased); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1780, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_3);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
         __pyx_t_6 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_4))) {
           __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
           if (likely(__pyx_t_5)) {
@@ -26708,43 +28963,46 @@
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_4, function);
             __pyx_t_6 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_4)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32_batch), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1574, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32_batch), __pyx_t_3, __pyx_v_allele_cts};
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1780, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_4)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32_batch), __pyx_t_3};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 2+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1574, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_5, ((PyObject *)__pyx_v_allele_int32_batch), __pyx_t_3, __pyx_v_allele_cts};
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_4, __pyx_temp+1-__pyx_t_6, 3+__pyx_t_6); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1780, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_5); __pyx_t_5 = 0;
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         } else
         #endif
         {
-          __pyx_t_7 = PyTuple_New(2+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1574, __pyx_L1_error)
+          __pyx_t_7 = PyTuple_New(3+__pyx_t_6); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 1780, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_7);
           if (__pyx_t_5) {
             __Pyx_GIVEREF(__pyx_t_5); PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_5); __pyx_t_5 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32_batch));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32_batch));
           PyTuple_SET_ITEM(__pyx_t_7, 0+__pyx_t_6, ((PyObject *)__pyx_v_allele_int32_batch));
           __Pyx_GIVEREF(__pyx_t_3);
           PyTuple_SET_ITEM(__pyx_t_7, 1+__pyx_t_6, __pyx_t_3);
+          __Pyx_INCREF(__pyx_v_allele_cts);
+          __Pyx_GIVEREF(__pyx_v_allele_cts);
+          PyTuple_SET_ITEM(__pyx_t_7, 2+__pyx_t_6, __pyx_v_allele_cts);
           __pyx_t_3 = 0;
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1574, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1780, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
         }
         __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -26759,291 +29017,788 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1575
+  /* "pgenlib.pyx":1781
  * 
- *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False):
+ *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False, object allele_cts = None):
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef int32_t* allele_codes
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  */
   __pyx_v_batch_size = ((uint32_t)(__pyx_v_allele_int32_batch->dimensions[0]));
 
-  /* "pgenlib.pyx":1576
- *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False):
+  /* "pgenlib.pyx":1782
+ *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False, object allele_cts = None):
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)             # <<<<<<<<<<<<<<
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
+ *         cdef uintptr_t* genovec = self._genovec
+ */
+  __pyx_v_sample_ct = plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr);
+
+  /* "pgenlib.pyx":1783
+ *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef int32_t* allele_codes
+ */
+  __pyx_t_8 = __pyx_v_self->_allele_ct_limit;
+  __pyx_v_allele_ct_limit = __pyx_t_8;
+
+  /* "pgenlib.pyx":1784
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ */
+  __pyx_t_9 = __pyx_v_self->_genovec;
+  __pyx_v_genovec = __pyx_t_9;
+
+  /* "pgenlib.pyx":1786
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef int32_t* allele_codes
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ */
+  __pyx_t_9 = __pyx_v_self->_patch_01_set;
+  __pyx_v_patch_01_set = __pyx_t_9;
+
+  /* "pgenlib.pyx":1787
+ *         cdef int32_t* allele_codes
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ */
+  __pyx_t_10 = __pyx_v_self->_patch_01_vals;
+  __pyx_v_patch_01_vals = __pyx_t_10;
+
+  /* "pgenlib.pyx":1788
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
  *         cdef uint32_t uii
  */
-  __pyx_t_8 = __pyx_v_self->_genovec;
-  __pyx_v_genovec = __pyx_t_8;
+  __pyx_t_9 = __pyx_v_self->_patch_10_set;
+  __pyx_v_patch_10_set = __pyx_t_9;
 
-  /* "pgenlib.pyx":1580
+  /* "pgenlib.pyx":1789
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals             # <<<<<<<<<<<<<<
  *         cdef uint32_t uii
+ *         cdef int32_t observed_allele_ct
+ */
+  __pyx_t_10 = __pyx_v_self->_patch_10_vals;
+  __pyx_v_patch_10_vals = __pyx_t_10;
+
+  /* "pgenlib.pyx":1797
+ *         cdef uint32_t patch_10_ct
  *         cdef PglErr reterr
  *         if not all_phased:             # <<<<<<<<<<<<<<
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  */
-  __pyx_t_9 = ((!(__pyx_v_all_phased != 0)) != 0);
-  if (__pyx_t_9) {
+  __pyx_t_11 = ((!(__pyx_v_all_phased != 0)) != 0);
+  if (__pyx_t_11) {
 
-    /* "pgenlib.pyx":1581
+    /* "pgenlib.pyx":1798
  *         cdef PglErr reterr
  *         if not all_phased:
  *             for uii in range(batch_size):             # <<<<<<<<<<<<<<
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
  */
-    __pyx_t_10 = __pyx_v_batch_size;
-    __pyx_t_11 = __pyx_t_10;
-    for (__pyx_t_12 = 0; __pyx_t_12 < __pyx_t_11; __pyx_t_12+=1) {
-      __pyx_v_uii = __pyx_t_12;
+    __pyx_t_8 = __pyx_v_batch_size;
+    __pyx_t_12 = __pyx_t_8;
+    for (__pyx_t_13 = 0; __pyx_t_13 < __pyx_t_12; __pyx_t_13+=1) {
+      __pyx_v_uii = __pyx_t_13;
 
-      /* "pgenlib.pyx":1582
+      /* "pgenlib.pyx":1799
  *         if not all_phased:
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))             # <<<<<<<<<<<<<<
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
+ *                 if observed_allele_ct == -1:
  */
-      __pyx_t_13 = __pyx_v_uii;
-      __pyx_t_14 = 0;
+      __pyx_t_14 = __pyx_v_uii;
+      __pyx_t_15 = 0;
       __pyx_t_6 = -1;
-      if (unlikely(__pyx_t_13 >= (size_t)__pyx_pybuffernd_allele_int32_batch.diminfo[0].shape)) __pyx_t_6 = 0;
-      if (__pyx_t_14 < 0) {
-        __pyx_t_14 += __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape;
-        if (unlikely(__pyx_t_14 < 0)) __pyx_t_6 = 1;
-      } else if (unlikely(__pyx_t_14 >= __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape)) __pyx_t_6 = 1;
+      if (unlikely(__pyx_t_14 >= (size_t)__pyx_pybuffernd_allele_int32_batch.diminfo[0].shape)) __pyx_t_6 = 0;
+      if (__pyx_t_15 < 0) {
+        __pyx_t_15 += __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape;
+        if (unlikely(__pyx_t_15 < 0)) __pyx_t_6 = 1;
+      } else if (unlikely(__pyx_t_15 >= __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape)) __pyx_t_6 = 1;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1582, __pyx_L1_error)
+        __PYX_ERR(0, 1799, __pyx_L1_error)
       }
-      __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.buf, __pyx_t_13, __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides, __pyx_t_14, __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides))));
+      __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.buf, __pyx_t_14, __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides, __pyx_t_15, __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1583
+      /* "pgenlib.pyx":1800
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)             # <<<<<<<<<<<<<<
- *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
- *                 if reterr != kPglRetSuccess:
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)             # <<<<<<<<<<<<<<
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
  */
-      plink2::AlleleCodesToGenoarrUnsafe(__pyx_v_allele_codes, NULL, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_genovec, NULL, NULL);
+      __pyx_v_observed_allele_ct = plink2::ConvertMultiAlleleCodesUnsafe(__pyx_v_allele_codes, NULL, __pyx_v_sample_ct, __pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, (&__pyx_v_patch_01_ct), (&__pyx_v_patch_10_ct), NULL, NULL);
 
-      /* "pgenlib.pyx":1584
+      /* "pgenlib.pyx":1801
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)             # <<<<<<<<<<<<<<
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
+ *                 if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+      __pyx_t_11 = ((__pyx_v_observed_allele_ct == -1L) != 0);
+      if (unlikely(__pyx_t_11)) {
+
+        /* "pgenlib.pyx":1802
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:
+ */
+        __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__40, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1802, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __PYX_ERR(0, 1802, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1801
+ *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
+ *                 if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+      }
+
+      /* "pgenlib.pyx":1803
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)             # <<<<<<<<<<<<<<
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ */
+      __pyx_v_write_allele_ct = ((uint32_t)__pyx_v_observed_allele_ct);
+
+      /* "pgenlib.pyx":1804
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:
+ */
+      __pyx_t_11 = ((__pyx_v_write_allele_ct > __pyx_v_allele_ct_limit) != 0);
+      if (unlikely(__pyx_t_11)) {
+
+        /* "pgenlib.pyx":1805
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ */
+        __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__41, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1805, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __PYX_ERR(0, 1805, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1804
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:
+ */
+      }
+
+      /* "pgenlib.pyx":1806
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:             # <<<<<<<<<<<<<<
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ */
+      __pyx_t_11 = (__pyx_v_allele_cts != Py_None);
+      __pyx_t_16 = (__pyx_t_11 != 0);
+      if (__pyx_t_16) {
+
+        /* "pgenlib.pyx":1807
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]             # <<<<<<<<<<<<<<
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ */
+        __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_allele_cts, __pyx_v_uii, uint32_t, 0, __Pyx_PyInt_From_uint32_t, 0, 0, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1807, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __pyx_t_17 = __Pyx_PyInt_As_uint32_t(__pyx_t_1); if (unlikely((__pyx_t_17 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1807, __pyx_L1_error)
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __pyx_v_casted_allele_ct = __pyx_t_17;
+
+        /* "pgenlib.pyx":1808
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ */
+        __pyx_t_16 = ((__pyx_v_casted_allele_ct < __pyx_v_write_allele_ct) != 0);
+        if (unlikely(__pyx_t_16)) {
+
+          /* "pgenlib.pyx":1809
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")             # <<<<<<<<<<<<<<
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ */
+          __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__42, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1809, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_1);
+          __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+          __PYX_ERR(0, 1809, __pyx_L1_error)
+
+          /* "pgenlib.pyx":1808
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ */
+        }
+
+        /* "pgenlib.pyx":1810
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ */
+        __pyx_t_16 = ((__pyx_v_casted_allele_ct > __pyx_v_allele_ct_limit) != 0);
+        if (unlikely(__pyx_t_16)) {
+
+          /* "pgenlib.pyx":1811
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ */
+          __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__43, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1811, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_1);
+          __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+          __PYX_ERR(0, 1811, __pyx_L1_error)
+
+          /* "pgenlib.pyx":1810
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ */
+        }
+
+        /* "pgenlib.pyx":1812
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct             # <<<<<<<<<<<<<<
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                     reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ */
+        __pyx_v_write_allele_ct = __pyx_v_casted_allele_ct;
+
+        /* "pgenlib.pyx":1806
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:             # <<<<<<<<<<<<<<
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ */
+      }
+
+      /* "pgenlib.pyx":1813
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                     reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *                 else:
+ */
+      __pyx_t_11 = ((__pyx_v_patch_01_ct == 0) != 0);
+      if (__pyx_t_11) {
+      } else {
+        __pyx_t_16 = __pyx_t_11;
+        goto __pyx_L12_bool_binop_done;
+      }
+      __pyx_t_11 = ((__pyx_v_patch_10_ct == 0) != 0);
+      __pyx_t_16 = __pyx_t_11;
+      __pyx_L12_bool_binop_done:;
+      if (__pyx_t_16) {
+
+        /* "pgenlib.pyx":1814
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                     reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)             # <<<<<<<<<<<<<<
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ */
+        __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovec(__pyx_v_genovec, __pyx_v_self->_state_ptr);
+
+        /* "pgenlib.pyx":1813
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                     reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *                 else:
+ */
+        goto __pyx_L11;
+      }
+
+      /* "pgenlib.pyx":1816
+ *                     reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  */
-      __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovec(__pyx_v_genovec, __pyx_v_self->_state_ptr);
+      /*else*/ {
+        __pyx_v_reterr = plink2::SpgwAppendMultiallelicSparse(__pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, __pyx_v_write_allele_ct, __pyx_v_patch_01_ct, __pyx_v_patch_10_ct, __pyx_v_self->_state_ptr);
+      }
+      __pyx_L11:;
 
-      /* "pgenlib.pyx":1585
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+      /* "pgenlib.pyx":1817
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         else:
  */
-      __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-      if (unlikely(__pyx_t_9)) {
+      __pyx_t_16 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+      if (unlikely(__pyx_t_16)) {
 
-        /* "pgenlib.pyx":1586
- *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+        /* "pgenlib.pyx":1818
+ *                     reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         else:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  */
-        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1586, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1818, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1586, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1818, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_alleles_batch_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1586, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_alleles_batch_error, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1818, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1586, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1818, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         __Pyx_Raise(__pyx_t_2, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __PYX_ERR(0, 1586, __pyx_L1_error)
+        __PYX_ERR(0, 1818, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1585
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
- *                 reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+        /* "pgenlib.pyx":1817
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         else:
  */
       }
     }
 
-    /* "pgenlib.pyx":1580
- *         cdef uint32_t uii
+    /* "pgenlib.pyx":1797
+ *         cdef uint32_t patch_10_ct
  *         cdef PglErr reterr
  *         if not all_phased:             # <<<<<<<<<<<<<<
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  */
     goto __pyx_L3;
   }
 
-  /* "pgenlib.pyx":1588
+  /* "pgenlib.pyx":1820
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         else:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
  *             for uii in range(batch_size):
  */
   /*else*/ {
-    __pyx_t_9 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalHardcallPhasePresent) == 0) != 0);
-    if (unlikely(__pyx_t_9)) {
+    __pyx_t_16 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalHardcallPhasePresent) == 0) != 0);
+    if (unlikely(__pyx_t_16)) {
 
-      /* "pgenlib.pyx":1589
+      /* "pgenlib.pyx":1821
  *         else:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  */
-      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__27, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1589, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__44, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1821, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_Raise(__pyx_t_2, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __PYX_ERR(0, 1589, __pyx_L1_error)
+      __PYX_ERR(0, 1821, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1588
+      /* "pgenlib.pyx":1820
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         else:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
  *             for uii in range(batch_size):
  */
     }
 
-    /* "pgenlib.pyx":1590
+    /* "pgenlib.pyx":1822
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
  *             for uii in range(batch_size):             # <<<<<<<<<<<<<<
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)
  */
-    __pyx_t_10 = __pyx_v_batch_size;
-    __pyx_t_11 = __pyx_t_10;
-    for (__pyx_t_12 = 0; __pyx_t_12 < __pyx_t_11; __pyx_t_12+=1) {
-      __pyx_v_uii = __pyx_t_12;
+    __pyx_t_8 = __pyx_v_batch_size;
+    __pyx_t_12 = __pyx_t_8;
+    for (__pyx_t_13 = 0; __pyx_t_13 < __pyx_t_12; __pyx_t_13+=1) {
+      __pyx_v_uii = __pyx_t_13;
 
-      /* "pgenlib.pyx":1591
+      /* "pgenlib.pyx":1823
  *                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))             # <<<<<<<<<<<<<<
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)
+ *                 if observed_allele_ct == -1:
  */
-      __pyx_t_13 = __pyx_v_uii;
-      __pyx_t_14 = 0;
+      __pyx_t_14 = __pyx_v_uii;
+      __pyx_t_15 = 0;
       __pyx_t_6 = -1;
-      if (unlikely(__pyx_t_13 >= (size_t)__pyx_pybuffernd_allele_int32_batch.diminfo[0].shape)) __pyx_t_6 = 0;
-      if (__pyx_t_14 < 0) {
-        __pyx_t_14 += __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape;
-        if (unlikely(__pyx_t_14 < 0)) __pyx_t_6 = 1;
-      } else if (unlikely(__pyx_t_14 >= __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape)) __pyx_t_6 = 1;
+      if (unlikely(__pyx_t_14 >= (size_t)__pyx_pybuffernd_allele_int32_batch.diminfo[0].shape)) __pyx_t_6 = 0;
+      if (__pyx_t_15 < 0) {
+        __pyx_t_15 += __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape;
+        if (unlikely(__pyx_t_15 < 0)) __pyx_t_6 = 1;
+      } else if (unlikely(__pyx_t_15 >= __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape)) __pyx_t_6 = 1;
       if (unlikely(__pyx_t_6 != -1)) {
         __Pyx_RaiseBufferIndexError(__pyx_t_6);
-        __PYX_ERR(0, 1591, __pyx_L1_error)
+        __PYX_ERR(0, 1823, __pyx_L1_error)
       }
-      __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.buf, __pyx_t_13, __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides, __pyx_t_14, __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides))));
+      __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.buf, __pyx_t_14, __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides, __pyx_t_15, __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides))));
 
-      /* "pgenlib.pyx":1592
+      /* "pgenlib.pyx":1824
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)             # <<<<<<<<<<<<<<
- *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
- *                 if reterr != kPglRetSuccess:
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)             # <<<<<<<<<<<<<<
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
  */
-      plink2::AlleleCodesToGenoarrUnsafe(__pyx_v_allele_codes, NULL, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_genovec, NULL, __pyx_v_self->_phaseinfo);
+      __pyx_v_observed_allele_ct = plink2::ConvertMultiAlleleCodesUnsafe(__pyx_v_allele_codes, NULL, __pyx_v_sample_ct, __pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, (&__pyx_v_patch_01_ct), (&__pyx_v_patch_10_ct), NULL, __pyx_v_self->_phaseinfo);
 
-      /* "pgenlib.pyx":1593
+      /* "pgenlib.pyx":1825
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)
+ *                 if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+      __pyx_t_16 = ((__pyx_v_observed_allele_ct == -1L) != 0);
+      if (unlikely(__pyx_t_16)) {
+
+        /* "pgenlib.pyx":1826
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:
+ */
+        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__40, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1826, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __PYX_ERR(0, 1826, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1825
+ *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)
+ *                 if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+      }
+
+      /* "pgenlib.pyx":1827
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)             # <<<<<<<<<<<<<<
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ */
+      __pyx_v_write_allele_ct = ((uint32_t)__pyx_v_observed_allele_ct);
+
+      /* "pgenlib.pyx":1828
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:
+ */
+      __pyx_t_16 = ((__pyx_v_write_allele_ct > __pyx_v_allele_ct_limit) != 0);
+      if (unlikely(__pyx_t_16)) {
+
+        /* "pgenlib.pyx":1829
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ */
+        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__41, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1829, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __PYX_ERR(0, 1829, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1828
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:
+ */
+      }
+
+      /* "pgenlib.pyx":1830
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:             # <<<<<<<<<<<<<<
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ */
+      __pyx_t_16 = (__pyx_v_allele_cts != Py_None);
+      __pyx_t_11 = (__pyx_t_16 != 0);
+      if (__pyx_t_11) {
+
+        /* "pgenlib.pyx":1831
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]             # <<<<<<<<<<<<<<
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ */
+        __pyx_t_2 = __Pyx_GetItemInt(__pyx_v_allele_cts, __pyx_v_uii, uint32_t, 0, __Pyx_PyInt_From_uint32_t, 0, 0, 1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1831, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __pyx_t_17 = __Pyx_PyInt_As_uint32_t(__pyx_t_2); if (unlikely((__pyx_t_17 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1831, __pyx_L1_error)
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __pyx_v_casted_allele_ct = __pyx_t_17;
+
+        /* "pgenlib.pyx":1832
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ */
+        __pyx_t_11 = ((__pyx_v_casted_allele_ct < __pyx_v_write_allele_ct) != 0);
+        if (unlikely(__pyx_t_11)) {
+
+          /* "pgenlib.pyx":1833
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")             # <<<<<<<<<<<<<<
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ */
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__42, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1833, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_2);
+          __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+          __PYX_ERR(0, 1833, __pyx_L1_error)
+
+          /* "pgenlib.pyx":1832
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ */
+        }
+
+        /* "pgenlib.pyx":1834
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ */
+        __pyx_t_11 = ((__pyx_v_casted_allele_ct > __pyx_v_allele_ct_limit) != 0);
+        if (unlikely(__pyx_t_11)) {
+
+          /* "pgenlib.pyx":1835
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ */
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__43, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1835, __pyx_L1_error)
+          __Pyx_GOTREF(__pyx_t_2);
+          __Pyx_Raise(__pyx_t_2, 0, 0, 0);
+          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+          __PYX_ERR(0, 1835, __pyx_L1_error)
+
+          /* "pgenlib.pyx":1834
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ */
+        }
+
+        /* "pgenlib.pyx":1836
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct             # <<<<<<<<<<<<<<
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                     reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ */
+        __pyx_v_write_allele_ct = __pyx_v_casted_allele_ct;
+
+        /* "pgenlib.pyx":1830
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *                 if allele_cts is not None:             # <<<<<<<<<<<<<<
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ */
+      }
+
+      /* "pgenlib.pyx":1837
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                     reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ *                 else:
+ */
+      __pyx_t_16 = ((__pyx_v_patch_01_ct == 0) != 0);
+      if (__pyx_t_16) {
+      } else {
+        __pyx_t_11 = __pyx_t_16;
+        goto __pyx_L24_bool_binop_done;
+      }
+      __pyx_t_16 = ((__pyx_v_patch_10_ct == 0) != 0);
+      __pyx_t_11 = __pyx_t_16;
+      __pyx_L24_bool_binop_done:;
+      if (__pyx_t_11) {
+
+        /* "pgenlib.pyx":1838
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                     reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, self._phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ */
+        __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, NULL, __pyx_v_self->_phaseinfo, __pyx_v_self->_state_ptr);
+
+        /* "pgenlib.pyx":1837
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                     reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ *                 else:
+ */
+        goto __pyx_L23;
+      }
+
+      /* "pgenlib.pyx":1840
+ *                     reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, self._phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  */
-      __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, NULL, __pyx_v_self->_phaseinfo, __pyx_v_self->_state_ptr);
+      /*else*/ {
+        __pyx_v_reterr = plink2::SpgwAppendMultiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, NULL, __pyx_v_self->_phaseinfo, __pyx_v_write_allele_ct, __pyx_v_patch_01_ct, __pyx_v_patch_10_ct, __pyx_v_self->_state_ptr);
+      }
+      __pyx_L23:;
 
-      /* "pgenlib.pyx":1594
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+      /* "pgenlib.pyx":1841
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, self._phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         return
  */
-      __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-      if (unlikely(__pyx_t_9)) {
+      __pyx_t_11 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+      if (unlikely(__pyx_t_11)) {
 
-        /* "pgenlib.pyx":1595
- *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+        /* "pgenlib.pyx":1842
+ *                     reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, self._phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-        __pyx_t_2 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1595, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1595, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_alleles_batch_error, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1595, __pyx_L1_error)
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_alleles_batch_error, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1595, __pyx_L1_error)
+        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1842, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_1);
         __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
         __Pyx_Raise(__pyx_t_1, 0, 0, 0);
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-        __PYX_ERR(0, 1595, __pyx_L1_error)
+        __PYX_ERR(0, 1842, __pyx_L1_error)
 
-        /* "pgenlib.pyx":1594
- *                 AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+        /* "pgenlib.pyx":1841
+ *                 else:
+ *                     reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, self._phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         return
  */
       }
     }
   }
   __pyx_L3:;
 
-  /* "pgenlib.pyx":1596
+  /* "pgenlib.pyx":1843
  *                 if reterr != kPglRetSuccess:
  *                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1574
+  /* "pgenlib.pyx":1780
  * 
  * 
- *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False):             # <<<<<<<<<<<<<<
+ *     cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False, object allele_cts = None):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
   __Pyx_XDECREF(__pyx_t_2);
   __Pyx_XDECREF(__pyx_t_3);
@@ -27068,27 +29823,31 @@
 }
 
 /* Python wrapper */
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_15append_alleles_batch(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_15append_alleles_batch(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
   PyArrayObject *__pyx_v_allele_int32_batch = 0;
   int __pyx_v_all_phased;
+  PyObject *__pyx_v_allele_cts = 0;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_alleles_batch (wrapper)", 0);
   {
-    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32_batch,&__pyx_n_s_all_phased,0};
-    PyObject* values[2] = {0,0};
+    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32_batch,&__pyx_n_s_all_phased,&__pyx_n_s_allele_cts,0};
+    PyObject* values[3] = {0,0,0};
+    values[2] = ((PyObject *)Py_None);
     if (unlikely(__pyx_kwds)) {
       Py_ssize_t kw_args;
       const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
       switch (pos_args) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         CYTHON_FALLTHROUGH;
         case  0: break;
         default: goto __pyx_L5_argtuple_error;
       }
@@ -27099,55 +29858,64 @@
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (kw_args > 0) {
           PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_all_phased);
           if (value) { values[1] = value; kw_args--; }
         }
+        CYTHON_FALLTHROUGH;
+        case  2:
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_cts);
+          if (value) { values[2] = value; kw_args--; }
+        }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_alleles_batch") < 0)) __PYX_ERR(0, 1574, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_alleles_batch") < 0)) __PYX_ERR(0, 1780, __pyx_L3_error)
       }
     } else {
       switch (PyTuple_GET_SIZE(__pyx_args)) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         break;
         default: goto __pyx_L5_argtuple_error;
       }
     }
     __pyx_v_allele_int32_batch = ((PyArrayObject *)values[0]);
     if (values[1]) {
-      __pyx_v_all_phased = __Pyx_PyObject_IsTrue(values[1]); if (unlikely((__pyx_v_all_phased == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1574, __pyx_L3_error)
+      __pyx_v_all_phased = __Pyx_PyObject_IsTrue(values[1]); if (unlikely((__pyx_v_all_phased == (int)-1) && PyErr_Occurred())) __PYX_ERR(0, 1780, __pyx_L3_error)
     } else {
       __pyx_v_all_phased = ((int)0);
     }
+    __pyx_v_allele_cts = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("append_alleles_batch", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1574, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("append_alleles_batch", 0, 1, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1780, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenWriter.append_alleles_batch", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_batch), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_batch", 0))) __PYX_ERR(0, 1574, __pyx_L1_error)
-  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_14append_alleles_batch(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32_batch, __pyx_v_all_phased);
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_batch), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_batch", 0))) __PYX_ERR(0, 1780, __pyx_L1_error)
+  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_14append_alleles_batch(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32_batch, __pyx_v_all_phased, __pyx_v_allele_cts);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_14append_alleles_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, int __pyx_v_all_phased) {
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_14append_alleles_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, int __pyx_v_all_phased, PyObject *__pyx_v_allele_cts) {
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_batch;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_batch;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch __pyx_t_2;
   int __pyx_lineno = 0;
@@ -27156,21 +29924,22 @@
   __Pyx_RefNannySetupContext("append_alleles_batch", 0);
   __pyx_pybuffer_allele_int32_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_batch.refcount = 0;
   __pyx_pybuffernd_allele_int32_batch.data = NULL;
   __pyx_pybuffernd_allele_int32_batch.rcbuffer = &__pyx_pybuffer_allele_int32_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1574, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1780, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[0].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_2.__pyx_n = 1;
+  __pyx_t_2.__pyx_n = 2;
   __pyx_t_2.all_phased = __pyx_v_all_phased;
-  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenWriter->append_alleles_batch(__pyx_v_self, __pyx_v_allele_int32_batch, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1574, __pyx_L1_error)
+  __pyx_t_2.allele_cts = __pyx_v_allele_cts;
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenWriter->append_alleles_batch(__pyx_v_self, __pyx_v_allele_int32_batch, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1780, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -27188,83 +29957,103 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1599
+/* "pgenlib.pyx":1846
  * 
  * 
- *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):             # <<<<<<<<<<<<<<
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_17append_partially_phased_batch(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
-static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch, int __pyx_skip_dispatch) {
+static PyObject *__pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch *__pyx_optional_args) {
+  PyObject *__pyx_v_allele_cts = ((PyObject *)Py_None);
   uint32_t __pyx_v_batch_size;
+  uint32_t __pyx_v_sample_ct;
+  uint32_t __pyx_v_allele_ct_limit;
   uintptr_t *__pyx_v_genovec;
+  uintptr_t *__pyx_v_patch_01_set;
+  plink2::AlleleCode *__pyx_v_patch_01_vals;
+  uintptr_t *__pyx_v_patch_10_set;
+  plink2::AlleleCode *__pyx_v_patch_10_vals;
   uintptr_t *__pyx_v_phasepresent_buf;
   uintptr_t *__pyx_v_phaseinfo;
   int32_t *__pyx_v_allele_codes;
   unsigned char *__pyx_v_phasepresent_bytes;
   uint32_t __pyx_v_uii;
+  int32_t __pyx_v_observed_allele_ct;
+  uint32_t __pyx_v_write_allele_ct;
+  uint32_t __pyx_v_casted_allele_ct;
+  uint32_t __pyx_v_patch_01_ct;
+  uint32_t __pyx_v_patch_10_ct;
   plink2::PglErr __pyx_v_reterr;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_batch;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_batch;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent_batch;
   __Pyx_Buffer __pyx_pybuffer_phasepresent_batch;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   PyObject *__pyx_t_2 = NULL;
   PyObject *__pyx_t_3 = NULL;
   PyObject *__pyx_t_4 = NULL;
   int __pyx_t_5;
   PyObject *__pyx_t_6 = NULL;
   int __pyx_t_7;
-  uintptr_t *__pyx_t_8;
-  uint32_t __pyx_t_9;
-  uint32_t __pyx_t_10;
+  uint32_t __pyx_t_8;
+  uintptr_t *__pyx_t_9;
+  plink2::AlleleCode *__pyx_t_10;
   uint32_t __pyx_t_11;
-  size_t __pyx_t_12;
-  Py_ssize_t __pyx_t_13;
+  uint32_t __pyx_t_12;
+  size_t __pyx_t_13;
+  Py_ssize_t __pyx_t_14;
+  int __pyx_t_15;
+  uint32_t __pyx_t_16;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_partially_phased_batch", 0);
+  if (__pyx_optional_args) {
+    if (__pyx_optional_args->__pyx_n > 0) {
+      __pyx_v_allele_cts = __pyx_optional_args->allele_cts;
+    }
+  }
   __pyx_pybuffer_allele_int32_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_batch.refcount = 0;
   __pyx_pybuffernd_allele_int32_batch.data = NULL;
   __pyx_pybuffernd_allele_int32_batch.rcbuffer = &__pyx_pybuffer_allele_int32_batch;
   __pyx_pybuffer_phasepresent_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_batch.refcount = 0;
   __pyx_pybuffernd_phasepresent_batch.data = NULL;
   __pyx_pybuffernd_phasepresent_batch.rcbuffer = &__pyx_pybuffer_phasepresent_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1599, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1846, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[0].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[1];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1599, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1846, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_batch.diminfo[0].strides = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_batch.diminfo[0].shape = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_phasepresent_batch.diminfo[1].strides = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_phasepresent_batch.diminfo[1].shape = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.shape[1];
   /* Check if called by wrapper */
   if (unlikely(__pyx_skip_dispatch)) ;
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_partially_phased_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1599, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_partially_phased_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1846, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_17append_partially_phased_batch)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         __pyx_t_5 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
@@ -27275,41 +30064,44 @@
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
             __pyx_t_5 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_3)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32_batch), ((PyObject *)__pyx_v_phasepresent_batch)};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 2+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1599, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32_batch), ((PyObject *)__pyx_v_phasepresent_batch), __pyx_v_allele_cts};
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1846, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_3)) {
-          PyObject *__pyx_temp[3] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32_batch), ((PyObject *)__pyx_v_phasepresent_batch)};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 2+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1599, __pyx_L1_error)
+          PyObject *__pyx_temp[4] = {__pyx_t_4, ((PyObject *)__pyx_v_allele_int32_batch), ((PyObject *)__pyx_v_phasepresent_batch), __pyx_v_allele_cts};
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1846, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         {
-          __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1599, __pyx_L1_error)
+          __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1846, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_6);
           if (__pyx_t_4) {
             __Pyx_GIVEREF(__pyx_t_4); PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __pyx_t_4 = NULL;
           }
           __Pyx_INCREF(((PyObject *)__pyx_v_allele_int32_batch));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_allele_int32_batch));
           PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_allele_int32_batch));
           __Pyx_INCREF(((PyObject *)__pyx_v_phasepresent_batch));
           __Pyx_GIVEREF(((PyObject *)__pyx_v_phasepresent_batch));
           PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_phasepresent_batch));
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1599, __pyx_L1_error)
+          __Pyx_INCREF(__pyx_v_allele_cts);
+          __Pyx_GIVEREF(__pyx_v_allele_cts);
+          PyTuple_SET_ITEM(__pyx_t_6, 2+__pyx_t_5, __pyx_v_allele_cts);
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1846, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         }
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -27324,214 +30116,492 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1600
+  /* "pgenlib.pyx":1847
  * 
- *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
  */
   __pyx_t_7 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalHardcallPhasePresent) == 0) != 0);
   if (unlikely(__pyx_t_7)) {
 
-    /* "pgenlib.pyx":1601
- *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):
+    /* "pgenlib.pyx":1848
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__28, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1601, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__45, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1848, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 1601, __pyx_L1_error)
+    __PYX_ERR(0, 1848, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1600
+    /* "pgenlib.pyx":1847
  * 
- *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
  */
   }
 
-  /* "pgenlib.pyx":1602
+  /* "pgenlib.pyx":1849
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]             # <<<<<<<<<<<<<<
- *         cdef uintptr_t* genovec = self._genovec
- *         cdef uintptr_t* phasepresent_buf = self._phasepresent
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  */
   __pyx_v_batch_size = ((uint32_t)(__pyx_v_allele_int32_batch->dimensions[0]));
 
-  /* "pgenlib.pyx":1603
+  /* "pgenlib.pyx":1850
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)             # <<<<<<<<<<<<<<
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
+ *         cdef uintptr_t* genovec = self._genovec
+ */
+  __pyx_v_sample_ct = plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr);
+
+  /* "pgenlib.pyx":1851
+ *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ */
+  __pyx_t_8 = __pyx_v_self->_allele_ct_limit;
+  __pyx_v_allele_ct_limit = __pyx_t_8;
+
+  /* "pgenlib.pyx":1852
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ */
+  __pyx_t_9 = __pyx_v_self->_genovec;
+  __pyx_v_genovec = __pyx_t_9;
+
+  /* "pgenlib.pyx":1853
+ *         cdef uint32_t allele_ct_limit = self._allele_ct_limit
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ */
+  __pyx_t_9 = __pyx_v_self->_patch_01_set;
+  __pyx_v_patch_01_set = __pyx_t_9;
+
+  /* "pgenlib.pyx":1854
+ *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals             # <<<<<<<<<<<<<<
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ */
+  __pyx_t_10 = __pyx_v_self->_patch_01_vals;
+  __pyx_v_patch_01_vals = __pyx_t_10;
+
+  /* "pgenlib.pyx":1855
+ *         cdef uintptr_t* patch_01_set = self._patch_01_set
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set             # <<<<<<<<<<<<<<
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
+ *         cdef uintptr_t* phasepresent_buf = self._phasepresent
+ */
+  __pyx_t_9 = __pyx_v_self->_patch_10_set;
+  __pyx_v_patch_10_set = __pyx_t_9;
+
+  /* "pgenlib.pyx":1856
+ *         cdef AlleleCode* patch_01_vals = self._patch_01_vals
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent
  *         cdef uintptr_t* phaseinfo = self._phaseinfo
  */
-  __pyx_t_8 = __pyx_v_self->_genovec;
-  __pyx_v_genovec = __pyx_t_8;
+  __pyx_t_10 = __pyx_v_self->_patch_10_vals;
+  __pyx_v_patch_10_vals = __pyx_t_10;
 
-  /* "pgenlib.pyx":1604
- *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":1857
+ *         cdef uintptr_t* patch_10_set = self._patch_10_set
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* phaseinfo = self._phaseinfo
  *         cdef int32_t* allele_codes
  */
-  __pyx_t_8 = __pyx_v_self->_phasepresent;
-  __pyx_v_phasepresent_buf = __pyx_t_8;
+  __pyx_t_9 = __pyx_v_self->_phasepresent;
+  __pyx_v_phasepresent_buf = __pyx_t_9;
 
-  /* "pgenlib.pyx":1605
- *         cdef uintptr_t* genovec = self._genovec
+  /* "pgenlib.pyx":1858
+ *         cdef AlleleCode* patch_10_vals = self._patch_10_vals
  *         cdef uintptr_t* phasepresent_buf = self._phasepresent
  *         cdef uintptr_t* phaseinfo = self._phaseinfo             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes
  *         cdef unsigned char* phasepresent_bytes
  */
-  __pyx_t_8 = __pyx_v_self->_phaseinfo;
-  __pyx_v_phaseinfo = __pyx_t_8;
+  __pyx_t_9 = __pyx_v_self->_phaseinfo;
+  __pyx_v_phaseinfo = __pyx_t_9;
 
-  /* "pgenlib.pyx":1610
- *         cdef uint32_t uii
+  /* "pgenlib.pyx":1868
+ *         cdef uint32_t patch_10_ct
  *         cdef PglErr reterr
  *         for uii in range(batch_size):             # <<<<<<<<<<<<<<
  *             allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  *             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))
  */
-  __pyx_t_9 = __pyx_v_batch_size;
-  __pyx_t_10 = __pyx_t_9;
-  for (__pyx_t_11 = 0; __pyx_t_11 < __pyx_t_10; __pyx_t_11+=1) {
-    __pyx_v_uii = __pyx_t_11;
+  __pyx_t_8 = __pyx_v_batch_size;
+  __pyx_t_11 = __pyx_t_8;
+  for (__pyx_t_12 = 0; __pyx_t_12 < __pyx_t_11; __pyx_t_12+=1) {
+    __pyx_v_uii = __pyx_t_12;
 
-    /* "pgenlib.pyx":1611
+    /* "pgenlib.pyx":1869
  *         cdef PglErr reterr
  *         for uii in range(batch_size):
  *             allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))             # <<<<<<<<<<<<<<
  *             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))
- *             AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
  */
-    __pyx_t_12 = __pyx_v_uii;
-    __pyx_t_13 = 0;
+    __pyx_t_13 = __pyx_v_uii;
+    __pyx_t_14 = 0;
     __pyx_t_5 = -1;
-    if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_allele_int32_batch.diminfo[0].shape)) __pyx_t_5 = 0;
-    if (__pyx_t_13 < 0) {
-      __pyx_t_13 += __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape;
-      if (unlikely(__pyx_t_13 < 0)) __pyx_t_5 = 1;
-    } else if (unlikely(__pyx_t_13 >= __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape)) __pyx_t_5 = 1;
+    if (unlikely(__pyx_t_13 >= (size_t)__pyx_pybuffernd_allele_int32_batch.diminfo[0].shape)) __pyx_t_5 = 0;
+    if (__pyx_t_14 < 0) {
+      __pyx_t_14 += __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape;
+      if (unlikely(__pyx_t_14 < 0)) __pyx_t_5 = 1;
+    } else if (unlikely(__pyx_t_14 >= __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape)) __pyx_t_5 = 1;
     if (unlikely(__pyx_t_5 != -1)) {
       __Pyx_RaiseBufferIndexError(__pyx_t_5);
-      __PYX_ERR(0, 1611, __pyx_L1_error)
+      __PYX_ERR(0, 1869, __pyx_L1_error)
     }
-    __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides, __pyx_t_13, __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides))));
+    __pyx_v_allele_codes = ((int32_t *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_int32_t *, __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.buf, __pyx_t_13, __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides, __pyx_t_14, __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides))));
 
-    /* "pgenlib.pyx":1612
+    /* "pgenlib.pyx":1870
  *         for uii in range(batch_size):
  *             allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  *             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))             # <<<<<<<<<<<<<<
- *             AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *             if observed_allele_ct == -1:
  */
-    __pyx_t_12 = __pyx_v_uii;
-    __pyx_t_13 = 0;
+    __pyx_t_13 = __pyx_v_uii;
+    __pyx_t_14 = 0;
     __pyx_t_5 = -1;
-    if (unlikely(__pyx_t_12 >= (size_t)__pyx_pybuffernd_phasepresent_batch.diminfo[0].shape)) __pyx_t_5 = 0;
-    if (__pyx_t_13 < 0) {
-      __pyx_t_13 += __pyx_pybuffernd_phasepresent_batch.diminfo[1].shape;
-      if (unlikely(__pyx_t_13 < 0)) __pyx_t_5 = 1;
-    } else if (unlikely(__pyx_t_13 >= __pyx_pybuffernd_phasepresent_batch.diminfo[1].shape)) __pyx_t_5 = 1;
+    if (unlikely(__pyx_t_13 >= (size_t)__pyx_pybuffernd_phasepresent_batch.diminfo[0].shape)) __pyx_t_5 = 0;
+    if (__pyx_t_14 < 0) {
+      __pyx_t_14 += __pyx_pybuffernd_phasepresent_batch.diminfo[1].shape;
+      if (unlikely(__pyx_t_14 < 0)) __pyx_t_5 = 1;
+    } else if (unlikely(__pyx_t_14 >= __pyx_pybuffernd_phasepresent_batch.diminfo[1].shape)) __pyx_t_5 = 1;
     if (unlikely(__pyx_t_5 != -1)) {
       __Pyx_RaiseBufferIndexError(__pyx_t_5);
-      __PYX_ERR(0, 1612, __pyx_L1_error)
+      __PYX_ERR(0, 1870, __pyx_L1_error)
     }
-    __pyx_v_phasepresent_bytes = ((unsigned char *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_uint8_t *, __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.buf, __pyx_t_12, __pyx_pybuffernd_phasepresent_batch.diminfo[0].strides, __pyx_t_13, __pyx_pybuffernd_phasepresent_batch.diminfo[1].strides))));
+    __pyx_v_phasepresent_bytes = ((unsigned char *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_uint8_t *, __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.buf, __pyx_t_13, __pyx_pybuffernd_phasepresent_batch.diminfo[0].strides, __pyx_t_14, __pyx_pybuffernd_phasepresent_batch.diminfo[1].strides))));
 
-    /* "pgenlib.pyx":1613
+    /* "pgenlib.pyx":1871
  *             allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  *             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))
- *             AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)             # <<<<<<<<<<<<<<
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
- *             if reterr != kPglRetSuccess:
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)             # <<<<<<<<<<<<<<
+ *             if observed_allele_ct == -1:
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
  */
-    plink2::AlleleCodesToGenoarrUnsafe(__pyx_v_allele_codes, __pyx_v_phasepresent_bytes, plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), __pyx_v_genovec, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo);
+    __pyx_v_observed_allele_ct = plink2::ConvertMultiAlleleCodesUnsafe(__pyx_v_allele_codes, __pyx_v_phasepresent_bytes, __pyx_v_sample_ct, __pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, (&__pyx_v_patch_01_ct), (&__pyx_v_patch_10_ct), __pyx_v_phasepresent_buf, __pyx_v_phaseinfo);
 
-    /* "pgenlib.pyx":1614
+    /* "pgenlib.pyx":1872
  *             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))
- *             AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *             if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+    __pyx_t_7 = ((__pyx_v_observed_allele_ct == -1L) != 0);
+    if (unlikely(__pyx_t_7)) {
+
+      /* "pgenlib.pyx":1873
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *             if observed_allele_ct == -1:
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ *             if write_allele_ct > allele_ct_limit:
+ */
+      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__46, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1873, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __PYX_ERR(0, 1873, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1872
+ *             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *             if observed_allele_ct == -1:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ */
+    }
+
+    /* "pgenlib.pyx":1874
+ *             if observed_allele_ct == -1:
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)             # <<<<<<<<<<<<<<
+ *             if write_allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ */
+    __pyx_v_write_allele_ct = ((uint32_t)__pyx_v_observed_allele_ct);
+
+    /* "pgenlib.pyx":1875
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ *             if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *             if allele_cts is not None:
+ */
+    __pyx_t_7 = ((__pyx_v_write_allele_ct > __pyx_v_allele_ct_limit) != 0);
+    if (unlikely(__pyx_t_7)) {
+
+      /* "pgenlib.pyx":1876
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ *             if write_allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *             if allele_cts is not None:
+ *                 casted_allele_ct = allele_cts[uii]
+ */
+      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__47, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1876, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __PYX_ERR(0, 1876, __pyx_L1_error)
+
+      /* "pgenlib.pyx":1875
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ *             if write_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *             if allele_cts is not None:
+ */
+    }
+
+    /* "pgenlib.pyx":1877
+ *             if write_allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *             if allele_cts is not None:             # <<<<<<<<<<<<<<
+ *                 casted_allele_ct = allele_cts[uii]
+ *                 if casted_allele_ct < write_allele_ct:
+ */
+    __pyx_t_7 = (__pyx_v_allele_cts != Py_None);
+    __pyx_t_15 = (__pyx_t_7 != 0);
+    if (__pyx_t_15) {
+
+      /* "pgenlib.pyx":1878
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *             if allele_cts is not None:
+ *                 casted_allele_ct = allele_cts[uii]             # <<<<<<<<<<<<<<
+ *                 if casted_allele_ct < write_allele_ct:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ */
+      __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_allele_cts, __pyx_v_uii, uint32_t, 0, __Pyx_PyInt_From_uint32_t, 0, 0, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1878, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __pyx_t_16 = __Pyx_PyInt_As_uint32_t(__pyx_t_1); if (unlikely((__pyx_t_16 == ((uint32_t)-1)) && PyErr_Occurred())) __PYX_ERR(0, 1878, __pyx_L1_error)
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __pyx_v_casted_allele_ct = __pyx_t_16;
+
+      /* "pgenlib.pyx":1879
+ *             if allele_cts is not None:
+ *                 casted_allele_ct = allele_cts[uii]
+ *                 if casted_allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                 if casted_allele_ct > allele_ct_limit:
+ */
+      __pyx_t_15 = ((__pyx_v_casted_allele_ct < __pyx_v_write_allele_ct) != 0);
+      if (unlikely(__pyx_t_15)) {
+
+        /* "pgenlib.pyx":1880
+ *                 casted_allele_ct = allele_cts[uii]
+ *                 if casted_allele_ct < write_allele_ct:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")             # <<<<<<<<<<<<<<
+ *                 if casted_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ */
+        __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__42, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1880, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __PYX_ERR(0, 1880, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1879
+ *             if allele_cts is not None:
+ *                 casted_allele_ct = allele_cts[uii]
+ *                 if casted_allele_ct < write_allele_ct:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                 if casted_allele_ct > allele_ct_limit:
+ */
+      }
+
+      /* "pgenlib.pyx":1881
+ *                 if casted_allele_ct < write_allele_ct:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                 if casted_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                 write_allele_ct = casted_allele_ct
+ */
+      __pyx_t_15 = ((__pyx_v_casted_allele_ct > __pyx_v_allele_ct_limit) != 0);
+      if (unlikely(__pyx_t_15)) {
+
+        /* "pgenlib.pyx":1882
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                 if casted_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *                 write_allele_ct = casted_allele_ct
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ */
+        __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__43, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1882, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_1);
+        __Pyx_Raise(__pyx_t_1, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+        __PYX_ERR(0, 1882, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1881
+ *                 if casted_allele_ct < write_allele_ct:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                 if casted_allele_ct > allele_ct_limit:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                 write_allele_ct = casted_allele_ct
+ */
+      }
+
+      /* "pgenlib.pyx":1883
+ *                 if casted_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                 write_allele_ct = casted_allele_ct             # <<<<<<<<<<<<<<
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ */
+      __pyx_v_write_allele_ct = __pyx_v_casted_allele_ct;
+
+      /* "pgenlib.pyx":1877
+ *             if write_allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+ *             if allele_cts is not None:             # <<<<<<<<<<<<<<
+ *                 casted_allele_ct = allele_cts[uii]
+ *                 if casted_allele_ct < write_allele_ct:
+ */
+    }
+
+    /* "pgenlib.pyx":1884
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                 write_allele_ct = casted_allele_ct
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *             else:
+ */
+    __pyx_t_7 = ((__pyx_v_patch_01_ct == 0) != 0);
+    if (__pyx_t_7) {
+    } else {
+      __pyx_t_15 = __pyx_t_7;
+      goto __pyx_L12_bool_binop_done;
+    }
+    __pyx_t_7 = ((__pyx_v_patch_10_ct == 0) != 0);
+    __pyx_t_15 = __pyx_t_7;
+    __pyx_L12_bool_binop_done:;
+    if (__pyx_t_15) {
+
+      /* "pgenlib.pyx":1885
+ *                 write_allele_ct = casted_allele_ct
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)             # <<<<<<<<<<<<<<
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
+ */
+      __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo, __pyx_v_self->_state_ptr);
+
+      /* "pgenlib.pyx":1884
+ *                     raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ *                 write_allele_ct = casted_allele_ct
+ *             if (patch_01_ct == 0) and (patch_10_ct == 0):             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *             else:
+ */
+      goto __pyx_L11;
+    }
+
+    /* "pgenlib.pyx":1887
+ *                 reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_partially_phased_batch() error " + str(reterr))
  */
-    __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo, __pyx_v_self->_state_ptr);
+    /*else*/ {
+      __pyx_v_reterr = plink2::SpgwAppendMultiallelicGenovecHphase(__pyx_v_genovec, __pyx_v_patch_01_set, __pyx_v_patch_01_vals, __pyx_v_patch_10_set, __pyx_v_patch_10_vals, __pyx_v_phasepresent_buf, __pyx_v_phaseinfo, __pyx_v_write_allele_ct, __pyx_v_patch_01_ct, __pyx_v_patch_10_ct, __pyx_v_self->_state_ptr);
+    }
+    __pyx_L11:;
 
-    /* "pgenlib.pyx":1615
- *             AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+    /* "pgenlib.pyx":1888
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_partially_phased_batch() error " + str(reterr))
  *         return
  */
-    __pyx_t_7 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
-    if (unlikely(__pyx_t_7)) {
+    __pyx_t_15 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+    if (unlikely(__pyx_t_15)) {
 
-      /* "pgenlib.pyx":1616
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+      /* "pgenlib.pyx":1889
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_partially_phased_batch() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1616, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1889, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1616, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1889, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_partially_phased_batch_er, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1616, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_partially_phased_batch_er, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1889, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1616, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1889, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       __Pyx_Raise(__pyx_t_2, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __PYX_ERR(0, 1616, __pyx_L1_error)
+      __PYX_ERR(0, 1889, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1615
- *             AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+      /* "pgenlib.pyx":1888
+ *             else:
+ *                 reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_partially_phased_batch() error " + str(reterr))
  *         return
  */
     }
   }
 
-  /* "pgenlib.pyx":1617
+  /* "pgenlib.pyx":1890
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_partially_phased_batch() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1599
+  /* "pgenlib.pyx":1846
  * 
  * 
- *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):             # <<<<<<<<<<<<<<
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
@@ -27559,27 +30629,31 @@
 }
 
 /* Python wrapper */
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_17append_partially_phased_batch(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_17append_partially_phased_batch(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
   PyArrayObject *__pyx_v_allele_int32_batch = 0;
   PyArrayObject *__pyx_v_phasepresent_batch = 0;
+  PyObject *__pyx_v_allele_cts = 0;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_partially_phased_batch (wrapper)", 0);
   {
-    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32_batch,&__pyx_n_s_phasepresent_batch,0};
-    PyObject* values[2] = {0,0};
+    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_allele_int32_batch,&__pyx_n_s_phasepresent_batch,&__pyx_n_s_allele_cts,0};
+    PyObject* values[3] = {0,0,0};
+    values[2] = ((PyObject *)Py_None);
     if (unlikely(__pyx_kwds)) {
       Py_ssize_t kw_args;
       const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
       switch (pos_args) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
         case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
         CYTHON_FALLTHROUGH;
         case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
         CYTHON_FALLTHROUGH;
         case  0: break;
         default: goto __pyx_L5_argtuple_error;
       }
@@ -27588,82 +30662,96 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_int32_batch)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_phasepresent_batch)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("append_partially_phased_batch", 1, 2, 2, 1); __PYX_ERR(0, 1599, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("append_partially_phased_batch", 0, 2, 3, 1); __PYX_ERR(0, 1846, __pyx_L3_error)
+        }
+        CYTHON_FALLTHROUGH;
+        case  2:
+        if (kw_args > 0) {
+          PyObject* value = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_allele_cts);
+          if (value) { values[2] = value; kw_args--; }
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_partially_phased_batch") < 0)) __PYX_ERR(0, 1599, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "append_partially_phased_batch") < 0)) __PYX_ERR(0, 1846, __pyx_L3_error)
       }
-    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
-      goto __pyx_L5_argtuple_error;
     } else {
-      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
-      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
+      switch (PyTuple_GET_SIZE(__pyx_args)) {
+        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
+        CYTHON_FALLTHROUGH;
+        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
+        values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
+        break;
+        default: goto __pyx_L5_argtuple_error;
+      }
     }
     __pyx_v_allele_int32_batch = ((PyArrayObject *)values[0]);
     __pyx_v_phasepresent_batch = ((PyArrayObject *)values[1]);
+    __pyx_v_allele_cts = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("append_partially_phased_batch", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1599, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("append_partially_phased_batch", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1846, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenWriter.append_partially_phased_batch", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_batch), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_batch", 0))) __PYX_ERR(0, 1599, __pyx_L1_error)
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_batch), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_batch", 0))) __PYX_ERR(0, 1599, __pyx_L1_error)
-  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_16append_partially_phased_batch(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32_batch, __pyx_v_phasepresent_batch);
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_allele_int32_batch), __pyx_ptype_5numpy_ndarray, 1, "allele_int32_batch", 0))) __PYX_ERR(0, 1846, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_phasepresent_batch), __pyx_ptype_5numpy_ndarray, 1, "phasepresent_batch", 0))) __PYX_ERR(0, 1846, __pyx_L1_error)
+  __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_16append_partially_phased_batch(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_allele_int32_batch, __pyx_v_phasepresent_batch, __pyx_v_allele_cts);
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-static PyObject *__pyx_pf_7pgenlib_10PgenWriter_16append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch) {
+static PyObject *__pyx_pf_7pgenlib_10PgenWriter_16append_partially_phased_batch(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, PyArrayObject *__pyx_v_allele_int32_batch, PyArrayObject *__pyx_v_phasepresent_batch, PyObject *__pyx_v_allele_cts) {
   __Pyx_LocalBuf_ND __pyx_pybuffernd_allele_int32_batch;
   __Pyx_Buffer __pyx_pybuffer_allele_int32_batch;
   __Pyx_LocalBuf_ND __pyx_pybuffernd_phasepresent_batch;
   __Pyx_Buffer __pyx_pybuffer_phasepresent_batch;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
+  struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch __pyx_t_2;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_partially_phased_batch", 0);
   __pyx_pybuffer_allele_int32_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_allele_int32_batch.refcount = 0;
   __pyx_pybuffernd_allele_int32_batch.data = NULL;
   __pyx_pybuffernd_allele_int32_batch.rcbuffer = &__pyx_pybuffer_allele_int32_batch;
   __pyx_pybuffer_phasepresent_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_phasepresent_batch.refcount = 0;
   __pyx_pybuffernd_phasepresent_batch.data = NULL;
   __pyx_pybuffernd_phasepresent_batch.rcbuffer = &__pyx_pybuffer_phasepresent_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1599, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_allele_int32_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_int32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1846, __pyx_L1_error)
   }
   __pyx_pybuffernd_allele_int32_batch.diminfo[0].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[0].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].strides = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_allele_int32_batch.diminfo[1].shape = __pyx_pybuffernd_allele_int32_batch.rcbuffer->pybuffer.shape[1];
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1599, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_phasepresent_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint8_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 1, __pyx_stack) == -1)) __PYX_ERR(0, 1846, __pyx_L1_error)
   }
   __pyx_pybuffernd_phasepresent_batch.diminfo[0].strides = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_phasepresent_batch.diminfo[0].shape = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_phasepresent_batch.diminfo[1].strides = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_phasepresent_batch.diminfo[1].shape = __pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer.shape[1];
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch(__pyx_v_self, __pyx_v_allele_int32_batch, __pyx_v_phasepresent_batch, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1599, __pyx_L1_error)
+  __pyx_t_2.__pyx_n = 1;
+  __pyx_t_2.allele_cts = __pyx_v_allele_cts;
+  __pyx_t_1 = __pyx_vtabptr_7pgenlib_PgenWriter->append_partially_phased_batch(__pyx_v_self, __pyx_v_allele_int32_batch, __pyx_v_phasepresent_batch, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1846, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -27683,15 +30771,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_phasepresent_batch.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1620
+/* "pgenlib.pyx":1893
  * 
  * 
  *     cdef append_dosages_batch_internal32(self, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_batch):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>floatarr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec
  */
 
@@ -27724,70 +30812,70 @@
   __Pyx_RefNannySetupContext("append_dosages_batch_internal32", 0);
   __pyx_pybuffer_floatarr_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_floatarr_batch.refcount = 0;
   __pyx_pybuffernd_floatarr_batch.data = NULL;
   __pyx_pybuffernd_floatarr_batch.rcbuffer = &__pyx_pybuffer_floatarr_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1620, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_floatarr_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float32_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1893, __pyx_L1_error)
   }
   __pyx_pybuffernd_floatarr_batch.diminfo[0].strides = __pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_floatarr_batch.diminfo[0].shape = __pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_floatarr_batch.diminfo[1].strides = __pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_floatarr_batch.diminfo[1].shape = __pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":1621
+  /* "pgenlib.pyx":1894
  * 
  *     cdef append_dosages_batch_internal32(self, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_batch):
  *         cdef uint32_t batch_size = <uint32_t>floatarr_batch.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  */
   __pyx_v_batch_size = ((uint32_t)(__pyx_v_floatarr_batch->dimensions[0]));
 
-  /* "pgenlib.pyx":1622
+  /* "pgenlib.pyx":1895
  *     cdef append_dosages_batch_internal32(self, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_batch):
  *         cdef uint32_t batch_size = <uint32_t>floatarr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main
  */
   __pyx_t_1 = __pyx_v_self->_genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1623
+  /* "pgenlib.pyx":1896
  *         cdef uint32_t batch_size = <uint32_t>floatarr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
  *         cdef uint16_t* dosage_main = self._dosage_main
  *         cdef uint32_t dosage_ct
  */
   __pyx_t_1 = __pyx_v_self->_dosage_present;
   __pyx_v_dosage_present = __pyx_t_1;
 
-  /* "pgenlib.pyx":1624
+  /* "pgenlib.pyx":1897
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t dosage_ct
  *         cdef uint32_t uii
  */
   __pyx_t_2 = __pyx_v_self->_dosage_main;
   __pyx_v_dosage_main = __pyx_t_2;
 
-  /* "pgenlib.pyx":1628
+  /* "pgenlib.pyx":1901
  *         cdef uint32_t uii
  *         cdef PglErr reterr
  *         for uii in range(batch_size):             # <<<<<<<<<<<<<<
  *             FloatsToDosage16(<float*>(&(floatarr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  */
   __pyx_t_3 = __pyx_v_batch_size;
   __pyx_t_4 = __pyx_t_3;
   for (__pyx_t_5 = 0; __pyx_t_5 < __pyx_t_4; __pyx_t_5+=1) {
     __pyx_v_uii = __pyx_t_5;
 
-    /* "pgenlib.pyx":1629
+    /* "pgenlib.pyx":1902
  *         cdef PglErr reterr
  *         for uii in range(batch_size):
  *             FloatsToDosage16(<float*>(&(floatarr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  */
     __pyx_t_6 = __pyx_v_uii;
@@ -27796,81 +30884,81 @@
     if (unlikely(__pyx_t_6 >= (size_t)__pyx_pybuffernd_floatarr_batch.diminfo[0].shape)) __pyx_t_8 = 0;
     if (__pyx_t_7 < 0) {
       __pyx_t_7 += __pyx_pybuffernd_floatarr_batch.diminfo[1].shape;
       if (unlikely(__pyx_t_7 < 0)) __pyx_t_8 = 1;
     } else if (unlikely(__pyx_t_7 >= __pyx_pybuffernd_floatarr_batch.diminfo[1].shape)) __pyx_t_8 = 1;
     if (unlikely(__pyx_t_8 != -1)) {
       __Pyx_RaiseBufferIndexError(__pyx_t_8);
-      __PYX_ERR(0, 1629, __pyx_L1_error)
+      __PYX_ERR(0, 1902, __pyx_L1_error)
     }
     plink2::FloatsToDosage16(((float *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_float32_t *, __pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer.buf, __pyx_t_6, __pyx_pybuffernd_floatarr_batch.diminfo[0].strides, __pyx_t_7, __pyx_pybuffernd_floatarr_batch.diminfo[1].strides)))), plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), 0x199A, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-    /* "pgenlib.pyx":1630
+    /* "pgenlib.pyx":1903
  *         for uii in range(batch_size):
  *             FloatsToDosage16(<float*>(&(floatarr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  */
     __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecDosage16(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_dosage_ct, __pyx_v_self->_state_ptr);
 
-    /* "pgenlib.pyx":1631
+    /* "pgenlib.pyx":1904
  *             FloatsToDosage16(<float*>(&(floatarr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  *         return
  */
     __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
     if (unlikely(__pyx_t_9)) {
 
-      /* "pgenlib.pyx":1632
+      /* "pgenlib.pyx":1905
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-      __pyx_t_10 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1632, __pyx_L1_error)
+      __pyx_t_10 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1905, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_10);
-      __pyx_t_11 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1632, __pyx_L1_error)
+      __pyx_t_11 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1905, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_11);
       __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-      __pyx_t_10 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_batch_error, __pyx_t_11); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1632, __pyx_L1_error)
+      __pyx_t_10 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_batch_error, __pyx_t_11); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1905, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_10);
       __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
-      __pyx_t_11 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1632, __pyx_L1_error)
+      __pyx_t_11 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1905, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_11);
       __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
       __Pyx_Raise(__pyx_t_11, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
-      __PYX_ERR(0, 1632, __pyx_L1_error)
+      __PYX_ERR(0, 1905, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1631
+      /* "pgenlib.pyx":1904
  *             FloatsToDosage16(<float*>(&(floatarr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  *         return
  */
     }
   }
 
-  /* "pgenlib.pyx":1633
+  /* "pgenlib.pyx":1906
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cdef append_dosages_batch_internal64(self, np.ndarray[np.float64_t,mode="c",ndim=2] doublearr_batch):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1620
+  /* "pgenlib.pyx":1893
  * 
  * 
  *     cdef append_dosages_batch_internal32(self, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_batch):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>floatarr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec
  */
 
@@ -27891,15 +30979,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_floatarr_batch.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1635
+/* "pgenlib.pyx":1908
  *         return
  * 
  *     cdef append_dosages_batch_internal64(self, np.ndarray[np.float64_t,mode="c",ndim=2] doublearr_batch):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>doublearr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec
  */
 
@@ -27932,70 +31020,70 @@
   __Pyx_RefNannySetupContext("append_dosages_batch_internal64", 0);
   __pyx_pybuffer_doublearr_batch.pybuffer.buf = NULL;
   __pyx_pybuffer_doublearr_batch.refcount = 0;
   __pyx_pybuffernd_doublearr_batch.data = NULL;
   __pyx_pybuffernd_doublearr_batch.rcbuffer = &__pyx_pybuffer_doublearr_batch;
   {
     __Pyx_BufFmt_StackElem __pyx_stack[1];
-    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_doublearr_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1635, __pyx_L1_error)
+    if (unlikely(__Pyx_GetBufferAndValidate(&__pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer, (PyObject*)__pyx_v_doublearr_batch, &__Pyx_TypeInfo_nn___pyx_t_5numpy_float64_t, PyBUF_FORMAT| PyBUF_C_CONTIGUOUS, 2, 0, __pyx_stack) == -1)) __PYX_ERR(0, 1908, __pyx_L1_error)
   }
   __pyx_pybuffernd_doublearr_batch.diminfo[0].strides = __pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer.strides[0]; __pyx_pybuffernd_doublearr_batch.diminfo[0].shape = __pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer.shape[0]; __pyx_pybuffernd_doublearr_batch.diminfo[1].strides = __pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer.strides[1]; __pyx_pybuffernd_doublearr_batch.diminfo[1].shape = __pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer.shape[1];
 
-  /* "pgenlib.pyx":1636
+  /* "pgenlib.pyx":1909
  * 
  *     cdef append_dosages_batch_internal64(self, np.ndarray[np.float64_t,mode="c",ndim=2] doublearr_batch):
  *         cdef uint32_t batch_size = <uint32_t>doublearr_batch.shape[0]             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  */
   __pyx_v_batch_size = ((uint32_t)(__pyx_v_doublearr_batch->dimensions[0]));
 
-  /* "pgenlib.pyx":1637
+  /* "pgenlib.pyx":1910
  *     cdef append_dosages_batch_internal64(self, np.ndarray[np.float64_t,mode="c",ndim=2] doublearr_batch):
  *         cdef uint32_t batch_size = <uint32_t>doublearr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec             # <<<<<<<<<<<<<<
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main
  */
   __pyx_t_1 = __pyx_v_self->_genovec;
   __pyx_v_genovec = __pyx_t_1;
 
-  /* "pgenlib.pyx":1638
+  /* "pgenlib.pyx":1911
  *         cdef uint32_t batch_size = <uint32_t>doublearr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present             # <<<<<<<<<<<<<<
  *         cdef uint16_t* dosage_main = self._dosage_main
  *         cdef uint32_t dosage_ct
  */
   __pyx_t_1 = __pyx_v_self->_dosage_present;
   __pyx_v_dosage_present = __pyx_t_1;
 
-  /* "pgenlib.pyx":1639
+  /* "pgenlib.pyx":1912
  *         cdef uintptr_t* genovec = self._genovec
  *         cdef uintptr_t* dosage_present = self._dosage_present
  *         cdef uint16_t* dosage_main = self._dosage_main             # <<<<<<<<<<<<<<
  *         cdef uint32_t dosage_ct
  *         cdef uint32_t uii
  */
   __pyx_t_2 = __pyx_v_self->_dosage_main;
   __pyx_v_dosage_main = __pyx_t_2;
 
-  /* "pgenlib.pyx":1643
+  /* "pgenlib.pyx":1916
  *         cdef uint32_t uii
  *         cdef PglErr reterr
  *         for uii in range(batch_size):             # <<<<<<<<<<<<<<
  *             DoublesToDosage16(<double*>(&(doublearr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  */
   __pyx_t_3 = __pyx_v_batch_size;
   __pyx_t_4 = __pyx_t_3;
   for (__pyx_t_5 = 0; __pyx_t_5 < __pyx_t_4; __pyx_t_5+=1) {
     __pyx_v_uii = __pyx_t_5;
 
-    /* "pgenlib.pyx":1644
+    /* "pgenlib.pyx":1917
  *         cdef PglErr reterr
  *         for uii in range(batch_size):
  *             DoublesToDosage16(<double*>(&(doublearr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)             # <<<<<<<<<<<<<<
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  */
     __pyx_t_6 = __pyx_v_uii;
@@ -28004,81 +31092,81 @@
     if (unlikely(__pyx_t_6 >= (size_t)__pyx_pybuffernd_doublearr_batch.diminfo[0].shape)) __pyx_t_8 = 0;
     if (__pyx_t_7 < 0) {
       __pyx_t_7 += __pyx_pybuffernd_doublearr_batch.diminfo[1].shape;
       if (unlikely(__pyx_t_7 < 0)) __pyx_t_8 = 1;
     } else if (unlikely(__pyx_t_7 >= __pyx_pybuffernd_doublearr_batch.diminfo[1].shape)) __pyx_t_8 = 1;
     if (unlikely(__pyx_t_8 != -1)) {
       __Pyx_RaiseBufferIndexError(__pyx_t_8);
-      __PYX_ERR(0, 1644, __pyx_L1_error)
+      __PYX_ERR(0, 1917, __pyx_L1_error)
     }
     plink2::DoublesToDosage16(((double *)(&(*__Pyx_BufPtrCContig2d(__pyx_t_5numpy_float64_t *, __pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer.buf, __pyx_t_6, __pyx_pybuffernd_doublearr_batch.diminfo[0].strides, __pyx_t_7, __pyx_pybuffernd_doublearr_batch.diminfo[1].strides)))), plink2::SpgwGetSampleCt(__pyx_v_self->_state_ptr), 0x199A, __pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, (&__pyx_v_dosage_ct));
 
-    /* "pgenlib.pyx":1645
+    /* "pgenlib.pyx":1918
  *         for uii in range(batch_size):
  *             DoublesToDosage16(<double*>(&(doublearr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)             # <<<<<<<<<<<<<<
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  */
     __pyx_v_reterr = plink2::SpgwAppendBiallelicGenovecDosage16(__pyx_v_genovec, __pyx_v_dosage_present, __pyx_v_dosage_main, __pyx_v_dosage_ct, __pyx_v_self->_state_ptr);
 
-    /* "pgenlib.pyx":1646
+    /* "pgenlib.pyx":1919
  *             DoublesToDosage16(<double*>(&(doublearr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  *         return
  */
     __pyx_t_9 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
     if (unlikely(__pyx_t_9)) {
 
-      /* "pgenlib.pyx":1647
+      /* "pgenlib.pyx":1920
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-      __pyx_t_10 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1647, __pyx_L1_error)
+      __pyx_t_10 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1920, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_10);
-      __pyx_t_11 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1647, __pyx_L1_error)
+      __pyx_t_11 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1920, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_11);
       __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
-      __pyx_t_10 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_batch_error, __pyx_t_11); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1647, __pyx_L1_error)
+      __pyx_t_10 = __Pyx_PyUnicode_Concat(__pyx_kp_u_append_dosages_batch_error, __pyx_t_11); if (unlikely(!__pyx_t_10)) __PYX_ERR(0, 1920, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_10);
       __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
-      __pyx_t_11 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1647, __pyx_L1_error)
+      __pyx_t_11 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_10); if (unlikely(!__pyx_t_11)) __PYX_ERR(0, 1920, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_11);
       __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
       __Pyx_Raise(__pyx_t_11, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
-      __PYX_ERR(0, 1647, __pyx_L1_error)
+      __PYX_ERR(0, 1920, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1646
+      /* "pgenlib.pyx":1919
  *             DoublesToDosage16(<double*>(&(doublearr_batch[uii, 0])), SpgwGetSampleCt(self._state_ptr), 6554, genovec, dosage_present, dosage_main, &dosage_ct)
  *             reterr = SpgwAppendBiallelicGenovecDosage16(genovec, dosage_present, dosage_main, dosage_ct, self._state_ptr)
  *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  *         return
  */
     }
   }
 
-  /* "pgenlib.pyx":1648
+  /* "pgenlib.pyx":1921
  *             if reterr != kPglRetSuccess:
  *                 raise RuntimeError("append_dosages_batch() error " + str(reterr))
  *         return             # <<<<<<<<<<<<<<
  * 
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1635
+  /* "pgenlib.pyx":1908
  *         return
  * 
  *     cdef append_dosages_batch_internal64(self, np.ndarray[np.float64_t,mode="c",ndim=2] doublearr_batch):             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>doublearr_batch.shape[0]
  *         cdef uintptr_t* genovec = self._genovec
  */
 
@@ -28099,15 +31187,15 @@
   __Pyx_SafeReleaseBuffer(&__pyx_pybuffernd_doublearr_batch.rcbuffer->pybuffer);
   __pyx_L2:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1650
+/* "pgenlib.pyx":1923
  *         return
  * 
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  */
 
@@ -28129,15 +31217,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_dosages_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1650, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_append_dosages_batch); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1923, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_19append_dosages_batch)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -28146,15 +31234,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_Call2Args(__pyx_t_3, __pyx_t_4, ((PyObject *)__pyx_v_floatarr_batch)) : __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_floatarr_batch));
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1650, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1923, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -28167,158 +31255,158 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1651
+  /* "pgenlib.pyx":1924
  * 
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr_batch.dtype == np.float32:
  */
   __pyx_t_5 = (((__pyx_v_self->_phase_dosage_gflags & plink2::kfPgenGlobalDosagePresent) == 0) != 0);
   if (unlikely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":1652
+    /* "pgenlib.pyx":1925
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")             # <<<<<<<<<<<<<<
  *         if floatarr_batch.dtype == np.float32:
  *             self.append_dosages_batch_internal32(floatarr_batch)
  */
-    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__29, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1652, __pyx_L1_error)
+    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__48, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1925, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_1);
     __Pyx_Raise(__pyx_t_1, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-    __PYX_ERR(0, 1652, __pyx_L1_error)
+    __PYX_ERR(0, 1925, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1651
+    /* "pgenlib.pyx":1924
  * 
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:             # <<<<<<<<<<<<<<
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr_batch.dtype == np.float32:
  */
   }
 
-  /* "pgenlib.pyx":1653
+  /* "pgenlib.pyx":1926
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr_batch.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.append_dosages_batch_internal32(floatarr_batch)
  *         elif floatarr_batch.dtype == np.float64:
  */
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_batch), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1653, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_batch), __pyx_n_s_dtype); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1926, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1653, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_np); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1926, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float32); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1653, __pyx_L1_error)
+  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_float32); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1926, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1653, __pyx_L1_error)
+  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1926, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1653, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1926, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":1654
+    /* "pgenlib.pyx":1927
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr_batch.dtype == np.float32:
  *             self.append_dosages_batch_internal32(floatarr_batch)             # <<<<<<<<<<<<<<
  *         elif floatarr_batch.dtype == np.float64:
  *             self.append_dosages_batch_internal64(floatarr_batch)
  */
-    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_batch_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr_batch)); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1654, __pyx_L1_error)
+    __pyx_t_2 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_batch_internal32(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr_batch)); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1927, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_2);
     __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
 
-    /* "pgenlib.pyx":1653
+    /* "pgenlib.pyx":1926
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  *         if floatarr_batch.dtype == np.float32:             # <<<<<<<<<<<<<<
  *             self.append_dosages_batch_internal32(floatarr_batch)
  *         elif floatarr_batch.dtype == np.float64:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":1655
+  /* "pgenlib.pyx":1928
  *         if floatarr_batch.dtype == np.float32:
  *             self.append_dosages_batch_internal32(floatarr_batch)
  *         elif floatarr_batch.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.append_dosages_batch_internal64(floatarr_batch)
  *         else:
  */
-  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_batch), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1655, __pyx_L1_error)
+  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_floatarr_batch), __pyx_n_s_dtype); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1928, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_2);
-  __Pyx_GetModuleGlobalName(__pyx_t_3, __pyx_n_s_np); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1655, __pyx_L1_error)
+  __Pyx_GetModuleGlobalName(__pyx_t_3, __pyx_n_s_np); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1928, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
-  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1655, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_float64); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1928, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-  __pyx_t_3 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1655, __pyx_L1_error)
+  __pyx_t_3 = PyObject_RichCompare(__pyx_t_2, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1928, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1655, __pyx_L1_error)
+  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_5 < 0)) __PYX_ERR(0, 1928, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
   if (likely(__pyx_t_5)) {
 
-    /* "pgenlib.pyx":1656
+    /* "pgenlib.pyx":1929
  *             self.append_dosages_batch_internal32(floatarr_batch)
  *         elif floatarr_batch.dtype == np.float64:
  *             self.append_dosages_batch_internal64(floatarr_batch)             # <<<<<<<<<<<<<<
  *         else:
  *             raise RuntimeError("Invalid append_dosages_batch() dosage array element type (float32 or float64 expected).")
  */
-    __pyx_t_3 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_batch_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr_batch)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1656, __pyx_L1_error)
+    __pyx_t_3 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->append_dosages_batch_internal64(__pyx_v_self, ((PyArrayObject *)__pyx_v_floatarr_batch)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1929, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
 
-    /* "pgenlib.pyx":1655
+    /* "pgenlib.pyx":1928
  *         if floatarr_batch.dtype == np.float32:
  *             self.append_dosages_batch_internal32(floatarr_batch)
  *         elif floatarr_batch.dtype == np.float64:             # <<<<<<<<<<<<<<
  *             self.append_dosages_batch_internal64(floatarr_batch)
  *         else:
  */
     goto __pyx_L4;
   }
 
-  /* "pgenlib.pyx":1658
+  /* "pgenlib.pyx":1931
  *             self.append_dosages_batch_internal64(floatarr_batch)
  *         else:
  *             raise RuntimeError("Invalid append_dosages_batch() dosage array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
   /*else*/ {
-    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__30, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1658, __pyx_L1_error)
+    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__49, NULL); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1931, __pyx_L1_error)
     __Pyx_GOTREF(__pyx_t_3);
     __Pyx_Raise(__pyx_t_3, 0, 0, 0);
     __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-    __PYX_ERR(0, 1658, __pyx_L1_error)
+    __PYX_ERR(0, 1931, __pyx_L1_error)
   }
   __pyx_L4:;
 
-  /* "pgenlib.pyx":1659
+  /* "pgenlib.pyx":1932
  *         else:
  *             raise RuntimeError("Invalid append_dosages_batch() dosage array element type (float32 or float64 expected).")
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1650
+  /* "pgenlib.pyx":1923
  *         return
  * 
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):             # <<<<<<<<<<<<<<
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")
  */
 
@@ -28341,15 +31429,15 @@
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_19append_dosages_batch(PyObject *__pyx_v_self, PyObject *__pyx_v_floatarr_batch) {
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   PyObject *__pyx_r = 0;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("append_dosages_batch (wrapper)", 0);
-  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_batch), __pyx_ptype_5numpy_ndarray, 1, "floatarr_batch", 0))) __PYX_ERR(0, 1650, __pyx_L1_error)
+  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_floatarr_batch), __pyx_ptype_5numpy_ndarray, 1, "floatarr_batch", 0))) __PYX_ERR(0, 1923, __pyx_L1_error)
   __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_18append_dosages_batch(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), ((PyArrayObject *)__pyx_v_floatarr_batch));
 
   /* function exit code */
   goto __pyx_L0;
   __pyx_L1_error:;
   __pyx_r = NULL;
   __pyx_L0:;
@@ -28362,15 +31450,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("append_dosages_batch", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_dosages_batch(__pyx_v_self, __pyx_v_floatarr_batch, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1650, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_append_dosages_batch(__pyx_v_self, __pyx_v_floatarr_batch, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1923, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -28379,24 +31467,25 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1662
+/* "pgenlib.pyx":1935
  * 
  * 
  *     cpdef close(self):             # <<<<<<<<<<<<<<
+ *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
- *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
  */
 
 static PyObject *__pyx_pw_7pgenlib_10PgenWriter_21close(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
 static PyObject *__pyx_f_7pgenlib_10PgenWriter_close(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self, int __pyx_skip_dispatch) {
+  plink2::PglErr __pyx_v_reterr;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   PyObject *__pyx_t_2 = NULL;
   PyObject *__pyx_t_3 = NULL;
   PyObject *__pyx_t_4 = NULL;
   int __pyx_t_5;
@@ -28409,15 +31498,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_close); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1662, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_close); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1935, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_21close)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
           __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
@@ -28426,15 +31515,15 @@
             __Pyx_INCREF(__pyx_t_4);
             __Pyx_INCREF(function);
             __Pyx_DECREF_SET(__pyx_t_3, function);
           }
         }
         __pyx_t_2 = (__pyx_t_4) ? __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4) : __Pyx_PyObject_CallNoArg(__pyx_t_3);
         __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
-        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1662, __pyx_L1_error)
+        if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1935, __pyx_L1_error)
         __Pyx_GOTREF(__pyx_t_2);
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
         goto __pyx_L0;
       }
@@ -28447,161 +31536,211 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1663
+  /* "pgenlib.pyx":1936
  * 
  *     cpdef close(self):
+ *         cdef PglErr reterr = kPglRetSuccess             # <<<<<<<<<<<<<<
+ *         if self._state_ptr:
+ *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
+ */
+  __pyx_v_reterr = plink2::kPglRetSuccess;
+
+  /* "pgenlib.pyx":1937
+ *     cpdef close(self):
+ *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:             # <<<<<<<<<<<<<<
  *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
  */
   __pyx_t_5 = (__pyx_v_self->_state_ptr != 0);
   if (__pyx_t_5) {
 
-    /* "pgenlib.pyx":1664
- *     cpdef close(self):
+    /* "pgenlib.pyx":1938
+ *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
- *             SpgwFinish(self._state_ptr)
+ *             reterr = SpgwFinish(self._state_ptr)
  */
     __pyx_t_5 = ((plink2::SpgwGetVidx(__pyx_v_self->_state_ptr) != plink2::SpgwGetVariantCt(__pyx_v_self->_state_ptr)) != 0);
     if (unlikely(__pyx_t_5)) {
 
-      /* "pgenlib.pyx":1665
+      /* "pgenlib.pyx":1939
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")             # <<<<<<<<<<<<<<
- *             SpgwFinish(self._state_ptr)
- *             if self._nonref_flags:
+ *             reterr = SpgwFinish(self._state_ptr)
+ *             if reterr != kPglRetSuccess:
  */
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(plink2::SpgwGetVidx(__pyx_v_self->_state_ptr)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(plink2::SpgwGetVidx(__pyx_v_self->_state_ptr)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_PgenWriter_close_called_when_num, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_PgenWriter_close_called_when_num, __pyx_t_2); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
-      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_unequal_to_initially_declared_v); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u_unequal_to_initially_declared_v); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_2);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(plink2::SpgwGetVariantCt(__pyx_v_self->_state_ptr)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyInt_From_uint32_t(plink2::SpgwGetVariantCt(__pyx_v_self->_state_ptr)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
-      __pyx_t_3 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_3 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_3);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
       __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-      __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_3 = __Pyx_PyUnicode_Concat(__pyx_t_1, __pyx_kp_u__5); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_3);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_3); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1665, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_3); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1939, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
       __Pyx_Raise(__pyx_t_1, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
-      __PYX_ERR(0, 1665, __pyx_L1_error)
+      __PYX_ERR(0, 1939, __pyx_L1_error)
 
-      /* "pgenlib.pyx":1664
- *     cpdef close(self):
+      /* "pgenlib.pyx":1938
+ *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):             # <<<<<<<<<<<<<<
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
- *             SpgwFinish(self._state_ptr)
+ *             reterr = SpgwFinish(self._state_ptr)
  */
     }
 
-    /* "pgenlib.pyx":1666
+    /* "pgenlib.pyx":1940
  *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
- *             SpgwFinish(self._state_ptr)             # <<<<<<<<<<<<<<
+ *             reterr = SpgwFinish(self._state_ptr)             # <<<<<<<<<<<<<<
+ *             if reterr != kPglRetSuccess:
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
+ */
+    __pyx_v_reterr = plink2::SpgwFinish(__pyx_v_self->_state_ptr);
+
+    /* "pgenlib.pyx":1941
+ *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
+ *             reterr = SpgwFinish(self._state_ptr)
+ *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
+ *             if self._nonref_flags:
+ */
+    __pyx_t_5 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+    if (unlikely(__pyx_t_5)) {
+
+      /* "pgenlib.pyx":1942
+ *             reterr = SpgwFinish(self._state_ptr)
+ *             if reterr != kPglRetSuccess:
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))             # <<<<<<<<<<<<<<
  *             if self._nonref_flags:
  *                 aligned_free(self._nonref_flags)
  */
-    (void)(plink2::SpgwFinish(__pyx_v_self->_state_ptr));
+      __pyx_t_1 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1942, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __pyx_t_3 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_1); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1942, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_3);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __pyx_t_1 = __Pyx_PyUnicode_Concat(__pyx_kp_u_PgenWriter_close_SpgwFinish_erro, __pyx_t_3); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1942, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_1);
+      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
+      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_1); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1942, __pyx_L1_error)
+      __Pyx_GOTREF(__pyx_t_3);
+      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
+      __Pyx_Raise(__pyx_t_3, 0, 0, 0);
+      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
+      __PYX_ERR(0, 1942, __pyx_L1_error)
 
-    /* "pgenlib.pyx":1667
+      /* "pgenlib.pyx":1941
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
- *             SpgwFinish(self._state_ptr)
+ *             reterr = SpgwFinish(self._state_ptr)
+ *             if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
+ *             if self._nonref_flags:
+ */
+    }
+
+    /* "pgenlib.pyx":1943
+ *             if reterr != kPglRetSuccess:
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
  *             if self._nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)
  */
     __pyx_t_5 = (__pyx_v_self->_nonref_flags != 0);
     if (__pyx_t_5) {
 
-      /* "pgenlib.pyx":1668
- *             SpgwFinish(self._state_ptr)
+      /* "pgenlib.pyx":1944
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
  *             if self._nonref_flags:
  *                 aligned_free(self._nonref_flags)             # <<<<<<<<<<<<<<
  *             PyMem_Free(self._state_ptr)
  *             self._state_ptr = NULL
  */
       plink2::aligned_free(__pyx_v_self->_nonref_flags);
 
-      /* "pgenlib.pyx":1667
- *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
- *             SpgwFinish(self._state_ptr)
+      /* "pgenlib.pyx":1943
+ *             if reterr != kPglRetSuccess:
+ *                 raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
  *             if self._nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)
  */
     }
 
-    /* "pgenlib.pyx":1669
+    /* "pgenlib.pyx":1945
  *             if self._nonref_flags:
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)             # <<<<<<<<<<<<<<
  *             self._state_ptr = NULL
  *         return
  */
     PyMem_Free(__pyx_v_self->_state_ptr);
 
-    /* "pgenlib.pyx":1670
+    /* "pgenlib.pyx":1946
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)
  *             self._state_ptr = NULL             # <<<<<<<<<<<<<<
  *         return
  * 
  */
     __pyx_v_self->_state_ptr = NULL;
 
-    /* "pgenlib.pyx":1663
- * 
+    /* "pgenlib.pyx":1937
  *     cpdef close(self):
+ *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:             # <<<<<<<<<<<<<<
  *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
  *                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
  */
   }
 
-  /* "pgenlib.pyx":1671
+  /* "pgenlib.pyx":1947
  *             PyMem_Free(self._state_ptr)
  *             self._state_ptr = NULL
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1662
+  /* "pgenlib.pyx":1935
  * 
  * 
  *     cpdef close(self):             # <<<<<<<<<<<<<<
+ *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
- *             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
  */
 
   /* function exit code */
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
   __Pyx_XDECREF(__pyx_t_2);
   __Pyx_XDECREF(__pyx_t_3);
@@ -28632,15 +31771,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("close", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_close(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1662, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter_close(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1935, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -28649,15 +31788,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1674
+/* "pgenlib.pyx":1950
  * 
  * 
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):             # <<<<<<<<<<<<<<
  *         self.close()
  *         return
  */
 
@@ -28680,15 +31819,15 @@
   /* Check if overridden in Python */
   else if (unlikely((Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0) || (Py_TYPE(((PyObject *)__pyx_v_self))->tp_flags & (Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HEAPTYPE)))) {
     #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     static PY_UINT64_T __pyx_tp_dict_version = __PYX_DICT_VERSION_INIT, __pyx_obj_dict_version = __PYX_DICT_VERSION_INIT;
     if (unlikely(!__Pyx_object_dict_version_matches(((PyObject *)__pyx_v_self), __pyx_tp_dict_version, __pyx_obj_dict_version))) {
       PY_UINT64_T __pyx_type_dict_guard = __Pyx_get_tp_dict_version(((PyObject *)__pyx_v_self));
       #endif
-      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_exit); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1674, __pyx_L1_error)
+      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_exit); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1950, __pyx_L1_error)
       __Pyx_GOTREF(__pyx_t_1);
       if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)(void*)__pyx_pw_7pgenlib_10PgenWriter_23__exit__)) {
         __Pyx_XDECREF(__pyx_r);
         __Pyx_INCREF(__pyx_t_1);
         __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
         __pyx_t_5 = 0;
         if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
@@ -28700,43 +31839,43 @@
             __Pyx_DECREF_SET(__pyx_t_3, function);
             __pyx_t_5 = 1;
           }
         }
         #if CYTHON_FAST_PYCALL
         if (PyFunction_Check(__pyx_t_3)) {
           PyObject *__pyx_temp[4] = {__pyx_t_4, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb};
-          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1674, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1950, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         #if CYTHON_FAST_PYCCALL
         if (__Pyx_PyFastCFunction_Check(__pyx_t_3)) {
           PyObject *__pyx_temp[4] = {__pyx_t_4, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb};
-          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1674, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyCFunction_FastCall(__pyx_t_3, __pyx_temp+1-__pyx_t_5, 3+__pyx_t_5); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1950, __pyx_L1_error)
           __Pyx_XDECREF(__pyx_t_4); __pyx_t_4 = 0;
           __Pyx_GOTREF(__pyx_t_2);
         } else
         #endif
         {
-          __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1674, __pyx_L1_error)
+          __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 1950, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_6);
           if (__pyx_t_4) {
             __Pyx_GIVEREF(__pyx_t_4); PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __pyx_t_4 = NULL;
           }
           __Pyx_INCREF(__pyx_v_exc_type);
           __Pyx_GIVEREF(__pyx_v_exc_type);
           PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, __pyx_v_exc_type);
           __Pyx_INCREF(__pyx_v_exc_val);
           __Pyx_GIVEREF(__pyx_v_exc_val);
           PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_exc_val);
           __Pyx_INCREF(__pyx_v_exc_tb);
           __Pyx_GIVEREF(__pyx_v_exc_tb);
           PyTuple_SET_ITEM(__pyx_t_6, 2+__pyx_t_5, __pyx_v_exc_tb);
-          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1674, __pyx_L1_error)
+          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1950, __pyx_L1_error)
           __Pyx_GOTREF(__pyx_t_2);
           __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
         }
         __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
         __pyx_r = __pyx_t_2;
         __pyx_t_2 = 0;
         __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
@@ -28751,37 +31890,37 @@
       #endif
       __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
       #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_PYTYPE_LOOKUP && CYTHON_USE_TYPE_SLOTS
     }
     #endif
   }
 
-  /* "pgenlib.pyx":1675
+  /* "pgenlib.pyx":1951
  * 
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):
  *         self.close()             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->close(__pyx_v_self, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1675, __pyx_L1_error)
+  __pyx_t_1 = ((struct __pyx_vtabstruct_7pgenlib_PgenWriter *)__pyx_v_self->__pyx_vtab)->close(__pyx_v_self, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1951, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-  /* "pgenlib.pyx":1676
+  /* "pgenlib.pyx":1952
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):
  *         self.close()
  *         return             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_r = Py_None; __Pyx_INCREF(Py_None);
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1674
+  /* "pgenlib.pyx":1950
  * 
  * 
  *     cpdef __exit__(self, exc_type, exc_val, exc_tb):             # <<<<<<<<<<<<<<
  *         self.close()
  *         return
  */
 
@@ -28833,40 +31972,40 @@
         case  0:
         if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_exc_type)) != 0)) kw_args--;
         else goto __pyx_L5_argtuple_error;
         CYTHON_FALLTHROUGH;
         case  1:
         if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_exc_val)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 1); __PYX_ERR(0, 1674, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 1); __PYX_ERR(0, 1950, __pyx_L3_error)
         }
         CYTHON_FALLTHROUGH;
         case  2:
         if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_exc_tb)) != 0)) kw_args--;
         else {
-          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 2); __PYX_ERR(0, 1674, __pyx_L3_error)
+          __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, 2); __PYX_ERR(0, 1950, __pyx_L3_error)
         }
       }
       if (unlikely(kw_args > 0)) {
-        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__exit__") < 0)) __PYX_ERR(0, 1674, __pyx_L3_error)
+        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__exit__") < 0)) __PYX_ERR(0, 1950, __pyx_L3_error)
       }
     } else if (PyTuple_GET_SIZE(__pyx_args) != 3) {
       goto __pyx_L5_argtuple_error;
     } else {
       values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
       values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
       values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
     }
     __pyx_v_exc_type = values[0];
     __pyx_v_exc_val = values[1];
     __pyx_v_exc_tb = values[2];
   }
   goto __pyx_L4_argument_unpacking_done;
   __pyx_L5_argtuple_error:;
-  __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1674, __pyx_L3_error)
+  __Pyx_RaiseArgtupleInvalid("__exit__", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(0, 1950, __pyx_L3_error)
   __pyx_L3_error:;
   __Pyx_AddTraceback("pgenlib.PgenWriter.__exit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __Pyx_RefNannyFinishContext();
   return NULL;
   __pyx_L4_argument_unpacking_done:;
   __pyx_r = __pyx_pf_7pgenlib_10PgenWriter_22__exit__(((struct __pyx_obj_7pgenlib_PgenWriter *)__pyx_v_self), __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb);
 
@@ -28880,15 +32019,15 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__exit__", 0);
   __Pyx_XDECREF(__pyx_r);
-  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter___exit__(__pyx_v_self, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1674, __pyx_L1_error)
+  __pyx_t_1 = __pyx_f_7pgenlib_10PgenWriter___exit__(__pyx_v_self, __pyx_v_exc_type, __pyx_v_exc_val, __pyx_v_exc_tb, 1); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1950, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
   /* function exit code */
   __pyx_L1_error:;
@@ -28897,15 +32036,15 @@
   __pyx_r = NULL;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "pgenlib.pyx":1679
+/* "pgenlib.pyx":1955
  * 
  * 
  *     def __dealloc__(self):             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
  */
 
@@ -28920,137 +32059,187 @@
   __Pyx_RefNannyFinishContext();
 }
 
 static void __pyx_pf_7pgenlib_10PgenWriter_24__dealloc__(struct __pyx_obj_7pgenlib_PgenWriter *__pyx_v_self) {
   plink2::PglErr __pyx_v_reterr;
   __Pyx_RefNannyDeclarations
   int __pyx_t_1;
+  PyObject *__pyx_t_2 = NULL;
+  PyObject *__pyx_t_3 = NULL;
+  int __pyx_lineno = 0;
+  const char *__pyx_filename = NULL;
+  int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__dealloc__", 0);
 
-  /* "pgenlib.pyx":1680
+  /* "pgenlib.pyx":1956
  * 
  *     def __dealloc__(self):
  *         cdef PglErr reterr = kPglRetSuccess             # <<<<<<<<<<<<<<
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
  */
   __pyx_v_reterr = plink2::kPglRetSuccess;
 
-  /* "pgenlib.pyx":1681
+  /* "pgenlib.pyx":1957
  *     def __dealloc__(self):
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:             # <<<<<<<<<<<<<<
  *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
- *                 SpgwFinish(self._state_ptr)
+ *                 reterr = SpgwFinish(self._state_ptr)
  */
   __pyx_t_1 = (__pyx_v_self->_state_ptr != 0);
   if (__pyx_t_1) {
 
-    /* "pgenlib.pyx":1682
+    /* "pgenlib.pyx":1958
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):             # <<<<<<<<<<<<<<
- *                 SpgwFinish(self._state_ptr)
- *             else:
+ *                 reterr = SpgwFinish(self._state_ptr)
+ *                 if reterr != kPglRetSuccess:
  */
     __pyx_t_1 = ((plink2::SpgwGetVidx(__pyx_v_self->_state_ptr) == plink2::SpgwGetVariantCt(__pyx_v_self->_state_ptr)) != 0);
     if (__pyx_t_1) {
 
-      /* "pgenlib.pyx":1683
+      /* "pgenlib.pyx":1959
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
- *                 SpgwFinish(self._state_ptr)             # <<<<<<<<<<<<<<
+ *                 reterr = SpgwFinish(self._state_ptr)             # <<<<<<<<<<<<<<
+ *                 if reterr != kPglRetSuccess:
+ *                     raise RuntimeError("PgenWriter.__dealloc__(): SpgwFinish() error " + str(reterr))
+ */
+      __pyx_v_reterr = plink2::SpgwFinish(__pyx_v_self->_state_ptr);
+
+      /* "pgenlib.pyx":1960
+ *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
+ *                 reterr = SpgwFinish(self._state_ptr)
+ *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("PgenWriter.__dealloc__(): SpgwFinish() error " + str(reterr))
+ *             else:
+ */
+      __pyx_t_1 = ((__pyx_v_reterr != plink2::kPglRetSuccess) != 0);
+      if (unlikely(__pyx_t_1)) {
+
+        /* "pgenlib.pyx":1961
+ *                 reterr = SpgwFinish(self._state_ptr)
+ *                 if reterr != kPglRetSuccess:
+ *                     raise RuntimeError("PgenWriter.__dealloc__(): SpgwFinish() error " + str(reterr))             # <<<<<<<<<<<<<<
  *             else:
  *                 CleanupSpgw(self._state_ptr, &reterr)
  */
-      (void)(plink2::SpgwFinish(__pyx_v_self->_state_ptr));
+        __pyx_t_2 = __Pyx_PyInt_From_plink2_3a__3a_PglErr(__pyx_v_reterr); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1961, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __pyx_t_3 = __Pyx_PyObject_CallOneArg(((PyObject *)(&PyUnicode_Type)), __pyx_t_2); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1961, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_3);
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __pyx_t_2 = __Pyx_PyUnicode_Concat(__pyx_kp_u_PgenWriter___dealloc___SpgwFinis, __pyx_t_3); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 1961, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_2);
+        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
+        __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_builtin_RuntimeError, __pyx_t_2); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 1961, __pyx_L1_error)
+        __Pyx_GOTREF(__pyx_t_3);
+        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
+        __Pyx_Raise(__pyx_t_3, 0, 0, 0);
+        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
+        __PYX_ERR(0, 1961, __pyx_L1_error)
+
+        /* "pgenlib.pyx":1960
+ *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
+ *                 reterr = SpgwFinish(self._state_ptr)
+ *                 if reterr != kPglRetSuccess:             # <<<<<<<<<<<<<<
+ *                     raise RuntimeError("PgenWriter.__dealloc__(): SpgwFinish() error " + str(reterr))
+ *             else:
+ */
+      }
 
-      /* "pgenlib.pyx":1682
+      /* "pgenlib.pyx":1958
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
  *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):             # <<<<<<<<<<<<<<
- *                 SpgwFinish(self._state_ptr)
- *             else:
+ *                 reterr = SpgwFinish(self._state_ptr)
+ *                 if reterr != kPglRetSuccess:
  */
       goto __pyx_L4;
     }
 
-    /* "pgenlib.pyx":1685
- *                 SpgwFinish(self._state_ptr)
+    /* "pgenlib.pyx":1963
+ *                     raise RuntimeError("PgenWriter.__dealloc__(): SpgwFinish() error " + str(reterr))
  *             else:
  *                 CleanupSpgw(self._state_ptr, &reterr)             # <<<<<<<<<<<<<<
  *             if self._nonref_flags:
  *                 aligned_free(self._nonref_flags)
  */
     /*else*/ {
       (void)(plink2::CleanupSpgw(__pyx_v_self->_state_ptr, (&__pyx_v_reterr)));
     }
     __pyx_L4:;
 
-    /* "pgenlib.pyx":1686
+    /* "pgenlib.pyx":1964
  *             else:
  *                 CleanupSpgw(self._state_ptr, &reterr)
  *             if self._nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)
  */
     __pyx_t_1 = (__pyx_v_self->_nonref_flags != 0);
     if (__pyx_t_1) {
 
-      /* "pgenlib.pyx":1687
+      /* "pgenlib.pyx":1965
  *                 CleanupSpgw(self._state_ptr, &reterr)
  *             if self._nonref_flags:
  *                 aligned_free(self._nonref_flags)             # <<<<<<<<<<<<<<
  *             PyMem_Free(self._state_ptr)
  *         return
  */
       plink2::aligned_free(__pyx_v_self->_nonref_flags);
 
-      /* "pgenlib.pyx":1686
+      /* "pgenlib.pyx":1964
  *             else:
  *                 CleanupSpgw(self._state_ptr, &reterr)
  *             if self._nonref_flags:             # <<<<<<<<<<<<<<
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)
  */
     }
 
-    /* "pgenlib.pyx":1688
+    /* "pgenlib.pyx":1966
  *             if self._nonref_flags:
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)             # <<<<<<<<<<<<<<
  *         return
  */
     PyMem_Free(__pyx_v_self->_state_ptr);
 
-    /* "pgenlib.pyx":1681
+    /* "pgenlib.pyx":1957
  *     def __dealloc__(self):
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:             # <<<<<<<<<<<<<<
  *             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
- *                 SpgwFinish(self._state_ptr)
+ *                 reterr = SpgwFinish(self._state_ptr)
  */
   }
 
-  /* "pgenlib.pyx":1689
+  /* "pgenlib.pyx":1967
  *                 aligned_free(self._nonref_flags)
  *             PyMem_Free(self._state_ptr)
  *         return             # <<<<<<<<<<<<<<
  */
   goto __pyx_L0;
 
-  /* "pgenlib.pyx":1679
+  /* "pgenlib.pyx":1955
  * 
  * 
  *     def __dealloc__(self):             # <<<<<<<<<<<<<<
  *         cdef PglErr reterr = kPglRetSuccess
  *         if self._state_ptr:
  */
 
   /* function exit code */
+  __pyx_L1_error:;
+  __Pyx_XDECREF(__pyx_t_2);
+  __Pyx_XDECREF(__pyx_t_3);
+  __Pyx_WriteUnraisable("pgenlib.PgenWriter.__dealloc__", __pyx_clineno, __pyx_lineno, __pyx_filename, 1, 0);
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
 }
 
 /* "(tree fragment)":1
  * def __reduce_cython__(self):             # <<<<<<<<<<<<<<
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
@@ -29081,15 +32270,15 @@
 
   /* "(tree fragment)":2
  * def __reduce_cython__(self):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  */
-  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__31, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 2, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__50, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 2, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_Raise(__pyx_t_1, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __PYX_ERR(1, 2, __pyx_L1_error)
 
   /* "(tree fragment)":1
  * def __reduce_cython__(self):             # <<<<<<<<<<<<<<
@@ -29137,15 +32326,15 @@
   __Pyx_RefNannySetupContext("__setstate_cython__", 0);
 
   /* "(tree fragment)":4
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  */
-  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__32, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 4, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_tuple__51, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 4, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_Raise(__pyx_t_1, 0, 0, 0);
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __PYX_ERR(1, 4, __pyx_L1_error)
 
   /* "(tree fragment)":3
  * def __reduce_cython__(self):
@@ -29160,15 +32349,15 @@
   __Pyx_AddTraceback("pgenlib.PgenWriter.__setstate_cython__", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = NULL;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":735
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":735
  * ctypedef npy_cdouble     complex_t
  * 
  * cdef inline object PyArray_MultiIterNew1(a):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(1, <void*>a)
  * 
  */
 
@@ -29177,29 +32366,29 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("PyArray_MultiIterNew1", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":736
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":736
  * 
  * cdef inline object PyArray_MultiIterNew1(a):
  *     return PyArray_MultiIterNew(1, <void*>a)             # <<<<<<<<<<<<<<
  * 
  * cdef inline object PyArray_MultiIterNew2(a, b):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_1 = PyArray_MultiIterNew(1, ((void *)__pyx_v_a)); if (unlikely(!__pyx_t_1)) __PYX_ERR(2, 736, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":735
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":735
  * ctypedef npy_cdouble     complex_t
  * 
  * cdef inline object PyArray_MultiIterNew1(a):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(1, <void*>a)
  * 
  */
 
@@ -29210,15 +32399,15 @@
   __pyx_r = 0;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":738
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":738
  *     return PyArray_MultiIterNew(1, <void*>a)
  * 
  * cdef inline object PyArray_MultiIterNew2(a, b):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
  * 
  */
 
@@ -29227,29 +32416,29 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("PyArray_MultiIterNew2", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":739
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":739
  * 
  * cdef inline object PyArray_MultiIterNew2(a, b):
  *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)             # <<<<<<<<<<<<<<
  * 
  * cdef inline object PyArray_MultiIterNew3(a, b, c):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_1 = PyArray_MultiIterNew(2, ((void *)__pyx_v_a), ((void *)__pyx_v_b)); if (unlikely(!__pyx_t_1)) __PYX_ERR(2, 739, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":738
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":738
  *     return PyArray_MultiIterNew(1, <void*>a)
  * 
  * cdef inline object PyArray_MultiIterNew2(a, b):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
  * 
  */
 
@@ -29260,15 +32449,15 @@
   __pyx_r = 0;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":741
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":741
  *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
  * 
  * cdef inline object PyArray_MultiIterNew3(a, b, c):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
  * 
  */
 
@@ -29277,29 +32466,29 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("PyArray_MultiIterNew3", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":742
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":742
  * 
  * cdef inline object PyArray_MultiIterNew3(a, b, c):
  *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)             # <<<<<<<<<<<<<<
  * 
  * cdef inline object PyArray_MultiIterNew4(a, b, c, d):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_1 = PyArray_MultiIterNew(3, ((void *)__pyx_v_a), ((void *)__pyx_v_b), ((void *)__pyx_v_c)); if (unlikely(!__pyx_t_1)) __PYX_ERR(2, 742, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":741
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":741
  *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
  * 
  * cdef inline object PyArray_MultiIterNew3(a, b, c):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
  * 
  */
 
@@ -29310,15 +32499,15 @@
   __pyx_r = 0;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":744
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":744
  *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
  * 
  * cdef inline object PyArray_MultiIterNew4(a, b, c, d):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
  * 
  */
 
@@ -29327,29 +32516,29 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("PyArray_MultiIterNew4", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":745
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":745
  * 
  * cdef inline object PyArray_MultiIterNew4(a, b, c, d):
  *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)             # <<<<<<<<<<<<<<
  * 
  * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_1 = PyArray_MultiIterNew(4, ((void *)__pyx_v_a), ((void *)__pyx_v_b), ((void *)__pyx_v_c), ((void *)__pyx_v_d)); if (unlikely(!__pyx_t_1)) __PYX_ERR(2, 745, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":744
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":744
  *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
  * 
  * cdef inline object PyArray_MultiIterNew4(a, b, c, d):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
  * 
  */
 
@@ -29360,15 +32549,15 @@
   __pyx_r = 0;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":747
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":747
  *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
  * 
  * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
  * 
  */
 
@@ -29377,29 +32566,29 @@
   __Pyx_RefNannyDeclarations
   PyObject *__pyx_t_1 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("PyArray_MultiIterNew5", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":748
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":748
  * 
  * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):
  *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)             # <<<<<<<<<<<<<<
  * 
  * cdef inline tuple PyDataType_SHAPE(dtype d):
  */
   __Pyx_XDECREF(__pyx_r);
   __pyx_t_1 = PyArray_MultiIterNew(5, ((void *)__pyx_v_a), ((void *)__pyx_v_b), ((void *)__pyx_v_c), ((void *)__pyx_v_d), ((void *)__pyx_v_e)); if (unlikely(!__pyx_t_1)) __PYX_ERR(2, 748, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_r = __pyx_t_1;
   __pyx_t_1 = 0;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":747
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":747
  *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
  * 
  * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):             # <<<<<<<<<<<<<<
  *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
  * 
  */
 
@@ -29410,212 +32599,212 @@
   __pyx_r = 0;
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":750
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":750
  *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
  * 
  * cdef inline tuple PyDataType_SHAPE(dtype d):             # <<<<<<<<<<<<<<
  *     if PyDataType_HASSUBARRAY(d):
  *         return <tuple>d.subarray.shape
  */
 
 static CYTHON_INLINE PyObject *__pyx_f_5numpy_PyDataType_SHAPE(PyArray_Descr *__pyx_v_d) {
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   int __pyx_t_1;
   __Pyx_RefNannySetupContext("PyDataType_SHAPE", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":751
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":751
  * 
  * cdef inline tuple PyDataType_SHAPE(dtype d):
  *     if PyDataType_HASSUBARRAY(d):             # <<<<<<<<<<<<<<
  *         return <tuple>d.subarray.shape
  *     else:
  */
   __pyx_t_1 = (PyDataType_HASSUBARRAY(__pyx_v_d) != 0);
   if (__pyx_t_1) {
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":752
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":752
  * cdef inline tuple PyDataType_SHAPE(dtype d):
  *     if PyDataType_HASSUBARRAY(d):
  *         return <tuple>d.subarray.shape             # <<<<<<<<<<<<<<
  *     else:
  *         return ()
  */
     __Pyx_XDECREF(__pyx_r);
     __Pyx_INCREF(((PyObject*)__pyx_v_d->subarray->shape));
     __pyx_r = ((PyObject*)__pyx_v_d->subarray->shape);
     goto __pyx_L0;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":751
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":751
  * 
  * cdef inline tuple PyDataType_SHAPE(dtype d):
  *     if PyDataType_HASSUBARRAY(d):             # <<<<<<<<<<<<<<
  *         return <tuple>d.subarray.shape
  *     else:
  */
   }
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":754
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":754
  *         return <tuple>d.subarray.shape
  *     else:
  *         return ()             # <<<<<<<<<<<<<<
  * 
  * 
  */
   /*else*/ {
     __Pyx_XDECREF(__pyx_r);
     __Pyx_INCREF(__pyx_empty_tuple);
     __pyx_r = __pyx_empty_tuple;
     goto __pyx_L0;
   }
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":750
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":750
  *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
  * 
  * cdef inline tuple PyDataType_SHAPE(dtype d):             # <<<<<<<<<<<<<<
  *     if PyDataType_HASSUBARRAY(d):
  *         return <tuple>d.subarray.shape
  */
 
   /* function exit code */
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":929
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":929
  *     int _import_umath() except -1
  * 
  * cdef inline void set_array_base(ndarray arr, object base):             # <<<<<<<<<<<<<<
  *     Py_INCREF(base) # important to do this before stealing the reference below!
  *     PyArray_SetBaseObject(arr, base)
  */
 
 static CYTHON_INLINE void __pyx_f_5numpy_set_array_base(PyArrayObject *__pyx_v_arr, PyObject *__pyx_v_base) {
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("set_array_base", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":930
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":930
  * 
  * cdef inline void set_array_base(ndarray arr, object base):
  *     Py_INCREF(base) # important to do this before stealing the reference below!             # <<<<<<<<<<<<<<
  *     PyArray_SetBaseObject(arr, base)
  * 
  */
   Py_INCREF(__pyx_v_base);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":931
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":931
  * cdef inline void set_array_base(ndarray arr, object base):
  *     Py_INCREF(base) # important to do this before stealing the reference below!
  *     PyArray_SetBaseObject(arr, base)             # <<<<<<<<<<<<<<
  * 
  * cdef inline object get_array_base(ndarray arr):
  */
   (void)(PyArray_SetBaseObject(__pyx_v_arr, __pyx_v_base));
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":929
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":929
  *     int _import_umath() except -1
  * 
  * cdef inline void set_array_base(ndarray arr, object base):             # <<<<<<<<<<<<<<
  *     Py_INCREF(base) # important to do this before stealing the reference below!
  *     PyArray_SetBaseObject(arr, base)
  */
 
   /* function exit code */
   __Pyx_RefNannyFinishContext();
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":933
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":933
  *     PyArray_SetBaseObject(arr, base)
  * 
  * cdef inline object get_array_base(ndarray arr):             # <<<<<<<<<<<<<<
  *     base = PyArray_BASE(arr)
  *     if base is NULL:
  */
 
 static CYTHON_INLINE PyObject *__pyx_f_5numpy_get_array_base(PyArrayObject *__pyx_v_arr) {
   PyObject *__pyx_v_base;
   PyObject *__pyx_r = NULL;
   __Pyx_RefNannyDeclarations
   int __pyx_t_1;
   __Pyx_RefNannySetupContext("get_array_base", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":934
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":934
  * 
  * cdef inline object get_array_base(ndarray arr):
  *     base = PyArray_BASE(arr)             # <<<<<<<<<<<<<<
  *     if base is NULL:
  *         return None
  */
   __pyx_v_base = PyArray_BASE(__pyx_v_arr);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":935
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":935
  * cdef inline object get_array_base(ndarray arr):
  *     base = PyArray_BASE(arr)
  *     if base is NULL:             # <<<<<<<<<<<<<<
  *         return None
  *     return <object>base
  */
   __pyx_t_1 = ((__pyx_v_base == NULL) != 0);
   if (__pyx_t_1) {
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":936
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":936
  *     base = PyArray_BASE(arr)
  *     if base is NULL:
  *         return None             # <<<<<<<<<<<<<<
  *     return <object>base
  * 
  */
     __Pyx_XDECREF(__pyx_r);
     __pyx_r = Py_None; __Pyx_INCREF(Py_None);
     goto __pyx_L0;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":935
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":935
  * cdef inline object get_array_base(ndarray arr):
  *     base = PyArray_BASE(arr)
  *     if base is NULL:             # <<<<<<<<<<<<<<
  *         return None
  *     return <object>base
  */
   }
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":937
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":937
  *     if base is NULL:
  *         return None
  *     return <object>base             # <<<<<<<<<<<<<<
  * 
  * # Versions of the import_* functions which are more suitable for
  */
   __Pyx_XDECREF(__pyx_r);
   __Pyx_INCREF(((PyObject *)__pyx_v_base));
   __pyx_r = ((PyObject *)__pyx_v_base);
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":933
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":933
  *     PyArray_SetBaseObject(arr, base)
  * 
  * cdef inline object get_array_base(ndarray arr):             # <<<<<<<<<<<<<<
  *     base = PyArray_BASE(arr)
  *     if base is NULL:
  */
 
   /* function exit code */
   __pyx_L0:;
   __Pyx_XGIVEREF(__pyx_r);
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":941
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":941
  * # Versions of the import_* functions which are more suitable for
  * # Cython code.
  * cdef inline int import_array() except -1:             # <<<<<<<<<<<<<<
  *     try:
  *         __pyx_import_array()
  */
 
@@ -29631,15 +32820,15 @@
   PyObject *__pyx_t_7 = NULL;
   PyObject *__pyx_t_8 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("import_array", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":942
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":942
  * # Cython code.
  * cdef inline int import_array() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         __pyx_import_array()
  *     except Exception:
  */
   {
@@ -29647,84 +32836,84 @@
     __Pyx_PyThreadState_assign
     __Pyx_ExceptionSave(&__pyx_t_1, &__pyx_t_2, &__pyx_t_3);
     __Pyx_XGOTREF(__pyx_t_1);
     __Pyx_XGOTREF(__pyx_t_2);
     __Pyx_XGOTREF(__pyx_t_3);
     /*try:*/ {
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":943
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":943
  * cdef inline int import_array() except -1:
  *     try:
  *         __pyx_import_array()             # <<<<<<<<<<<<<<
  *     except Exception:
  *         raise ImportError("numpy.core.multiarray failed to import")
  */
       __pyx_t_4 = _import_array(); if (unlikely(__pyx_t_4 == ((int)-1))) __PYX_ERR(2, 943, __pyx_L3_error)
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":942
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":942
  * # Cython code.
  * cdef inline int import_array() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         __pyx_import_array()
  *     except Exception:
  */
     }
     __Pyx_XDECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_XDECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_XDECREF(__pyx_t_3); __pyx_t_3 = 0;
     goto __pyx_L8_try_end;
     __pyx_L3_error:;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":944
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":944
  *     try:
  *         __pyx_import_array()
  *     except Exception:             # <<<<<<<<<<<<<<
  *         raise ImportError("numpy.core.multiarray failed to import")
  * 
  */
     __pyx_t_4 = __Pyx_PyErr_ExceptionMatches(((PyObject *)(&((PyTypeObject*)PyExc_Exception)[0])));
     if (__pyx_t_4) {
       __Pyx_AddTraceback("numpy.import_array", __pyx_clineno, __pyx_lineno, __pyx_filename);
       if (__Pyx_GetException(&__pyx_t_5, &__pyx_t_6, &__pyx_t_7) < 0) __PYX_ERR(2, 944, __pyx_L5_except_error)
       __Pyx_GOTREF(__pyx_t_5);
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_GOTREF(__pyx_t_7);
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":945
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":945
  *         __pyx_import_array()
  *     except Exception:
  *         raise ImportError("numpy.core.multiarray failed to import")             # <<<<<<<<<<<<<<
  * 
  * cdef inline int import_umath() except -1:
  */
-      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_ImportError, __pyx_tuple__33, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(2, 945, __pyx_L5_except_error)
+      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_ImportError, __pyx_tuple__52, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(2, 945, __pyx_L5_except_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_Raise(__pyx_t_8, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __PYX_ERR(2, 945, __pyx_L5_except_error)
     }
     goto __pyx_L5_except_error;
     __pyx_L5_except_error:;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":942
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":942
  * # Cython code.
  * cdef inline int import_array() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         __pyx_import_array()
  *     except Exception:
  */
     __Pyx_XGIVEREF(__pyx_t_1);
     __Pyx_XGIVEREF(__pyx_t_2);
     __Pyx_XGIVEREF(__pyx_t_3);
     __Pyx_ExceptionReset(__pyx_t_1, __pyx_t_2, __pyx_t_3);
     goto __pyx_L1_error;
     __pyx_L8_try_end:;
   }
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":941
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":941
  * # Versions of the import_* functions which are more suitable for
  * # Cython code.
  * cdef inline int import_array() except -1:             # <<<<<<<<<<<<<<
  *     try:
  *         __pyx_import_array()
  */
 
@@ -29739,15 +32928,15 @@
   __Pyx_AddTraceback("numpy.import_array", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = -1;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":947
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":947
  *         raise ImportError("numpy.core.multiarray failed to import")
  * 
  * cdef inline int import_umath() except -1:             # <<<<<<<<<<<<<<
  *     try:
  *         _import_umath()
  */
 
@@ -29763,15 +32952,15 @@
   PyObject *__pyx_t_7 = NULL;
   PyObject *__pyx_t_8 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("import_umath", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":948
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":948
  * 
  * cdef inline int import_umath() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         _import_umath()
  *     except Exception:
  */
   {
@@ -29779,84 +32968,84 @@
     __Pyx_PyThreadState_assign
     __Pyx_ExceptionSave(&__pyx_t_1, &__pyx_t_2, &__pyx_t_3);
     __Pyx_XGOTREF(__pyx_t_1);
     __Pyx_XGOTREF(__pyx_t_2);
     __Pyx_XGOTREF(__pyx_t_3);
     /*try:*/ {
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":949
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":949
  * cdef inline int import_umath() except -1:
  *     try:
  *         _import_umath()             # <<<<<<<<<<<<<<
  *     except Exception:
  *         raise ImportError("numpy.core.umath failed to import")
  */
       __pyx_t_4 = _import_umath(); if (unlikely(__pyx_t_4 == ((int)-1))) __PYX_ERR(2, 949, __pyx_L3_error)
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":948
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":948
  * 
  * cdef inline int import_umath() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         _import_umath()
  *     except Exception:
  */
     }
     __Pyx_XDECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_XDECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_XDECREF(__pyx_t_3); __pyx_t_3 = 0;
     goto __pyx_L8_try_end;
     __pyx_L3_error:;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":950
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":950
  *     try:
  *         _import_umath()
  *     except Exception:             # <<<<<<<<<<<<<<
  *         raise ImportError("numpy.core.umath failed to import")
  * 
  */
     __pyx_t_4 = __Pyx_PyErr_ExceptionMatches(((PyObject *)(&((PyTypeObject*)PyExc_Exception)[0])));
     if (__pyx_t_4) {
       __Pyx_AddTraceback("numpy.import_umath", __pyx_clineno, __pyx_lineno, __pyx_filename);
       if (__Pyx_GetException(&__pyx_t_5, &__pyx_t_6, &__pyx_t_7) < 0) __PYX_ERR(2, 950, __pyx_L5_except_error)
       __Pyx_GOTREF(__pyx_t_5);
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_GOTREF(__pyx_t_7);
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":951
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":951
  *         _import_umath()
  *     except Exception:
  *         raise ImportError("numpy.core.umath failed to import")             # <<<<<<<<<<<<<<
  * 
  * cdef inline int import_ufunc() except -1:
  */
-      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_ImportError, __pyx_tuple__34, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(2, 951, __pyx_L5_except_error)
+      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_ImportError, __pyx_tuple__53, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(2, 951, __pyx_L5_except_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_Raise(__pyx_t_8, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __PYX_ERR(2, 951, __pyx_L5_except_error)
     }
     goto __pyx_L5_except_error;
     __pyx_L5_except_error:;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":948
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":948
  * 
  * cdef inline int import_umath() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         _import_umath()
  *     except Exception:
  */
     __Pyx_XGIVEREF(__pyx_t_1);
     __Pyx_XGIVEREF(__pyx_t_2);
     __Pyx_XGIVEREF(__pyx_t_3);
     __Pyx_ExceptionReset(__pyx_t_1, __pyx_t_2, __pyx_t_3);
     goto __pyx_L1_error;
     __pyx_L8_try_end:;
   }
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":947
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":947
  *         raise ImportError("numpy.core.multiarray failed to import")
  * 
  * cdef inline int import_umath() except -1:             # <<<<<<<<<<<<<<
  *     try:
  *         _import_umath()
  */
 
@@ -29871,15 +33060,15 @@
   __Pyx_AddTraceback("numpy.import_umath", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = -1;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":953
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":953
  *         raise ImportError("numpy.core.umath failed to import")
  * 
  * cdef inline int import_ufunc() except -1:             # <<<<<<<<<<<<<<
  *     try:
  *         _import_umath()
  */
 
@@ -29895,15 +33084,15 @@
   PyObject *__pyx_t_7 = NULL;
   PyObject *__pyx_t_8 = NULL;
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("import_ufunc", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":954
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":954
  * 
  * cdef inline int import_ufunc() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         _import_umath()
  *     except Exception:
  */
   {
@@ -29911,84 +33100,84 @@
     __Pyx_PyThreadState_assign
     __Pyx_ExceptionSave(&__pyx_t_1, &__pyx_t_2, &__pyx_t_3);
     __Pyx_XGOTREF(__pyx_t_1);
     __Pyx_XGOTREF(__pyx_t_2);
     __Pyx_XGOTREF(__pyx_t_3);
     /*try:*/ {
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":955
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":955
  * cdef inline int import_ufunc() except -1:
  *     try:
  *         _import_umath()             # <<<<<<<<<<<<<<
  *     except Exception:
  *         raise ImportError("numpy.core.umath failed to import")
  */
       __pyx_t_4 = _import_umath(); if (unlikely(__pyx_t_4 == ((int)-1))) __PYX_ERR(2, 955, __pyx_L3_error)
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":954
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":954
  * 
  * cdef inline int import_ufunc() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         _import_umath()
  *     except Exception:
  */
     }
     __Pyx_XDECREF(__pyx_t_1); __pyx_t_1 = 0;
     __Pyx_XDECREF(__pyx_t_2); __pyx_t_2 = 0;
     __Pyx_XDECREF(__pyx_t_3); __pyx_t_3 = 0;
     goto __pyx_L8_try_end;
     __pyx_L3_error:;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":956
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":956
  *     try:
  *         _import_umath()
  *     except Exception:             # <<<<<<<<<<<<<<
  *         raise ImportError("numpy.core.umath failed to import")
  * 
  */
     __pyx_t_4 = __Pyx_PyErr_ExceptionMatches(((PyObject *)(&((PyTypeObject*)PyExc_Exception)[0])));
     if (__pyx_t_4) {
       __Pyx_AddTraceback("numpy.import_ufunc", __pyx_clineno, __pyx_lineno, __pyx_filename);
       if (__Pyx_GetException(&__pyx_t_5, &__pyx_t_6, &__pyx_t_7) < 0) __PYX_ERR(2, 956, __pyx_L5_except_error)
       __Pyx_GOTREF(__pyx_t_5);
       __Pyx_GOTREF(__pyx_t_6);
       __Pyx_GOTREF(__pyx_t_7);
 
-      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":957
+      /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":957
  *         _import_umath()
  *     except Exception:
  *         raise ImportError("numpy.core.umath failed to import")             # <<<<<<<<<<<<<<
  * 
  * cdef extern from *:
  */
-      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_ImportError, __pyx_tuple__34, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(2, 957, __pyx_L5_except_error)
+      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_builtin_ImportError, __pyx_tuple__53, NULL); if (unlikely(!__pyx_t_8)) __PYX_ERR(2, 957, __pyx_L5_except_error)
       __Pyx_GOTREF(__pyx_t_8);
       __Pyx_Raise(__pyx_t_8, 0, 0, 0);
       __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
       __PYX_ERR(2, 957, __pyx_L5_except_error)
     }
     goto __pyx_L5_except_error;
     __pyx_L5_except_error:;
 
-    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":954
+    /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":954
  * 
  * cdef inline int import_ufunc() except -1:
  *     try:             # <<<<<<<<<<<<<<
  *         _import_umath()
  *     except Exception:
  */
     __Pyx_XGIVEREF(__pyx_t_1);
     __Pyx_XGIVEREF(__pyx_t_2);
     __Pyx_XGIVEREF(__pyx_t_3);
     __Pyx_ExceptionReset(__pyx_t_1, __pyx_t_2, __pyx_t_3);
     goto __pyx_L1_error;
     __pyx_L8_try_end:;
   }
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":953
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":953
  *         raise ImportError("numpy.core.umath failed to import")
  * 
  * cdef inline int import_ufunc() except -1:             # <<<<<<<<<<<<<<
  *     try:
  *         _import_umath()
  */
 
@@ -30003,176 +33192,176 @@
   __Pyx_AddTraceback("numpy.import_ufunc", __pyx_clineno, __pyx_lineno, __pyx_filename);
   __pyx_r = -1;
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":967
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":967
  * 
  * 
  * cdef inline bint is_timedelta64_object(object obj):             # <<<<<<<<<<<<<<
  *     """
  *     Cython equivalent of `isinstance(obj, np.timedelta64)`
  */
 
 static CYTHON_INLINE int __pyx_f_5numpy_is_timedelta64_object(PyObject *__pyx_v_obj) {
   int __pyx_r;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("is_timedelta64_object", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":979
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":979
  *     bool
  *     """
  *     return PyObject_TypeCheck(obj, &PyTimedeltaArrType_Type)             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __pyx_r = PyObject_TypeCheck(__pyx_v_obj, (&PyTimedeltaArrType_Type));
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":967
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":967
  * 
  * 
  * cdef inline bint is_timedelta64_object(object obj):             # <<<<<<<<<<<<<<
  *     """
  *     Cython equivalent of `isinstance(obj, np.timedelta64)`
  */
 
   /* function exit code */
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":982
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":982
  * 
  * 
  * cdef inline bint is_datetime64_object(object obj):             # <<<<<<<<<<<<<<
  *     """
  *     Cython equivalent of `isinstance(obj, np.datetime64)`
  */
 
 static CYTHON_INLINE int __pyx_f_5numpy_is_datetime64_object(PyObject *__pyx_v_obj) {
   int __pyx_r;
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("is_datetime64_object", 0);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":994
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":994
  *     bool
  *     """
  *     return PyObject_TypeCheck(obj, &PyDatetimeArrType_Type)             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __pyx_r = PyObject_TypeCheck(__pyx_v_obj, (&PyDatetimeArrType_Type));
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":982
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":982
  * 
  * 
  * cdef inline bint is_datetime64_object(object obj):             # <<<<<<<<<<<<<<
  *     """
  *     Cython equivalent of `isinstance(obj, np.datetime64)`
  */
 
   /* function exit code */
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":997
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":997
  * 
  * 
  * cdef inline npy_datetime get_datetime64_value(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the int64 value underlying scalar numpy datetime64 object
  */
 
 static CYTHON_INLINE npy_datetime __pyx_f_5numpy_get_datetime64_value(PyObject *__pyx_v_obj) {
   npy_datetime __pyx_r;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1004
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1004
  *     also needed.  That can be found using `get_datetime64_unit`.
  *     """
  *     return (<PyDatetimeScalarObject*>obj).obval             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __pyx_r = ((PyDatetimeScalarObject *)__pyx_v_obj)->obval;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":997
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":997
  * 
  * 
  * cdef inline npy_datetime get_datetime64_value(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the int64 value underlying scalar numpy datetime64 object
  */
 
   /* function exit code */
   __pyx_L0:;
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1007
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1007
  * 
  * 
  * cdef inline npy_timedelta get_timedelta64_value(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the int64 value underlying scalar numpy timedelta64 object
  */
 
 static CYTHON_INLINE npy_timedelta __pyx_f_5numpy_get_timedelta64_value(PyObject *__pyx_v_obj) {
   npy_timedelta __pyx_r;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1011
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1011
  *     returns the int64 value underlying scalar numpy timedelta64 object
  *     """
  *     return (<PyTimedeltaScalarObject*>obj).obval             # <<<<<<<<<<<<<<
  * 
  * 
  */
   __pyx_r = ((PyTimedeltaScalarObject *)__pyx_v_obj)->obval;
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1007
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1007
  * 
  * 
  * cdef inline npy_timedelta get_timedelta64_value(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the int64 value underlying scalar numpy timedelta64 object
  */
 
   /* function exit code */
   __pyx_L0:;
   return __pyx_r;
 }
 
-/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1014
+/* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1014
  * 
  * 
  * cdef inline NPY_DATETIMEUNIT get_datetime64_unit(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the unit part of the dtype for a numpy datetime64 object.
  */
 
 static CYTHON_INLINE NPY_DATETIMEUNIT __pyx_f_5numpy_get_datetime64_unit(PyObject *__pyx_v_obj) {
   NPY_DATETIMEUNIT __pyx_r;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1018
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1018
  *     returns the unit part of the dtype for a numpy datetime64 object.
  *     """
  *     return <NPY_DATETIMEUNIT>(<PyDatetimeScalarObject*>obj).obmeta.base             # <<<<<<<<<<<<<<
  */
   __pyx_r = ((NPY_DATETIMEUNIT)((PyDatetimeScalarObject *)__pyx_v_obj)->obmeta.base);
   goto __pyx_L0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1014
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1014
  * 
  * 
  * cdef inline NPY_DATETIMEUNIT get_datetime64_unit(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the unit part of the dtype for a numpy datetime64 object.
  */
 
@@ -30486,31 +33675,37 @@
 #endif
 
 static __Pyx_StringTabEntry __pyx_string_tab[] = {
   {&__pyx_kp_u_0_based_sample_idx_too_large, __pyx_k_0_based_sample_idx_too_large, sizeof(__pyx_k_0_based_sample_idx_too_large), 0, 1, 0, 0},
   {&__pyx_kp_u_Empty_sample_subset_is_not_curre, __pyx_k_Empty_sample_subset_is_not_curre, sizeof(__pyx_k_Empty_sample_subset_is_not_curre), 0, 1, 0, 0},
   {&__pyx_kp_u_Haplotype_major_read_alleles_lis, __pyx_k_Haplotype_major_read_alleles_lis, sizeof(__pyx_k_Haplotype_major_read_alleles_lis), 0, 1, 0, 0},
   {&__pyx_kp_u_Haplotype_major_read_alleles_lis_2, __pyx_k_Haplotype_major_read_alleles_lis_2, sizeof(__pyx_k_Haplotype_major_read_alleles_lis_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_Haplotype_major_read_alleles_lis_3, __pyx_k_Haplotype_major_read_alleles_lis_3, sizeof(__pyx_k_Haplotype_major_read_alleles_lis_3), 0, 1, 0, 0},
   {&__pyx_kp_u_Haplotype_major_read_alleles_ran, __pyx_k_Haplotype_major_read_alleles_ran, sizeof(__pyx_k_Haplotype_major_read_alleles_ran), 0, 1, 0, 0},
   {&__pyx_kp_u_Haplotype_major_read_alleles_ran_2, __pyx_k_Haplotype_major_read_alleles_ran_2, sizeof(__pyx_k_Haplotype_major_read_alleles_ran_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_Haplotype_major_read_alleles_ran_3, __pyx_k_Haplotype_major_read_alleles_ran_3, sizeof(__pyx_k_Haplotype_major_read_alleles_ran_3), 0, 1, 0, 0},
   {&__pyx_n_s_ImportError, __pyx_k_ImportError, sizeof(__pyx_k_ImportError), 0, 0, 1, 1},
   {&__pyx_kp_u_Invalid_append_dosages_batch_dos, __pyx_k_Invalid_append_dosages_batch_dos, sizeof(__pyx_k_Invalid_append_dosages_batch_dos), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_append_dosages_dosage_ar, __pyx_k_Invalid_append_dosages_dosage_ar, sizeof(__pyx_k_Invalid_append_dosages_dosage_ar), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_arguments_for_PgenWriter, __pyx_k_Invalid_arguments_for_PgenWriter, sizeof(__pyx_k_Invalid_arguments_for_PgenWriter), 0, 1, 0, 0},
+  {&__pyx_kp_u_Invalid_arguments_for_PgenWriter_2, __pyx_k_Invalid_arguments_for_PgenWriter_2, sizeof(__pyx_k_Invalid_arguments_for_PgenWriter_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_Invalid_arguments_for_PgenWriter_3, __pyx_k_Invalid_arguments_for_PgenWriter_3, sizeof(__pyx_k_Invalid_arguments_for_PgenWriter_3), 0, 1, 0, 0},
+  {&__pyx_kp_u_Invalid_arguments_for_PgenWriter_4, __pyx_k_Invalid_arguments_for_PgenWriter_4, sizeof(__pyx_k_Invalid_arguments_for_PgenWriter_4), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_read_dosages_floatarr_ou, __pyx_k_Invalid_read_dosages_floatarr_ou, sizeof(__pyx_k_Invalid_read_dosages_floatarr_ou), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_read_dosages_list_floata, __pyx_k_Invalid_read_dosages_list_floata, sizeof(__pyx_k_Invalid_read_dosages_list_floata), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_read_dosages_range_float, __pyx_k_Invalid_read_dosages_range_float, sizeof(__pyx_k_Invalid_read_dosages_range_float), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_read_geno_int_out_array, __pyx_k_Invalid_read_geno_int_out_array, sizeof(__pyx_k_Invalid_read_geno_int_out_array), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_read_list_geno_int_out_a, __pyx_k_Invalid_read_list_geno_int_out_a, sizeof(__pyx_k_Invalid_read_list_geno_int_out_a), 0, 1, 0, 0},
   {&__pyx_kp_u_Invalid_read_range_geno_int_out, __pyx_k_Invalid_read_range_geno_int_out, sizeof(__pyx_k_Invalid_read_range_geno_int_out), 0, 1, 0, 0},
   {&__pyx_n_s_MemoryError, __pyx_k_MemoryError, sizeof(__pyx_k_MemoryError), 0, 0, 1, 1},
-  {&__pyx_kp_u_Multiallelic_phase_dosage_datase, __pyx_k_Multiallelic_phase_dosage_datase, sizeof(__pyx_k_Multiallelic_phase_dosage_datase), 0, 1, 0, 0},
-  {&__pyx_kp_u_Multiallelic_variants_aren_t_sup, __pyx_k_Multiallelic_variants_aren_t_sup, sizeof(__pyx_k_Multiallelic_variants_aren_t_sup), 0, 1, 0, 0},
   {&__pyx_n_s_PgenReader, __pyx_k_PgenReader, sizeof(__pyx_k_PgenReader), 0, 0, 1, 1},
+  {&__pyx_kp_u_PgenReader_multiallelic_variants, __pyx_k_PgenReader_multiallelic_variants, sizeof(__pyx_k_PgenReader_multiallelic_variants), 0, 1, 0, 0},
   {&__pyx_n_s_PgenWriter, __pyx_k_PgenWriter, sizeof(__pyx_k_PgenWriter), 0, 0, 1, 1},
+  {&__pyx_kp_u_PgenWriter___dealloc___SpgwFinis, __pyx_k_PgenWriter___dealloc___SpgwFinis, sizeof(__pyx_k_PgenWriter___dealloc___SpgwFinis), 0, 1, 0, 0},
+  {&__pyx_kp_u_PgenWriter_close_SpgwFinish_erro, __pyx_k_PgenWriter_close_SpgwFinish_erro, sizeof(__pyx_k_PgenWriter_close_SpgwFinish_erro), 0, 1, 0, 0},
   {&__pyx_kp_u_PgenWriter_close_called_when_num, __pyx_k_PgenWriter_close_called_when_num, sizeof(__pyx_k_PgenWriter_close_called_when_num), 0, 1, 0, 0},
   {&__pyx_kp_u_PgrInit_error, __pyx_k_PgrInit_error, sizeof(__pyx_k_PgrInit_error), 0, 1, 0, 0},
   {&__pyx_n_s_RuntimeError, __pyx_k_RuntimeError, sizeof(__pyx_k_RuntimeError), 0, 0, 1, 1},
   {&__pyx_kp_u_Sample_major_read_list_geno_int, __pyx_k_Sample_major_read_list_geno_int, sizeof(__pyx_k_Sample_major_read_list_geno_int), 0, 1, 0, 0},
   {&__pyx_kp_u_Sample_major_read_list_geno_int_2, __pyx_k_Sample_major_read_list_geno_int_2, sizeof(__pyx_k_Sample_major_read_list_geno_int_2), 0, 1, 0, 0},
   {&__pyx_kp_u_Sample_major_read_range_geno_int, __pyx_k_Sample_major_read_range_geno_int, sizeof(__pyx_k_Sample_major_read_range_geno_int), 0, 1, 0, 0},
   {&__pyx_kp_u_Sample_major_read_range_geno_int_2, __pyx_k_Sample_major_read_range_geno_int_2, sizeof(__pyx_k_Sample_major_read_range_geno_int_2), 0, 1, 0, 0},
@@ -30534,50 +33729,68 @@
   {&__pyx_kp_u_Variant_major_read_list_geno_int_2, __pyx_k_Variant_major_read_list_geno_int_2, sizeof(__pyx_k_Variant_major_read_list_geno_int_2), 0, 1, 0, 0},
   {&__pyx_kp_u_Variant_major_read_range_geno_in, __pyx_k_Variant_major_read_range_geno_in, sizeof(__pyx_k_Variant_major_read_range_geno_in), 0, 1, 0, 0},
   {&__pyx_kp_u_Variant_major_read_range_geno_in_2, __pyx_k_Variant_major_read_range_geno_in_2, sizeof(__pyx_k_Variant_major_read_range_geno_in_2), 0, 1, 0, 0},
   {&__pyx_kp_u__10, __pyx_k__10, sizeof(__pyx_k__10), 0, 1, 0, 0},
   {&__pyx_kp_u__5, __pyx_k__5, sizeof(__pyx_k__5), 0, 1, 0, 0},
   {&__pyx_kp_u__9, __pyx_k__9, sizeof(__pyx_k__9), 0, 1, 0, 0},
   {&__pyx_n_s_all_phased, __pyx_k_all_phased, sizeof(__pyx_k_all_phased), 0, 0, 1, 1},
+  {&__pyx_n_s_allele_ct, __pyx_k_allele_ct, sizeof(__pyx_k_allele_ct), 0, 0, 1, 1},
+  {&__pyx_n_s_allele_ct_limit, __pyx_k_allele_ct_limit, sizeof(__pyx_k_allele_ct_limit), 0, 0, 1, 1},
+  {&__pyx_n_s_allele_cts, __pyx_k_allele_cts, sizeof(__pyx_k_allele_cts), 0, 0, 1, 1},
   {&__pyx_n_s_allele_idx, __pyx_k_allele_idx, sizeof(__pyx_k_allele_idx), 0, 0, 1, 1},
   {&__pyx_n_s_allele_idx_offsets, __pyx_k_allele_idx_offsets, sizeof(__pyx_k_allele_idx_offsets), 0, 0, 1, 1},
   {&__pyx_n_s_allele_int32, __pyx_k_allele_int32, sizeof(__pyx_k_allele_int32), 0, 0, 1, 1},
   {&__pyx_n_s_allele_int32_batch, __pyx_k_allele_int32_batch, sizeof(__pyx_k_allele_int32_batch), 0, 0, 1, 1},
   {&__pyx_n_s_allele_int32_out, __pyx_k_allele_int32_out, sizeof(__pyx_k_allele_int32_out), 0, 0, 1, 1},
   {&__pyx_kp_u_and_column_count_should_be_twic, __pyx_k_and_column_count_should_be_twic, sizeof(__pyx_k_and_column_count_should_be_twic), 0, 1, 0, 0},
   {&__pyx_kp_u_and_column_count_should_be_twic_2, __pyx_k_and_column_count_should_be_twic_2, sizeof(__pyx_k_and_column_count_should_be_twic_2), 0, 1, 0, 0},
   {&__pyx_kp_u_and_row_count_should_be_twice_t, __pyx_k_and_row_count_should_be_twice_t, sizeof(__pyx_k_and_row_count_should_be_twice_t), 0, 1, 0, 0},
   {&__pyx_n_s_append_alleles, __pyx_k_append_alleles, sizeof(__pyx_k_append_alleles), 0, 0, 1, 1},
   {&__pyx_n_s_append_alleles_batch, __pyx_k_append_alleles_batch, sizeof(__pyx_k_append_alleles_batch), 0, 0, 1, 1},
   {&__pyx_kp_u_append_alleles_batch_called_with, __pyx_k_append_alleles_batch_called_with, sizeof(__pyx_k_append_alleles_batch_called_with), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_batch_called_with_2, __pyx_k_append_alleles_batch_called_with_2, sizeof(__pyx_k_append_alleles_batch_called_with_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_batch_called_with_3, __pyx_k_append_alleles_batch_called_with_3, sizeof(__pyx_k_append_alleles_batch_called_with_3), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_batch_called_with_4, __pyx_k_append_alleles_batch_called_with_4, sizeof(__pyx_k_append_alleles_batch_called_with_4), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_batch_called_with_5, __pyx_k_append_alleles_batch_called_with_5, sizeof(__pyx_k_append_alleles_batch_called_with_5), 0, 1, 0, 0},
   {&__pyx_kp_u_append_alleles_batch_error, __pyx_k_append_alleles_batch_error, sizeof(__pyx_k_append_alleles_batch_error), 0, 1, 0, 0},
   {&__pyx_kp_u_append_alleles_called_with_all_p, __pyx_k_append_alleles_called_with_all_p, sizeof(__pyx_k_append_alleles_called_with_all_p), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_called_with_allel, __pyx_k_append_alleles_called_with_allel, sizeof(__pyx_k_append_alleles_called_with_allel), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_called_with_allel_2, __pyx_k_append_alleles_called_with_allel_2, sizeof(__pyx_k_append_alleles_called_with_allel_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_called_with_allel_3, __pyx_k_append_alleles_called_with_allel_3, sizeof(__pyx_k_append_alleles_called_with_allel_3), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_alleles_called_with_inval, __pyx_k_append_alleles_called_with_inval, sizeof(__pyx_k_append_alleles_called_with_inval), 0, 1, 0, 0},
   {&__pyx_kp_u_append_alleles_error, __pyx_k_append_alleles_error, sizeof(__pyx_k_append_alleles_error), 0, 1, 0, 0},
   {&__pyx_n_s_append_biallelic, __pyx_k_append_biallelic, sizeof(__pyx_k_append_biallelic), 0, 0, 1, 1},
   {&__pyx_n_s_append_biallelic_batch, __pyx_k_append_biallelic_batch, sizeof(__pyx_k_append_biallelic_batch), 0, 0, 1, 1},
   {&__pyx_kp_u_append_biallelic_batch_error, __pyx_k_append_biallelic_batch_error, sizeof(__pyx_k_append_biallelic_batch_error), 0, 1, 0, 0},
   {&__pyx_kp_u_append_biallelic_error, __pyx_k_append_biallelic_error, sizeof(__pyx_k_append_biallelic_error), 0, 1, 0, 0},
   {&__pyx_n_s_append_dosages, __pyx_k_append_dosages, sizeof(__pyx_k_append_dosages), 0, 0, 1, 1},
   {&__pyx_n_s_append_dosages_batch, __pyx_k_append_dosages_batch, sizeof(__pyx_k_append_dosages_batch), 0, 0, 1, 1},
   {&__pyx_kp_u_append_dosages_batch_cannot_be_c, __pyx_k_append_dosages_batch_cannot_be_c, sizeof(__pyx_k_append_dosages_batch_cannot_be_c), 0, 1, 0, 0},
   {&__pyx_kp_u_append_dosages_batch_error, __pyx_k_append_dosages_batch_error, sizeof(__pyx_k_append_dosages_batch_error), 0, 1, 0, 0},
   {&__pyx_kp_u_append_dosages_cannot_be_called, __pyx_k_append_dosages_cannot_be_called, sizeof(__pyx_k_append_dosages_cannot_be_called), 0, 1, 0, 0},
   {&__pyx_kp_u_append_dosages_error, __pyx_k_append_dosages_error, sizeof(__pyx_k_append_dosages_error), 0, 1, 0, 0},
   {&__pyx_n_s_append_partially_phased, __pyx_k_append_partially_phased, sizeof(__pyx_k_append_partially_phased), 0, 0, 1, 1},
   {&__pyx_n_s_append_partially_phased_batch, __pyx_k_append_partially_phased_batch, sizeof(__pyx_k_append_partially_phased_batch), 0, 0, 1, 1},
   {&__pyx_kp_u_append_partially_phased_batch_ca, __pyx_k_append_partially_phased_batch_ca, sizeof(__pyx_k_append_partially_phased_batch_ca), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_partially_phased_batch_ca_2, __pyx_k_append_partially_phased_batch_ca_2, sizeof(__pyx_k_append_partially_phased_batch_ca_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_partially_phased_batch_ca_3, __pyx_k_append_partially_phased_batch_ca_3, sizeof(__pyx_k_append_partially_phased_batch_ca_3), 0, 1, 0, 0},
   {&__pyx_kp_u_append_partially_phased_batch_er, __pyx_k_append_partially_phased_batch_er, sizeof(__pyx_k_append_partially_phased_batch_er), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_partially_phased_called_w, __pyx_k_append_partially_phased_called_w, sizeof(__pyx_k_append_partially_phased_called_w), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_partially_phased_called_w_2, __pyx_k_append_partially_phased_called_w_2, sizeof(__pyx_k_append_partially_phased_called_w_2), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_partially_phased_called_w_3, __pyx_k_append_partially_phased_called_w_3, sizeof(__pyx_k_append_partially_phased_called_w_3), 0, 1, 0, 0},
+  {&__pyx_kp_u_append_partially_phased_called_w_4, __pyx_k_append_partially_phased_called_w_4, sizeof(__pyx_k_append_partially_phased_called_w_4), 0, 1, 0, 0},
   {&__pyx_kp_u_append_partially_phased_cannot_b, __pyx_k_append_partially_phased_cannot_b, sizeof(__pyx_k_append_partially_phased_cannot_b), 0, 1, 0, 0},
   {&__pyx_kp_u_append_partially_phased_error, __pyx_k_append_partially_phased_error, sizeof(__pyx_k_append_partially_phased_error), 0, 1, 0, 0},
   {&__pyx_n_s_change_sample_subset, __pyx_k_change_sample_subset, sizeof(__pyx_k_change_sample_subset), 0, 0, 1, 1},
   {&__pyx_n_s_cline_in_traceback, __pyx_k_cline_in_traceback, sizeof(__pyx_k_cline_in_traceback), 0, 0, 1, 1},
   {&__pyx_n_s_close, __pyx_k_close, sizeof(__pyx_k_close), 0, 0, 1, 1},
   {&__pyx_kp_u_close_error, __pyx_k_close_error, sizeof(__pyx_k_close_error), 0, 1, 0, 0},
   {&__pyx_n_s_count, __pyx_k_count, sizeof(__pyx_k_count), 0, 0, 1, 1},
   {&__pyx_kp_u_count_error, __pyx_k_count_error, sizeof(__pyx_k_count_error), 0, 1, 0, 0},
+  {&__pyx_kp_u_count_genocount_uint32_out_buffe, __pyx_k_count_genocount_uint32_out_buffe, sizeof(__pyx_k_count_genocount_uint32_out_buffe), 0, 1, 0, 0},
   {&__pyx_kp_u_current_sample_subset_has_size, __pyx_k_current_sample_subset_has_size, sizeof(__pyx_k_current_sample_subset_has_size), 0, 1, 0, 0},
   {&__pyx_n_s_dosage_phase_present, __pyx_k_dosage_phase_present, sizeof(__pyx_k_dosage_phase_present), 0, 0, 1, 1},
   {&__pyx_n_s_dosage_present, __pyx_k_dosage_present, sizeof(__pyx_k_dosage_present), 0, 0, 1, 1},
   {&__pyx_n_s_dtype, __pyx_k_dtype, sizeof(__pyx_k_dtype), 0, 0, 1, 1},
   {&__pyx_n_s_enter, __pyx_k_enter, sizeof(__pyx_k_enter), 0, 0, 1, 1},
   {&__pyx_n_s_exc_tb, __pyx_k_exc_tb, sizeof(__pyx_k_exc_tb), 0, 0, 1, 1},
   {&__pyx_n_s_exc_type, __pyx_k_exc_type, sizeof(__pyx_k_exc_type), 0, 0, 1, 1},
@@ -30622,14 +33835,15 @@
   {&__pyx_kp_u_read_alleles_and_phasepresent_al, __pyx_k_read_alleles_and_phasepresent_al, sizeof(__pyx_k_read_alleles_and_phasepresent_al), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_and_phasepresent_er, __pyx_k_read_alleles_and_phasepresent_er, sizeof(__pyx_k_read_alleles_and_phasepresent_er), 0, 1, 0, 0},
   {&__pyx_n_s_read_alleles_and_phasepresent_li, __pyx_k_read_alleles_and_phasepresent_li, sizeof(__pyx_k_read_alleles_and_phasepresent_li), 0, 0, 1, 1},
   {&__pyx_kp_u_read_alleles_and_phasepresent_li_2, __pyx_k_read_alleles_and_phasepresent_li_2, sizeof(__pyx_k_read_alleles_and_phasepresent_li_2), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_and_phasepresent_li_3, __pyx_k_read_alleles_and_phasepresent_li_3, sizeof(__pyx_k_read_alleles_and_phasepresent_li_3), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_and_phasepresent_li_4, __pyx_k_read_alleles_and_phasepresent_li_4, sizeof(__pyx_k_read_alleles_and_phasepresent_li_4), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_and_phasepresent_ph, __pyx_k_read_alleles_and_phasepresent_ph, sizeof(__pyx_k_read_alleles_and_phasepresent_ph), 0, 1, 0, 0},
+  {&__pyx_kp_u_read_alleles_and_phasepresent_r, __pyx_k_read_alleles_and_phasepresent_r, sizeof(__pyx_k_read_alleles_and_phasepresent_r), 0, 1, 0, 0},
   {&__pyx_n_s_read_alleles_and_phasepresent_ra, __pyx_k_read_alleles_and_phasepresent_ra, sizeof(__pyx_k_read_alleles_and_phasepresent_ra), 0, 0, 1, 1},
   {&__pyx_kp_u_read_alleles_and_phasepresent_ra_2, __pyx_k_read_alleles_and_phasepresent_ra_2, sizeof(__pyx_k_read_alleles_and_phasepresent_ra_2), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_and_phasepresent_va, __pyx_k_read_alleles_and_phasepresent_va, sizeof(__pyx_k_read_alleles_and_phasepresent_va), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_error, __pyx_k_read_alleles_error, sizeof(__pyx_k_read_alleles_error), 0, 1, 0, 0},
   {&__pyx_n_s_read_alleles_list, __pyx_k_read_alleles_list, sizeof(__pyx_k_read_alleles_list), 0, 0, 1, 1},
   {&__pyx_kp_u_read_alleles_list_error, __pyx_k_read_alleles_list_error, sizeof(__pyx_k_read_alleles_list_error), 0, 1, 0, 0},
   {&__pyx_kp_u_read_alleles_list_variant_index, __pyx_k_read_alleles_list_variant_index, sizeof(__pyx_k_read_alleles_list_variant_index), 0, 1, 0, 0},
@@ -30671,384 +33885,595 @@
   {&__pyx_n_s_setstate, __pyx_k_setstate, sizeof(__pyx_k_setstate), 0, 0, 1, 1},
   {&__pyx_n_s_setstate_cython, __pyx_k_setstate_cython, sizeof(__pyx_k_setstate_cython), 0, 0, 1, 1},
   {&__pyx_n_s_size, __pyx_k_size, sizeof(__pyx_k_size), 0, 0, 1, 1},
   {&__pyx_n_s_sys, __pyx_k_sys, sizeof(__pyx_k_sys), 0, 0, 1, 1},
   {&__pyx_n_s_test, __pyx_k_test, sizeof(__pyx_k_test), 0, 0, 1, 1},
   {&__pyx_kp_u_unequal_to_initially_declared_v, __pyx_k_unequal_to_initially_declared_v, sizeof(__pyx_k_unequal_to_initially_declared_v), 0, 1, 0, 0},
   {&__pyx_n_s_variant_ct, __pyx_k_variant_ct, sizeof(__pyx_k_variant_ct), 0, 0, 1, 1},
+  {&__pyx_n_s_variant_ct_limit, __pyx_k_variant_ct_limit, sizeof(__pyx_k_variant_ct_limit), 0, 0, 1, 1},
   {&__pyx_n_s_variant_idx, __pyx_k_variant_idx, sizeof(__pyx_k_variant_idx), 0, 0, 1, 1},
   {&__pyx_kp_u_variant_idx_2, __pyx_k_variant_idx_2, sizeof(__pyx_k_variant_idx_2), 0, 1, 0, 0},
   {&__pyx_n_s_variant_idx_end, __pyx_k_variant_idx_end, sizeof(__pyx_k_variant_idx_end), 0, 0, 1, 1},
   {&__pyx_kp_u_variant_idx_end_variant_idx_sta, __pyx_k_variant_idx_end_variant_idx_sta, sizeof(__pyx_k_variant_idx_end_variant_idx_sta), 0, 1, 0, 0},
   {&__pyx_n_s_variant_idx_start, __pyx_k_variant_idx_start, sizeof(__pyx_k_variant_idx_start), 0, 0, 1, 1},
   {&__pyx_n_s_variant_idxs, __pyx_k_variant_idxs, sizeof(__pyx_k_variant_idxs), 0, 0, 1, 1},
   {&__pyx_kp_u_variant_idxs_length_is, __pyx_k_variant_idxs_length_is, sizeof(__pyx_k_variant_idxs_length_is), 0, 1, 0, 0},
   {0, 0, 0, 0, 0, 0, 0}
 };
 static CYTHON_SMALL_CODE int __Pyx_InitCachedBuiltins(void) {
-  __pyx_builtin_RuntimeError = __Pyx_GetBuiltinName(__pyx_n_s_RuntimeError); if (!__pyx_builtin_RuntimeError) __PYX_ERR(0, 223, __pyx_L1_error)
-  __pyx_builtin_MemoryError = __Pyx_GetBuiltinName(__pyx_n_s_MemoryError); if (!__pyx_builtin_MemoryError) __PYX_ERR(0, 258, __pyx_L1_error)
-  __pyx_builtin_range = __Pyx_GetBuiltinName(__pyx_n_s_range); if (!__pyx_builtin_range) __PYX_ERR(0, 494, __pyx_L1_error)
+  __pyx_builtin_MemoryError = __Pyx_GetBuiltinName(__pyx_n_s_MemoryError); if (!__pyx_builtin_MemoryError) __PYX_ERR(0, 254, __pyx_L1_error)
+  __pyx_builtin_RuntimeError = __Pyx_GetBuiltinName(__pyx_n_s_RuntimeError); if (!__pyx_builtin_RuntimeError) __PYX_ERR(0, 266, __pyx_L1_error)
+  __pyx_builtin_range = __Pyx_GetBuiltinName(__pyx_n_s_range); if (!__pyx_builtin_range) __PYX_ERR(0, 553, __pyx_L1_error)
   __pyx_builtin_TypeError = __Pyx_GetBuiltinName(__pyx_n_s_TypeError); if (!__pyx_builtin_TypeError) __PYX_ERR(1, 2, __pyx_L1_error)
   __pyx_builtin_ImportError = __Pyx_GetBuiltinName(__pyx_n_s_ImportError); if (!__pyx_builtin_ImportError) __PYX_ERR(2, 945, __pyx_L1_error)
   return 0;
   __pyx_L1_error:;
   return -1;
 }
 
 static CYTHON_SMALL_CODE int __Pyx_InitCachedConstants(void) {
   __Pyx_RefNannyDeclarations
   __Pyx_RefNannySetupContext("__Pyx_InitCachedConstants", 0);
 
-  /* "pgenlib.pyx":223
+  /* "pgenlib.pyx":266
  *         cdef uint32_t subset_size = sample_subset.size
  *         if subset_size == 0:
  *             raise RuntimeError("Empty sample_subset is not currently permitted.")             # <<<<<<<<<<<<<<
  *         cdef sample_uidx = sample_subset[0]
  *         cdef uint32_t idx = 0
  */
-  __pyx_tuple_ = PyTuple_Pack(1, __pyx_kp_u_Empty_sample_subset_is_not_curre); if (unlikely(!__pyx_tuple_)) __PYX_ERR(0, 223, __pyx_L1_error)
+  __pyx_tuple_ = PyTuple_Pack(1, __pyx_kp_u_Empty_sample_subset_is_not_curre); if (unlikely(!__pyx_tuple_)) __PYX_ERR(0, 266, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple_);
   __Pyx_GIVEREF(__pyx_tuple_);
 
-  /* "pgenlib.pyx":239
+  /* "pgenlib.pyx":282
  *             # to be returned in a different order
  *             if next_uidx <= sample_uidx:
  *                 raise RuntimeError("sample_subset is not in strictly increasing order.")             # <<<<<<<<<<<<<<
  * 
  *             sample_uidx = next_uidx
  */
-  __pyx_tuple__2 = PyTuple_Pack(1, __pyx_kp_u_sample_subset_is_not_in_strictly); if (unlikely(!__pyx_tuple__2)) __PYX_ERR(0, 239, __pyx_L1_error)
+  __pyx_tuple__2 = PyTuple_Pack(1, __pyx_kp_u_sample_subset_is_not_in_strictly); if (unlikely(!__pyx_tuple__2)) __PYX_ERR(0, 282, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__2);
   __Pyx_GIVEREF(__pyx_tuple__2);
 
-  /* "pgenlib.pyx":297
- *             # todo: support this by wrapping pvar loader the same way as the R
- *             # interface
- *             raise RuntimeError("Multiallelic + phase/dosage datasets not supported yet")             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":350
+ *             # wrapping pvar loader the same way as the R interface
+ *             if self._info_ptr[0].allele_idx_offsets == NULL:
+ *                 raise RuntimeError("PgenReader: multiallelic variants present, but allele_idx_offsets not provided")             # <<<<<<<<<<<<<<
  * 
  *         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))
  */
-  __pyx_tuple__3 = PyTuple_Pack(1, __pyx_kp_u_Multiallelic_phase_dosage_datase); if (unlikely(!__pyx_tuple__3)) __PYX_ERR(0, 297, __pyx_L1_error)
+  __pyx_tuple__3 = PyTuple_Pack(1, __pyx_kp_u_PgenReader_multiallelic_variants); if (unlikely(!__pyx_tuple__3)) __PYX_ERR(0, 350, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__3);
   __Pyx_GIVEREF(__pyx_tuple__3);
 
-  /* "pgenlib.pyx":384
+  /* "pgenlib.pyx":447
  *             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if geno_int_out.ndim != 1:
  *             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         if geno_int_out.shape[0] < subset_size:
  */
-  __pyx_tuple__4 = PyTuple_Pack(1, __pyx_kp_u_read_requires_geno_int_out_to_be); if (unlikely(!__pyx_tuple__4)) __PYX_ERR(0, 384, __pyx_L1_error)
+  __pyx_tuple__4 = PyTuple_Pack(1, __pyx_kp_u_read_requires_geno_int_out_to_be); if (unlikely(!__pyx_tuple__4)) __PYX_ERR(0, 447, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__4);
   __Pyx_GIVEREF(__pyx_tuple__4);
 
-  /* "pgenlib.pyx":406
+  /* "pgenlib.pyx":469
  *             GenoarrToInt64sMinus9(genovec, subset_size, data64_ptr)
  *         else:
  *             raise RuntimeError("Invalid read() geno_int_out array element type (int8, int32, or int64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__6 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_geno_int_out_array); if (unlikely(!__pyx_tuple__6)) __PYX_ERR(0, 406, __pyx_L1_error)
+  __pyx_tuple__6 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_geno_int_out_array); if (unlikely(!__pyx_tuple__6)) __PYX_ERR(0, 469, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__6);
   __Pyx_GIVEREF(__pyx_tuple__6);
 
-  /* "pgenlib.pyx":414
+  /* "pgenlib.pyx":477
  *             raise RuntimeError("read_dosages() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
  *         if floatarr_out.ndim != 1:
  *             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")             # <<<<<<<<<<<<<<
  *         cdef uint32_t subset_size = self._subset_size
  *         if floatarr_out.shape[0] < subset_size:
  */
-  __pyx_tuple__7 = PyTuple_Pack(1, __pyx_kp_u_read_dosages_requires_floatarr_o); if (unlikely(!__pyx_tuple__7)) __PYX_ERR(0, 414, __pyx_L1_error)
+  __pyx_tuple__7 = PyTuple_Pack(1, __pyx_kp_u_read_dosages_requires_floatarr_o); if (unlikely(!__pyx_tuple__7)) __PYX_ERR(0, 477, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__7);
   __Pyx_GIVEREF(__pyx_tuple__7);
 
-  /* "pgenlib.pyx":432
- *             Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)
+  /* "pgenlib.pyx":495
+ *             Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)
  *         else:
  *             raise RuntimeError("Invalid read_dosages() floatarr_out array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__8 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_dosages_floatarr_ou); if (unlikely(!__pyx_tuple__8)) __PYX_ERR(0, 432, __pyx_L1_error)
+  __pyx_tuple__8 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_dosages_floatarr_ou); if (unlikely(!__pyx_tuple__8)) __PYX_ERR(0, 495, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__8);
   __Pyx_GIVEREF(__pyx_tuple__8);
 
-  /* "pgenlib.pyx":687
+  /* "pgenlib.pyx":746
  *             self.read_range_internal64(variant_idx_start, variant_idx_end, geno_int_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_range() geno_int_out array element type (int8, int32, or int64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__11 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_range_geno_int_out); if (unlikely(!__pyx_tuple__11)) __PYX_ERR(0, 687, __pyx_L1_error)
+  __pyx_tuple__11 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_range_geno_int_out); if (unlikely(!__pyx_tuple__11)) __PYX_ERR(0, 746, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__11);
   __Pyx_GIVEREF(__pyx_tuple__11);
 
-  /* "pgenlib.pyx":915
+  /* "pgenlib.pyx":974
  *             self.read_list_internal64(variant_idxs, geno_int_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_list() geno_int_out array element type (int8, int32, or int64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__12 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_list_geno_int_out_a); if (unlikely(!__pyx_tuple__12)) __PYX_ERR(0, 915, __pyx_L1_error)
+  __pyx_tuple__12 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_list_geno_int_out_a); if (unlikely(!__pyx_tuple__12)) __PYX_ERR(0, 974, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_tuple__12);
   __Pyx_GIVEREF(__pyx_tuple__12);
 
-  /* "pgenlib.pyx":1158
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+  /* "pgenlib.pyx":1017
+ *         if allele_idx_offsets != NULL:
+ *             if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+ *                 raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")             # <<<<<<<<<<<<<<
+ *         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
+ *         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
+ */
+  __pyx_tuple__13 = PyTuple_Pack(1, __pyx_kp_u_Haplotype_major_read_alleles_ran_3); if (unlikely(!__pyx_tuple__13)) __PYX_ERR(0, 1017, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__13);
+  __Pyx_GIVEREF(__pyx_tuple__13);
+
+  /* "pgenlib.pyx":1152
+ *                 if allele_idx_offsets != NULL:
+ *                     if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
+ *                         raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
+ *                 if reterr != kPglRetSuccess:
+ */
+  __pyx_tuple__14 = PyTuple_Pack(1, __pyx_kp_u_Haplotype_major_read_alleles_lis_3); if (unlikely(!__pyx_tuple__14)) __PYX_ERR(0, 1152, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__14);
+  __Pyx_GIVEREF(__pyx_tuple__14);
+
+  /* "pgenlib.pyx":1218
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return
  *         raise RuntimeError("read_alleles_and_phasepresent_range() does not support hap_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_tuple__13 = PyTuple_Pack(1, __pyx_kp_u_read_alleles_and_phasepresent_ra_2); if (unlikely(!__pyx_tuple__13)) __PYX_ERR(0, 1158, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__13);
-  __Pyx_GIVEREF(__pyx_tuple__13);
+  __pyx_tuple__15 = PyTuple_Pack(1, __pyx_kp_u_read_alleles_and_phasepresent_ra_2); if (unlikely(!__pyx_tuple__15)) __PYX_ERR(0, 1218, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__15);
+  __Pyx_GIVEREF(__pyx_tuple__15);
 
-  /* "pgenlib.pyx":1202
- *                 GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+  /* "pgenlib.pyx":1260
+ *                 GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
  *             return
  *         raise RuntimeError("read_alleles_and_phasepresent_list() does not support hap_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_tuple__14 = PyTuple_Pack(1, __pyx_kp_u_read_alleles_and_phasepresent_li_4); if (unlikely(!__pyx_tuple__14)) __PYX_ERR(0, 1202, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__14);
-  __Pyx_GIVEREF(__pyx_tuple__14);
+  __pyx_tuple__16 = PyTuple_Pack(1, __pyx_kp_u_read_alleles_and_phasepresent_li_4); if (unlikely(!__pyx_tuple__16)) __PYX_ERR(0, 1260, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__16);
+  __Pyx_GIVEREF(__pyx_tuple__16);
 
-  /* "pgenlib.pyx":1226
+  /* "pgenlib.pyx":1284
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return
  *         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_tuple__15 = PyTuple_Pack(1, __pyx_kp_u_read_dosages_range_does_not_supp); if (unlikely(!__pyx_tuple__15)) __PYX_ERR(0, 1226, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__15);
-  __Pyx_GIVEREF(__pyx_tuple__15);
+  __pyx_tuple__17 = PyTuple_Pack(1, __pyx_kp_u_read_dosages_range_does_not_supp); if (unlikely(!__pyx_tuple__17)) __PYX_ERR(0, 1284, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__17);
+  __Pyx_GIVEREF(__pyx_tuple__17);
 
-  /* "pgenlib.pyx":1261
+  /* "pgenlib.pyx":1319
  *             self.read_dosages_range_internal64(variant_idx_start, variant_idx_end, floatarr_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_dosages_range() floatarr_out array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__16 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_dosages_range_float); if (unlikely(!__pyx_tuple__16)) __PYX_ERR(0, 1261, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__16);
-  __Pyx_GIVEREF(__pyx_tuple__16);
+  __pyx_tuple__18 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_dosages_range_float); if (unlikely(!__pyx_tuple__18)) __PYX_ERR(0, 1319, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__18);
+  __Pyx_GIVEREF(__pyx_tuple__18);
 
-  /* "pgenlib.pyx":1295
+  /* "pgenlib.pyx":1353
  *                 Dosage16ToFloatsMinus9(genovec, dosage_present, dosage_main, subset_size, dosage_ct, data32_ptr)
  *             return
  *         raise RuntimeError("read_dosages_list() does not support sample_maj == 1 yet.")             # <<<<<<<<<<<<<<
  * 
  * 
  */
-  __pyx_tuple__17 = PyTuple_Pack(1, __pyx_kp_u_read_dosages_list_does_not_suppo); if (unlikely(!__pyx_tuple__17)) __PYX_ERR(0, 1295, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__17);
-  __Pyx_GIVEREF(__pyx_tuple__17);
+  __pyx_tuple__19 = PyTuple_Pack(1, __pyx_kp_u_read_dosages_list_does_not_suppo); if (unlikely(!__pyx_tuple__19)) __PYX_ERR(0, 1353, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__19);
+  __Pyx_GIVEREF(__pyx_tuple__19);
 
-  /* "pgenlib.pyx":1337
+  /* "pgenlib.pyx":1395
  *             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)
  *         else:
  *             raise RuntimeError("Invalid read_dosages_list() floatarr_out array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__18 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_dosages_list_floata); if (unlikely(!__pyx_tuple__18)) __PYX_ERR(0, 1337, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__18);
-  __Pyx_GIVEREF(__pyx_tuple__18);
+  __pyx_tuple__20 = PyTuple_Pack(1, __pyx_kp_u_Invalid_read_dosages_list_floata); if (unlikely(!__pyx_tuple__20)) __PYX_ERR(0, 1395, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__20);
+  __Pyx_GIVEREF(__pyx_tuple__20);
+
+  /* "pgenlib.pyx":1424
+ *                 type_ct = (allele_ct * (allele_ct + 1)) >> 1
+ *                 if genocount_uint32_out.shape[0] <= type_ct:
+ *                     raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")             # <<<<<<<<<<<<<<
+ *                 reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+ *                 if reterr != kPglRetSuccess:
+ */
+  __pyx_tuple__21 = PyTuple_Pack(1, __pyx_kp_u_count_genocount_uint32_out_buffe); if (unlikely(!__pyx_tuple__21)) __PYX_ERR(0, 1424, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__21);
+  __Pyx_GIVEREF(__pyx_tuple__21);
 
   /* "(tree fragment)":2
  * def __reduce_cython__(self):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  */
-  __pyx_tuple__19 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__19)) __PYX_ERR(1, 2, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__19);
-  __Pyx_GIVEREF(__pyx_tuple__19);
+  __pyx_tuple__22 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__22)) __PYX_ERR(1, 2, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__22);
+  __Pyx_GIVEREF(__pyx_tuple__22);
 
   /* "(tree fragment)":4
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  */
-  __pyx_tuple__20 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__20)) __PYX_ERR(1, 4, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__20);
-  __Pyx_GIVEREF(__pyx_tuple__20);
+  __pyx_tuple__23 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__23)) __PYX_ERR(1, 4, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__23);
+  __Pyx_GIVEREF(__pyx_tuple__23);
+
+  /* "pgenlib.pyx":1557
+ *             cur_variant_ct_limit = variant_ct_limit
+ *         else:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")             # <<<<<<<<<<<<<<
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
+ */
+  __pyx_tuple__24 = PyTuple_Pack(1, __pyx_kp_u_Invalid_arguments_for_PgenWriter); if (unlikely(!__pyx_tuple__24)) __PYX_ERR(0, 1557, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__24);
+  __Pyx_GIVEREF(__pyx_tuple__24);
+
+  /* "pgenlib.pyx":1559
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")
+ *         if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")             # <<<<<<<<<<<<<<
+ *         if dosage_phase_present and not dosage_present:
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
+ */
+  __pyx_tuple__25 = PyTuple_Pack(1, __pyx_kp_u_Invalid_arguments_for_PgenWriter_2); if (unlikely(!__pyx_tuple__25)) __PYX_ERR(0, 1559, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__25);
+  __Pyx_GIVEREF(__pyx_tuple__25);
 
-  /* "pgenlib.pyx":1431
- *                   bint dosage_phase_present = False):
+  /* "pgenlib.pyx":1561
+ *             raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
  *         if dosage_phase_present and not dosage_present:
  *             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")             # <<<<<<<<<<<<<<
- *         if allele_idx_offsets is not None:
- *             for uii in range(variant_ct + 1):
+ *         if allele_ct_limit != 2:
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
  */
-  __pyx_tuple__21 = PyTuple_Pack(1, __pyx_kp_u_Invalid_arguments_for_PgenWriter); if (unlikely(!__pyx_tuple__21)) __PYX_ERR(0, 1431, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__21);
-  __Pyx_GIVEREF(__pyx_tuple__21);
+  __pyx_tuple__26 = PyTuple_Pack(1, __pyx_kp_u_Invalid_arguments_for_PgenWriter_3); if (unlikely(!__pyx_tuple__26)) __PYX_ERR(0, 1561, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__26);
+  __Pyx_GIVEREF(__pyx_tuple__26);
 
-  /* "pgenlib.pyx":1435
- *             for uii in range(variant_ct + 1):
- *                 if allele_idx_offsets[uii] != uii * 2:
- *                     raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")             # <<<<<<<<<<<<<<
+  /* "pgenlib.pyx":1564
+ *         if allele_ct_limit != 2:
+ *             if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
+ *                 raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")             # <<<<<<<<<<<<<<
+ *             write_mode = kPgenWriteAndCopy
  * 
- *         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
  */
-  __pyx_tuple__22 = PyTuple_Pack(1, __pyx_kp_u_Multiallelic_variants_aren_t_sup); if (unlikely(!__pyx_tuple__22)) __PYX_ERR(0, 1435, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__22);
-  __Pyx_GIVEREF(__pyx_tuple__22);
+  __pyx_tuple__27 = PyTuple_Pack(1, __pyx_kp_u_Invalid_arguments_for_PgenWriter_4); if (unlikely(!__pyx_tuple__27)) __PYX_ERR(0, 1564, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__27);
+  __Pyx_GIVEREF(__pyx_tuple__27);
 
-  /* "pgenlib.pyx":1503
- *         else:
+  /* "pgenlib.pyx":1662
+ *         if all_phased:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *                 raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
- *             AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
- *             reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+ *             phaseinfo = self._phaseinfo
+ *         cdef uint32_t patch_01_ct
  */
-  __pyx_tuple__23 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_called_with_all_p); if (unlikely(!__pyx_tuple__23)) __PYX_ERR(0, 1503, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__23);
-  __Pyx_GIVEREF(__pyx_tuple__23);
+  __pyx_tuple__28 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_called_with_all_p); if (unlikely(!__pyx_tuple__28)) __PYX_ERR(0, 1662, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__28);
+  __Pyx_GIVEREF(__pyx_tuple__28);
+
+  /* "pgenlib.pyx":1668
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, phaseinfo)
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_alleles called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ */
+  __pyx_tuple__29 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_called_with_inval); if (unlikely(!__pyx_tuple__29)) __PYX_ERR(0, 1668, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__29);
+  __Pyx_GIVEREF(__pyx_tuple__29);
 
-  /* "pgenlib.pyx":1513
- *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):
+  /* "pgenlib.pyx":1671
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ */
+  __pyx_tuple__30 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_called_with_allel); if (unlikely(!__pyx_tuple__30)) __PYX_ERR(0, 1671, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__30);
+  __Pyx_GIVEREF(__pyx_tuple__30);
+
+  /* "pgenlib.pyx":1674
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")             # <<<<<<<<<<<<<<
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")
+ */
+  __pyx_tuple__31 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_called_with_allel_2); if (unlikely(!__pyx_tuple__31)) __PYX_ERR(0, 1674, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__31);
+  __Pyx_GIVEREF(__pyx_tuple__31);
+
+  /* "pgenlib.pyx":1676
+ *                 raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ */
+  __pyx_tuple__32 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_called_with_allel_3); if (unlikely(!__pyx_tuple__32)) __PYX_ERR(0, 1676, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__32);
+  __Pyx_GIVEREF(__pyx_tuple__32);
+
+  /* "pgenlib.pyx":1696
+ *     cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
  *         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
  *         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
  */
-  __pyx_tuple__24 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_cannot_b); if (unlikely(!__pyx_tuple__24)) __PYX_ERR(0, 1513, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__24);
-  __Pyx_GIVEREF(__pyx_tuple__24);
+  __pyx_tuple__33 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_cannot_b); if (unlikely(!__pyx_tuple__33)) __PYX_ERR(0, 1696, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__33);
+  __Pyx_GIVEREF(__pyx_tuple__33);
+
+  /* "pgenlib.pyx":1712
+ *         cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *         if observed_allele_ct == -1:
+ *             raise RuntimeError("append_partially_phased called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ */
+  __pyx_tuple__34 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_called_w); if (unlikely(!__pyx_tuple__34)) __PYX_ERR(0, 1712, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__34);
+  __Pyx_GIVEREF(__pyx_tuple__34);
 
-  /* "pgenlib.pyx":1550
+  /* "pgenlib.pyx":1715
+ *         cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+ *         if write_allele_ct > allele_ct_limit:
+ *             raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ */
+  __pyx_tuple__35 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_called_w_2); if (unlikely(!__pyx_tuple__35)) __PYX_ERR(0, 1715, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__35);
+  __Pyx_GIVEREF(__pyx_tuple__35);
+
+  /* "pgenlib.pyx":1718
+ *         if allele_ct is not None:
+ *             if allele_ct < write_allele_ct:
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")             # <<<<<<<<<<<<<<
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")
+ */
+  __pyx_tuple__36 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_called_w_3); if (unlikely(!__pyx_tuple__36)) __PYX_ERR(0, 1718, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__36);
+  __Pyx_GIVEREF(__pyx_tuple__36);
+
+  /* "pgenlib.pyx":1720
+ *                 raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+ *             if allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *             write_allele_ct = allele_ct
+ *         cdef PglErr reterr
+ */
+  __pyx_tuple__37 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_called_w_4); if (unlikely(!__pyx_tuple__37)) __PYX_ERR(0, 1720, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__37);
+  __Pyx_GIVEREF(__pyx_tuple__37);
+
+  /* "pgenlib.pyx":1756
  *     cpdef append_dosages(self, np.ndarray floatarr):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages cannot be called when PgenWriter was constructed with dosage_present False")             # <<<<<<<<<<<<<<
  *         if floatarr.dtype == np.float32:
  *             self.append_dosages_internal32(floatarr)
  */
-  __pyx_tuple__25 = PyTuple_Pack(1, __pyx_kp_u_append_dosages_cannot_be_called); if (unlikely(!__pyx_tuple__25)) __PYX_ERR(0, 1550, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__25);
-  __Pyx_GIVEREF(__pyx_tuple__25);
+  __pyx_tuple__38 = PyTuple_Pack(1, __pyx_kp_u_append_dosages_cannot_be_called); if (unlikely(!__pyx_tuple__38)) __PYX_ERR(0, 1756, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__38);
+  __Pyx_GIVEREF(__pyx_tuple__38);
 
-  /* "pgenlib.pyx":1556
+  /* "pgenlib.pyx":1762
  *             self.append_dosages_internal64(floatarr)
  *         else:
  *             raise RuntimeError("Invalid append_dosages() dosage array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__26 = PyTuple_Pack(1, __pyx_kp_u_Invalid_append_dosages_dosage_ar); if (unlikely(!__pyx_tuple__26)) __PYX_ERR(0, 1556, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__26);
-  __Pyx_GIVEREF(__pyx_tuple__26);
+  __pyx_tuple__39 = PyTuple_Pack(1, __pyx_kp_u_Invalid_append_dosages_dosage_ar); if (unlikely(!__pyx_tuple__39)) __PYX_ERR(0, 1762, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__39);
+  __Pyx_GIVEREF(__pyx_tuple__39);
+
+  /* "pgenlib.pyx":1802
+ *                 observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
+ *                 if observed_allele_ct == -1:
+ *                     raise RuntimeError("append_alleles_batch called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:
+ */
+  __pyx_tuple__40 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_batch_called_with); if (unlikely(!__pyx_tuple__40)) __PYX_ERR(0, 1802, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__40);
+  __Pyx_GIVEREF(__pyx_tuple__40);
+
+  /* "pgenlib.pyx":1805
+ *                 write_allele_ct = <uint32_t>(observed_allele_ct)
+ *                 if write_allele_ct > allele_ct_limit:
+ *                     raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *                 if allele_cts is not None:
+ *                     casted_allele_ct = allele_cts[uii]
+ */
+  __pyx_tuple__41 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_batch_called_with_2); if (unlikely(!__pyx_tuple__41)) __PYX_ERR(0, 1805, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__41);
+  __Pyx_GIVEREF(__pyx_tuple__41);
+
+  /* "pgenlib.pyx":1809
+ *                     casted_allele_ct = allele_cts[uii]
+ *                     if casted_allele_ct < write_allele_ct:
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")             # <<<<<<<<<<<<<<
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+ */
+  __pyx_tuple__42 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_batch_called_with_3); if (unlikely(!__pyx_tuple__42)) __PYX_ERR(0, 1809, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__42);
+  __Pyx_GIVEREF(__pyx_tuple__42);
+
+  /* "pgenlib.pyx":1811
+ *                         raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+ *                     if casted_allele_ct > allele_ct_limit:
+ *                         raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")             # <<<<<<<<<<<<<<
+ *                     write_allele_ct = casted_allele_ct
+ *                 if (patch_01_ct == 0) and (patch_10_ct == 0):
+ */
+  __pyx_tuple__43 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_batch_called_with_4); if (unlikely(!__pyx_tuple__43)) __PYX_ERR(0, 1811, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__43);
+  __Pyx_GIVEREF(__pyx_tuple__43);
 
-  /* "pgenlib.pyx":1589
+  /* "pgenlib.pyx":1821
  *         else:
  *             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
  *             for uii in range(batch_size):
  *                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
  */
-  __pyx_tuple__27 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_batch_called_with); if (unlikely(!__pyx_tuple__27)) __PYX_ERR(0, 1589, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__27);
-  __Pyx_GIVEREF(__pyx_tuple__27);
+  __pyx_tuple__44 = PyTuple_Pack(1, __pyx_kp_u_append_alleles_batch_called_with_5); if (unlikely(!__pyx_tuple__44)) __PYX_ERR(0, 1821, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__44);
+  __Pyx_GIVEREF(__pyx_tuple__44);
 
-  /* "pgenlib.pyx":1601
- *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):
+  /* "pgenlib.pyx":1848
+ *     cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):
  *         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
  *             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")             # <<<<<<<<<<<<<<
  *         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
- *         cdef uintptr_t* genovec = self._genovec
+ *         cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
  */
-  __pyx_tuple__28 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_batch_ca); if (unlikely(!__pyx_tuple__28)) __PYX_ERR(0, 1601, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__28);
-  __Pyx_GIVEREF(__pyx_tuple__28);
+  __pyx_tuple__45 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_batch_ca); if (unlikely(!__pyx_tuple__45)) __PYX_ERR(0, 1848, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__45);
+  __Pyx_GIVEREF(__pyx_tuple__45);
+
+  /* "pgenlib.pyx":1873
+ *             observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+ *             if observed_allele_ct == -1:
+ *                 raise RuntimeError("append_partially_phased_batch called with invalid allele codes")             # <<<<<<<<<<<<<<
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ *             if write_allele_ct > allele_ct_limit:
+ */
+  __pyx_tuple__46 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_batch_ca_2); if (unlikely(!__pyx_tuple__46)) __PYX_ERR(0, 1873, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__46);
+  __Pyx_GIVEREF(__pyx_tuple__46);
+
+  /* "pgenlib.pyx":1876
+ *             write_allele_ct = <uint32_t>(observed_allele_ct)
+ *             if write_allele_ct > allele_ct_limit:
+ *                 raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")             # <<<<<<<<<<<<<<
+ *             if allele_cts is not None:
+ *                 casted_allele_ct = allele_cts[uii]
+ */
+  __pyx_tuple__47 = PyTuple_Pack(1, __pyx_kp_u_append_partially_phased_batch_ca_3); if (unlikely(!__pyx_tuple__47)) __PYX_ERR(0, 1876, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__47);
+  __Pyx_GIVEREF(__pyx_tuple__47);
 
-  /* "pgenlib.pyx":1652
+  /* "pgenlib.pyx":1925
  *     cpdef append_dosages_batch(self, np.ndarray floatarr_batch):
  *         if (self._phase_dosage_gflags & kfPgenGlobalDosagePresent) == 0:
  *             raise RuntimeError("append_dosages_batch cannot be called when PgenWriter was constructed with dosage_present False")             # <<<<<<<<<<<<<<
  *         if floatarr_batch.dtype == np.float32:
  *             self.append_dosages_batch_internal32(floatarr_batch)
  */
-  __pyx_tuple__29 = PyTuple_Pack(1, __pyx_kp_u_append_dosages_batch_cannot_be_c); if (unlikely(!__pyx_tuple__29)) __PYX_ERR(0, 1652, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__29);
-  __Pyx_GIVEREF(__pyx_tuple__29);
+  __pyx_tuple__48 = PyTuple_Pack(1, __pyx_kp_u_append_dosages_batch_cannot_be_c); if (unlikely(!__pyx_tuple__48)) __PYX_ERR(0, 1925, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__48);
+  __Pyx_GIVEREF(__pyx_tuple__48);
 
-  /* "pgenlib.pyx":1658
+  /* "pgenlib.pyx":1931
  *             self.append_dosages_batch_internal64(floatarr_batch)
  *         else:
  *             raise RuntimeError("Invalid append_dosages_batch() dosage array element type (float32 or float64 expected).")             # <<<<<<<<<<<<<<
  *         return
  * 
  */
-  __pyx_tuple__30 = PyTuple_Pack(1, __pyx_kp_u_Invalid_append_dosages_batch_dos); if (unlikely(!__pyx_tuple__30)) __PYX_ERR(0, 1658, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__30);
-  __Pyx_GIVEREF(__pyx_tuple__30);
+  __pyx_tuple__49 = PyTuple_Pack(1, __pyx_kp_u_Invalid_append_dosages_batch_dos); if (unlikely(!__pyx_tuple__49)) __PYX_ERR(0, 1931, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__49);
+  __Pyx_GIVEREF(__pyx_tuple__49);
 
   /* "(tree fragment)":2
  * def __reduce_cython__(self):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  */
-  __pyx_tuple__31 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__31)) __PYX_ERR(1, 2, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__31);
-  __Pyx_GIVEREF(__pyx_tuple__31);
+  __pyx_tuple__50 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__50)) __PYX_ERR(1, 2, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__50);
+  __Pyx_GIVEREF(__pyx_tuple__50);
 
   /* "(tree fragment)":4
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")
  * def __setstate_cython__(self, __pyx_state):
  *     raise TypeError("no default __reduce__ due to non-trivial __cinit__")             # <<<<<<<<<<<<<<
  */
-  __pyx_tuple__32 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__32)) __PYX_ERR(1, 4, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__32);
-  __Pyx_GIVEREF(__pyx_tuple__32);
+  __pyx_tuple__51 = PyTuple_Pack(1, __pyx_kp_s_no_default___reduce___due_to_non); if (unlikely(!__pyx_tuple__51)) __PYX_ERR(1, 4, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__51);
+  __Pyx_GIVEREF(__pyx_tuple__51);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":945
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":945
  *         __pyx_import_array()
  *     except Exception:
  *         raise ImportError("numpy.core.multiarray failed to import")             # <<<<<<<<<<<<<<
  * 
  * cdef inline int import_umath() except -1:
  */
-  __pyx_tuple__33 = PyTuple_Pack(1, __pyx_kp_u_numpy_core_multiarray_failed_to); if (unlikely(!__pyx_tuple__33)) __PYX_ERR(2, 945, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__33);
-  __Pyx_GIVEREF(__pyx_tuple__33);
+  __pyx_tuple__52 = PyTuple_Pack(1, __pyx_kp_u_numpy_core_multiarray_failed_to); if (unlikely(!__pyx_tuple__52)) __PYX_ERR(2, 945, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__52);
+  __Pyx_GIVEREF(__pyx_tuple__52);
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":951
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":951
  *         _import_umath()
  *     except Exception:
  *         raise ImportError("numpy.core.umath failed to import")             # <<<<<<<<<<<<<<
  * 
  * cdef inline int import_ufunc() except -1:
  */
-  __pyx_tuple__34 = PyTuple_Pack(1, __pyx_kp_u_numpy_core_umath_failed_to_impor); if (unlikely(!__pyx_tuple__34)) __PYX_ERR(2, 951, __pyx_L1_error)
-  __Pyx_GOTREF(__pyx_tuple__34);
-  __Pyx_GIVEREF(__pyx_tuple__34);
+  __pyx_tuple__53 = PyTuple_Pack(1, __pyx_kp_u_numpy_core_umath_failed_to_impor); if (unlikely(!__pyx_tuple__53)) __PYX_ERR(2, 951, __pyx_L1_error)
+  __Pyx_GOTREF(__pyx_tuple__53);
+  __Pyx_GIVEREF(__pyx_tuple__53);
   __Pyx_RefNannyFinishContext();
   return 0;
   __pyx_L1_error:;
   __Pyx_RefNannyFinishContext();
   return -1;
 }
 
 static CYTHON_SMALL_CODE int __Pyx_InitGlobals(void) {
   if (__Pyx_InitStrings(__pyx_string_tab) < 0) __PYX_ERR(0, 1, __pyx_L1_error)
   __pyx_int_0 = PyInt_FromLong(0); if (unlikely(!__pyx_int_0)) __PYX_ERR(0, 1, __pyx_L1_error)
   __pyx_int_1 = PyInt_FromLong(1); if (unlikely(!__pyx_int_1)) __PYX_ERR(0, 1, __pyx_L1_error)
   __pyx_int_2 = PyInt_FromLong(2); if (unlikely(!__pyx_int_2)) __PYX_ERR(0, 1, __pyx_L1_error)
+  __pyx_int_3 = PyInt_FromLong(3); if (unlikely(!__pyx_int_3)) __PYX_ERR(0, 1, __pyx_L1_error)
   return 0;
   __pyx_L1_error:;
   return -1;
 }
 
 static CYTHON_SMALL_CODE int __Pyx_modinit_global_init_code(void); /*proto*/
 static CYTHON_SMALL_CODE int __Pyx_modinit_variable_export_code(void); /*proto*/
@@ -31086,14 +34511,15 @@
   __Pyx_RefNannyDeclarations
   int __pyx_lineno = 0;
   const char *__pyx_filename = NULL;
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__Pyx_modinit_type_init_code", 0);
   /*--- Type init code ---*/
   __pyx_vtabptr_7pgenlib_PgenReader = &__pyx_vtable_7pgenlib_PgenReader;
+  __pyx_vtable_7pgenlib_PgenReader.set_allele_idx_offsets_internal = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *))__pyx_f_7pgenlib_10PgenReader_set_allele_idx_offsets_internal;
   __pyx_vtable_7pgenlib_PgenReader.set_sample_subset_internal = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *))__pyx_f_7pgenlib_10PgenReader_set_sample_subset_internal;
   __pyx_vtable_7pgenlib_PgenReader.__pyx___enter__ = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenReader___enter__;
   __pyx_vtable_7pgenlib_PgenReader.get_raw_sample_ct = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenReader_get_raw_sample_ct;
   __pyx_vtable_7pgenlib_PgenReader.get_variant_ct = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenReader_get_variant_ct;
   __pyx_vtable_7pgenlib_PgenReader.hardcall_phase_present = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenReader_hardcall_phase_present;
   __pyx_vtable_7pgenlib_PgenReader.read = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, uint32_t, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_read;
   __pyx_vtable_7pgenlib_PgenReader.read_dosages = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, uint32_t, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_read_dosages;
@@ -31117,51 +34543,51 @@
   __pyx_vtable_7pgenlib_PgenReader.read_dosages_list_internal32 = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *, PyArrayObject *, struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list_internal32 *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_read_dosages_list_internal32;
   __pyx_vtable_7pgenlib_PgenReader.read_dosages_list_internal64 = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *, PyArrayObject *, struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list_internal64 *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_read_dosages_list_internal64;
   __pyx_vtable_7pgenlib_PgenReader.read_dosages_list = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_read_dosages_list *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_read_dosages_list;
   __pyx_vtable_7pgenlib_PgenReader.count = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, uint32_t, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_count *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_count;
   __pyx_vtable_7pgenlib_PgenReader.change_sample_subset = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenReader_change_sample_subset *__pyx_optional_args))__pyx_f_7pgenlib_10PgenReader_change_sample_subset;
   __pyx_vtable_7pgenlib_PgenReader.close = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenReader_close;
   __pyx_vtable_7pgenlib_PgenReader.__pyx___exit__ = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenReader *, PyObject *, PyObject *, PyObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenReader___exit__;
-  if (PyType_Ready(&__pyx_type_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 188, __pyx_L1_error)
+  if (PyType_Ready(&__pyx_type_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 227, __pyx_L1_error)
   #if PY_VERSION_HEX < 0x030800B1
   __pyx_type_7pgenlib_PgenReader.tp_print = 0;
   #endif
   if ((CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP) && likely(!__pyx_type_7pgenlib_PgenReader.tp_dictoffset && __pyx_type_7pgenlib_PgenReader.tp_getattro == PyObject_GenericGetAttr)) {
     __pyx_type_7pgenlib_PgenReader.tp_getattro = __Pyx_PyObject_GenericGetAttr;
   }
-  if (__Pyx_SetVtable(__pyx_type_7pgenlib_PgenReader.tp_dict, __pyx_vtabptr_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 188, __pyx_L1_error)
-  if (PyObject_SetAttr(__pyx_m, __pyx_n_s_PgenReader, (PyObject *)&__pyx_type_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 188, __pyx_L1_error)
-  if (__Pyx_setup_reduce((PyObject*)&__pyx_type_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 188, __pyx_L1_error)
+  if (__Pyx_SetVtable(__pyx_type_7pgenlib_PgenReader.tp_dict, __pyx_vtabptr_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 227, __pyx_L1_error)
+  if (PyObject_SetAttr(__pyx_m, __pyx_n_s_PgenReader, (PyObject *)&__pyx_type_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 227, __pyx_L1_error)
+  if (__Pyx_setup_reduce((PyObject*)&__pyx_type_7pgenlib_PgenReader) < 0) __PYX_ERR(0, 227, __pyx_L1_error)
   __pyx_ptype_7pgenlib_PgenReader = &__pyx_type_7pgenlib_PgenReader;
   __pyx_vtabptr_7pgenlib_PgenWriter = &__pyx_vtable_7pgenlib_PgenWriter;
   __pyx_vtable_7pgenlib_PgenWriter.__pyx___enter__ = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter___enter__;
   __pyx_vtable_7pgenlib_PgenWriter.append_biallelic = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_append_biallelic;
   __pyx_vtable_7pgenlib_PgenWriter.append_alleles = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles *__pyx_optional_args))__pyx_f_7pgenlib_10PgenWriter_append_alleles;
-  __pyx_vtable_7pgenlib_PgenWriter.append_partially_phased = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_append_partially_phased;
+  __pyx_vtable_7pgenlib_PgenWriter.append_partially_phased = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased *__pyx_optional_args))__pyx_f_7pgenlib_10PgenWriter_append_partially_phased;
   __pyx_vtable_7pgenlib_PgenWriter.append_dosages_internal32 = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *))__pyx_f_7pgenlib_10PgenWriter_append_dosages_internal32;
   __pyx_vtable_7pgenlib_PgenWriter.append_dosages_internal64 = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *))__pyx_f_7pgenlib_10PgenWriter_append_dosages_internal64;
   __pyx_vtable_7pgenlib_PgenWriter.append_dosages = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_append_dosages;
   __pyx_vtable_7pgenlib_PgenWriter.append_biallelic_batch = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_append_biallelic_batch;
   __pyx_vtable_7pgenlib_PgenWriter.append_alleles_batch = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_alleles_batch *__pyx_optional_args))__pyx_f_7pgenlib_10PgenWriter_append_alleles_batch;
-  __pyx_vtable_7pgenlib_PgenWriter.append_partially_phased_batch = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch;
+  __pyx_vtable_7pgenlib_PgenWriter.append_partially_phased_batch = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_7pgenlib_10PgenWriter_append_partially_phased_batch *__pyx_optional_args))__pyx_f_7pgenlib_10PgenWriter_append_partially_phased_batch;
   __pyx_vtable_7pgenlib_PgenWriter.append_dosages_batch_internal32 = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *))__pyx_f_7pgenlib_10PgenWriter_append_dosages_batch_internal32;
   __pyx_vtable_7pgenlib_PgenWriter.append_dosages_batch_internal64 = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *))__pyx_f_7pgenlib_10PgenWriter_append_dosages_batch_internal64;
   __pyx_vtable_7pgenlib_PgenWriter.append_dosages_batch = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyArrayObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_append_dosages_batch;
   __pyx_vtable_7pgenlib_PgenWriter.close = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter_close;
   __pyx_vtable_7pgenlib_PgenWriter.__pyx___exit__ = (PyObject *(*)(struct __pyx_obj_7pgenlib_PgenWriter *, PyObject *, PyObject *, PyObject *, int __pyx_skip_dispatch))__pyx_f_7pgenlib_10PgenWriter___exit__;
-  if (PyType_Ready(&__pyx_type_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1412, __pyx_L1_error)
+  if (PyType_Ready(&__pyx_type_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1524, __pyx_L1_error)
   #if PY_VERSION_HEX < 0x030800B1
   __pyx_type_7pgenlib_PgenWriter.tp_print = 0;
   #endif
   if ((CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP) && likely(!__pyx_type_7pgenlib_PgenWriter.tp_dictoffset && __pyx_type_7pgenlib_PgenWriter.tp_getattro == PyObject_GenericGetAttr)) {
     __pyx_type_7pgenlib_PgenWriter.tp_getattro = __Pyx_PyObject_GenericGetAttr;
   }
-  if (__Pyx_SetVtable(__pyx_type_7pgenlib_PgenWriter.tp_dict, __pyx_vtabptr_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1412, __pyx_L1_error)
-  if (PyObject_SetAttr(__pyx_m, __pyx_n_s_PgenWriter, (PyObject *)&__pyx_type_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1412, __pyx_L1_error)
-  if (__Pyx_setup_reduce((PyObject*)&__pyx_type_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1412, __pyx_L1_error)
+  if (__Pyx_SetVtable(__pyx_type_7pgenlib_PgenWriter.tp_dict, __pyx_vtabptr_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1524, __pyx_L1_error)
+  if (PyObject_SetAttr(__pyx_m, __pyx_n_s_PgenWriter, (PyObject *)&__pyx_type_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1524, __pyx_L1_error)
+  if (__Pyx_setup_reduce((PyObject*)&__pyx_type_7pgenlib_PgenWriter) < 0) __PYX_ERR(0, 1524, __pyx_L1_error)
   __pyx_ptype_7pgenlib_PgenWriter = &__pyx_type_7pgenlib_PgenWriter;
   __Pyx_RefNannyFinishContext();
   return 0;
   __pyx_L1_error:;
   __Pyx_RefNannyFinishContext();
   return -1;
 }
@@ -31174,52 +34600,67 @@
   int __pyx_clineno = 0;
   __Pyx_RefNannySetupContext("__Pyx_modinit_type_import_code", 0);
   /*--- Type import code ---*/
   __pyx_t_1 = PyImport_ImportModule(__Pyx_BUILTIN_MODULE_NAME); if (unlikely(!__pyx_t_1)) __PYX_ERR(3, 9, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_ptype_7cpython_4type_type = __Pyx_ImportType(__pyx_t_1, __Pyx_BUILTIN_MODULE_NAME, "type", 
   #if defined(PYPY_VERSION_NUM) && PYPY_VERSION_NUM < 0x050B0000
-  sizeof(PyTypeObject),
+  sizeof(PyTypeObject), __PYX_GET_STRUCT_ALIGNMENT(PyTypeObject),
   #else
-  sizeof(PyHeapTypeObject),
+  sizeof(PyHeapTypeObject), __PYX_GET_STRUCT_ALIGNMENT(PyHeapTypeObject),
   #endif
   __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_7cpython_4type_type) __PYX_ERR(3, 9, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __pyx_t_1 = PyImport_ImportModule("numpy"); if (unlikely(!__pyx_t_1)) __PYX_ERR(2, 200, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  __pyx_ptype_5numpy_dtype = __Pyx_ImportType(__pyx_t_1, "numpy", "dtype", sizeof(PyArray_Descr), __Pyx_ImportType_CheckSize_Ignore);
+  __pyx_ptype_5numpy_dtype = __Pyx_ImportType(__pyx_t_1, "numpy", "dtype", sizeof(PyArray_Descr), __PYX_GET_STRUCT_ALIGNMENT(PyArray_Descr),
+  __Pyx_ImportType_CheckSize_Ignore);
    if (!__pyx_ptype_5numpy_dtype) __PYX_ERR(2, 200, __pyx_L1_error)
-  __pyx_ptype_5numpy_flatiter = __Pyx_ImportType(__pyx_t_1, "numpy", "flatiter", sizeof(PyArrayIterObject), __Pyx_ImportType_CheckSize_Ignore);
+  __pyx_ptype_5numpy_flatiter = __Pyx_ImportType(__pyx_t_1, "numpy", "flatiter", sizeof(PyArrayIterObject), __PYX_GET_STRUCT_ALIGNMENT(PyArrayIterObject),
+  __Pyx_ImportType_CheckSize_Ignore);
    if (!__pyx_ptype_5numpy_flatiter) __PYX_ERR(2, 223, __pyx_L1_error)
-  __pyx_ptype_5numpy_broadcast = __Pyx_ImportType(__pyx_t_1, "numpy", "broadcast", sizeof(PyArrayMultiIterObject), __Pyx_ImportType_CheckSize_Ignore);
+  __pyx_ptype_5numpy_broadcast = __Pyx_ImportType(__pyx_t_1, "numpy", "broadcast", sizeof(PyArrayMultiIterObject), __PYX_GET_STRUCT_ALIGNMENT(PyArrayMultiIterObject),
+  __Pyx_ImportType_CheckSize_Ignore);
    if (!__pyx_ptype_5numpy_broadcast) __PYX_ERR(2, 227, __pyx_L1_error)
-  __pyx_ptype_5numpy_ndarray = __Pyx_ImportType(__pyx_t_1, "numpy", "ndarray", sizeof(PyArrayObject), __Pyx_ImportType_CheckSize_Ignore);
+  __pyx_ptype_5numpy_ndarray = __Pyx_ImportType(__pyx_t_1, "numpy", "ndarray", sizeof(PyArrayObject), __PYX_GET_STRUCT_ALIGNMENT(PyArrayObject),
+  __Pyx_ImportType_CheckSize_Ignore);
    if (!__pyx_ptype_5numpy_ndarray) __PYX_ERR(2, 239, __pyx_L1_error)
-  __pyx_ptype_5numpy_generic = __Pyx_ImportType(__pyx_t_1, "numpy", "generic", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_generic = __Pyx_ImportType(__pyx_t_1, "numpy", "generic", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_generic) __PYX_ERR(2, 771, __pyx_L1_error)
-  __pyx_ptype_5numpy_number = __Pyx_ImportType(__pyx_t_1, "numpy", "number", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_number = __Pyx_ImportType(__pyx_t_1, "numpy", "number", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_number) __PYX_ERR(2, 773, __pyx_L1_error)
-  __pyx_ptype_5numpy_integer = __Pyx_ImportType(__pyx_t_1, "numpy", "integer", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_integer = __Pyx_ImportType(__pyx_t_1, "numpy", "integer", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_integer) __PYX_ERR(2, 775, __pyx_L1_error)
-  __pyx_ptype_5numpy_signedinteger = __Pyx_ImportType(__pyx_t_1, "numpy", "signedinteger", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_signedinteger = __Pyx_ImportType(__pyx_t_1, "numpy", "signedinteger", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_signedinteger) __PYX_ERR(2, 777, __pyx_L1_error)
-  __pyx_ptype_5numpy_unsignedinteger = __Pyx_ImportType(__pyx_t_1, "numpy", "unsignedinteger", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_unsignedinteger = __Pyx_ImportType(__pyx_t_1, "numpy", "unsignedinteger", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_unsignedinteger) __PYX_ERR(2, 779, __pyx_L1_error)
-  __pyx_ptype_5numpy_inexact = __Pyx_ImportType(__pyx_t_1, "numpy", "inexact", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_inexact = __Pyx_ImportType(__pyx_t_1, "numpy", "inexact", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_inexact) __PYX_ERR(2, 781, __pyx_L1_error)
-  __pyx_ptype_5numpy_floating = __Pyx_ImportType(__pyx_t_1, "numpy", "floating", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_floating = __Pyx_ImportType(__pyx_t_1, "numpy", "floating", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_floating) __PYX_ERR(2, 783, __pyx_L1_error)
-  __pyx_ptype_5numpy_complexfloating = __Pyx_ImportType(__pyx_t_1, "numpy", "complexfloating", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_complexfloating = __Pyx_ImportType(__pyx_t_1, "numpy", "complexfloating", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_complexfloating) __PYX_ERR(2, 785, __pyx_L1_error)
-  __pyx_ptype_5numpy_flexible = __Pyx_ImportType(__pyx_t_1, "numpy", "flexible", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_flexible = __Pyx_ImportType(__pyx_t_1, "numpy", "flexible", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_flexible) __PYX_ERR(2, 787, __pyx_L1_error)
-  __pyx_ptype_5numpy_character = __Pyx_ImportType(__pyx_t_1, "numpy", "character", sizeof(PyObject), __Pyx_ImportType_CheckSize_Warn);
+  __pyx_ptype_5numpy_character = __Pyx_ImportType(__pyx_t_1, "numpy", "character", sizeof(PyObject), __PYX_GET_STRUCT_ALIGNMENT(PyObject),
+  __Pyx_ImportType_CheckSize_Warn);
    if (!__pyx_ptype_5numpy_character) __PYX_ERR(2, 789, __pyx_L1_error)
-  __pyx_ptype_5numpy_ufunc = __Pyx_ImportType(__pyx_t_1, "numpy", "ufunc", sizeof(PyUFuncObject), __Pyx_ImportType_CheckSize_Ignore);
+  __pyx_ptype_5numpy_ufunc = __Pyx_ImportType(__pyx_t_1, "numpy", "ufunc", sizeof(PyUFuncObject), __PYX_GET_STRUCT_ALIGNMENT(PyUFuncObject),
+  __Pyx_ImportType_CheckSize_Ignore);
    if (!__pyx_ptype_5numpy_ufunc) __PYX_ERR(2, 827, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __Pyx_RefNannyFinishContext();
   return 0;
   __pyx_L1_error:;
   __Pyx_XDECREF(__pyx_t_1);
   __Pyx_RefNannyFinishContext();
@@ -31438,49 +34879,49 @@
   (void)__Pyx_modinit_variable_import_code();
   (void)__Pyx_modinit_function_import_code();
   /*--- Execution code ---*/
   #if defined(__Pyx_Generator_USED) || defined(__Pyx_Coroutine_USED)
   if (__Pyx_patch_abc() < 0) __PYX_ERR(0, 1, __pyx_L1_error)
   #endif
 
-  /* "pgenlib.pyx":6
+  /* "pgenlib.pyx":7
  * from cpython.mem cimport PyMem_Malloc, PyMem_Free
  * # from cpython.view cimport array as cvarray
  * import numpy as np             # <<<<<<<<<<<<<<
  * cimport numpy as np
  * import sys
  */
-  __pyx_t_1 = __Pyx_Import(__pyx_n_s_numpy, 0, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 6, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_Import(__pyx_n_s_numpy, 0, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 7, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  if (PyDict_SetItem(__pyx_d, __pyx_n_s_np, __pyx_t_1) < 0) __PYX_ERR(0, 6, __pyx_L1_error)
+  if (PyDict_SetItem(__pyx_d, __pyx_n_s_np, __pyx_t_1) < 0) __PYX_ERR(0, 7, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-  /* "pgenlib.pyx":8
+  /* "pgenlib.pyx":9
  * import numpy as np
  * cimport numpy as np
  * import sys             # <<<<<<<<<<<<<<
  * 
- * cdef extern from "../plink2/pgenlib_ffi_support.h" namespace "plink2":
+ * cdef extern from "../plink2/include/pgenlib_misc.h" namespace "plink2":
  */
-  __pyx_t_1 = __Pyx_Import(__pyx_n_s_sys, 0, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 8, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_Import(__pyx_n_s_sys, 0, 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 9, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
-  if (PyDict_SetItem(__pyx_d, __pyx_n_s_sys, __pyx_t_1) < 0) __PYX_ERR(0, 8, __pyx_L1_error)
+  if (PyDict_SetItem(__pyx_d, __pyx_n_s_sys, __pyx_t_1) < 0) __PYX_ERR(0, 9, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
   /* "pgenlib.pyx":1
  * # cython: language_level=3             # <<<<<<<<<<<<<<
  * # from libc.stdlib cimport malloc, free
  * from libc.stdint cimport int64_t, uintptr_t, uint32_t, int32_t, uint16_t, uint8_t, int8_t
  */
   __pyx_t_1 = __Pyx_PyDict_NewPresized(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_1) < 0) __PYX_ERR(0, 1, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-op5cmpx1/lib/python3.10/site-packages/numpy/__init__.pxd":1014
+  /* "../../../../private/var/folders/hw/lw9w89gj5jncz6_gcyh0kj1m0000gn/T/build-env-h4atrljl/lib/python3.10/site-packages/numpy/__init__.pxd":1014
  * 
  * 
  * cdef inline NPY_DATETIMEUNIT get_datetime64_unit(object obj) nogil:             # <<<<<<<<<<<<<<
  *     """
  *     returns the unit part of the dtype for a numpy datetime64 object.
  */
 
@@ -32110,33 +35551,19 @@
   if (buf->suboffsets == NULL) buf->suboffsets = __Pyx_minusones;
   return 0;
 fail:;
   __Pyx_SafeReleaseBuffer(buf);
   return -1;
 }
 
-/* PyObjectCall */
-  #if CYTHON_COMPILING_IN_CPYTHON
-static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw) {
-    PyObject *result;
-    ternaryfunc call = Py_TYPE(func)->tp_call;
-    if (unlikely(!call))
-        return PyObject_Call(func, arg, kw);
-    if (unlikely(Py_EnterRecursiveCall((char*)" while calling a Python object")))
-        return NULL;
-    result = (*call)(func, arg, kw);
-    Py_LeaveRecursiveCall();
-    if (unlikely(!result) && unlikely(!PyErr_Occurred())) {
-        PyErr_SetString(
-            PyExc_SystemError,
-            "NULL result without error in PyObject_Call");
-    }
-    return result;
+/* BufferIndexError */
+  static void __Pyx_RaiseBufferIndexError(int axis) {
+  PyErr_Format(PyExc_IndexError,
+     "Out of bounds on buffer access (axis %d)", axis);
 }
-#endif
 
 /* PyErrFetchRestore */
   #if CYTHON_FAST_THREAD_STATE
 static CYTHON_INLINE void __Pyx_ErrRestoreInState(PyThreadState *tstate, PyObject *type, PyObject *value, PyObject *tb) {
     PyObject *tmp_type, *tmp_value, *tmp_tb;
     tmp_type = tstate->curexc_type;
     tmp_value = tstate->curexc_value;
@@ -32154,14 +35581,34 @@
     *tb = tstate->curexc_traceback;
     tstate->curexc_type = 0;
     tstate->curexc_value = 0;
     tstate->curexc_traceback = 0;
 }
 #endif
 
+/* PyObjectCall */
+  #if CYTHON_COMPILING_IN_CPYTHON
+static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw) {
+    PyObject *result;
+    ternaryfunc call = Py_TYPE(func)->tp_call;
+    if (unlikely(!call))
+        return PyObject_Call(func, arg, kw);
+    if (unlikely(Py_EnterRecursiveCall((char*)" while calling a Python object")))
+        return NULL;
+    result = (*call)(func, arg, kw);
+    Py_LeaveRecursiveCall();
+    if (unlikely(!result) && unlikely(!PyErr_Occurred())) {
+        PyErr_SetString(
+            PyExc_SystemError,
+            "NULL result without error in PyObject_Call");
+    }
+    return result;
+}
+#endif
+
 /* RaiseException */
   #if PY_MAJOR_VERSION < 3
 static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb,
                         CYTHON_UNUSED PyObject *cause) {
     __Pyx_PyThreadState_declare
     Py_XINCREF(type);
     if (!value || value == Py_None)
@@ -32291,42 +35738,36 @@
                             "BaseException");
             goto bad;
         }
         PyException_SetCause(value, fixed_cause);
     }
     PyErr_SetObject(type, value);
     if (tb) {
-#if CYTHON_COMPILING_IN_PYPY
-        PyObject *tmp_type, *tmp_value, *tmp_tb;
-        PyErr_Fetch(&tmp_type, &tmp_value, &tmp_tb);
-        Py_INCREF(tb);
-        PyErr_Restore(tmp_type, tmp_value, tb);
-        Py_XDECREF(tmp_tb);
-#else
+#if CYTHON_FAST_THREAD_STATE
         PyThreadState *tstate = __Pyx_PyThreadState_Current;
         PyObject* tmp_tb = tstate->curexc_traceback;
         if (tb != tmp_tb) {
             Py_INCREF(tb);
             tstate->curexc_traceback = tb;
             Py_XDECREF(tmp_tb);
         }
+#else
+        PyObject *tmp_type, *tmp_value, *tmp_tb;
+        PyErr_Fetch(&tmp_type, &tmp_value, &tmp_tb);
+        Py_INCREF(tb);
+        PyErr_Restore(tmp_type, tmp_value, tb);
+        Py_XDECREF(tmp_tb);
 #endif
     }
 bad:
     Py_XDECREF(owned_instance);
     return;
 }
 #endif
 
-/* BufferIndexError */
-  static void __Pyx_RaiseBufferIndexError(int axis) {
-  PyErr_Format(PyExc_IndexError,
-     "Out of bounds on buffer access (axis %d)", axis);
-}
-
 /* PyCFunctionFastCall */
   #if CYTHON_FAST_PYCCALL
 static CYTHON_INLINE PyObject * __Pyx_PyCFunction_FastCall(PyObject *func_obj, PyObject **args, Py_ssize_t nargs) {
     PyCFunctionObject *func = (PyCFunctionObject*)func_obj;
     PyCFunction meth = PyCFunction_GET_FUNCTION(func);
     PyObject *self = PyCFunction_GET_SELF(func);
     int flags = PyCFunction_GET_FLAGS(func);
@@ -32684,35 +36125,35 @@
     }
     PyErr_Format(PyExc_TypeError,
         "Argument '%.200s' has incorrect type (expected %.200s, got %.200s)",
         name, type->tp_name, Py_TYPE(obj)->tp_name);
     return 0;
 }
 
-/* DivInt[long] */
-  static CYTHON_INLINE long __Pyx_div_long(long a, long b) {
-    long q = a / b;
-    long r = a - q*b;
-    q -= ((r != 0) & ((r ^ b) < 0));
-    return q;
-}
-
 /* ExtTypeTest */
   static CYTHON_INLINE int __Pyx_TypeTest(PyObject *obj, PyTypeObject *type) {
     if (unlikely(!type)) {
         PyErr_SetString(PyExc_SystemError, "Missing type object");
         return 0;
     }
     if (likely(__Pyx_TypeCheck(obj, type)))
         return 1;
     PyErr_Format(PyExc_TypeError, "Cannot convert %.200s to %.200s",
                  Py_TYPE(obj)->tp_name, type->tp_name);
     return 0;
 }
 
+/* DivInt[long] */
+  static CYTHON_INLINE long __Pyx_div_long(long a, long b) {
+    long q = a / b;
+    long r = a - q*b;
+    q -= ((r != 0) & ((r ^ b) < 0));
+    return q;
+}
+
 /* PyDictVersioning */
   #if CYTHON_USE_DICT_VERSIONS && CYTHON_USE_TYPE_SLOTS
 static CYTHON_INLINE PY_UINT64_T __Pyx_get_tp_dict_version(PyObject *obj) {
     PyObject *dict = Py_TYPE(obj)->tp_dict;
     return likely(dict) ? __PYX_GET_DICT_VERSION(dict) : 0;
 }
 static CYTHON_INLINE PY_UINT64_T __Pyx_get_object_dict_version(PyObject *obj) {
@@ -32795,14 +36236,138 @@
 /* ModInt[long] */
   static CYTHON_INLINE long __Pyx_mod_long(long a, long b) {
     long r = a % b;
     r += ((r != 0) & ((r ^ b) < 0)) * b;
     return r;
 }
 
+/* PyIntBinop */
+  #if !CYTHON_COMPILING_IN_PYPY
+static PyObject* __Pyx_PyInt_AddObjC(PyObject *op1, PyObject *op2, CYTHON_UNUSED long intval, int inplace, int zerodivision_check) {
+    (void)inplace;
+    (void)zerodivision_check;
+    #if PY_MAJOR_VERSION < 3
+    if (likely(PyInt_CheckExact(op1))) {
+        const long b = intval;
+        long x;
+        long a = PyInt_AS_LONG(op1);
+            x = (long)((unsigned long)a + b);
+            if (likely((x^a) >= 0 || (x^b) >= 0))
+                return PyInt_FromLong(x);
+            return PyLong_Type.tp_as_number->nb_add(op1, op2);
+    }
+    #endif
+    #if CYTHON_USE_PYLONG_INTERNALS
+    if (likely(PyLong_CheckExact(op1))) {
+        const long b = intval;
+        long a, x;
+#ifdef HAVE_LONG_LONG
+        const PY_LONG_LONG llb = intval;
+        PY_LONG_LONG lla, llx;
+#endif
+        const digit* digits = ((PyLongObject*)op1)->ob_digit;
+        const Py_ssize_t size = Py_SIZE(op1);
+        if (likely(__Pyx_sst_abs(size) <= 1)) {
+            a = likely(size) ? digits[0] : 0;
+            if (size == -1) a = -a;
+        } else {
+            switch (size) {
+                case -2:
+                    if (8 * sizeof(long) - 1 > 2 * PyLong_SHIFT) {
+                        a = -(long) (((((unsigned long)digits[1]) << PyLong_SHIFT) | (unsigned long)digits[0]));
+                        break;
+#ifdef HAVE_LONG_LONG
+                    } else if (8 * sizeof(PY_LONG_LONG) - 1 > 2 * PyLong_SHIFT) {
+                        lla = -(PY_LONG_LONG) (((((unsigned PY_LONG_LONG)digits[1]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[0]));
+                        goto long_long;
+#endif
+                    }
+                    CYTHON_FALLTHROUGH;
+                case 2:
+                    if (8 * sizeof(long) - 1 > 2 * PyLong_SHIFT) {
+                        a = (long) (((((unsigned long)digits[1]) << PyLong_SHIFT) | (unsigned long)digits[0]));
+                        break;
+#ifdef HAVE_LONG_LONG
+                    } else if (8 * sizeof(PY_LONG_LONG) - 1 > 2 * PyLong_SHIFT) {
+                        lla = (PY_LONG_LONG) (((((unsigned PY_LONG_LONG)digits[1]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[0]));
+                        goto long_long;
+#endif
+                    }
+                    CYTHON_FALLTHROUGH;
+                case -3:
+                    if (8 * sizeof(long) - 1 > 3 * PyLong_SHIFT) {
+                        a = -(long) (((((((unsigned long)digits[2]) << PyLong_SHIFT) | (unsigned long)digits[1]) << PyLong_SHIFT) | (unsigned long)digits[0]));
+                        break;
+#ifdef HAVE_LONG_LONG
+                    } else if (8 * sizeof(PY_LONG_LONG) - 1 > 3 * PyLong_SHIFT) {
+                        lla = -(PY_LONG_LONG) (((((((unsigned PY_LONG_LONG)digits[2]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[1]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[0]));
+                        goto long_long;
+#endif
+                    }
+                    CYTHON_FALLTHROUGH;
+                case 3:
+                    if (8 * sizeof(long) - 1 > 3 * PyLong_SHIFT) {
+                        a = (long) (((((((unsigned long)digits[2]) << PyLong_SHIFT) | (unsigned long)digits[1]) << PyLong_SHIFT) | (unsigned long)digits[0]));
+                        break;
+#ifdef HAVE_LONG_LONG
+                    } else if (8 * sizeof(PY_LONG_LONG) - 1 > 3 * PyLong_SHIFT) {
+                        lla = (PY_LONG_LONG) (((((((unsigned PY_LONG_LONG)digits[2]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[1]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[0]));
+                        goto long_long;
+#endif
+                    }
+                    CYTHON_FALLTHROUGH;
+                case -4:
+                    if (8 * sizeof(long) - 1 > 4 * PyLong_SHIFT) {
+                        a = -(long) (((((((((unsigned long)digits[3]) << PyLong_SHIFT) | (unsigned long)digits[2]) << PyLong_SHIFT) | (unsigned long)digits[1]) << PyLong_SHIFT) | (unsigned long)digits[0]));
+                        break;
+#ifdef HAVE_LONG_LONG
+                    } else if (8 * sizeof(PY_LONG_LONG) - 1 > 4 * PyLong_SHIFT) {
+                        lla = -(PY_LONG_LONG) (((((((((unsigned PY_LONG_LONG)digits[3]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[2]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[1]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[0]));
+                        goto long_long;
+#endif
+                    }
+                    CYTHON_FALLTHROUGH;
+                case 4:
+                    if (8 * sizeof(long) - 1 > 4 * PyLong_SHIFT) {
+                        a = (long) (((((((((unsigned long)digits[3]) << PyLong_SHIFT) | (unsigned long)digits[2]) << PyLong_SHIFT) | (unsigned long)digits[1]) << PyLong_SHIFT) | (unsigned long)digits[0]));
+                        break;
+#ifdef HAVE_LONG_LONG
+                    } else if (8 * sizeof(PY_LONG_LONG) - 1 > 4 * PyLong_SHIFT) {
+                        lla = (PY_LONG_LONG) (((((((((unsigned PY_LONG_LONG)digits[3]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[2]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[1]) << PyLong_SHIFT) | (unsigned PY_LONG_LONG)digits[0]));
+                        goto long_long;
+#endif
+                    }
+                    CYTHON_FALLTHROUGH;
+                default: return PyLong_Type.tp_as_number->nb_add(op1, op2);
+            }
+        }
+                x = a + b;
+            return PyLong_FromLong(x);
+#ifdef HAVE_LONG_LONG
+        long_long:
+                llx = lla + llb;
+            return PyLong_FromLongLong(llx);
+#endif
+        
+        
+    }
+    #endif
+    if (PyFloat_CheckExact(op1)) {
+        const long b = intval;
+        double a = PyFloat_AS_DOUBLE(op1);
+            double result;
+            PyFPE_START_PROTECT("add", return NULL)
+            result = ((double)a) + (double)b;
+            PyFPE_END_PROTECT(result)
+            return PyFloat_FromDouble(result);
+    }
+    return (inplace ? PyNumber_InPlaceAdd : PyNumber_Add)(op1, op2);
+}
+#endif
+
 /* PyIntCompare */
   static CYTHON_INLINE PyObject* __Pyx_PyInt_NeObjC(PyObject *op1, PyObject *op2, CYTHON_UNUSED long intval, CYTHON_UNUSED long inplace) {
     if (op1 == op2) {
         Py_RETURN_FALSE;
     }
     #if PY_MAJOR_VERSION < 3
     if (likely(PyInt_CheckExact(op1))) {
@@ -32978,42 +36543,53 @@
     if (is_list || PySequence_Check(o)) {
         return PySequence_GetItem(o, i);
     }
 #endif
     return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
 }
 
-/* ObjectGetItem */
-  #if CYTHON_USE_TYPE_SLOTS
-static PyObject *__Pyx_PyObject_GetIndex(PyObject *obj, PyObject* index) {
-    PyObject *runerr = NULL;
-    Py_ssize_t key_value;
-    PySequenceMethods *m = Py_TYPE(obj)->tp_as_sequence;
-    if (unlikely(!(m && m->sq_item))) {
-        PyErr_Format(PyExc_TypeError, "'%.200s' object is not subscriptable", Py_TYPE(obj)->tp_name);
-        return NULL;
-    }
-    key_value = __Pyx_PyIndex_AsSsize_t(index);
-    if (likely(key_value != -1 || !(runerr = PyErr_Occurred()))) {
-        return __Pyx_GetItemInt_Fast(obj, key_value, 0, 1, 1);
-    }
-    if (PyErr_GivenExceptionMatches(runerr, PyExc_OverflowError)) {
-        PyErr_Clear();
-        PyErr_Format(PyExc_IndexError, "cannot fit '%.200s' into an index-sized integer", Py_TYPE(index)->tp_name);
+/* WriteUnraisableException */
+  static void __Pyx_WriteUnraisable(const char *name, CYTHON_UNUSED int clineno,
+                                  CYTHON_UNUSED int lineno, CYTHON_UNUSED const char *filename,
+                                  int full_traceback, CYTHON_UNUSED int nogil) {
+    PyObject *old_exc, *old_val, *old_tb;
+    PyObject *ctx;
+    __Pyx_PyThreadState_declare
+#ifdef WITH_THREAD
+    PyGILState_STATE state;
+    if (nogil)
+        state = PyGILState_Ensure();
+    else state = (PyGILState_STATE)0;
+#endif
+    __Pyx_PyThreadState_assign
+    __Pyx_ErrFetch(&old_exc, &old_val, &old_tb);
+    if (full_traceback) {
+        Py_XINCREF(old_exc);
+        Py_XINCREF(old_val);
+        Py_XINCREF(old_tb);
+        __Pyx_ErrRestore(old_exc, old_val, old_tb);
+        PyErr_PrintEx(1);
     }
-    return NULL;
-}
-static PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* key) {
-    PyMappingMethods *m = Py_TYPE(obj)->tp_as_mapping;
-    if (likely(m && m->mp_subscript)) {
-        return m->mp_subscript(obj, key);
+    #if PY_MAJOR_VERSION < 3
+    ctx = PyString_FromString(name);
+    #else
+    ctx = PyUnicode_FromString(name);
+    #endif
+    __Pyx_ErrRestore(old_exc, old_val, old_tb);
+    if (!ctx) {
+        PyErr_WriteUnraisable(Py_None);
+    } else {
+        PyErr_WriteUnraisable(ctx);
+        Py_DECREF(ctx);
     }
-    return __Pyx_PyObject_GetIndex(obj, key);
-}
+#ifdef WITH_THREAD
+    if (nogil)
+        PyGILState_Release(state);
 #endif
+}
 
 /* GetTopmostException */
   #if CYTHON_USE_EXC_INFO_STACK
 static _PyErr_StackItem *
 __Pyx_PyErr_GetTopmostException(PyThreadState *tstate)
 {
     _PyErr_StackItem *exc_info = tstate->exc_info;
@@ -33360,44 +36936,62 @@
     return ret;
 }
 
 /* TypeImport */
   #ifndef __PYX_HAVE_RT_ImportType
 #define __PYX_HAVE_RT_ImportType
 static PyTypeObject *__Pyx_ImportType(PyObject *module, const char *module_name, const char *class_name,
-    size_t size, enum __Pyx_ImportType_CheckSize check_size)
+    size_t size, size_t alignment, enum __Pyx_ImportType_CheckSize check_size)
 {
     PyObject *result = 0;
     char warning[200];
     Py_ssize_t basicsize;
+    Py_ssize_t itemsize;
 #ifdef Py_LIMITED_API
     PyObject *py_basicsize;
+    PyObject *py_itemsize;
 #endif
     result = PyObject_GetAttrString(module, class_name);
     if (!result)
         goto bad;
     if (!PyType_Check(result)) {
         PyErr_Format(PyExc_TypeError,
             "%.200s.%.200s is not a type object",
             module_name, class_name);
         goto bad;
     }
 #ifndef Py_LIMITED_API
     basicsize = ((PyTypeObject *)result)->tp_basicsize;
+    itemsize = ((PyTypeObject *)result)->tp_itemsize;
 #else
     py_basicsize = PyObject_GetAttrString(result, "__basicsize__");
     if (!py_basicsize)
         goto bad;
     basicsize = PyLong_AsSsize_t(py_basicsize);
     Py_DECREF(py_basicsize);
     py_basicsize = 0;
     if (basicsize == (Py_ssize_t)-1 && PyErr_Occurred())
         goto bad;
+    py_itemsize = PyObject_GetAttrString(result, "__itemsize__");
+    if (!py_itemsize)
+        goto bad;
+    itemsize = PyLong_AsSsize_t(py_itemsize);
+    Py_DECREF(py_itemsize);
+    py_itemsize = 0;
+    if (itemsize == (Py_ssize_t)-1 && PyErr_Occurred())
+        goto bad;
 #endif
-    if ((size_t)basicsize < size) {
+    if (itemsize) {
+        if (size % alignment) {
+            alignment = size % alignment;
+        }
+        if (itemsize < (Py_ssize_t)alignment)
+            itemsize = (Py_ssize_t)alignment;
+    }
+    if ((size_t)(basicsize + itemsize) < size) {
         PyErr_Format(PyExc_ValueError,
             "%.200s.%.200s size changed, may indicate binary incompatibility. "
             "Expected %zd from C header, got %zd from PyObject",
             module_name, class_name, size, basicsize);
         goto bad;
     }
     if (check_size == __Pyx_ImportType_CheckSize_Error && (size_t)basicsize != size) {
@@ -34447,52 +38041,14 @@
         int one = 1; int little = (int)*(unsigned char *)&one;
         unsigned char *bytes = (unsigned char *)&value;
         return _PyLong_FromByteArray(bytes, sizeof(Py_intptr_t),
                                      little, !is_unsigned);
     }
 }
 
-/* CIntToPy */
-  static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value) {
-#ifdef __Pyx_HAS_GCC_DIAGNOSTIC
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wconversion"
-#endif
-    const long neg_one = (long) -1, const_zero = (long) 0;
-#ifdef __Pyx_HAS_GCC_DIAGNOSTIC
-#pragma GCC diagnostic pop
-#endif
-    const int is_unsigned = neg_one > const_zero;
-    if (is_unsigned) {
-        if (sizeof(long) < sizeof(long)) {
-            return PyInt_FromLong((long) value);
-        } else if (sizeof(long) <= sizeof(unsigned long)) {
-            return PyLong_FromUnsignedLong((unsigned long) value);
-#ifdef HAVE_LONG_LONG
-        } else if (sizeof(long) <= sizeof(unsigned PY_LONG_LONG)) {
-            return PyLong_FromUnsignedLongLong((unsigned PY_LONG_LONG) value);
-#endif
-        }
-    } else {
-        if (sizeof(long) <= sizeof(long)) {
-            return PyInt_FromLong((long) value);
-#ifdef HAVE_LONG_LONG
-        } else if (sizeof(long) <= sizeof(PY_LONG_LONG)) {
-            return PyLong_FromLongLong((PY_LONG_LONG) value);
-#endif
-        }
-    }
-    {
-        int one = 1; int little = (int)*(unsigned char *)&one;
-        unsigned char *bytes = (unsigned char *)&value;
-        return _PyLong_FromByteArray(bytes, sizeof(long),
-                                     little, !is_unsigned);
-    }
-}
-
 /* CIntFromPy */
   static CYTHON_INLINE size_t __Pyx_PyInt_As_size_t(PyObject *x) {
 #ifdef __Pyx_HAS_GCC_DIAGNOSTIC
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wconversion"
 #endif
     const size_t neg_one = (size_t) -1, const_zero = (size_t) 0;
@@ -34681,14 +38237,52 @@
     return (size_t) -1;
 raise_neg_overflow:
     PyErr_SetString(PyExc_OverflowError,
         "can't convert negative value to size_t");
     return (size_t) -1;
 }
 
+/* CIntToPy */
+  static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value) {
+#ifdef __Pyx_HAS_GCC_DIAGNOSTIC
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wconversion"
+#endif
+    const long neg_one = (long) -1, const_zero = (long) 0;
+#ifdef __Pyx_HAS_GCC_DIAGNOSTIC
+#pragma GCC diagnostic pop
+#endif
+    const int is_unsigned = neg_one > const_zero;
+    if (is_unsigned) {
+        if (sizeof(long) < sizeof(long)) {
+            return PyInt_FromLong((long) value);
+        } else if (sizeof(long) <= sizeof(unsigned long)) {
+            return PyLong_FromUnsignedLong((unsigned long) value);
+#ifdef HAVE_LONG_LONG
+        } else if (sizeof(long) <= sizeof(unsigned PY_LONG_LONG)) {
+            return PyLong_FromUnsignedLongLong((unsigned PY_LONG_LONG) value);
+#endif
+        }
+    } else {
+        if (sizeof(long) <= sizeof(long)) {
+            return PyInt_FromLong((long) value);
+#ifdef HAVE_LONG_LONG
+        } else if (sizeof(long) <= sizeof(PY_LONG_LONG)) {
+            return PyLong_FromLongLong((PY_LONG_LONG) value);
+#endif
+        }
+    }
+    {
+        int one = 1; int little = (int)*(unsigned char *)&one;
+        unsigned char *bytes = (unsigned char *)&value;
+        return _PyLong_FromByteArray(bytes, sizeof(long),
+                                     little, !is_unsigned);
+    }
+}
+
 /* CIntFromPy */
   static CYTHON_INLINE long __Pyx_PyInt_As_long(PyObject *x) {
 #ifdef __Pyx_HAS_GCC_DIAGNOSTIC
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wconversion"
 #endif
     const long neg_one = (long) -1, const_zero = (long) 0;
```

### Comparing `Pgenlib-0.82.1/src/pgenlib/pgenlib.pyx` & `Pgenlib-0.83.0/src/pgenlib/pgenlib.pyx`

 * *Files 16% similar despite different names*

```diff
@@ -1,17 +1,18 @@
 # cython: language_level=3
 # from libc.stdlib cimport malloc, free
 from libc.stdint cimport int64_t, uintptr_t, uint32_t, int32_t, uint16_t, uint8_t, int8_t
+from libc.string cimport memcpy
 from cpython.mem cimport PyMem_Malloc, PyMem_Free
 # from cpython.view cimport array as cvarray
 import numpy as np
 cimport numpy as np
 import sys
 
-cdef extern from "../plink2/pgenlib_ffi_support.h" namespace "plink2":
+cdef extern from "../plink2/include/pgenlib_misc.h" namespace "plink2":
     ctypedef uint32_t BoolErr
     ctypedef enum PglErr:
         kPglRetSuccess
         kPglRetSkipped
         kPglRetNomem
         kPglRetOpenFail
         kPglRetReadFail
@@ -24,43 +25,14 @@
         kPglRetNetworkFail
         kPglRetSampleMajorBed
         kPglRetImproperFunctionCall
         kPglRetNotYetSupported
         kPglRetLongLine
         kPglRetEmptyFile
 
-    BoolErr cachealigned_malloc(uintptr_t size, void* aligned_pp)
-    void aligned_free(void* aligned_ptr)
-
-    uintptr_t DivUp(uintptr_t val, uintptr_t divisor)
-    void ZeroWArr(uintptr_t entry_ct, uintptr_t* ularr)
-    # void BitvecAnd(const uintptr_t* arg_bitvec, uintptr_t word_ct, uintptr_t* main_bitvec)
-    void FillInterleavedMaskVec(const uintptr_t* subset_mask, uint32_t base_vec_ct, uintptr_t* interleaved_mask_vec)
-    void FillCumulativePopcounts(const uintptr_t* subset_mask, uint32_t word_ct, uint32_t* cumulative_popcounts)
-
-    ctypedef uintptr_t VecW
-    void TransposeNypblock(const uintptr_t* read_iter, uint32_t read_ul_stride, uint32_t write_ul_stride, uint32_t read_batch_size, uint32_t write_batch_size, uintptr_t* write_iter, VecW* vecaligned_buf)
-    void TransposeBitblock(const uintptr_t* read_iter, uint32_t read_ul_stride, uint32_t write_ul_stride, uint32_t read_batch_size, uint32_t write_batch_size, uintptr_t* write_iter, VecW* vecaligned_buf)
-
-    void GenovecInvertUnsafe(uint32_t sample_ct, uintptr_t* genovec)
-    void BiallelicDosage16Invert(uint32_t dosage_ct, uint16_t* dosage_main)
-
-    void GenoarrToBytesMinus9(const uintptr_t* genoarr, uint32_t sample_ct, int8_t* genobytes)
-    void GenoarrToInt32sMinus9(const uintptr_t* genoarr, uint32_t sample_ct, int32_t* geno_int32)
-    void GenoarrToInt64sMinus9(const uintptr_t* genoarr, uint32_t sample_ct, int64_t* geno_int64)
-    void GenoarrPhasedToAlleleCodesMinus9(const uintptr_t* genoarr, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, uint32_t sample_ct, uint32_t phasepresent_ct, unsigned char* phasebytes, int32_t* allele_codes)
-    void GenoarrPhasedToHapCodes(const uintptr_t* genoarr, const uintptr_t* phaseinfo, uint32_t variant_batch_size, int32_t* hap0_codes_iter, int32_t* hap1_codes_iter)
-    void Dosage16ToFloatsMinus9(const uintptr_t* genoarr, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t sample_ct, uint32_t dosage_ct, float* geno_float)
-    void Dosage16ToDoublesMinus9(const uintptr_t* genoarr, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t sample_ct, uint32_t dosage_ct, double* geno_double)
-    void BytesToBitsUnsafe(const uint8_t* boolbytes, uint32_t sample_ct, uintptr_t* bitarr)
-    void BytesToGenoarrUnsafe(const int8_t* genobytes, uint32_t sample_ct, uintptr_t* genoarr)
-    void AlleleCodesToGenoarrUnsafe(const int32_t* allele_codes, const unsigned char* phasepresent_bytes, uint32_t sample_ct, uintptr_t* genoarr, uintptr_t* phasepresent, uintptr_t* phaseinfo)
-    void FloatsToDosage16(const float* floatarr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr)
-    void DoublesToDosage16(const double* doublearr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr)
-
     cdef enum:
         k1LU
     cdef enum:
         kCacheline
     cdef enum:
         kBitsPerWord
     cdef enum:
@@ -101,24 +73,82 @@
     cdef enum:
         kfPgenGlobalMultiallelicHardcallFound
     cdef enum:
         kfPgenGlobalHardcallPhasePresent
     cdef enum:
         kfPgenGlobalDosagePresent
 
+    BoolErr cachealigned_malloc(uintptr_t size, void* aligned_pp)
+    void aligned_free(void* aligned_ptr)
+
+    uintptr_t RoundUpPow2(uintptr_t val, uintptr_t alignment)
+    uintptr_t DivUp(uintptr_t val, uintptr_t divisor)
+    void ZeroTrailingNyps(uintptr_t nyp_ct, uintptr_t* bitarr)
+    void ZeroWArr(uintptr_t entry_ct, uintptr_t* ularr)
+    # void BitvecAnd(const uintptr_t* arg_bitvec, uintptr_t word_ct, uintptr_t* main_bitvec)
+    void FillInterleavedMaskVec(const uintptr_t* subset_mask, uint32_t base_vec_ct, uintptr_t* interleaved_mask_vec)
+    void FillCumulativePopcounts(const uintptr_t* subset_mask, uint32_t word_ct, uint32_t* cumulative_popcounts)
+
+    ctypedef uintptr_t VecW
+    void TransposeNypblock(const uintptr_t* read_iter, uint32_t read_ul_stride, uint32_t write_ul_stride, uint32_t read_batch_size, uint32_t write_batch_size, uintptr_t* write_iter, VecW* vecaligned_buf)
+    void TransposeBitblock(const uintptr_t* read_iter, uint32_t read_ul_stride, uint32_t write_ul_stride, uint32_t read_batch_size, uint32_t write_batch_size, uintptr_t* write_iter, VecW* vecaligned_buf)
+
+    void GenoarrCountFreqsUnsafe(const uintptr_t* genoarr, uint32_t sample_ct, uint32_t* genocounts)
+    void GenovecInvertUnsafe(uint32_t sample_ct, uintptr_t* genovec)
+    void BiallelicDosage16Invert(uint32_t dosage_ct, uint16_t* dosage_main)
+
     ctypedef unsigned char AlleleCode
 
+    cdef struct PgenVariantStruct:
+        uintptr_t* genovec
+        uintptr_t* patch_01_set
+        AlleleCode* patch_01_vals
+        uintptr_t* patch_10_set
+        AlleleCode* patch_10_vals
+        uintptr_t* phasepresent
+        uintptr_t* phaseinfo
+        uintptr_t* dosage_present
+        uint16_t* dosage_main
+        uint32_t patch_01_ct
+        uint32_t patch_10_ct
+        uint32_t phasepresent_ct
+    cdef enum:
+        kPglMaxAlleleCt
+    cdef enum:
+        kPglMaxVariantCt
+
+    uintptr_t PglComputeMaxAlleleCt(const uintptr_t* allele_idx_offsets, uint32_t variant_ct)
+
+
+cdef extern from "../plink2/pgenlib_ffi_support.h" namespace "plink2":
+
+    void GenoarrToBytesMinus9(const uintptr_t* genoarr, uint32_t sample_ct, int8_t* genobytes)
+    void GenoarrToInt32sMinus9(const uintptr_t* genoarr, uint32_t sample_ct, int32_t* geno_int32)
+    void GenoarrToInt64sMinus9(const uintptr_t* genoarr, uint32_t sample_ct, int64_t* geno_int64)
+    void GenoarrMPToAlleleCodesMinus9(const PgenVariantStruct* pgv, uint32_t sample_ct, unsigned char* phasebytes, int32_t* allele_codes)
+    void GenoarrPhasedToHapCodes(const uintptr_t* genoarr, const uintptr_t* phaseinfo, uint32_t variant_batch_size, int32_t* hap0_codes_iter, int32_t* hap1_codes_iter)
+    void Dosage16ToFloatsMinus9(const uintptr_t* genoarr, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t sample_ct, uint32_t dosage_ct, float* geno_float)
+    void Dosage16ToDoublesMinus9(const uintptr_t* genoarr, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t sample_ct, uint32_t dosage_ct, double* geno_double)
+    void BytesToBitsUnsafe(const uint8_t* boolbytes, uint32_t sample_ct, uintptr_t* bitarr)
+    void BytesToGenoarrUnsafe(const int8_t* genobytes, uint32_t sample_ct, uintptr_t* genoarr)
+
+    int32_t ConvertMultiAlleleCodesUnsafe(const int32_t* allele_codes, const unsigned char* phasepresent_bytes, uint32_t sample_ct, uintptr_t* genoarr, uintptr_t* patch_01_set, AlleleCode* patch_01_vals, uintptr_t* patch_10_set, AlleleCode* patch_10_vals, uint32_t* patch_01_ctp, uint32_t* patch_10_ctp, uintptr_t* phasepresent, uintptr_t* phaseinfo)
+    void FloatsToDosage16(const float* floatarr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr)
+    void DoublesToDosage16(const double* doublearr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr)
+
 
 cdef extern from "../plink2/include/pgenlib_read.h" namespace "plink2":
     cdef cppclass PgenFileInfo:
         uint32_t raw_variant_ct
         uint32_t raw_sample_ct
         unsigned char* vrtypes
+        uintptr_t* allele_idx_offsets
         uintptr_t* nonref_flags
         uint32_t gflags
+        uint32_t max_allele_ct
 
     ctypedef uint32_t PgenHeaderCtrl
 
     void PreinitPgfi(PgenFileInfo* pgfip)
 
     PglErr PgfiInitPhase1(const char* fname, const char* pgi_fname, uint32_t raw_variant_ct, uint32_t raw_sample_ct, PgenHeaderCtrl* header_ctrl_ptr, PgenFileInfo* pgfip, uintptr_t* pgfi_alloc_cacheline_ct_ptr, char* errstr_buf)
 
@@ -140,30 +170,35 @@
 
     void PreinitPgr(PgenReaderStruct* pgr_ptr)
 
     PglErr PgrInit(const char* fname, uint32_t max_vrec_width, PgenFileInfo* pgfip, PgenReaderStruct* pgr_ptr, unsigned char* pgr_alloc)
 
     PglErr PgrGet1(const uintptr_t* sample_include, PgrSampleSubsetIndexStruct pssi, uint32_t sample_ct, uint32_t vidx, uint32_t allele_idx, PgenReaderStruct* pgr_ptr, uintptr_t* allele_countvec)
 
+    PglErr PgrGetM(const uintptr_t* sample_include, PgrSampleSubsetIndexStruct pssi, uint32_t sample_ct, uint32_t vidx, PgenReaderStruct* pgr_ptr, PgenVariantStruct* pgvp)
+
     PglErr PgrGetP(const uintptr_t* sample_include, PgrSampleSubsetIndexStruct pssi, uint32_t sample_ct, uint32_t vidx, PgenReaderStruct* pgr_ptr, uintptr_t* genovec, uintptr_t* phasepresent, uintptr_t* phaseinfo, uint32_t* phasepresent_ct_ptr)
 
+    PglErr PgrGetMP(const uintptr_t* sample_include, PgrSampleSubsetIndexStruct pssi, uint32_t sample_ct, uint32_t vidx, PgenReaderStruct* pgr_ptr, PgenVariantStruct* pgvp)
+
     PglErr PgrGet1D(const uintptr_t* sample_include, PgrSampleSubsetIndexStruct pssi, uint32_t sample_ct, uint32_t vidx, AlleleCode allele_idx, PgenReaderStruct* pgr_ptr, uintptr_t* allele_countvec, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr)
 
     PglErr PgrGetCounts(const uintptr_t* sample_include, const uintptr_t* sample_include_interleaved_vec, PgrSampleSubsetIndexStruct pssi, uint32_t sample_ct, uint32_t vidx, PgenReaderStruct* pgr_ptr, uint32_t* genocounts)
 
     BoolErr CleanupPgfi(PgenFileInfo* pgfip, PglErr* reterrp)
     BoolErr CleanupPgr(PgenReaderStruct* pgr_ptr, PglErr* reterrp)
 
 
 cdef extern from "../plink2/include/pgenlib_write.h" namespace "plink2":
     cdef cppclass PgenWriterCommon:
-        uint32_t variant_ct
-        uint32_t sample_ct
-        uintptr_t* allele_idx_offsets
-        uint32_t vidx
+        pass
+
+    ctypedef enum PgenWriteMode:
+        kPgenWriteBackwardSeek
+        kPgenWriteAndCopy
 
     cdef cppclass STPgenWriter:
         pass
 
     uint32_t SpgwGetVariantCt(STPgenWriter* spgwp)
 
     uint32_t SpgwGetSampleCt(STPgenWriter* spgwp)
@@ -172,50 +207,58 @@
 
     PglErr SpgwInitPhase1(const char* fname, uintptr_t* allele_idx_offsets, uintptr_t* explicit_nonref_flags, uint32_t variant_ct, uint32_t sample_ct, uint32_t optional_max_allele_ct, uint32_t separate_index, PgenGlobalFlags phase_dosage_gflags, uint32_t nonref_flags_storage, STPgenWriter* spgwp, uintptr_t* alloc_cacheline_ct_ptr, uint32_t* max_vrec_len_ptr)
 
     void SpgwInitPhase2(uint32_t max_vrec_len, STPgenWriter* spgwp, unsigned char* spgw_alloc)
 
     PglErr SpgwAppendBiallelicGenovec(const uintptr_t* genovec, STPgenWriter* spgwp)
 
+    PglErr SpgwAppendMultiallelicSparse(const uintptr_t* genovec, const uintptr_t* patch_01_set, const AlleleCode* patch_01_vals, const uintptr_t* patch_10_set, const AlleleCode* patch_10_vals, uint32_t allele_ct, uint32_t patch_01_ct, uint32_t patch_10_ct, STPgenWriter* spgwp)
+
     PglErr SpgwAppendBiallelicGenovecHphase(const uintptr_t* genovec, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, STPgenWriter* spgwp)
 
+    PglErr SpgwAppendMultiallelicGenovecHphase(const uintptr_t* genovec, const uintptr_t* patch_01_set, const AlleleCode* patch_01_vals, const uintptr_t* patch_10_set, const AlleleCode* patch_10_vals, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, uint32_t allele_ct, uint32_t patch_01_ct, uint32_t patch_10_ct, STPgenWriter* spgwp)
+
     PglErr SpgwAppendBiallelicGenovecDosage16(const uintptr_t* genovec, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t dosage_ct, STPgenWriter* spgwp)
 
     PglErr SpgwFinish(STPgenWriter* spgwp)
 
     BoolErr CleanupSpgw(STPgenWriter* spgwp, PglErr* reterrp)
 
 
 cdef class PgenReader:
-    # todo: nonref_flags, multiallelic variant support
     cdef PgenFileInfo* _info_ptr
     cdef PgenReaderStruct* _state_ptr
     cdef uintptr_t* _subset_include_vec
     cdef uintptr_t* _subset_include_interleaved_vec
 
     cdef uint32_t* _subset_cumulative_popcounts
     cdef PgrSampleSubsetIndexStruct _subset_index
 
     cdef uint32_t _subset_size
     # preallocate buffers we'll use repeatedly
-    cdef uintptr_t* _genovec
-    cdef uintptr_t* _phasepresent
-    cdef uintptr_t* _phaseinfo
-    cdef uintptr_t* _dosage_present
-    cdef uint16_t* _dosage_main
+    cdef PgenVariantStruct _pgv
     cdef VecW* _transpose_batch_buf
     # for multi-variant load-and-transpose, we load up to
     # kPglNypTransposeBatch (= 256) variants at a time, and then transpose
     cdef uintptr_t* _multivar_vmaj_geno_buf
     cdef uintptr_t* _multivar_vmaj_phasepresent_buf
     cdef uintptr_t* _multivar_vmaj_phaseinfo_buf
     cdef uintptr_t* _multivar_smaj_geno_batch_buf
     cdef uintptr_t* _multivar_smaj_phaseinfo_batch_buf
     cdef uintptr_t* _multivar_smaj_phasepresent_batch_buf
 
+    cdef set_allele_idx_offsets_internal(self, np.ndarray[np.uintp_t,mode="c",ndim=1] allele_idx_offsets):
+        # Make a copy instead of trying to share this with the caller.
+        cdef uint32_t nvariant = self._info_ptr[0].raw_variant_ct
+        cdef uintptr_t nbytes = (1 + nvariant) * sizeof(uintptr_t)
+        if cachealigned_malloc(nbytes, &self._info_ptr[0].allele_idx_offsets):
+            raise MemoryError()
+        memcpy(self._info_ptr[0].allele_idx_offsets, &(allele_idx_offsets[0]), nbytes)
+        self._info_ptr[0].max_allele_ct = PglComputeMaxAlleleCt(self._info_ptr[0].allele_idx_offsets, nvariant)
+
     cdef set_sample_subset_internal(self, np.ndarray[np.uint32_t,mode="c",ndim=1] sample_subset):
         cdef uint32_t raw_sample_ct = self._info_ptr[0].raw_sample_ct
         cdef uint32_t raw_sample_ctv = DivUp(raw_sample_ct, kBitsPerVec)
         cdef uint32_t raw_sample_ctaw = raw_sample_ctv * kWordsPerVec
         cdef uintptr_t* sample_include = self._subset_include_vec
         ZeroWArr(raw_sample_ctaw, sample_include)
         cdef uint32_t subset_size = sample_subset.size
@@ -248,15 +291,16 @@
         PgrSetSampleSubsetIndex(self._subset_cumulative_popcounts, self._state_ptr, &(self._subset_index))
 
         self._subset_size = subset_size
         return
 
 
     def __cinit__(self, bytes filename, object raw_sample_ct = None,
-                  object variant_ct = None, object sample_subset = None):
+                  object variant_ct = None, object sample_subset = None,
+                  object allele_idx_offsets = None):
         self._info_ptr = <PgenFileInfo*>PyMem_Malloc(sizeof(PgenFileInfo))
         if not self._info_ptr:
             raise MemoryError()
         PreinitPgfi(self._info_ptr)
         # this depends on pgenlib_internal implementation.  could save
         # pgfi_alloc and pgr_alloc instead.
         self._info_ptr[0].vrtypes = NULL
@@ -268,50 +312,61 @@
             cur_variant_ct = variant_ct
         cdef const char* fname = <const char*>filename
         cdef PgenHeaderCtrl header_ctrl
         cdef uintptr_t pgfi_alloc_cacheline_ct
         cdef char errstr_buf[kPglErrstrBufBlen]
         if PgfiInitPhase1(fname, NULL, cur_variant_ct, cur_sample_ct, &header_ctrl, self._info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != kPglRetSuccess:
             raise RuntimeError(errstr_buf[7:])
-        assert (header_ctrl & 0x30) == 0 # no alt allele counts
-        cdef uintptr_t nonref_flags_byte_ct = DivUp(self._info_ptr[0].raw_variant_ct, kBitsPerWord) * kBytesPerWord
+        cdef uint32_t file_variant_ct = self._info_ptr[0].raw_variant_ct
+        if allele_idx_offsets is not None:
+            self.set_allele_idx_offsets_internal(allele_idx_offsets)
+        else:
+            if (header_ctrl & 0x30) != 0:
+                # Not tested yet, since C pgenlib doesn't yet have a mode which
+                # embeds a copy of allele_idx_offsets in the .pgen.
+                if cachealigned_malloc((file_variant_ct + 1) * sizeof(uintptr_t), &self._info_ptr[0].allele_idx_offsets):
+                    raise MemoryError()
+        cdef uintptr_t nonref_flags_byte_ct = DivUp(file_variant_ct, kBitsPerWord) * kBytesPerWord
         if (header_ctrl & 0xc0) == 0xc0:
             # explicit nonref_flags
             # might want to modify PgfiInitPhase2 to let us not load this.
             if cachealigned_malloc(nonref_flags_byte_ct, &self._info_ptr[0].nonref_flags):
                 raise MemoryError()
         cdef uint32_t file_sample_ct = self._info_ptr[0].raw_sample_ct
         assert file_sample_ct != 0
         cdef unsigned char* pgfi_alloc = NULL
         if pgfi_alloc_cacheline_ct != 0:
             if cachealigned_malloc(pgfi_alloc_cacheline_ct * kCacheline, &pgfi_alloc):
                 raise MemoryError()
         cdef uint32_t max_vrec_width
         cdef uintptr_t pgr_alloc_cacheline_ct
-        if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, self._info_ptr[0].raw_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
+        if PgfiInitPhase2(header_ctrl, 1, 0, 0, 0, file_variant_ct, &max_vrec_width, self._info_ptr, pgfi_alloc, &pgr_alloc_cacheline_ct, errstr_buf):
             if pgfi_alloc and not self._info_ptr[0].vrtypes:
                 aligned_free(pgfi_alloc)
             raise RuntimeError(errstr_buf[7:])
         if self._info_ptr[0].gflags & kfPgenGlobalMultiallelicHardcallFound:
-            # todo: support this by wrapping pvar loader the same way as the R
-            # interface
-            raise RuntimeError("Multiallelic + phase/dosage datasets not supported yet")
+            # todo: support easy initialization of allele_idx_offsets by
+            # wrapping pvar loader the same way as the R interface
+            if self._info_ptr[0].allele_idx_offsets == NULL:
+                raise RuntimeError("PgenReader: multiallelic variants present, but allele_idx_offsets not provided")
 
         self._state_ptr = <PgenReaderStruct*>PyMem_Malloc(sizeof(PgenReaderStruct))
         if not self._state_ptr:
             raise MemoryError()
         PreinitPgr(self._state_ptr)
         PgrSetFreadBuf(NULL, self._state_ptr)
         cdef uintptr_t pgr_alloc_main_byte_ct = pgr_alloc_cacheline_ct * kCacheline
         cdef uintptr_t sample_subset_byte_ct = DivUp(file_sample_ct, kBitsPerVec) * kBytesPerVec
         cdef uintptr_t cumulative_popcounts_byte_ct = DivUp(file_sample_ct, kBitsPerWord * kInt32PerVec) * kBytesPerVec
         cdef uintptr_t genovec_byte_ct = DivUp(file_sample_ct, kNypsPerVec) * kBytesPerVec
+        cdef uintptr_t patch_01_vals_byte_ct = RoundUpPow2(file_sample_ct * sizeof(AlleleCode), kBytesPerVec)
+        cdef uintptr_t patch_10_vals_byte_ct = RoundUpPow2(file_sample_ct * 2 * sizeof(AlleleCode), kBytesPerVec)
         cdef uintptr_t dosage_main_byte_ct = DivUp(file_sample_ct, (2 * kInt32PerVec)) * kBytesPerVec
         cdef unsigned char* pgr_alloc
-        if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 5) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
+        if cachealigned_malloc(pgr_alloc_main_byte_ct + (2 * kPglNypTransposeBatch + 7) * sample_subset_byte_ct + cumulative_popcounts_byte_ct + (1 + kPglNypTransposeBatch) * genovec_byte_ct + patch_01_vals_byte_ct + patch_10_vals_byte_ct + dosage_main_byte_ct + kPglBitTransposeBufbytes + 4 * (kPglNypTransposeBatch * kPglNypTransposeBatch // 8), &pgr_alloc):
             raise MemoryError()
         cdef PglErr reterr = PgrInit(fname, max_vrec_width, self._info_ptr, self._state_ptr, pgr_alloc)
         if reterr != kPglRetSuccess:
             if not PgrGetFreadBuf(self._state_ptr):
                 aligned_free(pgr_alloc)
             raise RuntimeError("PgrInit() error " + str(reterr))
         cdef unsigned char* pgr_alloc_iter = &(pgr_alloc[pgr_alloc_main_byte_ct])
@@ -321,23 +376,31 @@
         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
 
         # assumes kWordsPerVec <= 2
         self._subset_include_interleaved_vec[-1] = 0
 
         self._subset_cumulative_popcounts = <uint32_t*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[cumulative_popcounts_byte_ct])
-        self._genovec = <uintptr_t*>pgr_alloc_iter
+        self._pgv.genovec = <uintptr_t*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct])
-        self._phasepresent = <uintptr_t*>pgr_alloc_iter
+        self._pgv.patch_01_set = <uintptr_t*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
-        self._phaseinfo = <uintptr_t*>pgr_alloc_iter
+        self._pgv.patch_01_vals = <AlleleCode*>pgr_alloc_iter
+        pgr_alloc_iter = &(pgr_alloc_iter[patch_01_vals_byte_ct])
+        self._pgv.patch_10_set = <uintptr_t*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
-        self._dosage_present = <uintptr_t*>pgr_alloc_iter
+        self._pgv.patch_10_vals = <AlleleCode*>pgr_alloc_iter
+        pgr_alloc_iter = &(pgr_alloc_iter[patch_10_vals_byte_ct])
+        self._pgv.phasepresent = <uintptr_t*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
-        self._dosage_main = <uint16_t*>pgr_alloc_iter
+        self._pgv.phaseinfo = <uintptr_t*>pgr_alloc_iter
+        pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+        self._pgv.dosage_present = <uintptr_t*>pgr_alloc_iter
+        pgr_alloc_iter = &(pgr_alloc_iter[sample_subset_byte_ct])
+        self._pgv.dosage_main = <uint16_t*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[dosage_main_byte_ct])
         if sample_subset is not None:
             self.set_sample_subset_internal(sample_subset)
         else:
             self._subset_size = file_sample_ct
         self._transpose_batch_buf = <VecW*>pgr_alloc_iter
         pgr_alloc_iter = &(pgr_alloc_iter[kPglBitTransposeBufbytes])
@@ -369,28 +432,28 @@
 
 
     cpdef hardcall_phase_present(self):
         return ((self._info_ptr[0].gflags & kfPgenGlobalHardcallPhasePresent) != 0)
 
 
     cpdef read(self, uint32_t variant_idx, np.ndarray geno_int_out, uint32_t allele_idx = 1):
-        # for full genotype info for multiallelic variants, use read_phased()
+        # for full genotype info for multiallelic variants, use read_alleles()
         # instead
 
         # when this is too much bounds-checking, caller should be using
         # read_range() or read_list()
         if variant_idx >= self._info_ptr[0].raw_variant_ct:
             raise RuntimeError("read() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
         if geno_int_out.ndim != 1:
             raise RuntimeError("read() requires geno_int_out to be one-dimensional.")
         cdef uint32_t subset_size = self._subset_size
         if geno_int_out.shape[0] < subset_size:
             raise RuntimeError("read() geno_int_out is too small (" + str(geno_int_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
 
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef PglErr reterr = PgrGet1(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, genovec)
         if reterr != kPglRetSuccess:
             raise RuntimeError("read() error " + str(reterr))
         cdef int8_t* data8_ptr
         cdef int32_t* data32_ptr
         cdef int64_t* data64_ptr
         if geno_int_out.dtype == np.int8:
@@ -413,78 +476,74 @@
         if floatarr_out.ndim != 1:
             raise RuntimeError("read_dosages() requires floatarr_out to be one-dimensional.")
         cdef uint32_t subset_size = self._subset_size
         if floatarr_out.shape[0] < subset_size:
             raise RuntimeError("read_dosages() floatarr_out is too small (" + str(floatarr_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
 
         cdef uint32_t dosage_ct
-        cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._genovec, self._dosage_present, self._dosage_main, &dosage_ct)
+        cdef PglErr reterr = PgrGet1D(self._subset_include_vec, self._subset_index, subset_size, variant_idx, allele_idx, self._state_ptr, self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, &dosage_ct)
         if reterr != kPglRetSuccess:
             raise RuntimeError("read_dosages() error " + str(reterr))
         cdef float* data32_ptr
         cdef double* data64_ptr
         if floatarr_out.dtype == np.float32:
             data32_ptr = <float*>floatarr_out.data
-            Dosage16ToFloatsMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data32_ptr)
+            Dosage16ToFloatsMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data32_ptr)
         elif floatarr_out.dtype == np.float64:
             data64_ptr = <double*>floatarr_out.data
-            Dosage16ToDoublesMinus9(self._genovec, self._dosage_present, self._dosage_main, subset_size, dosage_ct, data64_ptr)
+            Dosage16ToDoublesMinus9(self._pgv.genovec, self._pgv.dosage_present, self._pgv.dosage_main, subset_size, dosage_ct, data64_ptr)
         else:
             raise RuntimeError("Invalid read_dosages() floatarr_out array element type (float32 or float64 expected).")
         return
 
 
     cpdef read_alleles(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out):
         if variant_idx >= self._info_ptr[0].raw_variant_ct:
             raise RuntimeError("read_alleles() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
         cdef uint32_t subset_size = self._subset_size
         if allele_int32_out.shape[0] < subset_size:
             raise RuntimeError("read_alleles() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
 
-        cdef uint32_t phasepresent_ct
-        # upgrade to multiallelic version of this function in the future
-        cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+        cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
         if reterr != kPglRetSuccess:
             raise RuntimeError("read_alleles() error " + str(reterr))
         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
-        GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+        GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
         return
 
 
     cpdef read_alleles_and_phasepresent(self, uint32_t variant_idx, np.ndarray[np.int32_t,mode="c",ndim=1] allele_int32_out, np.ndarray[np.uint8_t,mode="c",cast=True] phasepresent_out):
         if variant_idx >= self._info_ptr[0].raw_variant_ct:
             raise RuntimeError("read_alleles_and_phasepresent() variant_idx too large (" + str(variant_idx) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
         cdef uint32_t subset_size = self._subset_size
         if allele_int32_out.shape[0] < 2 * subset_size:
             raise RuntimeError("read_alleles_and_phasepresent() allele_int32_out is too small (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", , and column count should be twice that).")
         if phasepresent_out.shape[0] < subset_size:
             raise RuntimeError("read_alleles_and_phasepresent() phasepresent_out is too small (" + str(phasepresent_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ").")
 
-        cdef uint32_t phasepresent_ct
-        # upgrade to multiallelic version of this function in the future
-        cdef PglErr reterr = PgrGetP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, self._genovec, self._phasepresent, self._phaseinfo, &phasepresent_ct)
+        cdef PglErr reterr = PgrGetMP(self._subset_include_vec, self._subset_index, subset_size, variant_idx, self._state_ptr, &self._pgv)
         if reterr != kPglRetSuccess:
             raise RuntimeError("read_alleles_and_phasepresent() error " + str(reterr))
         cdef int32_t* main_data_ptr = <int32_t*>(&(allele_int32_out[0]))
         cdef unsigned char* phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[0]))
-        GenoarrPhasedToAlleleCodesMinus9(self._genovec, self._phasepresent, self._phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+        GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
         return
 
 
     cdef read_range_internal8(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         # todo: benchmark effect of adding @cython.boundscheck(False) and
         # @cython.wraparound(False) annotations to these multiple-variant-load
         # functions
         # todo: experiment with using multiple pthreads under the hood for
         # large jobs (this probably belongs under pgenlib_ffi_support, rather
         # than here)
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef int8_t* data_ptr
         cdef uint32_t variant_idx
         cdef PglErr reterr
         if sample_maj == 0:
             if geno_int8_out.shape[0] < variant_idx_ct:
@@ -542,15 +601,15 @@
             variant_idx_offset += kPglNypTransposeBatch
         return
 
     cdef read_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef int32_t* data_ptr
         cdef uint32_t variant_idx
         cdef PglErr reterr
         if sample_maj == 0:
             if geno_int32_out.shape[0] < variant_idx_ct:
@@ -608,15 +667,15 @@
             variant_idx_offset += kPglNypTransposeBatch
         return
 
     cdef read_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef int64_t* data_ptr
         cdef uint32_t variant_idx
         cdef PglErr reterr
         if sample_maj == 0:
             if geno_int64_out.shape[0] < variant_idx_ct:
@@ -689,15 +748,15 @@
 
 
     cdef read_list_internal8(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int8_t,mode="c",ndim=2] geno_int8_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef int8_t* data_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
         cdef PglErr reterr
         if sample_maj == 0:
@@ -761,15 +820,15 @@
         return
 
     cdef read_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] geno_int32_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef int32_t* data_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
         cdef PglErr reterr
         if sample_maj == 0:
@@ -833,15 +892,15 @@
         return
 
     cdef read_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int64_t,mode="c",ndim=2] geno_int64_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* genovec = self._pgv.genovec
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef int64_t* data_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
         cdef PglErr reterr
         if sample_maj == 0:
@@ -922,42 +981,44 @@
         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
         #   rows, variant_idx_ct columns
         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* phasepresent = self._phasepresent
-        cdef uintptr_t* phaseinfo = self._phaseinfo
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* phasepresent = self._pgv.phasepresent
+        cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef int32_t* main_data_ptr
         cdef uint32_t variant_idx
-        cdef uint32_t phasepresent_ct
         cdef PglErr reterr
         if hap_maj == 0:
             if allele_int32_out.shape[0] < variant_idx_ct:
                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
             if allele_int32_out.shape[1] < 2 * subset_size:
                 raise RuntimeError("Variant-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
             for variant_idx in range(variant_idx_start, variant_idx_end):
-                # upgrade to multiallelic version of this function later
-                reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+                reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
                 if reterr != kPglRetSuccess:
                     raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
-                GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+                GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
             return
         if variant_idx_start >= variant_idx_end:
             raise RuntimeError("read_alleles_range() variant_idx_start >= variant_idx_end (" + str(variant_idx_start) + ", " + str(variant_idx_end) + ")")
         if allele_int32_out.shape[0] < 2 * subset_size:
             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
         if allele_int32_out.shape[1] < variant_idx_ct:
             raise RuntimeError("Haplotype-major read_alleles_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
+        cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+        if allele_idx_offsets != NULL:
+            if allele_idx_offsets[variant_idx_end] - allele_idx_offsets[variant_idx_start] != (variant_idx_end - variant_idx_start) * k1LU * 2:
+                raise RuntimeError("Haplotype-major read_alleles_range() doesn't support multiallelic variants yet.")
         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
         cdef uint32_t variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
         cdef uint32_t variant_idx_offset = variant_idx_start
         cdef uint32_t sample_ctaw2 = kWordsPerVec * DivUp(subset_size, kBitsPerWordD2)
         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
@@ -970,14 +1031,15 @@
         cdef uintptr_t* vmaj_phaseinfo_iter
         cdef uintptr_t* smaj_geno_iter
         cdef uintptr_t* smaj_phaseinfo_iter
         cdef int32_t* main_data1_ptr
         cdef uint32_t variant_batch_idx
         cdef uint32_t sample_batch_size
         cdef uint32_t sample_batch_idx
+        cdef uint32_t phasepresent_ct
         cdef uint32_t uii
         for variant_batch_idx in range(variant_batch_ct):
             if variant_batch_idx == (variant_batch_ct - 1):
                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
             vmaj_geno_iter = multivar_vmaj_geno_buf
             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
@@ -1021,39 +1083,37 @@
         #   variant_idx_ct rows, 2 * sample_ct columns
         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
         #   rows, variant_idx_ct columns
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* phasepresent = self._phasepresent
-        cdef uintptr_t* phaseinfo = self._phaseinfo
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* phasepresent = self._pgv.phasepresent
+        cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef int32_t* main_data_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
-        cdef uint32_t phasepresent_ct
         cdef PglErr reterr
         if hap_maj == 0:
             if allele_int32_out.shape[0] < variant_idx_ct:
                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
             if allele_int32_out.shape[1] < 2 * subset_size:
                 raise RuntimeError("Variant-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
             for variant_list_idx in range(variant_idx_ct):
                 variant_idx = variant_idxs[variant_list_idx]
                 if variant_idx >= raw_variant_ct:
                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
-                # upgrade to multiallelic version of this function later
-                reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+                reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
                 if reterr != kPglRetSuccess:
                     raise RuntimeError("read_alleles_list() error " + str(reterr))
                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
-                GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, NULL, main_data_ptr)
+                GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, NULL, main_data_ptr)
             return
         if allele_int32_out.shape[0] < 2 * subset_size:
             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; current sample subset has size " + str(subset_size) + ", and row count should be twice that)")
         if allele_int32_out.shape[1] < variant_idx_ct:
             raise RuntimeError("Haplotype-major read_alleles_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
         cdef uint32_t variant_batch_ct = DivUp(variant_idx_ct, kPglNypTransposeBatch)
         cdef uint32_t variant_batch_size = kPglNypTransposeBatch
@@ -1062,33 +1122,38 @@
         cdef uint32_t sample_ctaw = kWordsPerVec * DivUp(subset_size, kBitsPerWord)
         cdef uint32_t sample_batch_ct = DivUp(subset_size, kPglNypTransposeBatch)
         cdef VecW* transpose_batch_buf = self._transpose_batch_buf
         cdef uintptr_t* multivar_vmaj_geno_buf = self._multivar_vmaj_geno_buf
         cdef uintptr_t* multivar_vmaj_phaseinfo_buf = self._multivar_vmaj_phaseinfo_buf
         cdef uintptr_t* multivar_smaj_geno_batch_buf = self._multivar_smaj_geno_batch_buf
         cdef uintptr_t* multivar_smaj_phaseinfo_batch_buf = self._multivar_smaj_phaseinfo_batch_buf
+        cdef uintptr_t* allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
         cdef uintptr_t* vmaj_geno_iter
         cdef uintptr_t* vmaj_phaseinfo_iter
         cdef uintptr_t* smaj_geno_iter
         cdef uintptr_t* smaj_phaseinfo_iter
         cdef int32_t* main_data1_ptr
         cdef uint32_t variant_batch_idx
         cdef uint32_t sample_batch_size
         cdef uint32_t sample_batch_idx
+        cdef uint32_t phasepresent_ct
         cdef uint32_t uii
         for variant_batch_idx in range(variant_batch_ct):
             if variant_batch_idx == (variant_batch_ct - 1):
                 variant_batch_size = 1 + <uint32_t>((variant_idx_ct - 1) % kPglNypTransposeBatch)
                 variant_batch_sizel = DivUp(variant_batch_size, kBitsPerWord)
             vmaj_geno_iter = multivar_vmaj_geno_buf
             vmaj_phaseinfo_iter = multivar_vmaj_phaseinfo_buf
             for variant_list_idx in range(variant_batch_idx * kPglNypTransposeBatch, variant_batch_idx * kPglNypTransposeBatch + variant_batch_size):
                 variant_idx = variant_idxs[variant_list_idx]
                 if variant_idx >= raw_variant_ct:
                     raise RuntimeError("read_alleles_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
+                if allele_idx_offsets != NULL:
+                    if allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx] != 2:
+                        raise RuntimeError("Haplotype-major read_alleles_list() doesn't support multiallelic variants yet.")
                 reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, vmaj_geno_iter, phasepresent, vmaj_phaseinfo_iter, &phasepresent_ct)
                 if reterr != kPglRetSuccess:
                     raise RuntimeError("read_alleles_list() error " + str(reterr))
                 if phasepresent_ct == 0:
                     ZeroWArr(sample_ctaw, vmaj_phaseinfo_iter)
                 # else:
                     # BitvecAnd(phasepresent, sample_ctaw, vmaj_phaseinfo_iter)
@@ -1123,96 +1188,89 @@
         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
         #   rows, variant_idx_ct columns
         if variant_idx_end > self._info_ptr[0].raw_variant_ct:
             raise RuntimeError("read_alleles_range() variant_idx_end too large (" + str(variant_idx_end) + "; only " + str(self._info_ptr[0].raw_variant_ct) + " in file)")
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* phasepresent = self._phasepresent
-        cdef uintptr_t* phaseinfo = self._phaseinfo
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef int32_t* main_data_ptr
         cdef unsigned char* phasepresent_data_ptr
         cdef uint32_t variant_idx
-        cdef uint32_t phasepresent_ct
         cdef PglErr reterr
         if hap_maj == 0:
             if allele_int32_out.shape[0] < variant_idx_ct:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
             if phasepresent_out.shape[0] < variant_idx_ct:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; (variant_idx_end - variant_idx_start) is " + str(variant_idx_ct) + ")")
             if allele_int32_out.shape[1] < 2 * subset_size:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
             if phasepresent_out.shape[1] < subset_size:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_range() phasepresent_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
             for variant_idx in range(variant_idx_start, variant_idx_end):
-                # upgrade to multiallelic version of this function later
-                reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+                reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
                 if reterr != kPglRetSuccess:
-                    raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_range() error " + str(reterr))
+                    raise RuntimeError("variant_idx " + str(variant_idx) + " read_alleles_and_phasepresent_range() error " + str(reterr))
                 main_data_ptr = <int32_t*>(&(allele_int32_out[(variant_idx - variant_idx_start), 0]))
                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[(variant_idx - variant_idx_start), 0]))
-                GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+                GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
             return
         raise RuntimeError("read_alleles_and_phasepresent_range() does not support hap_maj == 1 yet.")
 
 
     cpdef read_alleles_and_phasepresent_list(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_out, np.ndarray[np.uint8_t,cast=True,mode="c",ndim=2] phasepresent_out, bint hap_maj = 0):
         # if hap_maj == False, allele_int32_out must have at least
         #   variant_idx_ct rows, 2 * sample_ct columns
         # if hap_maj == True, allele_int32_out must have at least 2 * sample_ct
         #   rows, variant_idx_ct columns
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* phasepresent = self._phasepresent
-        cdef uintptr_t* phaseinfo = self._phaseinfo
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* phasepresent = self._pgv.phasepresent
+        cdef uintptr_t* phaseinfo = self._pgv.phaseinfo
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef int32_t* main_data_ptr
         cdef unsigned char* phasepresent_data_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
-        cdef uint32_t phasepresent_ct
         cdef PglErr reterr
         if hap_maj == 0:
             if allele_int32_out.shape[0] < variant_idx_ct:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few rows (" + str(allele_int32_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
             if phasepresent_out.shape[0] < variant_idx_ct:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few rows (" + str(phasepresent_out.shape[0]) + "; variant_idxs length is " + str(variant_idx_ct) + ")")
             if allele_int32_out.shape[1] < 2 * subset_size:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() allele_int32_out buffer has too few columns (" + str(allele_int32_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ", and column count should be twice that)")
             if phasepresent_out.shape[1] < subset_size:
                 raise RuntimeError("Variant-major read_alleles_and_phasepresent_list() phasepresent_out buffer has too few columns (" + str(phasepresent_out.shape[1]) + "; current sample subset has size " + str(subset_size) + ")")
             for variant_list_idx in range(variant_idx_ct):
                 variant_idx = variant_idxs[variant_list_idx]
                 if variant_idx >= raw_variant_ct:
                     raise RuntimeError("read_alleles_and_phasepresent_list() variant index too large (" + str(variant_idx) + "; only " + str(raw_variant_ct) + " in file)")
-                # upgrade to multiallelic version of this function later
-                reterr = PgrGetP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, genovec, phasepresent, phaseinfo, &phasepresent_ct)
+                reterr = PgrGetMP(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
                 if reterr != kPglRetSuccess:
                     raise RuntimeError("read_alleles_and_phasepresent_list() error " + str(reterr))
                 main_data_ptr = <int32_t*>(&(allele_int32_out[variant_list_idx, 0]))
                 phasepresent_data_ptr = <unsigned char*>(&(phasepresent_out[variant_list_idx, 0]))
-                GenoarrPhasedToAlleleCodesMinus9(genovec, phasepresent, phaseinfo, subset_size, phasepresent_ct, phasepresent_data_ptr, main_data_ptr)
+                GenoarrMPToAlleleCodesMinus9(&self._pgv, subset_size, phasepresent_data_ptr, main_data_ptr)
             return
         raise RuntimeError("read_alleles_and_phasepresent_list() does not support hap_maj == 1 yet.")
 
 
     cdef read_dosages_range_internal32(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* dosage_present = self._dosage_present
-        cdef uint16_t* dosage_main = self._dosage_main
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* dosage_present = self._pgv.dosage_present
+        cdef uint16_t* dosage_main = self._pgv.dosage_main
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef float* data32_ptr
         cdef uint32_t variant_idx
         cdef uint32_t dosage_ct
         cdef PglErr reterr
         if sample_maj == 0:
@@ -1226,17 +1284,17 @@
         raise RuntimeError("read_dosages_range() does not support sample_maj == 1 yet.")
 
 
     cdef read_dosages_range_internal64(self, uint32_t variant_idx_start, uint32_t variant_idx_end, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* dosage_present = self._dosage_present
-        cdef uint16_t* dosage_main = self._dosage_main
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* dosage_present = self._pgv.dosage_present
+        cdef uint16_t* dosage_main = self._pgv.dosage_main
         cdef uint32_t variant_idx_ct = variant_idx_end - variant_idx_start
         cdef uint32_t subset_size = self._subset_size
         cdef double* data64_ptr
         cdef uint32_t variant_idx
         cdef uint32_t dosage_ct
         cdef PglErr reterr
         if sample_maj == 0:
@@ -1263,17 +1321,17 @@
 
 
     cdef read_dosages_list_internal32(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* dosage_present = self._dosage_present
-        cdef uint16_t* dosage_main = self._dosage_main
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* dosage_present = self._pgv.dosage_present
+        cdef uint16_t* dosage_main = self._pgv.dosage_main
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef float* data32_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
         cdef uint32_t dosage_ct
         cdef PglErr reterr
@@ -1296,17 +1354,17 @@
 
 
     cdef read_dosages_list_internal64(self, np.ndarray[np.uint32_t] variant_idxs, np.ndarray[np.float64_t,mode="c",ndim=2] floatarr_out, uint32_t allele_idx = 1, bint sample_maj = 0):
         cdef uint32_t raw_variant_ct = self._info_ptr[0].raw_variant_ct
         cdef const uintptr_t* subset_include_vec = self._subset_include_vec
         cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
         cdef PgenReaderStruct* pgrp = self._state_ptr
-        cdef uintptr_t* genovec = self._genovec
-        cdef uintptr_t* dosage_present = self._dosage_present
-        cdef uint16_t* dosage_main = self._dosage_main
+        cdef uintptr_t* genovec = self._pgv.genovec
+        cdef uintptr_t* dosage_present = self._pgv.dosage_present
+        cdef uint16_t* dosage_main = self._pgv.dosage_main
         cdef uint32_t variant_idx_ct = <uint32_t>variant_idxs.shape[0]
         cdef uint32_t subset_size = self._subset_size
         cdef double* data64_ptr
         cdef uint32_t variant_list_idx
         cdef uint32_t variant_idx
         cdef uint32_t dosage_ct
         cdef PglErr reterr
@@ -1334,19 +1392,69 @@
         elif floatarr_out.dtype == np.float64:
             self.read_dosages_list_internal64(variant_idxs, floatarr_out, allele_idx, sample_maj)
         else:
             raise RuntimeError("Invalid read_dosages_list() floatarr_out array element type (float32 or float64 expected).")
         return
 
     cpdef count(self, uint32_t variant_idx, np.ndarray[np.uint32_t,mode="c"] genocount_uint32_out, object allele_idx = 1):
-        # todo: multiallelic variants
+        cdef const uintptr_t* subset_include_vec = self._subset_include_vec
+        cdef PgrSampleSubsetIndexStruct subset_index = self._subset_index
+        cdef uint32_t subset_size = self._subset_size
+        cdef PgenReaderStruct* pgrp = self._state_ptr
+        cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
+        cdef uintptr_t* allele_idx_offsets
+        cdef uintptr_t* genovec
+        cdef AlleleCode* patch_01_vals
+        cdef AlleleCode* patch_10_vals
+        cdef uintptr_t type_ct
+        cdef uintptr_t allele_ct
+        cdef uint32_t patch_01_ct
+        cdef uint32_t patch_10_ct
+        cdef uintptr_t ac0
+        cdef uintptr_t ac1
+        cdef PglErr reterr
         if allele_idx is None:
+            allele_ct = 2
+            allele_idx_offsets = self._info_ptr[0].allele_idx_offsets
+            if allele_idx_offsets != NULL:
+                allele_ct = allele_idx_offsets[variant_idx + 1] - allele_idx_offsets[variant_idx]
+            if allele_ct > 2:
+                # Multiallelic special case.
+                type_ct = (allele_ct * (allele_ct + 1)) >> 1
+                if genocount_uint32_out.shape[0] <= type_ct:
+                    raise RuntimeError("count() genocount_uint32_out buffer is too small for multiallelic variant")
+                reterr = PgrGetM(subset_include_vec, subset_index, subset_size, variant_idx, pgrp, &self._pgv)
+                if reterr != kPglRetSuccess:
+                    raise RuntimeError("count() error " + str(reterr))
+                genovec = self._pgv.genovec
+                ZeroTrailingNyps(subset_size, genovec)
+                GenoarrCountFreqsUnsafe(genovec, subset_size, data_ptr)
+                data_ptr[type_ct] = data_ptr[3]
+                for i in range(3, type_ct):
+                    data_ptr[i] = 0
+                # Now correct for patch_01 and patch_10.
+                patch_01_ct = self._pgv.patch_01_ct
+                if patch_01_ct > 0:
+                    data_ptr[1] -= patch_01_ct
+                    patch_01_vals = self._pgv.patch_01_vals
+                    # could special-case allele_ct == 3, etc.
+                    for i in range(patch_01_ct):
+                        ac1 = patch_01_vals[i]
+                        data_ptr[(ac1 * (ac1 + 1)) >> 1] += 1
+                patch_10_ct = self._pgv.patch_10_ct
+                if patch_10_ct > 0:
+                    data_ptr[2] -= patch_10_ct
+                    patch_10_vals = self._pgv.patch_10_vals
+                    for i in range(patch_10_ct):
+                        ac0 = patch_10_vals[2*i]
+                        ac1 = patch_10_vals[2*i+1]
+                        data_ptr[ac0 + ((ac1 * (ac1 + 1)) >> 1)] += 1
+                return
             allele_idx = 1
-        cdef uint32_t* data_ptr = <uint32_t*>(&(genocount_uint32_out[0]))
-        cdef PglErr reterr = PgrGetCounts(self._subset_include_vec, self._subset_include_interleaved_vec, self._subset_index, self._subset_size, variant_idx, self._state_ptr, data_ptr)
+        reterr = PgrGetCounts(subset_include_vec, self._subset_include_interleaved_vec, subset_index, subset_size, variant_idx, pgrp, data_ptr)
         if reterr != kPglRetSuccess:
             raise RuntimeError("count() error " + str(reterr))
         if allele_idx != 0:
             return
         cdef uint32_t tmp = data_ptr[0]
         data_ptr[0] = data_ptr[2]
         data_ptr[2] = tmp
@@ -1364,14 +1472,16 @@
 
     cpdef close(self):
         cdef PglErr reterr = kPglRetSuccess
         if self._info_ptr:
             CleanupPgfi(self._info_ptr, &reterr)
             if self._info_ptr[0].vrtypes:
                 aligned_free(self._info_ptr[0].vrtypes)
+            if self._info_ptr[0].allele_idx_offsets:
+                aligned_free(self._info_ptr[0].allele_idx_offsets)
             if self._info_ptr[0].nonref_flags:
                 aligned_free(self._info_ptr[0].nonref_flags)
             if self._state_ptr:
                 CleanupPgr(self._state_ptr, &reterr)
                 if PgrGetFreadBuf(self._state_ptr):
                     aligned_free(PgrGetFreadBuf(self._state_ptr))
                 PyMem_Free(self._state_ptr)
@@ -1390,53 +1500,73 @@
 
     def __dealloc__(self):
         cdef PglErr reterr = kPglRetSuccess
         if self._info_ptr:
             CleanupPgfi(self._info_ptr, &reterr)
             if self._info_ptr[0].vrtypes:
                 aligned_free(self._info_ptr[0].vrtypes)
+            if self._info_ptr[0].allele_idx_offsets:
+                aligned_free(self._info_ptr[0].allele_idx_offsets)
             if self._info_ptr[0].nonref_flags:
                 aligned_free(self._info_ptr[0].nonref_flags)
             if self._state_ptr:
                 CleanupPgr(self._state_ptr, &reterr)
                 if PgrGetFreadBuf(self._state_ptr):
                     aligned_free(PgrGetFreadBuf(self._state_ptr))
                 PyMem_Free(self._state_ptr)
             PyMem_Free(self._info_ptr)
         return
 
 
-
 cdef bytes_to_bits_internal(np.ndarray[np.uint8_t,mode="c",cast=True] boolbytes, uint32_t sample_ct, uintptr_t* bitarr):
     BytesToBitsUnsafe(boolbytes, sample_ct, bitarr)
 
+
 cdef class PgenWriter:
     cdef STPgenWriter* _state_ptr
     cdef uintptr_t* _nonref_flags
     cdef PgenGlobalFlags _phase_dosage_gflags
+    cdef uint32_t _allele_ct_limit
     # preallocate buffers we'll use repeatedly
     cdef uintptr_t* _genovec
+    cdef uintptr_t* _patch_01_set
+    cdef AlleleCode* _patch_01_vals
+    cdef uintptr_t* _patch_10_set
+    cdef AlleleCode* _patch_10_vals
     cdef uintptr_t* _phasepresent
     cdef uintptr_t* _phaseinfo
     cdef uintptr_t* _dosage_present
     cdef uint16_t* _dosage_main
 
-
     def __cinit__(self, bytes filename, uint32_t sample_ct,
-                  uint32_t variant_ct, object nonref_flags,
-                  object allele_idx_offsets = None,
+                  object variant_ct = None, object nonref_flags = True,
+                  uint32_t allele_ct_limit = 2,
                   bint hardcall_phase_present = False,
                   bint dosage_present = False,
-                  bint dosage_phase_present = False):
+                  bint dosage_phase_present = False,
+                  object variant_ct_limit = None):
+        cdef PgenWriteMode write_mode = kPgenWriteBackwardSeek
+        cdef uint32_t cur_variant_ct_limit
+        if variant_ct is not None:
+            # Could also enforce variant_ct >= variant_ct_limit when both are
+            # provided?
+            cur_variant_ct_limit = variant_ct
+        elif variant_ct_limit is not None:
+            write_mode = kPgenWriteAndCopy
+            cur_variant_ct_limit = variant_ct_limit
+        else:
+            raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct or variant_ct_upper_bound must be provided).")
+        if cur_variant_ct_limit == 0 or cur_variant_ct_limit > kPglMaxVariantCt:
+            raise RuntimeError("Invalid arguments for PgenWriter constructor (variant_ct must be positive, and less than ~2^31).")
         if dosage_phase_present and not dosage_present:
             raise RuntimeError("Invalid arguments for PgenWriter constructor (dosage_phase_present true but dosage_present false).")
-        if allele_idx_offsets is not None:
-            for uii in range(variant_ct + 1):
-                if allele_idx_offsets[uii] != uii * 2:
-                    raise RuntimeError("Multiallelic variants aren't supported by PgenWriter yet.")
+        if allele_ct_limit != 2:
+            if (allele_ct_limit < 2) or (allele_ct_limit > kPglMaxAlleleCt):
+                raise RuntimeError("Invalid arguments for PgenWriter constructor (allele_ct_limit must be in [2, 255]).")
+            write_mode = kPgenWriteAndCopy
 
         self._state_ptr = <STPgenWriter*>PyMem_Malloc(sizeof(STPgenWriter))
         if not self._state_ptr:
             raise MemoryError()
         self._nonref_flags = NULL
         cdef uint32_t nonref_flags_storage = 0
         cdef uint32_t bitvec_cacheline_ct = DivUp(sample_ct, kBitsPerCacheline)
@@ -1455,30 +1585,56 @@
         cdef PgenGlobalFlags phase_dosage_gflags = kfPgenGlobal0
         if hardcall_phase_present:
             phase_dosage_gflags |= kfPgenGlobalHardcallPhasePresent
         if dosage_present:
             phase_dosage_gflags |= kfPgenGlobalDosagePresent
         self._phase_dosage_gflags = phase_dosage_gflags
         assert not dosage_phase_present
+
         cdef uintptr_t alloc_cacheline_ct
         cdef uint32_t max_vrec_len
-        cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, variant_ct, sample_ct, 0, 0, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
+        cdef PglErr reterr = SpgwInitPhase1(fname, NULL, self._nonref_flags, cur_variant_ct_limit, sample_ct, allele_ct_limit, write_mode, phase_dosage_gflags, nonref_flags_storage, self._state_ptr, &alloc_cacheline_ct, &max_vrec_len)
         if reterr != kPglRetSuccess:
             raise RuntimeError("SpgwInitPhase1() error " + str(reterr))
         cdef uint32_t genovec_cacheline_ct = DivUp(sample_ct, kNypsPerCacheline)
+        cdef uint32_t patch_01_vals_cacheline_ct = DivUp(sample_ct * sizeof(AlleleCode), kCacheline)
+        cdef uint32_t patch_10_vals_cacheline_ct = DivUp(sample_ct * 2 * sizeof(AlleleCode), kCacheline)
         cdef uint32_t dosage_main_cacheline_ct = DivUp(sample_ct, (2 * kInt32PerCacheline))
         cdef unsigned char* spgw_alloc
-        if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
+        if cachealigned_malloc((alloc_cacheline_ct + genovec_cacheline_ct + 5 * bitvec_cacheline_ct + patch_01_vals_cacheline_ct + patch_10_vals_cacheline_ct + dosage_main_cacheline_ct) * kCacheline, &spgw_alloc):
             raise MemoryError()
         SpgwInitPhase2(max_vrec_len, self._state_ptr, spgw_alloc)
-        self._genovec = <uintptr_t*>(&(spgw_alloc[alloc_cacheline_ct * kCacheline]))
-        self._phasepresent = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct) * kCacheline]))
-        self._phaseinfo = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + bitvec_cacheline_ct) * kCacheline]))
-        self._dosage_present = <uintptr_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 2 * bitvec_cacheline_ct) * kCacheline]))
-        self._dosage_main = <uint16_t*>(&(spgw_alloc[(alloc_cacheline_ct + genovec_cacheline_ct + 3 * bitvec_cacheline_ct) * kCacheline]))
+        cdef unsigned char* spgw_alloc_iter = &(spgw_alloc[alloc_cacheline_ct * kCacheline])
+        self._allele_ct_limit = allele_ct_limit
+        self._genovec = <uintptr_t*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[genovec_cacheline_ct * kCacheline])
+
+        # Can't skimp on patch_{01,10}_{set,vals} allocations even when
+        # allele_ct_limit == 2, due to how ConvertMultiAlleleCodesUnsafe()
+        # works.
+        # Could skimp on dosage/phase, but that doesn't gain us much.
+        self._patch_01_set = <uintptr_t*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+        self._patch_01_vals = <AlleleCode*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[patch_01_vals_cacheline_ct * kCacheline])
+        self._patch_10_set = <uintptr_t*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+        self._patch_10_vals = <AlleleCode*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[patch_10_vals_cacheline_ct * kCacheline])
+        self._phasepresent = <uintptr_t*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+        self._phaseinfo = <uintptr_t*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+        self._dosage_present = <uintptr_t*>(spgw_alloc_iter)
+        spgw_alloc_iter = &(spgw_alloc_iter[bitvec_cacheline_ct * kCacheline])
+        self._dosage_main = <uint16_t*>(spgw_alloc_iter)
+        # bugfix (16 Apr 2023): SpgwAppendBiallelicGenovec[Hphase] assumes
+        # trailing bits are clear
+        self._genovec[(sample_ct - 1) // kBitsPerWordD2] = 0
+        self._phasepresent[(sample_ct - 1) // kBitsPerWord] = 0
         return
 
 
     cpdef __enter__(self):
         return self
 
 
@@ -1487,41 +1643,91 @@
         BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
         cdef PglErr reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
         if reterr != kPglRetSuccess:
             raise RuntimeError("append_biallelic() error " + str(reterr))
         return
 
 
-    cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False):
+    cpdef append_alleles(self, np.ndarray[np.int32_t,mode="c"] allele_int32, bint all_phased = False, object allele_ct = None):
         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
+        cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+        cdef uint32_t allele_ct_limit = self._allele_ct_limit
         cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* patch_01_set = self._patch_01_set
+        cdef AlleleCode* patch_01_vals = self._patch_01_vals
+        cdef uintptr_t* patch_10_set = self._patch_10_set
+        cdef AlleleCode* patch_10_vals = self._patch_10_vals
+        cdef uintptr_t* phaseinfo = NULL
+        if all_phased:
+            if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
+                raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
+            phaseinfo = self._phaseinfo
+        cdef uint32_t patch_01_ct
+        cdef uint32_t patch_10_ct
+        cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, phaseinfo)
+        if observed_allele_ct == -1:
+            raise RuntimeError("append_alleles called with invalid allele codes")
+        cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+        if write_allele_ct > allele_ct_limit:
+            raise RuntimeError("append_alleles called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+        if allele_ct is not None:
+            if allele_ct < write_allele_ct:
+                raise RuntimeError("append_alleles called with allele codes >= allele_ct argument")
+            if allele_ct > allele_ct_limit:
+                raise RuntimeError("append_alleles called with allele_ct > allele_ct_limit")
+            write_allele_ct = allele_ct
         cdef PglErr reterr
         if not all_phased:
-            AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
-            reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+            if (patch_01_ct == 0) and (patch_10_ct == 0):
+                reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+            else:
+                reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
         else:
-            if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
-                raise RuntimeError("append_alleles called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
-            AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
-            reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+            if (patch_01_ct == 0) and (patch_10_ct == 0):
+                reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, phaseinfo, self._state_ptr)
+            else:
+                reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
         if reterr != kPglRetSuccess:
             raise RuntimeError("append_alleles() error " + str(reterr))
         return
 
 
-    cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent):
+    cpdef append_partially_phased(self, np.ndarray[np.int32_t,mode="c"] allele_int32, np.ndarray[np.uint8_t,cast=True] phasepresent, object allele_ct = None):
         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
             raise RuntimeError("append_partially_phased cannot be called when PgenWriter was constructed with hardcall_phase_present False")
         cdef int32_t* allele_codes = <int32_t*>(&(allele_int32[0]))
         cdef unsigned char* phasepresent_bytes = <unsigned char*>(&(phasepresent[0]))
+        cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+        cdef uint32_t allele_ct_limit = self._allele_ct_limit
         cdef uintptr_t* genovec = self._genovec
         cdef uintptr_t* phasepresent_buf = self._phasepresent
         cdef uintptr_t* phaseinfo = self._phaseinfo
-        AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
-        cdef PglErr reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+        cdef uintptr_t* patch_01_set = self._patch_01_set
+        cdef AlleleCode* patch_01_vals = self._patch_01_vals
+        cdef uintptr_t* patch_10_set = self._patch_10_set
+        cdef AlleleCode* patch_10_vals = self._patch_10_vals
+        cdef uint32_t patch_01_ct
+        cdef uint32_t patch_10_ct
+        cdef int32_t observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+        if observed_allele_ct == -1:
+            raise RuntimeError("append_partially_phased called with invalid allele codes")
+        cdef uint32_t write_allele_ct = <uint32_t>(observed_allele_ct)
+        if write_allele_ct > allele_ct_limit:
+            raise RuntimeError("append_partially_phased called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+        if allele_ct is not None:
+            if allele_ct < write_allele_ct:
+                raise RuntimeError("append_partially_phased called with allele codes >= allele_ct argument")
+            if allele_ct > allele_ct_limit:
+                raise RuntimeError("append_partially_phased called with allele_ct > allele_ct_limit")
+            write_allele_ct = allele_ct
+        cdef PglErr reterr
+        if (patch_01_ct == 0) and (patch_10_ct == 0):
+            reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+        else:
+            reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
         if reterr != kPglRetSuccess:
             raise RuntimeError("append_partially_phased() error " + str(reterr))
         return
 
 
     cdef append_dosages_internal32(self, np.ndarray[np.float32_t,mode="c"] floatarr):
         cdef uintptr_t* genovec = self._genovec
@@ -1567,55 +1773,122 @@
             BytesToGenoarrUnsafe(genobytes, SpgwGetSampleCt(self._state_ptr), self._genovec)
             reterr = SpgwAppendBiallelicGenovec(self._genovec, self._state_ptr)
             if reterr != kPglRetSuccess:
                 raise RuntimeError("append_biallelic_batch() error " + str(reterr))
         return
 
 
-    cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False):
+    cpdef append_alleles_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, bint all_phased = False, object allele_cts = None):
         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
+        cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+        cdef uint32_t allele_ct_limit = self._allele_ct_limit
         cdef uintptr_t* genovec = self._genovec
         cdef int32_t* allele_codes
+        cdef uintptr_t* patch_01_set = self._patch_01_set
+        cdef AlleleCode* patch_01_vals = self._patch_01_vals
+        cdef uintptr_t* patch_10_set = self._patch_10_set
+        cdef AlleleCode* patch_10_vals = self._patch_10_vals
         cdef uint32_t uii
+        cdef int32_t observed_allele_ct
+        cdef uint32_t write_allele_ct
+        cdef uint32_t casted_allele_ct
+        cdef uint32_t patch_01_ct
+        cdef uint32_t patch_10_ct
         cdef PglErr reterr
         if not all_phased:
             for uii in range(batch_size):
                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
-                AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, NULL)
-                reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+                observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, NULL)
+                if observed_allele_ct == -1:
+                    raise RuntimeError("append_alleles_batch called with invalid allele codes")
+                write_allele_ct = <uint32_t>(observed_allele_ct)
+                if write_allele_ct > allele_ct_limit:
+                    raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+                if allele_cts is not None:
+                    casted_allele_ct = allele_cts[uii]
+                    if casted_allele_ct < write_allele_ct:
+                        raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+                    if casted_allele_ct > allele_ct_limit:
+                        raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+                    write_allele_ct = casted_allele_ct
+                if (patch_01_ct == 0) and (patch_10_ct == 0):
+                    reterr = SpgwAppendBiallelicGenovec(genovec, self._state_ptr)
+                else:
+                    reterr = SpgwAppendMultiallelicSparse(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
                 if reterr != kPglRetSuccess:
                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
         else:
             if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
                 raise RuntimeError("append_alleles_batch called with all_phased True, but PgenWriter was constructed with hardcall_phase_present False")
             for uii in range(batch_size):
                 allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
-                AlleleCodesToGenoarrUnsafe(allele_codes, NULL, SpgwGetSampleCt(self._state_ptr), genovec, NULL, self._phaseinfo)
-                reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+                observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, NULL, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, NULL, self._phaseinfo)
+                if observed_allele_ct == -1:
+                    raise RuntimeError("append_alleles_batch called with invalid allele codes")
+                write_allele_ct = <uint32_t>(observed_allele_ct)
+                if write_allele_ct > allele_ct_limit:
+                    raise RuntimeError("append_alleles_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+                if allele_cts is not None:
+                    casted_allele_ct = allele_cts[uii]
+                    if casted_allele_ct < write_allele_ct:
+                        raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+                    if casted_allele_ct > allele_ct_limit:
+                        raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+                    write_allele_ct = casted_allele_ct
+                if (patch_01_ct == 0) and (patch_10_ct == 0):
+                    reterr = SpgwAppendBiallelicGenovecHphase(genovec, NULL, self._phaseinfo, self._state_ptr)
+                else:
+                    reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, NULL, self._phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
                 if reterr != kPglRetSuccess:
                     raise RuntimeError("append_alleles_batch() error " + str(reterr))
         return
 
 
-    cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch):
+    cpdef append_partially_phased_batch(self, np.ndarray[np.int32_t,mode="c",ndim=2] allele_int32_batch, np.ndarray[np.uint8_t,mode="c",cast=True,ndim=2] phasepresent_batch, object allele_cts = None):
         if (self._phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) == 0:
             raise RuntimeError("append_partially_phased_batch cannot be called when PgenWriter was constructed with hardcall_phase_present False")
         cdef uint32_t batch_size = <uint32_t>allele_int32_batch.shape[0]
+        cdef uint32_t sample_ct = SpgwGetSampleCt(self._state_ptr)
+        cdef uint32_t allele_ct_limit = self._allele_ct_limit
         cdef uintptr_t* genovec = self._genovec
+        cdef uintptr_t* patch_01_set = self._patch_01_set
+        cdef AlleleCode* patch_01_vals = self._patch_01_vals
+        cdef uintptr_t* patch_10_set = self._patch_10_set
+        cdef AlleleCode* patch_10_vals = self._patch_10_vals
         cdef uintptr_t* phasepresent_buf = self._phasepresent
         cdef uintptr_t* phaseinfo = self._phaseinfo
         cdef int32_t* allele_codes
         cdef unsigned char* phasepresent_bytes
         cdef uint32_t uii
+        cdef int32_t observed_allele_ct
+        cdef uint32_t write_allele_ct
+        cdef uint32_t casted_allele_ct
+        cdef uint32_t patch_01_ct
+        cdef uint32_t patch_10_ct
         cdef PglErr reterr
         for uii in range(batch_size):
             allele_codes = <int32_t*>(&(allele_int32_batch[uii, 0]))
             phasepresent_bytes = <unsigned char*>(&(phasepresent_batch[uii, 0]))
-            AlleleCodesToGenoarrUnsafe(allele_codes, phasepresent_bytes, SpgwGetSampleCt(self._state_ptr), genovec, phasepresent_buf, phaseinfo)
-            reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+            observed_allele_ct = ConvertMultiAlleleCodesUnsafe(allele_codes, phasepresent_bytes, sample_ct, genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, &patch_01_ct, &patch_10_ct, phasepresent_buf, phaseinfo)
+            if observed_allele_ct == -1:
+                raise RuntimeError("append_partially_phased_batch called with invalid allele codes")
+            write_allele_ct = <uint32_t>(observed_allele_ct)
+            if write_allele_ct > allele_ct_limit:
+                raise RuntimeError("append_partially_phased_batch called with allele codes >= allele_ct_limit; you may need to construct the PgenWriter with a higher allele_ct_limit setting")
+            if allele_cts is not None:
+                casted_allele_ct = allele_cts[uii]
+                if casted_allele_ct < write_allele_ct:
+                    raise RuntimeError("append_alleles_batch called with allele codes >= allele_cts[] value")
+                if casted_allele_ct > allele_ct_limit:
+                    raise RuntimeError("append_alleles_batch called with allele_cts[] value > allele_ct_limit")
+                write_allele_ct = casted_allele_ct
+            if (patch_01_ct == 0) and (patch_10_ct == 0):
+                reterr = SpgwAppendBiallelicGenovecHphase(genovec, phasepresent_buf, phaseinfo, self._state_ptr)
+            else:
+                reterr = SpgwAppendMultiallelicGenovecHphase(genovec, patch_01_set, patch_01_vals, patch_10_set, patch_10_vals, phasepresent_buf, phaseinfo, write_allele_ct, patch_01_ct, patch_10_ct, self._state_ptr)
             if reterr != kPglRetSuccess:
                 raise RuntimeError("append_partially_phased_batch() error " + str(reterr))
         return
 
 
     cdef append_dosages_batch_internal32(self, np.ndarray[np.float32_t,mode="c",ndim=2] floatarr_batch):
         cdef uint32_t batch_size = <uint32_t>floatarr_batch.shape[0]
@@ -1656,18 +1929,21 @@
             self.append_dosages_batch_internal64(floatarr_batch)
         else:
             raise RuntimeError("Invalid append_dosages_batch() dosage array element type (float32 or float64 expected).")
         return
 
 
     cpdef close(self):
+        cdef PglErr reterr = kPglRetSuccess
         if self._state_ptr:
             if SpgwGetVidx(self._state_ptr) != SpgwGetVariantCt(self._state_ptr):
                 raise RuntimeError("PgenWriter.close() called when number of written variants (" + str(SpgwGetVidx(self._state_ptr)) + ") unequal to initially declared value (" + str(SpgwGetVariantCt(self._state_ptr)) + ").")
-            SpgwFinish(self._state_ptr)
+            reterr = SpgwFinish(self._state_ptr)
+            if reterr != kPglRetSuccess:
+                raise RuntimeError("PgenWriter.close(): SpgwFinish() error " + str(reterr))
             if self._nonref_flags:
                 aligned_free(self._nonref_flags)
             PyMem_Free(self._state_ptr)
             self._state_ptr = NULL
         return
 
 
@@ -1676,14 +1952,16 @@
         return
 
 
     def __dealloc__(self):
         cdef PglErr reterr = kPglRetSuccess
         if self._state_ptr:
             if SpgwGetVidx(self._state_ptr) == SpgwGetVariantCt(self._state_ptr):
-                SpgwFinish(self._state_ptr)
+                reterr = SpgwFinish(self._state_ptr)
+                if reterr != kPglRetSuccess:
+                    raise RuntimeError("PgenWriter.__dealloc__(): SpgwFinish() error " + str(reterr))
             else:
                 CleanupSpgw(self._state_ptr, &reterr)
             if self._nonref_flags:
                 aligned_free(self._nonref_flags)
             PyMem_Free(self._state_ptr)
         return
```

### Comparing `Pgenlib-0.82.1/src/plink2/include/hedley.h` & `Pgenlib-0.83.0/src/plink2/include/hedley.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/pgenlib_misc.cc` & `Pgenlib-0.83.0/src/plink2/include/pgenlib_misc.cc`

 * *Files 1% similar despite different names*

```diff
@@ -2881,10 +2881,29 @@
     for (uint32_t uii = 0; uii != patch_10_ct; ++uii) {
       const uintptr_t sample_idx = BitIter1(patch_10_set, &sample_idx_base, &cur_bits);
       wide_codes_alias[sample_idx] = patch_10_vals_alias[uii];
     }
   }
 }
 
+uintptr_t PglComputeMaxAlleleCt(const uintptr_t* allele_idx_offsets, uint32_t variant_ct) {
+  if ((!allele_idx_offsets) || (allele_idx_offsets[variant_ct] == 2 * variant_ct)) {
+    return 2;
+  }
+  // todo: try vectorizing this
+  uintptr_t max_allele_ct = 2;
+  uintptr_t prev_offset = allele_idx_offsets[0];
+  const uintptr_t* shifted_offsets = &(allele_idx_offsets[1]);
+  for (uintptr_t uii = 0; uii != variant_ct; ++uii) {
+    const uintptr_t cur_offset = shifted_offsets[uii];
+    const uintptr_t cur_allele_ct = cur_offset - prev_offset;
+    if (cur_allele_ct > max_allele_ct) {
+      max_allele_ct = cur_allele_ct;
+    }
+    prev_offset = cur_offset;
+  }
+  return max_allele_ct;
+}
+
 #ifdef __cplusplus
 }  // namespace plink2
 #endif
```

### Comparing `Pgenlib-0.82.1/src/plink2/include/pgenlib_misc.h` & `Pgenlib-0.83.0/src/plink2/include/pgenlib_misc.h`

 * *Files 1% similar despite different names*

```diff
@@ -75,15 +75,15 @@
 //   much memory.
 
 #include "plink2_bits.h"
 
 // 10000 * major + 100 * minor + patch
 // Exception to CONSTI32, since we want the preprocessor to have access to this
 // value.  Named with all caps as a consequence.
-#define PGENLIB_INTERNAL_VERNUM 1906
+#define PGENLIB_INTERNAL_VERNUM 1908
 
 #ifdef __cplusplus
 namespace plink2 {
 #endif
 
 // other configuration-ish values needed by plink2_common subset
 typedef unsigned char AlleleCode;
@@ -609,14 +609,16 @@
 }
 
 extern const uint16_t kHcToAlleleCodes[1024];
 
 // Permits missing codes, does not remap.
 void PglMultiallelicSparseToDenseMiss(const PgenVariant* pgvp, uint32_t sample_ct, AlleleCode* __restrict wide_codes);
 
+uintptr_t PglComputeMaxAlleleCt(const uintptr_t* allele_idx_offsets, uint32_t variant_ct);
+
 // The actual format:
 // 1. 2 magic bytes 0x6c 0x1b.
 //
 // 2. Mode byte.
 //      0x01 = plink1 variant-major.
 //      0x02 = plink2 basic variant-major.  variant/sample counts in header,
 //             00 = hom ref, 01 = het, 10 = hom alt, 11 = missing.  (vrtype 0)
```

### Comparing `Pgenlib-0.82.1/src/plink2/include/pgenlib_read.cc` & `Pgenlib-0.83.0/src/plink2/include/pgenlib_read.cc`

 * *Files 0% similar despite different names*

```diff
@@ -6870,14 +6870,16 @@
     return ReadGenovecHphaseSubsetUnsafe(sample_include, sample_include_cumulative_popcounts, sample_ct, vidx, pgrp, nullptr, nullptr, pgvp->genovec, pgvp->phasepresent, pgvp->phaseinfo, &(pgvp->phasepresent_ct));
   }
   const unsigned char* fread_ptr;
   const unsigned char* fread_end;
   uintptr_t* all_hets = VrtypeHphase(vrtype)? pgrp->workspace_all_hets : nullptr;
   PglErr reterr = GetMultiallelicCodes(sample_include, sample_include_cumulative_popcounts, sample_ct, vidx, pgrp, all_hets? (&fread_ptr) : nullptr, all_hets? (&fread_end) : nullptr, all_hets, pgvp);
   if (reterr || (!all_hets)) {
+    // bugfix (17 Apr 2023): need to zero out phasepresent_ct in this case
+    pgvp->phasepresent_ct = 0;
     return reterr;
   }
   const uint32_t raw_sample_ct = pgrp->fi.raw_sample_ct;
   return ParseAux2Subset(fread_end, (sample_ct != raw_sample_ct)? sample_include : nullptr, all_hets, nullptr, raw_sample_ct, sample_ct, &fread_ptr, pgvp->phasepresent, pgvp->phaseinfo, &(pgvp->phasepresent_ct), pgrp->workspace_subset);
 }
 
 // ok for sample_include to be nullptr if not subsetting, though this is not
@@ -8726,14 +8728,15 @@
   }
   const unsigned char* fread_ptr;
   const unsigned char* fread_end;
   uintptr_t* all_hets = VrtypeHphase(vrtype)? pgrp->workspace_all_hets : nullptr;
   if (VrtypeMultiallelicHc(vrtype)) {
     PglErr reterr = GetMultiallelicCodes(sample_include, sample_include_cumulative_popcounts, sample_ct, vidx, pgrp, all_hets? (&fread_ptr) : nullptr, all_hets? (&fread_end) : nullptr, all_hets, pgvp);
     if (reterr || (!all_hets)) {
+      pgvp->phasepresent_ct = 0;
       return reterr;
     }
     if (!(vrtype & 0x60)) {
       const uint32_t raw_sample_ct = pgrp->fi.raw_sample_ct;
       return ParseAux2Subset(fread_end, (sample_ct != raw_sample_ct)? sample_include : nullptr, all_hets, nullptr, raw_sample_ct, sample_ct, &fread_ptr, pgvp->phasepresent, pgvp->phaseinfo, &(pgvp->phasepresent_ct), pgrp->workspace_subset);
     }
   } else {
```

### Comparing `Pgenlib-0.82.1/src/plink2/include/pgenlib_read.h` & `Pgenlib-0.83.0/src/plink2/include/pgenlib_read.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/pgenlib_write.cc` & `Pgenlib-0.83.0/src/plink2/include/pgenlib_write.cc`

 * *Files 0% similar despite different names*

```diff
@@ -249,15 +249,16 @@
       }
     }
   } else if (allele_ct_upper_bound) {
     max_alt_ct_p1 = allele_ct_upper_bound;
   }
   if (max_alt_ct_p1 > 2) {
     // see comments in middle of MpgwInitPhase1()
-    max_vrec_len += 2 + sizeof(AlleleCode) + GetAux1bAlleleEntryByteCt(max_alt_ct_p1, sample_ct - 1);
+    // bugfix (29 Mar 2023): forgot (sample_ct + 6) / 8 term
+    max_vrec_len += 2 + sizeof(AlleleCode) + (sample_ct + 6) / 8 + GetAux1bAlleleEntryByteCt(max_alt_ct_p1, sample_ct - 1);
     // try to permit uncompressed records to be larger than this, only error
     // out when trying to write a larger compressed record.
   }
   if (phase_dosage_gflags & kfPgenGlobalHardcallPhasePresent) {
     // phasepresent, phaseinfo
     max_vrec_len += 2 * DivUp(sample_ct, CHAR_BIT);
   }
@@ -2399,15 +2400,16 @@
       break;
     }
     if (unlikely(!fwrite_unlocked(copybuf, nbyte, 1, header_ff))) {
       return kPglRetWriteFail;
     }
   }
   if (unlikely(fclose_null(pgen_outfile_ptr) ||
-               unlink(*fname_buf_ptr))) {
+               unlink(*fname_buf_ptr) ||
+               fclose_null(header_ff_ptr))) {
     return kPglRetWriteFail;
   }
   free(*fname_buf_ptr);
   *fname_buf_ptr = nullptr;
   return kPglRetSuccess;
 }
```

### Comparing `Pgenlib-0.82.1/src/plink2/include/pgenlib_write.h` & `Pgenlib-0.83.0/src/plink2/include/pgenlib_write.h`

 * *Files 1% similar despite different names*

```diff
@@ -301,15 +301,15 @@
     return kPglRetVarRecordTooLarge;
   }
   return kPglRetSuccess;
 }
 
 
 // dosage_main[] has length dosage_ct, not sample_ct
-// ok for traling bits of dosage_present to not be zeroed out
+// ok for trailing bits of dosage_present to not be zeroed out
 BoolErr PwcAppendBiallelicGenovecDosage16(const uintptr_t* __restrict genovec, const uintptr_t* __restrict dosage_present, const uint16_t* dosage_main, uint32_t dosage_ct, PgenWriterCommon* pwcp);
 
 HEADER_INLINE PglErr SpgwAppendBiallelicGenovecDosage16(const uintptr_t* __restrict genovec, const uintptr_t* __restrict dosage_present, const uint16_t* dosage_main, uint32_t dosage_ct, STPgenWriter* spgwp) {
   if (unlikely(SpgwFlush(spgwp))) {
     return kPglRetWriteFail;
   }
   PgenWriterCommon* pwcp = &GET_PRIVATE(*spgwp, pwc);
```

### Comparing `Pgenlib-0.82.1/src/plink2/include/plink2_base.cc` & `Pgenlib-0.83.0/src/plink2/include/plink2_base.cc`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/plink2_base.h` & `Pgenlib-0.83.0/src/plink2/include/plink2_base.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/plink2_bits.cc` & `Pgenlib-0.83.0/src/plink2/include/plink2_bits.cc`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/plink2_bits.h` & `Pgenlib-0.83.0/src/plink2/include/plink2_bits.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-align.h` & `Pgenlib-0.83.0/src/plink2/include/simde-align.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-arch.h` & `Pgenlib-0.83.0/src/plink2/include/simde-arch.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-common.h` & `Pgenlib-0.83.0/src/plink2/include/simde-common.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-constify.h` & `Pgenlib-0.83.0/src/plink2/include/simde-constify.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-detect-clang.h` & `Pgenlib-0.83.0/src/plink2/include/simde-detect-clang.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-diagnostic.h` & `Pgenlib-0.83.0/src/plink2/include/simde-diagnostic.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-features.h` & `Pgenlib-0.83.0/src/plink2/include/simde-features.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/simde-math.h` & `Pgenlib-0.83.0/src/plink2/include/simde-math.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/x86/mmx.h` & `Pgenlib-0.83.0/src/plink2/include/x86/mmx.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/x86/sse.h` & `Pgenlib-0.83.0/src/plink2/include/x86/sse.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/include/x86/sse2.h` & `Pgenlib-0.83.0/src/plink2/include/x86/sse2.h`

 * *Files identical despite different names*

### Comparing `Pgenlib-0.82.1/src/plink2/pgenlib_ffi_support.cc` & `Pgenlib-0.83.0/src/plink2/pgenlib_ffi_support.cc`

 * *Files 15% similar despite different names*

```diff
@@ -68,17 +68,17 @@
 }
 
 // missing = -9
 const double kGenoDoublePairs[32] ALIGNV16 = PAIR_TABLE16(0.0, 1.0, 2.0, -9.0);
 
 const uint64_t kGenoToIntcodeDPairs[32] ALIGNV16 = PAIR_TABLE16(0, 0x100000000LLU, 0x100000001LLU, 0xfffffff7fffffff7LLU);
 
-void GenoarrPhasedToAlleleCodes(const uint64_t* genoarr_to_intcode_pair_table, const uintptr_t* genoarr, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, uint32_t sample_ct, uint32_t phasepresent_ct, unsigned char* phasebytes, int32_t* allele_codes) {
+void GenoarrPhasedToAlleleCodes(const uint64_t* genoarr_to_intcode_dpair_table, const uintptr_t* genoarr, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, uint32_t sample_ct, uint32_t phasepresent_ct, unsigned char* phasebytes, int32_t* allele_codes) {
   // phasebytes can be nullptr, phasepresent cannot
-  GenoarrToAlleleCodes(genoarr_to_intcode_pair_table, genoarr, sample_ct, allele_codes);
+  GenoarrToAlleleCodes(genoarr_to_intcode_dpair_table, genoarr, sample_ct, allele_codes);
   uint64_t* allele_codes_alias64 = R_CAST(uint64_t*, allele_codes);
   uintptr_t sample_uidx_base = 0;
   uintptr_t cur_bits = phasepresent[0];
   if (!phasebytes) {
     for (uint32_t phased_idx = 0; phased_idx != phasepresent_ct; ++phased_idx) {
       const uintptr_t sample_uidx = BitIter1(phasepresent, &sample_uidx_base, &cur_bits);
       if (IsSet(phaseinfo, sample_uidx)) {
@@ -108,14 +108,103 @@
     phasebytes[sample_uidx] = 1;
     if (IsSet(phaseinfo, sample_uidx)) {
       allele_codes_alias64[sample_uidx] = 1;
     }
   }
 }
 
+void GenoarrMPToAlleleCodes(const uint64_t* geno_to_intcode_dpair_table, const PgenVariant* pgv, uint32_t sample_ct, unsigned char* phasebytes, int32_t* allele_codes) {
+  // phasebytes can be nullptr, phasepresent cannot
+  const uintptr_t* genoarr = pgv->genovec;
+  const uintptr_t* phasepresent = pgv->phasepresent;
+  const uintptr_t* phaseinfo = pgv->phaseinfo;
+  const uint32_t phasepresent_ct = pgv->phasepresent_ct;
+  const uint32_t patch_01_ct = pgv->patch_01_ct;
+  const uint32_t patch_10_ct = pgv->patch_10_ct;
+  if ((!patch_01_ct) && (!patch_10_ct)) {
+    GenoarrPhasedToAlleleCodes(geno_to_intcode_dpair_table, genoarr, phasepresent, phaseinfo, sample_ct, phasepresent_ct, phasebytes, allele_codes);
+    return;
+  }
+  GenoarrToAlleleCodes(geno_to_intcode_dpair_table, genoarr, sample_ct, allele_codes);
+  // See e.g. PglMultiallelicSparseToDense().
+  if (patch_01_ct) {
+    const uintptr_t* patch_01_set = pgv->patch_01_set;
+    const AlleleCode* patch_01_vals = pgv->patch_01_vals;
+    uintptr_t sample_idx_base = 0;
+    uintptr_t cur_bits = patch_01_set[0];
+    int32_t* allele_codes1 = &(allele_codes[1]);
+    for (uint32_t uii = 0; uii != patch_01_ct; ++uii) {
+      const uintptr_t sample_idx = BitIter1(patch_01_set, &sample_idx_base, &cur_bits);
+      allele_codes1[2 * sample_idx] = patch_01_vals[uii];
+    }
+  }
+  if (phasebytes) {
+    // Initialize 0/2 phasebytes before processing patch_10.
+    const uint32_t word_ct_m1 = (sample_ct - 1) / kBytesPerWord;
+    const Quarterword* read_alias = R_CAST(const Quarterword*, genoarr);
+    uintptr_t* write_walias = R_CAST(uintptr_t*, phasebytes);
+    for (uint32_t widx = 0; ; ++widx) {
+      uintptr_t qw = Unpack0303(read_alias[widx]);
+      qw = (~qw) & kMask0101;
+      if (widx == word_ct_m1) {
+        SubwordStore(qw, ModNz(sample_ct, kBytesPerWord), &(write_walias[widx]));
+        break;
+      }
+      write_walias[widx] = qw;
+    }
+  }
+  if (patch_10_ct) {
+    const uintptr_t* patch_10_set = pgv->patch_10_set;
+    const AlleleCode* patch_10_vals = pgv->patch_10_vals;
+    uintptr_t sample_idx_base = 0;
+    uintptr_t cur_bits = patch_10_set[0];
+    if (!phasebytes) {
+      for (uint32_t uii = 0; uii != patch_10_ct; ++uii) {
+        const uintptr_t sample_idx = BitIter1(patch_10_set, &sample_idx_base, &cur_bits);
+        allele_codes[2 * sample_idx] = patch_10_vals[2 * uii];
+        allele_codes[2 * sample_idx + 1] = patch_10_vals[2 * uii + 1];
+      }
+    } else {
+      for (uint32_t uii = 0; uii != patch_10_ct; ++uii) {
+        const uintptr_t sample_idx = BitIter1(patch_10_set, &sample_idx_base, &cur_bits);
+        const AlleleCode ac0 = patch_10_vals[2 * uii];
+        const AlleleCode ac1 = patch_10_vals[2 * uii + 1];
+        allele_codes[2 * sample_idx] = ac0;
+        allele_codes[2 * sample_idx + 1] = ac1;
+        if (ac0 != ac1) {
+          phasebytes[sample_idx] = 0;
+          // When phasepresent bit is set, we'll fix this up later.
+        }
+      }
+    }
+  }
+  uintptr_t sample_uidx_base = 0;
+  uintptr_t cur_bits = phasepresent[0];
+  if (!phasebytes) {
+    for (uint32_t phased_idx = 0; phased_idx != phasepresent_ct; ++phased_idx) {
+      const uintptr_t sample_uidx = BitIter1(phasepresent, &sample_uidx_base, &cur_bits);
+      if (IsSet(phaseinfo, sample_uidx)) {
+        const int32_t tmp_code = allele_codes[2 * sample_uidx];
+        allele_codes[2 * sample_uidx] = allele_codes[2 * sample_uidx + 1];
+        allele_codes[2 * sample_uidx + 1] = tmp_code;
+      }
+    }
+    return;
+  }
+  for (uint32_t phased_idx = 0; phased_idx != phasepresent_ct; ++phased_idx) {
+    const uintptr_t sample_uidx = BitIter1(phasepresent, &sample_uidx_base, &cur_bits);
+    phasebytes[sample_uidx] = 1;
+    if (IsSet(phaseinfo, sample_uidx)) {
+      const int32_t tmp_code = allele_codes[2 * sample_uidx];
+      allele_codes[2 * sample_uidx] = allele_codes[2 * sample_uidx + 1];
+      allele_codes[2 * sample_uidx + 1] = tmp_code;
+    }
+  }
+}
+
 // missing = -9
 // may want a double-lookup function for this
 static const int32_t kGenoToHap0Code[6] = {0, 0, 1, -9, 0, 1};
 static const int32_t kGenoToHap1Code[6] = {0, 1, 1, -9, 0, 0};
 
 // todo: write version of this which fills phasebytes
 void GenoarrPhasedToHapCodes(const uintptr_t* genoarr, const uintptr_t* phaseinfo, uint32_t variant_batch_size, int32_t* hap0_codes_iter, int32_t* hap1_codes_iter) {
@@ -399,14 +488,15 @@
 
 void AlleleCodesToGenoarrUnsafe(const int32_t* allele_codes, const unsigned char* phasepresent_bytes, uint32_t sample_ct, uintptr_t* genoarr, uintptr_t* phasepresent, uintptr_t* phaseinfo) {
   // - If phasepresent_bytes is nullptr, phasepresent is not updated.  In this
   //   case, phaseinfo is updated iff it's not nullptr.  It's okay for both
   //   phasepresent and phaseinfo to be nullptr here.
   // - Otherwise, phasepresent and phaseinfo are always updated; neither can be
   //   nullptr.
+  // - Trailing bits of phasepresent/phaseinfo may not be zeroed out.
   const uint32_t word_ct_m1 = (sample_ct - 1) / kBitsPerWordD2;
   uint32_t subgroup_len = kBitsPerWordD2;
   const uint32_t* read_alias = R_CAST(const uint32_t*, allele_codes);
   Halfword* phaseinfo_alias = R_CAST(Halfword*, phaseinfo);
   if (!phasepresent_bytes) {
     for (uint32_t widx = 0; ; ++widx) {
       if (widx >= word_ct_m1) {
@@ -486,14 +576,132 @@
     phasepresent_bytes_iter = &(phasepresent_bytes_iter[subgroup_len]);
     phasepresent_alias[widx] = phasepresent_write_hw;
     phaseinfo_alias[widx] = phaseinfo_write_hw;
     genoarr[widx] = geno_write_word;
   }
 }
 
+// Does not clear trailing bits of genovec, phasepresent, or phaseinfo.
+// Returns max(2, 1 + max allele code) if allele_codes is valid, -1 if invalid.
+int32_t ConvertMultiAlleleCodesUnsafe(const int32_t* allele_codes, const unsigned char* phasepresent_bytes, uint32_t sample_ct, uintptr_t* genoarr, uintptr_t* patch_01_set, AlleleCode* patch_01_vals, uintptr_t* patch_10_set, AlleleCode* patch_10_vals, uint32_t* patch_01_ctp, uint32_t* patch_10_ctp, uintptr_t* phasepresent, uintptr_t* phaseinfo) {
+  const uint32_t sample_ctl = DivUp(sample_ct, kBitsPerWord);
+  const uint32_t word_ct_m1 = (sample_ct - 1) / kBitsPerWordD2;
+  uint32_t subgroup_len = kBitsPerWordD2;
+  const uint32_t* read_alias = R_CAST(const uint32_t*, allele_codes);
+  if (phasepresent_bytes) {
+    BytesToBitsUnsafe(phasepresent_bytes, sample_ct, phasepresent);
+  }
+  Halfword* phasepresent_alias = R_CAST(Halfword*, phasepresent);
+  Halfword* phaseinfo_alias = R_CAST(Halfword*, phaseinfo);
+  // todo: try scanning allele_codes for its maximum value upfront, instead of
+  // checking in inner loops; then mirror PglMultiallelicDenseToSparse() in
+  // valid multiallelic case, and call AlleleCodesToGenoarrUnsafe() otherwise.
+  ZeroWArr(sample_ctl, patch_01_set);
+  ZeroWArr(sample_ctl, patch_10_set);
+  uint32_t max_allele_code = 1;
+  Halfword* patch_01_set_alias = R_CAST(Halfword*, patch_01_set);
+  Halfword* patch_10_set_alias = R_CAST(Halfword*, patch_10_set);
+  AlleleCode* patch_01_iter = patch_01_vals;
+  AlleleCode* patch_10_iter = patch_10_vals;
+  for (uint32_t widx = 0; ; ++widx) {
+    if (widx >= word_ct_m1) {
+      if (widx > word_ct_m1) {
+        if (max_allele_code >= kPglMaxAlleleCt) {
+          return -1;
+        }
+        *patch_01_ctp = patch_01_iter - patch_01_vals;
+        *patch_10_ctp = (patch_10_iter - patch_10_vals) >> 1;
+        return S_CAST(int32_t, max_allele_code + 1);
+      }
+      subgroup_len = ModNz(sample_ct, kBitsPerWordD2);
+    }
+    uintptr_t geno_write_word = 0;
+    Halfword phaseinfo_write_hw = 0;
+    Halfword het_2_hw = 0;
+    for (uint32_t uii = 0; uii != subgroup_len; ++uii) {
+      // 0,0 -> 0
+      // 0,x or x,0 -> 1
+      // x,y -> 2
+      // -9,-9 -> 3
+      const uint32_t first_code = *read_alias++;
+      const uint32_t second_code = *read_alias++;
+      uintptr_t cur_geno = 0;
+      if (first_code == 0) {
+        if (second_code != 0) {
+          cur_geno = 1;
+          if (second_code > 1) {
+            if (second_code > max_allele_code) {
+              max_allele_code = second_code;
+            }
+            patch_01_set_alias[widx] |= 1U << uii;
+            // If second_code is actually out-of-range, harmlessly truncate
+            // here, and error out before function return.
+            // (this code is correct without the static-cast, but may as well
+            // be explicit about where truncation could happen.)
+            *patch_01_iter++ = S_CAST(AlleleCode, second_code);
+          }
+        }
+      } else if (first_code == 0xfffffff7U) {
+        if (second_code != 0xfffffff7U) {
+          return -1;
+        }
+        cur_geno = 3;
+      } else {
+        // first_code >= 1
+        if (second_code == 0) {
+          cur_geno = 1;
+          phaseinfo_write_hw |= 1U << uii;
+          if (first_code > 1) {
+            if (first_code > max_allele_code) {
+              max_allele_code = first_code;
+            }
+            patch_01_set_alias[widx] |= 1U << uii;
+            *patch_01_iter++ = S_CAST(AlleleCode, first_code);
+          }
+        } else {
+          cur_geno = 2;
+          if (first_code <= second_code) {
+            if (second_code > 1) {
+              if (second_code > max_allele_code) {
+                max_allele_code = second_code;
+              }
+              patch_10_set_alias[widx] |= 1U << uii;
+              *patch_10_iter++ = S_CAST(AlleleCode, first_code);
+              *patch_10_iter++ = S_CAST(AlleleCode, second_code);
+              if (first_code != second_code) {
+                het_2_hw |= 1U << uii;
+              }
+            }
+          } else {
+            // first_code > second_code
+            if (first_code > max_allele_code) {
+              max_allele_code = first_code;
+            }
+            phaseinfo_write_hw |= 1U << uii;
+            patch_10_set_alias[widx] |= 1U << uii;
+            het_2_hw |= 1U << uii;
+            *patch_10_iter++ = S_CAST(AlleleCode, second_code);
+            *patch_10_iter++ = S_CAST(AlleleCode, first_code);
+          }
+        }
+      }
+      geno_write_word |= (cur_geno << (uii * 2));
+    }
+    genoarr[widx] = geno_write_word;
+    if (phasepresent_bytes) {
+      const uintptr_t het_1_word = geno_write_word & (~(geno_write_word >> 1)) & kMask5555;
+      Halfword het_hw = het_2_hw | PackWordToHalfword(het_1_word);
+      phasepresent_alias[widx] &= het_hw;
+    }
+    if (phaseinfo_alias) {
+      phaseinfo_alias[widx] = phaseinfo_write_hw;
+    }
+  }
+}
+
 static inline uint32_t BiallelicDosage16Halfdist(uint32_t dosage_int) {
   const uint32_t dosage_int_rem = dosage_int & 16383;
   return abs_i32(S_CAST(int32_t, dosage_int_rem) - 8192);
 }
 
 void FloatsToDosage16(const float* floatarr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr) {
   const uint32_t word_ct_m1 = (sample_ct - 1) / kBitsPerWordD2;
```

### Comparing `Pgenlib-0.82.1/src/plink2/pgenlib_ffi_support.h` & `Pgenlib-0.83.0/src/plink2/pgenlib_ffi_support.h`

 * *Files 18% similar despite different names*

```diff
@@ -55,14 +55,20 @@
 // phasepresent cannot be nullptr
 void GenoarrPhasedToAlleleCodes(const uint64_t* geno_to_intcode_dpair_table, const uintptr_t* genoarr, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, uint32_t sample_ct, uint32_t phasepresent_ct, unsigned char* phasebytes, int32_t* allele_codes);
 
 HEADER_INLINE void GenoarrPhasedToAlleleCodesMinus9(const uintptr_t* genoarr, const uintptr_t* phasepresent, const uintptr_t* phaseinfo, uint32_t sample_ct, uint32_t phasepresent_ct, unsigned char* phasebytes, int32_t* allele_codes) {
   GenoarrPhasedToAlleleCodes(kGenoToIntcodeDPairs, genoarr, phasepresent, phaseinfo, sample_ct, phasepresent_ct, phasebytes, allele_codes);
 }
 
+void GenoarrMPToAlleleCodes(const uint64_t* geno_to_intcode_dpair_table, const PgenVariant* pgv, uint32_t sample_ct, unsigned char* phasebytes, int32_t* allele_codes);
+
+HEADER_INLINE void GenoarrMPToAlleleCodesMinus9(const PgenVariant* pgv, uint32_t sample_ct, unsigned char* phasebytes, int32_t* allele_codes) {
+  GenoarrMPToAlleleCodes(kGenoToIntcodeDPairs, pgv, sample_ct, phasebytes, allele_codes);
+}
+
 // assumes transposed genoarr, phaseinfo
 void GenoarrPhasedToHapCodes(const uintptr_t* genoarr, const uintptr_t* phaseinfo, uint32_t variant_batch_size, int32_t* hap0_codes_iter, int32_t* hap1_codes_iter);
 
 void Dosage16ToFloatsMinus9(const uintptr_t* genoarr, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t sample_ct, uint32_t dosage_ct, float* geno_float);
 
 void Dosage16ToDoubles(const double* geno_double_pair_table, const uintptr_t* genoarr, const uintptr_t* dosage_present, const uint16_t* dosage_main, uint32_t sample_ct, uint32_t dosage_ct, double* geno_double);
 
@@ -80,21 +86,28 @@
 void BytesToBitsUnsafe(const uint8_t* boolbytes, uint32_t sample_ct, uintptr_t* bitarr);
 
 // Bottom 2 bits are extracted from every byte.  Conveniently, -9 and 3 are
 // treated identically.
 // Does not zero out trailing bits of genoarr.
 void BytesToGenoarrUnsafe(const int8_t* genobytes, uint32_t sample_ct, uintptr_t* genoarr);
 
+// - Assumes biallelic variant, does not validate that.
+// - Low bit of each element of phasepresent_bytes is significant.
 // - If phasepresent_bytes is nullptr, phasepresent is not updated.  In this
 //   case, phaseinfo is updated iff it's not nullptr.  It's okay for both
 //   phasepresent and phaseinfo to be nullptr here.
 // - Otherwise, phasepresent and phaseinfo are always updated; neither can be
 //   nullptr.
+// - Trailing bits of phasepresent/phaseinfo may not be zeroed out.
 void AlleleCodesToGenoarrUnsafe(const int32_t* allele_codes, const unsigned char* phasepresent_bytes, uint32_t sample_ct, uintptr_t* genoarr, uintptr_t* phasepresent, uintptr_t* phaseinfo);
 
+// Does not clear trailing bits of genovec, phasepresent, or phaseinfo.
+// Returns min(2, 1 + max allele code) if allele_codes is valid, -1 if invalid.
+int32_t ConvertMultiAlleleCodesUnsafe(const int32_t* allele_codes, const unsigned char* phasepresent_bytes, uint32_t sample_ct, uintptr_t* genoarr, uintptr_t* patch_01_set, AlleleCode* patch_01_vals, uintptr_t* patch_10_set, AlleleCode* patch_10_vals, uint32_t* patch_01_ctp, uint32_t* patch_10_ctp, uintptr_t* phasepresent, uintptr_t* phaseinfo);
+
 void FloatsToDosage16(const float* floatarr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr);
 
 void DoublesToDosage16(const double* doublearr, uint32_t sample_ct, uint32_t hard_call_halfdist, uintptr_t* genoarr, uintptr_t* dosage_present, uint16_t* dosage_main, uint32_t* dosage_ct_ptr);
 
 #ifdef __cplusplus
 }  // namespace plink2
 #endif
```

